// @microsoft.azure/openapi-validator-rulesets/dist/spectral/functions/Extensions/long-running-response-status-code.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.longRunningResponseStatusCodeDataPlane = void 0;
const longRunningResponseStatusCode = (methodOp, _opts, ctx, validResponseCodesList) => {
    var _a, _b, _c, _d;
    if (methodOp === null || typeof methodOp !== "object") {
        return [];
    }
    const path = ctx.path || [];
    const errors = [];
    const method = Object.keys(methodOp)[0];
    if (!["delete", "put", "patch", "post"].includes(method)) {
        return [];
    }
    const operationId = ((_a = methodOp === null || methodOp === void 0 ? void 0 : methodOp[method]) === null || _a === void 0 ? void 0 : _a.operationId) || "";
    if (!((_b = methodOp === null || methodOp === void 0 ? void 0 : methodOp[method]) === null || _b === void 0 ? void 0 : _b["x-ms-long-running-operation"])) {
        return [];
    }
    if ((_c = methodOp === null || methodOp === void 0 ? void 0 : methodOp[method]) === null || _c === void 0 ? void 0 : _c.responses) {
        const responseCodes = Object.keys((_d = methodOp === null || methodOp === void 0 ? void 0 : methodOp[method]) === null || _d === void 0 ? void 0 : _d.responses);
        const validResponseCodes = validResponseCodesList[method];
        const validResponseCodeString = validResponseCodes.join(" or ");
        const withTerminalCode = validResponseCodes.some((code) => responseCodes.includes(code));
        if (!withTerminalCode) {
            errors.push({
                message: `A '${method}' operation '${operationId}' with x-ms-long-running-operation extension must have a valid terminal success status code ${validResponseCodeString}.`,
                path: [...path, method],
            });
        }
    }
    return errors;
};
const longRunningResponseStatusCodeDataPlane = (methodOp, _opts, ctx) => {
    const validResponseCodesList = {
        delete: ["200", "204", "202"],
        post: ["200", "201", "202", "204"],
        put: ["200", "201", "202"],
        patch: ["200", "201", "202"],
    };
    return longRunningResponseStatusCode(methodOp, _opts, ctx, validResponseCodesList);
};
exports.longRunningResponseStatusCodeDataPlane = longRunningResponseStatusCodeDataPlane;
