// @azure-tools/typespec-azure-core/dist/src/decorators/private/default-final-state-via.js
import { ignoreDiagnostics } from "@typespec/compiler";
import { getHttpOperation } from "@typespec/http";
import { reportDiagnostic } from "../../lib.js";
import { FinalStateValue, setFinalStateOverride, validateFinalState, } from "../../state/final-state.js";
export const $defaultFinalStateVia = (context, target, states) => {
    const { program } = context;
    const finalStateValues = [];
    for (const finalState of states) {
        switch (finalState?.toLowerCase()) {
            case "operation-location":
                finalStateValues.push(FinalStateValue.operationLocation);
                break;
            case "location":
                finalStateValues.push(FinalStateValue.location);
                break;
            case "azure-async-operation":
                finalStateValues.push(FinalStateValue.azureAsyncOperation);
                break;
            default:
                reportDiagnostic(program, {
                    code: "invalid-final-state",
                    target: target,
                    messageId: "badValue",
                    format: { finalStateValue: finalState },
                });
                return;
        }
    }
    const storedValue = validateFinalStates(program, target, finalStateValues);
    if (storedValue !== undefined) {
        setFinalStateOverride(program, target, storedValue);
    }
};
function validateFinalStates(program, operation, finalStates) {
    const httpOp = ignoreDiagnostics(getHttpOperation(program, operation));
    for (const state of finalStates) {
        if (validateFinalState(program, httpOp, state))
            return state;
    }
    return undefined;
}
