// @azure-tools/typespec-azure-core/dist/src/decorators/private/ensure-verb.js
import { useStateMap } from "@typespec/compiler/utils";
import { getHttpOperation } from "@typespec/http";
import { AzureCoreStateKeys, reportDiagnostic } from "../../lib.js";
const [_, setEnsureVerb, getEnsureVerbStateMap] = useStateMap(AzureCoreStateKeys.ensureVerb);
export const $ensureVerb = (context, entity, templateName, verb) => {
    setEnsureVerb(context.program, entity, [templateName, verb]);
    context.program.stateMap(AzureCoreStateKeys.ensureVerb).set(entity, [templateName, verb]);
};
// TODO: migrate this to decorator post validation when/if become available
export function checkEnsureVerb(program) {
    const opMap = getEnsureVerbStateMap(program);
    for (const [operation, [templateName, requiredVerb]] of opMap.entries()) {
        const verb = getHttpOperation(program, operation)[0].verb.toString().toLowerCase();
        const reqVerb = requiredVerb.toLowerCase();
        if (verb !== reqVerb) {
            reportDiagnostic(program, {
                code: "verb-conflict",
                target: operation,
                format: {
                    templateName: templateName,
                    verb: verb.toUpperCase(),
                    requiredVerb: reqVerb.toUpperCase(),
                },
            });
        }
    }
}
