// oav/dist/lib/exampleQualityValidator/exampleQualityValidator.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var ExampleQualityValidator_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExampleQualityValidator = void 0;
const jsonpath_plus_1 = require("jsonpath-plus");
const inversify_1 = require("inversify");
const jsonLoader_1 = require("../swagger/jsonLoader");
const swaggerLoader_1 = require("../swagger/swaggerLoader");
const context_1 = require("../transform/context");
const xmsPathsTransformer_1 = require("../transform/xmsPathsTransformer");
const resolveNestedDefinitionTransformer_1 = require("../transform/resolveNestedDefinitionTransformer");
const referenceFieldsTransformer_1 = require("../transform/referenceFieldsTransformer");
const discriminatorTransformer_1 = require("../transform/discriminatorTransformer");
const allOfTransformer_1 = require("../transform/allOfTransformer");
const noAdditionalPropertiesTransformer_1 = require("../transform/noAdditionalPropertiesTransformer");
const nullableTransformer_1 = require("../transform/nullableTransformer");
const pureObjectTransformer_1 = require("../transform/pureObjectTransformer");
const ajvSchemaValidator_1 = require("../swaggerValidator/ajvSchemaValidator");
const diffUtils_1 = require("../apiScenario/diffUtils");
const bodyTransformer_1 = require("../apiScenario/bodyTransformer");
const constants_1 = require("../util/constants");
const inversifyUtils_1 = require("./../inversifyUtils");
const loader_1 = require("./../swagger/loader");
const traverseSwagger_1 = require("./../transform/traverseSwagger");
const transformer_1 = require("./../transform/transformer");
// ARM RPC guide: https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#provisioningstate-property
// https://docs.microsoft.com/en-us/javascript/api/@azure/arm-databricks/provisioningstate?view=azure-node-latest
const incorrectProvisioningState = async (example, _operation, _jsonLoader, exampleValidationContext) => {
    const ret = [];
    if ((example === null || example === void 0 ? void 0 : example.responses["200"]) !== undefined) {
        const provisioningStatePath = "$.responses.200..[?(@.provisioningState)]";
        const provisioningStates = (0, jsonpath_plus_1.JSONPath)({
            path: provisioningStatePath,
            json: example,
            resultType: "all",
        });
        const terminalStatues = ["succeeded", "failed", "canceled", "ready", "created", "deleted"];
        for (const it of provisioningStates) {
            if (!terminalStatues.includes(it.value.provisioningState.toLowerCase())) {
                ret.push({
                    id: constants_1.ErrorCodes.IncorrectProvisioningState.id,
                    severity: "Error",
                    code: constants_1.ErrorCodes.IncorrectProvisioningState.name,
                    jsonPath: `${it.pointer}/provisioningState`,
                    message: "The resource's provisioning state should be terminal status in http 200 response.",
                    exampleName: exampleValidationContext.exampleName,
                    exampleFilePath: exampleValidationContext.exampleFilePath,
                });
            }
        }
    }
    return ret;
};
const roundtripInconsistentProperty = async (example, operation, jsonLoader, exampleValidationContext) => {
    var _a, _b, _c;
    const ret = [];
    if (operation._method === "put") {
        const reqSchema = (_a = operation.parameters) === null || _a === void 0 ? void 0 : _a.filter((it) => (it === null || it === void 0 ? void 0 : it.in) === "body")[0];
        for (const statusCode of Object.keys(example.responses)) {
            if (operation.responses[statusCode] === undefined) {
                continue;
            }
            const respSchema = operation.responses[statusCode].schema;
            if (reqSchema && ((_b = reqSchema.schema) === null || _b === void 0 ? void 0 : _b.$ref) === (respSchema === null || respSchema === void 0 ? void 0 : respSchema.$ref)) {
                const schema = jsonLoader.resolveRefObj(respSchema);
                const responseObj = await ((_c = exampleValidationContext.bodyTransform) === null || _c === void 0 ? void 0 : _c.resourceToRequest(example.responses[statusCode].body, schema));
                const requestObj = example.parameters.parameters;
                if (!!requestObj && !!responseObj) {
                    const delta = (0, diffUtils_1.getJsonPatchDiff)(requestObj, responseObj, {
                        includeOldValue: true,
                        minimizeDiff: false,
                    });
                    // filter replace operation.
                    delta
                        .filter((it) => it.replace !== undefined)
                        .map((it) => {
                        it.replace = `/${statusCode}/body${it.replace}`;
                        return it;
                    })
                        .forEach((it) => ret.push({
                        id: constants_1.ErrorCodes.RoundtripInconsistentProperty.id,
                        severity: "Error",
                        code: constants_1.ErrorCodes.RoundtripInconsistentProperty.name,
                        jsonPath: it.replace,
                        detail: JSON.stringify(it),
                        message: `The property's value in the response is different from what was set in the request. Path: ${it.replace}. Request: ${it.oldValue}. Response: ${it.value}`,
                        exampleName: exampleValidationContext.exampleName,
                        exampleFilePath: exampleValidationContext.exampleFilePath,
                    }));
                }
            }
        }
    }
    return ret;
};
/*
const recommendUsingBooleanType: ExampleValidationFunc = async (
  example: SwaggerExample,
  _operation: Operation,
  _jsonLoader: JsonLoader,
  exampleValidationContext: exampleValidationContext
) => {
  const ret: ExampleValidationRule[] = [];
  const allElementPath = "$..*";
  const allElements = JSONPath({
    path: allElementPath,
    json: example,
    resultType: "all",
  });
  const stringBooleans = ["false", "true"];
  for (const it of allElements) {
    if (typeof it.value === "string" && stringBooleans.includes(it.value.toLowerCase())) {
      ret.push({
        id: ErrorCodes.RecommendUsingBooleanType.id,
        severity: Severity.Warning,
        code: ErrorCodes.RecommendUsingBooleanType.name,
        path: `${it.pointer}`,
        message:
          "If the property only return two string value 'true' or 'false', recommend using boolean type.",
        exampleName: exampleValidationContext.exampleName,
        exampleFilePath: exampleValidationContext.exampleFilePath,
      });
    }
  }
  //TODO: find schema by pointer. and check whether it type is string.
  return ret;
};
*/
let ExampleQualityValidator = ExampleQualityValidator_1 = class ExampleQualityValidator {
    // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
    constructor(opts, jsonLoader, swaggerLoader, bodyTransformer) {
        this.opts = opts;
        this.jsonLoader = jsonLoader;
        this.swaggerLoader = swaggerLoader;
        this.bodyTransformer = bodyTransformer;
        this.initialized = false;
        this.swaggerSpecs = [];
        this.validationFuncs = [incorrectProvisioningState, roundtripInconsistentProperty];
        this.schemaValidator = new ajvSchemaValidator_1.AjvSchemaValidator(this.jsonLoader);
        this.transformContext = (0, context_1.getTransformContext)(this.jsonLoader, this.schemaValidator, [
            xmsPathsTransformer_1.xmsPathsTransformer,
            resolveNestedDefinitionTransformer_1.resolveNestedDefinitionTransformer,
            referenceFieldsTransformer_1.referenceFieldsTransformer,
            discriminatorTransformer_1.discriminatorTransformer,
            allOfTransformer_1.allOfTransformer,
            noAdditionalPropertiesTransformer_1.noAdditionalPropertiesTransformer,
            nullableTransformer_1.nullableTransformer,
            pureObjectTransformer_1.pureObjectTransformer,
        ]);
        this.operationMapping = {};
    }
    static create(opts) {
        (0, loader_1.setDefaultOpts)(opts, {
            eraseXmsExamples: false,
            eraseDescription: false,
        });
        return (0, inversifyUtils_1.inversifyGetInstance)(ExampleQualityValidator_1, opts);
    }
    async initialize() {
        var _a;
        if (this.initialized) {
            throw new Error("Already initialized");
        }
        for (const swaggerFilePath of (_a = this.opts.swaggerFilePaths) !== null && _a !== void 0 ? _a : []) {
            const swaggerSpec = await this.swaggerLoader.load(swaggerFilePath);
            this.swaggerSpecs.push(swaggerSpec);
            (0, transformer_1.applySpecTransformers)(swaggerSpec, this.transformContext);
            await (0, traverseSwagger_1.traverseSwaggerAsync)(swaggerSpec, {
                onOperation: async (operation) => {
                    this.operationMapping[operation.operationId] = operation;
                },
            });
        }
        (0, transformer_1.applyGlobalTransformers)(this.transformContext);
    }
    async validateExternalExamples(examples) {
        var _a;
        if (!this.initialized) {
            await this.initialize();
        }
        let res = [];
        for (const example of examples) {
            const operationId = example.operationId;
            const operation = this.operationMapping[operationId];
            if (operation === undefined) {
                throw new Error(`Cannot find operation for ${example.exampleFilePath}`);
            }
            const ctx = {
                exampleName: example.exampleFilePath,
                exampleFilePath: example.exampleFilePath,
                bodyTransform: this.bodyTransformer,
            };
            const exampleObj = (_a = example.example) !== null && _a !== void 0 ? _a : (await this.jsonLoader.load(example.exampleFilePath));
            for (const func of this.validationFuncs) {
                const ruleRes = await func(exampleObj, operation, this.jsonLoader, ctx);
                res = res.concat(ruleRes);
            }
        }
        return res;
    }
    /**
     * Validate swagger example quality.
     * @param filteredOperationIds If filtered is undefined, validate the whole x-ms-examples in swagger.
     * @returns
     */
    async validateSwaggerExamples(filteredOperationIds = undefined) {
        if (!this.initialized) {
            await this.initialize();
        }
        let res = [];
        const onOperation = async (operation, _httpPath, _method) => {
            var _a;
            if (filteredOperationIds === undefined ||
                filteredOperationIds.includes(operation.operationId)) {
                const xMsExamples = (_a = operation["x-ms-examples"]) !== null && _a !== void 0 ? _a : {};
                for (const exampleName of Object.keys(xMsExamples)) {
                    const example = xMsExamples[exampleName];
                    if (typeof example.$ref !== "string") {
                        throw new Error(`Example doesn't use $ref: ${exampleName}`);
                    }
                    const exampleObj = (await this.jsonLoader.resolveFile(example.$ref));
                    const ctx = {
                        exampleName: exampleName,
                        exampleFilePath: this.jsonLoader.getRealPath(example.$ref),
                        bodyTransform: this.bodyTransformer,
                    };
                    for (const func of this.validationFuncs) {
                        const ruleRes = await func(exampleObj, operation, this.jsonLoader, ctx);
                        res = res.concat(ruleRes);
                    }
                }
            }
        };
        for (const spec of this.swaggerSpecs) {
            await (0, traverseSwagger_1.traverseSwaggerAsync)(spec, { onOperation: onOperation });
        }
        return res;
    }
};
exports.ExampleQualityValidator = ExampleQualityValidator;
exports.ExampleQualityValidator = ExampleQualityValidator = ExampleQualityValidator_1 = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(inversifyUtils_1.TYPES.opts)),
    __metadata("design:paramtypes", [Object, jsonLoader_1.JsonLoader,
        swaggerLoader_1.SwaggerLoader,
        bodyTransformer_1.BodyTransformer])
], ExampleQualityValidator);
