// oav/dist/lib/transform/referenceFieldsTransformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.referenceFieldsTransformer = void 0;
const resolveNestedDefinitionTransformer_1 = require("./resolveNestedDefinitionTransformer");
const transformer_1 = require("./transformer");
const traverseSwagger_1 = require("./traverseSwagger");
const defaultMime = ["application/json"];
const isDefaultMime = (mimes) => {
    return mimes.length === 1 && mimes[0] === defaultMime[0];
};
exports.referenceFieldsTransformer = {
    type: transformer_1.TransformerType.Spec,
    before: [resolveNestedDefinitionTransformer_1.resolveNestedDefinitionTransformer],
    transform: (spec) => {
        if (spec.consumes === undefined || isDefaultMime(spec.consumes)) {
            spec.consumes = defaultMime;
        }
        if (spec.produces === undefined || isDefaultMime(spec.produces)) {
            spec.produces = defaultMime;
        }
        (0, traverseSwagger_1.traverseSwagger)(spec, {
            onPath: (path, pathTemplate) => {
                path._spec = spec;
                path._pathTemplate = pathTemplate;
            },
            onOperation: (operation, path, method) => {
                operation._path = path;
                operation._method = method;
                if (operation.consumes === undefined) {
                    operation.consumes = spec.consumes;
                }
                else if (isDefaultMime(operation.consumes)) {
                    operation.consumes = defaultMime;
                }
                if (operation.produces === undefined) {
                    operation.produces = spec.produces;
                }
                else if (isDefaultMime(operation.produces)) {
                    operation.produces = defaultMime;
                }
            },
        });
    },
};
