// oav/dist/lib/transform/resolveNestedDefinitionTransformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveNestedDefinitionTransformer = void 0;
const json_schema_traverse_1 = require("json-schema-traverse");
const jsonLoader_1 = require("../swagger/jsonLoader");
const swaggerTypes_1 = require("../swagger/swaggerTypes");
const transformer_1 = require("./transformer");
const traverseSwagger_1 = require("./traverseSwagger");
const visited = new WeakSet();
exports.resolveNestedDefinitionTransformer = {
    type: transformer_1.TransformerType.Spec,
    transform(spec, { jsonLoader, objSchemas, arrSchemas, primSchemas, allParams }) {
        var _a;
        const queue = new Array();
        const visitNestedDefinitions = (s, ref) => {
            if (s === undefined || s === null || typeof s !== "object") {
                return;
            }
            const schema = jsonLoader.resolveRefObj(s);
            if (visited.has(schema)) {
                return;
            }
            visited.add(schema);
            const refSelf = schema === s ? ref : s.$ref;
            if (refSelf !== undefined) {
                schema[swaggerTypes_1.refSelfSymbol] = refSelf;
            }
            if (schema.type === undefined || schema.type === "object") {
                objSchemas.push(schema);
                if (schema.discriminator !== undefined && schema[swaggerTypes_1.refSelfSymbol] !== undefined) {
                    queue.push(schema[swaggerTypes_1.refSelfSymbol]);
                }
            }
            else if (schema.type === "array") {
                arrSchemas.push(schema);
            }
            else {
                primSchemas.push(schema);
            }
            for (const key of Object.keys(schema)) {
                const sch = schema[key];
                const refSch = refSelf === null || refSelf === void 0 ? void 0 : refSelf.concat("/", key);
                if (Array.isArray(sch)) {
                    if (key in json_schema_traverse_1.arrayKeywords) {
                        for (let idx = 0; idx < sch.length; ++idx) {
                            visitNestedDefinitions(sch[idx], refSch === null || refSch === void 0 ? void 0 : refSch.concat("/", idx.toString()));
                        }
                    }
                }
                else if (key in json_schema_traverse_1.propsKeywords) {
                    if (typeof sch === "object" && sch !== null) {
                        for (const prop of Object.keys(sch)) {
                            visitNestedDefinitions(sch[prop], refSch === null || refSch === void 0 ? void 0 : refSch.concat("/", prop));
                        }
                    }
                }
                else if (key in json_schema_traverse_1.keywords) {
                    visitNestedDefinitions(sch, refSch);
                }
            }
        };
        const visitParameters = (x) => {
            if (x.parameters !== undefined) {
                for (const p of x.parameters) {
                    const param = jsonLoader.resolveRefObj(p);
                    if (param.in === "body") {
                        visitNestedDefinitions(param.schema);
                    }
                    allParams.push(param);
                }
            }
        };
        (0, traverseSwagger_1.traverseSwagger)(spec, {
            onPath: visitParameters,
            onOperation: visitParameters,
            onResponse: (response) => {
                visitNestedDefinitions(response.schema);
            },
        });
        if (spec.definitions !== undefined) {
            for (const key of Object.keys(spec.definitions)) {
                visitNestedDefinitions(spec.definitions[key], `${spec[jsonLoader_1.$id]}#/definitions/${key}`);
            }
        }
        if (spec.parameters !== undefined) {
            for (const key of Object.keys(spec.parameters)) {
                spec.parameters[key][swaggerTypes_1.refSelfSymbol] = `${spec[jsonLoader_1.$id]}#/parameters/${key}`;
            }
        }
        while (queue.length > 0) {
            const ref = queue.shift();
            const idx = ref.indexOf("#");
            const mockName = idx === -1 ? ref : ref.substr(0, idx);
            const specFile = jsonLoader.resolveMockedFile(mockName);
            if (specFile.definitions !== undefined) {
                for (const key of Object.keys(specFile.definitions)) {
                    const sch = specFile.definitions[key];
                    if (((_a = sch.allOf) === null || _a === void 0 ? void 0 : _a.find(function (s) {
                        return s.$ref === ref;
                    })) !== undefined) {
                        visitNestedDefinitions(specFile.definitions[key], `${mockName}#/definitions/${key}`);
                        queue.push(`${mockName}#/definitions/${key}`);
                    }
                }
            }
        }
    },
};
