// oav/dist/lib/transform/transformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyGlobalTransformers = exports.applySpecTransformers = exports.sortTransformers = exports.TransformerType = void 0;
const toposort_1 = require("toposort");
var TransformerType;
(function (TransformerType) {
    TransformerType["Spec"] = "spec";
    TransformerType["Global"] = "global";
})(TransformerType || (exports.TransformerType = TransformerType = {}));
const sortTransformers = (transformers) => {
    const specTransformers = new Set();
    const globalTransformers = new Set();
    for (const t of transformers) {
        switch (t.type) {
            case TransformerType.Spec:
                specTransformers.add(t);
                break;
            case TransformerType.Global:
                globalTransformers.add(t);
                break;
        }
    }
    const sortTrans = (ts) => {
        const edges = [];
        for (const t of ts) {
            if (t.after !== undefined) {
                for (const tAfter of t.after) {
                    if (ts.has(tAfter)) {
                        edges.push([tAfter, t]);
                    }
                }
            }
            if (t.before !== undefined) {
                for (const tBefore of t.before) {
                    if (ts.has(tBefore)) {
                        edges.push([t, tBefore]);
                    }
                }
            }
        }
        return (0, toposort_1.array)([...ts], edges);
    };
    return {
        specTransformers: sortTrans(specTransformers),
        globalTransformers: sortTrans(globalTransformers),
    };
};
exports.sortTransformers = sortTransformers;
const applySpecTransformers = (spec, ctx) => {
    for (const t of ctx.specTransformers) {
        t.transform(spec, ctx);
    }
};
exports.applySpecTransformers = applySpecTransformers;
const applyGlobalTransformers = (ctx) => {
    for (const t of ctx.globalTransformers) {
        t.transform(ctx);
    }
};
exports.applyGlobalTransformers = applyGlobalTransformers;
