// oav/dist/lib/armValidator/roundTripValidator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.diffRequestResponse = diffRequestResponse;
exports.buildLiveValidationIssue = buildLiveValidationIssue;
const diffUtils_1 = require("../apiScenario/diffUtils");
const errorDefinitions_1 = require("../util/errorDefinitions");
const utils = __importStar(require("../util/utils"));
const schemaSearcher_1 = require("../apiScenario/schemaSearcher");
const allowed = true;
const notAllowed = false;
function checkReplacedSchemaInParameter(jsonPath, parameter, jsonLoader) {
    const subSet = (currentValue) => ["create", "read"].includes(currentValue);
    if (parameter.in === "body") {
        const schema = jsonLoader.resolveRefObj(parameter.schema);
        const foundSchema = schemaSearcher_1.SchemaSearcher.findSchemaByJsonPointer(jsonPath, schema, jsonLoader);
        if (foundSchema.readOnly ||
            foundSchema.default ||
            (foundSchema["x-ms-mutability"] && foundSchema["x-ms-mutability"].every(subSet))) {
            return allowed;
        }
        return notAllowed;
    }
    return notAllowed;
}
function checkRemovedSchemaInParameter(jsonPath, parameter, jsonLoader) {
    const subSet = (currentValue) => ["create", "update"].includes(currentValue);
    if (parameter.in === "body") {
        const schema = jsonLoader.resolveRefObj(parameter.schema);
        const foundSchema = schemaSearcher_1.SchemaSearcher.findSchemaByJsonPointer(jsonPath, schema, jsonLoader);
        if (foundSchema["x-ms-secret"] ||
            (foundSchema["x-ms-mutability"] && foundSchema["x-ms-mutability"].every(subSet))) {
            return allowed;
        }
        return notAllowed;
    }
    return notAllowed;
}
function checkSchemaInResponse(jsonPath, op, jsonLoader, responseStatusCode) {
    let statusCode;
    if (isNaN(+responseStatusCode)) {
        statusCode = utils.statusCodeStringToStatusCode[responseStatusCode.toLowerCase()];
        if (statusCode === undefined) {
            statusCode = "default";
        }
    }
    else {
        statusCode = responseStatusCode;
    }
    let responseSchema = op.responses[statusCode];
    if (responseSchema === undefined) {
        responseSchema = op.responses["default"];
    }
    if (responseSchema.schema) {
        responseSchema = responseSchema.schema;
    }
    const schema = jsonLoader.resolveRefObj(responseSchema);
    const foundSchema = schemaSearcher_1.SchemaSearcher.findSchemaByJsonPointer(jsonPath, schema, jsonLoader);
    if (foundSchema.readOnly || foundSchema.default) {
        return allowed;
    }
    return notAllowed;
}
function diffRequestResponse(payload, info, jsonLoader) {
    var _a, _b;
    const diffs = (0, diffUtils_1.getJsonPatchDiff)((_a = payload.liveRequest.body) !== null && _a !== void 0 ? _a : {}, (_b = payload.liveResponse.body) !== null && _b !== void 0 ? _b : {}, {
        includeOldValue: true,
        minimizeDiff: false,
    });
    const rest = diffs
        .map((it) => {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        const jsonPath = it.remove || it.add || it.replace;
        if (it.replace !== undefined) {
            let isAllowed = false;
            for (let parameter of (_b = (_a = info.operationMatch) === null || _a === void 0 ? void 0 : _a.operation.parameters) !== null && _b !== void 0 ? _b : []) {
                if (isAllowed) {
                    break;
                }
                isAllowed = checkReplacedSchemaInParameter(it.replace, parameter, jsonLoader);
            }
            for (let parameter of (_d = (_c = info.operationMatch) === null || _c === void 0 ? void 0 : _c.operation._path.parameters) !== null && _d !== void 0 ? _d : []) {
                if (isAllowed) {
                    break;
                }
                isAllowed = checkReplacedSchemaInParameter(it.replace, parameter, jsonLoader);
            }
            if (!isAllowed) {
                return buildLiveValidationIssue("ROUNDTRIP_INCONSISTENT_PROPERTY", jsonPath, it);
            }
        }
        else if (it.add !== undefined && it.value !== null) {
            // IF a property is not in request but returned in response as null, ignore.
            let isAllowed = checkSchemaInResponse(it.add, (_e = info.operationMatch) === null || _e === void 0 ? void 0 : _e.operation, jsonLoader, payload.liveResponse.statusCode);
            if (!isAllowed) {
                return buildLiveValidationIssue("ROUNDTRIP_ADDITIONAL_PROPERTY", jsonPath, it);
            }
        }
        else if (it.remove !== undefined) {
            let isAllowed = false;
            for (let parameter of (_g = (_f = info.operationMatch) === null || _f === void 0 ? void 0 : _f.operation.parameters) !== null && _g !== void 0 ? _g : []) {
                if (isAllowed) {
                    break;
                }
                isAllowed = checkRemovedSchemaInParameter(it.remove, parameter, jsonLoader);
            }
            for (let parameter of (_j = (_h = info.operationMatch) === null || _h === void 0 ? void 0 : _h.operation._path.parameters) !== null && _j !== void 0 ? _j : []) {
                if (isAllowed) {
                    break;
                }
                isAllowed = checkRemovedSchemaInParameter(it.remove, parameter, jsonLoader);
            }
            if (!isAllowed) {
                return buildLiveValidationIssue("ROUNDTRIP_MISSING_PROPERTY", jsonPath, it);
            }
        }
        return undefined;
    })
        .filter((a) => a !== undefined);
    return rest;
}
function buildLiveValidationIssue(errorCode, path, it) {
    let severity, message;
    const properties = path.split("/");
    let property = properties.pop();
    if (!isNaN(Number(property)) && properties.length > 0) {
        property = `${properties.pop()}/${property}`;
    }
    switch (errorCode) {
        case "ROUNDTRIP_INCONSISTENT_PROPERTY": {
            severity = errorDefinitions_1.roundTripValidationErrors.ROUNDTRIP_INCONSISTENT_PROPERTY.severity;
            message = errorDefinitions_1.roundTripValidationErrors.ROUNDTRIP_INCONSISTENT_PROPERTY.message({
                getValue: it.value,
                putValue: it.oldValue,
            });
            break;
        }
        case "ROUNDTRIP_ADDITIONAL_PROPERTY": {
            severity = errorDefinitions_1.roundTripValidationErrors.ROUNDTRIP_ADDITIONAL_PROPERTY.severity;
            message = errorDefinitions_1.roundTripValidationErrors.ROUNDTRIP_ADDITIONAL_PROPERTY.message({
                property: property,
            });
            break;
        }
        case "ROUNDTRIP_MISSING_PROPERTY": {
            severity = errorDefinitions_1.roundTripValidationErrors.ROUNDTRIP_MISSING_PROPERTY.severity;
            message = errorDefinitions_1.roundTripValidationErrors.ROUNDTRIP_MISSING_PROPERTY.message({
                property: property,
            });
            break;
        }
    }
    const ret = {
        code: errorCode,
        pathsInPayload: [path],
        severity: severity,
        message: message,
        jsonPathsInPayload: [],
        schemaPath: "",
        source: {
            url: "",
            position: {
                column: 0,
                line: 0,
            },
        },
    };
    return ret;
}
