// @ts-common/fs/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const path = tslib_1.__importStar(require("path"));
const asyncIt = tslib_1.__importStar(require("@ts-common/async-iterator"));
const util = tslib_1.__importStar(require("util"));
exports.readFile = util.promisify(fs.readFile);
exports.writeFile = util.promisify(fs.writeFile);
exports.exists = util.promisify(fs.exists);
exports.readdir = util.promisify(fs.readdir);
exports.mkdir = util.promisify(fs.mkdir);
exports.rmdir = util.promisify(fs.rmdir);
exports.unlink = util.promisify(fs.unlink);
exports.recursiveReaddir = (dir) => asyncIt.fromPromise(exports.readdir(dir, { withFileTypes: true })).flatMap(f => {
    const p = path.join(dir, f.name);
    return f.isDirectory() ? exports.recursiveReaddir(p) : asyncIt.fromSequence(p);
});
exports.recursiveRmdir = async (dir) => {
    const list = await exports.readdir(dir, { withFileTypes: true });
    await Promise.all(list.map(async (f) => {
        const p = path.join(dir, f.name);
        if (f.isDirectory()) {
            await exports.recursiveRmdir(p);
        }
        else {
            await exports.unlink(p);
        }
    }));
    await exports.rmdir(dir);
};
