// @microsoft.azure/openapi-validator/dist/openapi-validator-plugin-func.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.openapiValidatorPluginFunc = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const openapi_validator_rulesets_1 = require("@microsoft.azure/openapi-validator-rulesets");
const plugin_common_1 = require("./plugin-common");
const _1 = require(".");
async function openapiValidatorPluginFunc(initiator) {
    initiator.Message({
        Channel: "information",
        Text: `openapiValidatorPluginFunc: Enter`,
    });
    const files = await (await initiator.ListInputs()).filter((f) => !(0, plugin_common_1.isCommonTypes)(f));
    const openapiType = await (0, plugin_common_1.getOpenapiTypeStr)(initiator);
    const sendMessage = (msg) => {
        initiator.Message((0, plugin_common_1.convertLintMsgToAutoRestMsg)(msg));
    };
    const readFile = async (fileUri) => {
        if ((0, openapi_validator_core_1.isExample)(fileUri)) {
            return "";
        }
        let file = _1.cachedFiles.get(fileUri);
        if (!file) {
            file = await initiator.ReadFile(fileUri);
            if (!file) {
                throw new Error(`Could not read file: ${fileUri} .`);
            }
            _1.cachedFiles.set(fileUri, file);
        }
        return file;
    };
    const defaultFileSystem = {
        read: readFile,
    };
    initiator.Message({
        Channel: "information",
        Text: `openapiValidatorPluginFunc: Validating files:`,
    });
    for (const [index, file] of files.entries()) {
        initiator.Message({
            Channel: "information",
            Text: `openapiValidatorPluginFunc: File ${index + 1}/${files.length}: '${file}'`,
        });
    }
    try {
        const mergedRuleset = mergeRulesets(Object.values(openapi_validator_rulesets_1.nativeRulesets));
        const resolvedOpenapiType = (0, openapi_validator_core_1.getOpenapiType)(openapiType);
        printRuleNames(initiator, mergedRuleset, resolvedOpenapiType);
        await (0, openapi_validator_core_1.lint)(files, { ruleSet: mergedRuleset, fileSystem: defaultFileSystem, openapiType: resolvedOpenapiType }, sendMessage);
    }
    catch (e) {
        initiator.Message({
            Channel: "fatal",
            Text: `openapiValidatorPluginFunc: Failed validating: ` + e,
        });
    }
    initiator.Message({
        Channel: "information",
        Text: `openapiValidatorPluginFunc: Return`,
    });
}
exports.openapiValidatorPluginFunc = openapiValidatorPluginFunc;
const mergeRulesets = (rulesets) => {
    let rules = {};
    rulesets.forEach((set) => {
        rules = {
            ...rules,
            ...set.rules,
        };
    });
    const mergedRuleSet = {
        documentationUrl: "",
        rules,
    };
    return mergedRuleSet;
};
function printRuleNames(initiator, ruleset, resolvedOpenapiType) {
    const ruleNames = Object.keys(ruleset.rules)
        .sort(Intl.Collator().compare);
    initiator.Message({
        Channel: "information",
        Text: `Loaded ${ruleNames.length} native & legacy rules, for OpenAPI type '${openapi_validator_core_1.OpenApiTypes[resolvedOpenapiType]}':`,
    });
    for (const ruleName of ruleNames) {
        const severity = ruleset.rules[ruleName].severity;
        initiator.Message({
            Channel: "information",
            Text: `Native or legacy rule, severity '${severity}': '${ruleName}'`,
        });
    }
}
