// @microsoft.azure/openapi-validator/dist/plugin-common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isCommonTypes = exports.getOpenapiTypeStr = exports.convertLintMsgToAutoRestMsg = void 0;
function convertLintMsgToAutoRestMsg(message) {
    var _a, _b;
    const path = ((_a = message.jsonpath) === null || _a === void 0 ? void 0 : _a[1]) === "paths" && message.jsonpath[2];
    const pathComponents = typeof path === "string" && path.split("/");
    const pathComponentsProviderIndex = pathComponents && pathComponents.indexOf("providers");
    const pathComponentsTail = pathComponentsProviderIndex && pathComponentsProviderIndex >= 0 && pathComponents.slice(pathComponentsProviderIndex + 1);
    const pathComponentProviderNamespace = pathComponentsTail && pathComponentsTail[0];
    const pathComponentResourceType = pathComponentsTail && pathComponentsTail[1];
    const msg = {
        Channel: message.type,
        Text: message.message,
        Key: [message.code],
        Source: [
            {
                document: ((_b = message === null || message === void 0 ? void 0 : message.sources) === null || _b === void 0 ? void 0 : _b[0]) || "",
                Position: {
                    path: message.jsonpath,
                },
            },
        ],
        Details: {
            jsonpath: message.jsonpath,
            validationCategory: message.category,
            providerNamespace: pathComponentProviderNamespace,
            resourceType: pathComponentResourceType,
            rpcGuidelineCode: message.rpcGuidelineCode,
            range: message.range,
        },
    };
    return msg;
}
exports.convertLintMsgToAutoRestMsg = convertLintMsgToAutoRestMsg;
async function getOpenapiTypeStr(initiator) {
    let openapiType = await initiator.GetValue("openapi-type");
    let subType = await initiator.GetValue("openapi-subtype");
    subType = subType === "providerHub" ? "rpaas" : subType;
    if (subType === "rpaas") {
        openapiType = "rpaas";
    }
    return openapiType;
}
exports.getOpenapiTypeStr = getOpenapiTypeStr;
function isCommonTypes(filePath) {
    const regex = new RegExp(/.*common-types\/resource-management\/v.*.json/);
    return regex.test(filePath);
}
exports.isCommonTypes = isCommonTypes;
