// @microsoft.azure/openapi-validator/dist/spectral-plugin-func.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.spectralPluginFunc = void 0;
const url_1 = require("url");
const uri_1 = require("@azure-tools/uri");
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const openapi_validator_rulesets_1 = require("@microsoft.azure/openapi-validator-rulesets");
const json_ref_resolver_1 = require("@stoplight/json-ref-resolver");
const spectral_core_1 = require("@stoplight/spectral-core");
const js_yaml_1 = require("js-yaml");
const plugin_common_1 = require("./plugin-common");
const spectral_plugin_utils_1 = require("./spectral-plugin-utils");
const _1 = require(".");
async function spectralPluginFunc(initiator) {
    initiator.Message({
        Channel: "information",
        Text: `spectralPluginFunc: Enter`,
    });
    const openApiSpecFiles = (await initiator.ListInputs()).filter((f) => !(0, plugin_common_1.isCommonTypes)(f));
    const openapiType = await (0, plugin_common_1.getOpenapiTypeStr)(initiator);
    const isStagingRun = await initiator.GetValue("is-staging-run");
    const resolvedOpenapiType = (0, openapi_validator_core_1.getOpenapiType)(openapiType);
    const { rulesetPayload, rulesetForManualSpecs, rulesetForTypeSpecGeneratedSpecs, } = await getRulesets(initiator, resolvedOpenapiType, isStagingRun);
    for (const openApiSpecFile of openApiSpecFiles) {
        await validateOpenApiSpecFileUsingSpectral(initiator, rulesetPayload, rulesetForManualSpecs, rulesetForTypeSpecGeneratedSpecs, openApiSpecFile);
    }
    initiator.Message({
        Channel: "information",
        Text: `spectralPluginFunc: Return`,
    });
}
exports.spectralPluginFunc = spectralPluginFunc;
async function getRulesets(initiator, resolvedOpenapiType, isStagingRun) {
    const rulesetPayload = await (0, spectral_plugin_utils_1.getRulesetPayload)(initiator, resolvedOpenapiType);
    const namesOfRulesInStagingOnly = (0, openapi_validator_rulesets_1.getNamesOfRulesInPayloadWithPropertySetToTrue)(rulesetPayload, "stagingOnly");
    const namesOfRulesDisabledForTypespecDataPlane = resolvedOpenapiType === openapi_validator_core_1.OpenApiTypes.dataplane
        ? (0, openapi_validator_rulesets_1.getNamesOfRulesInPayloadWithPropertySetToTrue)(rulesetPayload, "disableForTypeSpecDataPlane")
        : [];
    const rulesetPayloadCopy = JSON.parse(JSON.stringify(rulesetPayload));
    (0, openapi_validator_rulesets_1.deleteRulesPropertiesInPayloadNotValidForSpectralRules)(rulesetPayload);
    const rulesetForManualSpecs = new spectral_core_1.Ruleset(rulesetPayload, { severity: "recommended" });
    (0, spectral_plugin_utils_1.ifNotStagingRunDisableRulesInStagingOnly)(initiator, namesOfRulesInStagingOnly, isStagingRun, rulesetForManualSpecs);
    (0, spectral_plugin_utils_1.printRuleNames)(initiator, rulesetForManualSpecs, resolvedOpenapiType, "manually written OpenAPI specs");
    const rulesetForTypeSpecGeneratedSpecs = new spectral_core_1.Ruleset(rulesetPayload, { severity: "recommended" });
    (0, spectral_plugin_utils_1.ifNotStagingRunDisableRulesInStagingOnly)(initiator, namesOfRulesInStagingOnly, isStagingRun, rulesetForTypeSpecGeneratedSpecs);
    (0, openapi_validator_rulesets_1.disableRulesInRuleset)(rulesetForTypeSpecGeneratedSpecs, namesOfRulesDisabledForTypespecDataPlane);
    (0, spectral_plugin_utils_1.printRuleNames)(initiator, rulesetForTypeSpecGeneratedSpecs, resolvedOpenapiType, "TypeSpec-generated OpenAPI specs");
    return {
        rulesetPayload: rulesetPayloadCopy,
        rulesetForManualSpecs: rulesetForManualSpecs,
        rulesetForTypeSpecGeneratedSpecs,
    };
}
async function validateOpenApiSpecFileUsingSpectral(initiator, rulesetPayload, rulesetForManualSpecs, rulesetForTypeSpecGeneratedSpecs, openApiSpecFile) {
    if (openApiSpecFile.includes("common-types/resource-management")) {
        initiator.Message({
            Channel: "information",
            Text: `spectralPluginFunc: Ignoring file matching to 'common-types/resource-management': '${openApiSpecFile}'`,
        });
        return;
    }
    try {
        const openApiSpecFilePath = openApiSpecFile.startsWith("file:///") ? (0, url_1.fileURLToPath)(openApiSpecFile) : openApiSpecFile;
        const openApiSpecContent = await readFileUsingCache(initiator, openApiSpecFile);
        const openApiSpecYml = (0, js_yaml_1.load)(openApiSpecContent);
        const specIsGeneratedFromTypeSpec = Boolean(openApiSpecYml.info["x-typespec-generated"]);
        initiator.Message({
            Channel: "information",
            Text: `spectralPluginFunc: Validating OpenAPI spec. TypeSpec-generated: ${specIsGeneratedFromTypeSpec}. Path: '${openApiSpecFile}'`,
        });
        const spectral = newSpectral(initiator, openApiSpecFile);
        spectral.setRuleset(specIsGeneratedFromTypeSpec ? rulesetForTypeSpecGeneratedSpecs : rulesetForManualSpecs);
        const sendMessage = initiator.Message.bind(initiator);
        await (0, spectral_plugin_utils_1.runSpectral)(sendMessage, spectral, rulesetPayload, openApiSpecFilePath, openApiSpecYml);
    }
    catch (error) {
        (0, spectral_plugin_utils_1.catchSpectralRunErrors)(openApiSpecFile, error, initiator);
    }
}
function newSpectral(initiator, openApiSpecFile) {
    const resolveFile = {
        resolve: (ref, ctx) => {
            const href = ref.href();
            const openApiSpecFileUri = (0, openapi_validator_core_1.isUriAbsolute)(href) ? href : (0, uri_1.resolveUri)(openApiSpecFile, href);
            const openApiSpecContent = readFileUsingCache(initiator, openApiSpecFileUri);
            return openApiSpecContent;
        },
    };
    const jsonRefResolver = new json_ref_resolver_1.Resolver({
        resolvers: {
            file: resolveFile,
            http: resolveFile,
            https: resolveFile,
        },
    });
    const spectral = new spectral_core_1.Spectral({ resolver: jsonRefResolver });
    return spectral;
}
async function readFileUsingCache(initiator, fileUri) {
    let file = _1.cachedFiles.get(fileUri);
    if (!file) {
        file = await initiator.ReadFile(fileUri);
        if (!file) {
            throw new Error(`Could not read file: ${fileUri} .`);
        }
        _1.cachedFiles.set(fileUri, file);
    }
    return file;
}
