// @microsoft.azure/openapi-validator/dist/spectral-plugin-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.catchSpectralRunErrors = exports.runSpectral = exports.printRuleNames = exports.ifNotStagingRunDisableRulesInStagingOnly = exports.getRulesetPayload = void 0;
const uri_1 = require("@azure-tools/uri");
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const openapi_validator_rulesets_1 = require("@microsoft.azure/openapi-validator-rulesets");
const types_1 = require("@stoplight/types");
const plugin_common_1 = require("./plugin-common");
async function getRulesetPayload(initiator, openapiType) {
    let rulesetPayload;
    switch (openapiType) {
        case openapi_validator_core_1.OpenApiTypes.arm: {
            rulesetPayload = openapi_validator_rulesets_1.spectralRulesets.azARM;
            break;
        }
        case openapi_validator_core_1.OpenApiTypes.dataplane: {
            rulesetPayload = openapi_validator_rulesets_1.spectralRulesets.azDataplane;
            break;
        }
        default: {
            rulesetPayload = openapi_validator_rulesets_1.spectralRulesets.azCommon;
        }
    }
    return rulesetPayload;
}
exports.getRulesetPayload = getRulesetPayload;
function ifNotStagingRunDisableRulesInStagingOnly(initiator, namesOfRulesInStagingOnly, isStagingRun, ruleset) {
    if (isStagingRun) {
        initiator.Message({
            Channel: "information",
            Text: "Detected staging run. Running all enabled rules.",
        });
    }
    else {
        initiator.Message({
            Channel: "information",
            Text: "Detected production run. As a result, disabling all Spectral rules that are denoted to run only in staging. Names of rules being disabled: " +
                namesOfRulesInStagingOnly.join(", ") +
                ".",
        });
        (0, openapi_validator_rulesets_1.disableRulesInRuleset)(ruleset, namesOfRulesInStagingOnly);
    }
}
exports.ifNotStagingRunDisableRulesInStagingOnly = ifNotStagingRunDisableRulesInStagingOnly;
function printRuleNames(initiator, ruleset, resolvedOpenapiType, description) {
    const ruleNames = Object.keys(ruleset.rules)
        .sort(Intl.Collator().compare);
    initiator.Message({
        Channel: "information",
        Text: `Loaded ${ruleNames.length} spectral rules, for OpenAPI type '${openapi_validator_core_1.OpenApiTypes[resolvedOpenapiType]}' for ${description}:`,
    });
    for (const ruleName of ruleNames) {
        const severity = ruleset.rules[ruleName].severity;
        const sevStr = Number(severity) == -1 ? "DISABLED" : types_1.DiagnosticSeverity[severity];
        initiator.Message({
            Channel: "information",
            Text: (sevStr == "DISABLED" ? "DISABLED " : "").concat(`Spectral rule for ${description}, severity '${sevStr}': '${ruleName}'`),
        });
    }
}
exports.printRuleNames = printRuleNames;
async function runSpectral(sendMessage, spectral, rulesetPayload, openApiSpecFilePath, openApiSpecYml) {
    const mergedResults = [];
    const convertSeverity = (severity) => {
        switch (severity) {
            case 0:
                return "error";
            case 1:
                return "warning";
            case 2:
                return "info";
            default:
                return "info";
        }
    };
    const convertRange = (range) => {
        return {
            start: {
                line: range.start.line + 1,
                column: range.start.character,
            },
            end: {
                line: range.end.line + 1,
                column: range.end.character,
            },
        };
    };
    const removeXmsExampleFromPath = (paths) => {
        const index = paths.findIndex((item) => item === "x-ms-examples");
        if (index !== -1 && paths.length > index + 2) {
            return paths.slice(0, index + 2);
        }
        return paths;
    };
    const formatAsLintResultMessage = (result, spec) => {
        var _a, _b;
        return {
            type: convertSeverity(result.severity),
            category: "",
            code: result.code,
            message: result.message,
            jsonpath: result.path && result.path.length ? removeXmsExampleFromPath(result.path) : [],
            sources: [`${spec}`],
            location: {
                line: result.range.start.line + 1,
                column: result.range.start.character,
            },
            range: convertRange(result.range),
            rpcGuidelineCode: (_b = (_a = rulesetPayload.rules[result.code]) === null || _a === void 0 ? void 0 : _a.rpcGuidelineCode) !== null && _b !== void 0 ? _b : "",
        };
    };
    const results = await spectral.run(openApiSpecYml);
    mergedResults.push(...results.map((result) => formatAsLintResultMessage(result, (0, uri_1.createFileOrFolderUri)(openApiSpecFilePath))));
    for (const message of mergedResults) {
        sendMessage((0, plugin_common_1.convertLintMsgToAutoRestMsg)(message));
    }
    return mergedResults;
}
exports.runSpectral = runSpectral;
function catchSpectralRunErrors(file, error, initiator) {
    const errorMessages = [error];
    if (error && error.errors && Array.isArray(error.errors)) {
        error.errors.forEach((error, index) => {
            errorMessages.push(`Error ${index + 1}: ${error.message}`);
        });
    }
    const combinedErrorMessages = errorMessages.join("\n");
    initiator.Message({
        Channel: "fatal",
        Text: `spectralPluginFunc: Failed validating: '${file}'. Errors encountered:\n${combinedErrorMessages}`,
    });
}
exports.catchSpectralRunErrors = catchSpectralRunErrors;
