// @scalar/openapi-types/dist/schemas/3.1/unprocessed/security-scheme-object.js
import {
  ApiKeyInValues as OriginalApiKeyInValues,
  ApiKeySchema as OriginalApiKeySchema,
  AuthorizationCodeFlowSchema as OriginalAuthorizationCodeFlowSchema,
  ClientCredentialsFlowSchema as OriginalClientCredentialsFlowSchema,
  HttpSchema as OriginalHttpSchema,
  ImplicitFlowSchema as OriginalImplicitFlowSchema,
  MutualTlsSchema as OriginalMutualTlsSchema,
  OAuthFlowObjectSchema as OriginalOAuthFlowObjectSchema,
  OAuthFlowsObjectSchema as OriginalOAuthFlowsObjectSchema,
  OpenIdConnectSchema as OriginalOpenIdConnectSchema,
  PasswordFlowSchema as OriginalPasswordFlowSchema,
  SecuritySchemeObjectSchema as OriginalSecuritySchemeObjectSchema
} from "../processed/security-scheme-object.js";
const SecuritySchemeObjectSchema = OriginalSecuritySchemeObjectSchema;
const ApiKeyInValues = OriginalApiKeyInValues;
const ApiKeySchema = OriginalApiKeySchema;
const HttpSchema = OriginalHttpSchema;
const MutualTlsSchema = OriginalMutualTlsSchema;
const OpenIdConnectSchema = OriginalOpenIdConnectSchema;
const OAuthFlowsObjectSchema = OriginalOAuthFlowsObjectSchema;
const OAuthFlowObjectSchema = OriginalOAuthFlowObjectSchema;
const AuthorizationCodeFlowSchema = OriginalAuthorizationCodeFlowSchema;
const ClientCredentialsFlowSchema = OriginalClientCredentialsFlowSchema;
const ImplicitFlowSchema = OriginalImplicitFlowSchema;
const PasswordFlowSchema = OriginalPasswordFlowSchema;
export {
  ApiKeyInValues,
  ApiKeySchema,
  AuthorizationCodeFlowSchema,
  ClientCredentialsFlowSchema,
  HttpSchema,
  ImplicitFlowSchema,
  MutualTlsSchema,
  OAuthFlowObjectSchema,
  OAuthFlowsObjectSchema,
  OpenIdConnectSchema,
  PasswordFlowSchema,
  SecuritySchemeObjectSchema
};
