// @microsoft.azure/openapi-validator-rulesets/dist/native/utilities/arm-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArmHelper = void 0;
const tslib_1 = require("tslib");
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const lodash_1 = tslib_1.__importDefault(require("lodash"));
const jsonpath_1 = require("./jsonpath");
const rules_helper_1 = require("./rules-helper");
const swagger_helper_1 = require("./swagger-helper");
const swagger_walker_1 = require("./swagger-walker");
const swagger_workspace_1 = require("./swagger-workspace");
function* jsonPathIt(doc, jsonPath) {
    if (doc) {
        for (const node of (0, jsonpath_1.nodes)(doc, jsonPath)) {
            yield node.value;
        }
    }
}
class ArmHelper {
    constructor(innerDoc, specPath, inventory) {
        this.innerDoc = innerDoc;
        this.specPath = specPath;
        this.inventory = inventory;
        this.BaseResourceModelNames = ["trackedresource", "proxyresource", "resource", "azureentityresource"];
        this.ResourceGroupWideResourceRegEx = new RegExp("^/subscriptions/{[^/]+}/resourceGroups/{[^/]+}/", "gi");
        this.SubscriptionsWideResourceRegEx = new RegExp("^/subscriptions/{[^/]+}/providers/", "gi");
        this.OperationApiRegEx = new RegExp("^/providers/[^/]+/operations$", "gi");
        this.SpecificResourcePathRegEx = new RegExp("/providers/[^/]+(?:/\\w+/default|/\\w+/{[^/]+})+$", "gi");
        this.ExtensionResourceFullyQualifiedPathReg = new RegExp(".+/providers/.+/providers/.+$", "gi");
        this.ExtensionResourceReg = new RegExp("^/{\\w+}/providers/.+$", "gi");
        this.XmsResources = new Set();
        this.resources = [];
        this.swaggerUtil = new swagger_helper_1.SwaggerHelper(this.innerDoc, this.specPath, this.inventory);
        this.getXmsResources();
        this.getAllResources();
    }
    getBodyParameter(parameters) {
        let bodyParameter;
        if (parameters && Array.isArray(parameters)) {
            parameters.forEach((param) => {
                if (param.$ref) {
                    const resolvedParam = this.swaggerUtil.resolveRef(this.enhancedSchema(param));
                    if (resolvedParam && resolvedParam.value && resolvedParam.value.in === "body") {
                        bodyParameter = this.enhancedSchema(resolvedParam.value.schema, resolvedParam.file);
                    }
                }
            });
        }
        return bodyParameter;
    }
    populateOperations(doc, specPath) {
        var _a, _b;
        const paths = { ...(doc.paths || {}), ...(doc["x-ms-paths"] || {}) };
        const operations = [];
        for (const [key, value] of Object.entries(paths)) {
            for (const [method, operation] of Object.entries(value)) {
                if (method !== "parameters") {
                    const op = operation;
                    const response = ((_a = op === null || op === void 0 ? void 0 : op.responses) === null || _a === void 0 ? void 0 : _a["200"]) || ((_b = op === null || op === void 0 ? void 0 : op.responses) === null || _b === void 0 ? void 0 : _b["201"]);
                    const requestBodyParameter = this.getBodyParameter(op.parameters);
                    if (response) {
                        operations.push({
                            specPath,
                            apiPath: key,
                            httpMethod: method,
                            requestBodyParameter,
                            responseSchema: response.schema,
                            operationId: op === null || op === void 0 ? void 0 : op.operationId,
                        });
                    }
                }
            }
        }
        return operations;
    }
    populateResources(doc, specPath) {
        const operations = this.populateOperations(doc, specPath);
        for (const op of operations) {
            const resourceInfo = this.extractResourceInfo(op.responseSchema, specPath);
            if (resourceInfo) {
                const existing = this.resources.find((re) => re.modelName === resourceInfo.modelName && re.specPath === resourceInfo.specPath);
                if (existing) {
                    existing.operations.push(op);
                }
                else {
                    this.resources.push({
                        ...resourceInfo,
                        operations: [op],
                    });
                }
            }
        }
    }
    getXmsResources() {
        for (const name of Object.keys(this.innerDoc.definitions || {})) {
            const model = this.getInternalModel(name);
            for (const extension of jsonPathIt(model === null || model === void 0 ? void 0 : model.value, `$..['x-ms-azure-resource']`)) {
                if (extension === true) {
                    this.XmsResources.add(name);
                    break;
                }
            }
            if (this.checkResource(name)) {
                this.XmsResources.add(name);
            }
        }
        let resources = this.getAllOfResources();
        while (resources && resources.length) {
            resources.forEach((re) => this.XmsResources.add(re));
            resources = this.getAllOfResources();
        }
    }
    getAllOfResources() {
        if (!this.innerDoc.definitions) {
            return [];
        }
        const keys = Object.keys(this.innerDoc.definitions || {}).filter((key) => !this.XmsResources.has(key));
        const AllResources = keys.reduce((pre, cur) => {
            if (this.getResourceHierarchy(cur).some((model) => this.checkResource(model))) {
                return [...pre, cur];
            }
            else {
                return pre;
            }
        }, []);
        return AllResources;
    }
    getInternalModel(modelName) {
        var _a, _b;
        if (!modelName) {
            return undefined;
        }
        return swagger_workspace_1.Workspace.createEnhancedSchema((_b = (_a = this.innerDoc) === null || _a === void 0 ? void 0 : _a.definitions) === null || _b === void 0 ? void 0 : _b[modelName], this.specPath);
    }
    getResourceByName(modelName) {
        var _a;
        if (!modelName) {
            return undefined;
        }
        const resourceInfo = this.resources.find((re) => re.modelName === modelName);
        if (!resourceInfo) {
            return undefined;
        }
        return swagger_workspace_1.Workspace.createEnhancedSchema((_a = this.inventory.getDocuments(resourceInfo.specPath).definitions) === null || _a === void 0 ? void 0 : _a[resourceInfo.modelName], resourceInfo.specPath);
    }
    checkResource(modelName) {
        if (this.BaseResourceModelNames.includes(modelName.toLowerCase())) {
            return true;
        }
        if (this.XmsResources.has(modelName)) {
            return true;
        }
        return false;
    }
    stripDefinitionPath(reference) {
        const refPrefix = "#/definitions/";
        if (!reference) {
            return undefined;
        }
        const index = reference.indexOf(refPrefix);
        if (index !== -1) {
            return reference.substr(index + refPrefix.length);
        }
        return undefined;
    }
    extractResourceInfo(schema, specPath) {
        if (schema && schema.$ref) {
            const segments = (0, openapi_validator_core_1.parseJsonRef)(schema.$ref);
            return {
                specPath: segments[0] || specPath || this.specPath,
                modelName: this.stripDefinitionPath("#" + segments[1]),
            };
        }
        return undefined;
    }
    getAllNestedResources() {
        const fullResources = this.getAllResources();
        const nestedResource = new Set();
        for (const re of fullResources) {
            const operations = re.operations;
            operations
                .filter((op) => op.apiPath.toLowerCase().startsWith("/subscriptions/"))
                .some((op) => {
                const hierarchy = this.getResourcesTypeHierarchy(op.apiPath);
                if (hierarchy.length > 1) {
                    nestedResource.add(re.modelName);
                    return true;
                }
                return false;
            });
        }
        return Array.from(nestedResource.values());
    }
    getTopLevelResources() {
        const fullResources = this.getAllResources();
        return fullResources.filter((re) => re.operations.some((op) => {
            const hierarchy = this.getResourcesTypeHierarchy(op.apiPath);
            if (hierarchy.length === 1 && !this.isPathOfExtensionResource(op.apiPath)) {
                return true;
            }
            return false;
        }));
    }
    getTopLevelResourceNames() {
        return lodash_1.default.uniq(Array.from(this.getTopLevelResources().map((re) => re.modelName)));
    }
    getTopLevelResourcesByRG() {
        return lodash_1.default.uniq(Array.from(this.getTopLevelResources()
            .filter((re) => re.operations.some((op) => this.isPathByResourceGroup(op.apiPath)))
            .map((re) => re.modelName)));
    }
    getAllResources(includeGet = false, useArmResources = true, includeListOperationPath = false) {
        if (useArmResources && this.armResources) {
            return this.armResources;
        }
        if (useArmResources) {
            this.populateResources(this.innerDoc, this.specPath);
        }
        const references = this.inventory.referencesOf(this.specPath);
        for (const [specPath, reference] of Object.entries(references)) {
            this.populateResources(reference, specPath);
        }
        const localResourceModels = this.resources.filter((re) => re.specPath === this.specPath);
        const resWithXmsRes = localResourceModels.filter((re) => this.XmsResources.has(re.modelName) && !this.BaseResourceModelNames.includes(re.modelName.toLowerCase()));
        const resWithPutOrPatch = includeGet
            ? localResourceModels.filter((re) => re.operations.some((op) => (op.httpMethod === "get" && (includeListOperationPath ? true : !(0, rules_helper_1.isListOperationPath)(op.apiPath))) ||
                op.httpMethod === "put" ||
                op.httpMethod === "patch"))
            : localResourceModels.filter((re) => re.operations.some((op) => op.httpMethod === "put" || op.httpMethod == "patch"));
        const reWithPostOnly = resWithXmsRes.filter((re) => re.operations.every((op) => op.httpMethod === "post"));
        const resources = lodash_1.default.uniqWith(resWithXmsRes.filter((re) => !reWithPostOnly.some((re1) => re1.modelName === re.modelName)).concat(resWithPutOrPatch), lodash_1.default.isEqual);
        if (useArmResources) {
            this.armResources = resources;
        }
        return resources;
    }
    getTrackedResources() {
        const isTrackedResource = (schema) => {
            const enhancedSchema = this.enhancedSchema(schema);
            return !!this.getProperty(enhancedSchema, "location");
        };
        const allTrackedResources = this.getAllResources().filter((re) => {
            const schema = re.operations.find((op) => op.responseSchema);
            if (schema) {
                return isTrackedResource(schema.responseSchema);
            }
            return false;
        });
        return allTrackedResources;
    }
    getProxyResources() {
        const isProxyResource = (schema) => {
            const enhancedSchema = this.enhancedSchema(schema);
            return !this.getProperty(enhancedSchema, "location");
        };
        const allProxyResources = this.getAllResources().filter((re) => {
            const schema = re.operations.find((op) => op.responseSchema);
            if (schema) {
                return isProxyResource(schema.responseSchema);
            }
            return false;
        });
        return allProxyResources;
    }
    getAllResourceNames() {
        const fullResources = this.getAllResources();
        const resources = new Set();
        for (const re of fullResources) {
            const operations = re.operations;
            operations.some((op) => {
                const hierarchy = this.getResourcesTypeHierarchy(op.apiPath);
                if (hierarchy.length > 0) {
                    resources.add(re.modelName);
                    return true;
                }
                return false;
            });
        }
        return [...resources.values()];
    }
    getResourcesTypeHierarchy(path) {
        const index = path.lastIndexOf("/providers/");
        if (index === -1) {
            return [];
        }
        const lastProvider = path.substr(index);
        const result = [];
        const matches = lastProvider.match(this.SpecificResourcePathRegEx);
        if (matches && matches.length) {
            const match = matches[0];
            const segments = match.split("/").slice(3);
            for (const segment of segments) {
                if (segment.startsWith("{") || segment === "default") {
                    continue;
                }
                result.push(segment);
            }
        }
        return result;
    }
    getResourceHierarchy(model) {
        let hierarchy = [];
        const enhancedModel = typeof model === "string" ? this.getInternalModel(model) : model;
        if (!enhancedModel) {
            return hierarchy;
        }
        const unwrappedSchema = swagger_workspace_1.Workspace.resolveRef(enhancedModel, this.inventory);
        for (const refs of jsonPathIt(unwrappedSchema.value, `$.allOf`)) {
            refs
                .filter((ref) => !!ref.$ref)
                .forEach((ref) => {
                const allOfModel = this.stripDefinitionPath(ref.$ref);
                if (allOfModel) {
                    hierarchy.push(allOfModel);
                    hierarchy = hierarchy.concat(this.getResourceHierarchy(this.enhancedSchema(ref, unwrappedSchema.file)));
                }
            });
        }
        return hierarchy;
    }
    containsDiscriminatorInternal(model) {
        var _a;
        if (model) {
            const unWrappedModel = swagger_workspace_1.Workspace.resolveRef(model, this.inventory);
            if ((unWrappedModel === null || unWrappedModel === void 0 ? void 0 : unWrappedModel.value) && (unWrappedModel === null || unWrappedModel === void 0 ? void 0 : unWrappedModel.value.allOf)) {
                for (const ref of unWrappedModel.value.allOf) {
                    const unWrappedRef = swagger_workspace_1.Workspace.resolveRef(this.enhancedSchema(ref), this.inventory);
                    if (((_a = unWrappedRef === null || unWrappedRef === void 0 ? void 0 : unWrappedRef.value) === null || _a === void 0 ? void 0 : _a.discriminator) || (unWrappedRef && this.containsDiscriminatorInternal(unWrappedRef))) {
                        return true;
                    }
                }
            }
        }
        return false;
    }
    containsDiscriminator(modelName) {
        let model;
        if (typeof modelName === "string") {
            model = this.getInternalModel(modelName);
        }
        if (model) {
            return this.containsDiscriminatorInternal(model);
        }
        return false;
    }
    getOperationApi() {
        const walker = new swagger_walker_1.SwaggerWalker(this.inventory);
        let result = undefined;
        walker.warkAll(["$.[paths,x-ms-paths].*"], (path, value, rootPath) => {
            var _a, _b;
            const apiPath = path[2];
            const matchResult = apiPath.match(this.OperationApiRegEx);
            if (matchResult) {
                result = [path, this.enhancedSchema((_b = (_a = value === null || value === void 0 ? void 0 : value.get) === null || _a === void 0 ? void 0 : _a.responses["200"]) === null || _b === void 0 ? void 0 : _b.schema, rootPath)];
            }
        });
        return result;
    }
    resourcesWithGetOperations() {
        return this.resources.filter((re) => re.operations.some((op) => op.httpMethod === "get"));
    }
    resourcesWithPutPatchOperations() {
        var _a;
        return ((_a = this.armResources) === null || _a === void 0 ? void 0 : _a.filter((re) => re.operations.some((op) => op.httpMethod === "put" || op.httpMethod == "patch"))) || [];
    }
    getCollectionApiInfo() {
        const getOperationModels = this.resourcesWithGetOperations();
        const allPathKeys = lodash_1.default.uniq(lodash_1.default.flattenDeep(this.resources.map((re) => re.operations.map((op) => op.apiPath))));
        const getResourcePaths = (res, name) => lodash_1.default.flattenDeep(res.filter((re) => re.modelName === name).map((re) => re.operations.map((op) => op.apiPath)));
        const collectionApis = [];
        for (const re of getOperationModels) {
            re.operations.forEach((op) => {
                const path = op.apiPath;
                if (collectionApis.find((re) => re.specificGetPath[0] === path)) {
                    return;
                }
                if (path.match(this.SpecificResourcePathRegEx)) {
                    const firstProviderIndex = path.lastIndexOf("/providers");
                    const lastIndex = path.lastIndexOf("/");
                    const possibleCollectionApiPath = path.substr(firstProviderIndex, lastIndex - firstProviderIndex);
                    const matchedPaths = allPathKeys.filter((p) => p
                        .substr(p.lastIndexOf("/providers"))
                        .replace(/{[^/]+}/gi, "{}")
                        .replace(/\/$/gi, "") === possibleCollectionApiPath.replace(/{[^/]+}/gi, "{}"));
                    if (matchedPaths && matchedPaths.length >= 1 && this.getOperationIdFromPath(matchedPaths[0])) {
                        collectionApis.push({
                            specificGetPath: [path],
                            collectionGetPath: matchedPaths,
                            modelName: this.getResponseModelNameFromPath(matchedPaths[0]),
                            childModelName: re.modelName,
                        });
                    }
                }
            });
        }
        const collectionResources = this.getCollectionResources();
        for (const resource of collectionResources) {
            if (getOperationModels.some((re) => re.modelName === resource[1])) {
                const index = collectionApis.findIndex((e) => e.modelName === resource[1]);
                const collectionInfo = {
                    specificGetPath: getResourcePaths(getOperationModels, resource[0]) || [],
                    collectionGetPath: getResourcePaths(getOperationModels, resource[1]) || [],
                    modelName: resource[1],
                    childModelName: resource[0],
                };
                if (index === -1) {
                    collectionApis.push(collectionInfo);
                }
            }
        }
        return collectionApis;
    }
    getCollectionResources() {
        const resourceModel = this.resourcesWithGetOperations();
        const resourceCollectMap = new Map();
        const doc = this.innerDoc;
        for (const resourceNode of (0, jsonpath_1.nodes)(doc, "$.definitions.*")) {
            for (const arrayNode of (0, jsonpath_1.nodes)(resourceNode.value, "$..[?(@property === 'type' && @ === 'array')]^")) {
                const arrayObj = arrayNode.value;
                const items = arrayObj === null || arrayObj === void 0 ? void 0 : arrayObj.items;
                const res = this.stripDefinitionPath(items === null || items === void 0 ? void 0 : items.$ref);
                if (res &&
                    resourceModel.some((re) => re.modelName === res) &&
                    arrayNode.path.length === 3 &&
                    arrayNode.path[1] === "properties" &&
                    arrayNode.path[2] === "value") {
                    resourceCollectMap.set(res, resourceNode.path[2]);
                }
            }
        }
        return resourceCollectMap;
    }
    isPathBySubscription(path) {
        return !!path.match(this.SubscriptionsWideResourceRegEx);
    }
    isPathByResourceGroup(path) {
        return !!path.match(this.ResourceGroupWideResourceRegEx);
    }
    isPathOfExtensionResource(path) {
        return !!path.match(this.ExtensionResourceFullyQualifiedPathReg) || !!path.match(this.ExtensionResourceReg);
    }
    getResponseModelFromPath(path) {
        var _a, _b;
        const getOperation = this.getOperation(path);
        if (getOperation && getOperation.responses["200"]) {
            return this.enhancedSchema((_b = (_a = getOperation.responses) === null || _a === void 0 ? void 0 : _a["200"]) === null || _b === void 0 ? void 0 : _b.schema);
        }
        return undefined;
    }
    getResponseModelNameFromPath(path) {
        const re = this.resources.find((re) => re.operations.some((op) => op.apiPath === path));
        if (re) {
            return re.modelName;
        }
        return undefined;
    }
    getOperation(path, code = "get", doc) {
        var _a;
        let pathObj;
        if (doc) {
            pathObj = (_a = doc.paths) === null || _a === void 0 ? void 0 : _a[path];
        }
        else {
            const walker = new swagger_walker_1.SwaggerWalker(this.inventory);
            walker.warkAll([`$.[paths,x-ms-paths][${path}]`], (path, value, rootPath) => {
                pathObj = value;
            });
        }
        if (pathObj && pathObj[code]) {
            return pathObj[code];
        }
        return undefined;
    }
    getOperationIdFromPath(path, code = "get", doc) {
        const operation = this.getOperation(path, code, doc);
        return operation === null || operation === void 0 ? void 0 : operation.operationId;
    }
    isPathXmsPageable(path, code = "get", doc) {
        const operation = this.getOperation(path, code, doc);
        return !!(operation === null || operation === void 0 ? void 0 : operation["x-ms-pageable"]);
    }
    findOperation(path, code = "get") {
        const op = this.getOperationIdFromPath(path, code, this.innerDoc);
        if (op) {
            return op;
        }
        const references = this.inventory.referencesOf(this.specPath);
        for (const reference of Object.values(references)) {
            const op = this.getOperationIdFromPath(path, code, reference);
            if (op) {
                return op;
            }
        }
        return undefined;
    }
    verifyCollectionModel(collectionModel, childModelName) {
        if (collectionModel) {
            if (collectionModel.type === "array" && collectionModel.items) {
                const itemsRef = collectionModel.items.$ref;
                if (this.stripDefinitionPath(itemsRef) === childModelName) {
                    return true;
                }
            }
        }
        return false;
    }
    getProperty(schema, property) {
        return this.swaggerUtil.getProperty(schema, property);
    }
    getResourceProperties(resourceName) {
        const schema = this.getResourceByName(resourceName);
        if (schema) {
            return this.getProperties(schema);
        }
        return [];
    }
    getProperties(schema) {
        return swagger_workspace_1.Workspace.getProperties(schema, this.inventory);
    }
    getAttribute(schema, attr) {
        return this.swaggerUtil.getAttribute(schema, attr);
    }
    enhancedSchema(schema, specPath) {
        return {
            value: schema,
            file: specPath || this.specPath,
        };
    }
}
exports.ArmHelper = ArmHelper;
