// @microsoft.azure/openapi-validator-rulesets/dist/native/utilities/rules-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getResourceProvider = exports.isListOperationPath = exports.stringify = exports.nodes = exports.getResolvedSchemaByPath = exports.transformEnum = exports.isValidEnum = exports.isValidOperation = exports.resourceTypeMustCamelCase = exports.resourceProviderMustPascalCase = exports.getAllWordsFromPath = exports.getProviderNamespaceFromPath = exports.getProviderNamespace = exports.getAllResourceProvidersFromPath = exports.getResponseSchema = exports.getMostSuccessfulResponseKey = exports.getSuccessfulResponseSchema = void 0;
const jsonpath_plus_1 = require("jsonpath-plus");
const ref_helper_1 = require("./ref-helper");
const swagger_walker_1 = require("./swagger-walker");
const swagger_workspace_1 = require("./swagger-workspace");
const matchAll = require("string.prototype.matchall");
function getSuccessfulResponseSchema(node, doc, inventory) {
    if (!node.responses) {
        return undefined;
    }
    const responses = Object.keys(node.responses);
    const response = getMostSuccessfulResponseKey(responses);
    return getResponseSchema(node.responses[response], doc, inventory);
}
exports.getSuccessfulResponseSchema = getSuccessfulResponseSchema;
function getMostSuccessfulResponseKey(responses) {
    let response = "default";
    if (responses.includes("200")) {
        response = "200";
    }
    else {
        const twoHundreds = [];
        responses.forEach(function (value) {
            if (value.startsWith("2")) {
                twoHundreds.push(value);
            }
        });
        if (twoHundreds.length > 0) {
            response = twoHundreds[0];
        }
    }
    return response;
}
exports.getMostSuccessfulResponseKey = getMostSuccessfulResponseKey;
function getResponseSchema(response, doc, inventory) {
    let schema = response === null || response === void 0 ? void 0 : response.schema;
    if (schema === undefined || schema === null) {
        return;
    }
    if ("$ref" in schema) {
        schema = (0, ref_helper_1.crawlReference)(doc, schema, inventory);
        if (!schema) {
            return;
        }
    }
    return schema;
}
exports.getResponseSchema = getResponseSchema;
function getAllResourceProvidersFromPath(path) {
    const resourceProviderRegex = new RegExp(/providers\/([\w.]+)/, "g");
    return Array.from(matchAll(path, resourceProviderRegex), (m) => m[1]);
}
exports.getAllResourceProvidersFromPath = getAllResourceProvidersFromPath;
function getProviderNamespace(apiPath) {
    const matches = getAllResourceProvidersFromPath(apiPath);
    if (matches.length) {
        return matches.pop();
    }
    return undefined;
}
exports.getProviderNamespace = getProviderNamespace;
function getProviderNamespaceFromPath(filePath) {
    if (!filePath) {
        return undefined;
    }
    const resourceProviderRegex = new RegExp(/\/(Microsoft\.\w+)\//i, "g");
    const match = Array.from(matchAll(filePath.replace(/\\/g, "/"), resourceProviderRegex), (m) => m[1]);
    if (match) {
        return match[0];
    }
    return undefined;
}
exports.getProviderNamespaceFromPath = getProviderNamespaceFromPath;
function getAllWordsFromPath(path) {
    const wordRegex = new RegExp(/([\w.]+)/, "g");
    return Array.from(matchAll(path, wordRegex), (m) => m[1]);
}
exports.getAllWordsFromPath = getAllWordsFromPath;
function resourceProviderMustPascalCase(resourceProvider) {
    if (resourceProvider.length === 0) {
        return false;
    }
    const pascalCase = new RegExp(`^[A-Z][a-z0-9]+(.([A-Z]{1,3}[a-z0-9]+)+[A-Z]{0,2})+$`);
    return pascalCase.test(resourceProvider);
}
exports.resourceProviderMustPascalCase = resourceProviderMustPascalCase;
function resourceTypeMustCamelCase(resourceType) {
    if (resourceType.length === 0) {
        return true;
    }
    const pascalCase = new RegExp("^[a-z][a-z0-9]+([A-Z]+[a-z0-9]*)*$");
    return pascalCase.test(resourceType);
}
exports.resourceTypeMustCamelCase = resourceTypeMustCamelCase;
function isValidOperation(operation) {
    const validOperations = ["put", "get", "patch", "post", "head", "options", "delete"];
    return validOperations.indexOf(operation.toLowerCase()) !== -1;
}
exports.isValidOperation = isValidOperation;
function isValidEnum(node) {
    if (!node || !node.type || typeof node.type !== "string") {
        return false;
    }
    return ["boolean", "integer", "number", "string"].indexOf(node.type) !== -1 && Array.isArray(node.enum);
}
exports.isValidEnum = isValidEnum;
function transformEnum(type, enumEntries) {
    return enumEntries.map((v) => {
        if (v === null) {
            return "";
        }
        return v.toString();
    });
}
exports.transformEnum = transformEnum;
function getResolvedSchemaByPath(path, specPath, inventory) {
    const result = swagger_workspace_1.Workspace.jsonPath(path, inventory.getDocuments(specPath));
    if (result) {
        return swagger_workspace_1.Workspace.resolveRef({ value: result, file: specPath }, inventory);
    }
    return undefined;
}
exports.getResolvedSchemaByPath = getResolvedSchemaByPath;
function nodes(obj, pathExpression) {
    const result = (0, jsonpath_plus_1.JSONPath)({ json: obj, path: pathExpression, resultType: "all" });
    return result.map((v) => ({ path: jsonpath_plus_1.JSONPath.toPathArray(v.path), value: v.value, parent: v.parent }));
}
exports.nodes = nodes;
function stringify(path) {
    const pathWithRoot = ["$", ...path];
    return jsonpath_plus_1.JSONPath.toPathString(pathWithRoot);
}
exports.stringify = stringify;
function isListOperationPath(path) {
    if (path.includes(".")) {
        const splitNamespace = path.split(".");
        if (path.includes("/")) {
            const segments = splitNamespace[splitNamespace.length - 1].split("/");
            if (segments.length % 2 == 0) {
                return true;
            }
        }
    }
    return false;
}
exports.isListOperationPath = isListOperationPath;
function getResourceProvider(inventory) {
    const walker = new swagger_walker_1.SwaggerWalker(inventory);
    let result = [];
    walker.warkAll(["$.paths.*"], (path) => {
        const apiPath = path[2];
        if (result.length === 0) {
            result = [...getAllResourceProvidersFromPath(apiPath)];
        }
    });
    return result.length ? result.pop() || "" : "";
}
exports.getResourceProvider = getResourceProvider;
