// @microsoft.azure/openapi-validator-rulesets/dist/native/utilities/swagger-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SwaggerHelper = void 0;
const tslib_1 = require("tslib");
const json_schema_ref_parser_1 = tslib_1.__importDefault(require("@apidevtools/json-schema-ref-parser"));
const lodash_1 = tslib_1.__importDefault(require("lodash"));
const ref_helper_1 = require("./ref-helper");
const swagger_workspace_1 = require("./swagger-workspace");
class SwaggerHelper {
    constructor(root, specPath, inventory) {
        this.root = root;
        this.specPath = specPath;
        this.inventory = inventory;
    }
    getOperationIdFromPath(path, code = "get") {
        let pathObj = this.root.paths[path];
        if (!pathObj && this.root["x-ms-paths"]) {
            pathObj = this.root["x-ms-paths"][path];
        }
        if (pathObj && pathObj[code]) {
            return pathObj[code].operationId;
        }
    }
    getSpecPath() {
        return this.specPath;
    }
    getProperty(sourceModel, propertyName) {
        if (!sourceModel) {
            return undefined;
        }
        return swagger_workspace_1.Workspace.getProperty(sourceModel, propertyName, this.inventory);
    }
    resolveRef(sourceModel) {
        if (!sourceModel) {
            return undefined;
        }
        return swagger_workspace_1.Workspace.resolveRef(sourceModel, this.inventory);
    }
    getAttribute(sourceModel, attributeName) {
        if (!sourceModel) {
            return undefined;
        }
        return swagger_workspace_1.Workspace.getAttribute(sourceModel, attributeName, this.inventory);
    }
    async resolveSchema(schema, skipResolveExample = true) {
        if (!schema) {
            return schema;
        }
        schema = lodash_1.default.cloneDeep(schema);
        function isLocalRef(schema) {
            return schema.$ref && schema.$ref.startsWith("#/");
        }
        if (isLocalRef(schema)) {
            schema.$ref = this.specPath + schema.$ref;
        }
        const inventory = this.inventory;
        const resolveOption = {
            resolve: {
                file: {
                    canRead: true,
                    read(file) {
                        if (skipResolveExample && (0, ref_helper_1.isExample)(file.url)) {
                            return {};
                        }
                        return lodash_1.default.cloneDeep(inventory === null || inventory === void 0 ? void 0 : inventory.getDocuments(file.url));
                    },
                },
            },
        };
        return await json_schema_ref_parser_1.default.dereference(schema, resolveOption);
    }
}
exports.SwaggerHelper = SwaggerHelper;
