// @microsoft.azure/openapi-validator-rulesets/dist/native/utilities/swagger-workspace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Workspace = void 0;
const ref_helper_1 = require("./ref-helper");
var Workspace;
(function (Workspace) {
    function getSchemaByName(modelName, swaggerPath, inventory) {
        var _a;
        const root = inventory.getDocuments(swaggerPath);
        if (!root || !modelName) {
            return undefined;
        }
        return (_a = root === null || root === void 0 ? void 0 : root.definitions) === null || _a === void 0 ? void 0 : _a[modelName];
    }
    Workspace.getSchemaByName = getSchemaByName;
    function jsonPath(paths, document) {
        let root = document;
        for (const seg of paths) {
            root = root[seg];
            if (!root) {
                break;
            }
        }
        return root;
    }
    Workspace.jsonPath = jsonPath;
    function resolveRef(schema, inventory) {
        function getRef(refValue, swaggerPath) {
            const root = inventory.getDocuments(swaggerPath);
            if (refValue.startsWith("/")) {
                refValue = refValue.substring(1);
            }
            const segments = refValue.split("/");
            return jsonPath(segments, root);
        }
        let currentSpec = schema.file;
        let currentSchema = schema.value;
        while (currentSchema && currentSchema.$ref) {
            const slices = (0, ref_helper_1.parseJsonRef)(currentSchema.$ref);
            currentSpec = slices[0] || currentSpec;
            currentSchema = getRef(slices[1], currentSpec);
        }
        return {
            file: currentSpec,
            value: currentSchema,
        };
    }
    Workspace.resolveRef = resolveRef;
    function getProperty(schema, propertyName, inventory) {
        let source = schema;
        const visited = new Set();
        while (source.value && source.value.$ref && !visited.has(source.value)) {
            visited.add(source.value);
            source = resolveRef(source, inventory);
        }
        if (!source || !source.value) {
            return undefined;
        }
        const model = source.value;
        if (model.properties && model.properties[propertyName]) {
            return resolveRef(createEnhancedSchema(model.properties[propertyName], source.file), inventory);
        }
        if (model.allOf) {
            for (const element of model.allOf) {
                const property = getProperty({ file: source.file, value: element }, propertyName, inventory);
                if (property) {
                    return property;
                }
            }
        }
        return undefined;
    }
    Workspace.getProperty = getProperty;
    function getProperties(schema, inventory) {
        var _a, _b, _c;
        let source = schema;
        const visited = new Set();
        while (source.value && source.value.$ref && !visited.has(source.value)) {
            visited.add(source.value);
            source = resolveRef(source, inventory);
        }
        if (!source || !source.value) {
            return [];
        }
        let result = {};
        let model = source.value;
        const reference = (_c = (_b = (_a = model.properties) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.items) === null || _c === void 0 ? void 0 : _c.$ref;
        if ((reference === null || reference === void 0 ? void 0 : reference.includes("/common-types/")) && (reference === null || reference === void 0 ? void 0 : reference.includes("/types.json#/definitions/"))) {
            result["common-types"] = createEnhancedSchema("", "");
            return result;
        }
        else if (reference) {
            const referenceSchema = resolveRef(createEnhancedSchema(model.properties.value.items, source.file), inventory);
            if (referenceSchema && referenceSchema.value && (referenceSchema.value.properties || referenceSchema.value.allOf)) {
                model = referenceSchema.value;
            }
        }
        if (model.properties) {
            for (const propertyName of Object.keys(model.properties)) {
                result[propertyName] = createEnhancedSchema(model.properties[propertyName], source.file);
            }
        }
        if (model.allOf) {
            for (const element of model.allOf) {
                const properties = getProperties({ file: source.file, value: element }, inventory);
                if (properties) {
                    result = { ...properties, ...result };
                }
            }
        }
        return result;
    }
    Workspace.getProperties = getProperties;
    function getAttribute(schema, attributeName, inventory) {
        let source = schema;
        const visited = new Set();
        while (source.value && source.value.$ref && !visited.has(source.value)) {
            visited.add(source.value);
            source = resolveRef(source, inventory);
        }
        if (!source || !source.value) {
            return undefined;
        }
        const attribute = source.value[attributeName];
        if (attribute) {
            return createEnhancedSchema(attribute, source.file);
        }
        return undefined;
    }
    Workspace.getAttribute = getAttribute;
    function createEnhancedSchema(schema, specPath) {
        return {
            file: specPath,
            value: schema,
        };
    }
    Workspace.createEnhancedSchema = createEnhancedSchema;
})(Workspace = exports.Workspace || (exports.Workspace = {}));
