// oav/dist/lib/util/logging.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.log = void 0;
const fs = __importStar(require("fs"));
const os = __importStar(require("os"));
const path = __importStar(require("path"));
const winston = __importStar(require("winston"));
let logDir = path.resolve(os.homedir(), "oav_output");
let currentLogFile;
/*
 * Provides current time in custom format that will be used in naming log files. Example:
 * '20140820_151113'
 * @return {string} Current time in a custom string format
 */
function getTimeStamp() {
    // We pad each value so that sorted directory listings show the files in chronological order
    function pad(num) {
        return num < 10 ? "0" + num.toString() : num.toString();
    }
    const now = new Date();
    return (pad(now.getFullYear()) +
        pad(now.getMonth() + 1) +
        pad(now.getDate()) +
        "_" +
        pad(now.getHours()) +
        pad(now.getMinutes()) +
        pad(now.getSeconds()));
}
const customLogLevels = {
    off: 0,
    json: 1,
    error: 2,
    warn: 3,
    info: 4,
    verbose: 5,
    debug: 6,
    silly: 7,
};
exports.log = winston.createLogger({
    transports: [
        new winston.transports.Console({
            level: "warn",
            format: winston.format.combine(winston.format.colorize(), winston.format.prettyPrint()),
        }),
    ],
    levels: customLogLevels,
});
Object.defineProperties(exports.log, {
    consoleLogLevel: {
        enumerable: true,
        get() {
            const transport = this.transports.find((t) => t instanceof winston.transports.Console);
            return transport !== undefined ? transport.level : undefined;
        },
        set(level) {
            if (!level) {
                level = "warn";
            }
            const validLevels = Object.keys(customLogLevels);
            if (!validLevels.some((item) => item === level)) {
                throw new Error(`The logging level provided is "${level}". Valid values are: "${validLevels}".`);
            }
            const transport = this.transports.find((t) => t instanceof winston.transports.Console);
            if (transport !== undefined) {
                transport.level = level;
            }
        },
    },
    directory: {
        enumerable: true,
        get() {
            return logDir;
        },
        set(logDirectory) {
            if (!logDirectory || (logDirectory && typeof logDirectory.valueOf() !== "string")) {
                throw new Error('logDirectory cannot be null or undefined and must be of type "string".');
            }
            if (!fs.existsSync(logDirectory)) {
                fs.mkdirSync(logDirectory);
            }
            logDir = logDirectory;
        },
    },
    filepath: {
        enumerable: true,
        get() {
            if (!currentLogFile) {
                const filename = `validate_log_${getTimeStamp()}.log`;
                currentLogFile = path.join(this.directory, filename);
            }
            return currentLogFile;
        },
        set(logFilePath) {
            const self = this;
            if (!logFilePath || (logFilePath && typeof logFilePath.valueOf() !== "string")) {
                throw new Error("filepath cannot be null or undefined and must be of type string. It must be " +
                    "an absolute file path.");
            }
            currentLogFile = logFilePath;
            self.directory = path.dirname(logFilePath);
            if (!self.transports.some((t) => t instanceof winston.transports.File)) {
                self.add(new winston.transports.File({
                    level: "silly",
                    format: winston.format.prettyPrint(),
                    silent: false,
                    filename: logFilePath,
                }));
            }
        },
    },
});
