// oav/dist/lib/util/validationError.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.errorCodeToErrorMetadata = exports.allErrorConstants = void 0;
exports.serializeErrors = serializeErrors;
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
const openapi_tools_common_1 = require("@azure-tools/openapi-tools-common");
const jsonpath_plus_1 = require("jsonpath-plus");
const lodash_1 = __importDefault(require("lodash"));
const errorConstants = {
    INVALID_TYPE: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    INVALID_FORMAT: { severity: 0 /* Severity.Critical */, docUrl: "" },
    ENUM_MISMATCH: { severity: 0 /* Severity.Critical */, docUrl: "" },
    ENUM_CASE_MISMATCH: { severity: 1 /* Severity.Error */, docUrl: "" },
    PII_MISMATCH: { severity: 2 /* Severity.Warning */, docUrl: "" },
    NOT_PASSED: { severity: 0 /* Severity.Critical */, docUrl: "" },
    ARRAY_LENGTH_SHORT: { severity: 0 /* Severity.Critical */, docUrl: "" },
    ARRAY_LENGTH_LONG: { severity: 0 /* Severity.Critical */, docUrl: "" },
    ARRAY_UNIQUE: { severity: 0 /* Severity.Critical */, docUrl: "" },
    ARRAY_ADDITIONAL_ITEMS: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    MULTIPLE_OF: { severity: 0 /* Severity.Critical */, docUrl: "" },
    MINIMUM: { severity: 0 /* Severity.Critical */, docUrl: "" },
    MINIMUM_EXCLUSIVE: { severity: 0 /* Severity.Critical */, docUrl: "" },
    MAXIMUM: { severity: 0 /* Severity.Critical */, docUrl: "" },
    MAXIMUM_EXCLUSIVE: { severity: 0 /* Severity.Critical */, docUrl: "" },
    READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    UNRESOLVABLE_REFERENCE: { severity: 0 /* Severity.Critical */, docUrl: "" },
    SECRET_PROPERTY: { severity: 0 /* Severity.Critical */, docUrl: "" },
    WRITEONLY_PROPERTY_NOT_ALLOWED_IN_RESPONSE: { severity: 0 /* Severity.Critical */, docUrl: "" },
    OBJECT_PROPERTIES_MINIMUM: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    OBJECT_PROPERTIES_MAXIMUM: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    OBJECT_MISSING_REQUIRED_PROPERTY: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    MISSING_REQUIRED_PARAMETER: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    OBJECT_ADDITIONAL_PROPERTIES: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    OBJECT_DEPENDENCY_KEY: { severity: 2 /* Severity.Warning */, docUrl: "" },
    MIN_LENGTH: { severity: 0 /* Severity.Critical */, docUrl: "" },
    MAX_LENGTH: { severity: 0 /* Severity.Critical */, docUrl: "" },
    PATTERN: { severity: 0 /* Severity.Critical */, docUrl: "" },
    INVALID_RESPONSE_CODE: { severity: 0 /* Severity.Critical */, docUrl: "" },
    INVALID_CONTENT_TYPE: { severity: 1 /* Severity.Error */, docUrl: "" },
    DISCRIMINATOR_VALUE_NOT_FOUND: { severity: 0 /* Severity.Critical */, docUrl: "" },
    INVALID_RESPONSE_HEADER: { severity: 1 /* Severity.Error */, docUrl: "" },
    INVALID_RESPONSE_BODY: { severity: 0 /* Severity.Critical */, docUrl: "" },
    MISSING_RESOURCE_ID: { severity: 0 /* Severity.Critical */, docUrl: "" },
    LRO_RESPONSE_CODE: { severity: 0 /* Severity.Critical */, docUrl: "" },
    LRO_RESPONSE_HEADER: { severity: 0 /* Severity.Critical */, docUrl: "" },
};
const wrapperErrorConstants = {
    ANY_OF_MISSING: { severity: 0 /* Severity.Critical */, docUrl: "" },
    ONE_OF_MISSING: { severity: 0 /* Severity.Critical */, docUrl: "" },
    ONE_OF_MULTIPLE: { severity: 0 /* Severity.Critical */, docUrl: "" },
    MULTIPLE_OPERATIONS_FOUND: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    INVALID_RESPONSE_HEADER: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    INVALID_RESPONSE_BODY: { severity: 0 /* Severity.Critical */, docUrl: "" },
    INVALID_REQUEST_PARAMETER: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
};
const runtimeErrorConstants = {
    OPERATION_NOT_FOUND_IN_CACHE: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    OPERATION_NOT_FOUND_IN_CACHE_WITH_VERB: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    OPERATION_NOT_FOUND_IN_CACHE_WITH_API: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    OPERATION_NOT_FOUND_IN_CACHE_WITH_PROVIDER: {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    },
    INTERNAL_ERROR: { severity: 0 /* Severity.Critical */, docUrl: "" },
};
exports.allErrorConstants = {
    ...errorConstants,
    ...wrapperErrorConstants,
    ...runtimeErrorConstants,
};
/**
 * Gets the validation error metadata from an error code. If the code is unknown assume critical.
 */
const errorCodeToErrorMetadata = (code) => ({
    ...(exports.allErrorConstants[code] || {
        severity: 0 /* Severity.Critical */,
        docUrl: "",
    }),
    code,
});
exports.errorCodeToErrorMetadata = errorCodeToErrorMetadata;
/**
 * Serializes error tree
 */
function serializeErrors(node, path) {
    if (isLeaf(node)) {
        if (isTrueError(node)) {
            setPathProperties(node, path);
            return [node];
        }
        return [];
    }
    if (node.path) {
        // in this case the path will be set to the url instead of the path to the property
        if (node.code === "INVALID_REQUEST_PARAMETER" && node.in === "body") {
            node.path = [];
        }
        else if ((node.in === "query" || node.in === "path") &&
            node.path[0] === "paths" &&
            node.name) {
            // in this case we will want to normalize the path with the uri and the paramter name
            node.path = [node.path[1], node.name];
        }
        path = consolidatePath(path, node.path);
    }
    const serializedErrors = (0, openapi_tools_common_1.flatMap)(node.errors, (validationError) => serializeErrors(validationError, path)).toArray();
    const serializedInner = (0, openapi_tools_common_1.fold)(node.inner, (acc, validationError) => {
        const errs = serializeErrors(validationError, path);
        errs.forEach((err) => {
            const similarErr = acc.find((el) => areErrorsSimilar(err, el));
            if (similarErr && similarErr.path) {
                if (!similarErr.similarPaths) {
                    similarErr.similarPaths = [];
                }
                similarErr.similarPaths.push(err.path);
                if (!similarErr.similarJsonPaths) {
                    similarErr.similarJsonPaths = [];
                }
                similarErr.similarJsonPaths.push(err.jsonPath);
            }
            else {
                acc.push(err);
            }
        });
        return acc;
    }, new Array());
    if (isDiscriminatorError(node)) {
        setPathProperties(node, path);
        node.inner = serializedInner;
        return [node];
    }
    return [...serializedErrors, ...serializedInner];
}
/**
 * Sets the path and jsonPath properties on an error node.
 */
function setPathProperties(node, path) {
    if (!node.path) {
        return;
    }
    let nodePath = typeof node.path === "string" ? [node.path] : node.path;
    if (node.code === "OBJECT_MISSING_REQUIRED_PROPERTY" ||
        node.code === "OBJECT_ADDITIONAL_PROPERTIES") {
        // For multiple missing/additional properties , each node would only contain one param.
        if (node.params && node.params.length > 0) {
            nodePath = nodePath.concat(node.params[0]);
        }
    }
    const pathSegments = consolidatePath(path, nodePath);
    pathSegments.unshift("$");
    node.path = pathSegments.join("/");
    node.jsonPath = (pathSegments.length && jsonpath_plus_1.JSONPath.toPathString(pathSegments)) || "";
}
/**
 * Checks if two errors are the same except their path.
 */
function areErrorsSimilar(node1, node2) {
    if (node1.code !== node2.code ||
        node1.title !== node2.title ||
        node1.message !== node2.message ||
        !arePathsSimilar(node1.path, node2.path)) {
        return false;
    }
    if (!node1.inner && !node2.inner) {
        return true;
    }
    if (!node1.inner || !node2.inner || node1.inner.length !== node2.inner.length) {
        return false;
    }
    for (let i = 0; i < node1.inner.length; i++) {
        if (!areErrorsSimilar(node1.inner[i], node2.inner[i])) {
            return false;
        }
    }
    return true;
}
/**
 * Checks if paths differ only in indexes
 */
const arePathsSimilar = (path1, path2) => {
    if (path1 === path2) {
        return true;
    }
    if (path1 === undefined || path2 === undefined) {
        return false;
    }
    const p1 = Array.isArray(path1) ? path1 : path1.split("/");
    const p2 = Array.isArray(path2) ? path2 : path2.split("/");
    return lodash_1.default.xor(p1, p2).every((v) => Number.isInteger(+v));
};
const isDiscriminatorError = (node) => node.code === "ONE_OF_MISSING" && node.inner && node.inner.length > 0;
const isTrueError = (node) => 
// this is necessary to filter out extra errors coming from doing the ONE_OF transformation on
// the models to allow "null"
!(node.code === "INVALID_TYPE" && node.params && node.params[0] === "null");
const isLeaf = (node) => !node.errors && !node.inner;
/**
 * Unifies a suffix path with a root path.
 */
function consolidatePath(path, suffixPath) {
    let newSuffixIndex = 0;
    let overlapIndex = path.lastIndexOf(suffixPath[newSuffixIndex]);
    let previousIndex = overlapIndex;
    if (overlapIndex === -1) {
        return path.concat(suffixPath);
    }
    for (newSuffixIndex = 1; newSuffixIndex < suffixPath.length; ++newSuffixIndex) {
        previousIndex = overlapIndex;
        overlapIndex = path.lastIndexOf(suffixPath[newSuffixIndex]);
        if (overlapIndex === -1 || overlapIndex !== previousIndex + 1) {
            break;
        }
    }
    let newPath = [];
    if (newSuffixIndex === suffixPath.length) {
        // if all elements are contained in the existing path, nothing to do.
        newPath = path.slice(0);
    }
    else if (overlapIndex === -1 && previousIndex === path.length - 1) {
        // if we didn't find element at x in the previous path and element at x -1 is the last one in
        // the path, append everything from x
        newPath = path.concat(suffixPath.slice(newSuffixIndex));
    }
    else {
        // otherwise it is not contained at all, so concat everything.
        newPath = path.concat(suffixPath);
    }
    return newPath;
}
