// oav/dist/lib/apiScenario/bodyTransformer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BodyTransformer = void 0;
const inversify_1 = require("inversify");
const lodash_1 = require("lodash");
const jp = __importStar(require("json-pointer"));
const inversifyUtils_1 = require("../inversifyUtils");
const jsonUtils_1 = require("../util/jsonUtils");
const diffUtils_1 = require("./diffUtils");
let BodyTransformer = class BodyTransformer {
    constructor(validator) {
        this.validator = validator;
    }
    async resourceToRequest(resource, responseSchema) {
        const validateFn = await this.validator.compileAsync(responseSchema);
        // Readonly field cannot be set in response, so we could filter readonly fields
        const errors = validateFn({ isResponse: false, includeErrors: ["READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST"] }, resource);
        const result = (0, lodash_1.cloneDeep)(resource);
        for (const err of errors) {
            for (const jsonPath of err.jsonPathsInPayload) {
                const jsonPointer = (0, jsonUtils_1.jsonPathToPointer)(jsonPath);
                (0, diffUtils_1.jsonPatchApply)(result, [{ remove: jsonPointer }]);
            }
        }
        // console.log(body);
        // console.log(errors);
        // console.log(result);
        return result;
    }
    async resourceToResponse(resource, requestSchema) {
        const validateFn = await this.validator.compileAsync(requestSchema);
        // Writeonly field cannot be set in request, so we could filter writeonly fields
        const errors = validateFn({
            isResponse: false,
            includeErrors: ["WRITEONLY_PROPERTY_NOT_ALLOWED_IN_RESPONSE", "SECRET_PROPERTY"],
        }, resource);
        const result = (0, lodash_1.cloneDeep)(resource);
        for (const err of errors) {
            for (const jsonPath of err.jsonPathsInPayload) {
                const jsonPointer = (0, jsonUtils_1.jsonPathToPointer)(jsonPath);
                (0, diffUtils_1.jsonPatchApply)(result, [{ remove: jsonPointer }]);
            }
        }
        // console.log(body);
        // console.log(errors);
        // console.log(result);
        return result;
    }
    deepMerge(dst, src) {
        const inconsistentPaths = [];
        const result = this.innerDeepMerge(src, dst, [], inconsistentPaths);
        return {
            result,
            inconsistentWarningPaths: inconsistentPaths,
        };
    }
    innerDeepMerge(dst, src, path, inconsistentPaths) {
        if (typeof src !== typeof dst) {
            inconsistentPaths.push(jp.compile(path));
            return src !== null && src !== void 0 ? src : dst;
        }
        if (Array.isArray(dst) || Array.isArray(src)) {
            if (!Array.isArray(src) || !Array.isArray(dst)) {
                inconsistentPaths.push(jp.compile(path));
                return src;
            }
            let length = dst.length;
            if (dst.length !== src.length) {
                inconsistentPaths.push(jp.compile(path));
                if (src.length < dst.length) {
                    length = src.length;
                }
            }
            const result = new Array(length);
            for (let idx = 0; idx < length; idx++) {
                result[idx] = this.innerDeepMerge(dst[idx], src[idx], path.concat(idx.toString()), inconsistentPaths);
            }
            return result;
        }
        if (typeof dst === "object" && typeof src === "object") {
            const result = { ...dst };
            if (dst !== null && src !== null) {
                for (const key of Object.keys(dst)) {
                    if (key in src) {
                        result[key] = this.innerDeepMerge(dst[key], src[key], path.concat([key]), inconsistentPaths);
                    }
                }
            }
            if (src !== null && dst !== null) {
                for (const key of Object.keys(src)) {
                    if (!(key in dst)) {
                        result[key] = src[key];
                    }
                }
            }
            return result;
        }
        if (dst !== src) {
            inconsistentPaths.push(jp.compile(path));
        }
        return src;
    }
};
exports.BodyTransformer = BodyTransformer;
exports.BodyTransformer = BodyTransformer = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(inversifyUtils_1.TYPES.schemaValidator)),
    __metadata("design:paramtypes", [Object])
], BodyTransformer);
