// oav/dist/lib/apiScenario/diffUtils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getJsonPointer = exports.getObjValueFromPointer = exports.getJsonPatchDiff = exports.jsonPatchApply = void 0;
const fast_json_stable_stringify_1 = __importDefault(require("fast-json-stable-stringify"));
const jsonPointer = __importStar(require("json-pointer"));
const openapi_tools_common_1 = require("@azure-tools/openapi-tools-common");
const difflib_1 = require("difflib");
const rootName = "ROOT";
const getCtx = (obj, path) => {
    if (path === "/") {
        path = "";
    }
    const pathSegments = jsonPointer.parse(path);
    pathSegments.unshift(rootName);
    const propertyName = pathSegments.pop();
    const target = jsonPointer.get(obj, jsonPointer.compile(pathSegments));
    const result = {
        root: obj,
        obj: target,
        propertyName,
        arrIdx: -1,
    };
    if (Array.isArray(obj)) {
        result.arrIdx = parseInt(propertyName);
    }
    return result;
};
const patchAdd = ({ obj, propertyName, arrIdx }, op) => {
    if (Array.isArray(obj)) {
        obj.splice(arrIdx, 0, op.value);
    }
    else {
        obj[propertyName] = op.value;
    }
};
const patchRemove = ({ obj, propertyName, arrIdx }) => {
    if (Array.isArray(obj)) {
        obj.splice(arrIdx, 1);
    }
    else {
        delete obj[propertyName];
    }
};
const patchReplace = ({ obj, propertyName }, op) => {
    obj[propertyName] = op.value;
};
const patchCopy = ({ root, obj, propertyName }, op) => {
    const val = (0, openapi_tools_common_1.cloneDeep)(obj[propertyName]);
    jsonPointer.set(root, `/${rootName}${op.copy}`, val);
};
const patchMove = (ctx, op) => {
    const { propertyName, obj, root } = ctx;
    const val = obj[propertyName];
    patchRemove(ctx);
    jsonPointer.set(root, `/${rootName}${op.move}`, val);
};
const patchTest = ({ obj, propertyName }, op) => {
    const val = obj[propertyName];
    const factStr = (0, fast_json_stable_stringify_1.default)(val);
    const expectStr = (0, fast_json_stable_stringify_1.default)(op.value);
    if (factStr !== expectStr) {
        throw new Error(`JsonPatch Test failed for path: ${op.test}\nExpect: ${factStr}\nActual: ${factStr}`);
    }
};
const jsonPatchApplyOp = (obj, op) => {
    if ("add" in op) {
        return patchAdd(getCtx(obj, op.add), op);
    }
    if ("remove" in op) {
        return patchRemove(getCtx(obj, op.remove));
    }
    if ("replace" in op) {
        return patchReplace(getCtx(obj, op.replace), op);
    }
    if ("copy" in op) {
        return patchCopy(getCtx(obj, op.from), op);
    }
    if ("move" in op) {
        return patchMove(getCtx(obj, op.from), op);
    }
    if ("test" in op) {
        return patchTest(getCtx(obj, op.test), op);
    }
    throw new Error(`Unknown jsonPatchOp: ${JSON.stringify(op)}`);
};
const jsonPatchApply = (obj, ops) => {
    const rootObj = {
        [rootName]: obj,
    };
    for (const op of ops) {
        jsonPatchApplyOp(rootObj, op);
    }
    return rootObj[rootName];
};
exports.jsonPatchApply = jsonPatchApply;
const getJsonPatchDiff = (from, to, opts = {}) => {
    var _a;
    const patches = calcDiff(from, to, [], opts);
    if (opts.includeOldValue) {
        for (const patch of patches) {
            const p = patch;
            const oldPath = (_a = p.remove) !== null && _a !== void 0 ? _a : p.replace;
            if (oldPath !== undefined) {
                p.oldValue = (0, exports.getObjValueFromPointer)(from, oldPath);
            }
        }
    }
    return patches;
};
exports.getJsonPatchDiff = getJsonPatchDiff;
const getObjValueFromPointer = (obj, pointer) => {
    return jsonPointer.get(obj, pointer === "/" ? "" : pointer);
};
exports.getObjValueFromPointer = getObjValueFromPointer;
const calcDiff = (from, to, path, opts) => {
    if (from === to) {
        return [];
    }
    const replaceOp = {
        replace: (0, exports.getJsonPointer)(path),
        value: to,
    };
    const fromType = typeof from;
    const toType = typeof to;
    if (fromType !== toType || fromType !== "object" || from === null || to === null) {
        return [replaceOp];
    }
    const isFromArray = Array.isArray(from);
    const isToArray = Array.isArray(to);
    if (isFromArray !== isToArray) {
        return [replaceOp];
    }
    const diffOps = isFromArray
        ? calcArrayDiff(from, to, path, opts)
        : calcObjDiff(from, to, path, opts);
    if (diffOps.length === 0 || !opts.minimizeDiff) {
        return diffOps;
    }
    const diffLen = JSON.stringify(diffOps).length;
    const replaceLen = JSON.stringify(replaceOp).length + 4;
    return diffLen < replaceLen ? diffOps : [replaceOp];
};
const calcObjDiff = (from, to, path, opts) => {
    const result = [];
    for (const key of Object.keys(from)) {
        if (from[key] === undefined) {
            continue;
        }
        if (to[key] === undefined) {
            result.push({
                remove: (0, exports.getJsonPointer)(path, key),
            });
        }
        else {
            const diff = calcDiff(from[key], to[key], path.concat([key]), opts);
            if (diff.length > 0) {
                result.push(...diff);
            }
        }
    }
    for (const key of Object.keys(to)) {
        if (to[key] === undefined || from[key] !== undefined) {
            continue;
        }
        result.push({
            add: (0, exports.getJsonPointer)(path, key),
            value: to[key],
        });
    }
    return result;
};
const calcArrayDiff = (from, to, path, opts) => {
    let matchSeq = calcArrayDiffWithIndex(from, to);
    let isKeyMatch = true;
    if (matchSeq === undefined) {
        const matcher = new difflib_1.SequenceMatcher(null, from, to);
        matchSeq = matcher.getOpcodes();
        isKeyMatch = false;
    }
    const addReplaceOps = [];
    const removeOps = [];
    for (const [op, i0, i1, j0, j1] of matchSeq) {
        switch (op) {
            case "equal":
                if (isKeyMatch) {
                    for (let ix = i0, jx = j0; ix < i1 && jx < j1; ++ix, ++jx) {
                        const diff = calcDiff(from[ix], to[jx], path.concat([jx.toString()]), opts);
                        addReplaceOps.push(...diff);
                    }
                }
                break;
            case "insert":
                for (let idx = j0; idx < j1; ++idx) {
                    addReplaceOps.push({
                        add: (0, exports.getJsonPointer)(path, idx.toString()),
                        value: to[idx],
                    });
                }
                break;
            case "delete":
                for (let idx = i0; idx < i1; ++idx) {
                    removeOps.push({
                        remove: (0, exports.getJsonPointer)(path.concat([idx.toString()])),
                    });
                }
                break;
            case "replace": {
                let ix = i0;
                let jx = j0;
                for (; ix < i1 && jx < j1; ++ix, ++jx) {
                    if (isKeyMatch) {
                        addReplaceOps.push({
                            replace: (0, exports.getJsonPointer)(path, jx.toString()),
                            value: to[jx],
                        });
                    }
                    else {
                        const diff = calcDiff(from[ix], to[jx], path.concat([jx.toString()]), opts);
                        addReplaceOps.push(...diff);
                    }
                }
                for (; ix < i1; ++ix) {
                    removeOps.push({
                        remove: (0, exports.getJsonPointer)(path, ix.toString()),
                    });
                }
                for (; jx < j1; ++jx) {
                    addReplaceOps.push({
                        add: (0, exports.getJsonPointer)(path, jx.toString()),
                        value: to[jx],
                    });
                }
            }
        }
    }
    return removeOps.reverse().concat(addReplaceOps);
};
const calcArrayDiffWithIndex = (from, to) => {
    const fromKeys = new Array(from.length);
    const toKeys = new Array(to.length);
    let hasKey = false;
    for (let idx = 0; idx < from.length; ++idx) {
        const key = getObjKey(from[idx]);
        if (key !== undefined) {
            hasKey = true;
        }
        fromKeys[idx] = key;
    }
    for (let idx = 0; idx < to.length; ++idx) {
        const key = getObjKey(to[idx]);
        if (key !== undefined) {
            hasKey = true;
        }
        toKeys[idx] = key;
    }
    if (!hasKey) {
        return undefined;
    }
    const matcher = new difflib_1.SequenceMatcher(null, fromKeys, toKeys);
    const matchSeq = matcher.getOpcodes();
    return matchSeq;
};
const getObjKey = (item) => {
    var _a, _b;
    if (item === undefined || item === null || typeof item !== "object") {
        return undefined;
    }
    return (_b = (_a = item.id) !== null && _a !== void 0 ? _a : item.name) !== null && _b !== void 0 ? _b : probeObjectKey(item);
};
const probeObjectKey = (item) => {
    for (const key of Object.keys(item)) {
        if (typeof item[key] !== "string") {
            continue;
        }
        const lowerKey = key.toLowerCase();
        if (lowerKey.endsWith("id") || lowerKey.endsWith("name")) {
            return item[key];
        }
    }
    return undefined;
};
const getJsonPointer = (input, additional) => {
    let result = jsonPointer.compile(input);
    if (additional !== undefined) {
        result = `${result}/${jsonPointer.escape(additional)}`;
    }
    if (result === "") {
        result = "/";
    }
    return result;
};
exports.getJsonPointer = getJsonPointer;
