// @azure-tools/openapi-tools-common/dist/tests/json-tests.js
"use strict";
// tslint:disable:no-null-keyword
// tslint:disable:no-magic-numbers
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
const index = __importStar(require("../libs/json"));
describe("isObject", () => {
    it("object", () => {
        const x = {};
        assert.strictEqual(index.isObject(x), true);
    });
    it("array", () => {
        const x = [];
        assert.strictEqual(index.isObject(x), false);
    });
    it("primitive", () => {
        const x = 5;
        assert.strictEqual(index.isObject(x), false);
    });
    it("null", () => {
        const x = null;
        assert.strictEqual(index.isObject(x), false);
    });
});
describe("isPrimitive", () => {
    it("primitive", () => {
        const x = 5;
        assert.strictEqual(index.isPrimitive(x), true);
    });
    it("null", () => {
        const x = null;
        assert.strictEqual(index.isPrimitive(x), true);
    });
    it("not primitive", () => {
        const x = {};
        assert.strictEqual(index.isPrimitive(x), false);
    });
});
describe("visit", () => {
    it("string", () => {
        const x = "str";
        index.visit(x, {
            asNull: () => { assert.fail(); },
            asBoolean: () => { assert.fail(); },
            asString: (v) => { assert.strictEqual(v, "str"); },
            asNumber: () => { assert.fail(); },
            asArray: () => { assert.fail(); },
            asObject: () => { assert.fail(); },
        });
    });
    it("null", () => {
        const x = null;
        index.visit(x, {
            asNull: () => { },
            asBoolean: () => { assert.fail(); },
            asString: () => { assert.fail(); },
            asNumber: () => { assert.fail(); },
            asArray: () => { assert.fail(); },
            asObject: () => { assert.fail(); },
        });
    });
    it("boolean", () => {
        const x = true;
        index.visit(x, {
            asNull: () => { assert.fail(); },
            asBoolean: v => { assert.strictEqual(v, true); },
            asString: () => { assert.fail(); },
            asNumber: () => { assert.fail(); },
            asArray: () => { assert.fail(); },
            asObject: () => { assert.fail(); },
        });
    });
    it("number", () => {
        const x = 5;
        index.visit(x, {
            asNull: () => { assert.fail(); },
            asBoolean: () => { assert.fail(); },
            asString: () => { assert.fail(); },
            asNumber: v => { assert.strictEqual(v, 5); },
            asArray: () => { assert.fail(); },
            asObject: () => { assert.fail(); },
        });
    });
    it("array", () => {
        const x = [76];
        index.visit(x, {
            asNull: () => { assert.fail(); },
            asBoolean: () => { assert.fail(); },
            asString: () => { assert.fail(); },
            asNumber: () => { assert.fail(); },
            asArray: v => { assert.deepStrictEqual(v, [76]); },
            asObject: () => { assert.fail(); },
        });
    });
    it("object", () => {
        const x = {};
        index.visit(x, {
            asNull: () => { assert.fail(); },
            asBoolean: () => { assert.fail(); },
            asString: () => { assert.fail(); },
            asNumber: () => { assert.fail(); },
            asArray: () => { assert.fail(); },
            asObject: v => { assert.deepStrictEqual(v, {}); },
        });
    });
});
describe("typeOf", () => {
    it("object", () => {
        const x = {};
        assert.strictEqual(index.typeOf(x), "object");
    });
    it("null", () => {
        assert.strictEqual(index.typeOf(null), "null");
    });
    it("string", () => {
        assert.strictEqual(index.typeOf("ssss"), "string");
    });
    it("number", () => {
        assert.strictEqual(index.typeOf(4), "number");
    });
    it("boolean", () => {
        assert.strictEqual(index.typeOf(false), "boolean");
    });
    it("array", () => {
        assert.strictEqual(index.typeOf([]), "array");
    });
});
describe("sealed object", () => {
    const x = { a: 54 };
    const j = x;
    assert.deepStrictEqual(j, { a: 54 });
});
