// @inversifyjs/core/lib/cjs/common/models/__mocks__/OneToManyMapStar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OneToManyMapStar = void 0;
const vitest_1 = require("vitest");
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const addMock = vitest_1.vitest.fn();
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const cloneMock = vitest_1.vitest.fn();
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const getMock = vitest_1.vitest.fn();
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const getAllKeysMock = vitest_1.vitest.fn();
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const removeByRelationMock = vitest_1.vitest.fn();
function castMock(mock) {
    return mock;
}
class OneToManyMapStar {
    add;
    clone;
    get;
    getAllKeys;
    removeByRelation;
    constructor() {
        this.add = castMock(addMock);
        this.clone = castMock(cloneMock);
        this.get = castMock(getMock);
        this.getAllKeys = castMock(getAllKeysMock);
        this.removeByRelation = castMock(removeByRelationMock);
    }
}
exports.OneToManyMapStar = OneToManyMapStar;
