// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/AllResourcesMustHaveGetOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllResourcesMustHaveGetOperation = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const arm_helper_1 = require("../utilities/arm-helper");
exports.AllResourcesMustHaveGetOperation = "AllResourcesMustHaveGetOperation";
openapi_validator_core_1.rules.push({
    id: "R4014",
    name: exports.AllResourcesMustHaveGetOperation,
    severity: "warning",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.composed,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$",
    *run(doc, node, path, ctx) {
        const msg = 'The resource "{0}" does not have get operation, please add it.';
        const utils = new arm_helper_1.ArmHelper(doc, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
        const allResources = utils.resourcesWithPutPatchOperations();
        const allResourcesHavingGetOperation = utils.resourcesWithGetOperations();
        for (const resource of allResources) {
            if (!utils.containsDiscriminator(resource.modelName) &&
                !allResourcesHavingGetOperation.some((re) => re.modelName === resource.modelName)) {
                yield {
                    message: msg.replace("{0}", resource.modelName),
                    location: ["$", "definitions", resource.modelName],
                };
            }
        }
    },
});
