// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/AzureResourceTagsSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureResourceTagsSchema = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const arm_helper_1 = require("../utilities/arm-helper");
exports.AzureResourceTagsSchema = "AzureResourceTagsSchema";
openapi_validator_core_1.rules.push({
    id: "R4034",
    name: exports.AzureResourceTagsSchema,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$",
    *run(doc, node, path, ctx) {
        const msg = 'The property tags in the resource "{0}" does not conform to the common type definition.';
        const utils = new arm_helper_1.ArmHelper(doc, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
        const allResources = utils.getAllResourceNames();
        for (const resourceName of allResources) {
            const resource = utils === null || utils === void 0 ? void 0 : utils.getResourceByName(resourceName);
            const tagSchema = utils === null || utils === void 0 ? void 0 : utils.getProperty(resource, "tags");
            if (tagSchema && (!tagSchema.value.additionalProperties || tagSchema.value.additionalProperties.type !== "string")) {
                yield {
                    message: msg.replace("{0}", resourceName),
                    location: ["$", "definitions", resourceName],
                };
            }
        }
    },
});
