// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/ControlCharactersAreNotAllowed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ControlCharactersAreNotAllowed = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
exports.ControlCharactersAreNotAllowed = "ControlCharactersAreNotAllowed";
openapi_validator_core_1.rules.push({
    id: "R2006",
    name: exports.ControlCharactersAreNotAllowed,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$..*",
    *run(doc, node, path) {
        const msg = "May not contain control characters: ";
        if (typeof node === "string") {
            const nodeValue = node;
            const controlChars = nodeValue.split("").filter((ch) => ch < " " && ch !== "\t" && ch !== "\n" && ch !== "\r");
            if (controlChars.length > 0) {
                yield { message: `${msg} Characters:'${controlChars}' in:'${nodeValue}'`, location: path };
            }
        }
    },
});
