// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/DefaultErrorResponseSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultErrorResponseSchema = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const swagger_workspace_1 = require("../utilities/swagger-workspace");
exports.DefaultErrorResponseSchema = "DefaultErrorResponseSchema";
openapi_validator_core_1.rules.push({
    id: "R4007",
    name: exports.DefaultErrorResponseSchema,
    severity: "error",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$.paths.*.*.responses",
    *run(doc, node, path, ctx) {
        const msg = "the default error response schema does not correspond to the schema documented at https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-details.md#error-response-content.";
        const response = node;
        if (response.default) {
            const paths = path.concat(["default"]);
            if (response.default.schema) {
                const schema = swagger_workspace_1.Workspace.jsonPath(paths.concat("schema"), doc);
                if (schema) {
                    const errorDefinition = swagger_workspace_1.Workspace.getProperty({ file: ctx === null || ctx === void 0 ? void 0 : ctx.specPath, value: schema }, "error", ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
                    if (errorDefinition && errorDefinition.value) {
                        const code = swagger_workspace_1.Workspace.getProperty(errorDefinition, "code", ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
                        const message = swagger_workspace_1.Workspace.getProperty(errorDefinition, "message", ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
                        if (code && message) {
                            return;
                        }
                    }
                }
            }
            yield { message: `${msg}`, location: paths };
        }
    },
});
