// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/DeleteOperationResponses.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteOperationResponses = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
exports.DeleteOperationResponses = "DeleteOperationResponses";
openapi_validator_core_1.rules.push({
    id: "R4011",
    name: exports.DeleteOperationResponses,
    severity: "error",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$.paths.*.delete.responses",
    *run(doc, node, path) {
        if (node["202"] || node["x-ms-long-running-operation"]) {
            return;
        }
        if (!node["200"] || !node["204"] || !Object.keys(node["200"]) || !Object.keys(node["204"])) {
            yield {
                message: `The delete operation is defined without a 200 or 204 error response implementation,please add it.'`,
                location: path
            };
        }
    }
});
