// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/DeprecatedXmsCodeGenerationSetting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeprecatedXmsCodeGenerationSetting = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
exports.DeprecatedXmsCodeGenerationSetting = "DeprecatedXmsCodeGenerationSetting";
openapi_validator_core_1.rules.push({
    id: "R4006",
    name: exports.DeprecatedXmsCodeGenerationSetting,
    severity: "warning",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm | openapi_validator_core_1.OpenApiTypes.dataplane,
    appliesTo_JsonQuery: '$.info["x-ms-code-generation-settings"]',
    *run(doc, node, path) {
        if (node) {
            yield {
                message: `The x-ms-code-generation-setting extension is being deprecated. Please remove it and move settings to readme file for code generation.`,
                location: path
            };
        }
    }
});
