// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/DescriptionMustNotBeNodeName.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DescriptionMustNotBeNodeName = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
exports.DescriptionMustNotBeNodeName = "DescriptionMustNotBeNodeName";
openapi_validator_core_1.rules.push({
    id: "R3011",
    name: exports.DescriptionMustNotBeNodeName,
    severity: "error",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$..*[?(@property === 'description')]^",
    *run(doc, node, path) {
        const msg = "Description must not match the name of the node it is supposed to describe.";
        if (typeof node.description !== "string") {
            return;
        }
        const nodeName = path[path.length - 1];
        const description = TrimDescription(node.description);
        if ("name" in node && typeof node.name === "string" && TrimDescription(node.name.toLowerCase()) === description) {
            yield { message: `${msg} Node name:'${node.name}' Description:'${node.description}'`, location: path.concat(["description"]) };
        }
        else if (typeof nodeName === "string" && TrimDescription(nodeName.toLowerCase()) === description) {
            yield { message: `${msg} Node name:'${nodeName}' Description:'${node.description}'`, location: path.concat(["description"]) };
        }
        else if (description === "description") {
            yield { message: `${msg} Node name:'description' Description:'${node.description}'`, location: path.concat(["description"]) };
        }
    },
});
function TrimDescription(description) {
    return description.trim().replace(/\./g, "").toLowerCase();
}
