// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/EnumMustNotHaveEmptyValue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumMustNotHaveEmptyValue = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
exports.EnumMustNotHaveEmptyValue = "EnumMustNotHaveEmptyValue";
const rules_helper_1 = require("../utilities/rules-helper");
openapi_validator_core_1.rules.push({
    id: "R3029",
    name: exports.EnumMustNotHaveEmptyValue,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm | openapi_validator_core_1.OpenApiTypes.dataplane,
    appliesTo_JsonQuery: "$..*[?(@property === 'enum')]^",
    *run(doc, node, path) {
        const msg = `Enum value must not contain empty value.`;
        if (path.indexOf("x-ms-examples") === -1 && (0, rules_helper_1.isValidEnum)(node)) {
            const enumList = (0, rules_helper_1.transformEnum)(node.type, node.enum);
            if (enumList.some((value) => value.trim().length === 0)) {
                yield { message: `${msg}`, location: path };
            }
        }
    }
});
