// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/EnumMustRespectType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumMustRespectType = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const rules_helper_1 = require("../utilities/rules-helper");
exports.EnumMustRespectType = "EnumMustRespectType";
openapi_validator_core_1.rules.push({
    id: "R4040",
    name: exports.EnumMustRespectType,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm | openapi_validator_core_1.OpenApiTypes.dataplane,
    appliesTo_JsonQuery: ['$.parameters..[?(@property === "enum")]^', '$.definitions..[?(@property === "enum")]^'],
    *run(doc, node, path) {
        const msg = `Enum values should respect the type.`;
        if (node.enum && (0, rules_helper_1.isValidEnum)(node)) {
            if (node.enum.some((value) => {
                if (node.type === "integer") {
                    return !Number.isInteger(value);
                }
                return typeof value !== node.type;
            })) {
                yield { message: `${msg}`, location: path };
            }
        }
    }
});
