// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/EnumUniqueValue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumUniqueValue = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const rules_helper_1 = require("../utilities/rules-helper");
exports.EnumUniqueValue = "EnumUniqueValue";
openapi_validator_core_1.rules.push({
    id: "R3024",
    name: exports.EnumUniqueValue,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm | openapi_validator_core_1.OpenApiTypes.dataplane,
    appliesTo_JsonQuery: "$..*[?(@property === 'enum')]^",
    *run(doc, node, path) {
        const msg = `Enum must not contain duplicated value (case insentive).`;
        if (node.enum && path.indexOf("x-ms-examples") === -1 && (0, rules_helper_1.isValidEnum)(node)) {
            const enumList = (0, rules_helper_1.transformEnum)(node.type, node.enum);
            const caseInsensitiveSet = new Set();
            if (enumList.some((value) => {
                if (caseInsensitiveSet.has(value.toLowerCase())) {
                    return true;
                }
                caseInsensitiveSet.add(value.toLowerCase());
                return false;
            })) {
                yield { message: `${msg}`, location: path };
            }
        }
    }
});
