// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/EnumMustHaveType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumMustHaveType = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const rules_helper_1 = require("../utilities/rules-helper");
exports.EnumMustHaveType = "EnumMustHaveType";
openapi_validator_core_1.rules.push({
    id: "R3015",
    name: exports.EnumMustHaveType,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm | openapi_validator_core_1.OpenApiTypes.dataplane,
    appliesTo_JsonQuery: "$..*[?(@property==='enum')]^",
    *run(doc, node, path) {
        const msg = `Enum must define its type and "object" type is not allowed due to Autorest refuse to parse it.`;
        if (path.indexOf("x-ms-examples") === -1) {
            if (!(0, rules_helper_1.isValidEnum)(node)) {
                yield { message: `${msg}`, location: path };
            }
        }
    }
});
