// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/GetCollectionResponseSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetCollectionResponseSchema = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const arm_helper_1 = require("../utilities/arm-helper");
exports.GetCollectionResponseSchema = "GetCollectionResponseSchema";
openapi_validator_core_1.rules.push({
    id: "R4019",
    name: exports.GetCollectionResponseSchema,
    severity: "error",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.composed,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$",
    *run(doc, node, path, ctx) {
        const msg = 'The response in the GET collection operation "{0}" does not match the response definition in the individual GET  operation "{1}" .';
        const utils = new arm_helper_1.ArmHelper(doc, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
        const allCollectionPath = utils.getCollectionApiInfo();
        for (const collection of allCollectionPath) {
            let hasMatched = false;
            const resource = utils.getResourceByName(collection.modelName);
            const modelValue = utils === null || utils === void 0 ? void 0 : utils.getProperty(resource, "value");
            if (modelValue) {
                hasMatched = utils.verifyCollectionModel(modelValue.value, collection.childModelName);
            }
            if (!hasMatched) {
                const collectionOperationId = utils.getOperationIdFromPath(collection.collectionGetPath[0]);
                const specificOperationId = utils.getOperationIdFromPath(collection.specificGetPath[0]);
                yield {
                    message: msg.replace("{0}", collectionOperationId).replace("{1}", specificOperationId),
                    location: ["$", "paths", collection.collectionGetPath[0]],
                };
            }
        }
    },
});
