// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/IntegerTypeMustHaveFormat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntegerTypeMustHaveFormat = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
exports.IntegerTypeMustHaveFormat = "IntegerTypeMustHaveFormat";
openapi_validator_core_1.rules.push({
    id: "R4013",
    name: exports.IntegerTypeMustHaveFormat,
    severity: "error",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm | openapi_validator_core_1.OpenApiTypes.dataplane,
    appliesTo_JsonQuery: "$..*[?(@property === 'type' && @ === 'integer')]^",
    *run(doc, node, path) {
        if (path.indexOf("x-ms-examples") === -1) {
            const msg = `The integer type does not have a format, please add it.`;
            const formats = ["int32", "int64"];
            if (!node.format || formats.indexOf(node.format) === -1) {
                yield { message: `${msg}`, location: path };
            }
        }
    }
});
