// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/LicenseHeaderMustNotBeSpecified.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LicenseHeaderMustNotBeSpecified = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
exports.LicenseHeaderMustNotBeSpecified = "LicenseHeaderMustNotBeSpecified";
openapi_validator_core_1.rules.push({
    id: "R2065",
    name: exports.LicenseHeaderMustNotBeSpecified,
    severity: "warning",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm | openapi_validator_core_1.OpenApiTypes.dataplane,
    appliesTo_JsonQuery: "$..info['x-ms-code-generation-settings']",
    *run(doc, node, path) {
        const msg = `License header must not be specified inside x-ms-code-generation-settings. This is different for different sdks generated and is passed via command line/config file when generating the sdk.`;
        if (node.header !== undefined) {
            yield { message: `${msg}`, location: path };
        }
    }
});
