// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/MissingTypeObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MissingTypeObject = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
exports.MissingTypeObject = "MissingTypeObject";
openapi_validator_core_1.rules.push({
    id: "R4037",
    name: exports.MissingTypeObject,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm | openapi_validator_core_1.OpenApiTypes.dataplane,
    appliesTo_JsonQuery: "$.definitions.*",
    *run(doc, node, path) {
        const msg = `The schema '{0}' is considered an object but without a 'type:object', please add the missing 'type:object'.`;
        const isMissingTypeObject = (node) => {
            return node && (node.properties || node.additionalProperties) && !node.type;
        };
        const keys = [];
        keys.push({ node, path });
        while (true) {
            const tail = keys.pop();
            if (!tail) {
                break;
            }
            const node = tail.node;
            const path = tail.path;
            if (isMissingTypeObject(node)) {
                yield {
                    message: msg.replace(`{0}`, path[path.length - 1].toString()),
                    location: path
                };
            }
            if (node.properties) {
                Object.keys(node.properties).forEach(p => keys.push({ node: node.properties[p], path: path.concat(["properties", p]) }));
            }
            if (node.additionalProperties) {
                keys.push({ node: node.additionalProperties, path: path.concat(["additionalProperties"]) });
            }
            if (node.type === "array" && node.items) {
                keys.push({ node: node.items, path: path.concat(["items"]) });
            }
            if (node.allOf && Array.isArray(node.allOf)) {
                node.allOf.forEach((element, index) => {
                    keys.push({ node: element, path: path.concat(["allOf", index]) });
                });
            }
        }
    }
});
