// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/MissingXmsErrorResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MissingXmsErrorResponse = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
exports.MissingXmsErrorResponse = "MissingXmsErrorResponse";
openapi_validator_core_1.rules.push({
    id: "R4032",
    name: exports.MissingXmsErrorResponse,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$.paths.*.*.responses.*",
    *run(doc, node, path) {
        const msg = "Response code {0} is defined without a x-ms-error-response.";
        const httpMethod = path[path.length - 3].toString();
        const errorCode = path[path.length - 1].toString();
        if (httpMethod === "head" && errorCode === "404") {
            return;
        }
        if (errorCode.startsWith("4") || errorCode.startsWith("5")) {
            if (!node["x-ms-error-response"]) {
                yield { message: `${msg}`.replace("{0}", errorCode), location: path };
            }
        }
    }
});
