// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/NestedResourcesMustHaveListOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NestedResourcesMustHaveListOperation = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const arm_helper_1 = require("../utilities/arm-helper");
exports.NestedResourcesMustHaveListOperation = "NestedResourcesMustHaveListOperation";
openapi_validator_core_1.rules.push({
    id: "R4015",
    name: exports.NestedResourcesMustHaveListOperation,
    severity: "error",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.composed,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    *run(doc, node, path, ctx) {
        const msg = 'The nested resource "{0}" does not have list operation, please add it.';
        const utils = new arm_helper_1.ArmHelper(doc, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
        const nestedResource = utils.getAllNestedResources();
        const allCollectionApis = utils.getCollectionApiInfo();
        for (const resource of nestedResource) {
            const hasMatched = allCollectionApis.some((collection) => resource === collection.childModelName);
            if (!hasMatched) {
                yield {
                    message: msg.replace("{0}", resource),
                    location: ["$", "definitions", resource],
                };
            }
        }
    },
});
