// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/OperationIdRequired.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OperationIdRequired = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const rules_helper_1 = require("../utilities/rules-helper");
exports.OperationIdRequired = "OperationIdRequired";
openapi_validator_core_1.rules.push({
    id: "R4004",
    name: exports.OperationIdRequired,
    severity: "error",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$.paths",
    *run(doc, node, path) {
        if (node) {
            for (const pathKey in node) {
                for (const op in node[pathKey]) {
                    if (!(0, rules_helper_1.isValidOperation)(op)) {
                        continue;
                    }
                    if (!node[pathKey][op].operationId) {
                        yield {
                            message: `Missing operationId in path:'${pathKey}', operation:'${op}', consider adding the operationId .`,
                            location: path.concat([pathKey, op])
                        };
                    }
                }
            }
        }
    }
});
