// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/PageableOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PageableOperation = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const rules_helper_1 = require("../utilities/rules-helper");
exports.PageableOperation = "PageableOperation";
openapi_validator_core_1.rules.push({
    id: "R2029",
    name: exports.PageableOperation,
    severity: "warning",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm | openapi_validator_core_1.OpenApiTypes.dataplane,
    appliesTo_JsonQuery: "$.paths.*[?(@property==='get')]^",
    *run(doc, node, path, ctx) {
        var _a;
        const operations = Object.keys(node);
        const getKey = operations.find((key) => {
            return key.toLowerCase() === "get";
        });
        if (!getKey) {
            return;
        }
        const schemaProperties = (_a = (0, rules_helper_1.getSuccessfulResponseSchema)(node[getKey], doc, ctx === null || ctx === void 0 ? void 0 : ctx.inventory)) === null || _a === void 0 ? void 0 : _a.properties;
        function hasArrayProperty(schema) {
            let arrayPresent = false;
            Object.keys(schema).forEach(function (key) {
                if (schema[key].type === "array") {
                    arrayPresent = true;
                }
            });
            return arrayPresent;
        }
        if (schemaProperties !== undefined && schemaProperties != null && Object.keys(schemaProperties).length <= 3) {
            if (hasArrayProperty(schemaProperties) && !("x-ms-pageable" in node[getKey])) {
                yield {
                    message: `Based on the response model schema, operation '${node[getKey].operationId}' might be pageable. Consider adding the x-ms-pageable extension.`,
                    location: path.concat(getKey),
                };
            }
        }
    },
});
