// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/ParametersOrder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParametersOrder = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const swagger_workspace_1 = require("../utilities/swagger-workspace");
exports.ParametersOrder = "ParametersOrder";
function getParametersFromPath(apiPath) {
    const parameters = apiPath
        .split("/")
        .filter((x) => x.startsWith("{") && x.endsWith("}"))
        .map((x) => x.slice(1, -1));
    return parameters;
}
function getPathMethodParameter(schema, ctx) {
    var _a;
    if (!schema.$ref && schema.in === "path") {
        return schema.name;
    }
    const enhancedSchema = { file: ctx.specPath, value: schema };
    const resolvedParameter = (_a = swagger_workspace_1.Workspace.resolveRef(enhancedSchema, ctx === null || ctx === void 0 ? void 0 : ctx.inventory)) === null || _a === void 0 ? void 0 : _a.value;
    if (resolvedParameter && resolvedParameter.in === "path" && resolvedParameter["x-ms-parameter-location"] === "method") {
        return resolvedParameter.name;
    }
    return "";
}
openapi_validator_core_1.rules.push({
    id: "R4039",
    name: exports.ParametersOrder,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$.paths",
    *run(doc, node, path, ctx) {
        const msg = `The parameters:{0} should be kept in the same order as they present in the path.`;
        for (const apiPath of Object.keys(node)) {
            const commonParameters = node[apiPath].parameters || [];
            const httpMethods = Object.keys(node[apiPath]).filter((k) => k.toLowerCase() !== "parameters");
            for (const method of httpMethods) {
                const resolvedPathParameters = commonParameters
                    .concat(node[apiPath][method].parameters || [])
                    .map((x) => getPathMethodParameter(x, ctx))
                    .filter((x) => x);
                const parametersInPath = getParametersFromPath(apiPath).filter((p) => resolvedPathParameters.includes(p));
                if (parametersInPath.some((value, index) => index < resolvedPathParameters.length && resolvedPathParameters[index] !== value)) {
                    yield { message: msg.replace("{0}", resolvedPathParameters.join(",")), location: path.concat(apiPath, method) };
                }
            }
        }
    },
});
