// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/PathResourceTypeNameCamelCase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PathResourceTypeNameCamelCase = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const rules_helper_1 = require("../utilities/rules-helper");
exports.PathResourceTypeNameCamelCase = "PathResourceTypeNameCamelCase";
openapi_validator_core_1.rules.push({
    id: "R3021",
    name: exports.PathResourceTypeNameCamelCase,
    severity: "error",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    *run(doc, node, path) {
        if (node.paths !== undefined) {
            const msg = "Resource type naming must follow camel case.";
            const paths = Object.keys(node.paths);
            for (const it of paths) {
                const allWords = (0, rules_helper_1.getAllWordsFromPath)(it);
                const resourceProviders = new Set((0, rules_helper_1.getAllResourceProvidersFromPath)(it));
                const resourceTypes = allWords.filter(subPath => !resourceProviders.has(subPath));
                if (resourceTypes.some(it => !(0, rules_helper_1.resourceTypeMustCamelCase)(it))) {
                    yield {
                        message: `${msg} Path: '${it}'`,
                        location: path.concat(["paths", it])
                    };
                }
            }
        }
    }
});
