// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/PathResourceProviderNamePascalCase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PathResourceProviderNamePascalCase = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const rules_helper_1 = require("../utilities/rules-helper");
exports.PathResourceProviderNamePascalCase = "PathResourceProviderNamePascalCase";
openapi_validator_core_1.rules.push({
    id: "R3020",
    name: exports.PathResourceProviderNamePascalCase,
    severity: "error",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    *run(doc, node, path) {
        if (node.paths !== undefined) {
            const msg = "Resource provider naming must follow the pascal case.";
            const paths = Object.keys(node.paths);
            for (const it of paths) {
                const resourceProviders = (0, rules_helper_1.getAllResourceProvidersFromPath)(it);
                if (resourceProviders.some(rp => !(0, rules_helper_1.resourceProviderMustPascalCase)(rp))) {
                    yield {
                        message: `${msg} Path: '${it}'`,
                        location: path.concat(["paths", it])
                    };
                }
            }
        }
    }
});
