// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/PostOperationIdContainsUrlVerb.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostOperationIdContainsUrlVerb = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
exports.PostOperationIdContainsUrlVerb = "PostOperationIdContainsUrlVerb";
openapi_validator_core_1.rules.push({
    id: "R2066",
    name: exports.PostOperationIdContainsUrlVerb,
    severity: "warning",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$.paths.*[?(@property==='post')]^",
    *run(doc, node, path) {
        const pathNodes = path[path.length - 1].toLowerCase().split("/");
        const urlVerb = pathNodes[pathNodes.length - 1];
        const keys = Object.keys(node);
        const postKey = keys.find(key => {
            return key.toLowerCase() === "post";
        });
        if (postKey) {
            const operationId = node[postKey].operationId;
            if (operationId &&
                operationId
                    .toLowerCase()
                    .split("_")
                    .pop()
                    .indexOf(urlVerb) === -1) {
                yield {
                    message: `OperationId should contain the verb: '${urlVerb}' in:'${operationId}'. Consider updating the operationId`,
                    location: path.concat(postKey).concat("operationId")
                };
            }
        }
    }
});
