// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/PrivateEndpointResourceSchemaValidation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivateEndpointResourceSchemaValidation = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const arm_helper_1 = require("../utilities/arm-helper");
const swagger_helper_1 = require("../utilities/swagger-helper");
exports.PrivateEndpointResourceSchemaValidation = "PrivateEndpointResourceSchemaValidation";
openapi_validator_core_1.rules.push({
    id: "R4035",
    name: exports.PrivateEndpointResourceSchemaValidation,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.composed,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$.paths.*",
    async *run(doc, node, path, ctx) {
        const msg = 'The private endpoint model "{0}" schema does not conform to the common type definition.';
        const privateEndpointConnection = /.*\/privateEndpointConnections(\/\{[^/]+\})*$/;
        const privateLinkResources = /.*\/privateLinkResources$/;
        const utils = new arm_helper_1.ArmHelper(doc, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
        const swaggerUtil = new swagger_helper_1.SwaggerHelper(doc, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
        const apiPath = path[path.length - 1];
        const checkPrivateEndpoint = (model) => {
            const properties = swaggerUtil === null || swaggerUtil === void 0 ? void 0 : swaggerUtil.getProperty(model, "properties");
            if (!properties) {
                return false;
            }
            const requiredProperties = ["privateEndpoint", "privateLinkServiceConnectionState"];
            if (requiredProperties.some((p) => !(swaggerUtil === null || swaggerUtil === void 0 ? void 0 : swaggerUtil.getProperty(properties, p)))) {
                return false;
            }
            return true;
        };
        const checkPrivateResources = (model) => {
            const properties = swaggerUtil === null || swaggerUtil === void 0 ? void 0 : swaggerUtil.getProperty(model, "properties");
            if (!properties) {
                return false;
            }
            const requiredProperties = ["groupId", "requiredMembers", "requiredZoneNames"];
            if (requiredProperties.some((p) => !(swaggerUtil === null || swaggerUtil === void 0 ? void 0 : swaggerUtil.getProperty(properties, p)))) {
                return false;
            }
            return true;
        };
        if (privateEndpointConnection.test(apiPath)) {
            const model = utils.getResponseModelFromPath(apiPath);
            if (model) {
                const modelName = utils.stripDefinitionPath(model.value.$ref) || "";
                if (apiPath.endsWith("privateEndpointConnections")) {
                    const privateEndpoint = utils === null || utils === void 0 ? void 0 : utils.getProperty(model, "value");
                    const items = utils === null || utils === void 0 ? void 0 : utils.getAttribute(privateEndpoint, "items");
                    if (!privateEndpoint || !items) {
                        yield {
                            message: msg.replace("{0}", modelName),
                            location: [...path, "get", "responses", "200"],
                        };
                    }
                    else if (!checkPrivateEndpoint(items)) {
                        yield {
                            message: msg.replace("{0}", modelName),
                            location: [...path, "get", "responses", "200"],
                        };
                    }
                }
                else {
                    if (!checkPrivateEndpoint(model)) {
                        yield {
                            message: msg.replace("{0}", modelName),
                            location: [...path, "get", "responses", "200"],
                        };
                    }
                }
            }
        }
        if (privateLinkResources.test(apiPath)) {
            const model = utils.getResponseModelFromPath(apiPath);
            if (model) {
                const modelName = utils.stripDefinitionPath(model.value.$ref) || "";
                const privateResources = utils === null || utils === void 0 ? void 0 : utils.getProperty(model, "value");
                const items = utils === null || utils === void 0 ? void 0 : utils.getAttribute(privateResources, "items");
                if (!privateResources) {
                    yield {
                        message: msg.replace("{0}", modelName),
                        location: [...path, "get", "responses", "200"],
                    };
                }
                else if (!items || !checkPrivateResources(items)) {
                    yield {
                        message: msg.replace("{0}", modelName),
                        location: [...path, "get", "responses", "200"],
                    };
                }
            }
        }
    },
});
