// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/RequiredReadOnlySystemData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequiredReadOnlySystemData = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const arm_helper_1 = require("../utilities/arm-helper");
const jsonpath_1 = require("../utilities/jsonpath");
const swagger_helper_1 = require("../utilities/swagger-helper");
const swagger_workspace_1 = require("../utilities/swagger-workspace");
exports.RequiredReadOnlySystemData = "RequiredReadOnlySystemData";
openapi_validator_core_1.rules.push({
    id: "R4009",
    name: exports.RequiredReadOnlySystemData,
    severity: "warning",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    *run(doc, node, path, ctx) {
        if (doc.info && doc.info.version) {
            const apiVersion = doc.info.version;
            const matched = apiVersion.match(/\d{4}-\d{2}-\d{2}/g);
            const apiVersionFormated = matched && matched.length > 0 ? matched[0] : undefined;
            if (!apiVersionFormated || apiVersionFormated < "2020-05-01") {
                return;
            }
            const utils = new arm_helper_1.ArmHelper(doc, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
            const swaggerUtil = new swagger_helper_1.SwaggerHelper(doc, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
            const allResources = utils.getAllResourceNames();
            for (const value of ["get", "put", "patch"]) {
                for (const responses of (0, jsonpath_1.nodes)(doc, `$.paths.*.${value}.responses`)) {
                    let hasSystemData = true;
                    let isReadOnly = true;
                    for (const key of Object.keys(responses.value)) {
                        if (key.startsWith("20")) {
                            const response = responses.value[key];
                            const toValidateSchema = response.schema;
                            if (!toValidateSchema || !toValidateSchema.$ref) {
                                continue;
                            }
                            const toValidateModelName = utils.stripDefinitionPath(toValidateSchema.$ref);
                            if (toValidateModelName && allResources.includes(toValidateModelName)) {
                                const systemData = swaggerUtil === null || swaggerUtil === void 0 ? void 0 : swaggerUtil.getProperty(swagger_workspace_1.Workspace.createEnhancedSchema(toValidateSchema, ctx === null || ctx === void 0 ? void 0 : ctx.specPath), "systemData");
                                if (!systemData) {
                                    hasSystemData = false;
                                    break;
                                }
                                if (!systemData.value.readOnly) {
                                    isReadOnly = false;
                                    break;
                                }
                            }
                        }
                    }
                    if (!hasSystemData) {
                        const operationId = (0, jsonpath_1.nodes)(doc, (0, jsonpath_1.stringify)(responses.path.slice(0, -1)))[0].value.operationId;
                        yield {
                            message: `The response of operation:'${operationId}' is defined without 'systemData'. Consider adding the systemData to the response.`,
                            location: responses.path.slice(0, -1),
                        };
                    }
                    if (!isReadOnly) {
                        const operationId = (0, jsonpath_1.nodes)(doc, (0, jsonpath_1.stringify)(responses.path.slice(0, -1)))[0].value.operationId;
                        yield {
                            message: `The property systemData in the response of operation:'${operationId}' is not read only. Please add the readonly for the systemData.`,
                            location: responses.path.slice(0, -1),
                        };
                    }
                }
            }
        }
    },
});
