// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/Rpaas_ResourceProvisioningState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rpaas_ResourceProvisioningState = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const arm_helper_1 = require("../utilities/arm-helper");
const swagger_helper_1 = require("../utilities/swagger-helper");
exports.Rpaas_ResourceProvisioningState = "Rpaas_ResourceProvisioningState";
openapi_validator_core_1.rules.push({
    id: "R4031",
    name: exports.Rpaas_ResourceProvisioningState,
    severity: "error",
    category: "RPaaSViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.rpaas,
    appliesTo_JsonQuery: "$",
    *run(doc, node, path, ctx) {
        const msg = `[RPaaS] The resource {0} is defined without 'provisioningState' in properties bag, consider adding the provisioningState for it.`;
        const utils = new arm_helper_1.ArmHelper(doc, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
        const swaggerUtil = new swagger_helper_1.SwaggerHelper(doc, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
        const allResources = utils.getAllResourceNames();
        for (const resource of allResources) {
            const model = utils.getResourceByName(resource);
            const properties = swaggerUtil === null || swaggerUtil === void 0 ? void 0 : swaggerUtil.getProperty(model, "properties");
            let hasProvisioningState = false;
            if (properties && (!properties.value.type || properties.value.type === "object")) {
                if (swaggerUtil === null || swaggerUtil === void 0 ? void 0 : swaggerUtil.getProperty(properties, "provisioningState")) {
                    hasProvisioningState = true;
                }
            }
            if (!hasProvisioningState) {
                yield { message: msg.replace("{0}", resource), location: ["$", "definitions", resource] };
            }
        }
    },
});
