// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/TopLevelResourcesListBySubscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TopLevelResourcesListBySubscription = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const arm_helper_1 = require("../utilities/arm-helper");
exports.TopLevelResourcesListBySubscription = "TopLevelResourcesListBySubscription";
openapi_validator_core_1.rules.push({
    id: "R4017",
    name: exports.TopLevelResourcesListBySubscription,
    severity: "error",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.composed,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$",
    *run(doc, node, path, ctx) {
        var _a;
        const msg = 'The top-level resource "{0}" does not have list by subscription operation, please add it.';
        const utils = new arm_helper_1.ArmHelper(doc, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
        const topLevelResources = utils.getTopLevelResources();
        const allCollectionApis = utils.getCollectionApiInfo();
        const validatedModels = new Set();
        for (const resource of topLevelResources) {
            const resourceName = resource.modelName;
            if (validatedModels.has(resourceName)) {
                continue;
            }
            const hasSubscriptionScopedOperation = (_a = resource.operations) === null || _a === void 0 ? void 0 : _a.some((op) => utils.isPathBySubscription(op.apiPath));
            if (!hasSubscriptionScopedOperation) {
                continue;
            }
            validatedModels.add(resourceName);
            const hasMatched = allCollectionApis.some((collection) => resourceName === collection.childModelName && collection.collectionGetPath.some((p) => utils.isPathBySubscription(p)));
            if (!hasMatched) {
                yield {
                    message: msg.replace("{0}", resourceName),
                    location: ["$", "definitions", resourceName],
                };
            }
        }
    },
});
