// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/TopLevelResourcesListByResourceGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TopLevelResourcesListByResourceGroup = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const arm_helper_1 = require("../utilities/arm-helper");
exports.TopLevelResourcesListByResourceGroup = "TopLevelResourcesListByResourceGroup";
openapi_validator_core_1.rules.push({
    id: "R4016",
    name: exports.TopLevelResourcesListByResourceGroup,
    severity: "error",
    category: "ARMViolation",
    mergeState: openapi_validator_core_1.MergeStates.composed,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$",
    *run(doc, node, path, ctx) {
        const msg = 'The top-level resource "{0}" does not have list by resource group operation, please add it.';
        const utils = new arm_helper_1.ArmHelper(doc, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
        const topLevelResources = utils.getTopLevelResourcesByRG();
        const allCollectionApis = utils.getCollectionApiInfo();
        for (const resource of topLevelResources) {
            const hasMatched = allCollectionApis.some((collection) => resource === collection.childModelName && collection.collectionGetPath.some((p) => utils.isPathByResourceGroup(p)));
            if (!hasMatched) {
                yield {
                    message: msg.replace("{0}", resource),
                    location: ["$", "definitions", resource],
                };
            }
        }
    },
});
