// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/UniqueXmsExample.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniqueXmsExample = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const jsonpath_1 = require("../utilities/jsonpath");
exports.UniqueXmsExample = "UniqueXmsExample";
openapi_validator_core_1.rules.push({
    id: "R4030",
    name: exports.UniqueXmsExample,
    severity: "warning",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.composed,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$",
    *run(doc, node, path) {
        const msg = `Do not have duplicate name of x-ms-example, make sure every x-ms-example name unique. Duplicate x-ms-example: `;
        const examples = new Set();
        for (const it of (0, jsonpath_1.nodes)(node, '$.paths.*.*["x-ms-examples"]')) {
            if (it.value !== null) {
                for (const exampleName of Object.keys(it.value)) {
                    if (examples.has(exampleName)) {
                        yield { message: msg + exampleName, location: path.concat(it.path.slice(1)) };
                    }
                    examples.add(exampleName);
                }
            }
        }
    }
});
