// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/UniqueClientParameterName.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniqueClientParameterName = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const jsonpath_1 = require("../utilities/jsonpath");
exports.UniqueClientParameterName = "UniqueClientParameterName";
openapi_validator_core_1.rules.push({
    id: "R4029",
    name: exports.UniqueClientParameterName,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.composed,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm | openapi_validator_core_1.OpenApiTypes.dataplane,
    appliesTo_JsonQuery: "$",
    *run(doc, node, path) {
        const clientParameterName = new Map();
        const getRefModel = (refValue) => {
            const pathExpression = refValue.replace(/^#\//, "").split("/");
            try {
                const result = (0, jsonpath_1.nodes)(doc, (0, jsonpath_1.stringify)(pathExpression));
                return result.length !== 0 ? result[0] : undefined;
            }
            catch (err) {
                return undefined;
            }
        };
        const checkParameterNameUnique = (parameter) => {
            if (parameter) {
                const ref = parameter.$ref;
                if (ref === undefined) {
                    return true;
                }
                const refModels = getRefModel(ref);
                if (refModels === undefined) {
                    return true;
                }
                const refModel = refModels.value;
                const parameterName = refModel.name;
                const xMsParameterLocation = refModel["x-ms-parameter-location"];
                const parameterIn = refModel.in;
                if (xMsParameterLocation === "method" || parameterIn !== "path") {
                    return true;
                }
                if (clientParameterName.has(parameterName)) {
                    return clientParameterName.get(parameterName) === ref;
                }
                else {
                    clientParameterName.set(parameterName, ref);
                }
            }
            return true;
        };
        const msg = `Do not have duplicate name of client parameter name, make sure every client parameter name unique. `;
        for (const it of (0, jsonpath_1.nodes)(doc, "$.paths.*.*.parameters")) {
            for (const parameter of Object.values(it.value)) {
                if (!checkParameterNameUnique(parameter)) {
                    const ref = parameter.$ref;
                    const conflictMsg = `Client parameter ${parameter.$ref} conflicted with ${clientParameterName.get(getRefModel(ref).value.name)}.`;
                    yield {
                        location: path.concat(it.path.slice(1)),
                        message: msg + conflictMsg
                    };
                }
            }
        }
        for (const it of (0, jsonpath_1.nodes)(doc, "$.paths.*.parameters")) {
            for (const parameter of Object.values(it.value)) {
                if (!checkParameterNameUnique(parameter)) {
                    const ref = parameter.$ref;
                    const conflictMsg = `Client parameter ${parameter.$ref} conflicted with ${clientParameterName.get(getRefModel(ref).value.name)}.`;
                    yield {
                        location: path.concat(it.path.slice(1)),
                        message: msg + conflictMsg
                    };
                }
            }
        }
    }
});
