// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/UniqueModelName.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniqueModelName = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
exports.UniqueModelName = "UniqueModelName";
openapi_validator_core_1.rules.push({
    id: "R4033",
    name: exports.UniqueModelName,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.composed,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$.definitions",
    *run(doc, node, path) {
        const msg = `The model name {0} is duplicated with {1} .`;
        const models = new Map();
        for (const key of Object.keys(node)) {
            const model = models.get(key.toLowerCase());
            if (model) {
                yield { message: msg.replace("{0}", key).replace("{1}", model), location: path.concat(key) };
            }
            else {
                models.set(key.toLowerCase(), key);
            }
        }
    }
});
