// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/UniqueXmsEnumName.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniqueXmsEnumName = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const jsonpath_1 = require("../utilities/jsonpath");
const rules_helper_1 = require("../utilities/rules-helper");
exports.UniqueXmsEnumName = "UniqueXmsEnumName";
openapi_validator_core_1.rules.push({
    id: "R4005",
    name: exports.UniqueXmsEnumName,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.composed,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm | openapi_validator_core_1.OpenApiTypes.dataplane,
    appliesTo_JsonQuery: "$.definitions",
    *run(doc, node, path) {
        const msg = `Must not have duplicate name of x-ms-enum extension , make sure every x-ms-enum name unique.`;
        if (node) {
            const enumMap = new Map();
            for (const section of (0, jsonpath_1.nodes)(node, "$..*[?(@property === 'enum')]^")) {
                if ((0, rules_helper_1.isValidEnum)(section.value) && section.value["x-ms-enum"] && section.value["x-ms-enum"].name) {
                    const enumName = section.value["x-ms-enum"].name.toLowerCase();
                    if (enumMap.has(enumName)) {
                        const curEnum = section.value.enum;
                        const existingNode = enumMap.get(enumName);
                        const existingEnum = existingNode.value.enum;
                        const existingEnumModelAsString = existingNode.value["x-ms-enum"].modelAsString || false;
                        const currentEnumModelAsString = section.value["x-ms-enum"].modelAsString || false;
                        if (section.value.type !== existingNode.value.type ||
                            currentEnumModelAsString !== existingEnumModelAsString ||
                            existingEnum.length !== curEnum.length ||
                            currentEnumModelAsString
                            ? existingEnum.some((value) => !curEnum.includes(value))
                            : existingEnum.some((value, index) => curEnum[index] !== value)) {
                            yield {
                                message: `${msg} The duplicate x-ms-enum name: ${enumName}, path: ${(0, jsonpath_1.stringify)(existingNode.path.slice(1))}`,
                                location: path.concat(section.path.slice(1)),
                            };
                        }
                    }
                    else {
                        enumMap.set(enumName, section);
                    }
                }
            }
        }
    },
});
