// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/ValidResponseCodeRequired.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidResponseCodeRequired = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
exports.ValidResponseCodeRequired = "ValidResponseCodeRequired";
openapi_validator_core_1.rules.push({
    id: "R4028",
    name: exports.ValidResponseCodeRequired,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm | openapi_validator_core_1.OpenApiTypes.dataplane,
    appliesTo_JsonQuery: "$.paths.*.*.responses",
    *run(doc, node, path) {
        const msg = "There is no declared valid status code.";
        const response = node;
        if (response.default && Object.keys(response).length === 1) {
            yield { message: `${msg}`, location: path };
        }
        else {
            if (!Object.keys(response).some(v => v.startsWith("20"))) {
                yield { message: `${msg}`, location: path };
            }
        }
    }
});
