// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/XmsEnumValidation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XmsEnumValidation = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const rules_helper_1 = require("../utilities/rules-helper");
exports.XmsEnumValidation = "XmsEnumValidation";
openapi_validator_core_1.rules.push({
    id: "R2018",
    name: exports.XmsEnumValidation,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm | openapi_validator_core_1.OpenApiTypes.dataplane,
    appliesTo_JsonQuery: [
        "$.definitions..[?(@property==='enum')]^",
        "$..parameters..[?(@property==='enum')]^",
        "$..responses..[?(@property==='enum')]^",
    ],
    *run(doc, node, path) {
        const msg = `The enum types should have x-ms-enum type extension set with appropriate options.`;
        if (node.enum && (0, rules_helper_1.isValidEnum)(node)) {
            if (!node["x-ms-enum"]) {
                yield { message: `${msg}`, location: path };
            }
        }
    },
});
