// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/XmsIdentifierValidation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XmsIdentifierValidation = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const swagger_workspace_1 = require("../utilities/swagger-workspace");
exports.XmsIdentifierValidation = "XmsIdentifierValidation";
openapi_validator_core_1.rules.push({
    id: "R4041",
    name: exports.XmsIdentifierValidation,
    severity: "warning",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: ["$.definitions..[?(@property === 'items')]^", "$.parameters..[?(@property === 'items')]^"],
    async *run(doc, node, path, ctx) {
        var _a;
        if (node.type !== "array") {
            return;
        }
        const identifiers = (_a = node["x-ms-identifiers"]) !== null && _a !== void 0 ? _a : ["id"];
        const items = await swagger_workspace_1.Workspace.resolveRef({ file: ctx === null || ctx === void 0 ? void 0 : ctx.specPath, value: node.items }, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
        if (items && items.value.type && items.value.type !== "object") {
            return;
        }
        for (const identifier of identifiers) {
            const pathExpression = identifier.replace(/^\//, "").split("/");
            let item = items;
            for (const key of pathExpression) {
                item = swagger_workspace_1.Workspace.getProperty(item, key, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
                if (item === undefined) {
                    yield { message: `Missing identifier ${identifier} in array item property`, location: path };
                    break;
                }
            }
        }
    },
});
