// @microsoft.azure/openapi-validator-rulesets/dist/native/legacyRules/XmsPageableMustHaveCorrespondingResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XmsPageableMustHaveCorrespondingResponse = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const rules_helper_1 = require("../utilities/rules-helper");
const swagger_workspace_1 = require("../utilities/swagger-workspace");
exports.XmsPageableMustHaveCorrespondingResponse = "XmsPageableMustHaveCorrespondingResponse";
openapi_validator_core_1.rules.push({
    id: "R4012",
    name: exports.XmsPageableMustHaveCorrespondingResponse,
    severity: "error",
    category: "SDKViolation",
    mergeState: openapi_validator_core_1.MergeStates.individual,
    openapiType: openapi_validator_core_1.OpenApiTypes.arm,
    appliesTo_JsonQuery: "$.paths..[?(@property==='x-ms-pageable')]^",
    async *run(doc, node, path, ctx) {
        const nextLinkValue = node["x-ms-pageable"].nextLinkName;
        if (!nextLinkValue) {
            return;
        }
        const mostSuccesskey = (0, rules_helper_1.getMostSuccessfulResponseKey)(Object.keys(node.responses));
        const msg = `The operation: '${node.operationId}' is defined with x-ms-pageable enabled,but can not find the corresponding nextLink property in the response, please add it.`;
        if (node.responses && node.responses[mostSuccesskey]) {
            const schemaPath = path.concat(["responses", mostSuccesskey, "schema"]);
            const resolvedSchema = (0, rules_helper_1.getResolvedSchemaByPath)(schemaPath, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
            if (resolvedSchema && !(swagger_workspace_1.Workspace === null || swagger_workspace_1.Workspace === void 0 ? void 0 : swagger_workspace_1.Workspace.getProperty(resolvedSchema, nextLinkValue, ctx === null || ctx === void 0 ? void 0 : ctx.inventory))) {
                yield { message: `${msg}`, location: path };
            }
        }
    },
});
