// oav/dist/lib/swaggerValidator/ajv.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ajvEnableArmRule = exports.ajvEnableAll = exports.ajvEnableArmIdFormat = exports.ajvEnableByteFormat = exports.ajvEnableDurationFormat = exports.ajvEnableDateTimeRfc1123Format = exports.ajvAddFormatsDefaultValidation = exports.ajvEnableUnixTimeFormat = exports.ajvEnableInt32AndInt64Format = exports.ajvEnableXmsAzureResource = exports.ajvEnableXmsSecret = exports.ajvEnableReadOnlyAndXmsMutability = void 0;
const constants_1 = require("../util/constants");
const ajvDiscriminatorMap_1 = require("./ajvDiscriminatorMap");
const ajvEnableReadOnlyAndXmsMutability = (ajv) => {
    ajv.removeKeyword("readOnly");
    ajv.addKeyword("readOnly", {
        metaSchema: { type: "boolean" },
        inline: (it, _keyword, isReadOnly, parentSchema) => {
            if ((parentSchema === null || parentSchema === void 0 ? void 0 : parentSchema[constants_1.xmsMutability]) !== undefined) {
                return "1";
            }
            const data = `data${it.dataLevel || ""}`;
            return isReadOnly ? `this.isResponse || ${data} === null || ${data} === undefined` : "1";
        },
    });
    ajv.addKeyword(constants_1.xmsMutability, {
        metaSchema: { type: "array", items: { enum: ["create", "update", "read"] } },
        inline: (it, _keyword, mutability) => {
            const validInRequest = mutability.includes("create") || mutability.includes("update");
            const validInResponse = mutability.includes("read");
            if (validInRequest && validInResponse) {
                return "1";
            }
            if (!validInRequest && !validInResponse) {
                throw new Error(`Invalid ${constants_1.xmsMutability} value: ${JSON.stringify(mutability)}`);
            }
            const data = `data${it.dataLevel || ""}`;
            return `${validInRequest ? "!" : ""}this.isResponse || ${data} === null || ${data} === undefined`;
        },
    });
};
exports.ajvEnableReadOnlyAndXmsMutability = ajvEnableReadOnlyAndXmsMutability;
const ajvEnableXmsSecret = (ajv) => {
    ajv.addKeyword(constants_1.xmsSecret, {
        metaSchema: { type: "boolean" },
        inline: (it, _keyword, isSecret) => {
            const data = `data${it.dataLevel || ""}`;
            return isSecret ? `!this.isResponse || ${data} === null || ${data} === undefined` : "1";
        },
    });
};
exports.ajvEnableXmsSecret = ajvEnableXmsSecret;
const ajvEnableXmsAzureResource = (ajv) => {
    ajv.addKeyword(constants_1.xmsAzureResource, {
        metaSchema: { type: "boolean" },
        inline: (it, _keyword, isResource) => {
            const data = `data${it.dataLevel || ""}`;
            return isResource
                ? `!(this.isResponse && (this.httpMethod === 'get' || this.httpMethod === 'put')) || (${data}.id !== null && ${data}.id !== undefined)`
                : "1";
        },
    });
};
exports.ajvEnableXmsAzureResource = ajvEnableXmsAzureResource;
const ajvEnableInt32AndInt64Format = (ajv) => {
    ajv.addFormat("int32", {
        type: "number",
        validate: (x) => x % 1 === 0 && x >= -2147483648 && x <= 2147483647,
    });
    // TODO int64 range exceed Number.MAX_SAFE_INTEGER so we will lost precision when JSON.parse
    const int64Max = BigInt(2) ** BigInt(63) - BigInt(1);
    const int64Min = BigInt(2) ** BigInt(63) * BigInt(-1);
    ajv.addFormat("int64", {
        type: "number",
        validate: (x) => x % 1 === 0 && x >= int64Min && x <= int64Max,
    });
};
exports.ajvEnableInt32AndInt64Format = ajvEnableInt32AndInt64Format;
const ajvEnableUnixTimeFormat = (ajv) => {
    ajv.addFormat("unixtime", {
        type: "number",
        validate: (x) => x % 1 === 0,
    });
};
exports.ajvEnableUnixTimeFormat = ajvEnableUnixTimeFormat;
const ajvAddFormatsDefaultValidation = (ajv, type, formats) => {
    for (const format of formats) {
        ajv.addFormat(format, {
            type,
            validate: () => true,
        });
    }
};
exports.ajvAddFormatsDefaultValidation = ajvAddFormatsDefaultValidation;
const ajvEnableDateTimeRfc1123Format = (ajv) => {
    // https://tools.ietf.org/html/rfc822#section-5
    ajv.addFormat("date-time-rfc1123", {
        type: "string",
        validate: /^(?:(?:Mon|Tue|Wed|Thu|Fri|Sat|Sun), )?[0-3]\d (?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec) \d\d(?:\d\d)? (?:[0-2]\d:[0-5]\d(?::[0-5]\d)|23:59:60) (?:[A-Z]{1,3})?(?:[+-]\d\d\d\d)?$/,
    });
};
exports.ajvEnableDateTimeRfc1123Format = ajvEnableDateTimeRfc1123Format;
const ajvEnableDurationFormat = (ajv) => {
    // https://en.wikipedia.org/wiki/ISO_8601#Durations
    ajv.addFormat("duration", {
        type: "string",
        validate: /^P([0-9]+(?:[,\.][0-9]+)?Y)?([0-9]+(?:[,\.][0-9]+)?M)?([0-9]+(?:[,\.][0-9]+)?D)?(?:T([0-9]+(?:[,\.][0-9]+)?H)?([0-9]+(?:[,\.][0-9]+)?M)?([0-9]+(?:[,\.][0-9]+)?S)?)?$/,
    });
};
exports.ajvEnableDurationFormat = ajvEnableDurationFormat;
const ajvEnableByteFormat = (ajv) => {
    // https://datatracker.ietf.org/doc/html/rfc4648#section-4
    ajv.addFormat("byte", {
        type: "string",
        validate: (x) => {
            const decodedValue = Buffer.from(x, "base64");
            const reencodedValue = Buffer.from(decodedValue).toString("base64");
            return reencodedValue === x;
        },
    });
};
exports.ajvEnableByteFormat = ajvEnableByteFormat;
// TODO: This could be more advanced, looking at the allowedResources field (see https://github.com/Azure/autorest/tree/main/docs/extensions#schema) and ensuring that only the allowed resources are referenced, but
// TODO: for now a generic ARM ID check is better than nothing.
const ajvEnableArmIdFormat = (ajv) => {
    ajv.addFormat("arm-id", {
        type: "string",
        // Note that this regex isn't perfect but it does an OK job. See https://regex101.com/r/96g3K5/1
        validate: new RegExp("(^(/subscriptions/([^/]+)(/resourcegroups/([^/]+))?)?/providers/([^/]+)/([^/]+/[^/]+)(/([^/]+/[^/]+))*$|^/subscriptions/([^/]+)(/resourcegroups/([^/]+))?$)", "i"),
    });
};
exports.ajvEnableArmIdFormat = ajvEnableArmIdFormat;
// for (const keyword of [
//   "name",
//   "in",
//   "example",
//   "parameters",
//   "externalDocs",
//   "x-nullable",
//   "x-ms-enum",
//   "x-ms-azure-resource",
//   "x-ms-parameter-location",
//   "x-ms-client-name",
//   "x-ms-external",
//   "x-ms-skip-url-encoding",
//   "x-ms-client-flatten",
//   "x-ms-api-version",
//   "x-ms-parameter-grouping",
//   "x-ms-discriminator-value",
//   "x-ms-client-request-id",
//   "x-apim-code-nillable",
//   "x-new-pattern",
//   "x-previous-pattern",
//   "x-comment",
//   "x-abstract",
//   "allowEmptyValue",
//   "collectionFormat",
// ]) {
//   ajv.addKeyword(keyword, {});
// }
const ajvEnableAll = (ajv, jsonLoader) => {
    (0, ajvDiscriminatorMap_1.ajvEnableDiscriminatorMap)(ajv, jsonLoader);
    (0, exports.ajvEnableXmsSecret)(ajv);
    (0, exports.ajvEnableReadOnlyAndXmsMutability)(ajv);
    (0, exports.ajvEnableUnixTimeFormat)(ajv);
    (0, exports.ajvEnableInt32AndInt64Format)(ajv);
    (0, exports.ajvEnableDurationFormat)(ajv);
    (0, exports.ajvEnableDateTimeRfc1123Format)(ajv);
    (0, exports.ajvEnableByteFormat)(ajv);
    (0, exports.ajvAddFormatsDefaultValidation)(ajv, "string", [
        "password",
        "file",
        "base64url",
        "",
        "binary",
        "non-iso-duration",
        "char",
    ]);
    (0, exports.ajvAddFormatsDefaultValidation)(ajv, "number", ["double", "float", "decimal"]);
};
exports.ajvEnableAll = ajvEnableAll;
const ajvEnableArmRule = (ajv) => {
    (0, exports.ajvEnableXmsAzureResource)(ajv);
};
exports.ajvEnableArmRule = ajvEnableArmRule;
