// oav/dist/lib/swaggerValidator/ajvDiscriminatorMap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ajvEnableDiscriminatorMap = void 0;
const ajvEnableDiscriminatorMap = (ajv, loader) => {
    ajv.addKeyword("discriminatorMap", {
        errors: "full",
        metaSchema: { type: "object", additionalProperty: { type: "object,null" } },
        compile(schemas, parentSchema) {
            const compiled = {};
            const schemaMap = {};
            for (const value of Object.keys(schemas)) {
                compiled[value] = schemas[value] === null ? null : ajv.compile(schemas[value]);
                schemaMap[value] =
                    schemas[value] === null ? parentSchema : loader.resolveRefObj(schemas[value]);
            }
            const discriminator = parentSchema.discriminator;
            if (discriminator === undefined) {
                throw new Error("Discriminator is absent");
            }
            const validated = new WeakSet();
            const allowedValues = Object.keys(schemas);
            return function v(data, dataPath) {
                var _a;
                if (data === null || data === undefined || typeof data !== "object") {
                    // Should be validated by other schema property.
                    return true;
                }
                dataPath = dataPath !== null && dataPath !== void 0 ? dataPath : "";
                const discriminatorValue = data[discriminator];
                const validate = discriminatorValue !== undefined && discriminatorValue !== null
                    ? compiled[discriminatorValue]
                    : undefined;
                if (validated.has(data)) {
                    return true;
                }
                validated.add(data);
                const errors = [];
                const sch = (_a = schemaMap[discriminatorValue]) !== null && _a !== void 0 ? _a : parentSchema;
                if (validate === undefined || validate === null) {
                    v.errors = [
                        {
                            keyword: "discriminatorMap",
                            data,
                            dataPath,
                            params: { allowedValues, discriminatorValue },
                            schemaPath: "/discriminator",
                            schema: schemas,
                            parentSchema,
                            _realSchema: schemas,
                            _realData: data,
                        },
                    ];
                    return false;
                }
                if (validate !== undefined && validate !== null) {
                    const valid = validate.call(this, data);
                    if (!valid && validate.errors) {
                        for (const err of validate.errors) {
                            err.dataPath = dataPath + err.dataPath;
                            if (err._realSchema === undefined) {
                                err._realSchema = err.schema;
                                err._realData = err.data;
                            }
                        }
                        errors.push(...validate.errors);
                    }
                }
                if (sch.properties !== undefined &&
                    sch.additionalProperties === undefined &&
                    parentSchema.properties !== undefined &&
                    parentSchema.additionalProperties === undefined) {
                    // We validate { additionalProperties: false } manually if it's not set
                    for (const key of Object.keys(data)) {
                        if (!(key in sch.properties)) {
                            errors.push({
                                keyword: "additionalProperties",
                                data,
                                dataPath,
                                params: { additionalProperty: key },
                                schemaPath: "/additionalProperties",
                                schema: false,
                                parentSchema,
                                _realSchema: false,
                                _realData: data,
                            });
                        }
                    }
                }
                if (errors.length > 0) {
                    v.errors = errors;
                    return false;
                }
                else {
                    v.errors = null;
                    return true;
                }
            };
        },
    });
};
exports.ajvEnableDiscriminatorMap = ajvEnableDiscriminatorMap;
