// oav/dist/lib/swaggerValidator/ajvSchemaValidator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ajvErrorCodeToOavErrorCode = exports.ajvErrorToSchemaValidateIssue = exports.sourceMapInfoToSourceLocation = exports.ajvErrorListToSchemaValidateIssueList = exports.AjvSchemaValidator = void 0;
const lodash = __importStar(require("lodash"));
const openapi_tools_common_1 = require("@azure-tools/openapi-tools-common");
const ajv_1 = __importDefault(require("ajv"));
const inversify_1 = require("inversify");
const inversifyUtils_1 = require("../inversifyUtils");
const jsonLoader_1 = require("../swagger/jsonLoader");
const suppressionLoader_1 = require("../swagger/suppressionLoader");
const swaggerTypes_1 = require("../swagger/swaggerTypes");
const context_1 = require("../transform/context");
const constants_1 = require("../util/constants");
const errorDefinitions_1 = require("../util/errorDefinitions");
const ajv_2 = require("./ajv");
const schemaValidator_1 = require("./schemaValidator");
let AjvSchemaValidator = class AjvSchemaValidator {
    constructor(loader, schemaValidatorOption) {
        this.ajv = (0, ajv_1.default)({
            // tslint:disable-next-line: no-submodule-imports
            meta: require("ajv/lib/refs/json-schema-draft-04.json"),
            schemaId: "auto",
            extendRefs: "fail",
            format: "full",
            missingRefs: true,
            addUsedSchema: false,
            removeAdditional: false,
            nullable: true,
            allErrors: true,
            messages: true,
            verbose: true,
            inlineRefs: false,
            passContext: true,
            loopRequired: 2,
            unknownFormats: "ignore",
            loadSchema: async (uri) => {
                const spec = await loader.resolveFile(uri);
                return { [jsonLoader_1.$id]: spec[jsonLoader_1.$id], definitions: spec.definitions, parameters: spec.parameters };
            },
        });
        (0, ajv_2.ajvEnableAll)(this.ajv, loader);
        // always enable the armId format validation
        (0, ajv_2.ajvEnableArmIdFormat)(this.ajv);
        if ((schemaValidatorOption === null || schemaValidatorOption === void 0 ? void 0 : schemaValidatorOption.isArmCall) === true) {
            (0, ajv_2.ajvEnableArmRule)(this.ajv);
        }
    }
    async compileAsync(schema) {
        const validate = await this.ajv.compileAsync(schema);
        return this.getValidateFunction(validate);
    }
    compile(schema) {
        const validate = this.ajv.compile(schema);
        return this.getValidateFunction(validate);
    }
    getValidateFunction(validate) {
        const ret = function validateSchema(ctx, data) {
            const result = [];
            const isValid = validateSchema.validate.call(ctx, data);
            if (!isValid) {
                const errors = ReValidateIfNeed(validateSchema.validate.errors, ctx, data, validateSchema.validate);
                if (errors.length > 0) {
                    (0, exports.ajvErrorListToSchemaValidateIssueList)(errors, ctx, result);
                }
                validateSchema.validate.errors = null;
            }
            return result;
        };
        ret.validate = validate;
        return ret;
    }
};
exports.AjvSchemaValidator = AjvSchemaValidator;
exports.AjvSchemaValidator = AjvSchemaValidator = __decorate([
    (0, inversify_1.injectable)(),
    __param(1, (0, inversify_1.inject)(inversifyUtils_1.TYPES.opts)),
    __metadata("design:paramtypes", [jsonLoader_1.JsonLoader, Object])
], AjvSchemaValidator);
const ajvErrorListToSchemaValidateIssueList = (errors, ctx, result) => {
    const includeErrorsSet = (0, schemaValidator_1.getIncludeErrorsMap)(ctx.includeErrors);
    const similarIssues = new Map();
    for (const error of errors) {
        const issue = (0, exports.ajvErrorToSchemaValidateIssue)(error, ctx);
        if (issue === undefined ||
            (includeErrorsSet !== undefined && !includeErrorsSet.has(issue.code))) {
            continue;
        }
        const issueHashKey = [
            issue.code,
            issue.message,
            issue.source.url,
            issue.source.position.column.toString(),
            issue.source.position.line.toString(),
        ].join("|");
        const similarIssue = similarIssues.get(issueHashKey);
        if (similarIssue === undefined) {
            similarIssues.set(issueHashKey, issue);
            result.push(issue);
            continue;
        }
        similarIssue.jsonPathsInPayload.push(issue.jsonPathsInPayload[0]);
    }
    for (const issue of result) {
        if (issue.jsonPathsInPayload.length > 1) {
            issue.jsonPathsInPayload = [...new Set(issue.jsonPathsInPayload)];
        }
    }
};
exports.ajvErrorListToSchemaValidateIssueList = ajvErrorListToSchemaValidateIssueList;
const sourceMapInfoToSourceLocation = (info) => {
    return info === undefined
        ? {
            url: "",
            position: { line: -1, column: -1 },
        }
        : {
            url: (0, openapi_tools_common_1.getRootObjectInfo)(info).url,
            position: { line: info.position.line, column: info.position.column },
        };
};
exports.sourceMapInfoToSourceLocation = sourceMapInfoToSourceLocation;
const ajvErrorToSchemaValidateIssue = (err, ctx) => {
    var _a, _b;
    const { parentSchema, params } = err;
    let { schema } = err;
    if (shouldSkipError(err, ctx)) {
        return undefined;
    }
    let dataPath = err.dataPath;
    const extraDataPath = (_b = (_a = params.additionalProperty) !== null && _a !== void 0 ? _a : params.missingProperty) !== null && _b !== void 0 ? _b : parentSchema.discriminator;
    if (extraDataPath !== undefined) {
        dataPath = `${dataPath}.${extraDataPath}`;
        if (schema[extraDataPath] !== undefined) {
            schema = schema[extraDataPath];
        }
    }
    const errInfo = (0, exports.ajvErrorCodeToOavErrorCode)(err, ctx);
    if (errInfo === undefined) {
        return undefined;
    }
    if ((0, suppressionLoader_1.isSuppressed)(err.parentSchema, errInfo.code, errInfo.message)) {
        return undefined;
    }
    let sch = parentSchema;
    let info = (0, openapi_tools_common_1.getInfo)(sch);
    if (info === undefined) {
        sch = schema;
        info = (0, openapi_tools_common_1.getInfo)(sch);
    }
    const source = (0, exports.sourceMapInfoToSourceLocation)(info);
    if ((sch === null || sch === void 0 ? void 0 : sch[swaggerTypes_1.refSelfSymbol]) !== undefined) {
        source.jsonRef = sch[swaggerTypes_1.refSelfSymbol].substr(sch[swaggerTypes_1.refSelfSymbol].indexOf("#"));
    }
    const result = errInfo;
    result.jsonPathsInPayload = [dataPath];
    result.schemaPath = err.schemaPath;
    result.source = source;
    return result;
};
exports.ajvErrorToSchemaValidateIssue = ajvErrorToSchemaValidateIssue;
const ReValidateIfNeed = (originalErrors, ctx, data, validate) => {
    const result = [];
    const newData = lodash.cloneDeep(data);
    for (const originalError of originalErrors) {
        validate.errors = null;
        const { schema, parentSchema: parentSch, keyword, data: errorData, dataPath } = originalError;
        const parentSchema = parentSch;
        // If the value of query parameter is in string format, we can revalidate this error
        if (!ctx.isResponse &&
            keyword === "type" &&
            schema === "array" &&
            typeof errorData === "string" &&
            (parentSchema === null || parentSchema === void 0 ? void 0 : parentSchema["in"]) === "query") {
            const arrayData = errorData.split(",").map((item) => {
                // when item is number
                const numberRegex = /^[+-]?\d+(\.\d+)?([Ee]\+?\d+)?$/g;
                if (numberRegex.test(item)) {
                    return parseFloat(item);
                }
                // when item is boolean
                if (item === "true" || item === "false") {
                    return item === "true";
                }
                return item;
            });
            const position = dataPath.substr(1);
            lodash.set(newData, position, arrayData);
            const isValid = validate.call(ctx, newData);
            if (!isValid) {
                // if validate.errors have new errors, add them to result
                for (const newError of validate.errors) {
                    let [includedInResult, includedInOriginalErrors] = [false, false];
                    for (const resultError of result) {
                        if (lodash.isEqual(newError, resultError)) {
                            // error is included in result
                            includedInResult = true;
                            break;
                        }
                    }
                    if (!includedInResult) {
                        for (const eachOriginalError of originalErrors) {
                            if (lodash.isEqual(newError, eachOriginalError)) {
                                // error is included in originalErrors
                                includedInOriginalErrors = true;
                                break;
                            }
                        }
                        if (!includedInOriginalErrors) {
                            result.push(newError);
                        }
                    }
                }
            }
            continue;
        }
        result.push(originalError);
    }
    return result;
};
const shouldSkipError = (error, cxt) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
    const { parentSchema: parentSch, params, keyword } = error;
    const parentSchema = parentSch;
    // If schema has allof property, we can get schema in "_realschema", so is data
    const schema = Object.keys(error).includes("_realSchema")
        ? error._realSchema
        : error.schema;
    const data = Object.keys(error).includes("_realData") ? error._realData : error.data;
    if ((schema === null || schema === void 0 ? void 0 : schema._skipError) || parentSchema._skipError) {
        return true;
    }
    // If we're erroring on the added property refWithReadOnly simply ignore the error
    if (error.keyword === "additionalProperties" &&
        params.additionalProperty === "refWithReadOnly") {
        return true;
    }
    // If a response has x-ms-mutability property and its missing the read we can skip this error
    if (cxt.isResponse &&
        ((keyword === "required" &&
            (((_c = (_b = (_a = parentSchema.properties) === null || _a === void 0 ? void 0 : _a[params.missingProperty]) === null || _b === void 0 ? void 0 : _b[constants_1.xmsMutability]) === null || _c === void 0 ? void 0 : _c.indexOf("read")) === -1 ||
                // required check is ignored when x-ms-secret is true
                ((_e = (_d = parentSchema.properties) === null || _d === void 0 ? void 0 : _d[params.missingProperty]) === null || _e === void 0 ? void 0 : _e[constants_1.xmsSecret]) ===
                    true)) ||
            (keyword === "type" && data === null && ((_f = parentSchema[constants_1.xmsMutability]) === null || _f === void 0 ? void 0 : _f.indexOf("read")) === -1))) {
        return true;
    }
    // If a request is missing a required property that is readOnly we can skip this error
    if (!cxt.isResponse &&
        keyword === "required" &&
        (((_h = (_g = parentSchema.properties) === null || _g === void 0 ? void 0 : _g[params.missingProperty]) === null || _h === void 0 ? void 0 : _h[constants_1.xmsReadonlyRef]) ||
            ((_k = (_j = parentSchema.properties) === null || _j === void 0 ? void 0 : _j[params.missingProperty]) === null || _k === void 0 ? void 0 : _k.readOnly))) {
        return true;
    }
    // If a response has property which x-ms-secret value is "true" in post we can skip this error
    if (cxt.isResponse &&
        (cxt === null || cxt === void 0 ? void 0 : cxt.httpMethod) === "post" &&
        // should skip error when x-ms-secret is "true"
        ((keyword === "x-ms-secret" && (parentSchema === null || parentSchema === void 0 ? void 0 : parentSchema[constants_1.xmsSecret]) === true) ||
            // should skip error when x-ms-secret is "true" and x-ms-mutability is "create" and "update"
            (keyword === "x-ms-mutability" &&
                (parentSchema === null || parentSchema === void 0 ? void 0 : parentSchema[constants_1.xmsSecret]) === true &&
                ((_l = parentSchema[constants_1.xmsMutability]) === null || _l === void 0 ? void 0 : _l.indexOf("read")) === -1))) {
        return true;
    }
    // If payload has property with date-time parameter and its value is valid except missing "Z" in the end we can skip this error
    if (keyword === "format" && schema === "date-time" && typeof data === "string") {
        const reg = /^\d+-(0\d|1[0-2])-([0-2]\d|3[01])T([01]\d|2[0-3]):[0-5][0-9]:[0-5][0-9]/;
        // intercept time, example: 2008-09-22T14:01:54
        const time = data.slice(0, 19);
        if (reg.test(time)) {
            const dateZ = new Date(data + "Z").toUTCString();
            // validate hour
            const ifHoursAreSame = time.slice(11, 13) === dateZ.slice(17, 19);
            // validate day for leap year, example: 2008-02-29
            const ifDaysAreSame = time.slice(8, 10) === dateZ.slice(5, 7);
            return ifHoursAreSame && ifDaysAreSame;
        }
    }
    // If a response data has multipleOf property, and it divided by multipleOf value is an integer, we can skip this error
    if (keyword === "multipleOf" && typeof schema === "number" && typeof data === "number") {
        let [newSchema, newData] = [schema, data];
        while (newSchema < 1) {
            newSchema *= 10;
            newData *= 10;
        }
        const result = newData / newSchema;
        // should skip error when response data divided by multipleOf value is an integer
        return result === parseInt(String(result));
    }
    return false;
};
const errorKeywordsMapping = {
    additionalProperties: "OBJECT_ADDITIONAL_PROPERTIES",
    required: "OBJECT_MISSING_REQUIRED_PROPERTY",
    format: "INVALID_FORMAT",
    type: "INVALID_TYPE",
    pattern: "PATTERN",
    minimum: "MINIMUM",
    maximum: "MAXIMUM",
    exclusiveMinimum: "MINIMUM_EXCLUSIVE",
    exclusiveMaximum: "MAXIMUM_EXCLUSIVE",
    minLength: "MIN_LENGTH",
    maxLength: "MAX_LENGTH",
    maxItems: "ARRAY_LENGTH_LONG",
    minItems: "ARRAY_LENGTH_SHORT",
    maxProperties: "OBJECT_PROPERTIES_MAXIMUM",
    minProperties: "OBJECT_PROPERTIES_MINIMUM",
    uniqueItems: "ARRAY_UNIQUE",
    additionalItems: "ARRAY_ADDITIONAL_ITEMS",
    anyOf: "ANY_OF_MISSING",
    dependencies: "OBJECT_DEPENDENCY_KEY",
    multiple: "MULTIPLE_OF",
    discriminatorMap: "DISCRIMINATOR_VALUE_NOT_FOUND",
    [constants_1.xmsAzureResource]: "MISSING_RESOURCE_ID",
};
// Should be type "never" to ensure we've covered all the errors
// export type MissingErrorCode = Exclude<
//   ExtendedErrorCode,
//   | keyof typeof validateErrorMessages
//   | "PII_MISMATCH" // Used in openapi-validate
//   | "INTERNAL_ERROR" // Used in liveValidator
//   | "UNRESOLVABLE_REFERENCE"
//   | "NOT_PASSED" // If keyword mapping not found then we use this error
//   | "OPERATION_NOT_FOUND_IN_CACHE_WITH_PROVIDER" // Covered by liveValidator
//   | "OPERATION_NOT_FOUND_IN_CACHE_WITH_API"
//   | "OPERATION_NOT_FOUND_IN_CACHE_WITH_VERB"
//   | "OPERATION_NOT_FOUND_IN_CACHE"
//   | "MULTIPLE_OPERATIONS_FOUND"
//   | "INVALID_RESPONSE_HEADER"
//   | "INVALID_REQUEST_PARAMETER"
// >;
const transformParamsKeyword = new Set([
    "pattern",
    "additionalProperties",
    "type",
    "format",
    "multipleOf",
    "required",
    constants_1.xmsAzureResource,
]);
const transformReverseParamsKeyword = new Set([
    "minLength",
    "maxLength",
    "maxItems",
    "minItems",
    "maxProperties",
    "minProperties",
    "minimum",
    "maximum",
    "exclusiveMinimum",
    "exclusiveMaximum",
    "discriminatorMap",
]);
const ajvErrorCodeToOavErrorCode = (error, ctx) => {
    var _a;
    const { keyword, parentSchema: parentSch } = error;
    const parentSchema = parentSch;
    let { params, data } = error;
    let result = {
        code: "NOT_PASSED",
        message: error.message,
        severity: 4 /* Severity.Verbose */,
    };
    // Workaround for incorrect ajv behavior.
    // See https://github.com/ajv-validator/ajv/blob/v6/lib/dot/custom.jst#L74
    if (error._realData !== undefined) {
        data = error._realData;
    }
    switch (keyword) {
        case "enum":
            const { allowedValues } = params;
            result =
                data === null && (parentSchema === null || parentSchema === void 0 ? void 0 : parentSchema.nullable)
                    ? undefined
                    : isEnumCaseMismatch(data, allowedValues)
                        ? (0, errorDefinitions_1.getOavErrorMeta)("ENUM_CASE_MISMATCH", { data })
                        : ((_a = parentSchema === null || parentSchema === void 0 ? void 0 : parentSchema[constants_1.xmsEnum]) === null || _a === void 0 ? void 0 : _a.modelAsString)
                            ? undefined
                            : (0, errorDefinitions_1.getOavErrorMeta)("ENUM_MISMATCH", { data });
            params = [data, allowedValues];
            break;
        case "readOnly":
        case constants_1.xmsMutability:
        case constants_1.xmsSecret:
            const param = {
                key: (0, context_1.getNameFromRef)(parentSchema),
                value: Array.isArray(data) ? data.join(",") : JSON.stringify(data),
            };
            params = [param.key, null];
            result =
                keyword === constants_1.xmsSecret
                    ? (0, errorDefinitions_1.getOavErrorMeta)("SECRET_PROPERTY", param)
                    : ctx.isResponse
                        ? (0, errorDefinitions_1.getOavErrorMeta)("WRITEONLY_PROPERTY_NOT_ALLOWED_IN_RESPONSE", param)
                        : (0, errorDefinitions_1.getOavErrorMeta)("READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST", param);
            break;
        case "oneOf":
            result =
                params.passingSchemas === null
                    ? (0, errorDefinitions_1.getOavErrorMeta)("ONE_OF_MISSING", {})
                    : (0, errorDefinitions_1.getOavErrorMeta)("ONE_OF_MULTIPLE", {});
            params = [];
            break;
        case "type":
            params.type = params.type.replace(",null", "");
            data = schemaType(data);
            break;
        case "discriminatorMap":
            data = params.discriminatorValue;
            break;
        case "maxLength":
        case "minLength":
        case "maxItems":
        case "minItems":
            data = data.length;
            break;
        case "maxProperties":
        case "minProperties":
            data = Object.keys(data).length;
            break;
        case "minimum":
        case "maximum":
        case "exclusiveMinimum":
        case "exclusiveMaximum":
            params = { limit: params.limit };
            break;
    }
    const code = errorKeywordsMapping[keyword];
    if (code !== undefined) {
        result = (0, errorDefinitions_1.getOavErrorMeta)(code, { ...params, data });
    }
    if (transformParamsKeyword.has(keyword)) {
        params = Object.values(params);
        if (typeof data !== "object") {
            params.push(data);
        }
    }
    else if (transformReverseParamsKeyword.has(keyword)) {
        params = Object.values(params);
        params.unshift(data);
    }
    if (result !== undefined) {
        result.params = params;
    }
    return result;
};
exports.ajvErrorCodeToOavErrorCode = ajvErrorCodeToOavErrorCode;
const isEnumCaseMismatch = (data, enumList) => {
    if (typeof data !== "string") {
        return false;
    }
    data = data.toLowerCase();
    for (const val of enumList) {
        if (typeof val === "string" && val.toLowerCase() === data) {
            return true;
        }
    }
    return false;
};
const schemaType = (what) => {
    const to = typeof what;
    if (to === "object") {
        if (what === null) {
            return "null";
        }
        if (Array.isArray(what)) {
            return "array";
        }
        return "object"; // typeof what === 'object' && what === Object(what) && !Array.isArray(what);
    }
    if (to === "number") {
        if (Number.isFinite(what)) {
            if (what % 1 === 0) {
                return "integer";
            }
            else {
                return "number";
            }
        }
        if (Number.isNaN(what)) {
            return "not-a-number";
        }
        return "unknown-number";
    }
    return to; // undefined, boolean, string, function
};
