// oav/dist/lib/swaggerValidator/modelValidator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewModelValidator = exports.SwaggerExampleValidator = void 0;
const inversify_1 = require("inversify");
const openapi_tools_common_1 = require("@azure-tools/openapi-tools-common");
const openapiToolsCommon = __importStar(require("@azure-tools/openapi-tools-common"));
const json_pointer_1 = __importDefault(require("json-pointer"));
const lodash_1 = require("lodash");
const C = __importStar(require("../util/constants"));
const inversifyUtils_1 = require("../inversifyUtils");
const liveValidatorLoader_1 = require("../liveValidation/liveValidatorLoader");
const jsonLoader_1 = require("../swagger/jsonLoader");
const suppressionLoader_1 = require("../swagger/suppressionLoader");
const swaggerTypes_1 = require("../swagger/swaggerTypes");
const errorDefinitions_1 = require("../util/errorDefinitions");
const traverseSwagger_1 = require("../transform/traverseSwagger");
const xmsPathsTransformer_1 = require("../transform/xmsPathsTransformer");
const resolveNestedDefinitionTransformer_1 = require("../transform/resolveNestedDefinitionTransformer");
const referenceFieldsTransformer_1 = require("../transform/referenceFieldsTransformer");
const pathRegexTransformer_1 = require("../transform/pathRegexTransformer");
const discriminatorTransformer_1 = require("../transform/discriminatorTransformer");
const allOfTransformer_1 = require("../transform/allOfTransformer");
const noAdditionalPropertiesTransformer_1 = require("../transform/noAdditionalPropertiesTransformer");
const nullableTransformer_1 = require("../transform/nullableTransformer");
const pureObjectTransformer_1 = require("../transform/pureObjectTransformer");
const context_1 = require("../transform/context");
const transformer_1 = require("../transform/transformer");
const operationValidator_1 = require("../liveValidation/operationValidator");
const logging_1 = require("../util/logging");
const jsonUtils_1 = require("../util/jsonUtils");
const utils_1 = require("../util/utils");
const validationResultSource_1 = require("../util/validationResultSource");
let SwaggerExampleValidator = class SwaggerExampleValidator {
    constructor(_opts, jsonLoader, suppressionLoader, liveValidatorLoader, schemaValidator) {
        this.jsonLoader = jsonLoader;
        this.suppressionLoader = suppressionLoader;
        this.liveValidatorLoader = liveValidatorLoader;
        this.schemaValidator = schemaValidator;
        this.errors = [];
        this.getNotSuppressedErrorPath = (err) => {
            let node;
            let jsonRef;
            if (err.source.jsonRef === undefined) {
                return undefined;
            }
            const idx = err.source.jsonRef.indexOf("#");
            jsonRef = err.source.jsonRef.substr(idx + 1);
            /*eslint no-constant-condition: ["error", { "checkLoops": false }]*/
            while (true) {
                try {
                    if (err.source.url === this.swagger._filePath) {
                        node = json_pointer_1.default.get(this.swagger, jsonRef);
                    }
                    else {
                        const externalSwagger = this.jsonLoader.getFileContentFromCache(err.source.url);
                        node = json_pointer_1.default.get(externalSwagger, jsonRef);
                    }
                    const isSuppressed = (0, suppressionLoader_1.isSuppressedInPath)(node, err.code, err.message);
                    if (isSuppressed) {
                        return undefined;
                    }
                    else {
                        break;
                    }
                }
                catch (e) {
                    let isContinue = false;
                    // the jsonRef will include non-existed path, so it needs to walk back to
                    // exclude the unexisted path
                    if (e.message.includes("Invalid reference token:")) {
                        const token = e.message.substring("Invalid reference token:".length + 1);
                        const index = jsonRef.lastIndexOf(token);
                        if (index > 0) {
                            jsonRef = jsonRef.substring(0, index);
                            if (jsonRef.endsWith(".") || jsonRef.endsWith("/")) {
                                jsonRef = jsonRef.substring(0, jsonRef.length - 1);
                            }
                            isContinue = true;
                        }
                    }
                    // if it's not the case of containing unexisted path, then throw this error
                    if (isContinue === false) {
                        logging_1.log.error(`Exception in filtering not suppressed error path:${jsonRef}.`);
                        throw e;
                    }
                }
            }
            return node;
        };
        this.issueFromErrorCode = (operationId, examplePath, code, param, schemaObj, schemaJsonPath, exampleObj, exampleJsonPath, source) => {
            const meta = (0, errorDefinitions_1.getOavErrorMeta)(code, param);
            const schemaInfo = schemaObj ? (0, openapi_tools_common_1.getInfo)(schemaObj) : schemaObj;
            const exampleInfo = exampleObj ? (0, openapi_tools_common_1.getInfo)(exampleObj) : exampleObj;
            return {
                code,
                message: meta.message,
                schemaUrl: this.specPath,
                exampleUrl: examplePath,
                schemaPosition: schemaInfo === null || schemaInfo === void 0 ? void 0 : schemaInfo.position,
                schemaJsonPath: schemaJsonPath,
                examplePosition: exampleInfo === null || exampleInfo === void 0 ? void 0 : exampleInfo.position,
                exampleJsonPath: exampleJsonPath,
                severity: meta.severity,
                source: source !== null && source !== void 0 ? source : validationResultSource_1.ValidationResultSource.GLOBAL,
                operationId,
            };
        };
        this.validateBodyType = (operationId, responseSchema, body, exampleFileUrl, responseDefinition, exampleResponsesObj, exampleResponseStatusCode) => {
            var _a, _b;
            let expectedType = responseSchema.schema.type;
            const actualType = typeof body;
            let bodySchema = {};
            if (responseSchema.schema.format === "file" || expectedType === "file") {
                // ignore validation in case of file type
                return true;
            }
            else if (expectedType === undefined) {
                // in case of body providing primitive type
                bodySchema = this.jsonLoader.resolveRefObj(responseSchema.schema);
                // set excpectedType as 'object' if the schema has properties key or additionalProperties key
                if (bodySchema.type === undefined &&
                    (Object.keys(bodySchema).includes("properties") ||
                        Object.keys(bodySchema).includes("additionalProperties"))) {
                    expectedType = "object";
                }
                else {
                    expectedType = bodySchema.type;
                }
            }
            let invalidTypeError = false;
            if ((actualType === "number" && expectedType === "integer") ||
                (actualType === "object" && expectedType === "array") ||
                (((_a = Object.keys(bodySchema)) === null || _a === void 0 ? void 0 : _a.length) === 0 && ((_b = Object.keys(body)) === null || _b === void 0 ? void 0 : _b.length) === 0)) {
                invalidTypeError = false;
            }
            else if (expectedType !== actualType) {
                invalidTypeError = true;
            }
            if (invalidTypeError) {
                const meta = (0, errorDefinitions_1.getOavErrorMeta)(errorDefinitions_1.ErrorCodeConstants.INVALID_TYPE, {
                    type: expectedType,
                    data: actualType,
                });
                this.addErrorsFromErrorCode(operationId, exampleFileUrl, meta, responseDefinition, undefined, exampleResponsesObj, `$response.${exampleResponseStatusCode}/body`, validationResultSource_1.ValidationResultSource.RESPONSE);
                return false;
            }
            return true;
        };
        this.validateContentType = (operationId, examplePath, allowedContentTypes, headers, isRequest, statusCode, schemaObj, exampleObj) => {
            var _a;
            const contentType = ((_a = headers["content-type"]) === null || _a === void 0 ? void 0 : _a.split(";")[0]) ||
                (isRequest ? undefined : "application/octet-stream");
            if (contentType !== undefined && !allowedContentTypes.includes(contentType)) {
                // in some cases, produces value could have colon in type like 'application/json;odata=minimalmetadata'
                for (const allowedContentType of allowedContentTypes) {
                    if (allowedContentType.includes(";")) {
                        const subAllowedContentType = allowedContentType.split(";")[0];
                        if (subAllowedContentType.includes(contentType)) {
                            return;
                        }
                    }
                }
                this.errors.push(this.issueFromErrorCode(operationId, examplePath, "INVALID_CONTENT_TYPE", {
                    contentType,
                    supported: allowedContentTypes.join(", "),
                }, schemaObj, undefined, exampleObj, `responses/${statusCode}/headers`, validationResultSource_1.ValidationResultSource.RESPONSE));
            }
        };
        this.validateLroOperation = (examplePath, operation, statusCode, headers, exampleObj) => {
            if (operation["x-ms-long-running-operation"] === true && (0, lodash_1.parseInt)(statusCode, 10) < 300) {
                if (operation._method === "post") {
                    if (statusCode === "202" || statusCode === "201") {
                        this.validateLroHeader(examplePath, operation, statusCode, headers);
                    }
                    else if (statusCode !== "200" && statusCode !== "204") {
                        this.errors.push(this.issueFromErrorCode(operation.operationId, examplePath, "LRO_RESPONSE_CODE", { statusCode }, operation.responses, undefined, exampleObj, `responses/${statusCode}`, validationResultSource_1.ValidationResultSource.RESPONSE));
                    }
                }
                else if (operation._method === "patch") {
                    if (statusCode === "202" || statusCode === "201") {
                        this.validateLroHeader(examplePath, operation, statusCode, headers);
                    }
                    else if (statusCode !== "200") {
                        this.errors.push(this.issueFromErrorCode(operation.operationId, examplePath, "LRO_RESPONSE_CODE", { statusCode }, operation.responses, undefined, exampleObj, `responses/${statusCode}`, validationResultSource_1.ValidationResultSource.RESPONSE));
                    }
                }
                else if (operation._method === "delete") {
                    if (statusCode === "202") {
                        this.validateLroHeader(examplePath, operation, statusCode, headers);
                    }
                    else if (statusCode !== "200" && statusCode !== "204") {
                        this.errors.push(this.issueFromErrorCode(operation.operationId, examplePath, "LRO_RESPONSE_CODE", { statusCode }, operation.responses, undefined, exampleObj, `responses/${statusCode}`, validationResultSource_1.ValidationResultSource.RESPONSE));
                    }
                }
                else if (operation._method === "put") {
                    if (statusCode !== "200" && statusCode !== "201" && statusCode !== "202") {
                        this.errors.push(this.issueFromErrorCode(operation.operationId, examplePath, "LRO_RESPONSE_CODE", { statusCode }, operation.responses, undefined, exampleObj, `responses/${statusCode}`, validationResultSource_1.ValidationResultSource.RESPONSE));
                    }
                }
            }
        };
        this.validateLroHeader = (examplePath, operation, statusCode, headers, exampleObj) => {
            if (statusCode === "201") {
                // Ignore LRO header check cause RPC says azure-AsyncOperation is optional if using 201/200+ provisioningState
                return;
            }
            const provider = (0, utils_1.getProviderFromSpecPath)(this.specPath);
            const providerType = provider === null || provider === void 0 ? void 0 : provider.type;
            if (providerType !== undefined && this.isMissingHeader(providerType, headers)) {
                this.errors.push(this.issueFromErrorCode(operation.operationId, examplePath, "LRO_RESPONSE_HEADER", {
                    header: providerType === "resource-manager"
                        ? "location or azure-AsyncOperation"
                        : "Operation-Id or Operation-Location or azure-AsyncOperation", // providerType === "data-plane"
                }, operation.responses, undefined, exampleObj, `responses/${statusCode}/headers`, validationResultSource_1.ValidationResultSource.RESPONSE));
            }
        };
        this.isMissingHeader = (providerType, header) => {
            const headerConfig = {
                "resource-manager": {
                    checkFirst: ["Location", "location", "azure-AsyncOperation", "azure-asyncoperation"],
                    checkSecond: [],
                },
                "data-plane": {
                    checkFirst: ["Operation-Id", "operation-id", "Operation-Location", "operation-location"],
                    checkSecond: ["azure-AsyncOperation", "azure-asyncoperation"],
                },
            };
            const config = headerConfig[providerType];
            for (const checkFirst of config.checkFirst) {
                const value = header[checkFirst];
                if (typeof value === "string" && value !== "") {
                    return false;
                }
            }
            for (const checkSecond of config.checkSecond) {
                const value = header[checkSecond];
                if (typeof value === "string" && value !== "") {
                    return false;
                }
            }
            return true;
        };
    }
    async validateOperations(operationIds) {
        try {
            if (operationIds !== null &&
                operationIds !== undefined &&
                typeof operationIds.valueOf() !== "string") {
                throw new Error(`please pass in operationIds parameter correctly. It must be a string separated by comma if containing multiple operationIds.`);
            }
            let operationIdArray = [];
            if (operationIds !== undefined) {
                operationIdArray = operationIds.trim().split(",");
            }
            // Parse the swagger resolving the ref
            await this.loadSwagger(this.specPath, false);
            if (this.swagger === undefined || this.errors.length > 0) {
                return;
            }
            await this.suppressionLoader.load(this.swagger);
            await this.transformSwagger(this.swagger);
            await this.liveValidatorLoader.buildAjvValidator(this.swagger);
            await (0, traverseSwagger_1.traverseSwaggerAsync)(this.swagger, {
                onOperation: async (_operation, _path, _method) => {
                    if (operationIdArray.length === 0 ||
                        (operationIdArray.length > 0 && operationIdArray.includes(_operation.operationId))) {
                        if (!_operation["x-ms-examples"]) {
                            const meta = (0, errorDefinitions_1.getOavErrorMeta)(errorDefinitions_1.ErrorCodeConstants.XMS_EXAMPLE_NOTFOUND_ERROR, {
                                operationId: _operation.operationId,
                            });
                            this.addErrorsFromErrorCode(_operation.operationId, undefined, meta, _operation);
                            return;
                        }
                        await this.validateOperation(_operation);
                    }
                },
            });
        }
        catch (e) {
            logging_1.log.error(`validateOperations - ErrorMessage:${e === null || e === void 0 ? void 0 : e.message}.ErrorStack:${e === null || e === void 0 ? void 0 : e.stack}`);
            throw e;
        }
    }
    async initialize(specPath) {
        if (specPath === null ||
            specPath === undefined ||
            typeof specPath.valueOf() !== "string" ||
            !specPath.trim().length) {
            throw new Error("specPath is a required parameter of type string and it cannot be an empty string.");
        }
        this.specPath = (0, utils_1.checkAndResolveGithubUrl)(specPath);
    }
    async validateOperation(operation) {
        // load example content
        let exampleContent;
        let exampleFileUrl;
        for (const scenario of Object.keys(operation["x-ms-examples"])) {
            const mockUrl = operation["x-ms-examples"][scenario].$ref;
            if (mockUrl === undefined) {
                throw new Error(`$ref is undefined in x-ms-examples defintion for operation ${operation.operationId}.`);
            }
            exampleContent = this.jsonLoader.resolveMockedFile(mockUrl);
            exampleFileUrl = this.jsonLoader.getRealPath(mockUrl);
            this.validateRequest(operation, exampleContent, exampleFileUrl);
            this.validateResponse(operation, exampleContent, exampleFileUrl);
        } // end of scenario for loop
    }
    validateResponse(operation, exampleContent, exampleFileUrl) {
        const exampleResponsesObj = exampleContent.responses;
        if (exampleResponsesObj === null ||
            exampleResponsesObj === undefined ||
            typeof exampleResponsesObj !== "object") {
            throw new Error(`For operation:${operation.operationId}, responses in example file ${exampleFileUrl} cannot be null or undefined and must be of type 'object'.`);
        }
        const definedResponsesInSwagger = [];
        for (const statusCode of openapiToolsCommon.keys(operation.responses)) {
            definedResponsesInSwagger.push(statusCode);
        }
        // loop for each response code
        for (const exampleResponseStatusCode of openapiToolsCommon.keys(exampleResponsesObj)) {
            const responseDefinition = operation.responses;
            const responseSchema = responseDefinition[exampleResponseStatusCode];
            if (!responseSchema) {
                const meta = (0, errorDefinitions_1.getOavErrorMeta)(errorDefinitions_1.ErrorCodeConstants.RESPONSE_STATUS_CODE_NOT_IN_SPEC, {
                    exampleResponseStatusCode,
                    operationId: operation.operationId,
                });
                this.addErrorsFromErrorCode(operation.operationId, exampleFileUrl, meta, responseDefinition, undefined, exampleResponsesObj, `$response.${exampleResponseStatusCode}`, validationResultSource_1.ValidationResultSource.RESPONSE);
                continue;
            }
            // remove response entry after validation
            const responseIndex = definedResponsesInSwagger.indexOf(exampleResponseStatusCode);
            if (responseIndex > -1) {
                definedResponsesInSwagger.splice(responseIndex, 1);
            }
            const exampleResponseHeaders = exampleResponsesObj[exampleResponseStatusCode].headers || {};
            const exampleResponseBody = exampleResponsesObj[exampleResponseStatusCode].body;
            // Fail when example provides the response body but the swagger spec doesn't define the schema for the response.
            if (exampleResponseBody !== undefined) {
                if (!responseSchema.schema) {
                    // having response body but doesn't have schema defined in response swagger definition
                    const meta = (0, errorDefinitions_1.getOavErrorMeta)(errorDefinitions_1.ErrorCodeConstants.RESPONSE_SCHEMA_NOT_IN_SPEC, {
                        exampleResponseStatusCode,
                        operationId: operation.operationId,
                    });
                    this.addErrorsFromErrorCode(operation.operationId, exampleFileUrl, meta, responseDefinition, undefined, exampleResponsesObj, `$response.${exampleResponseStatusCode}/body`, validationResultSource_1.ValidationResultSource.RESPONSE);
                    continue;
                }
                else if (responseSchema.schema.type !== typeof exampleResponseBody) {
                    const validBody = this.validateBodyType(operation.operationId, responseSchema, exampleResponseBody, exampleFileUrl, responseDefinition, exampleResponsesObj, exampleResponseStatusCode);
                    if (!validBody) {
                        continue;
                    }
                }
            }
            else if (exampleResponseBody === undefined && responseSchema.schema) {
                // Fail when example doesn't provide the response body but the swagger spec define the schema for the response.
                const meta = (0, errorDefinitions_1.getOavErrorMeta)(errorDefinitions_1.ErrorCodeConstants.RESPONSE_BODY_NOT_IN_EXAMPLE, {
                    exampleResponseStatusCode,
                    operationId: operation.operationId,
                });
                this.addErrorsFromErrorCode(operation.operationId, exampleFileUrl, meta, responseDefinition, undefined, exampleResponsesObj, `$response.${exampleResponseStatusCode}`, validationResultSource_1.ValidationResultSource.RESPONSE);
                continue;
            }
            // validate headers
            const headers = (0, operationValidator_1.transformLiveHeader)(exampleResponseHeaders, responseSchema);
            this.validateLroOperation(exampleFileUrl, operation, exampleResponseStatusCode, headers, exampleResponseHeaders);
            if (responseSchema.schema !== undefined) {
                if (headers["content-type"] !== undefined) {
                    this.validateContentType(operation.operationId, exampleFileUrl, operation.produces, headers, false, exampleResponseStatusCode, operation, exampleResponseHeaders);
                }
                const validate = responseSchema._validate;
                const ctx = {
                    isResponse: true,
                    statusCode: exampleResponseStatusCode,
                    httpMethod: operation._method,
                };
                const ajvValidatorErrors = validate(ctx, {
                    headers,
                    body: exampleResponseBody,
                });
                this.schemaIssuesToModelValidationIssues(operation.operationId, false, ajvValidatorErrors, exampleFileUrl, exampleContent, undefined, exampleResponseStatusCode, operation._method);
            }
        }
        // report missing response in example
        for (const statusCodeInSwagger of definedResponsesInSwagger) {
            if (statusCodeInSwagger !== "default") {
                const meta = (0, errorDefinitions_1.getOavErrorMeta)(errorDefinitions_1.ErrorCodeConstants.RESPONSE_STATUS_CODE_NOT_IN_EXAMPLE, { statusCodeInSwagger, operationId: operation.operationId });
                this.addErrorsFromErrorCode(operation.operationId, exampleFileUrl, meta, operation.responses[statusCodeInSwagger], undefined, exampleResponsesObj, `response`, validationResultSource_1.ValidationResultSource.RESPONSE);
            }
        }
    }
    validateRequest(operation, exampleContent, exampleFileUrl) {
        var _a;
        const parameterizedHostDef = operation._path._spec["x-ms-parameterized-host"];
        const useSchemePrefix = parameterizedHostDef
            ? parameterizedHostDef.useSchemePrefix === undefined
                ? true
                : parameterizedHostDef.useSchemePrefix
            : null;
        const pathTemplate = operation._path._pathTemplate;
        const parameters = operation.parameters;
        const publicParameters = operation._path.parameters;
        const mergedParameters = [...(parameters !== null && parameters !== void 0 ? parameters : []), ...(publicParameters !== null && publicParameters !== void 0 ? publicParameters : [])];
        const pathParameters = {};
        let bodyParameter = {};
        const queryParameters = {};
        const formData = {};
        const exampleRequestHeaders = {};
        if (mergedParameters === undefined) {
            return;
        }
        for (const p of mergedParameters) {
            const parameter = this.jsonLoader.resolveRefObj(p);
            let parameterValue = exampleContent === null || exampleContent === void 0 ? void 0 : exampleContent.parameters[parameter.name];
            if (!parameterValue) {
                if (parameter.required) {
                    const meta = (0, errorDefinitions_1.getOavErrorMeta)(errorDefinitions_1.ErrorCodeConstants.REQUIRED_PARAMETER_EXAMPLE_NOT_FOUND, { operationId: operation.operationId, name: parameter.name });
                    this.addErrorsFromErrorCode(operation.operationId, exampleFileUrl, meta, operation, undefined, p);
                    break;
                }
                continue;
            }
            const location = parameter.in;
            if (location === "path") {
                if (location === "path" && parameterValue && typeof parameterValue === "string") {
                    // "/{scope}/scopes/resourceGroups/{resourceGroupName}" In the aforementioned path
                    // template, we will search for the path parameter based on it's name
                    // for example: "scope". Find it's index in the string and move backwards by 2 positions.
                    // If the character at that position is a forward slash "/" and
                    // the value for the parameter starts with a forward slash "/" then we have found the case
                    // where there will be duplicate forward slashes in the url.
                    if (pathTemplate.charAt(pathTemplate.indexOf(`${parameter.name}`) - 2) === "/" &&
                        parameterValue.startsWith("/")) {
                        const meta = (0, errorDefinitions_1.getOavErrorMeta)(errorDefinitions_1.ErrorCodeConstants.DOUBLE_FORWARD_SLASHES_IN_URL, {
                            operationId: operation.operationId,
                            parameterName: parameter.name,
                            parameterValue,
                            pathTemplate,
                        });
                        this.addErrorsFromErrorCode(operation.operationId, exampleFileUrl, meta, operation, undefined, p);
                        break;
                    }
                    // replacing characters that may cause validator failed  with empty string because this messes up Sways regex
                    // validation of path segment.
                    parameterValue = parameterValue.replace(/\//gi, "");
                    // replacing scheme that may cause validator failed when x-ms-parameterized-host enbaled & useSchemePrefix enabled
                    // because if useSchemePrefix enabled ,the parameter value in x-ms-parameterized-host should not has the scheme (http://|https://)
                    if (useSchemePrefix) {
                        parameterValue = parameterValue.replace(/^https{0,1}:/gi, "");
                    }
                }
                // todo skip url encoding
                pathParameters[parameter.name] = parameterValue;
            }
            else if (location === "query") {
                // validate the api version value
                if (parameter.name === "api-version" && parameterValue !== this.swagger.info.version) {
                    const meta = (0, errorDefinitions_1.getOavErrorMeta)("INVALID_REQUEST_PARAMETER", {
                        parameterName: "api-version",
                        apiVersion: parameterValue,
                    });
                    this.addErrorsFromErrorCode(operation.operationId, exampleFileUrl, meta, operation, undefined, exampleContent === null || exampleContent === void 0 ? void 0 : exampleContent.parameters, `$parameters["api-version"]`);
                    continue;
                }
                queryParameters[parameter.name] = parameterValue;
            }
            else if (location === "body") {
                if (((_a = parameter.schema) === null || _a === void 0 ? void 0 : _a.format) === "file") {
                    continue;
                }
                bodyParameter = parameterValue;
            }
            else if (location === "header") {
                exampleRequestHeaders[parameter.name] = parameterValue;
            }
            else if (location === "formData") {
                formData[parameter.name] = parameterValue;
            }
        } // end of mergedParameters for loop
        (0, operationValidator_1.transformMapValue)(queryParameters, operation._queryTransform);
        (0, operationValidator_1.transformMapValue)(pathParameters, operation._pathTransform);
        const validate = operation._validate;
        const ctx = { isResponse: false };
        const headers = (0, operationValidator_1.transformLiveHeader)(exampleRequestHeaders, operation);
        const ajvValidatorErrors = validate(ctx, {
            path: pathParameters,
            body: (0, operationValidator_1.transformBodyValue)(bodyParameter, operation),
            query: queryParameters,
            headers,
            formData,
        });
        this.schemaIssuesToModelValidationIssues(operation.operationId, true, ajvValidatorErrors, exampleFileUrl, exampleContent, mergedParameters);
    }
    async loadSwagger(swaggerFilePath, skipResolveRef) {
        try {
            this.swagger = (await this.jsonLoader.load(swaggerFilePath, skipResolveRef));
            this.swagger._filePath = swaggerFilePath;
        }
        catch (e) {
            if (typeof e.kind === "string") {
                const ex = e;
                const errInfo = (0, errorDefinitions_1.getOavErrorMeta)("JSON_PARSING_ERROR", { details: ex.code });
                this.errors.push({
                    code: errInfo.code,
                    message: errInfo.message,
                    schemaPosition: ex.position,
                    schemaUrl: ex.url,
                });
            }
            else if (e instanceof jsonLoader_1.JsonLoaderRefError) {
                const errInfo = (0, errorDefinitions_1.getOavErrorMeta)("UNRESOLVABLE_REFERENCE", { ref: e.ref });
                this.errors.push({
                    code: errInfo.code,
                    message: errInfo.message,
                    schemaPosition: e.position,
                    schemaUrl: e.url,
                });
            }
            else {
                throw e;
            }
            return;
        }
    }
    async transformSwagger(spec) {
        const transformCtx = (0, context_1.getTransformContext)(this.jsonLoader, this.schemaValidator, [
            xmsPathsTransformer_1.xmsPathsTransformer,
            resolveNestedDefinitionTransformer_1.resolveNestedDefinitionTransformer,
            referenceFieldsTransformer_1.referenceFieldsTransformer,
            pathRegexTransformer_1.pathRegexTransformer,
            discriminatorTransformer_1.discriminatorTransformer,
            allOfTransformer_1.allOfTransformer,
            noAdditionalPropertiesTransformer_1.noAdditionalPropertiesTransformer,
            nullableTransformer_1.nullableTransformer,
            pureObjectTransformer_1.pureObjectTransformer,
        ]);
        (0, transformer_1.applySpecTransformers)(spec, transformCtx);
        (0, transformer_1.applyGlobalTransformers)(transformCtx);
    }
    addErrorsFromErrorCode(operationId, exampleUrl, meta, schemaObj, schemaJsonPath, exampleObj, exampleJsonPath, source) {
        if ((0, suppressionLoader_1.isSuppressedInPath)(schemaObj, meta.id, meta.message) ||
            (0, suppressionLoader_1.isSuppressedInPath)(schemaObj, meta.code, meta.message)) {
            return;
        }
        const schemaInfo = (0, openapi_tools_common_1.getInfo)(schemaObj);
        const exampleInfo = (0, openapi_tools_common_1.getInfo)(exampleObj);
        this.errors.push({
            code: meta.code,
            message: meta.message,
            schemaUrl: this.specPath,
            exampleUrl,
            schemaPosition: schemaInfo === null || schemaInfo === void 0 ? void 0 : schemaInfo.position,
            schemaJsonPath: schemaJsonPath !== null && schemaJsonPath !== void 0 ? schemaJsonPath : schemaObj === null || schemaObj === void 0 ? void 0 : schemaObj[swaggerTypes_1.refSelfSymbol],
            examplePosition: exampleInfo === null || exampleInfo === void 0 ? void 0 : exampleInfo.position,
            exampleJsonPath: exampleJsonPath,
            severity: meta.severity,
            source: source !== null && source !== void 0 ? source : validationResultSource_1.ValidationResultSource.GLOBAL,
            operationId,
        });
    }
    schemaIssuesToModelValidationIssues(operationId, isRequest, errors, exampleUrl, exampleContent, parameters, statusCode, httpMethod) {
        var _a, _b, _c;
        for (const err of errors) {
            // ignore below schema errors
            if ((err.code === "NOT_PASSED" &&
                (err.message.includes('should match "else" schema') ||
                    err.message.includes('should match "then" schema'))) ||
                (err.code === "SECRET_PROPERTY" && httpMethod === "post")) {
                continue;
            }
            let isSuppressed = false;
            let schemaPosition;
            let examplePosition;
            let externalSwagger;
            let schemaUrl = this.specPath;
            const exampleJsonPaths = [];
            if (isRequest) {
                for (const p of parameters) {
                    const parameter = this.jsonLoader.resolveRefObj(p);
                    if (err.schemaPath.indexOf(parameter.name) > 0 ||
                        (err.jsonPathsInPayload.length > 0 &&
                            (err.jsonPathsInPayload[0].includes(parameter.name) ||
                                (err.jsonPathsInPayload[0].includes(".body") && parameter.in === "body")))) {
                        schemaPosition = err.source.position;
                        if (parameter.in !== "body") {
                            const index = err.schemaPath.indexOf(parameter.name);
                            err.schemaPath = index > 0 ? err.schemaPath.substr(index) : err.schemaPath;
                        }
                        if ((0, suppressionLoader_1.isSuppressedInPath)(parameter, err.code, err.message)) {
                            isSuppressed = true;
                        }
                        exampleJsonPaths.push(`$parameters.${parameter.name}`);
                        break;
                    }
                }
            }
            else {
                // response error case
                let sourceSwagger = this.swagger;
                if (err.source.url !== this.swagger._filePath) {
                    // use external swagger when the error points to external file
                    sourceSwagger = this.jsonLoader.getFileContentFromCache(err.source.url);
                    schemaUrl = err.source.url;
                }
                // check x-nullable value when body is null
                if (err.jsonPathsInPayload.length === 1 && err.jsonPathsInPayload[0] === ".body") {
                    const idx = (_a = err.source.jsonRef) === null || _a === void 0 ? void 0 : _a.indexOf("#");
                    if (idx !== undefined && idx !== -1) {
                        const jsonRef = (_b = err.source.jsonRef) === null || _b === void 0 ? void 0 : _b.substr(idx + 1);
                        const bodySchema = json_pointer_1.default.get(sourceSwagger, jsonRef);
                        if ((bodySchema === null || bodySchema === void 0 ? void 0 : bodySchema[C.xNullable]) === true) {
                            continue;
                        }
                    }
                }
                if (err.code === "MISSING_RESOURCE_ID" &&
                    exampleContent.responses[statusCode].body &&
                    Object.keys(exampleContent.responses[statusCode].body).length === 0) {
                    // ignore this error when whole body of response is empty
                    continue;
                }
                const node = this.getNotSuppressedErrorPath(err);
                if (node === undefined) {
                    continue;
                }
                if (externalSwagger === undefined) {
                    schemaPosition = (_c = (0, openapi_tools_common_1.getInfo)(node)) === null || _c === void 0 ? void 0 : _c.position;
                }
                else {
                    schemaPosition = err.source.position;
                }
                for (let path of err.jsonPathsInPayload) {
                    // If parameter name includes ".", path should use "[]" for better understand.
                    for (const parameter of err.params) {
                        if (typeof parameter === "string" &&
                            path.includes(`.${parameter}`) &&
                            parameter.includes(".")) {
                            path = path.substring(0, path.indexOf(`.${parameter}`)) + `['${parameter}']`;
                        }
                    }
                    exampleJsonPaths.push(`$responses.${statusCode}${path}`);
                }
            }
            if (!isSuppressed) {
                for (const jsonPath of exampleJsonPaths) {
                    examplePosition = (0, jsonUtils_1.getFilePositionFromJsonPath)(exampleContent, jsonPath);
                    this.errors.push({
                        code: err.code,
                        message: err.message,
                        schemaUrl,
                        exampleUrl,
                        schemaPosition: schemaPosition,
                        schemaJsonPath: err.schemaPath,
                        examplePosition,
                        exampleJsonPath: jsonPath,
                        severity: err.severity,
                        source: isRequest ? validationResultSource_1.ValidationResultSource.REQUEST : validationResultSource_1.ValidationResultSource.RESPONSE,
                        operationId,
                    });
                }
            }
        }
    }
};
exports.SwaggerExampleValidator = SwaggerExampleValidator;
exports.SwaggerExampleValidator = SwaggerExampleValidator = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(inversifyUtils_1.TYPES.opts)),
    __param(4, (0, inversify_1.inject)(inversifyUtils_1.TYPES.schemaValidator)),
    __metadata("design:paramtypes", [Object, jsonLoader_1.JsonLoader,
        suppressionLoader_1.SuppressionLoader,
        liveValidatorLoader_1.LiveValidatorLoader, Object])
], SwaggerExampleValidator);
// load all error codes of model validation errors to as suppression option
const loadSuppression = [];
for (const errorCode of Object.keys(errorDefinitions_1.modelValidationErrors)) {
    const meta = errorDefinitions_1.modelValidationErrors[errorCode];
    if ("id" in meta) {
        loadSuppression.push(meta.id);
    }
    loadSuppression.push(errorCode);
}
// Set 'isArmCall flag to true so that the special ARM rules can be applied to examples validation too'
const defaultOpts = {
    eraseDescription: false,
    eraseXmsExamples: false,
    useJsonParser: true,
    loadSuppression,
    isArmCall: true,
};
// Compatible wrapper for old ModelValidator
class NewModelValidator {
    constructor(specPath) {
        this.specPath = specPath;
        this.result = [];
        const container = (0, inversifyUtils_1.inversifyGetContainer)();
        this.validator = (0, inversifyUtils_1.inversifyGetInstance)(SwaggerExampleValidator, {
            ...defaultOpts,
            container,
        });
        this.result = this.validator.errors;
    }
    async initialize() {
        await this.validator.initialize(this.specPath);
        // API compatible
        return null;
    }
    async validateOperations(operationIds) {
        await this.validator.validateOperations(operationIds);
    }
}
exports.NewModelValidator = NewModelValidator;
