// oav/dist/lib/swaggerValidator/semanticValidator.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SemanticValidator = exports.SwaggerSemanticValidator = void 0;
const openapi_tools_common_1 = require("@azure-tools/openapi-tools-common");
const inversify_1 = require("inversify");
const swagger_extensions_json_1 = __importDefault(require("@autorest/schemas/swagger-extensions.json"));
const example_schema_json_1 = __importDefault(require("@autorest/schemas/example-schema.json"));
const json_pointer_1 = __importDefault(require("json-pointer"));
const inversifyUtils_1 = require("../inversifyUtils");
const jsonLoader_1 = require("../swagger/jsonLoader");
const swaggerTypes_1 = require("../swagger/swaggerTypes");
const errorDefinitions_1 = require("../util/errorDefinitions");
const liveValidatorLoader_1 = require("../liveValidation/liveValidatorLoader");
const context_1 = require("../transform/context");
const referenceFieldsTransformer_1 = require("../transform/referenceFieldsTransformer");
const resolveNestedDefinitionTransformer_1 = require("../transform/resolveNestedDefinitionTransformer");
const xmsPathsTransformer_1 = require("../transform/xmsPathsTransformer");
const transformer_1 = require("../transform/transformer");
const traverseSwagger_1 = require("../transform/traverseSwagger");
const fileLoader_1 = require("../swagger/fileLoader");
const jsonUtils_1 = require("../util/jsonUtils");
const pathRegexTransformer_1 = require("../transform/pathRegexTransformer");
const discriminatorTransformer_1 = require("../transform/discriminatorTransformer");
const allOfTransformer_1 = require("../transform/allOfTransformer");
const noAdditionalPropertiesTransformer_1 = require("../transform/noAdditionalPropertiesTransformer");
const nullableTransformer_1 = require("../transform/nullableTransformer");
const pureObjectTransformer_1 = require("../transform/pureObjectTransformer");
const suppressionLoader_1 = require("../swagger/suppressionLoader");
const constants_1 = require("../util/constants");
const swagger_2_0_json_1 = __importDefault(require("./swagger-2.0.json"));
const loadSuppression = [];
for (const errorCode of Object.keys(errorDefinitions_1.semanticValidationErrors)) {
    const meta = errorDefinitions_1.semanticValidationErrors[errorCode];
    if ("id" in meta) {
        loadSuppression.push(meta.id);
    }
}
// Set isArmCall flag to true so that the ARM rules schema will be applied to swaggers too
const defaultOpts = {
    eraseDescription: false,
    eraseXmsExamples: false,
    useJsonParser: true,
    loadSuppression,
    isArmCall: true,
};
let SwaggerSemanticValidator = class SwaggerSemanticValidator {
    constructor(_opts, jsonLoader, fileLoader, suppressionLoader, liveValidatorLoader, schemaValidator) {
        this.jsonLoader = jsonLoader;
        this.fileLoader = fileLoader;
        this.suppressionLoader = suppressionLoader;
        this.liveValidatorLoader = liveValidatorLoader;
        this.schemaValidator = schemaValidator;
    }
    async initialize() {
        this.fileLoader.preloadExtraFile("https://raw.githubusercontent.com/Azure/autorest/master/schema/example-schema.json", JSON.stringify(example_schema_json_1.default));
        this.fileLoader.preloadExtraFile("http://json.schemastore.org/swagger-2.0", //DevSkim: ignore DS137138
        JSON.stringify(swagger_2_0_json_1.default));
        const properties = swagger_extensions_json_1.default.properties;
        properties[jsonLoader_1.$id] = {};
        properties._filePath = {};
        this.validateSwaggerSch = await this.schemaValidator.compileAsync(swagger_extensions_json_1.default);
    }
    async validateSwaggerSpec(swaggerFilePath) {
        const errors = await this.validateSwaggerSpecPipeline(swaggerFilePath);
        return errors;
    }
    async validateSwaggerSpecPipeline(swaggerFilePath) {
        const errors = [];
        try {
            const swagger = await this.loadSwagger(swaggerFilePath, errors);
            if (swagger === undefined || errors.length > 0) {
                return errors;
            }
            await this.suppressionLoader.load(swagger);
            // validate x-ms-* extensions
            await this.validateSwaggerSchema(swagger, errors);
            if (errors.length > 0) {
                return errors;
            }
            // compile swagger schema
            const transformCtx = await this.validateCompile(swagger, errors);
            await this.validateDiscriminator(transformCtx, errors);
            await this.validateDefaultValue(transformCtx, swagger._filePath, errors);
            await this.validateSchemaRequiredProperties(transformCtx, swagger._filePath, errors);
            await this.validateOperation(swagger, errors);
        }
        catch (e) {
            const errInfo = (0, errorDefinitions_1.getOavErrorMeta)("INTERNAL_ERROR", { message: `${e.message}\n${e.stack}` });
            errors.unshift({
                code: errInfo.code,
                message: errInfo.message,
                url: swaggerFilePath,
            });
        }
        return errors;
    }
    async loadSwagger(swaggerFilePath, errors) {
        try {
            const swagger = (await this.jsonLoader.load(swaggerFilePath));
            swagger._filePath = swaggerFilePath;
            return swagger;
        }
        catch (e) {
            if (typeof e.kind === "string") {
                const ex = e;
                const errInfo = (0, errorDefinitions_1.getOavErrorMeta)("JSON_PARSING_ERROR", { details: ex.code });
                errors.push({
                    code: errInfo.code,
                    message: errInfo.message,
                    position: ex.position,
                    url: ex.url,
                });
            }
            else if (e instanceof jsonLoader_1.JsonLoaderRefError) {
                const errInfo = (0, errorDefinitions_1.getOavErrorMeta)("UNRESOLVABLE_REFERENCE", { ref: e.ref });
                errors.push({
                    code: errInfo.code,
                    message: errInfo.message,
                    position: e.position,
                    url: e.url,
                });
            }
            else {
                throw e;
            }
            return;
        }
    }
    async validateSwaggerSchema(swagger, errors) {
        const result = this.validateSwaggerSch({}, swagger);
        const rootInfo = (0, openapi_tools_common_1.getRootObjectInfo)((0, openapi_tools_common_1.getInfo)(swagger));
        this.addErrorsFromSchemaValidation(result, rootInfo.url, swagger, errors);
    }
    addErrorsFromSchemaValidation(result, url, rootObj, errors) {
        const existedJsonPaths = [];
        for (const err of result) {
            // ignore below schema errors
            if (err.code === "NOT_PASSED" &&
                (err.message.includes('should match "else" schema') ||
                    err.message.includes('should match "then" schema'))) {
                continue;
            }
            else if (err.code === "NOT_PASSED" && err.schemaPath === "#/additionalProperties/not") {
                err.message = "path DOES NOT start with /";
            }
            err.jsonPathsInPayload = err.jsonPathsInPayload.filter((jsonPath) => {
                let node;
                /*eslint no-constant-condition: ["error", { "checkLoops": false }]*/
                while (true) {
                    try {
                        node = json_pointer_1.default.get(rootObj, (0, jsonUtils_1.jsonPathToPointer)(jsonPath));
                        const isSuppressed = (0, suppressionLoader_1.isSuppressedInPath)(node, err.code, err.message);
                        if (!isSuppressed) {
                            existedJsonPaths.push(jsonPath);
                        }
                        return !isSuppressed;
                    }
                    catch (e) {
                        let isContinue = false;
                        // the jsonPathsInPayload will include non-existed path, so it needs to walk back to
                        // exclude the unexisted path
                        if (e.message.includes("Invalid reference token:")) {
                            const token = e.message.substring("Invalid reference token:".length + 1);
                            const index = jsonPath.lastIndexOf(token);
                            if (index > 0) {
                                jsonPath = jsonPath.substring(0, index);
                                if (jsonPath.endsWith(".") || jsonPath.endsWith("/")) {
                                    jsonPath = jsonPath.substring(0, jsonPath.length - 1);
                                }
                                isContinue = true;
                            }
                        }
                        // if it's not the case of containing unexisted path, then throw this error
                        if (isContinue === false) {
                            throw e;
                        }
                    }
                }
            });
            if (err.jsonPathsInPayload.length === 0) {
                continue;
            }
            const jsonPath = existedJsonPaths[0];
            const position = (0, jsonUtils_1.getFilePositionFromJsonPath)(rootObj, jsonPath);
            errors.push({
                code: err.code,
                message: err.message,
                url,
                position,
                jsonPath,
            });
        }
    }
    addErrorsFromErrorCode(errors, url, meta, obj, jsonPath) {
        if ((0, suppressionLoader_1.isSuppressedInPath)(obj, meta.id, meta.message) ||
            (0, suppressionLoader_1.isSuppressedInPath)(obj, meta.code, meta.message)) {
            return;
        }
        const info = (0, openapi_tools_common_1.getInfo)(obj);
        errors.push({
            code: meta.code,
            message: meta.message,
            url,
            position: info === null || info === void 0 ? void 0 : info.position,
            jsonPath: jsonPath !== null && jsonPath !== void 0 ? jsonPath : obj === null || obj === void 0 ? void 0 : obj[swaggerTypes_1.refSelfSymbol],
        });
    }
    async validateCompile(spec, errors) {
        const transformCtx = (0, context_1.getTransformContext)(this.jsonLoader, this.schemaValidator, [
            xmsPathsTransformer_1.xmsPathsTransformer,
            resolveNestedDefinitionTransformer_1.resolveNestedDefinitionTransformer,
            referenceFieldsTransformer_1.referenceFieldsTransformer,
            pathRegexTransformer_1.pathRegexTransformer,
            discriminatorTransformer_1.discriminatorTransformer,
            allOfTransformer_1.allOfTransformer,
            noAdditionalPropertiesTransformer_1.noAdditionalPropertiesTransformer,
            nullableTransformer_1.nullableTransformer,
            pureObjectTransformer_1.pureObjectTransformer,
        ]);
        (0, transformer_1.applySpecTransformers)(spec, transformCtx);
        (0, transformer_1.applyGlobalTransformers)(transformCtx);
        await (0, traverseSwagger_1.traverseSwaggerAsync)(spec, {
            onOperation: async (operation) => {
                try {
                    await this.liveValidatorLoader.getRequestValidator(operation);
                }
                catch (e) {
                    const info = (0, openapi_tools_common_1.getInfo)(operation);
                    errors.push({
                        code: "INTERNAL_ERROR",
                        message: `Failed to compile validator on operation\n${operation.operationId} ${operation._method}\n${e.message}\n${e.stack}`,
                        url: spec._filePath,
                        position: info === null || info === void 0 ? void 0 : info.position,
                    });
                }
            },
            onResponse: async (response, operation, _, statusCode) => {
                try {
                    await this.liveValidatorLoader.getResponseValidator(response);
                }
                catch (e) {
                    const info = (0, openapi_tools_common_1.getInfo)(operation);
                    errors.push({
                        code: "INTERNAL_ERROR",
                        message: `Failed to compile validator on operation response\n${statusCode} ${operation.operationId} ${operation._method}\n${e.message}\n${e.stack}`,
                        url: spec._filePath,
                        position: info === null || info === void 0 ? void 0 : info.position,
                    });
                }
            },
        });
        return transformCtx;
    }
    async validateDiscriminator(transformCtx, errors) {
        var _a, _b;
        const { objSchemas, jsonLoader } = transformCtx;
        for (const sch of objSchemas) {
            const d = sch.discriminator;
            if (d === undefined) {
                continue;
            }
            const info = (0, openapi_tools_common_1.getInfo)(sch);
            const rootInfo = (0, openapi_tools_common_1.getRootObjectInfo)(info);
            if (((_a = sch.required) === null || _a === void 0 ? void 0 : _a.find((x) => x === d)) === undefined) {
                const meta = (0, errorDefinitions_1.getOavErrorMeta)("DISCRIMINATOR_NOT_REQUIRED", { property: d });
                this.addErrorsFromErrorCode(errors, rootInfo.url, meta, sch);
            }
            if (((_b = sch.properties) === null || _b === void 0 ? void 0 : _b[d]) === undefined) {
                const meta = (0, errorDefinitions_1.getOavErrorMeta)("OBJECT_MISSING_REQUIRED_PROPERTY_DEFINITION", {
                    property: d,
                });
                this.addErrorsFromErrorCode(errors, rootInfo.url, meta, sch);
            }
            else {
                const discriminatorProp = jsonLoader.resolveRefObj(sch.properties[d]);
                if (discriminatorProp.type !== "string") {
                    const meta = (0, errorDefinitions_1.getOavErrorMeta)("INVALID_DISCRIMINATOR_TYPE", {
                        property: d,
                    });
                    this.addErrorsFromErrorCode(errors, rootInfo.url, meta, sch);
                    continue;
                }
                if (discriminatorProp.enum !== undefined && sch.discriminatorMap !== undefined) {
                    for (const childSchRef of Object.values(sch.discriminatorMap)) {
                        if (childSchRef === null) {
                            continue;
                        }
                        const childSch = jsonLoader.resolveRefObj(childSchRef);
                        const discriminatorValue = childSch[constants_1.xmsDiscriminatorValue];
                        if (discriminatorValue !== undefined &&
                            !discriminatorProp.enum.includes(discriminatorValue)) {
                            const meta = (0, errorDefinitions_1.getOavErrorMeta)("INVALID_XMS_DISCRIMINATOR_VALUE", {
                                value: discriminatorValue,
                            });
                            const url = (0, openapi_tools_common_1.getRootObjectInfo)((0, openapi_tools_common_1.getInfo)(childSch)).url;
                            this.addErrorsFromErrorCode(errors, url, meta, childSch);
                        }
                    }
                }
            }
        }
        for (const sch of objSchemas) {
            if (!sch._missingDiscriminator) {
                continue;
            }
            const info = (0, openapi_tools_common_1.getInfo)(sch);
            const rootInfo = (0, openapi_tools_common_1.getRootObjectInfo)(info);
            const meta = (0, errorDefinitions_1.getOavErrorMeta)("DISCRIMINATOR_PROPERTY_NOT_FOUND", {
                value: sch[constants_1.xmsDiscriminatorValue],
            });
            this.addErrorsFromErrorCode(errors, rootInfo.url, meta, sch);
        }
    }
    async validateDefaultValue(transformCtx, url, errors) {
        for (const sch of transformCtx.objSchemas) {
            if (sch.default === undefined) {
                continue;
            }
            const validate = await this.schemaValidator.compileAsync({
                properties: {
                    default: sch,
                },
            });
            const result = validate({}, sch);
            this.addErrorsFromSchemaValidation(result, url, sch, errors);
        }
    }
    async validateSchemaRequiredProperties(transformCtx, url, errors) {
        var _a;
        for (const sch of transformCtx.objSchemas) {
            if (sch.required === undefined) {
                continue;
            }
            for (const name of sch.required) {
                if (((_a = sch.properties) === null || _a === void 0 ? void 0 : _a[name]) !== undefined) {
                    continue;
                }
                const meta = (0, errorDefinitions_1.getOavErrorMeta)("OBJECT_MISSING_REQUIRED_PROPERTY_DEFINITION", {
                    property: name,
                });
                this.addErrorsFromErrorCode(errors, url, meta, sch);
            }
        }
        for (const sch of transformCtx.arrSchemas) {
            if (sch.items !== undefined) {
                continue;
            }
            const meta = (0, errorDefinitions_1.getOavErrorMeta)("OBJECT_MISSING_REQUIRED_PROPERTY_SCHEMA", {
                property: "items",
            });
            this.addErrorsFromErrorCode(errors, url, meta, sch);
        }
    }
    async validateOperation(spec, errors) {
        const visitedOperationId = new Set();
        const visitedPathTemplate = new Set();
        const url = spec._filePath;
        const pathArgs = new Set();
        let pathParams;
        (0, traverseSwagger_1.traverseSwagger)(spec, {
            onPath: (path) => {
                pathArgs.clear();
                pathParams = path.parameters;
                const pathTemplate = path._pathTemplate;
                let normalizedPath = pathTemplate;
                const argMatches = normalizedPath.match(/\{.*?\}/g);
                let idx = 0;
                for (const arg of argMatches !== null && argMatches !== void 0 ? argMatches : []) {
                    if (arg === "{}") {
                        const meta = (0, errorDefinitions_1.getOavErrorMeta)("EMPTY_PATH_PARAMETER_DECLARATION", { pathTemplate });
                        this.addErrorsFromErrorCode(errors, url, meta, path);
                    }
                    else {
                        normalizedPath = normalizedPath.replace(arg, `arg${idx}`);
                        ++idx;
                        pathArgs.add(arg.substr(1, arg.length - 2));
                    }
                }
                if (visitedPathTemplate.has(normalizedPath)) {
                    const meta = (0, errorDefinitions_1.getOavErrorMeta)("EQUIVALENT_PATH", { pathTemplate });
                    this.addErrorsFromErrorCode(errors, url, meta, path);
                }
                visitedPathTemplate.add(normalizedPath);
            },
            onOperation: (operation) => {
                let bodyParam;
                const requiredPathArgs = new Set(pathArgs);
                const visitedParamName = new Set();
                const { operationId, parameters, consumes } = operation;
                const mergedParameters = [...(parameters !== null && parameters !== void 0 ? parameters : []), ...(pathParams !== null && pathParams !== void 0 ? pathParams : [])];
                if (operationId !== undefined) {
                    if (visitedOperationId.has(operationId)) {
                        const meta = (0, errorDefinitions_1.getOavErrorMeta)("DUPLICATE_OPERATIONID", { operationId });
                        this.addErrorsFromErrorCode(errors, url, meta, operation);
                    }
                    else {
                        visitedOperationId.add(operationId);
                    }
                }
                for (const p of mergedParameters) {
                    const param = this.jsonLoader.resolveRefObj(p);
                    const { name } = param;
                    if (visitedParamName.has(name)) {
                        const meta = (0, errorDefinitions_1.getOavErrorMeta)("DUPLICATE_PARAMETER", { name });
                        this.addErrorsFromErrorCode(errors, url, meta, operation);
                    }
                    visitedParamName.add(name);
                    if (param.in === "body" || param.in === "formData") {
                        if (bodyParam !== undefined) {
                            const meta = (0, errorDefinitions_1.getOavErrorMeta)(param.in === bodyParam.in
                                ? "MULTIPLE_BODY_PARAMETERS"
                                : "INVALID_PARAMETER_COMBINATION", {});
                            if (!(meta.code === "MULTIPLE_BODY_PARAMETERS" &&
                                param.in === "formData" &&
                                consumes !== undefined &&
                                // Must support both "multipart/form-data" and "application/x-www-form-urlencoded" per spec:
                                // https://swagger.io/docs/specification/v2_0/describing-parameters/#form-parameters
                                //
                                // Must support "multipart/mixed" for storage/data-plane/Microsoft.BlobStorage
                                (consumes.includes("multipart/form-data") ||
                                    consumes.includes("multipart/mixed") ||
                                    consumes.includes("application/x-www-form-urlencoded")))) {
                                this.addErrorsFromErrorCode(errors, url, meta, operation);
                            }
                        }
                        bodyParam = param;
                    }
                    if (param.in === "path") {
                        if (!requiredPathArgs.has(name)) {
                            const meta = (0, errorDefinitions_1.getOavErrorMeta)("MISSING_PATH_PARAMETER_DECLARATION", { name });
                            this.addErrorsFromErrorCode(errors, url, meta, operation);
                        }
                        requiredPathArgs.delete(name);
                    }
                }
                for (const name of requiredPathArgs) {
                    const meta = (0, errorDefinitions_1.getOavErrorMeta)("MISSING_PATH_PARAMETER_DEFINITION", { name });
                    this.addErrorsFromErrorCode(errors, url, meta, operation);
                }
            },
        });
    }
};
exports.SwaggerSemanticValidator = SwaggerSemanticValidator;
exports.SwaggerSemanticValidator = SwaggerSemanticValidator = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(inversifyUtils_1.TYPES.opts)),
    __param(5, (0, inversify_1.inject)(inversifyUtils_1.TYPES.schemaValidator)),
    __metadata("design:paramtypes", [Object, jsonLoader_1.JsonLoader,
        fileLoader_1.FileLoader,
        suppressionLoader_1.SuppressionLoader,
        liveValidatorLoader_1.LiveValidatorLoader, Object])
], SwaggerSemanticValidator);
// Compatible wrapper for old SemanticValidator
class SemanticValidator {
    constructor(specPath, specInJson) {
        this.specPath = specPath;
        this.specValidationResult = { validityStatus: true, operations: {} };
        const container = (0, inversifyUtils_1.inversifyGetContainer)();
        this.validator = (0, inversifyUtils_1.inversifyGetInstance)(SwaggerSemanticValidator, {
            ...defaultOpts,
            container,
        });
        if (specInJson) {
            const fileLoader = container.get(fileLoader_1.FileLoader);
            fileLoader.preloadExtraFile(specPath, JSON.stringify(specInJson));
        }
    }
    async initialize() {
        await this.validator.initialize();
        // API compatible
        return null;
    }
    async validateSpec() {
        const errors = await this.validator.validateSwaggerSpec(this.specPath);
        this.specValidationResult.validateSpec = {
            isValid: errors.length === 0,
            errors,
        };
        this.specValidationResult.validityStatus = errors.length === 0;
        return { errors, warnings: [] };
    }
}
exports.SemanticValidator = SemanticValidator;
