// oav/dist/lib/swaggerValidator/trafficValidator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrafficValidator = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const path_1 = require("path");
const glob = __importStar(require("glob"));
const liveValidator_1 = require("../liveValidation/liveValidator");
const constants_1 = require("../util/constants");
const errorDefinitions_1 = require("../util/errorDefinitions");
const inversifyUtils_1 = require("../inversifyUtils");
const utils_1 = require("../util/utils");
const swaggerLoader_1 = require("../swagger/swaggerLoader");
const jsonUtils_1 = require("../util/jsonUtils");
const liveValidatorLoader_1 = require("../liveValidation/liveValidatorLoader");
const traverseSwagger_1 = require("../transform/traverseSwagger");
class TrafficValidator {
    constructor(specPath, trafficPath) {
        this.trafficValidationResult = [];
        this.trafficFiles = [];
        this.trafficOperation = new Map();
        this.validationFailOperations = new Map();
        this.coverageData = new Map();
        this.operationSpecMapper = new Map();
        this.operationCoverageResult = [];
        this.operationUndefinedResult = 0;
        this.specPath = (0, path_1.resolve)(specPath);
        this.trafficPath = (0, path_1.resolve)(trafficPath);
    }
    async initialize() {
        const specPathStats = fs.statSync(this.specPath);
        const trafficPathStats = fs.statSync(this.trafficPath);
        let specFileDirectory = "";
        let swaggerPathsPattern = "**/*.json";
        if (specPathStats.isFile()) {
            specFileDirectory = path.dirname(this.specPath);
            swaggerPathsPattern = path.basename(this.specPath);
        }
        else if (specPathStats.isDirectory()) {
            specFileDirectory = this.specPath;
        }
        if (trafficPathStats.isFile()) {
            this.trafficFiles.push(this.trafficPath);
        }
        else if (trafficPathStats.isDirectory()) {
            const searchPattern = path.join(this.trafficPath, "**/*.json");
            const matchedPaths = glob
                .sync(searchPattern, {
                nodir: true,
            })
                // Sort alphabetically for compat with glob <= 8 (https://github.com/isaacs/node-glob/blob/v7.2.3/common.js#L20).
                // I'm not sure why oav requires these to be sorted, but without sorting, test "validate data-plane traffic"
                // fails with a small snapshot diff (jsonRef, column, and line are all undefined).
                .sort((a, b) => a.localeCompare(b, "en"));
            for (const filePath of matchedPaths) {
                this.trafficFiles.push(filePath);
            }
        }
        const liveValidationOptions = {
            checkUnderFileRoot: false,
            loadValidatorInBackground: false,
            directory: specFileDirectory,
            swaggerPathsPattern: [swaggerPathsPattern],
            excludedSwaggerPathsPattern: constants_1.DefaultConfig.ExcludedExamplesAndCommonFiles,
            git: {
                shouldClone: false,
            },
        };
        this.liveValidator = new liveValidator_1.LiveValidator(liveValidationOptions);
        await this.liveValidator.initialize();
        const container = (0, inversifyUtils_1.inversifyGetContainer)();
        this.loader = (0, inversifyUtils_1.inversifyGetInstance)(liveValidatorLoader_1.LiveValidatorLoader, {
            container,
            fileRoot: liveValidationOptions.directory,
            ...liveValidationOptions,
            loadSuppression: Object.keys(errorDefinitions_1.apiValidationErrors),
        });
        const swaggerPaths = this.liveValidator.swaggerList;
        while (swaggerPaths.length > 0) {
            const swaggerPath = swaggerPaths.shift();
            let spec;
            try {
                spec = await this.loader.load((0, path_1.resolve)(swaggerPath));
            }
            catch (e) {
                console.log(`Exception when loading spec, ErrorMessage: ${e === null || e === void 0 ? void 0 : e.message}; ErrorStack: ${e === null || e === void 0 ? void 0 : e.stack}.`);
            }
            if (spec !== undefined) {
                // Get Swagger - operation mapper.
                if (this.operationSpecMapper.get(swaggerPath) === undefined) {
                    this.operationSpecMapper.set(swaggerPath, []);
                }
                (0, traverseSwagger_1.traverseSwagger)(spec, {
                    onOperation: (operation) => {
                        var _a;
                        if (operation.operationId !== undefined &&
                            !((_a = this.operationSpecMapper.get(swaggerPath)) === null || _a === void 0 ? void 0 : _a.includes(operation.operationId))) {
                            this.operationSpecMapper.get(swaggerPath).push(operation.operationId);
                        }
                    },
                });
            }
        }
    }
    async validate() {
        var _a, _b, _c, _d, _e;
        let payloadFilePath;
        const swaggerOpts = {
            setFilePath: false,
        };
        this.swaggerLoader = (0, inversifyUtils_1.inversifyGetInstance)(swaggerLoader_1.SwaggerLoader, swaggerOpts);
        try {
            for (const trafficFile of this.trafficFiles) {
                payloadFilePath = trafficFile;
                const payload = require(trafficFile);
                const validationResult = await this.liveValidator.validateLiveRequestResponse(payload);
                let operationInfo = (_a = validationResult.requestValidationResult) === null || _a === void 0 ? void 0 : _a.operationInfo;
                const liveRequest = payload.liveRequest;
                const opInfo = await this.liveValidator.getOperationInfo(liveRequest);
                const errorResult = [];
                const runtimeExceptions = [];
                if (validationResult.requestValidationResult.isSuccessful === undefined) {
                    runtimeExceptions.push(validationResult.requestValidationResult.runtimeException);
                }
                else if (validationResult.requestValidationResult.isSuccessful === false) {
                    errorResult.push(...validationResult.requestValidationResult.errors.map((e) => {
                        e.issueSource = "request";
                        return e;
                    }));
                }
                if (validationResult.responseValidationResult.isSuccessful === undefined) {
                    runtimeExceptions.push(validationResult.responseValidationResult.runtimeException);
                }
                else if (validationResult.responseValidationResult.isSuccessful === false) {
                    errorResult.push(...validationResult.responseValidationResult.errors.map((e) => {
                        e.issueSource = "response";
                        return e;
                    }));
                }
                const trafficSpec = await this.swaggerLoader.load(payloadFilePath);
                let liveRequestResponseList;
                if (validationResult.requestValidationResult.isSuccessful) {
                    liveRequestResponseList = (0, utils_1.findPathsToKey)({ key: "liveResponse", obj: trafficSpec });
                }
                else {
                    liveRequestResponseList = (0, utils_1.findPathsToKey)({ key: "liveRequest", obj: trafficSpec });
                }
                const liveRequestResponsePosition = (0, jsonUtils_1.getFilePositionFromJsonPath)(trafficSpec, liveRequestResponseList[0]);
                let swaggerFiles = [];
                if (liveRequest.url.includes("provider")) {
                    // This is for validation of resource-manager
                    swaggerFiles = this.findSwaggerByOperationInfo(opInfo.info);
                }
                else {
                    // This is for validation of data-plane
                    swaggerFiles = this.findSwaggerByOperationId(opInfo.info);
                }
                if (swaggerFiles.length !== 0) {
                    for (const swaggerFile of swaggerFiles) {
                        if (this.trafficOperation.get(swaggerFile) === undefined) {
                            this.trafficOperation.set(swaggerFile, []);
                        }
                        if (!((_b = this.trafficOperation.get(swaggerFile)) === null || _b === void 0 ? void 0 : _b.includes(opInfo.info.operationId))) {
                            (_c = this.trafficOperation.get(swaggerFile)) === null || _c === void 0 ? void 0 : _c.push(opInfo.info.operationId);
                        }
                        if (validationResult.requestValidationResult.isSuccessful === false ||
                            validationResult.requestValidationResult.isSuccessful === undefined ||
                            validationResult.responseValidationResult.isSuccessful === false ||
                            validationResult.responseValidationResult.isSuccessful === undefined ||
                            validationResult.runtimeException !== undefined) {
                            if (this.validationFailOperations.get(swaggerFile) === undefined) {
                                this.validationFailOperations.set(swaggerFile, []);
                            }
                            if (!((_d = this.validationFailOperations.get(swaggerFile)) === null || _d === void 0 ? void 0 : _d.includes(opInfo.info.operationId))) {
                                (_e = this.validationFailOperations.get(swaggerFile)) === null || _e === void 0 ? void 0 : _e.push(opInfo.info.operationId);
                            }
                            const spec = swaggerFile && (await this.swaggerLoader.load(swaggerFile));
                            const operationIdList = (0, utils_1.findPathsToKey)({ key: "operationId", obj: spec });
                            const operationId = (0, utils_1.findPathToValue)(operationIdList, spec, operationInfo.operationId);
                            const operationIdPosition = (0, jsonUtils_1.getFilePositionFromJsonPath)(spec, operationId[0]);
                            operationInfo = Object.assign(operationInfo, { position: operationIdPosition });
                            this.trafficValidationResult.push({
                                specFilePath: swaggerFile,
                                payloadFilePath,
                                payloadFilePathPosition: liveRequestResponsePosition,
                                errors: errorResult,
                                runtimeExceptions,
                                operationInfo,
                            });
                        }
                    }
                }
                else {
                    console.log(`Error: Undefined operation ${JSON.stringify(opInfo.info)}`);
                    this.operationUndefinedResult = this.operationUndefinedResult + 1;
                }
            }
        }
        catch (err) {
            const msg = `Detail error message:${err === null || err === void 0 ? void 0 : err.message}. ErrorStack:${err === null || err === void 0 ? void 0 : err.Stack}`;
            this.trafficValidationResult.push({
                payloadFilePath,
                runtimeExceptions: [
                    {
                        code: errorDefinitions_1.ErrorCodeConstants.RUNTIME_ERROR,
                        message: msg,
                    },
                ],
            });
        }
        let coveredOperations;
        let coverageRate;
        let validationFailOperations;
        let coveredOperationsList;
        let unCoveredOperationsList;
        this.operationSpecMapper.forEach((value, key) => {
            // identify the spec has been match traffic file
            let isMatch = true;
            const unCoveredOperationsListFormat = [];
            coveredOperationsList = [];
            unCoveredOperationsList = [];
            if (this.trafficOperation.get(key) === undefined) {
                coveredOperations = 0;
                coverageRate = 0;
                this.coverageData.set(key, 0);
                isMatch = false;
            }
            else if (value !== undefined && value.length !== 0) {
                const validatedOperations = this.trafficOperation.get(key);
                coveredOperations = validatedOperations.length;
                coverageRate = coveredOperations / value.length;
                this.coverageData.set(key, coverageRate);
                const unValidatedOperations = [...value];
                validatedOperations.forEach((element) => {
                    coveredOperationsList.push({ operationId: element });
                    unValidatedOperations.splice(unValidatedOperations.indexOf(element), 1);
                });
                unValidatedOperations.forEach((element) => {
                    unCoveredOperationsList.push({
                        key: element.split("_")[0],
                        operationId: element,
                    });
                });
                const unCoveredOperationsInnerList = Object.values(unCoveredOperationsList.reduce((res, item) => {
                    /* eslint-disable no-unused-expressions */
                    res[item.key] ? res[item.key].push(item) : (res[item.key] = [item]);
                    /* eslint-enable no-unused-expressions */
                    return res;
                }, {}));
                unCoveredOperationsInnerList.forEach((element) => {
                    unCoveredOperationsListFormat.push({
                        operationIdList: element,
                    });
                });
            }
            else {
                isMatch = false;
                coveredOperations = 0;
                coverageRate = 0;
                this.coverageData.set(key, 0);
            }
            if (this.validationFailOperations.get(key) === undefined) {
                validationFailOperations = 0;
            }
            else {
                validationFailOperations = this.validationFailOperations.get(key).length;
            }
            const sortedUnCoveredOperationsList = unCoveredOperationsList.sort(function (op1, op2) {
                const opId1 = op1.operationId;
                const opId2 = op2.operationId;
                if (opId1 < opId2) {
                    return -1;
                }
                if (opId1 > opId2) {
                    return 1;
                }
                return 0;
            });
            const sortedCoveredOperationsList = coveredOperationsList.sort(function (op1, op2) {
                const opId1 = op1.operationId;
                const opId2 = op2.operationId;
                if (opId1 < opId2) {
                    return -1;
                }
                if (opId1 > opId2) {
                    return 1;
                }
                return 0;
            });
            /**
             * Sort untested operationId by bubble sort
             * Controlling the results of localeCompare can set the sorting method
             * X.localeCompare(Y) > 0 descending sort
             * X.localeCompare(Y) < 0 ascending sort
             */
            for (let i = 0; i < unCoveredOperationsListFormat.length - 1; i++) {
                for (let j = 0; j < unCoveredOperationsListFormat.length - 1 - i; j++) {
                    if (unCoveredOperationsListFormat[j].operationIdList[0].key.localeCompare(unCoveredOperationsListFormat[j + 1].operationIdList[0].key) > 0) {
                        var temp = unCoveredOperationsListFormat[j];
                        unCoveredOperationsListFormat[j] = unCoveredOperationsListFormat[j + 1];
                        unCoveredOperationsListFormat[j + 1] = temp;
                    }
                }
            }
            isMatch &&
                this.operationCoverageResult.push({
                    spec: key,
                    apiVersion: (0, utils_1.getApiVersionFromFilePath)(key),
                    coveredOperations,
                    coverageRate,
                    unCoveredOperations: value.length - coveredOperations,
                    totalOperations: value.length,
                    validationFailOperations: validationFailOperations,
                    coveredOperationsList: sortedCoveredOperationsList,
                    unCoveredOperationsList: sortedUnCoveredOperationsList,
                    unCoveredOperationsListGen: unCoveredOperationsListFormat,
                });
        });
        return this.trafficValidationResult;
    }
    findSwaggerByOperationInfo(operationInfo) {
        var _a;
        let result = [];
        if (operationInfo.validationRequest === undefined) {
            return result;
        }
        for (const key of this.operationSpecMapper.keys()) {
            const value = this.operationSpecMapper.get(key);
            if (key.toLowerCase().includes((_a = operationInfo.validationRequest) === null || _a === void 0 ? void 0 : _a.providerNamespace) &&
                (key.includes(operationInfo.apiVersion) ||
                    key.toLowerCase().includes(operationInfo.apiVersion))) {
                if (value.includes(operationInfo.operationId)) {
                    result.push(key);
                }
            }
        }
        return result;
    }
    findSwaggerByOperationId(operationInfo) {
        let result = [];
        for (const key of this.operationSpecMapper.keys()) {
            const value = this.operationSpecMapper.get(key);
            if (value.includes(operationInfo.operationId) &&
                (key.includes(operationInfo.apiVersion) ||
                    key.toLowerCase().includes(operationInfo.apiVersion))) {
                result.push(key);
            }
        }
        return result;
    }
}
exports.TrafficValidator = TrafficValidator;
