// oav/dist/lib/report/generateReport.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReportGenerator = exports.CoverageView = void 0;
exports.loadErrorDefinitions = loadErrorDefinitions;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const Mustache = __importStar(require("mustache"));
const fileLoader_1 = require("../swagger/fileLoader");
async function loadErrorDefinitions() {
    const errorDefinitionDoc = require("../../../documentation/error-definitions.json");
    const errorsMap = new Map();
    errorDefinitionDoc.ErrorDefinitions.forEach((def) => {
        errorsMap.set(def.code, def);
    });
    return errorsMap;
}
// used to pass data to the template rendering engine
class CoverageView {
    constructor(validationResults, coverageResults, undefinedOperationCount = 0, packageName = "", language = "", markdownPath = "", overrideLinkInReport = false, outputExceptionInReport = false, specLinkPrefix = "", payloadLinkPrefix = "") {
        this.apiVersion = "unknown";
        this.undefinedOperationCount = 0;
        this.operationValidated = 0;
        this.operationFailed = 0;
        this.operationUnValidated = 0;
        this.validationResultsForRendering = [];
        this.coverageResultsForRendering = [];
        this.resultsForRendering = [];
        this.package = packageName;
        this.markdownPath = markdownPath;
        this.validationResults = validationResults;
        this.coverageResults = coverageResults;
        this.undefinedOperationCount = undefinedOperationCount;
        this.generatedDate = new Date();
        this.generalErrorResults = new Map();
        this.language = language;
        this.overrideLinkInReport = overrideLinkInReport;
        this.outputExceptionInReport = outputExceptionInReport;
        this.specLinkPrefix = specLinkPrefix;
        this.payloadLinkPrefix = payloadLinkPrefix;
        if (this.overrideLinkInReport === true) {
            if (this.specLinkPrefix.endsWith("/")) {
                this.specLinkPrefix = this.specLinkPrefix.substring(0, this.specLinkPrefix.length - 1);
            }
            if (this.payloadLinkPrefix.endsWith("/")) {
                this.payloadLinkPrefix = this.payloadLinkPrefix.substring(0, this.payloadLinkPrefix.length - 1);
            }
        }
        this.setMetrics();
        this.sortOperationIds();
    }
    async prepareDataForRendering() {
        try {
            this.markdown = await this.readMarkdown();
            const errorDefinitions = await loadErrorDefinitions();
            let errorsForRendering;
            this.sortedValidationResults.forEach((element) => {
                var _a, _b, _c, _d, _e, _f;
                const payloadFile = (_a = element.payloadFilePath) === null || _a === void 0 ? void 0 : _a.substring(element.payloadFilePath.lastIndexOf("/") + 1);
                errorsForRendering = [];
                (_b = element.errors) === null || _b === void 0 ? void 0 : _b.forEach((error) => {
                    var _a, _b, _c, _d, _e;
                    const errorDef = errorDefinitions.get(error.code);
                    errorsForRendering.push({
                        friendlyName: (_a = errorDef === null || errorDef === void 0 ? void 0 : errorDef.friendlyName) !== null && _a !== void 0 ? _a : error.code,
                        link: errorDef === null || errorDef === void 0 ? void 0 : errorDef.link,
                        code: error.code,
                        message: error.message,
                        schemaPath: error.schemaPath,
                        schemaPathWithPosition: this.overrideLinkInReport
                            ? `${this.specLinkPrefix}/${(_b = element.specFilePath) === null || _b === void 0 ? void 0 : _b.substring((_c = element.specFilePath) === null || _c === void 0 ? void 0 : _c.indexOf("specification"))}#L${error.source.position.line}`
                            : `${element.specFilePath}#L${error.source.position.line}`,
                        pathsInPayload: error.pathsInPayload,
                        jsonPathsInPayload: error.jsonPathsInPayload,
                        severity: error.severity,
                        source: error.source,
                        params: error.params,
                        payloadFilePath: this.overrideLinkInReport
                            ? `${this.payloadLinkPrefix}/${payloadFile}`
                            : element.payloadFilePath,
                        payloadFilePathWithPosition: this.overrideLinkInReport
                            ? `${this.payloadLinkPrefix}/${payloadFile}#L${(_d = element.payloadFilePathPosition) === null || _d === void 0 ? void 0 : _d.line}`
                            : `${element.payloadFilePath}#L${(_e = element.payloadFilePathPosition) === null || _e === void 0 ? void 0 : _e.line}`,
                        payloadFileLinkLabel: payloadFile,
                    });
                });
                this.validationResultsForRendering.push({
                    payloadFilePath: this.overrideLinkInReport
                        ? `${this.payloadLinkPrefix}/${payloadFile}`
                        : element.payloadFilePath,
                    payloadFileLinkLabel: payloadFile,
                    payloadFilePathWithPosition: this.overrideLinkInReport
                        ? `${this.payloadLinkPrefix}/${payloadFile}#L${(_c = element.payloadFilePathPosition) === null || _c === void 0 ? void 0 : _c.line}`
                        : `${element.payloadFilePath}#L${(_d = element.payloadFilePathPosition) === null || _d === void 0 ? void 0 : _d.line}`,
                    errors: element.errors,
                    specFilePath: this.overrideLinkInReport
                        ? `${this.specLinkPrefix}/${(_e = element.specFilePath) === null || _e === void 0 ? void 0 : _e.substring((_f = element.specFilePath) === null || _f === void 0 ? void 0 : _f.indexOf("specification"))}`
                        : element.specFilePath,
                    errorsForRendering: errorsForRendering,
                    errorCodeLen: errorsForRendering.length,
                    operationInfo: element.operationInfo,
                    runtimeExceptions: element.runtimeExceptions,
                });
            });
            const generalErrorsInnerOrigin = this.validationResultsForRendering.filter((x) => {
                return x.errors && x.errors.length > 0;
            });
            this.coverageResults.forEach((element) => {
                var _a, _b, _c, _d;
                const specLink = this.overrideLinkInReport
                    ? `${this.specLinkPrefix}/${(_a = element.spec) === null || _a === void 0 ? void 0 : _a.substring((_b = element.spec) === null || _b === void 0 ? void 0 : _b.indexOf("specification"))}`
                    : `${element.spec}`;
                let errorOperationIds = generalErrorsInnerOrigin.map((item) => { var _a; return (_a = item.operationInfo) === null || _a === void 0 ? void 0 : _a.operationId; });
                let passOperations = element.coveredOperationsList
                    .filter((item) => errorOperationIds.indexOf(item.operationId) === -1)
                    .map((item) => {
                    return {
                        key: item.operationId.split("_")[0],
                        operationId: item.operationId,
                    };
                });
                const passOperationsInnerList = Object.values(passOperations.reduce((res, item) => {
                    /* eslint-disable no-unused-expressions */
                    res[item.key] ? res[item.key].push(item) : (res[item.key] = [item]);
                    /* eslint-enable no-unused-expressions */
                    return res;
                }, {}));
                const passOperationsListFormat = [];
                passOperationsInnerList.forEach((element) => {
                    passOperationsListFormat.push({
                        operationIdList: element,
                    });
                });
                /**
                 * Sort untested operationId by bubble sort
                 * Controlling the results of localeCompare can set the sorting method
                 * X.localeCompare(Y) > 0 descending sort
                 * X.localeCompare(Y) < 0 ascending sort
                 */
                for (let i = 0; i < passOperationsListFormat.length - 1; i++) {
                    for (let j = 0; j < passOperationsListFormat.length - 1 - i; j++) {
                        if (passOperationsListFormat[j].operationIdList[0].key.localeCompare(passOperationsListFormat[j + 1].operationIdList[0].key) > 0) {
                            var temp = passOperationsListFormat[j];
                            passOperationsListFormat[j] = passOperationsListFormat[j + 1];
                            passOperationsListFormat[j + 1] = temp;
                        }
                    }
                }
                this.coverageResultsForRendering.push({
                    spec: specLink,
                    specLinkLabel: (_c = element.spec) === null || _c === void 0 ? void 0 : _c.substring(((_d = element.spec) === null || _d === void 0 ? void 0 : _d.lastIndexOf("/")) + 1),
                    apiVersion: element.apiVersion,
                    coveredOperations: element.coveredOperations,
                    coveredOperationsList: element.coveredOperationsList,
                    validationPassOperations: element.coveredOperations - element.validationFailOperations,
                    validationPassOperationList: passOperationsListFormat,
                    validationFailOperations: element.validationFailOperations,
                    unCoveredOperations: element.unCoveredOperations,
                    unCoveredOperationsList: element.unCoveredOperationsList,
                    unCoveredOperationsListGen: element.unCoveredOperationsListGen,
                    totalOperations: element.totalOperations,
                    coverageRate: element.coverageRate,
                    generalErrorsInnerList: [],
                });
            });
            this.resultsForRendering = this.coverageResultsForRendering.map((item) => {
                const data = this.validationResultsForRendering.find((i) => i.specFilePath &&
                    item.spec.split(path.win32.sep).join(path.posix.sep).includes(i.specFilePath));
                return {
                    ...item,
                    ...data,
                };
            });
            const generalErrorsInnerFormat = Object.values(generalErrorsInnerOrigin.reduce((res, item) => {
                /* eslint-disable no-unused-expressions */
                res[item.operationInfo.operationId + item.specFilePath]
                    ? res[item.operationInfo.operationId + item.specFilePath].push(item)
                    : (res[item.operationInfo.operationId + item.specFilePath] = [item]);
                /* eslint-enable no-unused-expressions */
                return res;
            }, {}));
            const generalErrorsInnerList = [];
            generalErrorsInnerFormat.forEach((element) => {
                var _a, _b, _c, _d;
                let errorCodeLen = 0;
                element.forEach((item) => {
                    errorCodeLen = errorCodeLen + item.errorCodeLen;
                });
                let errorsForRendering = [];
                element.forEach((item) => {
                    errorsForRendering = errorsForRendering.concat(item.errorsForRendering);
                });
                generalErrorsInnerList.push({
                    generalErrorsInner: element,
                    errorCodeLen: errorCodeLen,
                    errorsForRendering: errorsForRendering,
                    operationInfo: element[0].operationInfo,
                    specFilePath: this.overrideLinkInReport
                        ? `${this.specLinkPrefix}/${(_a = element[0].specFilePath) === null || _a === void 0 ? void 0 : _a.substring((_b = element[0].specFilePath) === null || _b === void 0 ? void 0 : _b.indexOf("specification"))}`
                        : element[0].specFilePath,
                    specFilePathWithPosition: this.overrideLinkInReport
                        ? `${this.specLinkPrefix}/${(_c = element[0].specFilePath) === null || _c === void 0 ? void 0 : _c.substring((_d = element[0].specFilePath) === null || _d === void 0 ? void 0 : _d.indexOf("specification"))}#L${element[0].operationInfo.position.line}`
                        : `${element[0].specFilePath}#L${element[0].operationInfo.position.line}`,
                });
            });
            for (const [index, e] of this.resultsForRendering.entries()) {
                e.index = index;
                for (const i of generalErrorsInnerList) {
                    if (e.specFilePath === i.specFilePath && i) {
                        e.generalErrorsInnerList.push(i);
                    }
                }
            }
        }
        catch (e) {
            console.error(`Failed in prepareDataForRendering with err:${e === null || e === void 0 ? void 0 : e.stack};message:${e === null || e === void 0 ? void 0 : e.message}`);
        }
    }
    async readMarkdown() {
        try {
            const loader = new fileLoader_1.FileLoader({});
            const res = await loader.load(this.markdownPath);
            return res;
        }
        catch (e) {
            console.error(`Failed in read report.md file`);
            return "";
        }
    }
    sortOperationIds() {
        this.sortedValidationResults = this.validationResults.sort(function (op1, op2) {
            const opId1 = op1.operationInfo.operationId;
            const opId2 = op2.operationInfo.operationId;
            if (opId1 < opId2) {
                return -1;
            }
            if (opId1 > opId2) {
                return 1;
            }
            return 0;
        });
    }
    setMetrics() {
        var _a;
        if (((_a = this.coverageResults) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            this.apiVersion = this.coverageResults[0].apiVersion;
        }
    }
    formatGeneratedDate() {
        const day = this.generatedDate.getDate();
        const month = this.generatedDate.getMonth() + 1;
        const year = this.generatedDate.getFullYear();
        const hours = this.generatedDate.getHours();
        const minutes = this.generatedDate.getMinutes();
        return (year +
            "-" +
            (month < 10 ? "0" + month : month) +
            "-" +
            (day < 10 ? "0" + day : day) +
            " at " +
            hours +
            ":" +
            (minutes < 10 ? "0" + minutes : minutes) +
            (hours < 13 ? "AM" : "PM"));
    }
    getTotalErrors() {
        return this.validationResults.length;
    }
    getGeneralErrors() {
        return this.validationResultsForRendering.filter((x) => {
            return x.errors && x.errors.length > 0;
        });
    }
    getTotalGeneralErrors() {
        return this.getGeneralErrors().length;
    }
    getRunTimeErrors() {
        if (this.outputExceptionInReport) {
            const res = this.validationResults.filter((x) => {
                return x.runtimeExceptions && x.runtimeExceptions.length > 0;
            });
            const resFormat = [];
            res.forEach((element) => {
                element.runtimeExceptions &&
                    element.runtimeExceptions.forEach((i) => {
                        const specList = i.spec;
                        const specListFormat = [];
                        specList &&
                            specList.forEach((k) => {
                                const specLink = this.overrideLinkInReport
                                    ? `${this.specLinkPrefix}/${k === null || k === void 0 ? void 0 : k.substring(k === null || k === void 0 ? void 0 : k.indexOf("specification"))}`
                                    : k;
                                const specLabel = k === null || k === void 0 ? void 0 : k.substring((k === null || k === void 0 ? void 0 : k.lastIndexOf("/")) + 1);
                                specListFormat.push({ specLabel, specLink });
                            });
                        resFormat.push({
                            code: i.code,
                            message: i.message,
                            payloadFilePath: element.payloadFilePath,
                            specList: specListFormat,
                        });
                    });
            });
            return resFormat;
        }
        else {
            return [];
        }
    }
    getTotalRunTimeErrors() {
        return this.getRunTimeErrors().length;
    }
}
exports.CoverageView = CoverageView;
class ReportGenerator {
    constructor(validationResults, coverageResults, undefinedOperationResults, options) {
        this.validationResults = validationResults;
        this.coverageResults = coverageResults;
        this.undefinedOperationsCount = undefinedOperationResults;
        this.reportPath = path.resolve(process.cwd(), options.reportPath);
        this.sdkLanguage = options.sdkLanguage;
        this.sdkPackage = options.sdkPackage;
        this.markdownPath = options.markdownPath;
        this.overrideLinkInReport = options.overrideLinkInReport;
        this.outputExceptionInReport = options.outputExceptionInReport;
        this.specLinkPrefix = options.specLinkPrefix;
        this.payloadLinkPrefix = options.payloadLinkPrefix;
    }
    async generateHtmlReport() {
        const templatePath = path.join(__dirname, "../templates/baseLayout.mustache");
        const template = fs.readFileSync(templatePath, "utf-8");
        const view = new CoverageView(this.validationResults, this.coverageResults, this.undefinedOperationsCount, this.sdkPackage, this.sdkLanguage, this.markdownPath, this.overrideLinkInReport, this.outputExceptionInReport, this.specLinkPrefix, this.payloadLinkPrefix);
        await view.prepareDataForRendering();
        const general_errors = view.getGeneralErrors();
        const runtime_errors = view.getRunTimeErrors();
        console.log(JSON.stringify(general_errors, null, 2));
        console.log(JSON.stringify(runtime_errors, null, 2));
        const text = Mustache.render(template, view);
        fs.writeFileSync(this.reportPath, text, "utf-8");
    }
}
exports.ReportGenerator = ReportGenerator;
