// @azure-tools/typespec-azure-core/dist/src/experimental/decorators.js
import { reportDiagnostic } from "../lib.js";
export function $changePropertyType(ctx, target, newType) {
    reportDiagnostic(ctx.program, {
        code: "experimental-feature",
        messageId: "dangerous",
        format: {
            feature: "@changePropertyType",
        },
        target: ctx.decoratorTarget,
    });
    target.type = newType;
}
export function $copyProperties(ctx, target, baseModel) {
    reportDiagnostic(ctx.program, {
        code: "experimental-feature",
        messageId: "dangerous",
        format: {
            feature: "@copyProperties",
        },
        target: ctx.decoratorTarget,
    });
    for (const [propName, prop] of baseModel.properties) {
        if (target.properties.has(propName)) {
            continue;
        }
        target.properties.set(propName, prop);
    }
}
export function $copyVariants(ctx, target, sourceUnion) {
    reportDiagnostic(ctx.program, {
        code: "experimental-feature",
        messageId: "dangerous",
        format: {
            feature: "@copyVariants",
        },
        target: ctx.decoratorTarget,
    });
    for (const [variantName, variantType] of sourceUnion.variants) {
        if (target.variants.has(variantName)) {
            continue;
        }
        target.variants.set(variantName, variantType);
    }
}
