// oav/dist/lib/generator/exampleCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reBuildExample = exports.createTrunkItem = exports.createLeafItem = exports.buildItemOption = exports.PayloadCache = exports.MockerCache = void 0;
const isBaseResource = (cacheKey) => {
    const pieces = cacheKey.split("/");
    if (pieces.length < 2) {
        return false;
    }
    return ["resource", "proxyresource", "trackedresource", "azureentityresource"].some((r) => r === pieces[pieces.length - 1].toLowerCase());
};
class MockerCache {
    constructor() {
        this.caches = new Map();
    }
    get(modelName) {
        if (this.has(modelName)) {
            return this.caches.get(modelName);
        }
        return undefined;
    }
    set(modelName, example) {
        if (!this.has(modelName)) {
            this.caches.set(modelName, example);
        }
    }
    has(modelName) {
        return this.caches.has(modelName);
    }
    checkAndCache(schema, example) {
        if (!schema || !example) {
            return;
        }
        const cacheKey = schema.$ref && schema.$ref.includes("#") ? schema.$ref.split("#")[1] : undefined;
        if (cacheKey && !isBaseResource(cacheKey) && !this.has(cacheKey)) {
            this.set(cacheKey, example);
        }
    }
}
exports.MockerCache = MockerCache;
class PayloadCache {
    constructor() {
        this.requestCaches = new Map();
        this.responseCaches = new Map();
        this.mergedCaches = new Map();
    }
    hasByDirection(modelName, isRequest) {
        const cache = isRequest ? this.requestCaches : this.responseCaches;
        return cache.has(modelName);
    }
    setByDirection(modelName, example, isRequest) {
        const cache = isRequest ? this.requestCaches : this.responseCaches;
        if (!cache.has(modelName)) {
            cache.set(modelName, example);
        }
    }
    getByDirection(modelName, isRequest) {
        const cache = isRequest ? this.requestCaches : this.responseCaches;
        if (cache.has(modelName)) {
            return cache.get(modelName);
        }
        return undefined;
    }
    get(modelName) {
        if (this.mergedCaches.has(modelName)) {
            return this.mergedCaches.get(modelName);
        }
        return undefined;
    }
    set(key, value) {
        if (!this.mergedCaches.has(key)) {
            this.mergedCaches.set(key, value);
        }
    }
    has(modelName) {
        return this.mergedCaches.has(modelName);
    }
    checkAndCache(schema, example, isRequest) {
        if (!schema || !example) {
            return;
        }
        const cacheKey = schema.$ref && schema.$ref.includes("#") ? schema.$ref.split("#")[1] : undefined;
        if (cacheKey && !isBaseResource(cacheKey) && !this.hasByDirection(cacheKey, isRequest)) {
            this.setByDirection(cacheKey, example, isRequest);
        }
    }
    /**
     *  picking value priority : non-mocked value > mocked value , target value > source value
     * @param target The target item that to be merged into
     * @param source The source item that needs to merge
     */
    mergeItem(target, source) {
        const result = target;
        if (!source || !target) {
            return target ? target : source;
        }
        if (Array.isArray(result.child) && Array.isArray(source.child)) {
            const resultArr = result.child;
            const sourceArr = source.child;
            if (resultArr.length === 0 || sourceArr.length === 0) {
                return resultArr.length === 0 ? source : result;
            }
            // only when source is not mocked and target is mocked , choose source cache.
            if (resultArr[0].isMocked && !sourceArr[0].isMocked) {
                return source;
            }
            for (let i = 0; i < resultArr.length; i++) {
                if (i < sourceArr.length) {
                    resultArr[i] = this.mergeItem(resultArr[i], sourceArr[i]);
                }
            }
        }
        else if (source.child && result.child) {
            const resultObj = result.child;
            const sourceObj = source.child;
            for (const key of Object.keys(sourceObj)) {
                if (!resultObj[key]) {
                    resultObj[key] = sourceObj[key];
                }
                else {
                    resultObj[key] = this.mergeItem(resultObj[key], sourceObj[key]);
                }
            }
        }
        else {
            return result.isMocked && !source.isMocked ? source : result;
        }
        return result;
    }
    /**
     * 1 for each request cache , if exists in response cache, merge it with response cache and put into merged cache .
     * 2 for each response cache, if not exists in merged cache, then put into merged cache.
     */
    mergeCache() {
        for (const [key, requestCache] of this.requestCaches.entries()) {
            if (this.hasByDirection(key, false) && !requestCache.isLeaf) {
                const responseCache = this.getByDirection(key, false);
                if (responseCache) {
                    if (responseCache.isLeaf) {
                        console.error(`the response cache and request cache is inconsistent! key:${key}`);
                    }
                    else {
                        const mergedCache = this.mergeItem(requestCache, responseCache);
                        this.set(key, mergedCache);
                        continue;
                    }
                }
            }
            this.set(key, requestCache);
        }
        for (const [key, responseCache] of this.responseCaches.entries()) {
            if (!this.hasByDirection(key, true)) {
                this.set(key, responseCache);
            }
        }
        this.requestCaches.clear();
        this.responseCaches.clear();
    }
}
exports.PayloadCache = PayloadCache;
const buildItemOption = (schema) => {
    if (schema) {
        const isReadonly = !!schema.readOnly;
        const isXmsSecret = !!schema["x-ms-secret"];
        const isRequired = !!schema.required;
        const isWriteOnly = schema["x-ms-mutability"]
            ? schema["x-ms-mutability"].indexOf("read") === -1
            : false;
        if (!isReadonly && !isXmsSecret && !isRequired && !isWriteOnly) {
            return undefined;
        }
        let option = {};
        if (isReadonly) {
            option = { isReadonly: true };
        }
        if (isXmsSecret) {
            option = { ...option, isXmsSecret: true };
        }
        if (isWriteOnly) {
            option = { ...option, isWriteOnly: true };
        }
        if (schema.required === true) {
            option = { ...option, isRequired: true };
        }
        return option;
    }
    return undefined;
};
exports.buildItemOption = buildItemOption;
const createLeafItem = (itemValue, option = undefined) => {
    const item = {
        isLeaf: true,
        value: itemValue,
    };
    if (option) {
        item.options = option;
    }
    return item;
};
exports.createLeafItem = createLeafItem;
const createTrunkItem = (itemValue, option) => {
    const item = {
        isLeaf: false,
        child: itemValue,
    };
    if (option) {
        item.options = option;
    }
    return item;
};
exports.createTrunkItem = createTrunkItem;
const reBuildExample = (cache, isRequest, schema, validator) => {
    if (!cache) {
        return undefined;
    }
    if (validator && !validator({ schemaCache: cache, isRequest })) {
        return undefined;
    }
    if (cache.isLeaf) {
        return cache.value;
    }
    if (Array.isArray(cache.child)) {
        const result = [];
        for (const item of cache.child) {
            if (validator && !validator({ schemaCache: item, isRequest, schema })) {
                continue;
            }
            result.push((0, exports.reBuildExample)(item, isRequest, schema, validator));
        }
        return result;
    }
    else if (cache.child) {
        const result = {};
        for (const key of Object.keys(cache.child)) {
            if (!validator || validator({ schemaCache: cache, propertyName: key, isRequest, schema })) {
                const value = (0, exports.reBuildExample)(cache.child[key], isRequest, schema, validator);
                if (value !== undefined) {
                    result[key] = value;
                }
            }
        }
        return result;
    }
    return undefined;
};
exports.reBuildExample = reBuildExample;
