// oav/dist/lib/generator/exampleGenerator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = __importStar(require("path"));
const deepdash_1 = __importDefault(require("deepdash"));
const lodash_1 = __importDefault(require("lodash"));
const jsonLoader_1 = require("../swagger/jsonLoader");
const traverseSwagger_1 = require("../transform/traverseSwagger");
const validate = __importStar(require("../validate"));
const ajvSchemaValidator_1 = require("../swaggerValidator/ajvSchemaValidator");
const context_1 = require("../transform/context");
const xmsPathsTransformer_1 = require("../transform/xmsPathsTransformer");
const resolveNestedDefinitionTransformer_1 = require("../transform/resolveNestedDefinitionTransformer");
const referenceFieldsTransformer_1 = require("../transform/referenceFieldsTransformer");
const discriminatorTransformer_1 = require("../transform/discriminatorTransformer");
const allOfTransformer_1 = require("../transform/allOfTransformer");
const noAdditionalPropertiesTransformer_1 = require("../transform/noAdditionalPropertiesTransformer");
const transformer_1 = require("../transform/transformer");
const logging_1 = require("../util/logging");
const inversifyUtils_1 = require("../inversifyUtils");
const util = __importStar(require("./util"));
const translator_1 = __importDefault(require("./translator"));
const swaggerMocker_1 = __importDefault(require("./swaggerMocker"));
const exampleCache_1 = require("./exampleCache");
const _ = (0, deepdash_1.default)(lodash_1.default);
class Generator {
    constructor(specFilePath, payloadDir, generationRule) {
        this.generationRule = generationRule;
        this.shouldMock = payloadDir ? false : true;
        this.specFilePath = specFilePath;
        this.payloadDir = payloadDir;
        this.jsonLoader = (0, inversifyUtils_1.inversifyGetInstance)(jsonLoader_1.JsonLoader, {
            useJsonParser: false,
            eraseXmsExamples: false,
        });
        this.mockerCache = new exampleCache_1.MockerCache();
        this.payloadCache = new exampleCache_1.PayloadCache();
        this.swaggerMocker = new swaggerMocker_1.default(this.jsonLoader, this.mockerCache, this.payloadCache);
        this.translator = new translator_1.default(this.jsonLoader, this.payloadCache, this.shouldMock ? this.swaggerMocker : undefined);
        const schemaValidator = new ajvSchemaValidator_1.AjvSchemaValidator(this.jsonLoader);
        this.transformContext = (0, context_1.getTransformContext)(this.jsonLoader, schemaValidator, [
            xmsPathsTransformer_1.xmsPathsTransformer,
            resolveNestedDefinitionTransformer_1.resolveNestedDefinitionTransformer,
            referenceFieldsTransformer_1.referenceFieldsTransformer,
            discriminatorTransformer_1.discriminatorTransformer,
            allOfTransformer_1.allOfTransformer,
            noAdditionalPropertiesTransformer_1.noAdditionalPropertiesTransformer,
        ]);
    }
    getSpecItem(spec, operationId) {
        const paths = spec.paths;
        for (const pathName of Object.keys(paths)) {
            for (const methodName of Object.keys(paths[pathName])) {
                if (paths[pathName][methodName].operationId === operationId) {
                    return {
                        path: pathName,
                        methodName,
                        content: paths[pathName][methodName],
                    };
                }
            }
        }
        return null;
    }
    async load() {
        this.spec = (await this.jsonLoader.load(this.specFilePath));
        (0, transformer_1.applySpecTransformers)(this.spec, this.transformContext);
        (0, transformer_1.applyGlobalTransformers)(this.transformContext);
        await this.cacheExistingExamples();
    }
    async generateAll() {
        if (!this.spec) {
            await this.load();
        }
        const errs = [];
        await (0, traverseSwagger_1.traverseSwaggerAsync)(this.spec, {
            onPath: async (apiPath, pathTemplate) => {
                apiPath._pathTemplate = pathTemplate;
            },
            onOperation: async (operation, pathObject, methodName) => {
                const pathName = pathObject._pathTemplate;
                const specItem = {
                    path: pathName,
                    methodName,
                    content: operation,
                };
                const operationId = operation.operationId || "";
                const errors = await this.generate(operationId, specItem);
                if (errors.length > 0) {
                    errs.push(...errors);
                    return false;
                }
                return true;
            },
        });
        return errs;
    }
    async cacheExistingExamples() {
        if (!this.shouldMock) {
            return;
        }
        await (0, traverseSwagger_1.traverseSwaggerAsync)(this.spec, {
            onOperation: async (operation, pathObject, methodName) => {
                const pathName = pathObject._pathTemplate;
                const specItem = {
                    path: pathName,
                    methodName,
                    content: operation,
                };
                const examples = operation["x-ms-examples"] || undefined;
                if (!examples) {
                    return;
                }
                const operationId = operation.operationId;
                /*
                const validateErrors = await validate.validateExamples(this.specFilePath, operationId, {
                });
                if(validateErrors.length > 0) {
                  console.warn(`invalid examples for operation:${operationId}.`);
                  console.warn(validateErrors);
                  return
                } */
                for (const key of Object.keys(examples)) {
                    if (key.match(new RegExp(`^${operationId}_.*_Gen$`))) {
                        continue;
                    }
                    const example = this.jsonLoader.resolveRefObj(examples[key]);
                    if (!example) {
                        continue;
                    }
                    this.translator.extractParameters(specItem, example.parameters);
                    for (const code of Object.keys(operation.responses)) {
                        if (example.responses && example.responses[code]) {
                            this.translator.extractResponse(specItem, example.responses[code], code);
                        }
                    }
                }
                return true;
            },
        });
        // reuse the payloadCache as exampleCache.
        this.payloadCache.mergeCache();
    }
    async generateExample(operationId, specItem, rule) {
        var _a;
        this.translator.setRule(rule);
        this.swaggerMocker.setRule(rule);
        let example;
        console.log(`start generated example for ${operationId}, rule:${rule.ruleName}`);
        if (!this.shouldMock) {
            example = this.getExampleFromPayload(operationId, specItem, rule);
            if (!example) {
                return [];
            }
        }
        else {
            const xMsExamples = ((_a = specItem === null || specItem === void 0 ? void 0 : specItem.content) === null || _a === void 0 ? void 0 : _a["x-ms-examples"]) || {};
            const xMsExampleKeys = Object.getOwnPropertyNames(xMsExamples);
            const title = xMsExampleKeys.length > 0 ? xMsExampleKeys[0] : "";
            example = {
                title: title.length > 0
                    ? title.concat(" - generated by [", rule.ruleName, "] rule")
                    : specItem.content.summary
                        ? specItem.content.summary
                        : specItem.content.description
                            ? specItem.content.description
                            : `${operationId}_${rule.exampleNamePostfix}`,
                operationId: operationId,
                parameters: {},
                responses: this.extractResponse(specItem, {}),
            };
            this.swaggerMocker.mockForExample(example, specItem, this.spec, util.getBaseName(this.specFilePath).split(".")[0]);
        }
        logging_1.log.info(example);
        const unifiedExample = this.unifyCommonProperty(example);
        const newSpec = util.referenceExmInSpec(this.specFilePath, specItem.path, specItem.methodName, `${operationId}_${rule.exampleNamePostfix}_Gen`);
        util.updateExmAndSpecFile(unifiedExample, newSpec, this.specFilePath, `${operationId}_${rule.exampleNamePostfix}_Gen.json`);
        logging_1.log.info(`start validating generated example for ${operationId}`);
        const validateErrors = await validate.validateExamples(this.specFilePath, operationId, {
        //   consoleLogLevel: "error"
        });
        if (validateErrors.length > 0) {
            logging_1.log.error(`the validation raised below error:`);
            logging_1.log.error(validateErrors);
            return validateErrors;
        }
        console.log(`generated example for ${operationId}, rule:${rule.ruleName} successfully!`);
        return [];
    }
    async generate(operationId, specItem) {
        if (!this.spec) {
            await this.load();
        }
        if (!specItem) {
            specItem = this.getSpecItem(this.spec, operationId);
            if (!specItem) {
                console.error(`no specItem for the operation id ${operationId}`);
                return [];
            }
        }
        const ruleSet = [];
        if (this.generationRule) {
            ruleSet.push({
                exampleNamePostfix: `${this.generationRule}imumSet`,
                ruleName: `${this.generationRule}imumSet`,
            });
        }
        else {
            ruleSet.push({
                exampleNamePostfix: "MaximumSet",
                ruleName: "MaximumSet",
            }, {
                exampleNamePostfix: "MinimumSet",
                ruleName: "MinimumSet",
            });
        }
        for (const rule of ruleSet) {
            const error = await this.generateExample(operationId, specItem, rule);
            if (error.length) {
                return error;
            }
        }
        return [];
    }
    unifyCommonProperty(example) {
        if (!example || !example.parameters || !example.responses) {
            return;
        }
        const requestPaths = _.paths(example.parameters, { pathFormat: "array" }).map((v) => v.reverse());
        /**
         * construct a inverted index , the key is leaf property key, value is reverse of the path from the root to the leaf property.
         */
        const invertedIndex = new Map();
        requestPaths.forEach((v) => {
            if (v.length && typeof v[0] === "string") {
                const parentPaths = invertedIndex.get(v[0]);
                if (!parentPaths) {
                    invertedIndex.set(v[0], [v.slice(1)]);
                }
                else {
                    parentPaths.push(v.slice(1));
                }
            }
        });
        /**
         * get two paths' common properties' count
         */
        const getMatchedNodeCnt = (baseNode, destNode) => {
            let count = 0;
            baseNode.some((v, k) => {
                if (k < destNode.length && destNode[k] === v) {
                    count++;
                    return false;
                }
                else {
                    return true;
                }
            });
            return count;
        };
        /**
         * update the property value of response using the same value which is found in the request
         */
        const res = _.mapValuesDeep(example.responses, (value, key, parentValue, context) => {
            if (!parentValue) {
                logging_1.log.warn(`parent is null`);
            }
            if (["integer", "number", "string"].some((type) => typeof value === type) &&
                typeof key === "string") {
                const possiblePaths = invertedIndex.get(key);
                if (context.path && possiblePaths) {
                    const basePath = context.path.slice().reverse().slice(1);
                    /**
                     * to find out the most matchable path in the parameters
                     */
                    const candidates = possiblePaths.filter((apiPath) => getMatchedNodeCnt(basePath, apiPath) > 1);
                    if (candidates.length === 0) {
                        return value;
                    }
                    /**
                     * if only one matched one path, just use it.
                     */
                    if (candidates.length === 1) {
                        const pathOfParameter = _.pathToString([key, ...candidates[0]].reverse());
                        const parameterValue = _.get(example.parameters, pathOfParameter);
                        // console.debug(`use path ${pathOfParameter} ,value :${parameterValue}
                        // -- original path:${_.pathToString(context.path as pathNodes)},value:${value}`);
                        return parameterValue;
                    }
                    const mostMatched = candidates.reduce((previous, current) => {
                        const countPrevious = getMatchedNodeCnt(basePath, previous);
                        const countCurrent = getMatchedNodeCnt(basePath, current);
                        return countPrevious < countCurrent ? current : previous;
                    });
                    return _.get(example.parameters, _.pathToString([key, ...mostMatched].reverse()));
                }
            }
            return value;
        }, {
            leavesOnly: true,
            pathFormat: "array",
        });
        // console.debug(`unify common properties end!`);
        example.responses = res;
        return example;
    }
    getExampleFromPayload(operationId, specItem, rule) {
        if (this.payloadDir) {
            const subPaths = path.dirname(this.specFilePath).split(/\\|\//).slice(-3).join("/");
            const payloadDir = path.join(this.payloadDir, subPaths);
            const payload = util.readPayloadFile(payloadDir, operationId);
            if (!payload) {
                logging_1.log.info(`no payload file for operationId ${operationId} under directory ${path.resolve(payloadDir, operationId)} named with <statusCode>.json`);
                return;
            }
            this.validatePayload(specItem, payload, operationId);
            this.cachePayload(specItem, payload);
            const example = {
                title: specItem.content.summary
                    ? specItem.content.summary
                    : specItem.content.description
                        ? specItem.content.description
                        : `${operationId}_${rule.exampleNamePostfix}`,
                operationId: operationId,
                parameters: this.extractRequest(specItem, payload),
                responses: this.extractResponse(specItem, payload),
            };
            return example;
        }
        return undefined;
    }
    cachePayload(specItem, payload) {
        /**
         *  1 cache parameter model
         *
         *  2 cache response model
         *
         *  3 merged cache
         */
        this.extractRequest(specItem, payload);
        this.extractResponse(specItem, payload);
        this.payloadCache.mergeCache();
    }
    validatePayload(specItem, payload, operationId) {
        const specApiVersion = this.spec.info.version;
        for (const statusCode of Object.keys(payload)) {
            // remove payload with undefined status code
            if (!(statusCode in specItem.content.responses)) {
                delete payload[statusCode];
                continue;
            }
            // remove payload with inconsistent api-version
            if (!("query" in payload[statusCode].liveRequest)) {
                continue;
            }
            const realApiVersion = payload[statusCode].liveRequest.query["api-version"];
            if (realApiVersion && realApiVersion !== specApiVersion) {
                delete payload[statusCode];
                logging_1.log.error(`${operationId} payload ${statusCode}.json's api-version is ${realApiVersion}, inconsistent with swagger spec's api-version ${specApiVersion}`);
            }
        }
    }
    extractRequest(specItem, payload) {
        logging_1.log.info("extractRequest");
        const liveRequest = this.getRequestPayload(specItem, payload);
        if (!liveRequest) {
            logging_1.log.warn(`no live request in payload`);
            return {};
        }
        const request = this.translator.extractRequest(specItem, liveRequest) || {};
        return request;
    }
    getRequestPayload(specItem, payload) {
        const longRunning = util.isLongRunning(specItem);
        for (const statusCode in payload) {
            if (longRunning && statusCode === "200") {
                continue;
            }
            if ("liveRequest" in payload[statusCode]) {
                return payload[statusCode].liveRequest;
            }
        }
    }
    extractResponse(specItem, payload) {
        logging_1.log.info("extractResponse");
        const specResp = specItem.content.responses;
        const longRunning = specItem.content["x-ms-long-running-operation"];
        // below handled status code also should add in swaggerMocker.ts mockForExample() preHandledStatusCode array
        if (longRunning && !("202" in specResp) && !("201" in specResp)) {
            // console.warn('x-ms-long-running-operation is true, but no 202 or 201 response');
            return {};
        }
        if (longRunning && !("200" in specResp || "204" in specResp)) {
            // console.warn('x-ms-long-running-operation is true, but no 200 or 204 response');
        }
        if (!longRunning && ("202" in specResp || "201" in specResp)) {
            // console.warn('x-ms-long-running-operation is not set true, but 202 or 201 response is provided');
            return {};
        }
        const resp = {};
        if (!longRunning && "200" in specResp) {
            this.getResponseExample(specItem, payload, resp, "200", false);
        }
        if ("201" in specResp) {
            this.getResponseExample(specItem, payload, resp, "201", "200" in specResp);
        }
        if ("202" in specResp) {
            this.getResponseExample(specItem, payload, resp, "202", "200" in specResp);
        }
        if ("204" in specResp) {
            resp["204"] = {};
        }
        return resp;
    }
    getLongrunResp(specItem, payload) {
        const payload200 = payload[200];
        if (!payload200 || !("liveResponse" in payload200)) {
            // console.warn(`Payload doesn't have the response result for long running case`);
            return {};
        }
        return {
            body: "schema" in specItem.content.responses["200"]
                ? this.translator.filterBodyContent(payload200.liveResponse.body, specItem.content.responses["200"].schema, false)
                : undefined,
        };
    }
    getResponseExample(specItem, payloadGeneral, resp, statusCode, getAsyncResp) {
        const payload = payloadGeneral[statusCode];
        if (!payload || !("liveResponse" in payload)) {
            // console.warn(`no payload recording for status code = ${statusCode}`);
            resp[statusCode] = {};
        }
        else {
            resp[statusCode] = this.translator.extractResponse(specItem, payload.liveResponse, statusCode);
        }
        if (getAsyncResp) {
            resp["200"] = this.getLongrunResp(specItem, payloadGeneral);
        }
    }
}
exports.default = Generator;
