// oav/dist/lib/generator/swaggerMocker.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const logging_1 = require("../util/logging");
const exampleCache_1 = require("./exampleCache");
const mocker_1 = __importDefault(require("./mocker"));
const util = __importStar(require("./util"));
const exampleRule_1 = require("./exampleRule");
class SwaggerMocker {
    constructor(jsonLoader, mockerCache, payloadCache) {
        this.jsonLoader = jsonLoader;
        this.mocker = new mocker_1.default();
        this.mockCache = mockerCache;
        this.exampleCache = payloadCache;
    }
    setRule(exampleRule) {
        this.exampleRule = exampleRule;
    }
    mockForExample(example, specItem, spec, rp) {
        const preHandledStatusCode = ["200", "201", "202", "204"]; // above status code prehandle in exampleGenerator.ts extractResponse()
        this.spec = spec;
        if (Object.keys(example.responses).length === 0) {
            for (const statusCode of Object.keys(specItem.content.responses)) {
                if (statusCode !== "default") {
                    example.responses[`${statusCode}`] = {};
                }
            }
        }
        else {
            for (const statusCode of Object.keys(specItem.content.responses)) {
                if (statusCode !== "default" && !preHandledStatusCode.includes(statusCode)) {
                    example.responses[`${statusCode}`] = {};
                }
            }
        }
        example.parameters = this.mockRequest(example.parameters, specItem.content.parameters, rp);
        example.responses = this.mockResponse(example.responses, specItem);
    }
    getMockCachedObj(objName, schema, isRequest) {
        return this.mockCachedObj(objName, schema, undefined, new Set(), isRequest);
    }
    mockResponse(responseExample, specItem) {
        for (const statusCode of Object.keys(responseExample)) {
            const mockedResp = this.mockEachResponse(statusCode, responseExample[statusCode], specItem);
            responseExample[statusCode] = mockedResp;
        }
        return responseExample;
    }
    mockEachResponse(statusCode, responseExample, specItem) {
        const visited = new Set();
        const validator = (0, exampleRule_1.getRuleValidator)(this.exampleRule).onResponseBody;
        const responseSpec = specItem.content.responses[statusCode];
        if (validator && !validator({ schema: responseSpec })) {
            return undefined;
        }
        return {
            headers: responseExample.hearders || this.mockHeaders(statusCode, specItem),
            body: "schema" in responseSpec
                ? this.mockObj("response body", responseSpec.schema, responseExample.body || {}, visited, false) || {}
                : undefined,
        };
    }
    mockHeaders(statusCode, specItem) {
        if (statusCode !== "201" && statusCode !== "202") {
            return undefined;
        }
        const validator = (0, exampleRule_1.getRuleValidator)(this.exampleRule).onResponseHeader;
        if (validator && !validator({ schema: specItem })) {
            return undefined;
        }
        const headerAttr = util.getPollingAttr(specItem);
        if (!headerAttr) {
            return;
        }
        return {
            [headerAttr]: "https://contoso.com/operationstatus",
        };
    }
    mockRequest(paramExample, paramSpec, rp) {
        const validator = (0, exampleRule_1.getRuleValidator)(this.exampleRule).onParameter;
        for (const pName of Object.keys(paramSpec)) {
            const element = paramSpec[pName];
            const visited = new Set();
            const paramEle = this.getDefSpec(element, visited);
            if (paramEle.name === "resourceGroupName") {
                paramExample.resourceGroupName = `rg${rp}`;
            }
            else if (paramEle.name === "api-version") {
                paramExample["api-version"] = this.spec.info.version;
            }
            else if ("schema" in paramEle) {
                // {
                //     "name": "parameters",
                //     "in": "body",
                //     "required": false,
                //     "schema": {
                //       "$ref": "#/definitions/SignalRResource"
                //     }
                // }
                if (!validator || validator({ schema: paramEle })) {
                    paramExample[paramEle.name] = this.mockObj(paramEle.name, paramEle.schema, paramExample[paramEle.name] || {}, visited, true);
                }
            }
            else {
                if (paramEle.name in paramExample) {
                    continue;
                }
                // {
                //     "name": "api-version",
                //     "in": "query",
                //     "required": true,
                //     "type": "string"
                // }
                if (!validator || validator({ schema: paramEle })) {
                    paramExample[paramEle.name] = this.mockObj(paramEle.name, element, // use the original schema  containing "$ref" which will hit the cached value
                    paramExample[paramEle.name], new Set(), true);
                }
            }
        }
        return paramExample;
    }
    removeFromSet(schema, visited) {
        if ("$ref" in schema && visited.has(schema.$ref)) {
            visited.delete(schema.$ref);
        }
    }
    getCache(schema) {
        if ("$ref" in schema) {
            for (const cache of [this.exampleCache, this.mockCache]) {
                if (cache.has(schema.$ref.split("#")[1])) {
                    return cache.get(schema.$ref.split("#")[1]);
                }
            }
        }
        return undefined;
    }
    mockObj(objName, schema, example, visited, isRequest) {
        const cache = this.mockCachedObj(objName, schema, example, visited, isRequest);
        const validator = (0, exampleRule_1.getRuleValidator)(this.exampleRule).onSchema;
        return (0, exampleCache_1.reBuildExample)(cache, isRequest, schema, validator);
    }
    mockCachedObj(objName, schema, example, visited, isRequest, discriminatorValue = undefined) {
        if (!schema || typeof schema !== "object") {
            logging_1.log.warn(`invalid schema.`);
            return undefined;
        }
        // use visited set to avoid circular dependency
        if ("$ref" in schema && visited.has(schema.$ref)) {
            return undefined;
        }
        const cache = this.getCache(schema);
        if (cache) {
            return cache;
        }
        const definitionSpec = this.getDefSpec(schema, visited);
        if (util.isObject(definitionSpec)) {
            // circular inherit will not be handled
            const properties = this.getProperties(definitionSpec, visited);
            example = example || {};
            const discriminator = this.getDiscriminator(definitionSpec, visited);
            if (discriminator &&
                !discriminatorValue &&
                properties &&
                Object.keys(properties).includes(discriminator)) {
                return (this.mockForDiscriminator(definitionSpec, example, discriminator, isRequest, visited) ||
                    undefined);
            }
            else {
                Object.keys(properties).forEach((key) => {
                    // the objName is the discriminator when discriminatorValue is specified.
                    if (key === objName && discriminatorValue) {
                        example[key] = (0, exampleCache_1.createLeafItem)(discriminatorValue, (0, exampleCache_1.buildItemOption)(properties[key]));
                    }
                    else {
                        example[key] = this.mockCachedObj(key, properties[key], example[key], visited, isRequest, discriminatorValue);
                    }
                });
            }
            if ("additionalProperties" in definitionSpec && definitionSpec.additionalProperties) {
                const newKey = util.randomKey();
                if (newKey in properties) {
                    console.error(`generate additionalProperties for ${objName} fail`);
                }
                else {
                    example[newKey] = this.mockCachedObj(newKey, definitionSpec.additionalProperties, undefined, visited, isRequest, discriminatorValue);
                }
            }
        }
        else if (definitionSpec.type === "array") {
            example = example || [];
            const arrItem = this.mockCachedObj(`${objName}'s item`, definitionSpec.items, example[0], visited, isRequest);
            example = this.mocker.mock(definitionSpec, objName, arrItem);
        }
        else {
            /** type === number or integer  */
            example = example ? example : this.mocker.mock(definitionSpec, objName);
        }
        // return value for primary type: string, number, integer, boolean
        // "aaaa"
        // removeFromSet: once we try all roads started from present node, we should remove it and backtrack
        this.removeFromSet(schema, visited);
        let cacheItem;
        if (Array.isArray(example)) {
            const cacheChild = [];
            for (const item of example) {
                cacheChild.push(item);
            }
            cacheItem = (0, exampleCache_1.createTrunkItem)(cacheChild, (0, exampleCache_1.buildItemOption)(definitionSpec));
        }
        else if (typeof example === "object") {
            const cacheChild = {};
            for (const [key, item] of Object.entries(example)) {
                cacheChild[key] = item;
            }
            cacheItem = (0, exampleCache_1.createTrunkItem)(cacheChild, (0, exampleCache_1.buildItemOption)(definitionSpec));
        }
        else {
            cacheItem = (0, exampleCache_1.createLeafItem)(example, (0, exampleCache_1.buildItemOption)(definitionSpec));
        }
        cacheItem.isMocked = true;
        const requiredProperties = this.getRequiredProperties(definitionSpec);
        if (requiredProperties && requiredProperties.length > 0) {
            cacheItem.required = requiredProperties;
        }
        this.mockCache.checkAndCache(schema, cacheItem);
        return cacheItem;
    }
    /**
     * return all required properties of the object, including parent's properties defined by 'allOf'
     * It will not spread properties' properties.
     * @param definitionSpec
     */
    getRequiredProperties(definitionSpec) {
        var _a;
        let requiredProperties = Array.isArray(definitionSpec.required)
            ? definitionSpec.required
            : [];
        (_a = definitionSpec.allOf) === null || _a === void 0 ? void 0 : _a.map((item) => {
            requiredProperties = [
                ...requiredProperties,
                ...this.getRequiredProperties(this.getDefSpec(item, new Set())),
            ];
        });
        return requiredProperties;
    }
    // TODO: handle discriminator without enum options
    mockForDiscriminator(schema, example, discriminator, isRequest, visited) {
        const disDetail = this.getDefSpec(schema, visited);
        if (disDetail.discriminatorMap && Object.keys(disDetail.discriminatorMap).length > 0) {
            const properties = this.getProperties(disDetail, new Set());
            let discriminatorValue;
            if (properties[discriminator] && Array.isArray(properties[discriminator].enum)) {
                discriminatorValue = properties[discriminator].enum[0];
            }
            else {
                discriminatorValue = Object.keys(disDetail.discriminatorMap)[0];
            }
            const discriminatorSpec = disDetail.discriminatorMap[discriminatorValue];
            if (!discriminatorSpec) {
                this.removeFromSet(schema, visited);
                return example;
            }
            const cacheItem = this.mockCachedObj(discriminator, discriminatorSpec, {}, new Set(), isRequest, discriminatorValue) || undefined;
            this.removeFromSet(schema, visited);
            return cacheItem;
        }
        this.removeFromSet(schema, visited);
        return undefined;
    }
    // {
    //  "$ref": "#/parameters/ApiVersionParameter"
    // },
    // to
    // {
    //     "name": "api-version",
    //     "in": "query",
    //     "required": true,
    //     "type": "string"
    // }
    getDefSpec(schema, visited) {
        if ("$ref" in schema) {
            visited.add(schema.$ref);
        }
        const content = this.jsonLoader.resolveRefObj(schema);
        if (!content) {
            return undefined;
        }
        return content;
    }
    getProperties(definitionSpec, visited) {
        var _a;
        let properties = {};
        (_a = definitionSpec.allOf) === null || _a === void 0 ? void 0 : _a.map((item) => {
            properties = {
                ...properties,
                ...this.getProperties(this.getDefSpec(item, visited), visited),
            };
            this.removeFromSet(item, visited);
        });
        return {
            ...properties,
            ...definitionSpec.properties,
        };
    }
    getDiscriminator(definitionSpec, visited) {
        var _a;
        let discriminator = undefined;
        if (definitionSpec.discriminator) {
            return definitionSpec.discriminator;
        }
        (_a = definitionSpec.allOf) === null || _a === void 0 ? void 0 : _a.some((item) => {
            discriminator = this.getDiscriminator(this.getDefSpec(item, visited), visited);
            this.removeFromSet(item, visited);
            return !!discriminator;
        });
        return discriminator;
    }
}
exports.default = SwaggerMocker;
