// oav/dist/lib/generator/translator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const logging_1 = require("../util/logging");
const exampleCache_1 = require("./exampleCache");
const utils = __importStar(require("./util"));
const exampleRule_1 = require("./exampleRule");
class Translator {
    constructor(jsonLoader, payloadCache, mocker) {
        this.jsonLoader = jsonLoader;
        this.payloadCache = payloadCache;
        this.mocker = mocker;
    }
    setRule(exampleRule) {
        this.exampleRule = exampleRule;
    }
    extractRequest(specItem, request) {
        const path = request.url.split("?")[0];
        const pathValues = this.getPathParameters(specItem.path, path);
        const queryValues = this.getQueryParameters(specItem, request.query);
        const parameters = this.getBodyParameters(specItem, request.body);
        const result = {
            ...parameters,
            ...pathValues,
            ...queryValues,
        };
        this.maskSpecialValue(result);
        return result;
    }
    // for example cache.
    extractParameters(specItem, request) {
        const parameters = this.getMatchedParameters(specItem.content.parameters, request);
        const result = {
            ...parameters,
        };
        this.maskSpecialValue(result);
        return result;
    }
    maskSpecialValue(requestExample) {
        if (requestExample.subscriptionId) {
            requestExample.subscriptionId = requestExample.subscriptionId.replace(/[a-z0-9]/g, "0");
        }
    }
    getMatchedParameters(paramterSchema, parameters) {
        const bodyRes = {};
        const validator = (0, exampleRule_1.getRuleValidator)(this.exampleRule).onParameter;
        paramterSchema.forEach((item) => {
            const itemSchema = this.getDefSpec(item);
            if (parameters[itemSchema.name]) {
                if (!validator || validator({ schema: itemSchema })) {
                    bodyRes[itemSchema.name] = this.filterBodyContent(parameters[itemSchema.name], itemSchema.in === "body" ? itemSchema.schema : item);
                }
            }
        });
        return bodyRes;
    }
    getQueryParameters(specItem, body) {
        if (!specItem || !body) {
            return;
        }
        const parametersSpec = specItem.content.parameters
            .map((item) => this.getDefSpec(item))
            .filter((item) => "in" in item && item.in === "query");
        if (parametersSpec.length === 0) {
            logging_1.log.info("no query parameter definition in spec file");
            return;
        }
        return this.getMatchedParameters(parametersSpec, body);
    }
    // for payload cache
    getBodyParameters(specItem, body) {
        if (!specItem || !body) {
            return;
        }
        const parametersSpec = specItem.content.parameters
            .map((item) => this.getDefSpec(item))
            .filter((item) => "in" in item && item.in === "body");
        if (parametersSpec.length === 0) {
            logging_1.log.info("no body parameter definition in spec file");
            return;
        }
        const bodyParameterName = parametersSpec[0].name;
        const exampleBody = {};
        exampleBody[bodyParameterName] = body;
        return this.getMatchedParameters(parametersSpec, exampleBody);
    }
    /**
     * filter body's fields which is not defined in spec file
     * @param body
     * @param schema
     */
    filterBodyContent(body, schema, isRequest = true) {
        const cache = this.cacheBodyContent(body, schema, isRequest);
        const validator = (0, exampleRule_1.getRuleValidator)(this.exampleRule).onSchema;
        return (0, exampleCache_1.reBuildExample)(cache, isRequest, schema, validator);
    }
    cacheBodyContent(body, schema, isRequest) {
        var _a, _b;
        if (!schema) {
            return undefined;
        }
        if (schema.$ref && this.payloadCache.get(schema.$ref.split("#")[1])) {
            return this.payloadCache.get(schema.$ref.split("#")[1]);
        }
        const definitionSpec = this.getDefSpec(schema);
        const bodyContent = {};
        let cacheItem;
        if (utils.isObject(definitionSpec)) {
            const properties = this.getProperties(definitionSpec);
            if (body) {
                Object.keys(body)
                    .filter((key) => key in properties)
                    .forEach((key) => {
                    bodyContent[key] = this.cacheBodyContent(body[key], properties[key], isRequest);
                });
            }
            // to mock the properties that not exists in the body.
            // it's not needed when generating from payload.
            if (this.mocker !== undefined) {
                Object.keys(properties)
                    .filter((key) => !(body === null || body === void 0 ? void 0 : body[key]))
                    .forEach((key) => {
                    var _a;
                    bodyContent[key] = (_a = this.mocker) === null || _a === void 0 ? void 0 : _a.getMockCachedObj(key, properties[key], isRequest);
                });
            }
            cacheItem = (0, exampleCache_1.createTrunkItem)(bodyContent, (0, exampleCache_1.buildItemOption)(definitionSpec));
        }
        else if (definitionSpec.type === "array") {
            if (body) {
                const result = body.map((i) => {
                    return this.cacheBodyContent(i, schema.items, isRequest);
                });
                cacheItem = (0, exampleCache_1.createTrunkItem)(result, (0, exampleCache_1.buildItemOption)(definitionSpec));
            }
            else if (this.mocker !== undefined) {
                return (_a = this.mocker) === null || _a === void 0 ? void 0 : _a.getMockCachedObj("mock array", schema, isRequest);
            }
        }
        else {
            if (body !== undefined) {
                cacheItem = (0, exampleCache_1.createLeafItem)(body, (0, exampleCache_1.buildItemOption)(definitionSpec));
            }
            else if (this.mocker) {
                return (_b = this.mocker) === null || _b === void 0 ? void 0 : _b.getMockCachedObj("", schema, isRequest);
            }
        }
        const requiredProperties = this.getRequiredProperties(definitionSpec);
        if (requiredProperties && requiredProperties.length > 0 && cacheItem !== undefined) {
            cacheItem.required = requiredProperties;
        }
        this.payloadCache.checkAndCache(schema, cacheItem, isRequest);
        return cacheItem;
    }
    /**
     * return all properties of the object, including parent's properties defined by 'allOf'
     * It will not spread properties' properties.
     * @param definitionSpec
     */
    getProperties(definitionSpec) {
        var _a;
        let properties = {};
        (_a = definitionSpec.allOf) === null || _a === void 0 ? void 0 : _a.map((item) => {
            properties = {
                ...properties,
                ...this.getProperties(this.getDefSpec(item)),
            };
        });
        return {
            ...properties,
            ...definitionSpec.properties,
        };
    }
    /**
     * return all required properties of the object, including parent's properties defined by 'allOf'
     * It will not spread properties' properties.
     * @param definitionSpec
     */
    getRequiredProperties(definitionSpec) {
        var _a;
        let requiredProperties = Array.isArray(definitionSpec.required)
            ? definitionSpec.required
            : [];
        (_a = definitionSpec.allOf) === null || _a === void 0 ? void 0 : _a.map((item) => {
            requiredProperties = [
                ...requiredProperties,
                ...this.getRequiredProperties(this.getDefSpec(item)),
            ];
        });
        return requiredProperties;
    }
    getDefSpec(item) {
        if (item) {
            return this.jsonLoader.resolveRefObj(item);
        }
    }
    extractResponse(specItem, response, statusCode) {
        const specItemContent = specItem.content;
        const resp = {};
        if (statusCode === "201" || statusCode === "202") {
            resp.headers = {
                Location: response.headers && "location" in response.headers
                    ? response.headers.location
                    : undefined,
                "Azure-AsyncOperation": response.headers && "azure-AsyncOperation" in response.headers
                    ? response.headers["azure-AsyncOperation"]
                    : undefined,
            };
        }
        if ("schema" in specItemContent.responses[statusCode] && response.body) {
            resp.body =
                this.filterBodyContent(response.body, specItemContent.responses[statusCode].schema, false) || {};
        }
        return resp;
    }
    getPathParameters(pathName, path) {
        const pathTemplateItems = pathName.split("/");
        const urlItems = path.split("/") || [];
        let res = {};
        pathTemplateItems.forEach((item, idx) => {
            if (item !== urlItems[idx]) {
                const matchParam = /^{(.*)}$/.exec(item);
                if (matchParam && matchParam.length >= 2) {
                    const paramName = matchParam[1];
                    res = {
                        ...res,
                        [paramName]: urlItems[idx],
                    };
                }
            }
        });
        return res;
    }
}
exports.default = Translator;
