// oav/dist/lib/generator/util.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBaseName = getBaseName;
exports.isObject = isObject;
exports.randomKey = randomKey;
exports.getRp = getRp;
exports.fileExist = fileExist;
exports.isLongRunning = isLongRunning;
exports.readPayloadFile = readPayloadFile;
exports.getPollingAttr = getPollingAttr;
exports.getPollingUrl = getPollingUrl;
exports.updateExmAndSpecFile = updateExmAndSpecFile;
exports.referenceExmInSpec = referenceExmInSpec;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const url_1 = require("url");
const ATTR_MAP = {
    location: "location",
    "azure-async-operation": "Azure-AsyncOperation",
    "original-uri": "original-uri",
};
function getBaseName(filePath) {
    return path.basename(filePath);
}
function isObject(definitionSpec) {
    return (definitionSpec.type === "object" || "properties" in definitionSpec || "allOf" in definitionSpec);
}
function randomKey() {
    return `key${Math.ceil(Math.random() * 10000)}`;
}
function getRp(specFilePath) {
    return "microsoft." + specFilePath.split("/").slice(-1)[0].split(".")[0].toLowerCase();
}
function fileExist(...pathName) {
    return fs.existsSync(path.resolve(...pathName));
}
function isLongRunning(specItem) {
    return specItem.content["x-ms-long-running-operation"];
}
function readPayloadFile(payloadDir, operationId) {
    if (!fileExist(payloadDir, operationId)) {
        return null;
    }
    const inputPath = path.resolve(payloadDir, operationId);
    const filenames = fs.readdirSync(inputPath).filter((filename) => {
        return (filename.endsWith(".json") &&
            ["200", "201", "202", "204"].includes(filename.split(".json")[0]));
    });
    if (filenames.length === 0) {
        return null;
    }
    const payload = {};
    filenames.forEach((filename) => {
        const statusCode = filename.split(".json")[0];
        const data = fs.readFileSync(path.resolve(inputPath, filename));
        payload[statusCode] = JSON.parse(data.toString());
    });
    return payload;
}
function getPollingAttr(specItem) {
    if (!specItem.content["x-ms-long-running-operation"]) {
        return;
    }
    const lrOption = specItem.content["x-ms-long-running-operation-options"];
    if (lrOption && lrOption["final-state-via"]) {
        return ATTR_MAP[lrOption["final-state-via"]];
    }
    return "location";
}
function getPollingUrl(payload, specItem) {
    let attr = "location";
    const asyncOperation = payload.liveResponse.headers["azure-AsyncOperation"];
    const lrOption = specItem.content["x-ms-long-running-operation-options"];
    let location = payload.liveResponse.headers.location;
    if (lrOption && lrOption["final-state-via"]) {
        attr = ATTR_MAP[lrOption["final-state-via"]];
        location = payload.liveResponse.headers[attr];
    }
    else if (asyncOperation) {
        location = asyncOperation;
    }
    const url = new url_1.URL(location);
    return url.pathname + url.search;
}
function updateExmAndSpecFile(example, newSpec, specFilePath, exampleName) {
    const outputDir = path.resolve(specFilePath, "../", "examples");
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir);
    }
    const outputPath = path.resolve(outputDir, exampleName);
    console.log("example file path: " + outputPath);
    fs.writeFileSync(outputPath, JSON.stringify(example, null, 2), "utf8");
    if (newSpec) {
        //log.info("updated swagger file path: " + specFilePath);
        fs.writeFileSync(specFilePath, JSON.stringify(newSpec, null, 2), "utf8");
    }
}
function referenceExmInSpec(specFilePath, apiPath, methodName, exampleName) {
    const data = fs.readFileSync(specFilePath);
    const spec = JSON.parse(data.toString());
    if (!spec.paths[apiPath][methodName]["x-ms-examples"]) {
        spec.paths[apiPath][methodName]["x-ms-examples"] = {};
    }
    if (!(exampleName in spec.paths[apiPath][methodName]["x-ms-examples"])) {
        spec.paths[apiPath][methodName]["x-ms-examples"][exampleName] = {
            $ref: `./examples/${exampleName}.json`,
        };
        return spec;
    }
}
