// @microsoft.azure/openapi-validator-rulesets/dist/native/rulesets/arm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.armRuleset = void 0;
const openapi_validator_core_1 = require("@microsoft.azure/openapi-validator-core");
const arm_resource_validation_1 = require("../functions/arm-resource-validation");
const implement_private_endpoint_apis_1 = require("../functions/implement-private-endpoint-apis");
const provider_namespace_1 = require("../functions/provider-namespace");
exports.armRuleset = {
    documentationUrl: "https://github.com/Azure/azure-openapi-validator/blob/develop/docs/rules.md",
    rules: {
        AllTrackedResourcesMustHaveDelete: {
            category: "ARMViolation",
            openapiType: openapi_validator_core_1.OpenApiTypes.arm,
            severity: "error",
            given: "$",
            then: {
                execute: arm_resource_validation_1.allResourcesHaveDelete,
                options: {
                    isTrackedResource: true,
                },
            },
        },
        AllProxyResourcesShouldHaveDelete: {
            category: "ARMViolation",
            openapiType: openapi_validator_core_1.OpenApiTypes.arm,
            severity: "warning",
            given: "$",
            then: {
                execute: arm_resource_validation_1.allResourcesHaveDelete,
                options: {
                    isTrackedResource: false,
                },
            },
        },
        RequiredPropertiesMissingInResourceModel: {
            description: `Per [ARM guidelines](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md), a 'Resource' model must have the 'name', 'id' and 'type' properties defined as 'readOnly' in its hierarchy.`,
            category: "ARMViolation",
            openapiType: openapi_validator_core_1.OpenApiTypes.arm,
            severity: "error",
            given: "$",
            then: {
                execute: arm_resource_validation_1.resourcesHaveRequiredProperties,
            },
        },
        ArmResourcePropertiesBag: {
            description: "Per [ARM guidelines](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md), top level properties should not be repeated inside the properties bag for ARM resources.",
            category: "ARMViolation",
            openapiType: openapi_validator_core_1.OpenApiTypes.arm,
            severity: "error",
            given: "$",
            then: {
                execute: arm_resource_validation_1.armResourcePropertiesBag,
            },
        },
        TrackedResourcePatchOperation: {
            category: "ARMViolation",
            openapiType: openapi_validator_core_1.OpenApiTypes.arm,
            severity: "error",
            given: "$",
            then: {
                execute: arm_resource_validation_1.trackedResourcesHavePatch,
            },
        },
        BodyTopLevelProperties: {
            description: "Per [ARM guidelines](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md), top level properties of a resource should be only ones from the allowed set.",
            category: "ARMViolation",
            openapiType: openapi_validator_core_1.OpenApiTypes.arm,
            severity: "error",
            given: "$",
            then: {
                execute: arm_resource_validation_1.bodyTopLevelProperties,
            },
        },
        TrackedResourceBeyondsThirdLevel: {
            category: "ARMViolation",
            openapiType: openapi_validator_core_1.OpenApiTypes.arm,
            severity: "error",
            given: "$",
            then: {
                execute: arm_resource_validation_1.trackedResourceBeyondsThirdLevel,
            },
        },
        TrackedResourcesMustHavePut: {
            category: "ARMViolation",
            openapiType: openapi_validator_core_1.OpenApiTypes.arm,
            severity: "error",
            given: "$",
            then: {
                execute: arm_resource_validation_1.trackedResourcesMustHavePut,
            },
        },
        PathResourceProviderMatchNamespace: {
            description: `Verifies whether the last resource provider matches namespace or not. E.g the path /providers/Microsoft.Compute/virtualMachines/{vmName}/providers/Microsoft.Insights/extResource/{extType}' is allowed only if Microsoft.Insights matches the namespace (Microsoft.Insights).`,
            severity: "error",
            category: "ARMViolation",
            openapiType: openapi_validator_core_1.OpenApiTypes.arm,
            given: ["$[paths,'x-ms-paths'].*~"],
            then: {
                execute: provider_namespace_1.providerNamespace,
            },
        },
        OperationsAPIImplementation: {
            description: "Per [ARM guidelines](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md), each RP must expose an operations API that returns information about all the operations available with the service.",
            category: "ARMViolation",
            openapiType: openapi_validator_core_1.OpenApiTypes.arm,
            scope: "Global",
            severity: "error",
            given: "$",
            then: {
                execute: arm_resource_validation_1.operationsAPIImplementation,
            },
        },
        XmsPageableListByRGAndSubscriptions: {
            description: `When a tracked resource has list by resource group and subscription operations, the x-ms-pageable extension values must be same for both operations. A tracked resource is a resource with a 'location' property as required. If this rule flags a resource which does not have a 'location' property, then it might be a false positive.`,
            severity: "warning",
            category: "ARMViolation",
            openapiType: openapi_validator_core_1.OpenApiTypes.arm,
            then: {
                execute: arm_resource_validation_1.xmsPageableListByRGAndSubscriptions,
            },
        },
        ImplementPrivateEndpointAPIs: {
            description: "This rule is to check if all the APIs for private endpoint are implemented.",
            severity: "warning",
            category: "SDKViolation",
            scope: "Global",
            given: "$.paths",
            openapiType: openapi_validator_core_1.OpenApiTypes.arm,
            then: {
                execute: implement_private_endpoint_apis_1.implementPrivateEndpointApis,
            },
        },
    },
};
exports.default = exports.armRuleset;
