// @microsoft.azure/openapi-validator-rulesets/dist/spectral/functions/path-for-tracked-resource-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("./utils");
function matchAnyPatterns(patterns, path) {
    return patterns.some((p) => p.test(path));
}
function notMatchPatterns(invalidPatterns, path) {
    return invalidPatterns.every((p) => !p.test(path));
}
function verifyResourceGroupScope(path) {
    const patterns = [
        /^\/subscriptions\/{\w+}\/resourceGroups\/{\w+}\/providers\/.+/gi,
        /^\/subscriptions\/{\w+}\/resourceGroups\/{\w+}\/providers\/\w+\.\w+\/\w+.*/gi,
    ];
    return matchAnyPatterns(patterns, path);
}
function verifyNestResourceGroupScope(path) {
    const invalidPatterns = [
        /^\/subscriptions\/{\w+}\/resourceGroups\/{\w+}\/providers\/\w+\.\w+\/\w+\/{\w+}(?:\/\w+\/(?!default)\w+){1,2}$/gi,
        /^\/subscriptions\/{\w+}\/resourceGroups\/{\w+}\/providers\/\w+\.\w+(?:\/\w+\/(default|{\w+})){1,2}(?:\/\w+\/(?!default)\w+)+$/gi,
        /^\/subscriptions\/{\w+}\/resourceGroups\/{\w+}\/providers\/\w+\.\w+\/\w+\/(?:\/\w+\/(default|{\w+})){0,3}{\w+}(?:\/{\w+})+.*$/gi,
        /^\/subscriptions\/{\w+}\/resourceGroups\/{\w+}\/providers\/\w+\.\w+(?:\/\w+\/(default|{\w+})){0,2}(?:\/\w+\/(?!default)\w+)+\/{\w+}.*$/gi,
    ];
    return notMatchPatterns(invalidPatterns, path);
}
function checkTrackedForPut(allParams, path) {
    const bodyParam = allParams.find((p) => p.in === "body");
    if (bodyParam) {
        const properties = (0, utils_1.getProperties)(bodyParam.schema);
        if ("location" in properties) {
            if (!verifyResourceGroupScope(path[1]) || !verifyNestResourceGroupScope(path[1])) {
                return true;
            }
        }
    }
    return false;
}
function checkTrackedForGet(allParams, path) {
    const properties = (0, utils_1.getProperties)(allParams);
    if ("location" in properties) {
        if (!verifyResourceGroupScope(path[1]) || !verifyNestResourceGroupScope(path[1])) {
            return true;
        }
    }
    return false;
}
const pathForTrackedResourceTypes = (pathItem, _opts, paths) => {
    if (pathItem === null || typeof pathItem !== "object") {
        return [];
    }
    const path = paths.path || [];
    const errors = [];
    if (pathItem["put"] && Array.isArray(pathItem["put"].parameters)) {
        const allParams = [...pathItem["put"].parameters];
        if (checkTrackedForPut(allParams, path)) {
            errors.push({
                message: "The path must be under a subscription and resource group for tracked resource types.",
                path,
            });
        }
    }
    if (pathItem["get"]) {
        const resp = Object.keys(pathItem.get.responses).find((code) => code.startsWith("2"));
        if (!resp) {
            return [];
        }
        const responseSchema = pathItem.get.responses[resp].schema || {};
        if (checkTrackedForGet(responseSchema, path)) {
            errors.push({
                message: "The path must be under a subscription and resource group for tracked resource types.",
                path,
            });
        }
    }
    return errors;
};
exports.default = pathForTrackedResourceTypes;
