// @microsoft.azure/openapi-validator-rulesets/dist/spectral/functions/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deepFindObjectKeyPath = exports.getResourcesPathHierarchyBasedOnResourceType = exports.isPointOperation = exports.isSchemaEqual = exports.isXmsResource = exports.getReturnedSchema = exports.getReturnedType = exports.isPageableOperation = exports.getGetOperationSchema = exports.diffSchema = exports.jsonPath = exports.getRequiredProperties = exports.findBodyParam = exports.getProperty = exports.getAllPropertiesIncludingDeeplyNestedProperties = exports.getProperties = exports.isPathOfExtensionResource = exports.isLatestCommonTypesVersionForFile = exports.LATEST_VERSION_BY_COMMON_TYPES_FILENAME = void 0;
const tslib_1 = require("tslib");
const lodash_1 = tslib_1.__importDefault(require("lodash"));
exports.LATEST_VERSION_BY_COMMON_TYPES_FILENAME = new Map([
    ["types.json", "v6"],
    ["managedidentity.json", "v6"],
    ["privatelinks.json", "v6"],
    ["customermanagedkeys.json", "v5"],
    ["managedidentitywithdelegation.json", "v5"],
    ["networksecurityperimeter.json", "v5"],
    ["mobo.json", "v5"],
]);
function isLatestCommonTypesVersionForFile(version, fileName) {
    return exports.LATEST_VERSION_BY_COMMON_TYPES_FILENAME.get(fileName) === version.toLowerCase();
}
exports.isLatestCommonTypesVersionForFile = isLatestCommonTypesVersionForFile;
const ExtensionResourceFullyQualifiedPathReg = new RegExp(".+/providers/.+/providers/.+$", "gi");
const ExtensionResourceReg = new RegExp("^((/{\\w+})|({\\w+}))/providers/.+$", "gi");
function isPathOfExtensionResource(path) {
    return !!path.match(ExtensionResourceFullyQualifiedPathReg) || !!path.match(ExtensionResourceReg);
}
exports.isPathOfExtensionResource = isPathOfExtensionResource;
function getProperties(schema) {
    if (!schema) {
        return {};
    }
    let properties = {};
    if (schema.allOf && Array.isArray(schema.allOf)) {
        schema.allOf.forEach((base) => {
            properties = { ...getProperties(base), ...properties };
        });
    }
    if (schema.properties) {
        properties = { ...properties, ...schema.properties };
    }
    return properties;
}
exports.getProperties = getProperties;
function getAllPropertiesIncludingDeeplyNestedProperties(schema, properties) {
    if (!schema) {
        return {};
    }
    if (schema.allOf && Array.isArray(schema.allOf)) {
        schema.allOf.forEach((base) => {
            getAllPropertiesIncludingDeeplyNestedProperties(base, properties);
        });
    }
    if (schema.properties) {
        const props = schema.properties;
        Object.entries(props).forEach(([key, value]) => {
            if (!value.properties) {
                properties.push(Object.fromEntries([[key, value]]));
            }
            else {
                getAllPropertiesIncludingDeeplyNestedProperties(props[key], properties);
            }
        });
    }
    return properties;
}
exports.getAllPropertiesIncludingDeeplyNestedProperties = getAllPropertiesIncludingDeeplyNestedProperties;
function getProperty(schema, propName) {
    if (!schema) {
        return {};
    }
    if (schema.allOf && Array.isArray(schema.allOf)) {
        for (const base of schema.allOf) {
            const result = getProperty(base, propName);
            if (result) {
                return result;
            }
        }
    }
    if (schema.properties) {
        if (propName in schema.properties) {
            return schema.properties[propName];
        }
    }
    return undefined;
}
exports.getProperty = getProperty;
function findBodyParam(params) {
    const isBody = (elem) => elem.in === "body";
    if (params && Array.isArray(params)) {
        return params.filter(isBody).shift();
    }
    return undefined;
}
exports.findBodyParam = findBodyParam;
function getRequiredProperties(schema) {
    if (!schema) {
        return [];
    }
    let requires = [];
    if (schema.allOf && Array.isArray(schema.allOf)) {
        schema.allOf.forEach((base) => {
            requires = [...getRequiredProperties(base), ...requires];
        });
    }
    if (schema.required) {
        requires = [...schema.required, requires];
    }
    return requires;
}
exports.getRequiredProperties = getRequiredProperties;
function jsonPath(paths, root) {
    let result = undefined;
    paths.some((p) => {
        if (typeof root !== "object" && root !== null) {
            result = undefined;
            return true;
        }
        root = root[p];
        result = root;
        return false;
    });
    return result;
}
exports.jsonPath = jsonPath;
function diffSchema(a, b) {
    const notMatchedProperties = [];
    function diffSchemaInternal(a, b, paths) {
        if (!(a || b)) {
            return;
        }
        if (a && b) {
            const propsA = getProperties(a);
            const propsB = getProperties(b);
            Object.keys(propsA).forEach((p) => {
                if (propsB[p]) {
                    diffSchemaInternal(propsA[p], propsB[p], [...paths, p]);
                }
                else {
                    notMatchedProperties.push([...paths, p].join("."));
                }
            });
        }
    }
    diffSchemaInternal(a, b, []);
    return notMatchedProperties;
}
exports.diffSchema = diffSchema;
function getGetOperationSchema(paths, ctx) {
    var _a, _b, _c;
    const getOperationPath = [...paths, "get"];
    const getOperation = jsonPath(getOperationPath, (_a = ctx === null || ctx === void 0 ? void 0 : ctx.documentInventory) === null || _a === void 0 ? void 0 : _a.resolved);
    if (!getOperation) {
        return undefined;
    }
    return ((_b = getOperation === null || getOperation === void 0 ? void 0 : getOperation.responses["200"]) === null || _b === void 0 ? void 0 : _b.schema) || ((_c = getOperation === null || getOperation === void 0 ? void 0 : getOperation.responses["201"]) === null || _c === void 0 ? void 0 : _c.schema);
}
exports.getGetOperationSchema = getGetOperationSchema;
function isPageableOperation(operation) {
    return !!(operation === null || operation === void 0 ? void 0 : operation["x-ms-pageable"]);
}
exports.isPageableOperation = isPageableOperation;
function getReturnedType(operation) {
    var _a;
    const succeededCodes = ["200", "201", "202"];
    for (const code of succeededCodes) {
        const response = operation.responses[code];
        if (response) {
            return (_a = response === null || response === void 0 ? void 0 : response.schema) === null || _a === void 0 ? void 0 : _a.$ref;
        }
    }
}
exports.getReturnedType = getReturnedType;
function getReturnedSchema(operation) {
    const succeededCodes = ["200", "201"];
    for (const code of succeededCodes) {
        const response = operation.responses[code];
        if (response === null || response === void 0 ? void 0 : response.schema) {
            return response === null || response === void 0 ? void 0 : response.schema;
        }
    }
}
exports.getReturnedSchema = getReturnedSchema;
function isXmsResource(schema) {
    if (!schema) {
        return false;
    }
    if (schema["x-ms-azure-resource"]) {
        return true;
    }
    if (schema.allOf && Array.isArray(schema.allOf)) {
        for (const base of schema.allOf) {
            if (isXmsResource(base)) {
                return true;
            }
        }
    }
    return false;
}
exports.isXmsResource = isXmsResource;
function isSchemaEqual(a, b) {
    if (a && b) {
        const propsA = Object.getOwnPropertyNames(a);
        const propsB = Object.getOwnPropertyNames(b);
        if (propsA.length === propsB.length) {
            if (propsA.length === 0) {
                return true;
            }
            for (let i = 0; i < propsA.length; i++) {
                const propsAName = propsA[i];
                const [propA, propB] = [a[propsAName], b[propsAName]];
                if (typeof propA === "object") {
                    if (!isSchemaEqual(propA, propB)) {
                        return false;
                    }
                    else if (i === propsA.length - 1) {
                        return true;
                    }
                }
                else if (propA !== propB) {
                    return false;
                }
                else if (propA === propB && i === propsA.length - 1) {
                    return true;
                }
            }
        }
    }
    return false;
}
exports.isSchemaEqual = isSchemaEqual;
const providerAndNamespace = "/providers/[^/]+";
const resourceTypeAndResourceName = "(?:/\\w+/default|/\\w+/{[^/]+})";
const queryParam = "(?:\\?\\w+)";
const resourcePathRegEx = new RegExp(`${providerAndNamespace}${resourceTypeAndResourceName}+${queryParam}?$`, "gi");
function isPointOperation(path) {
    const index = path.lastIndexOf("/providers/");
    if (index === -1) {
        return false;
    }
    const lastProvider = path.substr(index);
    const matches = lastProvider.match(resourcePathRegEx);
    if (matches) {
        return true;
    }
    return false;
}
exports.isPointOperation = isPointOperation;
function getResourcesPathHierarchyBasedOnResourceType(path) {
    const index = path.lastIndexOf("/providers/");
    if (index === -1) {
        return [];
    }
    const lastProvider = path.substr(index);
    const result = [];
    const matches = lastProvider.match(resourcePathRegEx);
    if (matches && matches.length) {
        const match = matches[0];
        const resourcePathSegments = match.split("/").slice(3);
        for (const resourcePathSegment of resourcePathSegments) {
            if (resourcePathSegment.startsWith("{") || resourcePathSegment === "default") {
                continue;
            }
            result.push(resourcePathSegment);
        }
    }
    return result;
}
exports.getResourcesPathHierarchyBasedOnResourceType = getResourcesPathHierarchyBasedOnResourceType;
function deepFindObjectKeyPath(object, keyName, path = []) {
    if (!lodash_1.default.isObject(object)) {
        return [];
    }
    if (lodash_1.default.has(object, keyName)) {
        return lodash_1.default.concat(path, keyName);
    }
    for (const [key, value] of Object.entries(object)) {
        const result = deepFindObjectKeyPath(value, keyName, lodash_1.default.concat(path, key));
        if (result) {
            return result;
        }
    }
    return [];
}
exports.deepFindObjectKeyPath = deepFindObjectKeyPath;
