// @microsoft.azure/openapi-validator-rulesets/dist/spectral/functions/validate-patch-body-param-properties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validatePatchBodyParamProperties = void 0;
const spectral_core_1 = require("@stoplight/spectral-core");
const utils_1 = require("./utils");
exports.validatePatchBodyParamProperties = (0, spectral_core_1.createRulesetFunction)({
    input: null,
    options: {
        type: "object",
        properties: {
            should: {
                type: "array",
                items: {
                    type: "string",
                },
            },
            shouldNot: {
                type: "array",
                items: {
                    type: "string",
                },
            },
            propertiesThatMustNotBeInPropertiesBagAsWritable: {
                type: "array",
                items: {
                    type: "string",
                },
            },
        },
        additionalProperties: false,
    },
}, (patchOp, _opts, ctx) => {
    var _a, _b, _c, _d, _e, _f;
    if (patchOp === null || typeof patchOp !== "object") {
        return [];
    }
    const path = ctx.path || [];
    const errors = [];
    const bodyParameter = (_b = (_a = patchOp.parameters) === null || _a === void 0 ? void 0 : _a.find((p) => p.in === "body")) === null || _b === void 0 ? void 0 : _b.schema;
    if (bodyParameter) {
        const index = patchOp.parameters.findIndex((p) => p.in === "body");
        if (_opts.should) {
            const responseSchema = ((_d = (_c = patchOp.responses) === null || _c === void 0 ? void 0 : _c["200"]) === null || _d === void 0 ? void 0 : _d.schema) || ((_f = (_e = patchOp.responses) === null || _e === void 0 ? void 0 : _e["202"]) === null || _f === void 0 ? void 0 : _f.schema) || (0, utils_1.getGetOperationSchema)(path.slice(0, -1), ctx);
            _opts.should.forEach((p) => {
                var _a, _b;
                if (!((_a = (0, utils_1.getProperties)(bodyParameter)) === null || _a === void 0 ? void 0 : _a[p]) && ((_b = (0, utils_1.getProperties)(responseSchema)) === null || _b === void 0 ? void 0 : _b[p])) {
                    errors.push({
                        message: `The patch operation body parameter schema should contain property '${p}'.`,
                        path: [...path, "parameters", index],
                    });
                }
            });
        }
        if (_opts.shouldNot) {
            _opts.shouldNot.forEach((p) => {
                var _a;
                const property = (_a = (0, utils_1.getProperties)(bodyParameter)) === null || _a === void 0 ? void 0 : _a[p];
                if (property) {
                    let isPropertyReadOnly = false;
                    let isPropertyImmutable = false;
                    if (property["readOnly"] && property["readOnly"] === true) {
                        isPropertyReadOnly = true;
                    }
                    if (property["x-ms-mutability"] && Array.isArray(property["x-ms-mutability"])) {
                        const schemaArray = property["x-ms-mutability"];
                        if (!schemaArray.includes("update")) {
                            isPropertyImmutable = true;
                        }
                    }
                    if (!(isPropertyReadOnly || isPropertyImmutable)) {
                        errors.push({
                            message: `Mark the top-level property "${p}", specified in the patch operation body, as readOnly or immutable. You could also choose to remove it from the request payload of the Patch operation. This property is not patchable.`,
                            path: [...path, "parameters", index],
                        });
                    }
                }
            });
        }
        if (_opts.propertiesThatMustNotBeInPropertiesBagAsWritable) {
            _opts.propertiesThatMustNotBeInPropertiesBagAsWritable.forEach((p) => {
                var _a, _b;
                const propertiesProperty = (_a = (0, utils_1.getProperties)(bodyParameter)) === null || _a === void 0 ? void 0 : _a["properties"];
                const property = (_b = (0, utils_1.getProperties)(propertiesProperty)) === null || _b === void 0 ? void 0 : _b[p];
                if (property) {
                    let isPropertyReadOnly = false;
                    let isPropertyImmutable = false;
                    if (property["readOnly"] && property["readOnly"] === true) {
                        isPropertyReadOnly = true;
                    }
                    if (property["x-ms-mutability"] && Array.isArray(property["x-ms-mutability"])) {
                        const schemaArray = property["x-ms-mutability"];
                        if (!schemaArray.includes("update")) {
                            isPropertyImmutable = true;
                        }
                    }
                    if (!(isPropertyReadOnly || isPropertyImmutable)) {
                        errors.push({
                            message: `Mark the property "properties.${p}", specified in the patch operation body, as readOnly or immutable. You could also choose to remove it from the request payload of the Patch operation. This property is not patchable.`,
                            path: [...path, "parameters", index],
                        });
                    }
                }
            });
        }
    }
    return errors;
});
