// @microsoft.azure/openapi-validator-rulesets/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.disableRulesInRuleset = exports.deleteRulesPropertiesInPayloadNotValidForSpectralRules = exports.getNamesOfRulesInPayloadWithPropertySetToTrue = exports.nativeRulesets = exports.spectralRulesets = exports.spectralDataplaneRulesetFile = exports.spectralArmRulesetFile = exports.spectralCommonRulesetFile = void 0;
const tslib_1 = require("tslib");
const path_1 = require("path");
const arm_1 = tslib_1.__importDefault(require("./native/rulesets/arm"));
const common_1 = tslib_1.__importDefault(require("./native/rulesets/common"));
const az_arm_1 = tslib_1.__importDefault(require("./spectral/az-arm"));
const az_common_1 = tslib_1.__importDefault(require("./spectral/az-common"));
const az_dataplane_1 = tslib_1.__importDefault(require("./spectral/az-dataplane"));
const spectralRulesetDir = (0, path_1.join)(__dirname, "spectral");
exports.spectralCommonRulesetFile = (0, path_1.join)(spectralRulesetDir, "az-common.js");
exports.spectralArmRulesetFile = (0, path_1.join)(spectralRulesetDir, "az-arm.js");
exports.spectralDataplaneRulesetFile = (0, path_1.join)(spectralRulesetDir, "az-dataplane.js");
exports.spectralRulesets = {
    azARM: az_arm_1.default,
    azCommon: az_common_1.default,
    azDataplane: az_dataplane_1.default,
};
exports.nativeRulesets = {
    azCommon: common_1.default,
    azArm: arm_1.default,
};
var utils_1 = require("./spectral/utils");
Object.defineProperty(exports, "getNamesOfRulesInPayloadWithPropertySetToTrue", { enumerable: true, get: function () { return utils_1.getNamesOfRulesInPayloadWithPropertySetToTrue; } });
Object.defineProperty(exports, "deleteRulesPropertiesInPayloadNotValidForSpectralRules", { enumerable: true, get: function () { return utils_1.deleteRulesPropertiesInPayloadNotValidForSpectralRules; } });
Object.defineProperty(exports, "disableRulesInRuleset", { enumerable: true, get: function () { return utils_1.disableRulesInRuleset; } });
