// @microsoft.azure/openapi-validator/dist/jsonrpc/plugin-host.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoRestPluginHost = void 0;
const vscode_jsonrpc_1 = require("vscode-jsonrpc");
var IAutoRestPluginTarget_Types;
(function (IAutoRestPluginTarget_Types) {
    IAutoRestPluginTarget_Types.GetPluginNames = new vscode_jsonrpc_1.RequestType0("GetPluginNames");
    IAutoRestPluginTarget_Types.Process = new vscode_jsonrpc_1.RequestType2("Process");
})(IAutoRestPluginTarget_Types || (IAutoRestPluginTarget_Types = {}));
var IAutoRestPluginInitiator_Types;
(function (IAutoRestPluginInitiator_Types) {
    IAutoRestPluginInitiator_Types.ReadFile = new vscode_jsonrpc_1.RequestType2("ReadFile");
    IAutoRestPluginInitiator_Types.GetValue = new vscode_jsonrpc_1.RequestType2("GetValue");
    IAutoRestPluginInitiator_Types.ListInputs = new vscode_jsonrpc_1.RequestType1("ListInputs");
    IAutoRestPluginInitiator_Types.WriteFile = new vscode_jsonrpc_1.NotificationType4("WriteFile");
    IAutoRestPluginInitiator_Types.Message = new vscode_jsonrpc_1.NotificationType2("Message");
})(IAutoRestPluginInitiator_Types || (IAutoRestPluginInitiator_Types = {}));
class AutoRestPluginHost {
    constructor() {
        this.plugins = {};
    }
    Add(name, handler) {
        this.plugins[name] = handler;
    }
    async Run() {
        const channel = (0, vscode_jsonrpc_1.createMessageConnection)(process.stdin, process.stdout, {
            error(message) {
                console.error("error: ", message);
            },
            info(message) {
                console.error("info: ", message);
            },
            log(message) {
                console.error("log: ", message);
            },
            warn(message) {
                console.error("warn: ", message);
            },
        });
        channel.onRequest(IAutoRestPluginTarget_Types.GetPluginNames, async () => Object.keys(this.plugins));
        channel.onRequest(IAutoRestPluginTarget_Types.Process, async (pluginName, sessionId) => {
            try {
                const handler = this.plugins[pluginName];
                if (!handler) {
                    throw new Error(`Plugin host could not find requested plugin '${pluginName}'.`);
                }
                await handler({
                    async ReadFile(filename) {
                        return await channel.sendRequest(IAutoRestPluginInitiator_Types.ReadFile, sessionId, filename);
                    },
                    async GetValue(key) {
                        return await channel.sendRequest(IAutoRestPluginInitiator_Types.GetValue, sessionId, key);
                    },
                    async ListInputs() {
                        return await channel.sendRequest(IAutoRestPluginInitiator_Types.ListInputs, sessionId);
                    },
                    WriteFile(filename, content, sourceMap) {
                        channel.sendNotification(IAutoRestPluginInitiator_Types.WriteFile, sessionId, filename, content, sourceMap);
                    },
                    Message(message) {
                        channel.sendNotification(IAutoRestPluginInitiator_Types.Message, sessionId, message);
                    },
                });
                return true;
            }
            catch (e) {
                channel.sendNotification(IAutoRestPluginInitiator_Types.Message, sessionId, {
                    Channel: "fatal",
                    Text: "" + e,
                    Details: e,
                });
                return false;
            }
        });
        channel.listen();
    }
}
exports.AutoRestPluginHost = AutoRestPluginHost;
