// @microsoft.azure/openapi-validator-rulesets/dist/spectral/az-arm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const spectral_formats_1 = require("@stoplight/spectral-formats");
const spectral_functions_1 = require("@stoplight/spectral-functions");
const az_common_1 = tslib_1.__importDefault(require("./az-common"));
const arm_path_validation_1 = tslib_1.__importDefault(require("./functions/arm-path-validation"));
const body_param_repeated_info_1 = tslib_1.__importDefault(require("./functions/body-param-repeated-info"));
const camel_case_1 = require("./functions/camel-case");
const collection_object_properties_naming_1 = tslib_1.__importDefault(require("./functions/collection-object-properties-naming"));
const consistent_patch_properties_1 = require("./functions/consistent-patch-properties");
const consistent_response_schema_for_put_1 = tslib_1.__importDefault(require("./functions/consistent-response-schema-for-put"));
const delete_response_codes_1 = require("./functions/delete-response-codes");
const get_collection_only_has_value_and_next_link_1 = require("./functions/get-collection-only-has-value-and-next-link");
const get_response_codes_1 = require("./functions/get-response-codes");
const has_api_version_parameter_1 = tslib_1.__importDefault(require("./functions/has-api-version-parameter"));
const has_header_1 = tslib_1.__importDefault(require("./functions/has-header"));
const https_supported_scheme_1 = tslib_1.__importDefault(require("./functions/https-supported-scheme"));
const latest_version_of_common_types_must_be_used_1 = require("./functions/latest-version-of-common-types-must-be-used");
const location_must_have_xms_mutability_1 = tslib_1.__importDefault(require("./functions/location-must-have-xms-mutability"));
const lro_original_uri_1 = tslib_1.__importDefault(require("./functions/lro-original-uri"));
const lro_patch_202_1 = require("./functions/lro-patch-202");
const lro_patch_provisioning_state_specified_1 = tslib_1.__importDefault(require("./functions/lro-patch-provisioning-state-specified"));
const lro_put_provisioning_state_specified_1 = tslib_1.__importDefault(require("./functions/lro-put-provisioning-state-specified"));
const missing_segments_in_nested_resource_list_operation_1 = require("./functions/missing-segments-in-nested-resource-list-operation");
const no_duplicate_paths_for_scope_parameter_1 = tslib_1.__importDefault(require("./functions/no-duplicate-paths-for-scope-parameter"));
const no_error_code_responses_1 = require("./functions/no-error-code-responses");
const operations_api_schema_1 = tslib_1.__importDefault(require("./functions/operations-api-schema"));
const operations_api_tenant_level_only_1 = require("./functions/operations-api-tenant-level-only");
const parameter_not_defined_in_global_parameters_1 = require("./functions/parameter-not-defined-in-global-parameters");
const parameter_not_using_common_types_1 = require("./functions/parameter-not-using-common-types");
const parameters_in_point_get_1 = require("./functions/parameters-in-point-get");
const parameters_in_post_1 = require("./functions/parameters-in-post");
const patch_body_parameters_1 = tslib_1.__importDefault(require("./functions/patch-body-parameters"));
const patch_properties_correspond_to_put_properties_1 = require("./functions/patch-properties-correspond-to-put-properties");
const patch_response_codes_1 = require("./functions/patch-response-codes");
const path_for_tracked_resource_types_1 = tslib_1.__importDefault(require("./functions/path-for-tracked-resource-types"));
const path_segment_casing_1 = tslib_1.__importDefault(require("./functions/path-segment-casing"));
const post_response_codes_1 = require("./functions/post-response-codes");
const properties_type_object_no_definition_1 = require("./functions/properties-type-object-no-definition");
const provisioning_state_1 = tslib_1.__importDefault(require("./functions/provisioning-state"));
const provisioning_state_must_be_read_only_1 = require("./functions/provisioning-state-must-be-read-only");
const put_get_patch_schema_1 = tslib_1.__importDefault(require("./functions/put-get-patch-schema"));
const put_request_response_scheme_1 = require("./functions/put-request-response-scheme");
const put_response_codes_1 = require("./functions/put-response-codes");
const query_parameters_in_collection_get_1 = require("./functions/query-parameters-in-collection-get");
const request_body_must_exist_for_put_patch_1 = require("./functions/request-body-must-exist-for-put-patch");
const reserved_resource_names_model_as_enum_1 = require("./functions/reserved-resource-names-model-as-enum");
const resource_name_restriction_1 = tslib_1.__importDefault(require("./functions/resource-name-restriction"));
const response_schema_specified_for_success_status_code_1 = tslib_1.__importDefault(require("./functions/response-schema-specified-for-success-status-code"));
const security_definitions_structure_1 = require("./functions/security-definitions-structure");
const sku_validation_1 = tslib_1.__importDefault(require("./functions/sku-validation"));
const system_data_in_properties_bag_1 = require("./functions/system-data-in-properties-bag");
const tags_are_not_allowed_for_proxy_resources_1 = require("./functions/tags-are-not-allowed-for-proxy-resources");
const tenant_level_apis_not_allowed_1 = require("./functions/tenant-level-apis-not-allowed");
const tracked_extension_resources_are_not_allowed_1 = require("./functions/tracked-extension-resources-are-not-allowed");
const trackedresource_tags_property_in_request_1 = tslib_1.__importDefault(require("./functions/trackedresource-tags-property-in-request"));
const valid_query_parameters_for_point_operations_1 = require("./functions/valid-query-parameters-for-point-operations");
const validate_patch_body_param_properties_1 = require("./functions/validate-patch-body-param-properties");
const with_xms_resource_1 = tslib_1.__importDefault(require("./functions/with-xms-resource"));
const xms_long_running_operation_property_1 = tslib_1.__importDefault(require("./functions/xms-long-running-operation-property"));
const xms_pageable_for_list_calls_1 = tslib_1.__importDefault(require("./functions/xms-pageable-for-list-calls"));
const xms_secret_in_response_1 = tslib_1.__importDefault(require("./functions/xms-secret-in-response"));
const ruleset = {
    extends: [az_common_1.default],
    rules: {
        ApiHost: {
            description: "The host is required for management plane specs.",
            message: "{{description}}",
            severity: "error",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: ["$.host"],
            then: {
                function: spectral_functions_1.truthy,
            },
        },
        SubscriptionsAndResourceGroupCasing: {
            description: "The subscriptions and resourceGroup in resource uri should follow lower camel case.",
            message: "{{error}}",
            severity: "error",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: ["$.paths", "$.x-ms-paths"],
            then: {
                function: path_segment_casing_1.default,
                functionOptions: {
                    segments: ["resourceGroups", "subscriptions"],
                },
            },
        },
        PutResponseCodes: {
            rpcGuidelineCode: "RPC-Async-V1-01, RPC-Put-V1-11",
            description: "LRO and Synchronous PUT must have 200 & 201 return codes.",
            severity: "error",
            disableForTypeSpecDataPlane: true,
            disableForTypeSpecDataPlaneReason: "Covered by TSP's '@azure-tools/typespec-azure-resource-manager/arm-put-operation-response-codes' rule.",
            message: "{{error}}",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*[put]"],
            then: {
                function: put_response_codes_1.PutResponseCodes,
            },
        },
        ProvisioningStateSpecifiedForLROPut: {
            rpcGuidelineCode: "RPC-Async-V1-02",
            description: 'A LRO PUT operation\'s response schema must have "ProvisioningState" property specified for the 200 and 201 status codes.',
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*[put][?(@property === 'x-ms-long-running-operation' && @ === true)]^"],
            then: {
                function: lro_put_provisioning_state_specified_1.default,
            },
        },
        ProvisioningStateSpecifiedForLROPatch: {
            rpcGuidelineCode: "RPC-Async-V1-02",
            description: 'A long running Patch operation\'s response schema must have "ProvisioningState" property specified for the 200 status code.',
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: [
                "$[paths,'x-ms-paths'].*[patch][?(@property === 'x-ms-long-running-operation' && @ === true)]^.responses[?(@property == '200')]",
            ],
            then: {
                function: lro_patch_provisioning_state_specified_1.default,
            },
        },
        ProvisioningStateValidation: {
            rpcGuidelineCode: "RPC-Async-V1-03",
            description: "ProvisioningState must have terminal states: Succeeded, Failed and Canceled.",
            message: "{{error}}",
            severity: "error",
            disableForTypeSpecDataPlane: true,
            disableForTypeSpecDataPlaneReason: "Covered by TSP's '@azure-tools/typespec-azure-resource-manager/arm-resource-provisioning-state' rule.",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: ["$.definitions..provisioningState[?(@property === 'enum')]^", "$.definitions..ProvisioningState[?(@property === 'enum')]^"],
            then: {
                function: provisioning_state_1.default,
            },
        },
        LroLocationHeader: {
            rpcGuidelineCode: "RPC-Async-V1-07",
            description: "Location header must be supported for all async operations that return 202.",
            message: "A 202 response should include an Location response header.",
            severity: "error",
            disableForTypeSpecDataPlane: true,
            disableForTypeSpecDataPlaneReason: "Covered by TSP's '@azure-tools/typespec-azure-resource-manager/arm-location-header' rule.",
            formats: [spectral_formats_1.oas2],
            given: "$.paths[*][*].responses[?(@property == '202')]",
            then: {
                function: has_header_1.default,
                functionOptions: {
                    name: "Location",
                },
            },
        },
        PostResponseCodes: {
            rpcGuidelineCode: "RPC-Async-V1-11, RPC-Async-V1-14, RPC-POST-V1-02, RPC-POST-V1-03",
            description: "Synchronous POST must return 200 when a response body is required or 204 when no body is needed; LRO POST must initially return 202, with the final response returning 200 if a body is expected or 204 if not.",
            severity: "error",
            disableForTypeSpecDataPlane: true,
            disableForTypeSpecDataPlaneReason: "Covered by TSP's '@azure-tools/typespec-azure-resource-manager/arm-post-operation-response-codes' rule.",
            message: "{{error}}",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*[post]"],
            then: {
                function: post_response_codes_1.PostResponseCodes,
            },
        },
        XMSLongRunningOperationProperty: {
            rpcGuidelineCode: "RPC-Async-V1-15",
            description: "If an operation's (PUT/POST/PATCH/DELETE) responses have `Location` or `Azure-AsyncOperation` headers then it MUST have the property `x-ms-long-running-operation` set to `true`",
            message: "If an operation's (PUT/POST/PATCH/DELETE) responses have `Location` or `Azure-AsyncOperation` headers then it MUST have the property `x-ms-long-running-operation` set to `true`",
            severity: "error",
            formats: [spectral_formats_1.oas2],
            resolved: true,
            given: "$[paths,'x-ms-paths'].*[put,patch,post,delete]",
            then: {
                function: xms_long_running_operation_property_1.default,
            },
        },
        ProvisioningStateMustBeReadOnly: {
            rpcGuidelineCode: "RPC-Async-V1-16",
            description: "This is a rule introduced to validate if provisioningState property is set to readOnly or not.",
            message: "{{error}}",
            severity: "error",
            disableForTypeSpecDataPlane: true,
            disableForTypeSpecDataPlaneReason: "Covered by TSP's '@azure-tools/typespec-azure-resource-manager/arm-resource-provisioning-state' rule.",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*.*.responses.*.schema"],
            then: {
                function: provisioning_state_must_be_read_only_1.provisioningStateMustBeReadOnly,
            },
        },
        LroErrorContent: {
            rpcGuidelineCode: "RPC-Common-V1-05",
            description: "Error response content of long running operations must follow the error schema provided in the common types v2 and above.",
            message: "{{description}}",
            severity: "error",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*.*[?(@property === 'x-ms-long-running-operation' && @ === true)]^.responses[?(@property === 'default' || @property.startsWith('5') || @property.startsWith('4'))].schema.$ref",
            then: {
                function: spectral_functions_1.pattern,
                functionOptions: {
                    match: ".*/common-types/resource-management/v(([1-9]\\d+)|[2-9])/types.json#/definitions/ErrorResponse",
                },
            },
        },
        DeleteResponseCodes: {
            rpcGuidelineCode: "RPC-Delete-V1-01, RPC-Async-V1-09",
            description: "Synchronous DELETE must have 200 & 204 return codes and LRO DELETE must have 202 & 204 return codes.",
            severity: "error",
            message: "{{error}}",
            disableForTypeSpecDataPlane: true,
            disableForTypeSpecDataPlaneReason: "Covered by TSP's '@azure-tools/typespec-azure-resource-manager/arm-delete-operation-response-codes' rule.",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*[delete]"],
            then: {
                function: delete_response_codes_1.DeleteResponseCodes,
            },
        },
        DeleteMustNotHaveRequestBody: {
            rpcGuidelineCode: "RPC-Delete-V1-02",
            description: "The delete operation must not have a request body.",
            severity: "error",
            message: "{{description}}",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$.paths.*.delete.parameters[?(@.in === 'body')]",
            then: {
                function: spectral_functions_1.falsy,
            },
        },
        DeleteResponseBodyEmpty: {
            rpcGuidelineCode: "RPC-Delete-V1-04",
            description: "The delete response body must be empty.",
            message: "{{description}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*[delete].responses['200','204'].schema"],
            then: {
                function: spectral_functions_1.falsy,
            },
        },
        AvoidAdditionalProperties: {
            rpcGuidelineCode: "RPC-Policy-V1-05, RPC-Put-V1-23",
            description: "Definitions must not have properties named additionalProperties except for user defined tags or predefined references.",
            severity: "error",
            message: "{{description}}",
            disableForTypeSpecDataPlane: true,
            disableForTypeSpecDataPlaneReason: "Covered by TSP's '@azure-tools/typespec-azure-resource-manager/no-record' rule.",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: "$.definitions..[?(@property !== 'tags' && @property !== 'delegatedResources' && @property !== 'userAssignedIdentities' && @ && @.additionalProperties)]",
            then: {
                function: spectral_functions_1.falsy,
            },
        },
        PropertiesTypeObjectNoDefinition: {
            rpcGuidelineCode: "RPC-Policy-V1-03",
            description: "Properties with type:object that don't reference a model definition are not allowed. ARM doesn't allow generic type definitions as this leads to bad customer experience.",
            severity: "error",
            stagingOnly: true,
            message: "{{error}}",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: [
                "$.definitions..[?(@property === 'type' && @ ==='object')]^",
                "$.definitions..[?(@property === 'type' && @ === '')]^",
                "$.definitions..[?(@property === 'undefined')]^",
            ],
            then: {
                function: properties_type_object_no_definition_1.propertiesTypeObjectNoDefinition,
            },
        },
        GetResponseCodes: {
            rpcGuidelineCode: "RPC-Get-V1-01",
            description: 'The GET operation should only return 200. In addition, it can return 202 only if it has "Location" header defined',
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*[get]"],
            then: {
                function: get_response_codes_1.getResponseCodes,
            },
        },
        GetMustNotHaveRequestBody: {
            rpcGuidelineCode: "RPC-Get-V1-02",
            description: "The Get operation must not have a request body.",
            severity: "error",
            message: "{{description}}",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*.get.parameters[?(@.in === 'body')]",
            then: {
                function: spectral_functions_1.falsy,
            },
        },
        GetCollectionOnlyHasValueAndNextLink: {
            rpcGuidelineCode: "RPC-Get-V1-09, RPC-Arg-V1-01, RPC-Get-V1-06",
            description: "Get endpoints for collections of resources must only have the `value` and `nextLink` properties in their model.",
            message: "{{description}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'][?(!@property.endsWith('}') && !@property.endsWith('operations') && !@property.endsWith('default'))][get].responses.200.schema.properties",
            then: {
                function: get_collection_only_has_value_and_next_link_1.getCollectionOnlyHasValueAndNextLink,
            },
        },
        ParametersInPointGet: {
            rpcGuidelineCode: "RPC-Get-V1-08",
            description: "Point Get's MUST not have query parameters other than api version.",
            severity: "error",
            message: "{{error}}",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths']",
            then: {
                function: parameters_in_point_get_1.parametersInPointGet,
            },
        },
        MissingSegmentsInNestedResourceListOperation: {
            rpcGuidelineCode: "RPC-Get-V1-11",
            description: "A nested resource type's List operation must include all the parent segments in its api path.",
            severity: "warn",
            message: "{{error}}",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*[get]^~",
            then: {
                function: missing_segments_in_nested_resource_list_operation_1.missingSegmentsInNestedResourceListOperation,
            },
        },
        XmsPageableForListCalls: {
            rpcGuidelineCode: "RPC-Get-V1-11",
            description: "`x-ms-pageable` extension must be specified for LIST APIs.",
            severity: "error",
            message: "{{error}}",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'][?(!@property.endsWith('}') && !@property.endsWith('/default'))].get",
            then: {
                function: xms_pageable_for_list_calls_1.default,
            },
        },
        GetOperationMustNotBeLongRunning: {
            rpcGuidelineCode: "RPC-Get-V1-14",
            description: "The GET operation cannot be long running. It must not have the `x-ms-long-running-operation` and `x-ms-long-running-operation-options` properties defined.",
            severity: "error",
            message: "{{description}}",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: [
                "$[paths,'x-ms-paths'].*[get].x-ms-long-running-operation-options",
                "$[paths,'x-ms-paths'].*[get][?(@property === 'x-ms-long-running-operation' && @ === true)]",
            ],
            then: {
                function: spectral_functions_1.falsy,
            },
        },
        QueryParametersInCollectionGet: {
            rpcGuidelineCode: "RPC-Get-V1-15",
            description: "Collection Get's/List operations MUST not have query parameters other than api-version & OData filter.",
            severity: "error",
            message: "{{error}}",
            stagingOnly: true,
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths']",
            then: {
                function: query_parameters_in_collection_get_1.queryParametersInCollectionGet,
            },
        },
        PatchPropertiesCorrespondToPutProperties: {
            rpcGuidelineCode: "RPC-Patch-V1-01",
            description: "PATCH request body must only contain properties present in the corresponding PUT request body, and must contain at least one property.",
            message: "{{error}}",
            severity: "error",
            stagingOnly: true,
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*"],
            then: {
                function: patch_properties_correspond_to_put_properties_1.patchPropertiesCorrespondToPutProperties,
            },
        },
        ConsistentPatchProperties: {
            rpcGuidelineCode: "RPC-Patch-V1-01",
            description: "The properties in the patch body must be present in the resource model and follow json merge patch.",
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$.paths.*.patch"],
            then: {
                function: consistent_patch_properties_1.consistentPatchProperties,
            },
        },
        UnSupportedPatchProperties: {
            rpcGuidelineCode: "RPC-Patch-V1-02",
            description: "Patch may not change the name, location, or type of the resource.",
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*.patch"],
            then: {
                function: validate_patch_body_param_properties_1.validatePatchBodyParamProperties,
                functionOptions: {
                    shouldNot: ["id", "name", "type", "location"],
                    propertiesThatMustNotBeInPropertiesBagAsWritable: ["provisioningState"],
                },
            },
        },
        PatchResponseCodes: {
            rpcGuidelineCode: "RPC-Patch-V1-06",
            description: "Synchronous PATCH must have 200 return code and LRO PATCH must have 200 and 202 return codes.",
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*[patch]"],
            then: {
                function: patch_response_codes_1.patchResponseCodes,
            },
        },
        LroPatch202: {
            rpcGuidelineCode: "RPC-Patch-V1-06, RPC-Async-V1-08",
            description: "Async PATCH should return 202.",
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*[patch][?(@property === 'x-ms-long-running-operation' && @ === true)]^"],
            then: {
                function: lro_patch_202_1.lroPatch202,
            },
        },
        PatchSkuProperty: {
            rpcGuidelineCode: "RPC-Patch-V1-09",
            description: "RP should consider implementing Patch for the 'SKU' envelope property if it's defined in the resource model and the service supports its updation.",
            message: "{{error}}",
            severity: "warn",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*.patch"],
            then: {
                function: validate_patch_body_param_properties_1.validatePatchBodyParamProperties,
                functionOptions: {
                    should: ["sku"],
                },
            },
        },
        PatchBodyParametersSchema: {
            rpcGuidelineCode: "RPC-Patch-V1-10",
            description: "A request parameter of the Patch Operation must not have a required/default/'x-ms-mutability: [\"create\"]' value.",
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$.paths.*.patch.parameters[?(@.in === 'body')]"],
            then: {
                function: patch_body_parameters_1.default,
            },
        },
        PatchIdentityProperty: {
            rpcGuidelineCode: "RPC-Patch-V1-11",
            description: "RP must implement PATCH for the 'identity' envelope property If it's defined in the resource model.",
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*.patch"],
            then: {
                function: validate_patch_body_param_properties_1.validatePatchBodyParamProperties,
                functionOptions: {
                    should: ["identity"],
                },
            },
        },
        PathForTrackedResourceTypes: {
            rpcGuidelineCode: "RPC-Put-V1-01, RPC-Get-V1-11",
            description: "The path must be under a subscription and resource group for tracked resource types.",
            message: "{{description}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*[get,put]^"],
            then: {
                function: path_for_tracked_resource_types_1.default,
            },
        },
        EvenSegmentedPathForPutOperation: {
            rpcGuidelineCode: "RPC-Put-V1-02",
            description: "API path with PUT operation defined MUST have even number of segments (i.e. end in {resourceType}/{resourceName} segments).",
            message: "{{description}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*.put^~",
            then: {
                function: spectral_functions_1.pattern,
                functionOptions: {
                    match: ".*/providers/\\w+.\\w+(/\\w+/(default|{\\w+}))+$",
                },
            },
        },
        RepeatedPathInfo: {
            rpcGuidelineCode: "RPC-Put-V1-05",
            description: "Information in the Path should not be repeated in the request body (i.e. subscription ID, resource group name, resource name).",
            message: "The '{{error}}' already appears in the path, please don't repeat it in the request body.",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*.put^",
            then: {
                function: body_param_repeated_info_1.default,
            },
        },
        RequestSchemaForTrackedResourcesMustHaveTags: {
            rpcGuidelineCode: "RPC-Put-V1-07",
            description: "A tracked resource MUST always have tags as a top level optional property",
            message: "{{description}}. {{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*.put^",
            then: {
                function: trackedresource_tags_property_in_request_1.default,
            },
        },
        PutGetPatchResponseSchema: {
            rpcGuidelineCode: "RPC-Put-V1-12",
            description: `For a given path with PUT, GET and PATCH operations, the schema of the response must be the same.`,
            message: "{{property}} has different responses for PUT/GET/PATCH operations. The PUT/GET/PATCH operations must have same schema response.",
            severity: "error",
            resolved: false,
            given: ["$[paths,'x-ms-paths'].*.put^"],
            then: {
                function: put_get_patch_schema_1.default,
            },
        },
        XmsResourceInPutResponse: {
            rpcGuidelineCode: "RPC-Put-V1-12",
            description: `The 200 response model for an ARM PUT operation must have x-ms-azure-resource extension set to true in its hierarchy.`,
            message: "{{error}}",
            severity: "error",
            resolved: true,
            given: ["$[paths,'x-ms-paths'].*.put"],
            then: {
                function: with_xms_resource_1.default,
            },
        },
        XMSSecretInResponse: {
            rpcGuidelineCode: "RPC-Put-V1-13",
            description: `When defining the response model for an ARM PUT/GET/POST operation, any property that contains sensitive information (such as passwords, keys, tokens, credentials, or other secrets) must include the "x-ms-secret": true annotation. This ensures that secrets are properly identified and handled according to ARM security guidelines.`,
            message: "{{error}}",
            severity: "error",
            resolved: true,
            given: ["$.definitions.*.properties"],
            then: {
                function: xms_secret_in_response_1.default,
            },
        },
        LocationMustHaveXmsMutability: {
            rpcGuidelineCode: "RPC-Put-V1-14",
            description: "A tracked resource's location property must have the x-ms-mutability properties set as read, create.",
            message: 'Property `location` must have `"x-ms-mutability":["read", "create"]` extension defined.',
            severity: "warn",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: ["$.definitions[*].properties.location"],
            then: {
                function: location_must_have_xms_mutability_1.default,
            },
        },
        ResponseSchemaSpecifiedForSuccessStatusCode: {
            rpcGuidelineCode: "RPC-Put-V1-24",
            description: "The 200 and 201 success status codes for an ARM PUT operation must have a response schema specified.",
            message: "{{error}}",
            severity: "error",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*.put"],
            then: {
                function: response_schema_specified_for_success_status_code_1.default,
            },
        },
        PutRequestResponseSchemeArm: {
            rpcGuidelineCode: "RPC-Put-V1-25",
            description: "The request & response('200') schema of the PUT operation must be same.",
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*[put][responses][?(@property === '200' || @property === '201')]^^"],
            then: {
                function: put_request_response_scheme_1.putRequestResponseScheme,
            },
        },
        RequestBodyMustExistForPutPatch: {
            rpcGuidelineCode: "RPC-Put-V1-28, RPC-Patch-V1-12",
            description: "Every Put and Patch operation must have a request body",
            message: "{{error}}",
            severity: "error",
            stagingOnly: true,
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*[put,patch].parameters",
            then: {
                function: request_body_must_exist_for_put_patch_1.requestBodyMustExistForPutPatch,
            },
        },
        ConsistentResponseSchemaForPut: {
            rpcGuidelineCode: "RPC-Put-V1-29",
            description: "A Put operation must return the same schema for 200 and 201 response codes",
            message: "{{error}}",
            severity: "error",
            stagingOnly: true,
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$.paths.*",
            then: {
                function: consistent_response_schema_for_put_1.default,
            },
        },
        TagsAreNotAllowedForProxyResources: {
            rpcGuidelineCode: "RPC-Put-V1-31",
            description: "Tags should not be specified in the properties bag for proxy resources. Consider using a Tracked resource instead.",
            severity: "error",
            stagingOnly: true,
            message: "{{error}}",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$.definitions.*.properties^"],
            then: {
                function: tags_are_not_allowed_for_proxy_resources_1.tagsAreNotAllowedForProxyResources,
            },
        },
        ParametersInPost: {
            rpcGuidelineCode: "RPC-POST-V1-05",
            description: "For a POST action parameters MUST be in the payload and not in the URI.",
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*[post][parameters]",
            then: {
                function: parameters_in_post_1.parametersInPost,
            },
        },
        ParametersSchemaAsTypeObject: {
            rpcGuidelineCode: "RPC-POST-V1-05",
            description: "The schema for body parameters must specify type:object and include a definition for its reference model.",
            message: "{{description}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*.*.parameters.*.schema[?(@property === 'type' && @ !=='object')]",
            then: {
                function: spectral_functions_1.falsy,
            },
        },
        PathContainsSubscriptionId: {
            rpcGuidelineCode: "RPC-Uri-V1-01",
            description: "Path for resource group scoped CRUD methods MUST contain a subscriptionId parameter.",
            message: "{{error}}",
            severity: "error",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*[get,patch,put,delete]^~",
            then: {
                function: arm_path_validation_1.default,
                functionOptions: {
                    segmentToCheck: "subscriptionIdParam",
                },
            },
        },
        PathContainsResourceGroup: {
            rpcGuidelineCode: "RPC-Uri-V1-02",
            description: "Path for resource group scoped CRUD methods MUST contain a resourceGroupName parameter.",
            message: "{{error}}",
            severity: "error",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*[get,patch,put,delete]^~"],
            then: {
                function: arm_path_validation_1.default,
                functionOptions: {
                    segmentToCheck: "resourceGroupParam",
                },
            },
        },
        PathContainsResourceType: {
            rpcGuidelineCode: "RPC-Uri-V1-04",
            description: "Path for resource CRUD methods MUST contain a resource type.",
            message: "{{error}}",
            severity: "error",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*[get,patch,put,delete]^~",
            then: {
                function: arm_path_validation_1.default,
                functionOptions: {
                    segmentToCheck: "resourceType",
                },
            },
        },
        ResourceNameRestriction: {
            rpcGuidelineCode: "RPC-Uri-V1-05",
            description: "This rule ensures that the authors explicitly define these restrictions as a regex on the resource name.",
            message: "{{error}}",
            severity: "error",
            disableForTypeSpecDataPlane: true,
            disableForTypeSpecDataPlaneReason: "Covered by TSP's '@azure-tools/typespec-azure-resource-manager/arm-resource-name-pattern' rule.",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*.^",
            then: {
                function: resource_name_restriction_1.default,
            },
        },
        PathForNestedResource: {
            rpcGuidelineCode: "RPC-Uri-V1-06",
            description: "Path for CRUD methods on a nested resource type MUST follow valid resource naming.",
            message: "{{error}}",
            severity: "error",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*[get,patch,delete,put]^~",
            then: {
                function: arm_path_validation_1.default,
                functionOptions: {
                    segmentToCheck: "nestedResourceType",
                },
            },
        },
        PathForResourceAction: {
            rpcGuidelineCode: "RPC-Uri-V1-07, RPC-POST-V1-01, RPC-POST-V1-07",
            description: "Path for 'post' method on a resource type MUST follow valid resource naming.",
            message: "{{description}}",
            severity: "error",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*.post^~",
            then: {
                function: spectral_functions_1.pattern,
                functionOptions: {
                    match: ".*/providers/[\\w\\.]+(?:/\\w+/(default|{\\w+}))*/\\w+$",
                },
            },
        },
        ApiVersionParameterRequired: {
            rpcGuidelineCode: "RPC-Uri-V1-08",
            description: "All operations should have api-version query parameter.",
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$.paths.*", "$.x-ms-paths.*"],
            then: {
                function: has_api_version_parameter_1.default,
                functionOptions: {
                    methods: ["get", "put", "patch", "post", "delete", "options", "head", "trace"],
                },
            },
        },
        NoDuplicatePathsForScopeParameter: {
            rpcGuidelineCode: "RPC-Uri-V1-10",
            description: 'Paths with explicitly defined scope should not be present if there is an equivalent path with the "scope" parameter.',
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$.paths[?(@property.match(/.*{scope}.*/))]~))", "$.x-ms-paths[?(@property.match(/.*{scope}.*/))]~))"],
            then: {
                function: no_duplicate_paths_for_scope_parameter_1.default,
            },
        },
        TenantLevelAPIsNotAllowed: {
            rpcGuidelineCode: "RPC-Uri-V1-11",
            description: "Tenant level APIs require a review from the security RBAC team during manifest check-in. For details, refer to the Manifest security review process: https://eng.ms/docs/microsoft-security/identity/auth-authz/access-control-managed-identityacmi/policy-administration-service/pas-wiki/livesite/security/manifest",
            message: "{{error}}",
            severity: "warn",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths']",
            then: {
                function: tenant_level_apis_not_allowed_1.tenantLevelAPIsNotAllowed,
            },
        },
        TrackedExtensionResourcesAreNotAllowed: {
            rpcGuidelineCode: "RPC-Uri-V1-12",
            description: "Extension resources are always considered to be proxy and must not be of the type tracked.",
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'].*~",
            then: {
                function: tracked_extension_resources_are_not_allowed_1.trackedExtensionResourcesAreNotAllowed,
            },
        },
        ValidQueryParametersForPointOperations: {
            rpcGuidelineCode: "RPC-Uri-V1-13",
            description: "Point operations (GET, PUT, PATCH, DELETE) must not include any query parameters other than api-version.",
            message: "{{error}}",
            stagingOnly: true,
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths']",
            then: {
                function: valid_query_parameters_for_point_operations_1.validQueryParametersForPointOperations,
            },
        },
        SystemDataDefinitionsCommonTypes: {
            rpcGuidelineCode: "RPC-SystemData-V1-01, RPC-SystemData-V1-02",
            description: "System data references must utilize common types.",
            message: "{{description}}",
            severity: "error",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: "$.definitions.*.properties.[systemData,SystemData].$ref",
            then: {
                function: spectral_functions_1.pattern,
                functionOptions: {
                    match: ".*/common-types/resource-management/v\\d+/types.json#/definitions/systemData",
                },
            },
        },
        SystemDataInPropertiesBag: {
            rpcGuidelineCode: "RPC-SystemData-V1-01, RPC-SystemData-V1-02",
            description: "System data must be defined as a top-level property, not in the properties bag.",
            message: "{{description}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$.definitions.*.properties[?(@property === 'properties')]^"],
            then: {
                function: system_data_in_properties_bag_1.systemDataInPropertiesBag,
            },
        },
        ReservedResourceNamesModelAsEnum: {
            rpcGuidelineCode: "RPC-ConstrainedCollections-V1-04",
            description: "Service-defined (reserved) resource names should be represented as an enum type with modelAsString set to true, not as a static string in the path.",
            message: "{{error}}",
            severity: "warn",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths']"],
            then: {
                function: reserved_resource_names_model_as_enum_1.reservedResourceNamesModelAsEnum,
            },
        },
        OperationsApiSchemaUsesCommonTypes: {
            rpcGuidelineCode: "RPC-Operations-V1-01",
            description: "Operations API path must follow the schema provided in the common types.",
            message: "{{description}}",
            severity: "error",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: "$[paths,'x-ms-paths'][?(@property.match(/\\/providers\\/\\w+\\.\\w+\\/operations$/i))].get.responses.200.schema.$ref",
            then: {
                function: spectral_functions_1.pattern,
                functionOptions: {
                    match: ".*/common-types/resource-management/v\\d+/types.json#/definitions/OperationListResult",
                },
            },
        },
        OperationsApiTenantLevelOnly: {
            rpcGuidelineCode: "RPC-Operations-V1-02",
            description: "The get operations endpoint must only be at the tenant level.",
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$.[paths,'x-ms-paths']",
            then: {
                function: operations_api_tenant_level_only_1.operationsApiTenantLevelOnly,
            },
        },
        LatestVersionOfCommonTypesMustBeUsed: {
            description: "This rule checks for references that aren't using latest version of common-types.",
            message: "{{error}}",
            severity: "warn",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: "$..['$ref']",
            then: {
                function: latest_version_of_common_types_must_be_used_1.latestVersionOfCommonTypesMustBeUsed,
            },
        },
        ArrayMustHaveType: {
            description: "Array type must have a type except for any type.",
            message: "{{error}}",
            severity: "error",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: ["$.definitions..items[?(@object())]^"],
            then: {
                function: spectral_functions_1.truthy,
                field: "type",
            },
        },
        NoErrorCodeResponses: {
            description: "Invalid status code specified. Please refer to the documentation for the allowed set.",
            message: "{{description}}",
            severity: "error",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: ["$.paths.*.*.responses.*~"],
            then: {
                function: no_error_code_responses_1.noErrorCodeResponses,
            },
        },
        LroWithOriginalUriAsFinalState: {
            description: "The long running operation with final-state-via:original-uri should have a sibling 'get' operation.",
            message: "{{description}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: [
                "$[paths,'x-ms-paths'].*[put,patch,delete].x-ms-long-running-operation-options[?(@property === 'final-state-via' && @ === 'original-uri')]^",
            ],
            then: {
                function: lro_original_uri_1.default,
            },
        },
        LroPostMustNotUseOriginalUriAsFinalState: {
            description: "The long running post operation must not use final-stat-via:original-uri.",
            message: "{{description}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: [
                "$[paths,'x-ms-paths'].*.post.x-ms-long-running-operation-options[?(@property === 'final-state-via' && @ === 'original-uri')]^",
            ],
            then: {
                function: spectral_functions_1.falsy,
            },
        },
        APIVersionPattern: {
            description: "The API Version parameter MUST be in the Year-Month-Date format (i.e. 2016-07-04.)  NOTE that this is the en-US ordering of month and date.",
            severity: "error",
            message: "{{description}}",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$.info.version",
            then: {
                function: spectral_functions_1.pattern,
                functionOptions: {
                    match: "^(20\\d{2})-(0[1-9]|1[0-2])-((0[1-9])|[12][0-9]|3[01])(-(preview))?$",
                },
            },
        },
        ParameterNotDefinedInGlobalParameters: {
            description: "Per ARM guidelines, if `subscriptionId` is used anywhere as a path parameter, it must always be defined as global parameter. `api-version` is almost always an input parameter in any ARM spec and must also be defined as a global parameter.",
            message: "{{error}}",
            severity: "warn",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*.*[?(@property === 'parameters')]"],
            then: {
                function: parameter_not_defined_in_global_parameters_1.parameterNotDefinedInGlobalParameters,
            },
        },
        ParameterNotUsingCommonTypes: {
            description: "This rule checks for parameters defined in common-types that are not using the common-types definition.",
            message: "{{error}}",
            severity: "warn",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*.*.parameters.*.name", "$[parameters].*.name"],
            then: {
                function: parameter_not_using_common_types_1.parameterNotUsingCommonTypes,
            },
        },
        CollectionObjectPropertiesNaming: {
            description: "Per ARM guidelines, a model returned by an `x-ms-pageable` operation must have a property named `value`. This property indicates what type of array the object is.",
            severity: "error",
            message: "{{error}}",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: "$.paths.*[get,post]",
            then: {
                function: collection_object_properties_naming_1.default,
            },
        },
        DefinitionsPropertiesNamesCamelCase: {
            description: "Property names should be camel case.",
            message: "Property name should be camel case.",
            severity: "error",
            resolved: false,
            given: "$.definitions..[?(@property === 'type' && @ === 'object')]^.properties[?(@property.match(/^[^@].+$/))]~",
            then: {
                function: camel_case_1.camelCase,
            },
        },
        GuidUsage: {
            description: `Verifies whether format is specified as "uuid" or not.`,
            message: "Usage of Guid is not recommended. If GUIDs are absolutely required in your service, please get sign off from the Azure API review board.",
            severity: "error",
            resolved: false,
            given: "$..[?(@property === 'format' && @ === 'uuid')]",
            then: {
                function: spectral_functions_1.falsy,
            },
        },
        InvalidSkuModel: {
            description: `A Sku model must have 'name' property. It can also have 'tier', 'size', 'family', 'capacity' as optional properties.`,
            message: "{{error}}",
            severity: "warn",
            resolved: true,
            given: "$.definitions[?(@property.match(/^sku$/i))]",
            then: {
                function: sku_validation_1.default,
            },
        },
        NonApplicationJsonType: {
            description: `Verifies whether operation supports "application/json" as consumes or produces section.`,
            message: "Only content-type 'application/json' is supported by ARM",
            severity: "warn",
            resolved: true,
            given: ["$[produces,consumes].*", "$[paths,'x-ms-paths'].*.*[produces,consumes].*"],
            then: {
                function: spectral_functions_1.pattern,
                functionOptions: {
                    match: "application/json",
                },
            },
        },
        SecurityDefinitionsStructure: {
            description: `Each OpenAPI json document must contain a security definitions section and the section must adhere to a certain format.`,
            message: "{{error}}",
            severity: "error",
            resolved: true,
            given: ["$"],
            then: {
                function: security_definitions_structure_1.securityDefinitionsStructure,
            },
        },
        SubscriptionIdParameterInOperations: {
            description: `'subscriptionId' must not be an operation parameter and must be declared in the global parameters section.`,
            message: "Parameter 'subscriptionId' is not allowed in the operations section, define it in the global parameters section instead/Parameter '{{path}}' is referenced but not defined in the global parameters section of Service Definition",
            severity: "error",
            resolved: false,
            given: [
                "$[paths,'x-ms-paths'].*.*.parameters.*[?(@property === 'name' && @.match(/^subscriptionid$/i))]^",
                "$[paths,'x-ms-paths'].*.parameters.*[?(@property === 'name' && @.match(/^subscriptionid$/i))]^",
            ],
            then: {
                function: spectral_functions_1.falsy,
            },
        },
        OperationsApiResponseSchema: {
            severity: "error",
            message: "The response schema of operations API '{{error}}' does not match the ARM specification. Please standardize the schema.",
            resolved: true,
            given: "$.paths[?(@property.match(/\\/providers\\/\\w+\\.\\w+\\/operations$/i))].get.responses.200.schema",
            then: {
                function: operations_api_schema_1.default,
            },
        },
        HttpsSupportedScheme: {
            description: "Verifies whether specification supports HTTPS scheme or not.",
            message: "Azure Resource Management only supports HTTPS scheme.",
            severity: "warn",
            resolved: false,
            formats: [spectral_formats_1.oas2],
            given: ["$.schemes"],
            then: {
                function: https_supported_scheme_1.default,
            },
        },
        MissingDefaultResponse: {
            description: "All operations should have a default (error) response.",
            message: "Operation is missing a default response.",
            severity: "error",
            given: "$.paths.*.*.responses.*~",
            then: {
                field: "default",
                function: spectral_functions_1.truthy,
            },
        },
    },
};
exports.default = ruleset;
