// @microsoft.azure/openapi-validator-rulesets/dist/spectral/az-dataplane.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const spectral_formats_1 = require("@stoplight/spectral-formats");
const spectral_functions_1 = require("@stoplight/spectral-functions");
const az_common_1 = tslib_1.__importDefault(require("./az-common"));
const consistent_response_body_1 = tslib_1.__importDefault(require("./functions/consistent-response-body"));
const delete_204_response_1 = tslib_1.__importDefault(require("./functions/delete-204-response"));
const error_response_1 = tslib_1.__importDefault(require("./functions/error-response"));
const long_running_response_status_code_1 = require("./functions/Extensions/long-running-response-status-code");
const has_header_1 = tslib_1.__importDefault(require("./functions/has-header"));
const host_parameters_1 = tslib_1.__importDefault(require("./functions/host-parameters"));
const operation_id_1 = tslib_1.__importDefault(require("./functions/operation-id"));
const pagination_response_1 = tslib_1.__importDefault(require("./functions/pagination-response"));
const param_names_1 = tslib_1.__importDefault(require("./functions/param-names"));
const param_names_unique_1 = tslib_1.__importDefault(require("./functions/param-names-unique"));
const param_order_1 = tslib_1.__importDefault(require("./functions/param-order"));
const patch_content_type_1 = tslib_1.__importDefault(require("./functions/patch-content-type"));
const path_param_names_1 = tslib_1.__importDefault(require("./functions/path-param-names"));
const path_param_schema_1 = tslib_1.__importDefault(require("./functions/path-param-schema"));
const put_request_response_scheme_1 = require("./functions/put-request-response-scheme");
const schema_type_and_format_1 = tslib_1.__importDefault(require("./functions/schema-type-and-format"));
const version_policy_1 = tslib_1.__importDefault(require("./functions/version-policy"));
const ruleset = {
    extends: [az_common_1.default],
    rules: {
        AdditionalPropertiesAndProperties: {
            description: "Don't specify additionalProperties as a sibling of properties.",
            disableForTypeSpecDataPlane: true,
            disableForTypeSpecDataPlaneReason: "Covered by TSP's '@azure-tools/typespec-azure-core/bad-record-type' rule.",
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            given: "$..[?(@object() && @.type === 'object' && @.properties)]",
            then: {
                field: "additionalProperties",
                function: spectral_functions_1.falsy,
            },
        },
        AdditionalPropertiesObject: {
            description: "additionalProperties with type object is a common error.",
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            resolved: false,
            given: "$..[?(@property == 'additionalProperties' && @.type == 'object' && @.properties == undefined)]",
            then: {
                function: spectral_functions_1.falsy,
            },
        },
        ApiVersionEnum: {
            description: "The api-version parameter should not be an enum.",
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            given: [
                "$.paths[*].parameters[?(@.name == 'ApiVersion')]",
                "$.paths.*[get,put,post,patch,delete,options,head].parameters[?(@.name == 'api-version')]",
            ],
            then: {
                field: "enum",
                function: spectral_functions_1.falsy,
            },
        },
        ConsistentResponseBody: {
            description: "Ensure the get, put, and patch response body schemas are consistent.",
            message: "{{error}}",
            severity: "warn",
            formats: [spectral_formats_1.oas2],
            given: "$.paths.*",
            then: {
                function: consistent_response_body_1.default,
            },
        },
        DefaultResponse: {
            description: "All operations should have a default (error) response.",
            message: "Operation is missing a default response.",
            severity: "warn",
            given: "$.paths.*.*.responses",
            then: {
                field: "default",
                function: spectral_functions_1.truthy,
            },
        },
        Delete204Response: {
            description: "A delete operation should have a 204 response.",
            message: "A delete operation should have a `204` response.",
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            given: "$.paths[*].delete.responses",
            then: {
                function: delete_204_response_1.default,
            },
        },
        ErrorResponse: {
            description: "Error response body should conform to Microsoft Azure API Guidelines.",
            message: "{{error}}",
            severity: "warn",
            formats: [spectral_formats_1.oas2],
            given: "$.paths[*][*].responses",
            then: {
                function: error_response_1.default,
            },
        },
        Formdata: {
            description: "Check for appropriate use of formData parameters.",
            severity: "warn",
            formats: [spectral_formats_1.oas2],
            given: '$.paths.*[get,put,post,patch,delete,options,head].parameters[?(@.in == "formData")]',
            then: {
                function: spectral_functions_1.falsy,
            },
        },
        HeaderDisallowed: {
            description: "Authorization, Content-type, and Accept headers should not be defined explicitly.",
            message: 'Header parameter "{{value}}" should not be defined explicitly.',
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            given: [
                "$.paths[*].parameters[?(@.in == 'header')]",
                "$.paths.*[get,put,post,patch,delete,options,head].parameters[?(@.in == 'header')]",
            ],
            then: {
                function: spectral_functions_1.pattern,
                field: "name",
                functionOptions: {
                    notMatch: "/^(authorization|content-type|accept)$/i",
                },
            },
        },
        LroHeaders: {
            description: "A 202 response should include an Operation-Location response header.",
            message: "A 202 response should include an Operation-Location response header.",
            severity: "warn",
            formats: [spectral_formats_1.oas2],
            given: "$.paths[*][*].responses[?(@property == '202')]",
            then: {
                function: has_header_1.default,
                functionOptions: {
                    name: "Operation-location",
                },
            },
        },
        MsPaths: {
            description: "Don't use x-ms-paths except where necessary to support legacy APIs.",
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            given: "$.x-ms-paths",
            then: {
                function: spectral_functions_1.falsy,
            },
        },
        Nullable: {
            description: "Avoid the use of x-nullable.",
            severity: "warn",
            disableForTypeSpecDataPlane: true,
            disableForTypeSpecDataPlaneReason: "Covered by Azure Core '@azure-tools/typespec-azure-core/no-nullable' rule",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            resolved: false,
            given: "$..x-nullable",
            then: {
                function: spectral_functions_1.undefined,
            },
        },
        OperationId: {
            description: "OperationId should conform to Azure API Guidelines",
            message: "{{error}}",
            severity: "warn",
            given: ["$.paths.*[get,put,post,patch,delete,options,head]"],
            then: {
                function: operation_id_1.default,
            },
        },
        PaginationResponse: {
            description: "An operation that returns a list that is potentially large should support pagination.",
            message: "{{error}}",
            severity: "warn",
            formats: [spectral_formats_1.oas2],
            given: ["$.paths.*[get,post]"],
            then: {
                function: pagination_response_1.default,
            },
        },
        ParameterDefaultNotAllowed: {
            description: "A required parameter should not specify a default value.",
            severity: "warn",
            given: ["$.paths[*].parameters[?(@.required)]", "$.paths.*[get,put,post,patch,delete,options,head].parameters[?(@.required)]"],
            then: {
                field: "default",
                function: spectral_functions_1.falsy,
            },
        },
        ParameterNamesConvention: {
            description: "Parameter names should conform to Azure naming conventions.",
            message: "{{error}}",
            severity: "warn",
            given: ["$.paths[*].parameters.*", "$.paths.*[get,put,post,patch,delete,options,head].parameters.*"],
            then: {
                function: param_names_1.default,
            },
        },
        ParameterNamesUnique: {
            description: "All parameter names for an operation should be case-insensitive unique.",
            message: "{{error}}",
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            given: "$.paths[*]",
            then: {
                function: param_names_unique_1.default,
            },
        },
        ParameterOrder: {
            description: "Path parameters must be in the same order as in the path.",
            message: "{{error}}",
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            given: "$.paths",
            then: {
                function: param_order_1.default,
            },
        },
        PathParameterNames: {
            description: "Path parameter names should be consistent across all paths.",
            message: "{{error}}",
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            given: "$.paths",
            resolved: false,
            then: {
                function: path_param_names_1.default,
            },
        },
        PatchContentType: {
            description: "The request body content type for patch operations should be JSON merge patch.",
            message: "{{error}}",
            severity: "warn",
            formats: [spectral_formats_1.oas2],
            given: "$",
            then: {
                function: patch_content_type_1.default,
            },
        },
        PathCharacters: {
            description: "Path should contain only recommended characters.",
            message: "Path contains non-recommended characters.",
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            given: "$.paths.*~",
            then: {
                function: spectral_functions_1.pattern,
                functionOptions: {
                    match: "^(/([0-9A-Za-z._~-]+|{[^}]+}))*(/([0-9A-Za-z._~:-]+|{[^}]*}(:[0-9A-Za-z._~-]+)?))$",
                },
            },
        },
        PathParameterSchema: {
            description: "Path parameter should be type: string and specify maxLength and pattern.",
            message: "{{error}}",
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            given: [
                "$.paths[*].parameters[?(@.in == 'path')]",
                "$.paths.*[get,put,post,patch,delete,options,head].parameters[?(@.in == 'path')]",
            ],
            then: {
                function: path_param_schema_1.default,
            },
        },
        Post201Response: {
            description: "Using post for a create operation is discouraged.",
            message: "Using post for a create operation is discouraged.",
            severity: "warn",
            formats: [spectral_formats_1.oas2],
            given: "$.paths[*].post.responses",
            then: {
                field: "201",
                function: spectral_functions_1.falsy,
            },
        },
        PropertyDescription: {
            description: "All schema properties should have a description.",
            message: "Property should have a description.",
            severity: "warn",
            disableForTypeSpecDataPlane: true,
            disableForTypeSpecDataPlaneReason: "Covered by TSP's '@azure-tools/typespec-azure-core/documentation-required' rule.",
            resolved: false,
            given: "$..properties[?(@object() && @.$ref == undefined)]",
            then: {
                field: "description",
                function: spectral_functions_1.truthy,
            },
        },
        PropertyType: {
            description: "All schema properties should have a defined type.",
            message: "Property should have a defined type.",
            severity: "warn",
            resolved: false,
            given: "$..properties[?(@object() && @.$ref == undefined)]",
            then: {
                field: "type",
                function: spectral_functions_1.truthy,
            },
        },
        PutPath: {
            description: "The path for a put should have a final path parameter.",
            message: "The path for a put should have a final path parameter.",
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            given: "$.paths[*].put^~",
            then: {
                function: spectral_functions_1.pattern,
                functionOptions: {
                    match: "/\\}$/",
                },
            },
        },
        RequestBodyNotAllowed: {
            description: "A get or delete operation must not accept a body parameter.",
            severity: "warn",
            formats: [spectral_formats_1.oas2],
            given: ["$.paths[*][get,delete].parameters[*]"],
            then: {
                field: "in",
                function: spectral_functions_1.pattern,
                functionOptions: {
                    notMatch: "/^body$/",
                },
            },
        },
        RequestBodyOptional: {
            description: "Flag optional request body -- common oversight.",
            message: "The body parameter is not marked as required.",
            severity: "warn",
            formats: [spectral_formats_1.oas2],
            given: ["$.paths[*][put,post,patch].parameters[?(@.in == 'body')]"],
            then: {
                field: "required",
                function: spectral_functions_1.truthy,
            },
        },
        SchemaNamesConvention: {
            description: "Schema names should be Pascal case.",
            message: "Schema name should be Pascal case.",
            severity: "warn",
            disableForTypeSpecDataPlane: true,
            disableForTypeSpecDataPlaneReason: "Covered by TSP's '@azure-tools/typespec-azure-core/casing-style' rule.",
            formats: [spectral_formats_1.oas2],
            given: "$.definitions.*~",
            then: {
                function: spectral_functions_1.casing,
                functionOptions: {
                    type: "pascal",
                },
            },
        },
        SchemaTypeAndFormat: {
            description: "Schema should use well-defined type and format.",
            message: "{{error}}",
            severity: "warn",
            formats: [spectral_formats_1.oas2],
            given: [
                "$.paths[*][put,post,patch].parameters[?(@.in == 'body')].schema",
                "$.paths[*][get,put,post,patch,delete].responses[*].schema",
            ],
            then: {
                function: schema_type_and_format_1.default,
            },
        },
        SecurityDefinitionDescription: {
            description: "A security definition should have a description.",
            message: "Security definition should have a description.",
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            given: ["$.securityDefinitions[*]", "$.components.securitySchemes[*]"],
            then: {
                field: "description",
                function: spectral_functions_1.truthy,
            },
        },
        SuccessResponseBody: {
            description: "All success responses except 202 & 204 should define a response body.",
            severity: "warn",
            formats: [spectral_formats_1.oas2],
            given: "$.paths[*][*].responses[?(@property >= 200 && @property < 300 && @property != '202' && @property != '204')]",
            then: {
                field: "schema",
                function: spectral_functions_1.truthy,
            },
        },
        VersionConvention: {
            description: "API version should be a date in YYYY-MM-DD format, optionally suffixed with '-preview'.",
            severity: "warn",
            formats: [spectral_formats_1.oas2, spectral_formats_1.oas3],
            given: "$.info.version",
            then: {
                function: spectral_functions_1.pattern,
                functionOptions: {
                    match: "^\\d\\d\\d\\d-\\d\\d-\\d\\d(-preview)?$",
                },
            },
        },
        VersionPolicy: {
            description: "Specify API version using `api-version` query parameter, not in path.",
            message: "{{error}}",
            severity: "warn",
            formats: [spectral_formats_1.oas2],
            given: "$",
            then: {
                function: version_policy_1.default,
            },
        },
        HostParametersValidation: {
            description: "Validate the parameters in x-ms-parameterized-host.",
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$.x-ms-parameterized-host"],
            then: {
                function: host_parameters_1.default,
            },
        },
        LongRunningResponseStatusCodeDataPlane: {
            description: 'A LRO Post operation with return schema must have "x-ms-long-running-operation-options" extension enabled.',
            message: "{{error}}",
            severity: "error",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*.*[?(@property === 'x-ms-long-running-operation' && @ === true)]^^"],
            then: {
                function: long_running_response_status_code_1.longRunningResponseStatusCodeDataPlane,
            },
        },
        PutRequestResponseScheme: {
            description: "The request & response('200') schema of the PUT operation must be same.",
            message: "{{error}}",
            severity: "warn",
            resolved: true,
            formats: [spectral_formats_1.oas2],
            given: ["$[paths,'x-ms-paths'].*[put][responses][?(@property === '200' || @property === '201')]^^"],
            then: {
                function: put_request_response_scheme_1.putRequestResponseScheme,
            },
        },
    },
};
exports.default = ruleset;
