// temporal-polyfill/fns/instant.cjs
"use strict";

require("../chunks/internal.cjs");

var funcApi = require("../chunks/funcApi.cjs");

exports.add = funcApi.add, exports.compare = funcApi.compare, exports.create = funcApi.create, 
exports.epochMicroseconds = funcApi.epochMicroseconds, exports.epochMilliseconds = funcApi.epochMilliseconds, 
exports.epochNanoseconds = funcApi.epochNanoseconds, exports.epochSeconds = funcApi.epochSeconds, 
exports.equals = funcApi.equals, exports.fromEpochMicroseconds = funcApi.fromEpochMicroseconds, 
exports.fromEpochMilliseconds = funcApi.fromEpochMilliseconds, exports.fromEpochNanoseconds = funcApi.fromEpochNanoseconds, 
exports.fromEpochSeconds = funcApi.fromEpochSeconds, exports.fromString = funcApi.fromString, 
exports.isInstance = funcApi.isInstance, exports.rangeToLocaleString = funcApi.rangeToLocaleString, 
exports.rangeToLocaleStringParts = funcApi.rangeToLocaleStringParts, exports.round = funcApi.round, 
exports.since = funcApi.since, exports.subtract = funcApi.subtract, exports.toLocaleString = funcApi.toLocaleString, 
exports.toLocaleStringParts = funcApi.toLocaleStringParts, exports.toString = funcApi.toString, 
exports.toZonedDateTime = funcApi.toZonedDateTime, exports.toZonedDateTimeISO = funcApi.toZonedDateTimeISO, 
exports.until = funcApi.until;
