// temporal-polyfill/fns/plaindatetime.cjs
"use strict";

require("../chunks/internal.cjs");

var funcApi = require("../chunks/funcApi.cjs");

exports.add = funcApi.add$2, exports.addDays = funcApi.addDays$1, exports.addHours = funcApi.addHours$1, 
exports.addMicroseconds = funcApi.addMicroseconds$1, exports.addMilliseconds = funcApi.addMilliseconds$1, 
exports.addMinutes = funcApi.addMinutes$1, exports.addMonths = funcApi.addMonths$1, 
exports.addNanoseconds = funcApi.addNanoseconds$1, exports.addSeconds = funcApi.addSeconds$1, 
exports.addWeeks = funcApi.addWeeks$1, exports.addYears = funcApi.addYears$1, exports.calendarId = funcApi.calendarId$1, 
exports.compare = funcApi.compare$2, exports.create = funcApi.create$2, exports.dayOfWeek = funcApi.dayOfWeek$1, 
exports.dayOfYear = funcApi.dayOfYear$1, exports.daysInMonth = funcApi.daysInMonth$1, 
exports.daysInWeek = funcApi.daysInWeek$1, exports.daysInYear = funcApi.daysInYear$1, 
exports.diffDays = funcApi.diffDays$1, exports.diffHours = funcApi.diffHours$1, 
exports.diffMicroseconds = funcApi.diffMicroseconds$1, exports.diffMilliseconds = funcApi.diffMilliseconds$1, 
exports.diffMinutes = funcApi.diffMinutes$1, exports.diffMonths = funcApi.diffMonths$1, 
exports.diffNanoseconds = funcApi.diffNanoseconds$1, exports.diffSeconds = funcApi.diffSeconds$1, 
exports.diffWeeks = funcApi.diffWeeks$1, exports.diffYears = funcApi.diffYears$1, 
exports.endOfDay = funcApi.endOfDay$1, exports.endOfHour = funcApi.endOfHour$1, 
exports.endOfMicrosecond = funcApi.endOfMicrosecond$1, exports.endOfMillisecond = funcApi.endOfMillisecond$1, 
exports.endOfMinute = funcApi.endOfMinute$1, exports.endOfMonth = funcApi.endOfMonth$1, 
exports.endOfSecond = funcApi.endOfSecond$1, exports.endOfWeek = funcApi.endOfWeek$1, 
exports.endOfYear = funcApi.endOfYear$1, exports.equals = funcApi.equals$2, exports.fromFields = funcApi.fromFields$1, 
exports.fromString = funcApi.fromString$2, exports.getFields = funcApi.getFields$1, 
exports.getISOFields = funcApi.getISOFields$1, exports.inLeapYear = funcApi.inLeapYear$1, 
exports.isInstance = funcApi.isInstance$2, exports.monthsInYear = funcApi.monthsInYear$1, 
exports.rangeToLocaleString = funcApi.rangeToLocaleString$2, exports.rangeToLocaleStringParts = funcApi.rangeToLocaleStringParts$2, 
exports.round = funcApi.round$2, exports.roundToMonth = funcApi.roundToMonth$1, 
exports.roundToWeek = funcApi.roundToWeek$1, exports.roundToYear = funcApi.roundToYear$1, 
exports.since = funcApi.since$2, exports.startOfDay = funcApi.startOfDay$1, exports.startOfHour = funcApi.startOfHour$1, 
exports.startOfMicrosecond = funcApi.startOfMicrosecond$1, exports.startOfMillisecond = funcApi.startOfMillisecond$1, 
exports.startOfMinute = funcApi.startOfMinute$1, exports.startOfMonth = funcApi.startOfMonth$1, 
exports.startOfSecond = funcApi.startOfSecond$1, exports.startOfWeek = funcApi.startOfWeek$1, 
exports.startOfYear = funcApi.startOfYear$1, exports.subtract = funcApi.subtract$2, 
exports.subtractDays = funcApi.subtractDays$1, exports.subtractHours = funcApi.subtractHours$1, 
exports.subtractMicroseconds = funcApi.subtractMicroseconds$1, exports.subtractMilliseconds = funcApi.subtractMilliseconds$1, 
exports.subtractMinutes = funcApi.subtractMinutes$1, exports.subtractMonths = funcApi.subtractMonths$1, 
exports.subtractNanoseconds = funcApi.subtractNanoseconds$1, exports.subtractSeconds = funcApi.subtractSeconds$1, 
exports.subtractWeeks = funcApi.subtractWeeks$1, exports.subtractYears = funcApi.subtractYears$1, 
exports.toLocaleString = funcApi.toLocaleString$2, exports.toLocaleStringParts = funcApi.toLocaleStringParts$2, 
exports.toPlainDate = funcApi.toPlainDate$1, exports.toPlainMonthDay = funcApi.toPlainMonthDay$1, 
exports.toPlainTime = funcApi.toPlainTime$1, exports.toPlainYearMonth = funcApi.toPlainYearMonth$1, 
exports.toString = funcApi.toString$2, exports.toZonedDateTime = funcApi.toZonedDateTime$1, 
exports.until = funcApi.until$2, exports.weekOfYear = funcApi.weekOfYear$1, exports.withCalendar = funcApi.withCalendar$1, 
exports.withDayOfMonth = funcApi.withDayOfMonth$1, exports.withDayOfWeek = funcApi.withDayOfWeek$1, 
exports.withDayOfYear = funcApi.withDayOfYear$1, exports.withFields = funcApi.withFields$1, 
exports.withPlainDate = funcApi.withPlainDate$1, exports.withPlainTime = funcApi.withPlainTime$1, 
exports.withWeekOfYear = funcApi.withWeekOfYear$1, exports.yearOfWeek = funcApi.yearOfWeek$1;
