// temporal-polyfill/fns/plainyearmonth.cjs
"use strict";

require("../chunks/internal.cjs");

var funcApi = require("../chunks/funcApi.cjs");

exports.add = funcApi.add$5, exports.calendarId = funcApi.calendarId$3, exports.compare = funcApi.compare$5, 
exports.create = funcApi.create$5, exports.daysInMonth = funcApi.daysInMonth$3, 
exports.daysInYear = funcApi.daysInYear$3, exports.equals = funcApi.equals$5, exports.fromFields = funcApi.fromFields$4, 
exports.fromString = funcApi.fromString$5, exports.getFields = funcApi.getFields$4, 
exports.getISOFields = funcApi.getISOFields$4, exports.inLeapYear = funcApi.inLeapYear$3, 
exports.isInstance = funcApi.isInstance$5, exports.monthsInYear = funcApi.monthsInYear$3, 
exports.rangeToLocaleString = funcApi.rangeToLocaleString$5, exports.rangeToLocaleStringParts = funcApi.rangeToLocaleStringParts$5, 
exports.since = funcApi.since$5, exports.subtract = funcApi.subtract$5, exports.toLocaleString = funcApi.toLocaleString$5, 
exports.toLocaleStringParts = funcApi.toLocaleStringParts$5, exports.toPlainDate = funcApi.toPlainDate$2, 
exports.toString = funcApi.toString$5, exports.until = funcApi.until$5, exports.withFields = funcApi.withFields$4;
