// temporal-polyfill/fns/plaintime.cjs
"use strict";

require("../chunks/internal.cjs");

var funcApi = require("../chunks/funcApi.cjs");

exports.add = funcApi.add$4, exports.compare = funcApi.compare$4, exports.create = funcApi.create$4, 
exports.equals = funcApi.equals$4, exports.fromFields = funcApi.fromFields$3, exports.fromString = funcApi.fromString$4, 
exports.getFields = funcApi.getFields$3, exports.getISOFields = funcApi.getISOFields$3, 
exports.isInstance = funcApi.isInstance$4, exports.rangeToLocaleString = funcApi.rangeToLocaleString$4, 
exports.rangeToLocaleStringParts = funcApi.rangeToLocaleStringParts$4, exports.round = funcApi.round$3, 
exports.since = funcApi.since$4, exports.subtract = funcApi.subtract$4, exports.toLocaleString = funcApi.toLocaleString$4, 
exports.toLocaleStringParts = funcApi.toLocaleStringParts$4, exports.toPlainDateTime = funcApi.toPlainDateTime$2, 
exports.toString = funcApi.toString$4, exports.toZonedDateTime = funcApi.toZonedDateTime$3, 
exports.until = funcApi.until$4, exports.withFields = funcApi.withFields$3;
