// temporal-polyfill/fns/plaindate.cjs
"use strict";

require("../chunks/internal.cjs");

var funcApi = require("../chunks/funcApi.cjs");

exports.add = funcApi.add$3, exports.addDays = funcApi.addDays$2, exports.addMonths = funcApi.addMonths$2, 
exports.addWeeks = funcApi.addWeeks$2, exports.addYears = funcApi.addYears$2, exports.calendarId = funcApi.calendarId$2, 
exports.compare = funcApi.compare$3, exports.create = funcApi.create$3, exports.dayOfWeek = funcApi.dayOfWeek$2, 
exports.dayOfYear = funcApi.dayOfYear$2, exports.daysInMonth = funcApi.daysInMonth$2, 
exports.daysInWeek = funcApi.daysInWeek$2, exports.daysInYear = funcApi.daysInYear$2, 
exports.diffDays = funcApi.diffDays$2, exports.diffMonths = funcApi.diffMonths$2, 
exports.diffWeeks = funcApi.diffWeeks$2, exports.diffYears = funcApi.diffYears$2, 
exports.endOfMonth = funcApi.endOfMonth$2, exports.endOfWeek = funcApi.endOfWeek$2, 
exports.endOfYear = funcApi.endOfYear$2, exports.equals = funcApi.equals$3, exports.fromFields = funcApi.fromFields$2, 
exports.fromString = funcApi.fromString$3, exports.getFields = funcApi.getFields$2, 
exports.getISOFields = funcApi.getISOFields$2, exports.inLeapYear = funcApi.inLeapYear$2, 
exports.isInstance = funcApi.isInstance$3, exports.monthsInYear = funcApi.monthsInYear$2, 
exports.rangeToLocaleString = funcApi.rangeToLocaleString$3, exports.rangeToLocaleStringParts = funcApi.rangeToLocaleStringParts$3, 
exports.roundToMonth = funcApi.roundToMonth$2, exports.roundToWeek = funcApi.roundToWeek$2, 
exports.roundToYear = funcApi.roundToYear$2, exports.since = funcApi.since$3, exports.startOfMonth = funcApi.startOfMonth$2, 
exports.startOfWeek = funcApi.startOfWeek$2, exports.startOfYear = funcApi.startOfYear$2, 
exports.subtract = funcApi.subtract$3, exports.subtractDays = funcApi.subtractDays$2, 
exports.subtractMonths = funcApi.subtractMonths$2, exports.subtractWeeks = funcApi.subtractWeeks$2, 
exports.subtractYears = funcApi.subtractYears$2, exports.toLocaleString = funcApi.toLocaleString$3, 
exports.toLocaleStringParts = funcApi.toLocaleStringParts$3, exports.toPlainDateTime = funcApi.toPlainDateTime$1, 
exports.toPlainMonthDay = funcApi.toPlainMonthDay$2, exports.toPlainYearMonth = funcApi.toPlainYearMonth$2, 
exports.toString = funcApi.toString$3, exports.toZonedDateTime = funcApi.toZonedDateTime$2, 
exports.until = funcApi.until$3, exports.weekOfYear = funcApi.weekOfYear$2, exports.withCalendar = funcApi.withCalendar$2, 
exports.withDayOfMonth = funcApi.withDayOfMonth$2, exports.withDayOfWeek = funcApi.withDayOfWeek$2, 
exports.withDayOfYear = funcApi.withDayOfYear$2, exports.withFields = funcApi.withFields$2, 
exports.withWeekOfYear = funcApi.withWeekOfYear$2, exports.yearOfWeek = funcApi.yearOfWeek$2;
