// @azure-tools/typespec-azure-resource-manager/dist/src/private.decorators.js
import { $key, addVisibilityModifiers, clearVisibilityModifiersForClass, getKeyName, getLifecycleVisibilityEnum, getNamespaceFullName, getTypeName, isKey, isTemplateDeclarationOrInstance, isTemplateInstance, sealVisibilityModifiers, } from "@typespec/compiler";
import { $ } from "@typespec/compiler/typekit";
import { useStateMap } from "@typespec/compiler/utils";
import { $bodyRoot, getHttpOperation } from "@typespec/http";
import { $actionSegment, $createsOrReplacesResource, $deletesResource, $listsResource, $readsResource, $segment, $updatesResource, getActionSegment, getSegment, } from "@typespec/rest";
import { camelCase } from "change-case";
import pluralize from "pluralize";
import { getArmCommonTypesVersion } from "./common-types.js";
import { reportDiagnostic } from "./lib.js";
import { getArmProviderNamespace, isArmLibraryNamespace } from "./namespace.js";
import { $armResourceAction, $armResourceCheckExistence, $armResourceCreateOrUpdate, $armResourceDelete, $armResourceList, $armResourceRead, $armResourceUpdate, addArmResourceOperation, armRenameListByOperationInternal, getArmResourceOperations, setArmOperationIdentifier, } from "./operations.js";
import { getArmResourceKind, getArmVirtualResourceDetails, getResourceBaseType, isArmVirtualResource, isCustomAzureResource, resolveResourceBaseType, ResourceBaseType, setResourceBaseType, } from "./resource.js";
import { ArmStateKeys } from "./state.js";
export const namespace = "Azure.ResourceManager.Private";
/** @internal */
const $builtInResource = (context, resourceType) => {
    const { program } = context;
    setResourceBaseType(program, resourceType, ResourceBaseType.BuiltIn);
};
const $builtInSubscriptionResource = (context, resourceType) => {
    const { program } = context;
    setResourceBaseType(program, resourceType, ResourceBaseType.BuiltInSubscription);
};
const $builtInResourceGroupResource = (context, resourceType) => {
    const { program } = context;
    setResourceBaseType(program, resourceType, ResourceBaseType.BuiltInResourceGroup);
};
const $omitIfEmpty = (context, entity, propertyName) => {
    const modelProp = getProperty(entity, propertyName);
    if (modelProp &&
        modelProp.type.kind === "Model" &&
        !hasProperty(context.program, modelProp.type)) {
        entity.properties.delete(propertyName);
    }
};
function checkAllowedVirtualResource(program, target, type) {
    if (target.kind === "Model")
        return false;
    if (!isArmVirtualResource(program, type))
        return false;
    const [httpOp, _] = getHttpOperation(program, target);
    if (httpOp === undefined)
        return false;
    const method = httpOp.verb;
    switch (method) {
        case "post":
        case "delete":
            return true;
        default:
            return true;
    }
}
function isBuiltIn(baseType) {
    return (baseType === ResourceBaseType.BuiltIn ||
        baseType === ResourceBaseType.BuiltInSubscription ||
        baseType === ResourceBaseType.BuiltInResourceGroup);
}
const $enforceConstraint = (context, entity, sourceType, constraintType) => {
    if (sourceType !== undefined && constraintType !== undefined) {
        // walk the baseModel chain until find a match or fail
        let baseType = sourceType;
        do {
            if (baseType === constraintType ||
                isCustomAzureResource(context.program, baseType) ||
                isBuiltIn(getResourceBaseType(context.program, baseType)) ||
                checkAllowedVirtualResource(context.program, entity, baseType))
                return;
        } while ((baseType = baseType.baseModel) !== undefined);
        reportDiagnostic(context.program, {
            code: "template-type-constraint-no-met",
            target: entity,
            format: {
                entity: entity.name,
                sourceType: sourceType.name,
                constraintType: constraintType.name,
                actionMessage: `Please use the "TrackedResource", "ProxyResource", or "ExtensionResource" template to define the resource.`,
            },
        });
    }
};
const $resourceBaseParametersOf = (context, entity, resourceType) => {
    const targetResourceBaseType = getResourceBaseType(context.program, resourceType);
    const removedProperties = [];
    for (const [propertyName, property] of entity.properties) {
        if (!isResourceParameterBaseForInternal(context.program, property, targetResourceBaseType))
            removedProperties.push(propertyName);
    }
    for (const removedProperty of removedProperties) {
        entity.properties.delete(removedProperty);
    }
};
const $resourceParameterBaseFor = (context, entity, values) => {
    const resolvedValues = [];
    // TODO this will crash if passed anything other than a tuple
    for (const value of values.values) {
        if (value.kind !== "EnumMember") {
            return;
        }
        resolvedValues.push(value.name);
    }
    context.program.stateMap(ArmStateKeys.armResourceCollection).set(entity, resolvedValues);
};
const $defaultResourceKeySegmentName = (context, entity, resource, keyName, segment) => {
    const modelName = camelCase(resource.name);
    const pluralName = pluralize(modelName);
    if (keyName.length > 0) {
        context.call($key, entity, keyName);
    }
    else {
        context.call($key, entity, `${modelName}Name`);
    }
    if (segment.length > 0) {
        context.call($segment, entity, segment);
    }
    else {
        context.call($segment, entity, pluralName);
    }
};
export function getResourceParameterBases(program, property) {
    return program.stateMap(ArmStateKeys.armResourceCollection).get(property);
}
export function isResourceParameterBaseFor(program, property, resourceBaseType) {
    return isResourceParameterBaseForInternal(program, property, resolveResourceBaseType(resourceBaseType));
}
function isResourceParameterBaseForInternal(program, property, resolvedBaseType) {
    const resourceBases = getResourceParameterBases(program, property);
    if (resourceBases !== undefined) {
        for (const rawType of resourceBases) {
            if (resolveResourceBaseType(rawType) === resolvedBaseType)
                return true;
        }
    }
    return false;
}
/**
 * This decorator dynamically assigns the serviceNamespace from the containing
 * namespace to the string literal value of the path parameter to which this
 * decorator is applied.  Its purpose is to dynamically insert the provider
 * namespace (e.g. 'Microsoft.CodeSigning') into the path parameter list.
 * @param {DecoratorContext} context DecoratorContext
 * @param {Type} target Target of this decorator. Must be a string `ModelProperty`.
 * @param {Type} resourceType Must be a `Model`.
 */
const $assignProviderNameValue = (context, target, resourceType) => {
    const { program } = context;
    let armProviderNamespace = getArmProviderNamespace(program, resourceType);
    // Workaround for deprecated Legacy.Provider which by default point to TenantActionScope which is not able to resolve the provider namespace
    if (resourceType.name === "TenantActionScope" &&
        armProviderNamespace === undefined &&
        target.model?.namespace) {
        armProviderNamespace = getArmProviderNamespace(program, target.model.namespace);
    }
    if (armProviderNamespace &&
        target.type.kind === "String" &&
        target.type.value === "Microsoft.ThisWillBeReplaced") {
        target.type = $(context.program).literal.createString(armProviderNamespace);
    }
};
/**
 * This decorator allows setting a unique provider name value, for scenarios in which
 * multiple providers are allowed.
 * @param {DecoratorContext} context DecoratorContext
 * @param {Type} target Target of this decorator. Must be a string `ModelProperty`.
 * @param {Type} resourceType Must be a `Model`.
 */
const $assignUniqueProviderNameValue = (context, target, resourceType) => {
    const { program } = context;
    const armProviderNamespace = getArmProviderNamespace(program, resourceType);
    if (!armProviderNamespace && !isBuiltIn(getResourceBaseType(program, resourceType))) {
        reportDiagnostic(program, {
            code: "resource-without-provider-namespace",
            format: { resourceName: resourceType.name },
            target: resourceType,
        });
        return;
    }
    if (armProviderNamespace &&
        target.type.kind === "String" &&
        target.type.value !== armProviderNamespace) {
        target.type = $(program).literal.createString(armProviderNamespace);
    }
};
/**
 * Update the ARM provider namespace for a given entity.
 * @param {DecoratorContext} context DecoratorContext
 * @param {Type} entity Entity to set namespace. Must be a `Operation`.
 * @returns
 */
const $armUpdateProviderNamespace = (context, entity) => {
    const { program } = context;
    const operation = entity;
    const opInterface = operation.interface;
    if (opInterface && opInterface.namespace) {
        const armProviderNamespace = getArmProviderNamespace(program, opInterface.namespace);
        if (armProviderNamespace) {
            // Set the namespace constant on the 'provider' parameter
            const providerParam = operation.parameters.properties.get("provider");
            if (providerParam) {
                if (providerParam.type.kind !== "String") {
                    reportDiagnostic(program, {
                        code: "decorator-param-wrong-type",
                        messageId: "armUpdateProviderNamespace",
                        target: providerParam,
                    });
                    return;
                }
                if (providerParam.type.value === "Microsoft.ThisWillBeReplaced") {
                    providerParam.type.value = armProviderNamespace;
                }
            }
        }
    }
};
/**
 * Check if an interface is extending the Azure.ResourceManager.Operations interface.
 */
export function isArmOperationsListInterface(program, type) {
    if (type.name !== "Operations") {
        return false;
    }
    const listOperation = type.operations.get("list");
    if (listOperation) {
        if (getSegment(program, listOperation) === "operations") {
            return true;
        }
    }
    return false;
}
/**
 * This decorator is used to identify ARM resource types and extract their
 * metadata.  It is *not* meant to be used directly by a spec author, it instead
 * gets implicitly applied when the spec author defines a model type in this form:
 *
 *   `model Server is TrackedResource<ServerProperties>;`
 *
 * The `TrackedResource<T>` type (and other associated base types) use the @armResource
 * decorator, so it also gets applied to the type which absorbs the `TrackedResource<T>`
 * definition by using the `is` keyword.
 */
const $armResourceInternal = (context, resourceType, propertiesType) => {
    registerArmResource(context, resourceType);
};
const $armResourceWithParameter = (context, target, properties, type, nameParameter) => {
    registerArmResource(context, target, type, nameParameter);
};
function getPrimaryKeyProperty(program, resource) {
    const nameProperty = resource.properties.get("name");
    if (nameProperty !== undefined)
        return nameProperty;
    const keyProps = [...resource.properties.values()].filter((prop) => isKey(program, prop));
    if (keyProps.length !== 1)
        return undefined;
    return keyProps[0];
}
export function registerArmResource(context, resourceType, type, nameParameter) {
    const { program } = context;
    const namespaceName = resourceType.namespace ? getTypeName(resourceType.namespace) : undefined;
    if (namespaceName === undefined ||
        namespaceName === "Azure.ResourceManager" ||
        namespaceName === "Azure.ResourceManager.Legacy" ||
        namespaceName === "Azure.ResourceManager.CommonTypes") {
        // The @armResource decorator will be evaluated on instantiations of
        // base templated resource types like TrackedResource<SomeResource>,
        // so ignore in that case.
        return;
    }
    // The global namespace has an empty string as name
    if (!resourceType.namespace || resourceType.namespace.name === "") {
        reportDiagnostic(program, {
            code: "decorator-in-namespace",
            format: { decoratorName: "armResource" },
            target: resourceType,
        });
        return;
    }
    // Locate the ARM namespace in the namespace hierarchy
    const armProviderNamespace = getArmProviderNamespace(program, resourceType);
    const armLibraryNamespace = isArmLibraryNamespace(program, resourceType.namespace);
    const armExternalNamespace = getArmVirtualResourceDetails(program, resourceType)?.provider;
    if (!armProviderNamespace && !armLibraryNamespace && armExternalNamespace === undefined) {
        reportDiagnostic(program, { code: "arm-resource-missing-arm-namespace", target: resourceType });
        return;
    }
    let keyName = undefined;
    let collectionName = undefined;
    let kind = undefined;
    if (type !== undefined && nameParameter !== undefined) {
        keyName = nameParameter;
        collectionName = type;
        kind = "Proxy";
    }
    else {
        // Ensure the resource type has defined a name property that has a segment
        const primaryKeyProperty = getPrimaryKeyProperty(program, resourceType);
        if (!primaryKeyProperty) {
            reportDiagnostic(program, {
                code: "arm-resource-missing-name-property",
                target: resourceType,
            });
            return;
        }
        // Set the name property to be read only
        if (primaryKeyProperty.name === "name") {
            const Lifecycle = getLifecycleVisibilityEnum(program);
            clearVisibilityModifiersForClass(program, primaryKeyProperty, Lifecycle, context);
            addVisibilityModifiers(program, primaryKeyProperty, [Lifecycle.members.get("Read")], context);
            sealVisibilityModifiers(program, primaryKeyProperty, Lifecycle);
        }
        keyName = getKeyName(program, primaryKeyProperty);
        if (!keyName) {
            reportDiagnostic(program, {
                code: "arm-resource-missing-name-key-decorator",
                target: resourceType,
            });
            return;
        }
        collectionName = getSegment(program, primaryKeyProperty);
        if (!collectionName) {
            reportDiagnostic(program, {
                code: "arm-resource-missing-name-segment-decorator",
                target: resourceType,
            });
            return;
        }
        kind = getArmResourceKind(resourceType);
        if (isArmVirtualResource(program, resourceType))
            kind = "Virtual";
        if (isCustomAzureResource(program, resourceType))
            kind = "Custom";
    }
    if (!kind) {
        reportDiagnostic(program, {
            code: "arm-resource-invalid-base-type",
            target: resourceType,
        });
        return;
    }
    const armResourceDetails = {
        name: resourceType.name,
        kind,
        typespecType: resourceType,
        collectionName,
        keyName,
        armProviderNamespace: armProviderNamespace ?? armExternalNamespace ?? "",
        operations: {
            lifecycle: {},
            lists: {},
            actions: {},
        },
    };
    setArmResource(context.program, resourceType, armResourceDetails);
}
const [getArmResource, setArmResource, armResourceStateMap] = useStateMap(ArmStateKeys.armResources);
export { getArmResource };
export function listArmResources(program) {
    return [...armResourceStateMap(program).values()];
}
function getProperty(model, propertyName) {
    let returnProperty = model.properties?.get(propertyName);
    if (!returnProperty && model.baseModel) {
        returnProperty = getProperty(model.baseModel, propertyName);
    }
    return returnProperty;
}
function hasProperty(program, model) {
    if (model.properties.size > 0)
        return true;
    if (model.baseModel)
        return hasProperty(program, model.baseModel);
    return false;
}
const $azureResourceBase = (context, resourceType) => {
    context.program.stateMap(ArmStateKeys.azureResourceBase).set(resourceType, true);
};
export function isAzureResource(program, resourceType) {
    const isResourceBase = program.stateMap(ArmStateKeys.azureResourceBase).get(resourceType);
    return isResourceBase ?? false;
}
const $armRenameListByOperation = (context, entity, resourceType, parentTypeName, parentFriendlyTypeName, applyOperationRename) => {
    armRenameListByOperationInternal(context, entity, resourceType, parentTypeName, parentFriendlyTypeName, applyOperationRename);
};
const $armResourcePropertiesOptionality = (context, target, isOptional) => {
    if (target.name === "properties") {
        target.optional = isOptional;
    }
};
const $armBodyRoot = (context, target, isOptional) => {
    target.optional = isOptional;
    context.call($bodyRoot, target);
};
const $legacyType = (context, target) => {
    const { program } = context;
    if (target.namespace &&
        getNamespaceFullName(target.namespace).startsWith("Azure.ResourceManager")) {
        return;
    }
    reportDiagnostic(program, { code: "legacy-type-usage", target });
};
const $resourceParentType = (context, target, parentType) => {
    const { program } = context;
    setResourceBaseType(program, target, parentType);
};
function callOperationDecorator(context, target, resourceType, resourceName, operationType) {
    switch (operationType) {
        case "read":
            context.call($armResourceRead, target, resourceType, resourceName);
            break;
        case "createOrUpdate":
            context.call($armResourceCreateOrUpdate, target, resourceType, resourceName);
            break;
        case "update":
            context.call($armResourceUpdate, target, resourceType, resourceName);
            break;
        case "delete":
            context.call($armResourceDelete, target, resourceType, resourceName);
            break;
        case "list":
            context.call($armResourceList, target, resourceType, resourceName);
            break;
        case "action":
            context.call($armResourceAction, target, resourceType, resourceName);
            break;
        case "checkExistence":
            context.call($armResourceCheckExistence, target, resourceType, resourceName);
            break;
    }
}
function callLifecycleDecorator(context, target, resourceType, operationType) {
    switch (operationType) {
        case "read":
            context.call($readsResource, target, resourceType);
            break;
        case "createOrUpdate":
            context.call($createsOrReplacesResource, target, resourceType);
            break;
        case "update":
            context.call($updatesResource, target, resourceType);
            break;
        case "delete":
            context.call($deletesResource, target, resourceType);
            break;
        case "list":
            context.call($listsResource, target, resourceType);
            break;
        case "checkExistence":
            context.call($armResourceCheckExistence, target, resourceType);
            break;
    }
}
const $extensionResourceOperation = (context, target, targetResourceType, extensionResourceType, operationType, resourceName) => {
    if (target.interface === undefined ||
        target.interface.node === undefined ||
        target.interface.node.templateParameters.length > 0) {
        return;
    }
    const resolvedResourceName = resourceName === undefined || resourceName.length === 0
        ? `${targetResourceType.name}${extensionResourceType.name}`
        : resourceName;
    callOperationDecorator(context, target, extensionResourceType, resolvedResourceName, operationType);
};
const $builtInResourceOperation = (context, target, parentResourceType, builtInResourceType, operationType, resourceName) => {
    if (target.interface === undefined ||
        target.interface.node === undefined ||
        target.interface.node.templateParameters.length > 0) {
        return;
    }
    const resolvedResourceName = resourceName === undefined || resourceName.length === 0
        ? `${parentResourceType.name}${builtInResourceType.name}`
        : resourceName;
    callOperationDecorator(context, target, builtInResourceType, resolvedResourceName, operationType);
};
const $legacyResourceOperation = (context, target, resourceType, operationType, resourceName) => {
    if (target.interface === undefined ||
        target.interface.node === undefined ||
        target.interface.node.templateParameters.length > 0) {
        return;
    }
    const resolvedResourceName = resourceName !== undefined && resourceName.length > 0 ? resourceName : undefined;
    // We can't resolve the operation path yet so treat the operation as a partial
    // type so that we can fill in the missing details later
    const operations = getArmResourceOperations(context.program, resourceType);
    const operation = {
        name: target.name,
        kind: operationType,
        operation: target,
        operationGroup: target.interface.name,
        resourceName: resolvedResourceName,
        resourceModelName: resourceType.name,
        resourceKind: "legacy",
    };
    operations.lists[target.name] = operation;
    const opId = {
        name: target.name,
        kind: operationType,
        operation: target,
        operationGroup: target.interface.name,
        resource: resourceType,
        resourceName: resolvedResourceName,
        resourceKind: "legacy",
    };
    addArmResourceOperation(context.program, resourceType, opId);
    setArmOperationIdentifier(context.program, target, resourceType, {
        kind: operationType,
        name: target.name,
        operation: target,
        operationGroup: target.interface.name,
        resourceModelName: resourceType.name,
        resourceName: resolvedResourceName,
        resourceKind: "legacy",
    });
    const { program } = context;
    callLifecycleDecorator(context, target, resourceType, operationType);
    if (operationType === "action") {
        const segment = getSegment(program, target) ?? getActionSegment(program, target);
        if (!segment) {
            // Also apply the @actionSegment decorator to the operation
            context.call($actionSegment, target, uncapitalize(target.name));
        }
    }
};
function uncapitalize(str) {
    return str.charAt(0).toLowerCase() + str.slice(1);
}
const $legacyExtensionResourceOperation = (context, target, resourceType, operationType, resourceName) => {
    if (target.interface === undefined ||
        target.interface.node === undefined ||
        target.interface.node.templateParameters.length > 0) {
        return;
    }
    const resolvedResourceName = resourceName !== undefined && resourceName.length > 0 ? resourceName : undefined;
    // We can't resolve the operation path yet so treat the operation as a partial
    // type so that we can fill in the missing details later
    const operations = getArmResourceOperations(context.program, resourceType);
    const operation = {
        name: target.name,
        kind: operationType,
        operation: target,
        operationGroup: target.interface.name,
        resourceName: resolvedResourceName,
        resourceModelName: resourceType.name,
        resourceKind: "legacy-extension",
    };
    operations.lists[target.name] = operation;
    const opId = {
        name: target.name,
        kind: operationType,
        operation: target,
        operationGroup: target.interface.name,
        resource: resourceType,
        resourceName: resolvedResourceName,
        resourceKind: "legacy-extension",
    };
    addArmResourceOperation(context.program, resourceType, opId);
    setArmOperationIdentifier(context.program, target, resourceType, {
        kind: operationType,
        name: target.name,
        operation: target,
        operationGroup: target.interface.name,
        resourceModelName: resourceType.name,
        resourceName: resolvedResourceName,
        resourceKind: "legacy-extension",
    });
    const { program } = context;
    callLifecycleDecorator(context, target, resourceType, operationType);
    if (operationType === "action") {
        const segment = getSegment(program, target) ?? getActionSegment(program, target);
        if (!segment) {
            // Also apply the @actionSegment decorator to the operation
            context.call($actionSegment, target, uncapitalize(target.name));
        }
    }
};
const $validateCommonTypesVersionForResource = (context, target, version, resourceName) => {
    if (isTemplateDeclarationOrInstance(target) && !isTemplateInstance(target)) {
        return;
    }
    const lowestVersion = getLowestCommonTypeVersion(context.program, target.namespace);
    if (lowestVersion === undefined || lowestVersion < version) {
        reportDiagnostic(context.program, {
            code: "invalid-version-for-common-type",
            target,
            format: {
                resourceName: resourceName,
                requiredVersion: version,
                version: lowestVersion ?? "v3",
            },
        });
    }
};
function getLowestCommonTypeVersion(program, ns) {
    let lowestVersion = getArmCommonTypesVersion(program, ns);
    // If it is versioned namespace, we will check each Version enum member. If no
    // @armCommonTypeVersion decorator, add the one
    const versioned = ns.decorators.find((x) => x.definition?.name === "@versioned");
    if (versioned) {
        const versionEnum = versioned.args[0].value;
        versionEnum.members.forEach((v) => {
            const candidateVersion = getArmCommonTypesVersion(program, v);
            if (candidateVersion !== undefined &&
                (lowestVersion === undefined || lowestVersion > candidateVersion)) {
                lowestVersion = candidateVersion;
            }
        });
    }
    return lowestVersion;
}
/** @internal */
export const $decorators = {
    "Azure.ResourceManager.Private": {
        resourceBaseParametersOf: $resourceBaseParametersOf,
        resourceParameterBaseFor: $resourceParameterBaseFor,
        azureResourceBase: $azureResourceBase,
        omitIfEmpty: $omitIfEmpty,
        assignUniqueProviderNameValue: $assignUniqueProviderNameValue,
        assignProviderNameValue: $assignProviderNameValue,
        armUpdateProviderNamespace: $armUpdateProviderNamespace,
        armResourceInternal: $armResourceInternal,
        defaultResourceKeySegmentName: $defaultResourceKeySegmentName,
        enforceConstraint: $enforceConstraint,
        armRenameListByOperation: $armRenameListByOperation,
        armResourcePropertiesOptionality: $armResourcePropertiesOptionality,
        armBodyRoot: $armBodyRoot,
        armResourceWithParameter: $armResourceWithParameter,
        legacyType: $legacyType,
        resourceParentType: $resourceParentType,
        extensionResourceOperation: $extensionResourceOperation,
        legacyExtensionResourceOperation: $legacyExtensionResourceOperation,
        legacyResourceOperation: $legacyResourceOperation,
        builtInResourceOperation: $builtInResourceOperation,
        validateCommonTypesVersionForResource: $validateCommonTypesVersionForResource,
    },
    "Azure.ResourceManager.Extension.Private": {
        builtInResource: $builtInResource,
        builtInSubscriptionResource: $builtInSubscriptionResource,
        builtInResourceGroupResource: $builtInResourceGroupResource,
    },
};
