// @azure-tools/typespec-azure-resource-manager/dist/src/state.js
// FIXME - This is a workaround for the circular dependency issue when loading
// createStateSymbol.
// Issue: https://github.com/microsoft/typespec/issues/2301
function azureResourceManagerCreateStateSymbol(name) {
    return Symbol.for(`@azure-tools/typespec-azure-resource-manager.${name}`);
}
export const ArmStateKeys = {
    armProviderCache: azureResourceManagerCreateStateSymbol("armProviderCache"),
    armProviderNamespaces: azureResourceManagerCreateStateSymbol("armProviderNamespaces"),
    armResourceOperations: azureResourceManagerCreateStateSymbol("armResourceOperations"),
    armResourceCollectionAction: azureResourceManagerCreateStateSymbol("armResourceCollectionAction"),
    armResourceCollection: azureResourceManagerCreateStateSymbol("parameterBaseTypes"),
    armResources: azureResourceManagerCreateStateSymbol("armResources"),
    armLibraryNamespaces: azureResourceManagerCreateStateSymbol("armLibraryNamespaces"),
    usesArmLibraryNamespaces: azureResourceManagerCreateStateSymbol("usesArmLibraryNamespaces"),
    armCommonTypesVersion: azureResourceManagerCreateStateSymbol("armCommonTypesVersion"),
    armIdentifiers: azureResourceManagerCreateStateSymbol("armIdentifiers"),
    externalTypeRef: azureResourceManagerCreateStateSymbol("externalTypeRef"),
    armResourceOperationData: azureResourceManagerCreateStateSymbol("armResourceOperationData"),
    resourceOperationList: azureResourceManagerCreateStateSymbol("resourceOperationList"),
    // resource.ts
    armResolvedResources: azureResourceManagerCreateStateSymbol("armResolvedResources"),
    armResourcesCached: azureResourceManagerCreateStateSymbol("armResourcesCached"),
    armSingletonResources: azureResourceManagerCreateStateSymbol("armSingletonResources"),
    resourceBaseType: azureResourceManagerCreateStateSymbol("resourceBaseTypeKey"),
    armBuiltInResource: azureResourceManagerCreateStateSymbol("armExternalResource"),
    customAzureResource: azureResourceManagerCreateStateSymbol("azureCustomResource"),
    armFeature: azureResourceManagerCreateStateSymbol("armFeature"),
    armFeatureSet: azureResourceManagerCreateStateSymbol("armFeatureSet"),
    armFeatureOptions: azureResourceManagerCreateStateSymbol("armFeatureOptions"),
    // private.decorator.ts
    azureResourceBase: azureResourceManagerCreateStateSymbol("azureResourceBase"),
    renamePathParameters: azureResourceManagerCreateStateSymbol("renamePathParameters"),
    // commontypes.private.decorators.ts
    armCommonDefinitions: azureResourceManagerCreateStateSymbol("armCommonDefinitions"),
    armCommonParameters: azureResourceManagerCreateStateSymbol("armCommonParameters"),
    armCommonTypesVersions: azureResourceManagerCreateStateSymbol("armCommonTypesVersions"),
    armResourceRoute: azureResourceManagerCreateStateSymbol("armResourceRoute"),
    armExternalType: azureResourceManagerCreateStateSymbol("armExternalType"),
    inlineAzureType: azureResourceManagerCreateStateSymbol("inlineAzureType"),
};
