// @azure-tools/typespec-azure-resource-manager/dist/src/tsp-index.js
import { $armCommonTypesVersion, $externalTypeRef } from "./common-types.js";
import { $armLibraryNamespace, $armProviderNamespace, $useLibraryNamespace } from "./namespace.js";
import { $armOperationRoute, $armResourceAction, $armResourceCheckExistence, $armResourceCollectionAction, $armResourceCreateOrUpdate, $armResourceDelete, $armResourceList, $armResourceRead, $armResourceUpdate, $renamePathParameter, } from "./operations.js";
import { $armExternalType, $armProviderNameValue, $armResourceOperations, $armVirtualResource, $customAzureResource, $extensionResource, $feature, $featureOptions, $features, $identifiers, $locationResource, $resourceBaseType, $resourceGroupResource, $singleton, $subscriptionResource, $tenantResource, } from "./resource.js";
export { $lib } from "./lib.js";
/** @internal */
export const $decorators = {
    "Azure.ResourceManager": {
        armResourceCollectionAction: $armResourceCollectionAction,
        armProviderNameValue: $armProviderNameValue,
        armProviderNamespace: $armProviderNamespace,
        useLibraryNamespace: $useLibraryNamespace,
        armLibraryNamespace: $armLibraryNamespace,
        singleton: $singleton,
        tenantResource: $tenantResource,
        subscriptionResource: $subscriptionResource,
        locationResource: $locationResource,
        resourceGroupResource: $resourceGroupResource,
        extensionResource: $extensionResource,
        armResourceAction: $armResourceAction,
        armResourceCreateOrUpdate: $armResourceCreateOrUpdate,
        armResourceRead: $armResourceRead,
        armResourceUpdate: $armResourceUpdate,
        armResourceDelete: $armResourceDelete,
        armResourceList: $armResourceList,
        armResourceCheckExistence: $armResourceCheckExistence,
        armResourceOperations: $armResourceOperations,
        armCommonTypesVersion: $armCommonTypesVersion,
        armVirtualResource: $armVirtualResource,
        resourceBaseType: $resourceBaseType,
        identifiers: $identifiers,
    },
    "Azure.ResourceManager.Legacy": {
        customAzureResource: $customAzureResource,
        externalTypeRef: $externalTypeRef,
        armOperationRoute: $armOperationRoute,
        armExternalType: $armExternalType,
        renamePathParameter: $renamePathParameter,
        feature: $feature,
        features: $features,
        featureOptions: $featureOptions,
    },
};
