// oav/dist/lib/inversifyUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.inversifyGetInstance = exports.inversifyGetContainer = exports.TYPES = void 0;
exports.TYPES = {
    opts: Symbol("InversifyTYPES.opts"),
    emptyObject: Symbol("InversifyTYPES.emptyObject"),
    schemaValidator: Symbol("InversifyTYPES.schemaValidator"),
};
const inversify_1 = require("inversify");
const loader_1 = require("./swagger/loader");
const inversifyGetContainer = (opts = {}) => {
    (0, loader_1.setDefaultOpts)(opts, {
        defaultScope: "Singleton",
        autobind: true,
    });
    return new inversify_1.Container(opts);
};
exports.inversifyGetContainer = inversifyGetContainer;
const inversifyGetInstance = (claz, opts) => {
    if (opts.container === undefined) {
        opts.container = (0, exports.inversifyGetContainer)(opts);
    }
    opts.container.bind(exports.TYPES.opts).toConstantValue(opts);
    opts.container.bind(exports.TYPES.emptyObject).toConstantValue({});
    const { AjvSchemaValidator } = require("./swaggerValidator/ajvSchemaValidator");
    opts.container.bind(exports.TYPES.schemaValidator).to(AjvSchemaValidator);
    return opts.container.get(claz);
};
exports.inversifyGetInstance = inversifyGetInstance;
