// oav/dist/lib/validate.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateSpec = void 0;
exports.validateExamples = validateExamples;
exports.validateTraffic = validateTraffic;
exports.extractXMsExamples = extractXMsExamples;
exports.generateExamples = generateExamples;
/* eslint-disable no-console */
const path = __importStar(require("path"));
const fs = __importStar(require("fs"));
const js_yaml_1 = require("js-yaml");
const openapi_tools_common_1 = require("@azure-tools/openapi-tools-common");
const utils = __importStar(require("./util/utils"));
const modelValidator_1 = require("./swaggerValidator/modelValidator");
const XMsExampleExtractor = __importStar(require("./xMsExampleExtractor"));
const exampleGenerator_1 = __importDefault(require("./generator/exampleGenerator"));
const logging_1 = require("./util/logging");
const semanticValidator_1 = require("./swaggerValidator/semanticValidator");
const errorDefinitions_1 = require("./util/errorDefinitions");
const trafficValidator_1 = require("./swaggerValidator/trafficValidator");
const generateReport_1 = require("./report/generateReport");
const vsoLogIssueWrapper = (issueType, message) => {
    if (issueType === "error" || issueType === "warning") {
        return `##vso[task.logissue type=${issueType}]${message}`;
    }
    else {
        return `##vso[task.logissue type=${issueType}]${message}`;
    }
};
const validate = async (options, func) => {
    if (!options) {
        options = {};
    }
    logging_1.log.consoleLogLevel = options.consoleLogLevel || logging_1.log.consoleLogLevel;
    logging_1.log.filepath = options.logFilepath || logging_1.log.filepath;
    if (options.pretty) {
        logging_1.log.consoleLogLevel = "off";
    }
    return func(options);
};
const prettyPrint = (errors, errorType) => {
    if (errors !== undefined) {
        for (const error of errors) {
            const yaml = (0, js_yaml_1.dump)(error);
            if (process.env["Agent.Id"]) {
                // eslint-disable-next-line no-console
                console.error(vsoLogIssueWrapper(errorType, yaml));
            }
            else {
                // eslint-disable-next-line no-console
                console.error(yaml);
            }
        }
    }
};
const prettyPrintInfo = (errors, errorType) => {
    if (errors !== undefined) {
        for (const error of errors) {
            const yaml = (0, js_yaml_1.dump)(error);
            if (process.env["Agent.Id"]) {
                // eslint-disable-next-line no-console
                console.error(vsoLogIssueWrapper(errorType, yaml));
            }
            else {
                // eslint-disable-next-line no-console
                console.error(yaml);
            }
        }
    }
};
const validateSpec = async (specPath, options) => validate(options, async (o) => {
    const validator = new semanticValidator_1.SemanticValidator(specPath, null);
    try {
        await validator.initialize();
        logging_1.log.info(`Semantically validating  ${specPath}:\n`);
        const validationResults = await validator.validateSpec();
        if (o.pretty) {
            if (validationResults.errors.length > 0) {
                logMessage(`Semantically validating ${specPath}`, "error");
            }
            else {
                logMessage(`Semantically validating ${specPath} without error`, "info");
            }
            if (validationResults.errors.length > 0) {
                logMessage(`Errors reported:`, "error");
                prettyPrint(validationResults.errors, "error");
            }
        }
        else {
            if (validationResults.errors.length > 0) {
                logMessage(`Errors reported:`, "error");
                for (const error of validationResults.errors) {
                    // eslint-disable-next-line no-console
                    logging_1.log.error(error);
                }
            }
            else {
                logMessage(`Semantically validating ${specPath} without error`, "info");
            }
        }
        return validator.specValidationResult;
    }
    catch (err) {
        let outputMsg = err;
        if (typeof err === "object") {
            outputMsg = (0, js_yaml_1.dump)(err);
        }
        if (o.pretty) {
            logMessage(`Semantically validating ${specPath}`);
            logMessage(`${outputMsg}`, "error");
        }
        else {
            logging_1.log.error(`Detail error:${err === null || err === void 0 ? void 0 : err.message}.ErrorStack:${err === null || err === void 0 ? void 0 : err.stack}`);
        }
        validator.specValidationResult.validityStatus = false;
        return validator.specValidationResult;
    }
});
exports.validateSpec = validateSpec;
async function validateExamples(specPath, operationIds, options) {
    return validate(options, async (o) => {
        try {
            const validator = new modelValidator_1.NewModelValidator(specPath);
            await validator.initialize();
            logging_1.log.info(`Validating "examples" and "x-ms-examples" in  ${specPath}:\n`);
            await validator.validateOperations(operationIds);
            const errors = validator.result;
            if (o.pretty) {
                if (errors.length > 0) {
                    logMessage(`Validating "examples" and "x-ms-examples" in ${specPath}`, "error");
                    logMessage("Error reported:");
                    prettyPrint(errors, "error");
                }
                else {
                    logMessage("Validation completes without errors.", "info");
                }
            }
            else {
                if (errors.length > 0) {
                    logMessage("Error reported:");
                    for (const error of errors) {
                        logging_1.log.error(error);
                    }
                }
                else {
                    logMessage("Validation completes without errors.", "info");
                }
            }
            return errors;
        }
        catch (e) {
            logMessage(`Validating x-ms-examples in ${specPath}`, "error");
            logMessage("Unexpected runtime exception:");
            if (o.pretty) {
                logMessage(`Detail error:${e === null || e === void 0 ? void 0 : e.message}.ErrorStack:${e === null || e === void 0 ? void 0 : e.stack}`, "error");
            }
            else {
                logging_1.log.error(`Detail error:${e === null || e === void 0 ? void 0 : e.message}.ErrorStack:${e === null || e === void 0 ? void 0 : e.stack}`);
            }
            const error = {
                inner: e,
                message: "Unexpected internal error",
                code: errorDefinitions_1.ErrorCodeConstants.INTERNAL_ERROR,
            };
            return [error];
        }
    });
}
async function validateTraffic(specPath, trafficPath, options) {
    return validate(options, async (o) => {
        o.consoleLogLevel = logging_1.log.consoleLogLevel;
        o.logFilepath = logging_1.log.filepath;
        const validator = new trafficValidator_1.TrafficValidator(specPath, trafficPath);
        const trafficValidationResult = [];
        try {
            await validator.initialize();
            const result = await validator.validate();
            trafficValidationResult.push(...result);
        }
        catch (err) {
            const msg = `Detail error message:${err === null || err === void 0 ? void 0 : err.message}. ErrorStack:${err === null || err === void 0 ? void 0 : err.Stack}`;
            logging_1.log.error(msg);
            trafficValidationResult.push({
                payloadFilePath: specPath,
                runtimeExceptions: [
                    {
                        code: errorDefinitions_1.ErrorCodeConstants.RUNTIME_ERROR,
                        message: msg,
                    },
                ],
            });
        }
        if (options.jsonReportPath) {
            const errorDefinitions = await (0, generateReport_1.loadErrorDefinitions)();
            const report = {
                coveredSpecFiles: validator.operationCoverageResult.map((item) => {
                    var _a, _b;
                    return options.overrideLinkInReport
                        ? `${options.specLinkPrefix}/${(_a = item.spec) === null || _a === void 0 ? void 0 : _a.substring((_b = item.spec) === null || _b === void 0 ? void 0 : _b.indexOf("specification"))}`
                        : `${item.spec}`;
                }),
                allOperations: validator.operationCoverageResult
                    .map((item) => item.totalOperations)
                    .reduce((a, b) => a + b, 0),
                coveredOperations: validator.operationCoverageResult
                    .map((item) => item.coveredOperations)
                    .reduce((a, b) => a + b, 0),
                unCoveredOperationsList: validator.operationCoverageResult.map((item) => {
                    return {
                        spec: item.spec,
                        operationIds: item.unCoveredOperationsList.map((opeartion) => opeartion.operationId),
                    };
                }),
                passedOperationsList: validator.operationCoverageResult.map((item) => {
                    return {
                        spec: item.spec,
                        operationIds: item.coveredOperationsList
                            .map((opeartion) => opeartion.operationId)
                            .filter((id) => {
                            return !trafficValidationResult.some((error) => { var _a; return ((_a = error.operationInfo) === null || _a === void 0 ? void 0 : _a.operationId) === id; });
                        }),
                    };
                }),
                failedOperations: validator.operationCoverageResult
                    .map((item) => item.validationFailOperations)
                    .reduce((a, b) => a + b, 0),
                errors: Array.from((0, openapi_tools_common_1.flatMap)(trafficValidationResult, (item) => {
                    var _a, _b;
                    return (_b = (_a = item.errors) === null || _a === void 0 ? void 0 : _a.map((it) => {
                        var _a;
                        const errorDef = errorDefinitions.get(it.code);
                        const specFilePath = item.specFilePath || "";
                        const overrideLinkInReport = options.overrideLinkInReport || false;
                        const specLinkPrefix = options.specLinkPrefix || "";
                        return {
                            spec: specFilePath,
                            errorCode: it.code,
                            errorLink: errorDef === null || errorDef === void 0 ? void 0 : errorDef.link,
                            errorMessage: it.message,
                            issueSource: it.issueSource,
                            operationId: (_a = item.operationInfo) === null || _a === void 0 ? void 0 : _a.operationId,
                            schemaPathWithPosition: overrideLinkInReport
                                ? `${specLinkPrefix}/${specFilePath.substring(specFilePath.indexOf("specification"))}#L${it.source.position.line}`
                                : `${specFilePath}#L${it.source.position.line}`,
                        };
                    })) !== null && _b !== void 0 ? _b : [];
                })),
            };
            fs.writeFileSync(options.jsonReportPath, JSON.stringify(report, null, 2));
        }
        if (options.reportPath) {
            const generator = new generateReport_1.ReportGenerator(trafficValidationResult, validator.operationCoverageResult, validator.operationUndefinedResult, options);
            await generator.generateHtmlReport();
        }
        else if (trafficValidationResult.length > 0) {
            if (o.pretty) {
                prettyPrintInfo(trafficValidationResult, "error");
            }
            else {
                for (const error of trafficValidationResult) {
                    const errorInfo = JSON.stringify(error);
                    logging_1.log.error(errorInfo);
                }
            }
        }
        else {
            logging_1.log.info("No errors were found.");
        }
        return trafficValidationResult;
    });
}
async function extractXMsExamples(specPath, recordings, options) {
    if (!options) {
        options = {};
    }
    logging_1.log.consoleLogLevel = options.consoleLogLevel || logging_1.log.consoleLogLevel;
    logging_1.log.filepath = options.logFilepath || logging_1.log.filepath;
    const xMsExampleExtractor = new XMsExampleExtractor.XMsExampleExtractor(specPath, recordings, options);
    return xMsExampleExtractor.extract();
}
async function generateExamples(specPath, payloadDir, operationIds, readme, tag, generationRule, options) {
    if (!options) {
        options = {};
    }
    const wholeInputFiles = [];
    if (readme && tag) {
        const inputFiles = await utils.getInputFiles(readme, tag);
        if (!inputFiles) {
            throw Error("get input files from readme tag failed.");
        }
        inputFiles.forEach((file) => {
            if (path.isAbsolute(file)) {
                wholeInputFiles.push(file);
            }
            else {
                wholeInputFiles.push(path.join(path.dirname(readme), file));
            }
        });
    }
    else if (specPath) {
        wholeInputFiles.push(specPath);
    }
    if (wholeInputFiles.length === 0) {
        console.error(`no spec file specified !`);
    }
    logging_1.log.consoleLogLevel = options.consoleLogLevel || logging_1.log.consoleLogLevel;
    logging_1.log.filepath = options.logFilepath || logging_1.log.filepath;
    for (const file of wholeInputFiles) {
        const generator = new exampleGenerator_1.default(file, payloadDir, generationRule);
        if (operationIds) {
            const operationIdArray = operationIds.trim().split(",");
            for (const operationId of operationIdArray) {
                if (operationId) {
                    await generator.generate(operationId);
                }
            }
            continue;
        }
        await generator.generateAll();
    }
}
const logMessage = (message, level) => {
    const logLevel = level || "error";
    if (process.env["Agent.Id"]) {
        if (level === "error") {
            console.error(vsoLogIssueWrapper(`${logLevel}`, `${message}\n`));
        }
        else {
            console.info(vsoLogIssueWrapper(`${logLevel}`, `${message}\n`));
        }
    }
    else {
        if (level === "error") {
            console.error(`${message}\n`);
        }
        else {
            console.info(`${message}\n`);
        }
    }
};
