// oav/dist/lib/xMsExampleExtractor.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.XMsExampleExtractor = void 0;
const fs = __importStar(require("fs"));
const pathlib = __importStar(require("path"));
const url_1 = require("url");
const openapi_tools_common_1 = require("@azure-tools/openapi-tools-common");
const swagger_parser_1 = __importDefault(require("@apidevtools/swagger-parser"));
const logging_1 = require("./util/logging");
const utils_1 = require("./util/utils");
const mkdirRecursiveSync = (dir) => {
    if (!fs.existsSync(dir)) {
        const parent = pathlib.dirname(dir);
        if (parent !== dir) {
            mkdirRecursiveSync(parent);
        }
        fs.mkdirSync(dir);
    }
};
/**
 * @class
 */
class XMsExampleExtractor {
    /**
     * @constructor
     * Initializes a new instance of the xMsExampleExtractor class.
     *
     * @param {string} specPath the swagger spec path
     *
     * @param {object} recordings the folder for recordings
     *
     * @param {object} [options] The options object
     *
     * @param {object} [options.matchApiVersion] Only generate examples if api-version matches.
     * Default: false
     *
     * @param {object} [options.output] Output folder for the generated examples.
     */
    constructor(specPath, recordings, options) {
        if (specPath === null ||
            specPath === undefined ||
            typeof specPath.valueOf() !== "string" ||
            !specPath.trim().length) {
            throw new Error("specPath is a required property of type string and it cannot be an empty string.");
        }
        if (recordings === null ||
            recordings === undefined ||
            typeof recordings.valueOf() !== "string" ||
            !recordings.trim().length) {
            throw new Error("recordings is a required property of type string and it cannot be an empty string.");
        }
        this.specPath = specPath;
        this.recordings = recordings;
        if (!options) {
            options = {};
        }
        if (options.output === null || options.output === undefined) {
            options.output = process.cwd() + "/output";
        }
        if (options.shouldResolveXmsExamples === null ||
            options.shouldResolveXmsExamples === undefined) {
            options.shouldResolveXmsExamples = true;
        }
        if (options.matchApiVersion === null || options.matchApiVersion === undefined) {
            options.matchApiVersion = false;
        }
        this.options = options;
        logging_1.log.debug(`specPath : ${this.specPath}`);
        logging_1.log.debug(`recordings : ${this.recordings}`);
        logging_1.log.debug(`options.output : ${this.options.output}`);
        logging_1.log.debug(`options.matchApiVersion : ${this.options.matchApiVersion}`);
    }
    extractOne(relativeExamplesPath, outputExamples, 
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    api, recordingFileName) {
        const recording = JSON.parse(fs.readFileSync(recordingFileName).toString());
        const paths = api.paths;
        let pathIndex = 0;
        let pathParams = {};
        for (const path of (0, openapi_tools_common_1.keys)(paths)) {
            pathIndex++;
            const searchResult = path.match(/\/{\w*\}/g);
            const pathParts = path.split("/");
            let pathToMatch = path;
            pathParams = {};
            if (searchResult !== null) {
                for (const match of searchResult) {
                    const splitRegEx = /[{}]/;
                    const pathParam = match.split(splitRegEx)[1];
                    for (const [part, value] of (0, openapi_tools_common_1.entries)(pathParts)) {
                        const pathPart = "/" + value;
                        if (pathPart.localeCompare(match) === 0) {
                            pathParams[pathParam] = part;
                        }
                    }
                    pathToMatch = pathToMatch.replace(match, "/[^/]+");
                }
            }
            let newPathToMatch = pathToMatch.replace(/\//g, "\\/");
            newPathToMatch = newPathToMatch + "$";
            // for this API path (and method), try to find it in the recording file, and get
            // the data
            const recordingEntries = recording.Entries;
            let entryIndex = 0;
            let queryParams = {};
            for (const recordingEntry of (0, openapi_tools_common_1.values)(recordingEntries)) {
                entryIndex++;
                const parsedUrl = new url_1.URL(recordingEntry.RequestUri, "https://management.azure.com");
                let recordingPath = parsedUrl.href || "";
                queryParams = (0, utils_1.kvPairsToObject)(parsedUrl.searchParams) || {};
                const hostUrl = parsedUrl ? parsedUrl.protocol + "//" + parsedUrl.hostname : undefined;
                const headerParams = recordingEntry.RequestHeaders;
                // if command-line included check for API version, validate api-version from URI in
                // recordings matches the api-version of the spec
                if (!this.options.matchApiVersion ||
                    ("api-version" in queryParams && queryParams["api-version"] === api.info.version)) {
                    recordingPath = recordingPath.replace(/\?.*/, "");
                    const recordingPathParts = recordingPath.split("/");
                    // eslint-disable-next-line @typescript-eslint/prefer-regexp-exec
                    const match = recordingPath.match(newPathToMatch);
                    if (match !== null) {
                        logging_1.log.silly("path: " + path);
                        logging_1.log.silly("recording path: " + recordingPath);
                        const pathParamsValues = {};
                        for (const [p, v] of (0, openapi_tools_common_1.mapEntries)(pathParams)) {
                            const index = v;
                            pathParamsValues[p] = recordingPathParts[index];
                        }
                        if (hostUrl !== undefined) {
                            pathParamsValues.url = hostUrl;
                        }
                        // found a match in the recording
                        const requestMethodFromRecording = recordingEntry.RequestMethod;
                        const infoFromOperation = paths[path][requestMethodFromRecording.toLowerCase()];
                        if (typeof infoFromOperation !== "undefined") {
                            // need to consider each method in operation
                            const fileNameArray = recordingFileName.split("/");
                            let fileName = fileNameArray[fileNameArray.length - 1];
                            fileName = fileName.split(".json")[0];
                            fileName = fileName.replace(/\//g, "-");
                            const exampleFileName = `${fileName}-${requestMethodFromRecording}-example-${pathIndex}${entryIndex}.json`;
                            const ref = {
                                $ref: relativeExamplesPath + exampleFileName,
                            };
                            const exampleFriendlyName = `${fileName}${requestMethodFromRecording}${pathIndex}${entryIndex}`;
                            logging_1.log.debug(`exampleFriendlyName: ${exampleFriendlyName}`);
                            if (!("x-ms-examples" in infoFromOperation)) {
                                infoFromOperation["x-ms-examples"] = {};
                            }
                            infoFromOperation["x-ms-examples"][exampleFriendlyName] = ref;
                            const exampleL = {
                                parameters: {},
                                responses: {},
                            };
                            const paramsToProcess = [
                                ...(0, openapi_tools_common_1.mapEntries)(pathParamsValues),
                                ...(0, openapi_tools_common_1.mapEntries)(queryParams),
                                ...(0, openapi_tools_common_1.mapEntries)(headerParams),
                            ];
                            for (const paramEntry of paramsToProcess) {
                                const param = paramEntry[0];
                                const v = paramEntry[1];
                                exampleL.parameters[param] = v;
                            }
                            const params = infoFromOperation.parameters;
                            for (const param of (0, openapi_tools_common_1.keys)(infoFromOperation.parameters)) {
                                if (params[param].in === "body") {
                                    const bodyParamName = params[param].name;
                                    const bodyParamValue = recordingEntry.RequestBody;
                                    const bodyParamExample = {};
                                    bodyParamExample[bodyParamName] = bodyParamValue;
                                    exampleL.parameters[bodyParamName] =
                                        bodyParamValue !== "" ? JSON.parse(bodyParamValue) : "";
                                }
                            }
                            const parseResponseBody = (body) => {
                                try {
                                    return JSON.parse(body);
                                }
                                catch (err) {
                                    return body;
                                }
                            };
                            for (const _v of (0, openapi_tools_common_1.keys)(infoFromOperation.responses)) {
                                const statusCodeFromRecording = recordingEntry.StatusCode;
                                let responseBody = recordingEntry.ResponseBody;
                                if (typeof responseBody === "string" && responseBody !== "") {
                                    responseBody = parseResponseBody(responseBody);
                                }
                                exampleL.responses[statusCodeFromRecording] = {
                                    body: responseBody,
                                };
                            }
                            logging_1.log.info(`Writing x-ms-examples at ${pathlib.resolve(outputExamples, exampleFileName)}`);
                            const examplePath = pathlib.join(outputExamples, exampleFileName);
                            const dir = pathlib.dirname(examplePath);
                            mkdirRecursiveSync(dir);
                            fs.writeFileSync(examplePath, JSON.stringify(exampleL, null, 2));
                        }
                    }
                }
            }
        }
    }
    /**
     * Extracts x-ms-examples from the recordings
     */
    async extract() {
        if (this.options.output === undefined) {
            throw new Error("this.options.output === undefined");
        }
        this.mkdirSync(this.options.output);
        this.mkdirSync(this.options.output + "/examples");
        this.mkdirSync(this.options.output + "/swagger");
        const outputExamples = pathlib.join(this.options.output, "examples");
        const relativeExamplesPath = "../examples/";
        const specName = this.specPath.split("/");
        const outputSwagger = pathlib.join(this.options.output, "swagger", specName[specName.length - 1].split(".")[0] + ".json");
        const accErrors = {};
        const filesArray = [];
        this.getFileList(this.recordings, filesArray);
        const recordingFiles = filesArray;
        try {
            const api = await swagger_parser_1.default.parse(this.specPath);
            for (const recordingFileName of recordingFiles) {
                logging_1.log.debug(`Processing recording file: ${recordingFileName}`);
                try {
                    this.extractOne(relativeExamplesPath, outputExamples, api, recordingFileName);
                    logging_1.log.info(`Writing updated swagger with x-ms-examples at ${outputSwagger}`);
                    fs.writeFileSync(outputSwagger, JSON.stringify(api, null, 2));
                }
                catch (err) {
                    accErrors[recordingFileName] = err.toString();
                    logging_1.log.warn(`Error processing recording file: "${recordingFileName}"`);
                    logging_1.log.warn(`Error: "${err.toString()} "`);
                }
            }
            if (JSON.stringify(accErrors) !== "{}") {
                logging_1.log.error(`Errors loading/parsing recording files.`);
                logging_1.log.error(`${JSON.stringify(accErrors)}`);
            }
        }
        catch (err) {
            process.exitCode = 1;
            logging_1.log.error(err);
        }
        return accErrors;
    }
    mkdirSync(dir) {
        try {
            fs.mkdirSync(dir);
        }
        catch (e) {
            if (e.code !== "EEXIST") {
                throw e;
            }
        }
    }
    getFileList(dir, fileList) {
        const files = fs.readdirSync(dir);
        fileList = fileList || [];
        files.forEach((file) => {
            if (fs.statSync(pathlib.join(dir, file)).isDirectory()) {
                fileList = this.getFileList(pathlib.join(dir, file), fileList);
            }
            else {
                fileList.push(pathlib.join(dir, file));
            }
        });
        return fileList;
    }
}
exports.XMsExampleExtractor = XMsExampleExtractor;
