// oav/dist/lib/swagger/fileLoader.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileLoader = void 0;
const path = __importStar(require("path"));
const promises_1 = require("fs/promises");
const openapi_tools_common_1 = require("@azure-tools/openapi-tools-common");
const inversify_1 = require("inversify");
const inversifyUtils_1 = require("../inversifyUtils");
const utils_1 = require("../util/utils");
const loader_1 = require("./loader");
let FileLoader = class FileLoader {
    constructor(opts) {
        this.opts = opts;
        this.preloadCache = new Map();
        (0, loader_1.setDefaultOpts)(opts, {
            checkUnderFileRoot: true,
        });
        if (this.opts.fileRoot) {
            this.opts.fileRoot = (0, utils_1.checkAndResolveGithubUrl)((0, openapi_tools_common_1.pathResolve)(this.opts.fileRoot));
        }
    }
    async load(filePath) {
        filePath = this.resolvePath(filePath);
        const preloadContent = this.preloadCache.get(filePath);
        if (preloadContent !== undefined) {
            return preloadContent;
        }
        return (0, openapi_tools_common_1.readFile)(filePath);
    }
    relativePath(filePath) {
        if (this.opts.fileRoot) {
            filePath = this.resolvePath(filePath);
            const url = (0, openapi_tools_common_1.urlParse)(filePath);
            if (url) {
                const rootUrl = (0, openapi_tools_common_1.urlParse)(this.opts.fileRoot);
                filePath = rootUrl === undefined ? filePath : path.relative(rootUrl.path, url.path);
            }
            else {
                filePath = path.relative(this.opts.fileRoot, filePath);
            }
        }
        return filePath;
    }
    resolvePath(filePath) {
        const url = (0, openapi_tools_common_1.urlParse)(filePath);
        if (url) {
            filePath = (0, utils_1.checkAndResolveGithubUrl)(filePath);
        }
        else if (!path.isAbsolute(filePath)) {
            filePath = this.opts.fileRoot
                ? (0, openapi_tools_common_1.pathJoin)(this.opts.fileRoot, filePath)
                : path.resolve(filePath);
        }
        if (this.opts.fileRoot &&
            this.opts.checkUnderFileRoot &&
            !filePath.startsWith(this.opts.fileRoot)) {
            throw new Error(`Try to load file "${filePath}" outside of root folder ${this.opts.fileRoot}`);
        }
        return filePath;
    }
    isUnderFileRoot(filePath) {
        if (!this.opts.fileRoot) {
            return true;
        }
        if (path.isAbsolute(filePath) && !filePath.startsWith(this.opts.fileRoot)) {
            return false;
        }
        filePath = (0, openapi_tools_common_1.pathJoin)(this.opts.fileRoot, filePath);
        return filePath.startsWith(this.opts.fileRoot);
    }
    async writeFile(filePath, content) {
        filePath = this.resolvePath(filePath);
        await (0, promises_1.mkdir)(path.dirname(filePath), { recursive: true });
        return (0, openapi_tools_common_1.asyncWriteFile)(filePath, content);
    }
    async appendFile(filePath, content) {
        filePath = this.resolvePath(filePath);
        await (0, promises_1.appendFile)(filePath, content);
    }
    preloadExtraFile(filePath, content) {
        filePath = this.resolvePath(filePath);
        this.preloadCache.set(filePath, content);
    }
};
exports.FileLoader = FileLoader;
exports.FileLoader = FileLoader = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(inversifyUtils_1.TYPES.opts)),
    __metadata("design:paramtypes", [Object])
], FileLoader);
