// oav/dist/lib/swagger/jsonLoader.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRefLike = exports.JsonLoader = exports.JsonLoaderRefError = exports.$id = void 0;
const path_1 = require("path");
const openapi_tools_common_1 = require("@azure-tools/openapi-tools-common");
const js_yaml_1 = require("js-yaml");
const json_pointer_1 = __importDefault(require("json-pointer"));
const inversify_1 = require("inversify");
const constants_1 = require("../util/constants");
const lazyBuilder_1 = require("../util/lazyBuilder");
const inversifyUtils_1 = require("../inversifyUtils");
const fileLoader_1 = require("./fileLoader");
const loader_1 = require("./loader");
exports.$id = "$id";
class JsonLoaderRefError extends Error {
    constructor(source) {
        super(`Failed to resolve ref for ${source.$ref}`);
        const info = (0, openapi_tools_common_1.getInfo)(source);
        if (info !== undefined) {
            const rootInfo = (0, openapi_tools_common_1.getRootObjectInfo)(info);
            this.position = info.position;
            this.url = rootInfo.url;
            this.ref = source.$ref;
        }
    }
}
exports.JsonLoaderRefError = JsonLoaderRefError;
let JsonLoader = class JsonLoader {
    constructor(opts, fileLoader) {
        this.opts = opts;
        this.fileLoader = fileLoader;
        this.mockNameMap = {};
        this.globalMockNameId = 0;
        this.loadedFiles = [];
        this.fileCache = new Map();
        this.loadFile = (0, lazyBuilder_1.getLazyBuilder)("resolved", async (cache) => {
            const fileString = await this.fileLoader.load(cache.filePath);
            if (this.opts.keepOriginalContent || cache.resolveRef) {
                // eslint-disable-next-line require-atomic-updates
                cache.originalContent = fileString;
            }
            let fileContent = this.parseFileContent(cache, fileString);
            // eslint-disable-next-line require-atomic-updates
            cache.resolved = fileContent;
            fileContent[exports.$id] = cache.mockName;
            if (cache.skipResolveRef !== true) {
                fileContent = await this.resolveRef(fileContent, ["$"], fileContent, cache.filePath, false);
            }
            this.loadedFiles.push(fileContent);
            return fileContent;
        });
        this.loadFileWithRefSiblings = (0, lazyBuilder_1.getLazyBuilder)("resolved", async (cache) => {
            const fileString = await this.fileLoader.load(cache.filePath);
            if (this.opts.keepOriginalContent || cache.resolveRef) {
                // eslint-disable-next-line require-atomic-updates
                cache.originalContent = fileString;
            }
            let fileContent = this.parseFileContent(cache, fileString);
            // eslint-disable-next-line require-atomic-updates
            cache.resolved = fileContent;
            fileContent[exports.$id] = cache.mockName;
            if (cache.skipResolveRef !== true) {
                fileContent = await this.resolveRef(fileContent, ["$"], fileContent, cache.filePath, false, true);
            }
            this.loadedFiles.push(fileContent);
            return fileContent;
        });
        (0, loader_1.setDefaultOpts)(opts, {
            useJsonParser: true,
            eraseDescription: true,
            eraseXmsExamples: true,
            transformRef: true,
            supportYaml: false,
        });
        this.skipResolveRefKeys = new Set(opts.skipResolveRefKeys);
    }
    parseFileContent(cache, fileString) {
        if (this.opts.supportYaml &&
            (cache.filePath.endsWith(".yaml") || cache.filePath.endsWith(".yml"))) {
            return (0, js_yaml_1.load)(fileString, {
                filename: cache.filePath,
                json: true,
            });
        }
        return this.opts.useJsonParser ? (0, openapi_tools_common_1.parseJson)(cache.filePath, fileString) : JSON.parse(fileString);
        // throw new Error(`Unknown file format while loading file ${cache.filePath}`);
    }
    async load(inputFilePath, skipResolveRef, keepRefSiblings) {
        const filePath = this.fileLoader.relativePath(inputFilePath);
        let cache = this.fileCache.get(filePath);
        if (cache === undefined) {
            cache = {
                filePath,
                mockName: this.getNextMockName(filePath),
            };
            this.fileCache.set(filePath, cache);
        }
        if (skipResolveRef !== undefined) {
            cache.skipResolveRef = skipResolveRef;
        }
        if (keepRefSiblings) {
            return this.loadFileWithRefSiblings(cache);
        }
        return this.loadFile(cache);
    }
    getFileContentFromCache(filePath) {
        if (this.fileCache !== undefined) {
            const cache = this.fileCache.get(filePath);
            return cache === null || cache === void 0 ? void 0 : cache.resolved;
        }
        return undefined;
    }
    async resolveFile(mockName) {
        const filePath = this.mockNameMap[mockName];
        let cache = this.fileCache.get(filePath);
        if (cache !== undefined) {
            return this.loadFile(cache);
        }
        // Fallback for load file outside our swagger context
        const contentString = await this.fileLoader.load(mockName);
        cache = {
            filePath: mockName,
            mockName,
        };
        const content = this.parseFileContent(cache, contentString);
        return content;
    }
    resolveRefObj(object) {
        let refObj = object;
        while ((0, exports.isRefLike)(refObj)) {
            const $ref = refObj.$ref;
            const idx = $ref.indexOf("#");
            const mockName = idx === -1 ? $ref : $ref.substr(0, idx);
            const refObjPath = idx === -1 ? "" : $ref.substr(idx + 1);
            const filePath = this.mockNameMap[mockName];
            const cache = this.fileCache.get(filePath);
            if (cache === undefined) {
                throw new Error(`cache not found for ${filePath}`);
            }
            refObj = json_pointer_1.default.get(cache.resolved, refObjPath);
            object.$ref = $ref;
        }
        return refObj;
    }
    resolveMockedFile(fileName) {
        let refObj;
        if (!!fileName && fileName.startsWith("_")) {
            const filePath = this.mockNameMap[fileName];
            const cache = this.fileCache.get(filePath);
            if (cache === undefined) {
                throw new Error(`cache not found for ${filePath} and mockName ${fileName}`);
            }
            refObj = json_pointer_1.default.get(cache.resolved, "");
        }
        return refObj;
    }
    getRealPath(mockName) {
        return this.mockNameMap[mockName];
    }
    async resolveRef(object, pathArr, rootObject, relativeFilePath, skipResolveChildRef, keepRefSiblings) {
        if ((0, exports.isRefLike)(object)) {
            const refObjResult = {};
            if (object.readOnly !== undefined) {
                refObjResult[constants_1.xmsReadonlyRef] = true;
            }
            const ref = object.$ref;
            const sp = ref.split("#");
            if (sp.length > 2) {
                throw new Error("ref format error multiple #");
            }
            const [refFilePath, refObjPath] = sp;
            if (refFilePath === "") {
                // Local reference
                if (!json_pointer_1.default.has(rootObject, refObjPath)) {
                    throw new JsonLoaderRefError(object);
                }
                const mockName = rootObject[exports.$id];
                if (keepRefSiblings) {
                    object.$ref = `${mockName}#${refObjPath}`;
                    return object;
                }
                refObjResult.$ref = `${mockName}#${refObjPath}`;
                return refObjResult;
            }
            const refObj = await this.load((0, openapi_tools_common_1.pathJoin)((0, path_1.dirname)(relativeFilePath), refFilePath), skipResolveChildRef);
            const refMockName = refObj[exports.$id];
            if (refObjPath !== undefined) {
                if (!json_pointer_1.default.has(refObj, refObjPath)) {
                    throw new JsonLoaderRefError(object);
                }
                if (keepRefSiblings) {
                    object.$ref = `${refMockName}#${refObjPath}`;
                    return object;
                }
                refObjResult.$ref = `${refMockName}#${refObjPath}`;
                return refObjResult;
            }
            else {
                if (keepRefSiblings) {
                    object.$ref = refMockName;
                    return object;
                }
                refObjResult.$ref = refMockName;
                return refObjResult;
            }
        }
        if (Array.isArray(object)) {
            for (let idx = 0; idx < object.length; ++idx) {
                const item = object[idx];
                if (typeof item === "object" && item !== null) {
                    const newRef = await this.resolveRef(item, pathArr.concat([idx.toString()]), rootObject, relativeFilePath, skipResolveChildRef);
                    if (newRef !== item) {
                        // eslint-disable-next-line require-atomic-updates
                        object[idx] = newRef;
                    }
                }
            }
        }
        else if (typeof object === "object" && object !== null) {
            const obj = object;
            if (this.opts.eraseDescription && typeof obj.description === "string") {
                delete obj.description;
            }
            if (this.opts.eraseXmsExamples && obj[constants_1.xmsExamples] !== undefined) {
                delete obj[constants_1.xmsExamples];
            }
            for (const key of Object.keys(obj)) {
                const item = obj[key];
                if (typeof item === "object" && item !== null) {
                    const newRef = await this.resolveRef(item, pathArr.concat([key]), rootObject, relativeFilePath, skipResolveChildRef || this.skipResolveRefKeys.has(key));
                    if (newRef !== item) {
                        obj[key] = newRef;
                    }
                }
            }
        }
        else {
            throw new Error("Invalid json");
        }
        return object;
    }
    getNextMockName(filePath) {
        const id = this.globalMockNameId++;
        const mockName = `_${id.toString(36)}`;
        this.mockNameMap[mockName] = filePath;
        return mockName;
    }
};
exports.JsonLoader = JsonLoader;
exports.JsonLoader = JsonLoader = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(inversifyUtils_1.TYPES.opts)),
    __metadata("design:paramtypes", [Object, fileLoader_1.FileLoader])
], JsonLoader);
const isRefLike = (obj) => typeof obj.$ref === "string";
exports.isRefLike = isRefLike;
