// oav/dist/lib/swagger/suppressionLoader.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isSuppressedInPath = exports.isSuppressed = exports.SuppressionLoader = void 0;
const path_1 = require("path");
const openapi_tools_common_1 = require("@azure-tools/openapi-tools-common");
const openapi_markdown_1 = require("@azure/openapi-markdown");
const jsonpath_plus_1 = require("jsonpath-plus");
const inversify_1 = require("inversify");
const logging_1 = require("../util/logging");
const inversifyUtils_1 = require("../inversifyUtils");
const fileLoader_1 = require("./fileLoader");
let SuppressionLoader = class SuppressionLoader {
    constructor(opts, fileLoader) {
        var _a;
        this.fileLoader = fileLoader;
        this.suppressionCache = new Map();
        this.suppressionToLoad = new Set((_a = opts === null || opts === void 0 ? void 0 : opts.loadSuppression) !== null && _a !== void 0 ? _a : []);
    }
    async load(spec) {
        if (this.suppressionToLoad.size === 0) {
            return;
        }
        const filePath = this.fileLoader.resolvePath(spec._filePath);
        const readmePath = await (0, openapi_markdown_1.findReadMe)(filePath);
        if (readmePath === undefined) {
            return;
        }
        try {
            let items = this.suppressionCache.get(readmePath);
            if (items === undefined) {
                items = await this.getSuppression(readmePath);
                this.suppressionCache.set(readmePath, items);
            }
            for (const item of items) {
                if (!matchFileFrom(filePath, item.from)) {
                    continue;
                }
                applySuppression(spec, item);
            }
        }
        catch (e) {
            const msg = `Error in loading suppression from readme:${readmePath}.\nDetails:${e.message}\n${e.stack}`;
            throw new Error(msg);
        }
    }
    async getSuppression(readmePath) {
        if (!this.fileLoader.isUnderFileRoot(readmePath)) {
            return [];
        }
        const fileContent = await this.fileLoader.load(readmePath);
        const cmd = (0, openapi_tools_common_1.parseMarkdown)(fileContent);
        const suppressionCodeBlock = (0, openapi_markdown_1.getCodeBlocksAndHeadings)(cmd.markDown).Suppression;
        if (suppressionCodeBlock === undefined) {
            return [];
        }
        const suppression = (0, openapi_markdown_1.getYamlFromNode)(suppressionCodeBlock);
        const items = suppression.directive;
        if (!Array.isArray(items)) {
            return [];
        }
        return items.filter((item) => this.suppressionToLoad.has(item.suppress));
    }
};
exports.SuppressionLoader = SuppressionLoader;
exports.SuppressionLoader = SuppressionLoader = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(inversifyUtils_1.TYPES.opts)),
    __metadata("design:paramtypes", [Object, fileLoader_1.FileLoader])
], SuppressionLoader);
const matchFileFrom = (filePath, from) => {
    if (from === undefined) {
        return true;
    }
    if (!Array.isArray(from) && typeof from === "string") {
        return endsWithPath(filePath, from);
    }
    for (const fromName of from) {
        if (endsWithPath(filePath, fromName)) {
            return true;
        }
    }
    return false;
};
const endsWithPath = (filePath, searchPath) => {
    const sep = filePath[filePath.length - searchPath.length - 1];
    return filePath.endsWith(searchPath) && (sep === path_1.sep || sep === "/");
};
const applySuppression = (spec, item) => {
    var _a, _b;
    for (const node of getNodesFromWhere(spec, item.where)) {
        const info = (_a = (0, openapi_tools_common_1.getInfo)(node)) === null || _a === void 0 ? void 0 : _a.position;
        if (info !== undefined) {
            if (info.directives === undefined) {
                info.directives = {};
            }
            const directives = info.directives;
            directives[item.suppress] = (_b = item["text-matches"]) !== null && _b !== void 0 ? _b : ".*";
        }
    }
};
const getNodesFromWhere = (spec, where) => {
    if (where === undefined) {
        return [spec];
    }
    if (typeof where === "string") {
        where = [where];
    }
    const output = [];
    for (const wh of where) {
        try {
            output.push(...(0, jsonpath_plus_1.JSONPath)({ path: wh, json: spec, resultType: "value" }));
        }
        catch (e) {
            logging_1.log.error(e);
        }
    }
    return output;
};
const isSuppressed = (node, code, message) => {
    const info = (0, openapi_tools_common_1.getInfo)(node);
    if (info === undefined) {
        return false;
    }
    const directives = info.position.directives;
    if (directives === undefined) {
        return false;
    }
    const messageRegex = directives[code];
    if (messageRegex === undefined) {
        return false;
    }
    if (messageRegex === ".*") {
        return true;
    }
    return new RegExp(messageRegex).test(message);
};
exports.isSuppressed = isSuppressed;
const getParentNode = (info) => {
    return info.isChild ? (0, openapi_tools_common_1.getInfo)(info.parent) : undefined;
};
const isSuppressedInPath = (node, code, message) => {
    let info = (0, openapi_tools_common_1.getInfo)(node);
    while (info !== undefined) {
        const directives = info.position.directives;
        if (directives === undefined) {
            info = getParentNode(info);
            continue;
        }
        const messageRegex = directives[code];
        if (messageRegex === undefined) {
            info = getParentNode(info);
            continue;
        }
        if (messageRegex === ".*") {
            return true;
        }
        if (new RegExp(messageRegex).test(message)) {
            return true;
        }
        info = getParentNode(info);
    }
    return false;
};
exports.isSuppressedInPath = isSuppressedInPath;
