// oav/dist/lib/swagger/swaggerLoader.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SwaggerLoader = void 0;
const path_1 = require("path");
const inversify_1 = require("inversify");
const inversifyUtils_1 = require("../inversifyUtils");
const traverseSwagger_1 = require("../transform/traverseSwagger");
const constants_1 = require("../util/constants");
const utils_1 = require("../util/utils");
const fileLoader_1 = require("./fileLoader");
const jsonLoader_1 = require("./jsonLoader");
const loader_1 = require("./loader");
const suppressionLoader_1 = require("./suppressionLoader");
let SwaggerLoader = class SwaggerLoader {
    constructor(opts, suppressionLoader, jsonLoader, siblingsJsonLoader, fileLoader) {
        this.opts = opts;
        this.suppressionLoader = suppressionLoader;
        this.jsonLoader = jsonLoader;
        this.siblingsJsonLoader = siblingsJsonLoader;
        this.fileLoader = fileLoader;
        (0, loader_1.setDefaultOpts)(opts, {
            setFilePath: true,
        });
    }
    getResolvedJsonLoader() {
        return this.siblingsJsonLoader;
    }
    // TODO reportError
    async load(specFilePath, keepRefSiblings) {
        const swaggerSpec = keepRefSiblings
            ? (await this.siblingsJsonLoader.load(specFilePath, false, true))
            : (await this.jsonLoader.load(specFilePath));
        if (this.opts.setFilePath) {
            const pathProvider = (0, utils_1.getProviderFromSpecPath)(this.fileLoader.resolvePath(specFilePath));
            swaggerSpec._filePath = this.fileLoader.relativePath(specFilePath);
            swaggerSpec._providerNamespace = pathProvider ? pathProvider.provider : "unknown";
        }
        await this.suppressionLoader.load(swaggerSpec);
        return swaggerSpec;
    }
    async updateSwaggerAndExamples(entries) {
        const swaggerPaths = new Set();
        const examplePaths = new Set();
        const entriesGroupByOperationId = new Map();
        for (const entry of entries) {
            swaggerPaths.add(entry.swaggerPath);
            if (examplePaths.has(entry.exampleFilePath)) {
                throw new Error(`Duplicated example path: ${entry.exampleFilePath}`);
            }
            examplePaths.add(entry.exampleFilePath);
            let entriesGroup = entriesGroupByOperationId.get(entry.operationId);
            if (entriesGroup === undefined) {
                entriesGroup = [];
                entriesGroupByOperationId.set(entry.operationId, entriesGroup);
            }
            entriesGroup.push(entry);
        }
        const toWait = [];
        for (const swaggerPath of swaggerPaths) {
            const fileContent = await this.fileLoader.load(swaggerPath);
            const swaggerSpec = JSON.parse(fileContent);
            (0, traverseSwagger_1.traverseSwagger)(swaggerSpec, {
                onOperation: (operation) => {
                    var _a;
                    const operationId = (_a = operation.operationId) !== null && _a !== void 0 ? _a : "";
                    const entriesGroup = entriesGroupByOperationId.get(operationId);
                    if (entriesGroup === undefined) {
                        return;
                    }
                    let examples = operation[constants_1.xmsExamples];
                    if (examples === undefined) {
                        examples = {};
                        operation[constants_1.xmsExamples] = examples;
                    }
                    for (const entry of entriesGroup) {
                        let path = (0, path_1.relative)((0, path_1.dirname)(swaggerPath), entry.exampleFilePath).replace(/\\/g, "/");
                        if (!path.startsWith(".")) {
                            path = `./${path}`;
                        }
                        examples[entry.exampleName] = {
                            $ref: path,
                        };
                        toWait.push(this.fileLoader.writeFile(entry.exampleFilePath, formatJson(entry.exampleContent)));
                    }
                },
            });
            toWait.push(this.fileLoader.writeFile(swaggerPath, formatJson(swaggerSpec)));
        }
        await Promise.all(toWait);
    }
};
exports.SwaggerLoader = SwaggerLoader;
exports.SwaggerLoader = SwaggerLoader = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(inversifyUtils_1.TYPES.opts)),
    __metadata("design:paramtypes", [Object, suppressionLoader_1.SuppressionLoader,
        jsonLoader_1.JsonLoader,
        jsonLoader_1.JsonLoader,
        fileLoader_1.FileLoader])
], SwaggerLoader);
const formatJson = (obj) => {
    return JSON.stringify(obj, undefined, 2);
};
