// oav/dist/lib/commands/extract-xmsexamples.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.builder = exports.describe = exports.command = void 0;
exports.handler = handler;
const cliSuppressExceptions_1 = require("../cliSuppressExceptions");
const logging_1 = require("../util/logging");
const validate = __importStar(require("../validate"));
exports.command = "extract-xmsexamples <spec-path> <recordings>";
exports.describe = "Extracts the x-ms-examples for a given swagger from the .NET session recordings and saves " +
    "them in a file.";
exports.builder = {
    d: {
        alias: "outDir",
        describe: "The output directory where the x-ms-examples files need to be stored. If not provided " +
            'then the output will be stored in a folder name "output" adjacent to the working directory.',
        string: true,
    },
    m: {
        alias: "matchApiVersion",
        describe: "Only generate examples if api-version matches.",
        boolean: true,
        default: true,
    },
};
async function handler(argv) {
    await (0, cliSuppressExceptions_1.cliSuppressExceptions)(async () => {
        logging_1.log.debug(argv.toString());
        const specPath = argv.specPath;
        const recordings = argv.recordings;
        const vOptions = {
            consoleLogLevel: argv.logLevel,
            logFilepath: argv.f,
            output: argv.outDir,
            matchApiVersion: argv.matchApiVersion,
        };
        await validate.extractXMsExamples(specPath, recordings, vOptions);
        return 0;
    });
}
