// oav/dist/lib/commands/traffic-converter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.builder = exports.describe = exports.command = void 0;
exports.handler = handler;
// node ./build/cli.js convert --directory ./input-example/ --out ./output-example/
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const url_1 = require("url");
const cliSuppressExceptions_1 = require("../cliSuppressExceptions");
const logging_1 = require("../util/logging");
let InputDirectory = "";
let OutputDirectory = "";
const CONTENT_TYPE_KEY = "Content-Type";
function requestBodyConversion(body, headers) {
    if (CONTENT_TYPE_KEY in headers) {
        let content = headers[CONTENT_TYPE_KEY];
        if (content.indexOf("application/json") > -1) {
            // if RequestBody is a unicode string, it need JSON.parse to format
            // eg. JSON.parse("{\u0022TableName\u0022: \u0022pytablesync70281ff8\u0022}") => {TableName: 'pytablesync70281ff8'}
            // else return directly includes Object or Array
            try {
                return JSON.parse(body);
            }
            catch (ex) {
                logging_1.log.error(ex);
                return body;
            }
        }
    }
    return body;
}
function requestUriConversion(uri, version) {
    const parsedUrl = new url_1.URL(uri);
    if (!parsedUrl.searchParams.get("api-version")) {
        parsedUrl.searchParams.set("api-version", version);
    }
    return parsedUrl.toString();
}
function processFile(file, inputJson) {
    if (inputJson.Entries !== undefined && inputJson.Entries.length > 0) {
        const filePrefix = file.substring(0, file.lastIndexOf("."));
        inputJson.Entries.forEach((entry, idx) => {
            const outFile = `${filePrefix}_${String(idx).padStart(4, "0")}.json`;
            const newEntry = {
                liveRequest: {},
                liveResponse: {},
            };
            // manipulate the request URI
            newEntry.liveRequest.url = requestUriConversion(entry.RequestUri, entry.RequestHeaders["x-ms-version"]);
            newEntry.liveRequest.headers = entry.RequestHeaders;
            // the request body is expected to be a JSON entry. Force that conversion if we can.
            newEntry.liveRequest.body = requestBodyConversion(entry.RequestBody, entry.RequestHeaders);
            newEntry.liveRequest.method = entry.RequestMethod;
            newEntry.liveResponse.body = entry.ResponseBody;
            // ensure string status code
            newEntry.liveResponse.statusCode = entry.StatusCode.toString();
            newEntry.liveResponse.headers = entry.ResponseHeaders;
            outputFile(outFile, newEntry);
        });
    }
}
function readFile(file) {
    const input_location = path.join(InputDirectory, file);
    fs.readFile(input_location, "utf8", (err, data) => {
        if (err) {
            throw err;
        }
        let convertedJson = {};
        try {
            // handle byte order mark
            convertedJson = JSON.parse(data.charCodeAt(0) === 0xfeff ? data.slice(1) : data);
        }
        catch (ex) {
            logging_1.log.error(ex);
            throw ex;
        }
        processFile(file, convertedJson);
    });
}
function outputFile(file, outputJson) {
    const data = JSON.stringify(outputJson, null, 4);
    const outputLocation = path.join(OutputDirectory, file);
    fs.writeFile(outputLocation, data, (err) => {
        if (err) {
            throw err;
        }
    });
}
function convert(directory, outDirectory) {
    logging_1.log.info(`Converting files in folder ${directory} ${outDirectory}`);
    logging_1.log.info(`Input Directory: ${directory}. Output Directory: ${outDirectory}`);
    let files = fs.readdirSync(directory);
    logging_1.log.info(`Operating on ${files.length} files.`);
    files.forEach((file) => {
        readFile(file);
    });
}
exports.command = "traffic-convert <input-dir> <output-dir>";
exports.describe = "Convert a folder files to traffic files";
exports.builder = {
    inputDir: {
        alias: "d",
        describe: "The targeted input directory.",
        string: true,
    },
    outputDir: {
        alias: "o",
        describe: "The targeted output directory.",
        string: true,
    },
};
async function handler(argv) {
    await (0, cliSuppressExceptions_1.cliSuppressExceptions)(async () => {
        logging_1.log.debug(argv.toString());
        InputDirectory = argv.inputDir;
        OutputDirectory = argv.outputDir;
        convert(InputDirectory, OutputDirectory);
        return 0;
    });
}
