// oav/dist/lib/commands/validate-traffic.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.builder = exports.describe = exports.command = void 0;
exports.handler = handler;
const cliSuppressExceptions_1 = require("../cliSuppressExceptions");
const logging_1 = require("../util/logging");
const validate = __importStar(require("../validate"));
exports.command = "validate-traffic <traffic-path> <spec-path>";
exports.describe = "Validate traffic payload against the spec.";
exports.builder = {
    trafficPath: {
        alias: "t",
        describe: "The recording payload path.",
        string: true,
    },
    specPath: {
        alias: "s",
        describe: "The targeted swagger spec path.",
        string: true,
    },
    package: {
        alias: "pkg",
        describe: "The target SDK package name",
        string: true,
        default: "azure-data-tables",
    },
    language: {
        alias: "lang",
        describe: "The target language of SDK",
        string: true,
        default: "Dotnet",
    },
    report: {
        alias: "r",
        describe: "path and file name for the report",
        string: true,
        default: "./SwaggerAccuracyReport.html",
    },
    overrideLinkInReport: {
        alias: "overridelink",
        describe: "override spec link and payload link in report with github url",
        boolean: true,
        default: false,
    },
    outputExceptionInReport: {
        alias: "oe",
        describe: "Whether rendering the runtime exceptions in the validation report",
        boolean: true,
        default: false,
    },
    specLinkPrefix: {
        alias: "slp",
        describe: "github specification link prefix",
        string: true,
        default: "https://github.com/Azure/azure-rest-api-specs/blob/main/",
    },
    payloadLinkPrefix: {
        alias: "plp",
        describe: "traffic payload link prefix",
        string: true,
        default: "https://github.com/scbedd/oav-traffic-converter/blob/main/sample-tables-input/",
    },
    jsonReport: {
        describe: "path and file name for json report",
        string: true,
    },
};
async function handler(argv) {
    await (0, cliSuppressExceptions_1.cliSuppressExceptions)(async () => {
        logging_1.log.debug(argv.toString());
        const specPath = argv.specPath;
        const trafficPath = argv.trafficPath;
        const vOptions = {
            consoleLogLevel: argv.logLevel,
            logFilepath: argv.f,
            pretty: argv.p,
            sdkPackage: argv.package,
            sdkLanguage: argv.language,
            reportPath: argv.report,
            overrideLinkInReport: argv.overrideLinkInReport,
            outputExceptionInReport: argv.outputExceptionInReport,
            specLinkPrefix: argv.specLinkPrefix,
            payloadLinkPrefix: argv.payloadLinkPrefix,
            jsonReportPath: argv.jsonReport,
        };
        const errors = await validate.validateTraffic(specPath, trafficPath, vOptions);
        return errors.length > 0 ? 1 : 0;
    });
}
