// @azure-tools/typespec-client-generator-core/dist/src/cache.js
import { compilerAssert, isService, isTemplateDeclaration, isTemplateDeclarationOrInstance, } from "@typespec/compiler";
import { unsafe_Realm } from "@typespec/compiler/experimental";
import { getVersionDependencies, getVersions } from "@typespec/versioning";
import { getClientLocation, getClientNameOverride, isInScope } from "./decorators.js";
import { clientKey, clientLocationKey, findServiceForOperation, getScopedDecoratorData, hasExplicitClientOrOperationGroup, listAllUserDefinedNamespaces, listScopedDecoratorData, omitOperation, operationGroupKey, removeVersionsLargerThanExplicitlySpecified, } from "./internal-utils.js";
import { reportDiagnostic } from "./lib.js";
import { getLibraryName } from "./public-utils.js";
/**
 * Create TCGC client types and operation group types and prepare the cache for clients, operation groups and operations.
 *
 * @param context TCGCContext
 */
export function prepareClientAndOperationCache(context) {
    // initialize the caches
    context.__rawClientsOperationGroupsCache = new Map();
    context.__operationToClientCache = new Map();
    context.__clientToOperationsCache = new Map();
    // create clients
    const clients = getOrCreateClients(context);
    // handle versioning with mutated types
    context.__packageVersions = new Map();
    context.__packageVersionEnum = new Map();
    if (clients.length === 1 && clients[0].services.length > 1) {
        // multi-service client
        const versionDependencies = getVersionDependencies(context.program, clients[0].type);
        for (const specificService of clients[0].services) {
            if (context.__packageVersions.has(specificService)) {
                continue;
            }
            const versions = getVersions(context.program, specificService)[1]?.getVersions();
            if (!versions) {
                context.__packageVersions.set(specificService, []);
                continue;
            }
            context.__packageVersionEnum.set(specificService, versions[0].enumMember.enum);
            const versionDependency = versionDependencies?.get(specificService);
            compilerAssert(versionDependency !== undefined && "name" in versionDependency, "Client with multiple services is missing version dependency declaration.");
            let end = false;
            context.__packageVersions.set(specificService, versions
                .map((version) => version.value)
                .filter((v) => {
                if (end)
                    return false;
                if (v === versionDependency.value)
                    end = true;
                return true;
            }));
        }
    }
    else if (clients.length > 0) {
        // single-service client
        const versions = getVersions(context.program, clients[0].services[0])[1]?.getVersions();
        if (!versions || versions.length === 0) {
            context.__packageVersions.set(clients[0].services[0], []);
        }
        else {
            context.__packageVersionEnum.set(clients[0].services[0], versions[0].enumMember.enum);
            removeVersionsLargerThanExplicitlySpecified(context, versions);
            const filteredVersions = versions.map((version) => version.value);
            context.__packageVersions.set(clients[0].services[0], filteredVersions);
        }
    }
    // Track operation group names to detect conflicts in multi-service scenarios
    const operationGroupNameMap = new Map();
    // Track merged operation groups and their original types for later operations processing
    const mergedOperationGroupTypes = new Map();
    // create operation groups for each client
    for (const client of clients) {
        const groups = [];
        if (client.services.length > 1) {
            // Multiple services case will auto-merge all the services and add their nested operation groups
            for (const specificService of client.services) {
                createFirstLevelOperationGroup(context, specificService, specificService);
            }
        }
        else {
            // Single service case needs to use the client type since it could contain customizations
            createFirstLevelOperationGroup(context, client.type, client.services[0]);
        }
        function createFirstLevelOperationGroup(context, type, service) {
            // iterate client's interfaces and namespaces to find operation groups
            if (type.kind === "Namespace") {
                for (const subItem of type.namespaces.values()) {
                    const og = createOperationGroup(context, subItem, `${client.name}`, service, client);
                    if (og && !handleMultipleServicesOperationGroupNameConflict(og)) {
                        groups.push(og);
                    }
                }
                for (const subItem of type.interfaces.values()) {
                    if (isTemplateDeclaration(subItem)) {
                        // Skip template interfaces
                        continue;
                    }
                    const og = createOperationGroup(context, subItem, `${client.name}`, service, client);
                    if (og && !handleMultipleServicesOperationGroupNameConflict(og)) {
                        groups.push(og);
                    }
                }
            }
        }
        function handleMultipleServicesOperationGroupNameConflict(og) {
            if (client.services.length > 1 && og.type) {
                // Track for conflict detection
                const ogName = getLibraryName(context, og.type);
                const existingOg = operationGroupNameMap.get(ogName);
                if (!existingOg) {
                    operationGroupNameMap.set(ogName, og);
                }
                else {
                    // Conflict detected, update the existing operation group to have multiple services
                    existingOg.services.push(og.services[0]);
                    existingOg.service = existingOg.services[0]; // eslint-disable-line @typescript-eslint/no-deprecated
                    existingOg.subOperationGroups.push(...og.subOperationGroups);
                    if (existingOg.type !== undefined) {
                        mergedOperationGroupTypes.set(existingOg, [existingOg.type]);
                        existingOg.type = undefined;
                    }
                    // Store the merged types for later operations processing
                    const types = mergedOperationGroupTypes.get(existingOg);
                    if (og.type) {
                        types.push(og.type);
                    }
                    return true;
                }
            }
            return false;
        }
        // build client's cache
        context.__rawClientsOperationGroupsCache.set(client.type, client);
        client.subOperationGroups = groups;
        context.__clientToOperationsCache.set(client, []);
    }
    // create operation group for `@clientLocation` of  string value
    // if no explicit `@client` or `@operationGroup`
    if (!hasExplicitClientOrOperationGroup(context)) {
        const newOperationGroupWithServices = new Map();
        listScopedDecoratorData(context, clientLocationKey).forEach((v, k) => {
            // only deal with mutated types or without mutation
            if ((!context.__mutatedRealm && !unsafe_Realm.realmForType.has(k)) ||
                (context.__mutatedRealm && context.__mutatedRealm.hasType(k))) {
                // If the target operation group already exists, handle the multiple services case
                if (typeof v === "string") {
                    // Check if an operation group with this name already exists, only check first level og for string target
                    const existingOg = clients[0].subOperationGroups.find((og) => og.type && getLibraryName(context, og.type) === v);
                    const operationService = clients[0].services.length > 1
                        ? findServiceForOperation(clients[0].services, k)
                        : clients[0].services[0];
                    if (existingOg) {
                        // Operation group already exists - check if moving this operation would create a multi-service situation
                        // Check if the existing operation group's service matches the operation's service
                        if (!existingOg.services.includes(operationService)) {
                            // This would create a multi-service operation group - merge the services
                            existingOg.services.push(operationService);
                            existingOg.service = existingOg.services[0]; // eslint-disable-line @typescript-eslint/no-deprecated
                        }
                        // Operation will be moved to this existing operation group during operations processing
                        context.__rawClientsOperationGroupsCache.set(v, existingOg);
                        return;
                    }
                    if (newOperationGroupWithServices.has(v)) {
                        // Add the service to the list if it's not already there
                        const services = newOperationGroupWithServices.get(v);
                        if (!services.includes(operationService)) {
                            services.push(operationService);
                        }
                    }
                    else {
                        newOperationGroupWithServices.set(v, [operationService]);
                    }
                }
            }
        });
        if (newOperationGroupWithServices.size > 0) {
            newOperationGroupWithServices.forEach((services, ogName) => {
                const og = {
                    kind: "SdkOperationGroup",
                    groupPath: `${clients[0].name}.${ogName}`,
                    service: services[0],
                    services,
                    subOperationGroups: [],
                    parent: clients[0],
                };
                context.__rawClientsOperationGroupsCache.set(ogName, og);
                clients[0].subOperationGroups.push(og);
                context.__clientToOperationsCache.set(og, []);
            });
        }
    }
    // iterate all clients and operation groups and build a map of operations
    const queue = [...clients];
    while (queue.length > 0) {
        const group = queue.shift();
        // operations directly under the group
        const operations = [];
        // Check if this is a merged operation group (has multiple services but still has a type)
        const mergedTypes = group.kind === "SdkOperationGroup" ? mergedOperationGroupTypes.get(group) : undefined;
        if (group.kind === "SdkClient" && group.services.length > 1) {
            // multi-service client
            operations.push(...group.services.flatMap((service) => [...service.operations.values()]));
        }
        else if (mergedTypes) {
            // multi-service operation group
            for (const type of mergedTypes) {
                operations.push(...type.operations.values());
            }
        }
        else if (group.type) {
            // single-service client or operation group
            operations.push(...group.type.operations.values());
        }
        // when there is explicitly `@operationGroup` or `@client`
        // operations under namespace or interface that are not decorated with `@operationGroup` or `@client`
        // should be placed in the first accessor client or operation group
        if (group.type?.kind === "Namespace" && hasExplicitClientOrOperationGroup(context)) {
            const innerQueue = [group.type];
            while (innerQueue.length > 0) {
                const ns = innerQueue.shift();
                for (const subNs of ns.namespaces.values()) {
                    if (!context.__rawClientsOperationGroupsCache.has(subNs)) {
                        operations.push(...subNs.operations.values());
                        innerQueue.push(subNs);
                    }
                }
                for (const iface of ns.interfaces.values()) {
                    if (!context.__rawClientsOperationGroupsCache.has(iface)) {
                        operations.push(...iface.operations.values());
                    }
                }
            }
        }
        // add operations
        for (const op of operations) {
            // skip operations that are not in scope
            if (!isInScope(context, op)) {
                continue;
            }
            // skip templated operations, omit operations (has override decorator)
            if (!isTemplateDeclarationOrInstance(op) &&
                !context.program.stateMap(omitOperation).get(op)) {
                let pushGroup = group;
                const clientLocation = getClientLocation(context, op);
                if (clientLocation) {
                    // operation with `@clientLocation` decorator is placed in another operation group
                    if (context.__rawClientsOperationGroupsCache.has(clientLocation)) {
                        pushGroup = context.__rawClientsOperationGroupsCache.get(clientLocation);
                    }
                    else {
                        reportDiagnostic(context.program, {
                            code: "client-location-wrong-type",
                            target: op,
                        });
                    }
                }
                context.__clientToOperationsCache.get(pushGroup).push(op);
                context.__operationToClientCache.set(op, pushGroup);
            }
        }
        if (group.type)
            queue.push(...group.subOperationGroups);
    }
    // omit empty client or operation groups
    if (!hasExplicitClientOrOperationGroup(context)) {
        const removeEmptyGroups = (group) => {
            // recursively check and remove empty sub-operation groups
            group.subOperationGroups = group.subOperationGroups.filter((subGroup) => {
                const keep = removeEmptyGroups(subGroup);
                if (!keep) {
                    context.__rawClientsOperationGroupsCache.delete(subGroup.type);
                }
                return keep;
            });
            // check if the group has operations or non-empty sub-operation groups
            const hasOperations = context.__clientToOperationsCache.get(group).length > 0;
            const hasSubGroups = group.subOperationGroups?.length > 0;
            return hasOperations || hasSubGroups;
        };
        // start from the top-level clients and remove empty groups
        for (const client of clients) {
            const keepClient = removeEmptyGroups(client);
            if (!keepClient) {
                context.__rawClientsOperationGroupsCache.delete(client.type);
                context.__clientToOperationsCache.delete(client);
            }
        }
    }
}
/**
 * Get or create the TCGC clients.
 * If user has explicitly defined `@client` then we will use those clients.
 * If user has not defined any `@client` then we will create a client for the first service namespace.
 *
 * @param context TCGCContext
 * @returns
 */
function getOrCreateClients(context) {
    const namespaces = listAllUserDefinedNamespaces(context);
    const explicitClients = [];
    for (const ns of namespaces) {
        if (getScopedDecoratorData(context, clientKey, ns)) {
            explicitClients.push(getScopedDecoratorData(context, clientKey, ns));
        }
        for (const i of ns.interfaces.values()) {
            if (getScopedDecoratorData(context, clientKey, i)) {
                explicitClients.push(getScopedDecoratorData(context, clientKey, i));
            }
        }
    }
    if (explicitClients.length > 0) {
        if (explicitClients.some((client) => isArm(client.services))) {
            context.arm = true;
        }
        return explicitClients;
    }
    // if there is no explicit client, we will treat the first namespace with service decorator as client
    const serviceNamespaces = namespaces.filter((ns) => isService(context.program, ns));
    if (serviceNamespaces.length >= 1) {
        const service = serviceNamespaces.shift();
        serviceNamespaces.map((ns) => {
            reportDiagnostic(context.program, {
                code: "multiple-services",
                target: ns,
            });
        });
        let originalName = service.name;
        const clientNameOverride = getClientNameOverride(context, service);
        if (clientNameOverride) {
            originalName = clientNameOverride;
        }
        else {
            originalName = service.name;
        }
        const clientName = originalName.endsWith("Client") ? originalName : `${originalName}Client`;
        context.arm = isArm(service);
        return [
            {
                kind: "SdkClient",
                name: clientName,
                service: service,
                services: [service],
                type: service,
                subOperationGroups: [],
            },
        ];
    }
    return [];
}
/**
 * Create a TCGC operation group for the given type.
 * This function will also iterate through the type's namespaces and interfaces to build nested operation groups.
 *
 * @param context TCGCContext
 * @param type
 * @returns
 */
function createOperationGroup(context, type, groupPathPrefix, service, parent) {
    let operationGroup;
    if (hasExplicitClientOrOperationGroup(context)) {
        operationGroup = getScopedDecoratorData(context, operationGroupKey, type);
        if (operationGroup) {
            operationGroup.groupPath = `${groupPathPrefix}.${getLibraryName(context, type)}`;
            operationGroup.service = service; // eslint-disable-line @typescript-eslint/no-deprecated
            operationGroup.services = [service];
            operationGroup.subOperationGroups = [];
            operationGroup.parent = parent;
            if (type.kind === "Namespace") {
                operationGroup.subOperationGroups =
                    buildHierarchyOfOperationGroups(context, type, operationGroup.groupPath, service, operationGroup) ?? [];
            }
        }
    }
    else {
        // if there is no explicit client, we will treat non-client namespaces and all interfaces as operation group
        if (type.kind !== "Interface" || !isTemplateDeclaration(type)) {
            operationGroup = {
                kind: "SdkOperationGroup",
                type,
                groupPath: `${groupPathPrefix}.${getLibraryName(context, type)}`,
                service: service,
                services: [service],
                subOperationGroups: [],
                parent,
            };
        }
        if (operationGroup && type.kind === "Namespace") {
            operationGroup.subOperationGroups =
                buildHierarchyOfOperationGroups(context, type, operationGroup.groupPath, service, operationGroup) ?? [];
        }
    }
    // build operation group's cache
    if (operationGroup) {
        context.__rawClientsOperationGroupsCache.set(operationGroup.type, operationGroup);
        context.__clientToOperationsCache.set(operationGroup, []);
    }
    return operationGroup;
}
function buildHierarchyOfOperationGroups(context, type, groupPathPrefix, service, parent) {
    // build hierarchy of operation group
    const subOperationGroups = [];
    type.namespaces.forEach((ns) => {
        const subOperationGroup = createOperationGroup(context, ns, groupPathPrefix, service, parent);
        if (subOperationGroup) {
            subOperationGroups.push(subOperationGroup);
        }
    });
    type.interfaces.forEach((i) => {
        const subOperationGroup = createOperationGroup(context, i, groupPathPrefix, service, parent);
        if (subOperationGroup) {
            subOperationGroups.push(subOperationGroup);
        }
    });
    if (subOperationGroups.length > 0) {
        return subOperationGroups;
    }
    return undefined;
}
function isArm(service) {
    if (Array.isArray(service)) {
        return service.some((s) => isArm(s));
    }
    return service.decorators.some((decorator) => decorator.decorator.name === "$armProviderNamespace");
}
