// @azure-tools/typespec-client-generator-core/dist/src/context.js
import { createDiagnosticCollector, emitFile, getRelativePathFromDirectory, normalizePath, resolvePath, } from "@typespec/compiler";
import { isAbsolute } from "path";
import { stringify } from "yaml";
import { prepareClientAndOperationCache } from "./cache.js";
import { defaultDecoratorsAllowList } from "./configs.js";
import { handleClientExamples } from "./example.js";
import { UsageFlags, } from "./interfaces.js";
import { handleVersioningMutationForGlobalNamespace, parseEmitterName, } from "./internal-utils.js";
import { createDiagnostic } from "./lib.js";
import { createSdkPackage } from "./package.js";
export function createTCGCContext(program, emitterName, options) {
    const diagnostics = createDiagnosticCollector();
    return {
        program,
        diagnostics: diagnostics.diagnostics,
        emitterName: diagnostics.pipe(parseEmitterName(program, emitterName ?? program.emitters[0]?.metadata?.name)),
        previewStringRegex: /-preview$/,
        disableUsageAccessPropagationToBase: false,
        generateProtocolMethods: true,
        generateConvenienceMethods: true,
        __referencedTypeCache: new Map(),
        __arrayDictionaryCache: new Map(),
        __methodParameterCache: new Map(),
        __modelPropertyCache: new Map(),
        __responseHeaderCache: new Map(),
        __generatedNames: new Map(),
        __httpOperationCache: new Map(),
        __clientParametersCache: new Map(),
        __tspTypeToApiVersions: new Map(),
        __clientApiVersionDefaultValueCache: new Map(),
        __httpOperationExamples: new Map(),
        __pagedResultSet: new Set(),
        getMutatedGlobalNamespace() {
            if (options?.mutateNamespace === false) {
                // If we are not mutating the global namespace, return the original global namespace type.
                return program.getGlobalNamespaceType();
            }
            if (!this.__mutatedGlobalNamespace) {
                this.__mutatedGlobalNamespace = handleVersioningMutationForGlobalNamespace(this);
            }
            return this.__mutatedGlobalNamespace;
        },
        getApiVersionsForType(type) {
            return this.__tspTypeToApiVersions.get(type) ?? [];
        },
        setApiVersionsForType(type, apiVersions) {
            const existingApiVersions = this.__tspTypeToApiVersions.get(type) ?? [];
            const mergedApiVersions = [...existingApiVersions];
            for (const apiVersion of apiVersions) {
                if (!mergedApiVersions.includes(apiVersion)) {
                    mergedApiVersions.push(apiVersion);
                }
            }
            this.__tspTypeToApiVersions.set(type, mergedApiVersions);
        },
        getPackageVersions() {
            if (!this.__packageVersions) {
                prepareClientAndOperationCache(this);
            }
            return this.__packageVersions;
        },
        getPackageVersionEnum() {
            if (!this.__packageVersionEnum) {
                prepareClientAndOperationCache(this);
            }
            return this.__packageVersionEnum;
        },
        getClients() {
            if (!this.__rawClientsOperationGroupsCache) {
                prepareClientAndOperationCache(this);
            }
            return Array.from(this.__rawClientsOperationGroupsCache.values()).filter((item) => item.kind === "SdkClient");
        },
        getClientOrOperationGroup(type) {
            if (!this.__rawClientsOperationGroupsCache) {
                prepareClientAndOperationCache(this);
            }
            return this.__rawClientsOperationGroupsCache.get(type);
        },
        getOperationsForClient(client) {
            if (!this.__clientToOperationsCache) {
                prepareClientAndOperationCache(this);
            }
            return this.__clientToOperationsCache.get(client);
        },
        getClientForOperation(operation) {
            if (!this.__operationToClientCache) {
                prepareClientAndOperationCache(this);
            }
            return this.__operationToClientCache.get(operation);
        },
    };
}
export async function createSdkContext(context, emitterName, options) {
    const diagnostics = createDiagnosticCollector();
    const tcgcContext = createTCGCContext(context.program, emitterName ?? context.options["emitter-name"]);
    const generateProtocolMethods = context.options["generate-protocol-methods"] ?? tcgcContext.generateProtocolMethods;
    const generateConvenienceMethods = context.options["generate-convenience-methods"] ?? tcgcContext.generateConvenienceMethods;
    const sdkContext = {
        ...tcgcContext,
        emitContext: context,
        sdkPackage: undefined,
        generateProtocolMethods: generateProtocolMethods,
        generateConvenienceMethods: generateConvenienceMethods,
        namespaceFlag: context.options["namespace"],
        apiVersion: context.options["api-version"],
        license: context.options["license"],
        decoratorsAllowList: [...defaultDecoratorsAllowList, ...(options?.additionalDecorators ?? [])],
        previewStringRegex: options?.versioning?.previewStringRegex || tcgcContext.previewStringRegex,
        disableUsageAccessPropagationToBase: options?.disableUsageAccessPropagationToBase ?? false,
        flattenUnionAsEnum: options?.flattenUnionAsEnum ?? true,
        enableLegacyHierarchyBuilding: options?.enableLegacyHierarchyBuilding ?? true,
    };
    if (context.options["examples-dir"]) {
        const normalizeExamplesDir = normalizePath(context.options["examples-dir"]);
        if (isAbsolute(normalizeExamplesDir)) {
            sdkContext.examplesDir = getRelativePathFromDirectory(context.program.projectRoot, normalizeExamplesDir, false);
        }
        else {
            sdkContext.examplesDir = normalizeExamplesDir;
        }
    }
    sdkContext.sdkPackage = diagnostics.pipe(createSdkPackage(sdkContext));
    for (const client of sdkContext.sdkPackage.clients) {
        diagnostics.pipe(await handleClientExamples(sdkContext, client));
    }
    // Validate duplicate names within each type kind in each namespace (cross-kind duplicates are allowed).
    diagnostics.pipe(validateNamesUnderNamespaces(sdkContext));
    sdkContext.diagnostics = [...sdkContext.diagnostics, ...diagnostics.diagnostics];
    if (options?.exportTCGCoutput) {
        await exportTCGCOutput(sdkContext);
    }
    return sdkContext;
}
function validateNamesUnderNamespaces(context) {
    const diagnostics = createDiagnosticCollector();
    const validateItems = (namespaceItems) => {
        const seenNames = new Set();
        for (const item of namespaceItems) {
            if (seenNames.has(item.name)) {
                diagnostics.add(createDiagnostic({
                    code: "duplicate-client-name",
                    format: { name: item.name, scope: context.emitterName },
                    target: item.__raw,
                }));
            }
            else {
                seenNames.add(item.name);
            }
        }
    };
    const validateNamespace = (namespace) => {
        validateItems(namespace.models);
        validateItems(namespace.enums.filter((e) => (e.usage & UsageFlags.ApiVersionEnum) === 0));
        validateItems(namespace.unions.filter((u) => u.kind === "union"));
        for (const nestedNamespace of namespace.namespaces) {
            validateNamespace(nestedNamespace);
        }
    };
    for (const namespace of context.sdkPackage.namespaces) {
        validateNamespace(namespace);
    }
    return diagnostics.wrap(undefined);
}
async function exportTCGCOutput(context) {
    await emitFile(context.program, {
        path: resolvePath(context.emitContext.emitterOutputDir, "tcgc-output.yaml"),
        content: stringify(context.sdkPackage, (k, v) => {
            if (typeof k === "string" && k.startsWith("__")) {
                return undefined; // skip keys starting with "__" from the output
            }
            if (k === "scheme") {
                const { model, ...rest } = v;
                return rest; // remove credential schema's model property
            }
            if (k === "rawExample") {
                return undefined; // remove raw example
            }
            return v;
        }, { lineWidth: 0 }),
    });
}
export async function $onEmit(context) {
    if (!context.program.compilerOptions.noEmit) {
        const sdkContext = await createSdkContext(context, undefined, { exportTCGCoutput: true });
        context.program.reportDiagnostics(sdkContext.diagnostics);
    }
}
