// @azure-tools/typespec-client-generator-core/dist/src/decorators.js
import { getLroMetadata } from "@azure-tools/typespec-azure-core";
import { compilerAssert, getDiscriminator, getNamespaceFullName, ignoreDiagnostics, isErrorModel, isList, isNumeric, isService, isTemplateDeclaration, } from "@typespec/compiler";
import { SyntaxKind } from "@typespec/compiler/ast";
import { $ } from "@typespec/compiler/typekit";
import { getAuthentication, getHttpOperation, getServers, isBody, isBodyRoot, } from "@typespec/http";
import { $useDependency, getVersions } from "@typespec/versioning";
import { UsageFlags, } from "./interfaces.js";
import { AllScopes, clientKey, clientLocationKey, clientNameKey, clientNamespaceKey, compareModelProperties, findEntriesWithTarget, findRootSourceProperty, getScopedDecoratorData, hasExplicitClientOrOperationGroup, isSameAuth, isSameServers, listAllUserDefinedNamespaces, negationScopesKey, omitOperation, operationGroupKey, overrideKey, parseScopes, scopeKey, } from "./internal-utils.js";
import { createStateSymbol, reportDiagnostic } from "./lib.js";
import { getSdkEnum, getSdkModel, getSdkUnion } from "./types.js";
export const namespace = "Azure.ClientGenerator.Core";
function setScopedDecoratorData(context, decorator, key, target, value, scope) {
    const targetEntry = context.program.stateMap(key).get(target);
    // if no scope specified, then set with the new value
    if (!scope) {
        if (targetEntry && targetEntry[AllScopes]) {
            targetEntry[AllScopes] = value;
        }
        else {
            const newObject = Object.fromEntries([[AllScopes, value]]);
            context.program
                .stateMap(key)
                .set(target, !targetEntry ? newObject : { ...targetEntry, ...newObject });
        }
        return;
    }
    const [negationScopes, scopes] = parseScopes(scope);
    if (negationScopes !== undefined && negationScopes.length > 0) {
        // override the previous value for negation scopes
        const newObject = scopes !== undefined && scopes.length > 0
            ? Object.fromEntries([AllScopes, ...scopes].map((scope) => [scope, value]))
            : Object.fromEntries([[AllScopes, value]]);
        newObject[negationScopesKey] = negationScopes;
        context.program.stateMap(key).set(target, newObject);
        // if a scope exists in the target entry and it overlaps with the negation scope, it means negation scope doesn't override it
        if (targetEntry !== undefined) {
            const existingScopes = Object.getOwnPropertyNames(targetEntry);
            const intersections = existingScopes.filter((x) => negationScopes.includes(x));
            if (intersections !== undefined && intersections.length > 0) {
                for (const scopeToKeep of intersections) {
                    newObject[scopeToKeep] = targetEntry[scopeToKeep];
                }
            }
        }
    }
    else if (scopes !== undefined && scopes.length > 0) {
        // for normal scopes, add them incrementally
        const newObject = Object.fromEntries(scopes.map((scope) => [scope, value]));
        context.program
            .stateMap(key)
            .set(target, !targetEntry ? newObject : { ...targetEntry, ...newObject });
    }
}
export const $client = (context, target, options, scope) => {
    if (context.decoratorTarget.kind === SyntaxKind.AugmentDecoratorStatement) {
        reportDiagnostic(context.program, {
            code: "wrong-client-decorator",
            target: context.decoratorTarget,
        });
        return;
    }
    const explicitName = options?.kind === "Model" ? options?.properties.get("name")?.type : undefined;
    const name = explicitName?.kind === "String" ? explicitName.value : target.name;
    let services;
    const serviceConfig = options?.kind === "Model" ? options?.properties.get("service")?.type : undefined;
    if (serviceConfig?.kind === "Namespace") {
        services = [serviceConfig];
    }
    else if (serviceConfig?.kind === "Tuple" &&
        serviceConfig.values.every((v) => v.kind === "Namespace")) {
        if (target.kind === "Interface") {
            reportDiagnostic(context.program, {
                code: "invalid-client-service-multiple",
                target: context.decoratorTarget,
            });
            return;
        }
        services = serviceConfig.values;
        // validate all services has same server definition
        let servers = undefined;
        let auth = undefined;
        let isSame = true;
        for (const svc of services) {
            const currentServers = getServers(context.program, svc);
            if (currentServers === undefined)
                continue;
            if (servers === undefined) {
                servers = currentServers;
            }
            else {
                isSame = isSameServers(servers, currentServers);
                if (!isSame) {
                    break;
                }
            }
        }
        for (const svc of services) {
            const currentAuth = getAuthentication(context.program, svc);
            if (currentAuth === undefined)
                continue;
            if (auth === undefined) {
                auth = currentAuth;
            }
            else {
                isSame = isSameAuth(auth, currentAuth);
                if (!isSame) {
                    break;
                }
            }
        }
        if (!isSame) {
            reportDiagnostic(context.program, {
                code: "inconsistent-multiple-service",
                target: context.decoratorTarget,
            });
            return;
        }
        // no explicit versioning dependency
        if (!target.decorators.some((d) => d.definition?.name === "@useDependency" &&
            getNamespaceFullName(d.definition?.namespace) === "TypeSpec.Versioning")) {
            const versionRecords = [];
            // collect the latest version enum member from each service
            for (const svc of services) {
                const versions = getVersions(context.program, svc)[1]?.getVersions();
                if (versions && versions.length > 0) {
                    versionRecords.push(versions[versions.length - 1].enumMember);
                }
            }
            // set the versioning dependency
            if (versionRecords.length > 0) {
                context.call($useDependency, target, ...versionRecords);
            }
        }
    }
    else {
        const service = findClientService(context.program, target);
        if (service === undefined) {
            reportDiagnostic(context.program, {
                code: "client-service",
                format: { name },
                target: context.decoratorTarget,
            });
            return;
        }
        services = [service];
    }
    const client = {
        kind: "SdkClient",
        name,
        service: services.length === 1 ? services[0] : services,
        services,
        type: target,
        subOperationGroups: [],
    };
    setScopedDecoratorData(context, $client, clientKey, target, client, scope);
};
function judgeService(program, type) {
    return (isService(program, type) ||
        type.decorators.some((d) => d.definition?.name === "@service" && d.definition?.namespace.name === "TypeSpec"));
}
function findClientService(program, client) {
    let current = client;
    while (current) {
        if (judgeService(program, current)) {
            return current;
        }
        current = current.namespace;
    }
    return undefined;
}
/**
 * Return the client object for the given namespace or interface, or undefined if the given namespace or interface is not a client.
 *
 * @param context TCGCContext
 * @param type Type to check
 * @returns Client or undefined
 */
export function getClient(context, type) {
    for (const client of listClients(context)) {
        if (client.type === type) {
            return client;
        }
    }
    return undefined;
}
/**
 * List all the clients.
 *
 * @param context TCGCContext
 * @returns Array of clients
 */
export function listClients(context) {
    return context.getClients();
}
export const $operationGroup = (context, target, scope) => {
    if (context.decoratorTarget.kind === SyntaxKind.AugmentDecoratorStatement) {
        reportDiagnostic(context.program, {
            code: "wrong-client-decorator",
            target: context.decoratorTarget,
        });
        return;
    }
    setScopedDecoratorData(context, $operationGroup, operationGroupKey, target, {
        kind: "SdkOperationGroup",
        type: target,
    }, scope);
};
/**
 * Check a namespace or interface is an operation group.
 * @param context TCGCContext
 * @param type Type to check
 * @returns boolean
 */
export function isOperationGroup(context, type) {
    if (hasExplicitClientOrOperationGroup(context)) {
        return getScopedDecoratorData(context, operationGroupKey, type) !== undefined;
    }
    // if there is no explicit client, we will treat non-client namespaces and all interfaces as operation group
    if (type.kind === "Interface" && !isTemplateDeclaration(type)) {
        return true;
    }
    if (type.kind === "Namespace" &&
        !type.decorators.some((d) => d.definition?.name === "@service" && d.definition?.namespace.name === "TypeSpec")) {
        return true;
    }
    return false;
}
/**
 * Return the operation group object for the given namespace or interface or undefined is not an operation group.
 * @param context TCGCContext
 * @param type Type to check
 * @returns Operation group or undefined.
 */
export function getOperationGroup(context, type) {
    const operationGroup = context.getClientOrOperationGroup(type);
    return operationGroup?.kind === "SdkOperationGroup" ? operationGroup : undefined;
}
/**
 * List all the operation groups inside a client or an operation group. If ignoreHierarchy is true, the result will include all nested operation groups.
 *
 * @param context TCGCContext
 * @param group Client or operation group to list operation groups
 * @param ignoreHierarchy Whether to get all nested operation groups
 * @returns
 */
export function listOperationGroups(context, group, ignoreHierarchy = false) {
    if (!ignoreHierarchy)
        return group.subOperationGroups;
    const groups = [...group.subOperationGroups];
    let current = 0;
    while (current < groups.length) {
        const operationGroup = groups[current];
        if (operationGroup.subOperationGroups) {
            groups.push(...operationGroup.subOperationGroups);
        }
        current++;
    }
    return groups;
}
/**
 * List operations inside a client or an operation group. If ignoreHierarchy is true, the result will include all nested operations.
 * @param program TCGCContext
 * @param group Client or operation group to list operations
 * @param ignoreHierarchy Whether to get all nested operations
 * @returns
 */
export function listOperationsInOperationGroup(context, group, ignoreHierarchy = false) {
    if (!ignoreHierarchy)
        return context.getOperationsForClient(group);
    const groups = [...group.subOperationGroups];
    const operations = [...context.getOperationsForClient(group)];
    while (groups.length > 0) {
        const operationGroup = groups.shift();
        if (operationGroup.subOperationGroups) {
            groups.push(...operationGroup.subOperationGroups);
        }
        operations.push(...context.getOperationsForClient(operationGroup));
    }
    return operations;
}
const protocolAPIKey = createStateSymbol("protocolAPI");
export const $protocolAPI = (context, entity, value, scope) => {
    setScopedDecoratorData(context, $protocolAPI, protocolAPIKey, entity, value, scope);
};
const convenientAPIKey = createStateSymbol("convenientAPI");
export const $convenientAPI = (context, entity, value, scope) => {
    setScopedDecoratorData(context, $convenientAPI, convenientAPIKey, entity, value, scope);
};
function getConvenientOrProtocolValue(context, key, entity) {
    // First check if the operation itself has the decorator
    const value = getScopedDecoratorData(context, key, entity);
    if (value !== undefined) {
        return value;
    }
    // Check the parent interface if the operation is in an interface
    if (entity.interface) {
        const interfaceValue = getScopedDecoratorData(context, key, entity.interface);
        if (interfaceValue !== undefined) {
            return interfaceValue;
        }
    }
    // Check the parent namespace hierarchy
    let currentNamespace = entity.namespace;
    while (currentNamespace) {
        const namespaceValue = getScopedDecoratorData(context, key, currentNamespace);
        if (namespaceValue !== undefined) {
            return namespaceValue;
        }
        currentNamespace = currentNamespace.namespace;
    }
    return undefined;
}
export function shouldGenerateProtocol(context, entity) {
    const value = getConvenientOrProtocolValue(context, protocolAPIKey, entity);
    return value ?? Boolean(context.generateProtocolMethods);
}
export function shouldGenerateConvenient(context, entity) {
    const value = getConvenientOrProtocolValue(context, convenientAPIKey, entity);
    return value ?? Boolean(context.generateConvenienceMethods);
}
const usageKey = createStateSymbol("usage");
export const $usage = (context, entity, value, scope) => {
    const isValidValue = (value) => {
        // Allow the new usage values: input(2), output(4), json(256), xml(512)
        return (value === UsageFlags.Input ||
            value === UsageFlags.Output ||
            value === UsageFlags.Json ||
            value === UsageFlags.Xml);
    };
    let newUsage = 0;
    if (value.kind === "EnumMember") {
        if (typeof value.value === "number" && isValidValue(value.value)) {
            newUsage = value.value;
        }
        else {
            reportDiagnostic(context.program, {
                code: "invalid-usage",
                format: {},
                target: entity,
            });
            return;
        }
    }
    else {
        for (const variant of value.variants.values()) {
            if (variant.type.kind === "EnumMember" && typeof variant.type.value === "number") {
                if (isValidValue(variant.type.value)) {
                    newUsage |= variant.type.value;
                }
            }
            else {
                reportDiagnostic(context.program, {
                    code: "invalid-usage",
                    format: {},
                    target: entity,
                });
                return;
            }
        }
        if (newUsage === 0) {
            reportDiagnostic(context.program, {
                code: "invalid-usage",
                format: {},
                target: entity,
            });
            return;
        }
    }
    // Get existing usage and combine with new usage (additive behavior)
    const existingUsage = getScopedDecoratorData(context, usageKey, entity) || 0;
    const combinedUsage = existingUsage | newUsage;
    setScopedDecoratorData(context, $usage, usageKey, entity, combinedUsage, scope);
};
export function getUsageOverride(context, entity) {
    const usageFlags = getScopedDecoratorData(context, usageKey, entity);
    if (usageFlags || entity.namespace === undefined)
        return usageFlags;
    return getScopedDecoratorData(context, usageKey, entity.namespace);
}
export function getUsage(context, entity) {
    switch (entity.kind) {
        case "Union":
            const type = getSdkUnion(context, entity);
            if (type.kind === "enum" || type.kind === "union" || type.kind === "nullable") {
                return type.usage;
            }
            return UsageFlags.None;
        case "Model":
            return getSdkModel(context, entity).usage;
        case "Enum":
            return getSdkEnum(context, entity).usage;
    }
}
const accessKey = createStateSymbol("access");
export const $access = (context, entity, value, scope) => {
    if (typeof value.value !== "string" || (value.value !== "public" && value.value !== "internal")) {
        reportDiagnostic(context.program, {
            code: "invalid-access",
            format: {},
            target: entity,
        });
        return;
    }
    setScopedDecoratorData(context, $access, accessKey, entity, value.value, scope);
};
export function getAccessOverride(context, entity) {
    const accessOverride = getScopedDecoratorData(context, accessKey, entity);
    if (!accessOverride && entity.kind !== "ModelProperty" && entity.namespace) {
        return getAccessOverride(context, entity.namespace);
    }
    return accessOverride;
}
export function getAccess(context, entity) {
    const override = getAccessOverride(context, entity);
    if (override || entity.kind === "Operation" || entity.kind === "ModelProperty") {
        return override || "public";
    }
    switch (entity.kind) {
        case "Model":
            return getSdkModel(context, entity).access;
        case "Enum":
            return getSdkEnum(context, entity).access;
        case "Union": {
            const type = getSdkUnion(context, entity);
            if (type.kind === "enum" || type.kind === "union" || type.kind === "nullable") {
                return type.access;
            }
            return "public";
        }
    }
}
const flattenPropertyKey = createStateSymbol("flattenProperty");
/**
 * Whether a model property should be flattened.
 *
 * @param context DecoratorContext
 * @param target ModelProperty to mark as flattened
 * @param scope Names of the projection (e.g. "python", "csharp", "java", "javascript")
 */
export const $flattenProperty = (context, target, scope) => {
    if (getDiscriminator(context.program, target.type)) {
        reportDiagnostic(context.program, {
            code: "flatten-polymorphism",
            format: {},
            target: target,
        });
        return;
    }
    setScopedDecoratorData(context, $flattenProperty, flattenPropertyKey, target, true, scope);
};
/**
 * Whether a model property should be flattened or not.
 *
 * @param context TCGCContext
 * @param target ModelProperty that we want to check whether it should be flattened or not
 * @returns whether the model property should be flattened or not
 */
export function shouldFlattenProperty(context, target) {
    return getScopedDecoratorData(context, flattenPropertyKey, target) ?? false;
}
export const $clientName = (context, entity, value, scope) => {
    // workaround for current lack of functionality in compiler
    // https://github.com/microsoft/typespec/issues/2717
    if (entity.kind === "Model" || entity.kind === "Operation") {
        const target = context.decoratorTarget;
        if (target.kind === SyntaxKind.AugmentDecoratorStatement) {
            if (ignoreDiagnostics(context.program.checker.resolveTypeReference(target.targetType))?.node !== entity.node) {
                return;
            }
        }
        if (target.kind === SyntaxKind.DecoratorExpression) {
            if (target.parent !== entity.node) {
                return;
            }
        }
    }
    if (value.trim() === "") {
        reportDiagnostic(context.program, {
            code: "empty-client-name",
            format: {},
            target: entity,
        });
        return;
    }
    setScopedDecoratorData(context, $clientName, clientNameKey, entity, value, scope);
};
export function getClientNameOverride(context, entity, languageScope) {
    return getScopedDecoratorData(context, clientNameKey, entity, languageScope);
}
// Recursive function to collect parameter names
function collectParams(properties, params = []) {
    properties.forEach((value, key) => {
        // If the property is of type 'model', recurse into its properties
        if (params.filter((x) => compareModelProperties(undefined, x, value)).length === 0) {
            if (value.type.kind === "Model") {
                collectParams(value.type.properties, params);
            }
            else {
                params.push(findRootSourceProperty(value));
            }
        }
    });
    return params;
}
export const $override = (context, original, override, scope) => {
    // omit all override operation
    context.program.stateMap(omitOperation).set(override, true);
    // Extract and sort parameter names
    const originalParams = collectParams(original.parameters.properties).sort((a, b) => a.name.localeCompare(b.name));
    const overrideParams = collectParams(override.parameters.properties).sort((a, b) => a.name.localeCompare(b.name));
    // Check if the sorted parameter names arrays are equal, omit optional parameters
    let parametersMatch = true;
    let checkParameter = undefined;
    let index = 0;
    for (const originalParam of originalParams) {
        if (index > overrideParams.length - 1) {
            if (!originalParam.optional) {
                parametersMatch = false;
                checkParameter = originalParam;
                break;
            }
            else {
                continue;
            }
        }
        if (!compareModelProperties(undefined, originalParam, overrideParams[index])) {
            if (!originalParam.optional) {
                parametersMatch = false;
                checkParameter = originalParam;
                break;
            }
            else {
                continue;
            }
        }
        // Apply the alternate type to the original parameter
        const overrideParam = overrideParams[index];
        overrideParam.decorators
            .filter((d) => d.definition?.name === "@alternateType" &&
            getNamespaceFullName(d.definition?.namespace) === namespace)
            .map((d) => context.call($alternateType, originalParam, d.args[0].value, d.args[1]?.jsValue));
        index++;
    }
    if (!parametersMatch) {
        reportDiagnostic(context.program, {
            code: "override-parameters-mismatch",
            target: context.decoratorTarget,
            format: {
                methodName: original.name,
                checkParameter: checkParameter?.name ?? "",
            },
        });
    }
    setScopedDecoratorData(context, $override, overrideKey, original, override, scope);
};
/**
 * Gets additional information on how to serialize / deserialize TYPESPEC standard types depending
 * on whether additional serialization information is provided or needed
 *
 * @param context the Sdk Context
 * @param entity the entity whose client format we are going to get
 * @returns the format in which to serialize the typespec type or undefined
 */
export function getOverriddenClientMethod(context, entity) {
    return getScopedDecoratorData(context, overrideKey, entity);
}
/**
 * Check if a model is an external type.
 * The external type model has properties: identity (required), package (optional), minVersion (optional).
 */
function isExternalType(model) {
    if (model.indexer !== undefined) {
        return false;
    }
    const properties = [...model.properties.values()];
    // Check if it has an 'identity' property with String literal type
    const hasIdentity = properties.some((prop) => prop.name === "identity" && prop.type.kind === "String");
    if (!hasIdentity) {
        return false;
    }
    // Check that all other properties are only 'package' or 'minVersion' with String literal types
    const otherProps = properties.filter((prop) => prop.name !== "identity");
    const validProps = otherProps.every((prop) => (prop.name === "package" || prop.name === "minVersion") && prop.type.kind === "String");
    return validProps;
}
const alternateTypeKey = createStateSymbol("alternateType");
/**
 * Replace a source type with an alternate type in a specific scope.
 *
 * @param context the decorator context
 * @param source source type to be replaced
 * @param alternate target type to replace the source type or ExternalType object
 * @param scope Names of the projection (e.g. "python", "csharp", "java", "javascript")
 */
export const $alternateType = (context, source, alternate, scope) => {
    let alternateInput = alternate;
    if (alternate.kind === "Model" && isExternalType(alternate)) {
        // This means we're dealing with external type
        if (source.kind === "ModelProperty") {
            reportDiagnostic(context.program, {
                code: "external-type-on-model-property",
                target: source,
            });
            return;
        }
        if (!scope) {
            reportDiagnostic(context.program, {
                code: "missing-scope",
                format: {
                    decoratorName: "@alternateType",
                },
                target: source,
            });
        }
        const alternatePropertyValues = [...alternate.properties.values()];
        // Get identity if needed
        const identity = alternatePropertyValues
            .filter((x) => x.name === "identity")
            .map((x) => x.type)
            .filter((x) => x.kind === "String")
            .map((x) => x.value)[0];
        const packageName = alternatePropertyValues
            .filter((x) => x.name === "package")
            .map((x) => x.type)
            .filter((x) => x.kind === "String")
            .map((x) => x.value)[0];
        const minVersion = alternatePropertyValues
            .filter((x) => x.name === "minVersion")
            .map((x) => x.type)
            .filter((x) => x.kind === "String")
            .map((x) => x.value)[0];
        alternateInput = {
            kind: "externalTypeInfo",
            identity,
            package: packageName,
            minVersion,
        };
    }
    else {
        // Not external type
        if (source.kind === "Scalar" && alternate.kind !== "Scalar") {
            reportDiagnostic(context.program, {
                code: "invalid-alternate-type",
                format: {
                    kindName: alternate.kind,
                },
                target: alternate,
            });
            return;
        }
    }
    setScopedDecoratorData(context, $alternateType, alternateTypeKey, source, alternateInput, scope);
};
/**
 * Get the alternate type for a source type in a specific scope.
 *
 * @param context the Sdk Context
 * @param source source type to be replaced
 * @returns alternate type to replace the source type, or undefined if no alternate type is found
 */
export function getAlternateType(context, source) {
    const retval = getScopedDecoratorData(context, alternateTypeKey, source);
    if (retval !== undefined && retval.kind === "externalTypeInfo") {
        if (!context.__externalPackageToVersions) {
            context.__externalPackageToVersions = new Map();
        }
        const externalPackage = retval.package;
        const externalMinVersion = retval.minVersion;
        if (externalPackage && externalMinVersion) {
            const existingVersion = context.__externalPackageToVersions.get(externalPackage);
            if (existingVersion && existingVersion !== externalMinVersion) {
                reportDiagnostic(context.program, {
                    code: "external-library-version-mismatch",
                    format: {
                        libraryName: externalPackage,
                        versionA: existingVersion,
                        versionB: externalMinVersion,
                    },
                    target: source,
                });
            }
            context.__externalPackageToVersions.set(externalPackage, externalMinVersion);
        }
    }
    return retval;
}
export const $useSystemTextJsonConverter = (context, entity, scope) => { };
const clientInitializationKey = createStateSymbol("clientInitialization");
export const $clientInitialization = (context, target, options, scope) => {
    if (options.kind === "Model") {
        if (options.properties.get("initializedBy")) {
            const value = options.properties.get("initializedBy").type;
            const isValidValue = (value) => value === 4 || value === 1 || value === 2;
            if (value.kind === "EnumMember") {
                if (typeof value.value !== "number" || !isValidValue(value.value)) {
                    reportDiagnostic(context.program, {
                        code: "invalid-initialized-by",
                        format: {
                            message: "Please use `InitializedBy` enum to set the value.",
                        },
                        target: target,
                    });
                    return;
                }
            }
            else if (value.kind === "Union") {
                for (const variant of value.variants.values()) {
                    if (variant.type.kind !== "EnumMember" ||
                        typeof variant.type.value !== "number" ||
                        !isValidValue(variant.type.value)) {
                        reportDiagnostic(context.program, {
                            code: "invalid-initialized-by",
                            format: {
                                message: "Please use `InitializedBy` enum to set the value.",
                            },
                            target: target,
                        });
                        return;
                    }
                    if (variant.type.value === 4) {
                        reportDiagnostic(context.program, {
                            code: "invalid-initialized-by",
                            format: {
                                message: "`InitializedBy.customizeCode` cannot be combined with other values.",
                            },
                            target: target,
                        });
                        return;
                    }
                }
            }
        }
        setScopedDecoratorData(context, $clientInitialization, clientInitializationKey, target, options, scope);
    }
};
/**
 * Get client initialization options for namespace or interface. The info is from `@clientInitialization` decorator.
 *
 * @param context
 * @param entity namespace or interface which represents a client
 * @returns
 */
export function getClientInitializationOptions(context, entity) {
    const options = getScopedDecoratorData(context, clientInitializationKey, entity);
    // backward compatibility
    if (options &&
        options.properties.get("initializedBy") === undefined &&
        options.properties.get("parameters") === undefined) {
        return {
            parameters: options,
        };
    }
    let initializedBy = undefined;
    if (options?.properties.get("initializedBy")) {
        if (options.properties.get("initializedBy").type.kind === "EnumMember") {
            initializedBy = options.properties.get("initializedBy").type.value;
        }
        else if (options.properties.get("initializedBy").type.kind === "Union") {
            initializedBy = 0;
            for (const variant of options.properties.get("initializedBy").type.variants.values()) {
                initializedBy |= variant.type.value;
            }
        }
    }
    let parametersModel = options?.properties.get("parameters")?.type;
    let currEntity = entity;
    while (currEntity) {
        const movedParameters = findEntriesWithTarget(context, clientLocationKey, currEntity, "ModelProperty");
        const tk = $(context.program);
        if (movedParameters.length > 0) {
            if (parametersModel) {
                // If the parameters model already exists, we will merge the moved parameters into it.
                for (const movedParameter of movedParameters) {
                    parametersModel.properties.set(movedParameter.name, movedParameter);
                }
            }
            else {
                parametersModel = tk.model.create({
                    name: "ClientInitializationParameters",
                    properties: {
                        ...Object.fromEntries(movedParameters.map((movedParameter) => [movedParameter.name, movedParameter])),
                    },
                });
            }
        }
        currEntity = currEntity.namespace;
    }
    return {
        parameters: parametersModel,
        initializedBy: initializedBy,
    };
}
const paramAliasKey = createStateSymbol("paramAlias");
export const $paramAlias = (context, original, paramAlias, scope) => {
    const paramAliasDec = context.program.stateMap(paramAliasKey).get(original);
    const paramAliasVal = paramAliasDec?.[scope || AllScopes] ?? paramAliasDec?.[AllScopes];
    if (paramAliasVal) {
        reportDiagnostic(context.program, {
            code: "multiple-param-alias",
            format: {
                originalName: original.name,
                firstParamAlias: paramAliasVal,
            },
            target: context.decoratorTarget,
        });
        return;
    }
    setScopedDecoratorData(context, $paramAlias, paramAliasKey, original, paramAlias, scope);
};
export function getParamAlias(context, original) {
    return getScopedDecoratorData(context, paramAliasKey, original);
}
const apiVersionKey = createStateSymbol("apiVersion");
export const $apiVersion = (context, target, value, scope) => {
    setScopedDecoratorData(context, $apiVersion, apiVersionKey, target, value ?? true, scope);
};
export function getIsApiVersion(context, param) {
    return getScopedDecoratorData(context, apiVersionKey, param);
}
export const $clientNamespace = (context, entity, value, scope) => {
    if (value.trim() === "") {
        reportDiagnostic(context.program, {
            code: "empty-client-namespace",
            format: {},
            target: entity,
        });
        return;
    }
    setScopedDecoratorData(context, $clientNamespace, clientNamespaceKey, entity, value, scope);
};
/**
 * Find the shortest namespace that overlaps with the override string.
 * @param override
 * @param userDefinedNamespaces
 * @returns
 */
function findNamespaceOverlapClosestToRoot(override, userDefinedNamespaces) {
    for (const namespace of userDefinedNamespaces) {
        if (override.includes(namespace.name)) {
            return namespace;
        }
    }
    return undefined;
}
/**
 * Returns the client namespace for a given entity. The order of operations is as follows:
 *
 * 1. If `@clientNamespace` is applied to the entity, this wins out.
 *    a. If the `--namespace` flag is passed in during generation, we will replace the root of the client namespace with the flag.
 * 2. If the `--namespace` flag is passed in, we treat that as the only namespace in the entire spec, and return that namespace.
 * 3. We return the namespace of the entity retrieved from the original spec.
 * @param context
 * @param entity
 * @returns
 */
export function getClientNamespace(context, entity) {
    const override = getScopedDecoratorData(context, clientNamespaceKey, entity);
    if (override) {
        // if `@clientNamespace` is applied to the entity, this wins out
        // if the override matches or extends the namespace flag, no replacement is needed
        if (context.namespaceFlag &&
            (override === context.namespaceFlag || override.startsWith(context.namespaceFlag + "."))) {
            return override;
        }
        const userDefinedNamespace = findNamespaceOverlapClosestToRoot(override, listAllUserDefinedNamespaces(context));
        if (userDefinedNamespace && context.namespaceFlag) {
            // we still make sure to replace the root of the client namespace with the flag (if the flag exists)
            return override.replace(userDefinedNamespace.name, context.namespaceFlag);
        }
        return override;
    }
    if (!entity.namespace) {
        return "";
    }
    if (entity.kind === "Namespace") {
        return getNamespaceFullNameWithOverride(context, entity);
    }
    return getNamespaceFullNameWithOverride(context, entity.namespace);
}
function getNamespaceFullNameWithOverride(context, namespace) {
    const segments = [];
    let current = namespace;
    let isOverridden = false;
    while (current && current.name !== "") {
        const override = getScopedDecoratorData(context, clientNamespaceKey, current);
        if (override) {
            segments.unshift(override);
            isOverridden = true;
            break;
        }
        segments.unshift(current.name);
        current = current.namespace;
    }
    const joinedSegments = segments.join(".");
    if (isOverridden) {
        // if it's overridden, and there's a `@clientNamespace` flag, we want to do the shortest namespace overlap replacement
        const userDefinedNamespace = findNamespaceOverlapClosestToRoot(joinedSegments, listAllUserDefinedNamespaces(context));
        if (userDefinedNamespace && context.namespaceFlag) {
            // Check if replacement would cause duplication:
            // This happens when the namespace flag is an extension of the user-defined namespace
            // and joinedSegments already starts with the flag (meaning override already applied it)
            if (context.namespaceFlag.startsWith(userDefinedNamespace.name) &&
                (joinedSegments.startsWith(context.namespaceFlag + ".") ||
                    joinedSegments === context.namespaceFlag)) {
                return joinedSegments;
            }
            return joinedSegments.replace(userDefinedNamespace.name, context.namespaceFlag);
        }
        return joinedSegments;
    }
    if (context.namespaceFlag)
        return context.namespaceFlag;
    return joinedSegments;
}
export const $scope = (context, entity, scope) => {
    const [negationScopes, scopes] = parseScopes(scope);
    if (negationScopes !== undefined && negationScopes.length > 0) {
        // for negation scope, override the previous value
        setScopedDecoratorData(context, $scope, negationScopesKey, entity, negationScopes);
    }
    if (scopes !== undefined && scopes.length > 0) {
        // for normal scope, add them incrementally
        const targetEntry = context.program.stateMap(scopeKey).get(entity);
        setScopedDecoratorData(context, $scope, scopeKey, entity, !targetEntry ? scopes : [...Object.values(targetEntry), ...scopes]);
    }
};
const clientApiVersionsKey = createStateSymbol("clientApiVersions");
/**
 * Add additional api versions that are possible for the client to use.
 *
 * @param context
 * @param target Service namespace that has these additional api versions
 * @param value Enum with the additional api versions
 * @param scope
 */
export const $clientApiVersions = (context, target, value, scope) => {
    setScopedDecoratorData(context, $clientApiVersions, clientApiVersionsKey, target, value, scope);
};
/**
 * Get the explicit client api versions that are possible for the client to use denoted by `@clientApiVersions`
 *
 * @param context
 * @param target
 * @returns
 */
export function getExplicitClientApiVersions(context, target) {
    return getScopedDecoratorData(context, clientApiVersionsKey, target);
}
export const $deserializeEmptyStringAsNull = (context, target, scope) => {
    if (target.type.kind !== "Scalar") {
        reportDiagnostic(context.program, {
            code: "invalid-deserializeEmptyStringAsNull-target-type",
            format: {},
            target: target,
        });
        return;
    }
    if (target.type.name !== "string") {
        let scalarType = target.type;
        while (scalarType.baseScalar !== undefined) {
            scalarType = scalarType.baseScalar;
        }
        if (scalarType.name !== "string") {
            reportDiagnostic(context.program, {
                code: "invalid-deserializeEmptyStringAsNull-target-type",
                format: {},
                target: target,
            });
            return;
        }
    }
};
const responseAsBoolKey = createStateSymbol("responseAsBool");
export const $responseAsBool = (context, target, scope) => {
    if (!target.decorators.some((d) => d.definition?.name === "@head")) {
        reportDiagnostic(context.program, {
            code: "non-head-bool-response-decorator",
            format: {
                operationName: target.name,
            },
            target: target,
        });
        return;
    }
    setScopedDecoratorData(context, $responseAsBool, responseAsBoolKey, target, true, scope);
};
export function getResponseAsBool(context, target) {
    return getScopedDecoratorData(context, responseAsBoolKey, target);
}
const clientDocKey = createStateSymbol("clientDoc");
export const $clientDoc = (context, target, documentation, mode, scope) => {
    const docMode = mode.value;
    // Validate the mode value
    if (docMode !== "append" && docMode !== "replace") {
        reportDiagnostic(context.program, {
            code: "invalid-client-doc-mode",
            format: { mode: docMode },
            target: context.decoratorTarget,
        });
        return;
    }
    const docData = {
        documentation,
        mode: docMode,
    };
    setScopedDecoratorData(context, $clientDoc, clientDocKey, target, docData, scope);
};
/**
 * Gets the client documentation data for a type.
 *
 * @param context TCGCContext
 * @param target Type to get client documentation for
 * @returns ClientDocData or undefined if no client documentation exists
 */
export function getClientDocExplicit(context, target) {
    return getScopedDecoratorData(context, clientDocKey, target);
}
export const $clientLocation = (context, source, target, scope) => {
    if (source.kind === "Operation") {
        // can only move parameters to an operation, not another operation
        if (typeof target !== "string" && target.kind === "Operation") {
            reportDiagnostic(context.program, {
                code: "client-location-conflict",
                format: { operationName: source.name },
                target: context.decoratorTarget,
                messageId: "operationToOperation",
            });
            return;
        }
    }
    else if (source.kind === "ModelProperty") {
        // verify that there isn't a conflict with existing client initialization parameter
        if (typeof target !== "string" &&
            (target.kind === "Interface" || target.kind === "Namespace")) {
            const clientInitializationParams = target.decorators
                .filter((d) => d.decorator.name === "$clientInitialization")
                .map((d) => d.args[0].value)
                .filter((a) => a.entityKind === "Type" && a.kind === "Model")
                .filter((model) => model.properties.has(source.name))
                .map((model) => model.properties.get(source.name));
            if (clientInitializationParams.length > 0) {
                reportDiagnostic(context.program, {
                    code: "client-location-conflict",
                    format: { parameterName: source.name },
                    target: context.decoratorTarget,
                    messageId: "modelPropertyToClientInitialization",
                });
                return;
            }
        }
        if (typeof target === "string") {
            reportDiagnostic(context.program, {
                code: "client-location-conflict",
                format: { parameterName: source.name },
                target: context.decoratorTarget,
                messageId: "modelPropertyToString",
            });
            return;
        }
    }
    setScopedDecoratorData(context, $clientLocation, clientLocationKey, source, target, scope);
};
/**
 * Gets the `Namespace`, `Interface`, `Operation` or name of client where an operation / parameter change the location to.
 *
 * @param context TCGCContext
 * @param input Operation or parameter to be moved
 * @returns `Namespace`, `Interface`, `Operation`, `string` target or undefined if no location change.
 */
export function getClientLocation(context, input) {
    // if there is `@client` or `@operationGroup` decorator, `@clientLocation` on operation will be ignored
    if (input.kind === "Operation" && hasExplicitClientOrOperationGroup(context)) {
        return undefined;
    }
    return getScopedDecoratorData(context, clientLocationKey, input);
}
const legacyHierarchyBuildingKey = createStateSymbol("legacyHierarchyBuilding");
function isPropertySuperset(program, target, value) {
    const conflicts = [];
    // Check if all properties in value exist in target
    for (const name of value.properties.keys()) {
        if (!target.properties.has(name)) {
            conflicts.push({
                propertyName: name,
                reason: "missing",
            });
            continue;
        }
        const targetProperty = target.properties.get(name);
        const valueProperty = value.properties.get(name);
        // Compare properties to handle envelope/spread semantics correctly
        // Properties match if they come from the same source OR if they have the same type
        // This ensures properties from envelopes (e.g., ...ArmTagsProperty) are recognized
        // as equivalent to directly defined properties with the same name and type
        if (targetProperty.sourceProperty !== valueProperty.sourceProperty) {
            // Different sources - check if they have the same type
            if (targetProperty.type !== valueProperty.type) {
                conflicts.push({
                    propertyName: name,
                    reason: "type-mismatch",
                });
            }
        }
    }
    return conflicts;
}
export const $legacyHierarchyBuilding = (context, target, value, scope) => {
    // Validate that target has all properties from value
    const conflicts = isPropertySuperset(context.program, target, value);
    if (conflicts.length > 0) {
        for (const conflict of conflicts) {
            if (conflict.reason === "missing") {
                reportDiagnostic(context.program, {
                    code: "legacy-hierarchy-building-conflict",
                    messageId: "property-missing",
                    format: {
                        childModel: target.name,
                        parentModel: value.name,
                        propertyName: conflict.propertyName,
                    },
                    target: context.decoratorTarget,
                });
            }
            else if (conflict.reason === "type-mismatch") {
                reportDiagnostic(context.program, {
                    code: "legacy-hierarchy-building-conflict",
                    messageId: "type-mismatch",
                    format: {
                        childModel: target.name,
                        parentModel: value.name,
                        propertyName: conflict.propertyName,
                    },
                    target: context.decoratorTarget,
                });
            }
        }
        return;
    }
    setScopedDecoratorData(context, $legacyHierarchyBuilding, legacyHierarchyBuildingKey, target, value, scope);
};
export function getLegacyHierarchyBuilding(context, target) {
    // If legacy hierarchy building is not respected, ignore the decorator completely
    if (!context.enableLegacyHierarchyBuilding)
        return undefined;
    return getScopedDecoratorData(context, legacyHierarchyBuildingKey, target);
}
const markAsLroKey = createStateSymbol("markAsLro");
export const $markAsLro = (context, target, scope) => {
    const httpOperation = ignoreDiagnostics(getHttpOperation(context.program, target));
    const hasModelResponse = httpOperation.responses.filter((r) => r.type?.kind === "Model" && !(r.statusCodes === "*" || isErrorModel(context.program, r.type)))[0];
    if (!hasModelResponse) {
        reportDiagnostic(context.program, {
            code: "invalid-mark-as-lro-target",
            format: {
                operation: target.name,
            },
            target: context.decoratorTarget,
        });
        return;
    }
    if (getLroMetadata(context.program, target)) {
        reportDiagnostic(context.program, {
            code: "mark-as-lro-ineffective",
            format: {
                operation: target.name,
            },
            target: context.decoratorTarget,
        });
        return;
    }
    setScopedDecoratorData(context, $markAsLro, markAsLroKey, target, true, scope);
};
export function getMarkAsLro(context, entity) {
    return getScopedDecoratorData(context, markAsLroKey, entity) ?? false;
}
const markAsPageableKey = createStateSymbol("markAsPageable");
export const $markAsPageable = (context, target, scope) => {
    const httpOperation = ignoreDiagnostics(getHttpOperation(context.program, target));
    const modelResponse = httpOperation.responses.filter((r) => r.type?.kind === "Model" && !(r.statusCodes === "*" || isErrorModel(context.program, r.type)))[0];
    if (!modelResponse) {
        reportDiagnostic(context.program, {
            code: "invalid-mark-as-pageable-target",
            format: {
                operation: target.name,
            },
            target: context.decoratorTarget,
        });
        return;
    }
    // Check if already marked with @list decorator
    if (isList(context.program, target)) {
        reportDiagnostic(context.program, {
            code: "mark-as-pageable-ineffective",
            format: {
                operation: target.name,
            },
            target: context.decoratorTarget,
        });
        return;
    }
    // Check the response model for @pageItems decorator
    const responseType = getRealResponseModel(context.program, modelResponse.type);
    if (responseType.kind !== "Model") {
        reportDiagnostic(context.program, {
            code: "invalid-mark-as-pageable-target",
            format: {
                operation: target.name,
            },
            target: context.decoratorTarget,
        });
        return;
    }
    // Check if any property has @pageItems decorator by checking the program state
    // The @pageItems decorator uses a state symbol "TypeSpec.pageItems"
    const pageItemsStateKey = Symbol.for("TypeSpec.pageItems");
    let itemsProperty = undefined;
    for (const [, prop] of responseType.properties) {
        if (context.program.stateSet(pageItemsStateKey).has(prop)) {
            itemsProperty = prop;
            break;
        }
    }
    if (!itemsProperty) {
        // Try to find a property named "value"
        itemsProperty = responseType.properties.get("value");
        if (!itemsProperty) {
            // No @pageItems property and no "value" property found
            reportDiagnostic(context.program, {
                code: "invalid-mark-as-pageable-target",
                format: {
                    operation: target.name,
                },
                target: context.decoratorTarget,
            });
            return;
        }
    }
    // Store metadata that will be checked by TCGC to treat this operation as pageable
    setScopedDecoratorData(context, $markAsPageable, markAsPageableKey, target, { itemsProperty }, scope);
};
export function getMarkAsPageable(context, entity) {
    return getScopedDecoratorData(context, markAsPageableKey, entity);
}
const disablePageableKey = createStateSymbol("disablePageable");
export const $disablePageable = (context, target, scope) => {
    setScopedDecoratorData(context, $disablePageable, disablePageableKey, target, true, scope);
};
export function getDisablePageable(context, entity) {
    return getScopedDecoratorData(context, disablePageableKey, entity) ?? false;
}
function getRealResponseModel(program, responseModel) {
    let bodyProperty = undefined;
    for (const prop of responseModel.properties.values()) {
        if (isBody(program, prop) || isBodyRoot(program, prop)) {
            bodyProperty = prop;
            break;
        }
    }
    if (bodyProperty) {
        return bodyProperty.type;
    }
    return responseModel;
}
const nextLinkVerbKey = createStateSymbol("nextLinkVerb");
export const $nextLinkVerb = (context, target, verb, scope) => {
    compilerAssert(verb.kind === "String" && (verb.value === "POST" || verb.value === "GET"), "@nextLinkVerb decorator only supports 'POST' or 'GET' string literal values.");
    setScopedDecoratorData(context, $nextLinkVerb, nextLinkVerbKey, target, verb.value, scope);
};
/**
 * Get the HTTP verb specified for next link operations in paging scenarios.
 * @param context TCGCContext
 * @param entity Operation to check for nextLinkVerb decorator
 * @returns The HTTP verb string ("POST" or "GET"). Defaults to "GET" if decorator is not applied.
 */
export function getNextLinkVerb(context, entity) {
    return getScopedDecoratorData(context, nextLinkVerbKey, entity) ?? "GET";
}
const clientDefaultValueKey = createStateSymbol("clientDefaultValue");
export const $clientDefaultValue = (context, target, value, scope) => {
    const actualValue = isNumeric(value) ? value.asNumber() : value;
    setScopedDecoratorData(context, $clientDefaultValue, clientDefaultValueKey, target, actualValue, scope);
};
/**
 * Get the client-level default value for a model property.
 * @param context TCGCContext
 * @param entity ModelProperty to check for clientDefaultValue decorator
 * @returns The client-level default value if decorator is applied, undefined otherwise.
 */
export function getClientDefaultValue(context, entity) {
    return getScopedDecoratorData(context, clientDefaultValueKey, entity);
}
/**
 * Check if an operation or model property is in scope for the current emitter.
 * @param context TCGCContext
 * @param entity Operation or ModelProperty to check if it is in scope
 * @returns
 */
export function isInScope(context, entity) {
    const scopes = getScopedDecoratorData(context, scopeKey, entity);
    const negationScopes = getScopedDecoratorData(context, negationScopesKey, entity);
    if (scopes !== undefined) {
        if (scopes.includes(context.emitterName)) {
            return true;
        }
        if (negationScopes === undefined) {
            return false;
        }
    }
    if (negationScopes !== undefined && negationScopes.includes(context.emitterName)) {
        return false;
    }
    return true;
}
export const clientOptionKey = createStateSymbol("ClientOption");
/**
 * `@clientOption` decorator implementation.
 * Pass experimental flags or options to emitters without requiring TCGC reshipping.
 * The decorator data is stored as {name, value} and exposed via the decorators array.
 */
export const $clientOption = (context, target, name, value, scope) => {
    // Always emit warning that this is experimental
    reportDiagnostic(context.program, {
        code: "client-option",
        target: target,
    });
    // Emit additional warning if scope is not provided
    if (scope === undefined) {
        reportDiagnostic(context.program, {
            code: "client-option-requires-scope",
            target: target,
        });
    }
    // Store the option data - each decorator application is stored separately
    // The decorator info will be exposed via the decorators array on SDK types
    setScopedDecoratorData(context, $clientOption, clientOptionKey, target, { name, value }, scope);
};
