// @azure-tools/typespec-client-generator-core/dist/src/interfaces.js
/**
 * This enum represents the different ways a model can be used in a method.
 */
export var UsageFlags;
(function (UsageFlags) {
    UsageFlags[UsageFlags["None"] = 0] = "None";
    UsageFlags[UsageFlags["Input"] = 2] = "Input";
    UsageFlags[UsageFlags["Output"] = 4] = "Output";
    UsageFlags[UsageFlags["ApiVersionEnum"] = 8] = "ApiVersionEnum";
    /** Input and Json will also be set when JsonMergePatch is set. */
    UsageFlags[UsageFlags["JsonMergePatch"] = 16] = "JsonMergePatch";
    /** Input will also be set when MultipartFormData is set. */
    UsageFlags[UsageFlags["MultipartFormData"] = 32] = "MultipartFormData";
    /** Used in spread. */
    UsageFlags[UsageFlags["Spread"] = 64] = "Spread";
    /** Set when type is used in conjunction with an application/json content type. */
    UsageFlags[UsageFlags["Json"] = 256] = "Json";
    /** Set when type is used in conjunction with an application/xml content type. */
    UsageFlags[UsageFlags["Xml"] = 512] = "Xml";
    /** Set when type is used for exception output. */
    UsageFlags[UsageFlags["Exception"] = 1024] = "Exception";
    /** Set when type is used as LRO initial response. */
    UsageFlags[UsageFlags["LroInitial"] = 2048] = "LroInitial";
    /** Set when type is used as LRO polling response. */
    UsageFlags[UsageFlags["LroPolling"] = 4096] = "LroPolling";
    /** Set when type is used as LRO final envelop response. */
    UsageFlags[UsageFlags["LroFinalEnvelope"] = 8192] = "LroFinalEnvelope";
    /** Set when type is only referenced by external types. */
    UsageFlags[UsageFlags["External"] = 16384] = "External";
})(UsageFlags || (UsageFlags = {}));
/**
 * Flags used to indicate how a client is initialized.
 *
 * - `Default` (0): No user-specific initialization setting has been specified. This is the default value for sub clients when no explicit initialization decorator is set.
 * - `Individually` (1): The client could be initialized individually.
 * - `Parent` (2): The client could be initialized by its parent client.
 * - `CustomizeCode` (4): Indicates that the client initialization should be omitted from generated code and handled manually in custom code.
 * - `Individually` and `Parent` are bit flags that can be combined using bitwise OR.
 */
export var InitializedByFlags;
(function (InitializedByFlags) {
    InitializedByFlags[InitializedByFlags["Default"] = 0] = "Default";
    InitializedByFlags[InitializedByFlags["CustomizeCode"] = 4] = "CustomizeCode";
    InitializedByFlags[InitializedByFlags["Individually"] = 1] = "Individually";
    InitializedByFlags[InitializedByFlags["Parent"] = 2] = "Parent";
})(InitializedByFlags || (InitializedByFlags = {}));
// these two vars are used to validate whether our SdkBuiltInKinds are exhaustive for all possible values from typespec
// if it is not, a typescript compilation error will be thrown here.
const _ = true;
const __ = true;
var SdkIntKindsEnum;
(function (SdkIntKindsEnum) {
    SdkIntKindsEnum["numeric"] = "numeric";
    SdkIntKindsEnum["integer"] = "integer";
    SdkIntKindsEnum["safeint"] = "safeint";
    SdkIntKindsEnum["int8"] = "int8";
    SdkIntKindsEnum["int16"] = "int16";
    SdkIntKindsEnum["int32"] = "int32";
    SdkIntKindsEnum["int64"] = "int64";
    SdkIntKindsEnum["uint8"] = "uint8";
    SdkIntKindsEnum["uint16"] = "uint16";
    SdkIntKindsEnum["uint32"] = "uint32";
    SdkIntKindsEnum["uint64"] = "uint64";
})(SdkIntKindsEnum || (SdkIntKindsEnum = {}));
var SdkFloatingPointKindsEnum;
(function (SdkFloatingPointKindsEnum) {
    SdkFloatingPointKindsEnum["float"] = "float";
    SdkFloatingPointKindsEnum["float32"] = "float32";
    SdkFloatingPointKindsEnum["float64"] = "float64";
})(SdkFloatingPointKindsEnum || (SdkFloatingPointKindsEnum = {}));
var SdkFixedPointKindsEnum;
(function (SdkFixedPointKindsEnum) {
    SdkFixedPointKindsEnum["decimal"] = "decimal";
    SdkFixedPointKindsEnum["decimal128"] = "decimal128";
})(SdkFixedPointKindsEnum || (SdkFixedPointKindsEnum = {}));
var SdkGenericBuiltInStringKindsEnum;
(function (SdkGenericBuiltInStringKindsEnum) {
    SdkGenericBuiltInStringKindsEnum["string"] = "string";
    SdkGenericBuiltInStringKindsEnum["url"] = "url";
})(SdkGenericBuiltInStringKindsEnum || (SdkGenericBuiltInStringKindsEnum = {}));
var SdkBuiltInKindsMiscellaneousEnum;
(function (SdkBuiltInKindsMiscellaneousEnum) {
    SdkBuiltInKindsMiscellaneousEnum["bytes"] = "bytes";
    SdkBuiltInKindsMiscellaneousEnum["boolean"] = "boolean";
    SdkBuiltInKindsMiscellaneousEnum["plainDate"] = "plainDate";
    SdkBuiltInKindsMiscellaneousEnum["plainTime"] = "plainTime";
    SdkBuiltInKindsMiscellaneousEnum["unknown"] = "unknown";
})(SdkBuiltInKindsMiscellaneousEnum || (SdkBuiltInKindsMiscellaneousEnum = {}));
export function getKnownScalars() {
    const retval = {};
    const typespecNamespace = Object.keys(SdkBuiltInKindsMiscellaneousEnum)
        .concat(Object.keys(SdkIntKindsEnum))
        .concat(Object.keys(SdkFloatingPointKindsEnum))
        .concat(Object.keys(SdkFixedPointKindsEnum))
        .concat(Object.keys(SdkGenericBuiltInStringKindsEnum));
    for (const kind of typespecNamespace) {
        if (!isSdkBuiltInKind(kind))
            continue; // it will always be true
        retval[`TypeSpec.${kind}`] = kind;
    }
    return retval;
}
export function isSdkBuiltInKind(kind) {
    return (kind in SdkBuiltInKindsMiscellaneousEnum ||
        isSdkIntKind(kind) ||
        isSdkFloatKind(kind) ||
        isSdkFixedPointKind(kind) ||
        kind in SdkGenericBuiltInStringKindsEnum);
}
export function isSdkIntKind(kind) {
    return kind in SdkIntKindsEnum;
}
export function isSdkFloatKind(kind) {
    return kind in SdkFloatingPointKindsEnum;
}
function isSdkFixedPointKind(kind) {
    return kind in SdkFixedPointKindsEnum;
}
const SdkDateTimeEncodingsConst = ["rfc3339", "rfc7231", "unixTimestamp"];
export function isSdkDateTimeEncodings(encoding) {
    return SdkDateTimeEncodingsConst.includes(encoding);
}
