// @azure-tools/typespec-client-generator-core/dist/src/internal-utils.js
import { FinalStateValue, getLroMetadata, isPreviewVersion, } from "@azure-tools/typespec-azure-core";
import { compilerAssert, createDiagnosticCollector, getDeprecationDetails, getDoc, getLifecycleVisibilityEnum, getNamespaceFullName, getSummary, getVisibilityForClass, ignoreDiagnostics, isNeverType, isNullType, isVoidType, listServices, } from "@typespec/compiler";
import { unsafe_mutateSubgraphWithNamespace, unsafe_Realm, } from "@typespec/compiler/experimental";
import { $ } from "@typespec/compiler/typekit";
import { getAddedOnVersions, getRemovedOnVersions, getVersioningMutators, getVersions, } from "@typespec/versioning";
import assert from "assert";
import { getAlternateType, getClientDocExplicit, getClientLocation, getMarkAsLro, getOverriddenClientMethod, getParamAlias, } from "./decorators.js";
import { getSdkHttpParameter, isSdkHttpParameter } from "./http.js";
import { createDiagnostic, createStateSymbol, reportDiagnostic } from "./lib.js";
import { getSdkBasicServiceMethod } from "./methods.js";
import { getCrossLanguageDefinitionId, getHttpOperationWithCache, isApiVersion, } from "./public-utils.js";
import { getClientTypeWithDiagnostics } from "./types.js";
export const AllScopes = Symbol.for("@azure-core/typespec-client-generator-core/all-scopes");
export const clientNameKey = createStateSymbol("clientName");
export const clientNamespaceKey = createStateSymbol("clientNamespace");
export const negationScopesKey = createStateSymbol("negationScopes");
export const scopeKey = createStateSymbol("scope");
export const clientKey = createStateSymbol("client");
export const operationGroupKey = createStateSymbol("operationGroup");
export const clientLocationKey = createStateSymbol("clientLocation");
export const omitOperation = createStateSymbol("omitOperation");
export const overrideKey = createStateSymbol("override");
export function hasExplicitClientOrOperationGroup(context) {
    // Multiple services case is not considered explicit client. It is auto-merged.
    const explicitClients = listScopedDecoratorData(context, clientKey);
    let multiServices = false;
    explicitClients.forEach((value) => {
        if (value.services.length > 1) {
            multiServices = true;
        }
    });
    return ((explicitClients.size > 0 && !multiServices) ||
        listScopedDecoratorData(context, operationGroupKey).size > 0);
}
export function listScopedDecoratorData(context, key, languageScope) {
    const scope = languageScope ?? context.emitterName;
    const retval = new Map();
    for (const [type, data] of context.program.stateMap(key).entries()) {
        if (data[scope]) {
            // positive scope case
            retval.set(type, data[scope]);
        }
        else if (data[negationScopesKey]) {
            // negative scope case
            if (data[negationScopesKey].includes(scope)) {
                // if the scope is negated, we should not include it
                continue;
            }
            else {
                // if the scope is not negated, we should include it
                retval.set(type, data[AllScopes]);
            }
        }
        else if (data[AllScopes]) {
            // all scopes case
            retval.set(type, data[AllScopes]);
        }
    }
    return retval;
}
export function getScopedDecoratorData(context, key, target, languageScope) {
    const retval = context.program.stateMap(key).get(target);
    if (retval === undefined)
        return retval;
    if (languageScope === AllScopes) {
        return retval[languageScope];
    }
    if (languageScope === undefined || typeof languageScope === "string") {
        const scope = languageScope ?? context.emitterName;
        if (Object.keys(retval).includes(scope))
            return retval[scope];
        // if the scope is negated, we should return undefined
        // if the scope is not negated, we should return the value for AllScopes
        const negationScopes = retval[negationScopesKey];
        if (negationScopes !== undefined && negationScopes.includes(scope)) {
            return undefined;
        }
    }
    return retval[AllScopes]; // in this case it applies to all languages
}
/**
 * Parse a scope string to extract negation scopes and positive scopes.
 * Supports two syntax patterns:
 * 1. !(scope1, scope2,...) - Grouped negation
 * 2. !scope1, !scope2, scope3, ... - Individual negation with positive scopes
 *
 * @param scope The scope string to parse
 * @returns A tuple of [negationScopes, positiveScopes] where each can be undefined if not present
 */
export function parseScopes(scope) {
    if (scope === undefined) {
        return [undefined, undefined];
    }
    // handle !(scope1, scope2,...) syntax
    const negationScopeRegex = /!\((.*?)\)/;
    const negationScopeMatch = scope.match(negationScopeRegex);
    if (negationScopeMatch) {
        return [negationScopeMatch[1].split(",").map((s) => s.trim()), undefined];
    }
    // handle !scope1, !scope2, scope3, ... syntax
    const splitScopes = scope.split(",").map((s) => s.trim());
    const negationScopes = [];
    const scopes = [];
    for (const s of splitScopes) {
        if (s.startsWith("!")) {
            negationScopes.push(s.slice(1));
        }
        else {
            scopes.push(s);
        }
    }
    return [negationScopes, scopes];
}
/**
 * Check if a scope string is applicable to the given emitter name.
 * Handles negation scopes like "!python" or "!(java, python)".
 *
 * @param scopeArg The scope string from the decorator argument
 * @param emitterName The current emitter name
 * @returns true if the decorator should be included, false otherwise
 */
function isScopeApplicable(scopeArg, emitterName) {
    const [negationScopes, positiveScopes] = parseScopes(scopeArg);
    // If there are positive scopes specified
    if (positiveScopes !== undefined && positiveScopes.length > 0) {
        // If the emitter matches any positive scope, include it
        if (positiveScopes.includes(emitterName)) {
            return true;
        }
        // If positive scopes specified but emitter doesn't match any, and no negation scopes
        // then the decorator doesn't apply to this emitter
        if (negationScopes === undefined || negationScopes.length === 0) {
            return false;
        }
    }
    // If there are negation scopes
    if (negationScopes !== undefined && negationScopes.length > 0) {
        // If the emitter is in the negation list, exclude it
        if (negationScopes.includes(emitterName)) {
            return false;
        }
        // If not in negation list, include it (applies to all except negated scopes)
        return true;
    }
    // No scopes specified at all (empty string edge case)
    return true;
}
/**
 *
 * @param emitterName Full emitter name
 * @returns The language of the emitter. I.e. "@azure-tools/typespec-csharp" will return "csharp"
 */
export function parseEmitterName(program, emitterName) {
    const diagnostics = createDiagnosticCollector();
    if (!emitterName) {
        diagnostics.add(createDiagnostic({
            code: "no-emitter-name",
            format: {},
            target: program.getGlobalNamespaceType(),
        }));
        return diagnostics.wrap("none");
    }
    const regex = /(?:cadl|typespec|client|server)-([^\\/-]*)/;
    const match = emitterName.match(regex);
    if (!match || match.length < 2)
        return diagnostics.wrap("none");
    const language = match[1];
    if (["typescript", "ts"].includes(language))
        return diagnostics.wrap("javascript");
    return diagnostics.wrap(language);
}
/**
 * Find the service namespace that contains the given operation.
 * @param services Array of service namespaces
 * @param operation The operation to find the service for
 * @returns The service namespace that contains the operation
 */
export function findServiceForOperation(services, operation) {
    let namespace = operation.namespace;
    while (namespace) {
        if (services.includes(namespace)) {
            return namespace;
        }
        namespace = namespace.namespace;
    }
    // Fallback to the first service. This can happen when an operation is defined outside
    // of any service namespace (e.g., in Azure.ResourceManager or other shared namespaces)
    // and is imported into a client that combines multiple services. In such cases,
    // we use the first service's api version as the default.
    return services[0];
}
/**
 *
 * @param context
 * @param type The type that we are adding api version information onto
 * @param client The client or operation group that contains the operation
 * @param operation The operation that contains the api version parameter (needed for multi-service operation groups)
 * @returns Whether the type is the api version parameter and the default value for the client
 */
export function updateWithApiVersionInformation(context, type, client, operation) {
    const isApiVersionParam = isApiVersion(context, type);
    if (!isApiVersionParam || !client) {
        return { isApiVersionParam, clientDefaultValue: undefined };
    }
    // For single-service clients, use the cached value
    if (client.services.length <= 1) {
        return {
            isApiVersionParam,
            clientDefaultValue: context.__clientApiVersionDefaultValueCache.get(client),
        };
    }
    // For multi-service clients/operation groups, we need to find the api version
    // from the operation's specific service
    if (operation) {
        const service = findServiceForOperation(client.services, operation);
        const packageVersions = context.getPackageVersions().get(service) || [];
        return {
            isApiVersionParam,
            clientDefaultValue: packageVersions.length > 0 ? packageVersions[packageVersions.length - 1] : undefined,
        };
    }
    // No operation provided for multi-service client, return undefined
    return { isApiVersionParam, clientDefaultValue: undefined };
}
export function filterApiVersionsWithDecorators(context, type, apiVersions) {
    const addedOnVersions = getAddedOnVersions(context.program, type)?.map((x) => x.value) ?? [];
    const removedOnVersions = getRemovedOnVersions(context.program, type)?.map((x) => x.value) ?? [];
    let added = addedOnVersions.length ? false : true;
    let addedCounter = 0;
    let removeCounter = 0;
    const retval = [];
    for (let i = 0; i < apiVersions.length; i++) {
        const version = apiVersions[i];
        if (addedCounter < addedOnVersions.length && version === addedOnVersions[addedCounter]) {
            added = true;
            addedCounter++;
        }
        if (removeCounter < removedOnVersions.length && version === removedOnVersions[removeCounter]) {
            added = false;
            removeCounter++;
        }
        if (added) {
            // only add version smaller than config
            if (context.apiVersion === undefined ||
                context.apiVersion === "latest" ||
                context.apiVersion === "all" ||
                apiVersions.indexOf(context.apiVersion) >= i) {
                retval.push(version);
            }
        }
    }
    return retval;
}
/**
 *
 * @param context
 * @param type
 * @param client If it's associated with a client, meaning it's a param etc, we can see if it's available on that client
 * @returns All api versions the type is available on
 */
export function getAvailableApiVersions(context, type, wrapper) {
    let wrapperApiVersions = [];
    if (wrapper) {
        wrapperApiVersions = context.getApiVersionsForType(wrapper);
    }
    const allApiVersions = getVersions(context.program, type)[1]
        ?.getVersions()
        .map((x) => x.value) || [];
    const apiVersions = wrapperApiVersions.length ? wrapperApiVersions : allApiVersions;
    if (!apiVersions)
        return [];
    const explicitlyDecorated = filterApiVersionsWithDecorators(context, type, apiVersions);
    if (explicitlyDecorated.length) {
        context.setApiVersionsForType(type, explicitlyDecorated);
        return explicitlyDecorated;
    }
    context.setApiVersionsForType(type, wrapperApiVersions);
    return wrapperApiVersions;
}
/**
 *
 * @param type
 * @returns A unique id for each type so we can do set comparisons
 */
export function getHashForType(type) {
    if (type.kind === "array" || type.kind === "dict") {
        return `${type.kind}[${getHashForType(type.valueType)}]`;
    }
    if (type.kind === "enum" || type.kind === "model" || type.kind === "enumvalue")
        return type.name;
    if (type.kind === "union") {
        return type.variantTypes.map((x) => getHashForType(x)).join("|");
    }
    return type.kind;
}
/**
 * Helper function to return default values for encode etc
 * @param type
 */
export function getSdkTypeBaseHelper(context, type, kind) {
    const diagnostics = createDiagnosticCollector();
    const base = {
        __raw: type,
        deprecation: getDeprecationDetails(context.program, type)?.message,
        kind,
        decorators: diagnostics.pipe(getTypeDecorators(context, type)),
        doc: getClientDoc(context, type),
        summary: getSummary(context.program, type),
    };
    if (type.kind === "ModelProperty" ||
        type.kind === "Scalar" ||
        type.kind === "Model" ||
        type.kind === "Enum" ||
        type.kind === "Union") {
        const external = getAlternateType(context, type);
        // Only set external if it's an ExternalTypeInfo (has 'identity' but not 'kind' property), not a regular Type
        if (external && external.kind === "externalTypeInfo") {
            base.external = external;
        }
    }
    return diagnostics.wrap(base);
}
export function getNamespacePrefix(namespace) {
    return namespace ? getNamespaceFullName(namespace) + "." : "";
}
export function getTypeDecorators(context, type) {
    const diagnostics = createDiagnosticCollector();
    const retval = [];
    if ("decorators" in type) {
        for (const decorator of type.decorators) {
            // only process explicitly defined decorators
            if (decorator.definition) {
                const decoratorName = `${getNamespacePrefix(decorator.definition?.namespace)}${decorator.definition?.name}`;
                // white list filtering
                if (!context.decoratorsAllowList ||
                    !context.decoratorsAllowList.some((x) => new RegExp(x).test(decoratorName))) {
                    continue;
                }
                const decoratorInfo = {
                    name: decoratorName,
                    arguments: {},
                };
                for (let i = 0; i < decorator.args.length; i++) {
                    decoratorInfo.arguments[decorator.definition.parameters[i].name] = diagnostics.pipe(getDecoratorArgValue(context, decorator.args[i].jsValue, type, decoratorName));
                }
                // Filter by scope - only include decorators that match the current emitter or have no scope
                const scopeArg = decoratorInfo.arguments["scope"];
                if (scopeArg !== undefined && !isScopeApplicable(scopeArg, context.emitterName)) {
                    // Skip this decorator if its scope is not applicable to the current emitter
                    continue;
                }
                retval.push(decoratorInfo);
            }
        }
    }
    return diagnostics.wrap(retval);
}
function getDecoratorArgValue(context, arg, type, decoratorName) {
    const diagnostics = createDiagnosticCollector();
    if (typeof arg === "object" && arg !== null && "kind" in arg) {
        if (arg.kind === "EnumMember") {
            return diagnostics.wrap(diagnostics.pipe(getClientTypeWithDiagnostics(context, arg)));
        }
        if (arg.kind === "String" ||
            arg.kind === "Number" ||
            arg.kind === "Boolean" ||
            arg.kind === "Value") {
            return diagnostics.wrap(arg.value);
        }
        diagnostics.add(createDiagnostic({
            code: "unsupported-generic-decorator-arg-type",
            target: type,
            format: { decoratorName },
        }));
        return diagnostics.wrap(undefined);
    }
    return diagnostics.wrap(arg);
}
export function intOrFloat(value) {
    return value.toString().indexOf(".") === -1 ? "int32" : "float32";
}
/**
 * Whether a model or enum or union as enum is in Azure.Core[.Foundations] namespace
 * @param t
 * @returns
 */
export function isAzureCoreTspModel(t) {
    return ((t.kind === "Model" || t.kind === "Enum" || t.kind === "Union") &&
        t.namespace !== undefined &&
        ["Azure.Core", "Azure.Core.Foundations"].includes(getNamespaceFullName(t.namespace)));
}
export function isAcceptHeader(param) {
    return param.kind === "header" && param.serializedName.toLowerCase() === "accept";
}
export function isContentTypeHeader(param) {
    return param.kind === "header" && param.serializedName.toLowerCase() === "content-type";
}
export function isHttpOperation(context, obj) {
    return obj?.kind === "Operation" && getHttpOperationWithCache(context, obj) !== undefined;
}
export function getNonNullOptions(type) {
    return [...type.variants.values()].map((x) => x.type).filter((t) => !isNullType(t));
}
export function getNullOption(type) {
    return [...type.variants.values()].map((x) => x.type).filter((t) => isNullType(t))[0];
}
/**
 * Use this if you are trying to create a generated name for something without an original TypeSpec type.
 *
 * Otherwise, you should use the `getGeneratedName` function.
 * @param context
 */
export function createGeneratedName(context, type, suffix) {
    return `${getCrossLanguageDefinitionId(context, type).split(".").at(-1)}${suffix}`;
}
export function isSubscriptionId(context, parameter) {
    return Boolean(context.arm) && parameter.name === "subscriptionId";
}
export function isNeverOrVoidType(type) {
    return isNeverType(type) || isVoidType(type);
}
export function getAnyType(context, type) {
    const diagnostics = createDiagnosticCollector();
    return diagnostics.wrap({
        kind: "unknown",
        name: "unknown",
        encode: "string",
        crossLanguageDefinitionId: "",
        decorators: diagnostics.pipe(getTypeDecorators(context, type)),
    });
}
export function getHttpOperationResponseHeaders(response) {
    const headers = response.headers ? Object.values(response.headers) : [];
    if (response.body?.contentTypeProperty) {
        headers.push(response.body.contentTypeProperty);
    }
    return headers;
}
export function removeVersionsLargerThanExplicitlySpecified(context, versions) {
    // filter with specific api version
    if (context.apiVersion !== undefined &&
        context.apiVersion !== "latest" &&
        context.apiVersion !== "all") {
        const index = versions.findIndex((version) => version.value === context.apiVersion);
        if (index >= 0) {
            versions.splice(index + 1, versions.length - index - 1);
        }
    }
}
export function filterPreviewVersion(context, sdkVersionsEnum, defaultApiVersion) {
    // if they explicitly set an api version, remove larger versions
    removeVersionsLargerThanExplicitlySpecified(context, sdkVersionsEnum.values);
    if (!context.previewStringRegex.test(defaultApiVersion)) {
        sdkVersionsEnum.values = sdkVersionsEnum.values.filter((v) => {
            if (typeof v.value !== "string") {
                return true;
            }
            // Check if the version has `@previewVersion` decorator
            if (v.__raw && v.__raw.kind === "EnumMember") {
                const enumMember = v.__raw;
                if (isPreviewVersion(context.program, enumMember)) {
                    return false;
                }
            }
            // Fall back to regex check for backward compatibility
            return !context.previewStringRegex.test(v.value);
        });
    }
}
export function twoParamsEquivalent(context, param1, param2) {
    if (!param1 || !param2) {
        return false;
    }
    return (param1.type === param2.type &&
        (param1.name === param2.name ||
            getParamAlias(context, param1) === param2.name ||
            param1.name === getParamAlias(context, param2)));
}
/**
 * If body is from spread, then it does not directly from a model property.
 * @param httpBody
 * @param parameters
 * @returns
 */
export function isHttpBodySpread(httpBody) {
    return httpBody.bodyKind !== "file" && httpBody.property === undefined;
}
/**
 * If body is from simple spread, then we use the original model as body model. Else we return the body type directly.
 * @param type
 * @returns
 */
export function getHttpBodyType(httpBody) {
    const type = httpBody.type;
    if (isHttpBodySpread(httpBody) && type.kind === "Model") {
        if (type.sourceModels.length === 1 && type.sourceModels[0].usage === "spread") {
            const innerModel = type.sourceModels[0].model;
            // for case: `op test(...Model):void;`
            if (innerModel.name !== "" && innerModel.properties.size === type.properties.size) {
                return innerModel;
            }
            // for case: `op test(@header h: string, @query q: string, ...Model): void;`
            if (innerModel.sourceModels.length === 1 &&
                innerModel.sourceModels[0].usage === "spread" &&
                innerModel.sourceModels[0].model.name !== "" &&
                innerModel.sourceModels[0].model.properties.size === type.properties.size) {
                return innerModel.sourceModels[0].model;
            }
        }
        return type;
    }
    return type;
}
export function isOnClient(context, type, operation, versioning) {
    const clientLocation = getClientLocation(context, type);
    if (operation &&
        clientLocation === (getOverriddenClientMethod(context, operation) ?? operation)) {
        // if the type has explicitly been moved to the operation, it is not on the client
        return false;
    }
    return (isSubscriptionId(context, type) ||
        (isApiVersion(context, type) && versioning) ||
        (operation !== undefined && getCorrespondingClientParam(context, type, operation) !== undefined));
}
export function getCorrespondingClientParam(context, type, operation) {
    const clientParams = [];
    let client = context.getClientForOperation(operation);
    while (client) {
        const clientParamsForClient = context.__clientParametersCache.get(client);
        if (clientParamsForClient) {
            clientParams.push(...clientParamsForClient);
        }
        if (client.kind === "SdkClient") {
            break;
        }
        client = client.parent;
    }
    const correspondingClientParam = clientParams?.find((x) => twoParamsEquivalent(context, x.__raw, type));
    if (correspondingClientParam)
        return correspondingClientParam;
    return undefined;
}
export function getValueTypeValue(value) {
    switch (value.valueKind) {
        case "ArrayValue":
            return value.values.map((x) => getValueTypeValue(x));
        case "BooleanValue":
        case "StringValue":
        case "NullValue":
            return value.value;
        case "NumericValue":
            return value.value.asNumber();
        case "EnumValue":
            return value.value.value ?? value.value.name;
        case "ObjectValue":
            return Object.fromEntries([...value.properties.keys()].map((x) => [
                x,
                getValueTypeValue(value.properties.get(x).value),
            ]));
        case "ScalarValue":
            // TODO: handle scalar value
            return undefined;
    }
}
export function hasNoneVisibility(context, type) {
    const lifecycle = getLifecycleVisibilityEnum(context.program);
    const visibility = getVisibilityForClass(context.program, type, lifecycle);
    return visibility.size === 0;
}
function listAllNamespaces(context, namespace, retval) {
    if (!retval) {
        retval = [];
    }
    if (retval.includes(namespace))
        return retval;
    retval.push(namespace);
    for (const ns of namespace.namespaces.values()) {
        listAllNamespaces(context, ns, retval);
    }
    return retval;
}
export function listAllUserDefinedNamespaces(context) {
    return listAllNamespaces(context, context.getMutatedGlobalNamespace()).filter((ns) => $(context.program).type.isUserDefined(ns));
}
export function findRootSourceProperty(property) {
    while (property.sourceProperty) {
        property = property.sourceProperty;
    }
    return property;
}
export function getStreamAsBytes(context, type) {
    const diagnostics = createDiagnosticCollector();
    const unknownType = {
        ...diagnostics.pipe(getSdkTypeBaseHelper(context, type, "bytes")),
        name: "bytes",
        encode: "bytes",
        crossLanguageDefinitionId: "",
    };
    return diagnostics.wrap(unknownType);
}
function getVersioningMutator(context, service, apiVersion) {
    const versionMutator = getVersioningMutators(context.program, service);
    if (!versionMutator)
        return undefined;
    if (versionMutator.kind === "transient") {
        return versionMutator.mutator;
    }
    const mutators = versionMutator.snapshots
        .filter((snapshot) => apiVersion === snapshot.version.value)
        .map((x) => x.mutator);
    compilerAssert(mutators.length === 1, "One api version should not get multiple mutators");
    return mutators[0];
}
export function handleVersioningMutationForGlobalNamespace(context) {
    const globalNamespace = context.program.getGlobalNamespaceType();
    const services = listServices(context.program);
    // No service, thus no versioning mutation needed
    if (services.length === 0)
        return globalNamespace;
    // Explicit all API version setting, thus no versioning mutation needed
    if (context.apiVersion === "all")
        return globalNamespace;
    const explicitClientNamespaces = [];
    const explicitServices = new Set();
    listScopedDecoratorData(context, clientKey).forEach((v, k) => {
        // See all explicit clients that in TypeSpec program
        if (!unsafe_Realm.realmForType.has(k)) {
            const sdkClient = v;
            explicitClientNamespaces.push(k);
            sdkClient.services.forEach((s) => explicitServices.add(s));
        }
    });
    let mutator;
    // No explicit clients (choose first service) or explicit client with one service
    if (explicitClientNamespaces.length === 0 || explicitServices.size === 1) {
        const serviceNamespace = explicitClientNamespaces.length === 0
            ? services[0].type
            : explicitServices.values().next().value;
        const versions = getVersions(context.program, serviceNamespace)[1]?.getVersions();
        // If the single service has no versioning, no mutation needed
        if (!versions)
            return globalNamespace;
        // Filter versions based on `apiVersion` config
        removeVersionsLargerThanExplicitlySpecified(context, versions);
        const versionsValues = versions.map((v) => v.value);
        // Fix apiVersion setting problem only if there's only one service
        if (context.apiVersion !== undefined &&
            context.apiVersion !== "latest" &&
            context.apiVersion !== "all" &&
            !versionsValues.includes(context.apiVersion)) {
            reportDiagnostic(context.program, {
                code: "api-version-undefined",
                format: { version: context.apiVersion },
                target: services[0].type,
            });
            context.apiVersion = versionsValues[versionsValues.length - 1];
        }
        mutator = getVersioningMutator(context, serviceNamespace, versionsValues[versionsValues.length - 1]);
    }
    // Explicit clients with multiple services
    else {
        // Currently we do not support multiple explicit clients with multiple services
        if (explicitClientNamespaces.length > 1 && explicitServices.size > 1) {
            reportDiagnostic(context.program, {
                code: "multiple-explicit-clients-multiple-services",
                format: {},
                target: services[0].type,
            });
            return globalNamespace;
        }
        mutator = getVersioningMutator(context, explicitClientNamespaces[0]);
    }
    if (!mutator)
        return globalNamespace;
    const subgraph = unsafe_mutateSubgraphWithNamespace(context.program, [mutator], globalNamespace);
    compilerAssert(subgraph.type.kind === "Namespace", "Should not have mutated to another type");
    compilerAssert(subgraph.realm !== null, "Should have a realm after mutation");
    context.__mutatedRealm = subgraph.realm;
    return subgraph.type;
}
export function resolveDuplicateGenearatedName(context, type, createName) {
    let duplicateCount = 1;
    const rawCreateName = createName;
    const generatedNames = [...context.__generatedNames.values()];
    while (generatedNames.includes(createName)) {
        createName = `${rawCreateName}${duplicateCount++}`;
    }
    context.__generatedNames.set(type, createName);
    return createName;
}
export function resolveConflictGeneratedName(context) {
    const userDefinedNames = [...context.__referencedTypeCache.values()]
        .filter((x) => !x.isGeneratedName)
        .map((x) => x.name);
    const generatedNames = [...context.__generatedNames.values()];
    for (const sdkType of context.__referencedTypeCache.values()) {
        if (sdkType.__raw && sdkType.isGeneratedName && userDefinedNames.includes(sdkType.name)) {
            const rawName = sdkType.name;
            let duplicateCount = 1;
            let createName = `${rawName}${duplicateCount++}`;
            while (userDefinedNames.includes(createName) || generatedNames.includes(createName)) {
                createName = `${rawName}${duplicateCount++}`;
            }
            sdkType.name = createName;
            context.__generatedNames.set(sdkType.__raw, createName);
            generatedNames.push(createName);
        }
    }
}
export function getClientDoc(context, target) {
    const clientDocExplicit = getClientDocExplicit(context, target);
    const baseDoc = getDoc(context.program, target);
    if (clientDocExplicit) {
        switch (clientDocExplicit.mode) {
            case "append":
                return baseDoc
                    ? `${baseDoc}\n${clientDocExplicit.documentation}`
                    : clientDocExplicit.documentation;
            case "replace":
                return clientDocExplicit.documentation;
        }
    }
    return baseDoc;
}
export function compareModelProperties(context, modelPropA, modelPropB) {
    if (!modelPropA || !modelPropB)
        return false;
    if (modelPropA.name !== modelPropB.name || modelPropA.type !== modelPropB.type)
        return false;
    if (!context)
        return true; // if we don't have a context, we can't further compare the types. Assume true.
    // compare serialized names if they are http parameters
    if (!isSdkHttpParameter(context, modelPropA) || !isSdkHttpParameter(context, modelPropB))
        return true;
    const sdkA = ignoreDiagnostics(getSdkHttpParameter(context, modelPropA));
    const sdkB = ignoreDiagnostics(getSdkHttpParameter(context, modelPropB));
    return sdkA.kind === sdkB.kind && sdkA.serializedName === sdkB.serializedName;
}
export function* filterMapValuesIterator(iterator, predicate) {
    for (const value of iterator) {
        if (predicate(value)) {
            yield value;
        }
    }
}
/**
 * Find all entries in a scoped decorator state map where the target matches a specific value
 */
export function findEntriesWithTarget(context, stateKey, targetValue, sourceKind) {
    const results = [];
    for (const [type, target] of listScopedDecoratorData(context, stateKey)) {
        if (sourceKind && type.kind !== sourceKind) {
            continue;
        }
        if (target === targetValue) {
            results.push(type);
        }
    }
    return results;
}
/**
 * Retrieves Long Running Operation (LRO) metadata for a given operation.
 *
 * This function serves as a wrapper that:
 * 1. First tries to get LRO metadata using the `getLroMetadata` function from the Azure Core library
 * 2. If unavailable or undefined, it would check for the existence of a `getMarkAsLro` function
 *    and return a mock LRO metadata object if the operation is marked as LRO
 *
 * @param context - The TypeSpec client generator context
 * @param operation - The TypeSpec operation to check for LRO metadata
 * @returns The LRO metadata for the operation if available, otherwise undefined
 */
export function getTcgcLroMetadata(context, operation, client) {
    const lroMetaData = getLroMetadata(context.program, operation);
    if (lroMetaData) {
        return lroMetaData;
    }
    if (getMarkAsLro(context, operation)) {
        // we guard against this in the setting of `@markAsLro`
        const sdkMethod = ignoreDiagnostics(getSdkBasicServiceMethod(context, operation, client));
        let returnType;
        const sdkMethodResponseType = sdkMethod.response.type;
        switch (sdkMethodResponseType.kind) {
            case "nullable":
                returnType = sdkMethodResponseType.type.__raw;
                break;
            case "model":
                returnType = sdkMethodResponseType.__raw;
                break;
            default:
                throw new Error(`LRO method ${operation.name} with @markAsLro must have a model return type.`);
        }
        return {
            operation,
            logicalResult: returnType,
            finalStateVia: FinalStateValue.location,
            pollingInfo: {
                kind: "pollingOperationStep",
                responseModel: returnType,
                terminationStatus: {
                    kind: "status-code",
                },
            },
            envelopeResult: returnType,
            finalEnvelopeResult: returnType,
            finalResult: returnType,
        };
    }
    return undefined;
}
export function getActualClientType(client) {
    if (client.kind === "SdkClient")
        return client.type;
    if (client.type)
        return client.type;
    // Created operation group from `@clientLocation`. May have single or multiple services. Choose the first service for multi-service case.
    return client.services[0];
}
export function isSameServers(left, right) {
    if (left.length !== right.length) {
        return false;
    }
    for (let i = 0; i < left.length; i++) {
        if (left[i].url !== right[i].url) {
            return false;
        }
    }
    return true;
}
export function isSameAuth(left, right) {
    if (left.options.length !== right.options.length) {
        return false;
    }
    for (let i = 0; i < left.options.length; i++) {
        if (left.options[i].schemes.length !== right.options[i].schemes.length) {
            return false;
        }
        for (let j = 0; j < left.options[i].schemes.length; j++) {
            const leftScheme = left.options[i].schemes[j];
            const rightScheme = right.options[i].schemes[j];
            if (leftScheme.type !== rightScheme.type) {
                return false;
            }
            switch (leftScheme.type) {
                case "http":
                    assert(rightScheme.type === "http");
                    if (leftScheme.scheme !== rightScheme.scheme) {
                        return false;
                    }
                    break;
                case "apiKey":
                    assert(rightScheme.type === "apiKey");
                    if (leftScheme.name !== rightScheme.name || leftScheme.in !== rightScheme.in) {
                        return false;
                    }
                    break;
                case "oauth2":
                    assert(rightScheme.type === "oauth2");
                    if (leftScheme.flows.length !== rightScheme.flows.length) {
                        return false;
                    }
                    for (let k = 0; k < leftScheme.flows.length; k++) {
                        const leftFlow = leftScheme.flows[k];
                        const rightFlow = rightScheme.flows[k];
                        if (leftFlow.type !== rightFlow.type) {
                            return false;
                        }
                        if (leftFlow.scopes.length !== rightFlow.scopes.length) {
                            return false;
                        }
                        for (let l = 0; l < leftFlow.scopes.length; l++) {
                            if (leftFlow.scopes[l].value !== rightFlow.scopes[l].value) {
                                return false;
                            }
                        }
                        switch (leftFlow.type) {
                            case "authorizationCode":
                                assert(rightFlow.type === "authorizationCode");
                                if (leftFlow.authorizationUrl !== rightFlow.authorizationUrl ||
                                    leftFlow.tokenUrl !== rightFlow.tokenUrl ||
                                    leftFlow.refreshUrl !== rightFlow.refreshUrl) {
                                    return false;
                                }
                                break;
                            case "clientCredentials":
                                assert(rightFlow.type === "clientCredentials");
                                if (leftFlow.tokenUrl !== rightFlow.tokenUrl ||
                                    leftFlow.refreshUrl !== rightFlow.refreshUrl) {
                                    return false;
                                }
                                break;
                            case "implicit":
                                assert(rightFlow.type === "implicit");
                                if (leftFlow.authorizationUrl !== rightFlow.authorizationUrl ||
                                    leftFlow.refreshUrl !== rightFlow.refreshUrl) {
                                    return false;
                                }
                                break;
                            case "password":
                                assert(rightFlow.type === "password");
                                if (leftFlow.authorizationUrl !== rightFlow.authorizationUrl ||
                                    leftFlow.refreshUrl !== rightFlow.refreshUrl) {
                                    return false;
                                }
                                break;
                        }
                    }
                    break;
                case "openIdConnect":
                    assert(rightScheme.type === "openIdConnect");
                    if (leftScheme.openIdConnectUrl !== rightScheme.openIdConnectUrl) {
                        return false;
                    }
                    break;
            }
        }
    }
    return true;
}
