// @azure-tools/typespec-client-generator-core/dist/src/methods.js
import { getParameterizedNextLinkArguments, } from "@azure-tools/typespec-azure-core";
import { compilerAssert, createDiagnosticCollector, getSummary, ignoreDiagnostics, isList, } from "@typespec/compiler";
import { $ } from "@typespec/compiler/typekit";
import { getAccess, getDisablePageable, getMarkAsPageable, getNextLinkVerb, getOverriddenClientMethod, getResponseAsBool, listOperationsInOperationGroup, shouldGenerateConvenient, shouldGenerateProtocol, } from "./decorators.js";
import { getSdkHttpOperation } from "./http.js";
import { UsageFlags, } from "./interfaces.js";
import { createGeneratedName, findRootSourceProperty, getActualClientType, getAvailableApiVersions, getClientDoc, getCorrespondingClientParam, getHashForType, getTcgcLroMetadata, getTypeDecorators, isNeverOrVoidType, isSubscriptionId, } from "./internal-utils.js";
import { createDiagnostic } from "./lib.js";
import { getCrossLanguageDefinitionId, getHttpOperationWithCache, getLibraryName, } from "./public-utils.js";
import { getClientTypeWithDiagnostics, getSdkBuiltInType, getSdkModel, getSdkModelPropertyType, getSdkModelPropertyTypeBase, getSdkModelWithDiagnostics, } from "./types.js";
function getSdkServiceOperation(context, operation, methodParameters, client) {
    const diagnostics = createDiagnosticCollector();
    const httpOperation = getHttpOperationWithCache(context, operation);
    if (httpOperation) {
        const sdkHttpOperation = diagnostics.pipe(getSdkHttpOperation(context, httpOperation, methodParameters, client));
        return diagnostics.wrap(sdkHttpOperation);
    }
    diagnostics.add(createDiagnostic({
        code: "unsupported-protocol",
        target: operation,
        format: {},
    }));
    return diagnostics.wrap(undefined);
}
function getSdkLroPagingServiceMethod(context, operation, client) {
    const diagnostics = createDiagnosticCollector();
    return diagnostics.wrap({
        ...diagnostics.pipe(getSdkLroServiceMethod(context, operation, client)),
        ...diagnostics.pipe(getSdkPagingServiceMethod(context, operation, client)),
        kind: "lropaging",
    });
}
function getPageSizeParameterSegments(baseServiceMethod) {
    function recurseToFindPageSizeParameterInModel(param, model) {
        for (const prop of model.properties) {
            if (prop.__raw && prop.__raw.decorators.find((d) => d.definition?.name === "@pageSize")) {
                return [param, prop];
            }
            if (prop.type.kind === "model") {
                const nested = recurseToFindPageSizeParameterInModel(param, prop.type);
                if (nested.length > 0) {
                    return nested;
                }
            }
        }
        return [];
    }
    for (const p of baseServiceMethod.parameters) {
        if (p.__raw && p.__raw.decorators.find((d) => d.definition?.name === "@pageSize")) {
            return [p];
        }
        if (p.type.kind === "model") {
            return recurseToFindPageSizeParameterInModel(p, p.type);
        }
    }
    return [];
}
function getSdkPagingServiceMethod(context, operation, client) {
    const diagnostics = createDiagnosticCollector();
    const baseServiceMethod = diagnostics.pipe(getSdkBasicServiceMethod(context, operation, client));
    // If the response body type itself is nullable (e.g., {@body body: Type | null}), unwrap it for paging/LRO processing
    let responseType = baseServiceMethod.response.type;
    if (responseType?.kind === "nullable") {
        responseType = responseType.type;
    }
    // normal paging
    if (isList(context.program, operation)) {
        const pagingMetadata = $(context.program).operation.getPagingMetadata(getOverriddenClientMethod(context, operation) ?? operation);
        if (responseType?.__raw?.kind !== "Model" || responseType.kind !== "model" || !pagingMetadata) {
            diagnostics.add(createDiagnostic({
                code: "unexpected-pageable-operation-return-type",
                target: operation,
                format: {
                    operationName: operation.name,
                },
            }));
            // return as page method with no paging info
            return diagnostics.wrap({
                ...baseServiceMethod,
                kind: "paging",
                pagingMetadata: {},
            });
        }
        const resultSegments = mapFirstSegmentForResultSegments(pagingMetadata.output.pageItems.path, baseServiceMethod.response);
        const nextLinkSegments = mapFirstSegmentForResultSegments(pagingMetadata.output.nextLink?.path, baseServiceMethod.response);
        const continuationTokenResponseSegments = mapFirstSegmentForResultSegments(pagingMetadata.output.continuationToken?.path, baseServiceMethod.response);
        baseServiceMethod.response.resultSegments = resultSegments?.map((resultSegment) => context.__modelPropertyCache.get(resultSegment));
        context.__pagedResultSet.add(responseType);
        // tcgc will let all paging method return a list of items
        baseServiceMethod.response.type = diagnostics.pipe(getClientTypeWithDiagnostics(context, pagingMetadata.output.pageItems.property.type, operation));
        return diagnostics.wrap({
            ...baseServiceMethod,
            kind: "paging",
            pagingMetadata: {
                __raw: pagingMetadata,
                nextLinkSegments: nextLinkSegments?.map((segment) => context.__responseHeaderCache.get(segment) ??
                    context.__modelPropertyCache.get(segment)),
                nextLinkVerb: getNextLinkVerb(context, operation),
                continuationTokenParameterSegments: pagingMetadata.input.continuationToken?.path.map((r) => context.__methodParameterCache.get(r) ?? context.__modelPropertyCache.get(r)),
                continuationTokenResponseSegments: continuationTokenResponseSegments?.map((segment) => context.__responseHeaderCache.get(segment) ??
                    context.__modelPropertyCache.get(segment)),
                pageItemsSegments: baseServiceMethod.response.resultSegments,
                pageSizeParameterSegments: getPageSizeParameterSegments(baseServiceMethod),
                nextLinkReInjectedParametersSegments: pagingMetadata.output.nextLink?.property.type.kind === "Scalar"
                    ? (getParameterizedNextLinkArguments(context.program, pagingMetadata.output.nextLink.property.type) ?? []).map((t) => getPropertySegmentsFromModelOrParameters(baseServiceMethod.parameters, (p) => p.__raw?.kind === "ModelProperty" &&
                        findRootSourceProperty(p.__raw) === findRootSourceProperty(t)))
                    : undefined,
            },
        });
    }
    else {
        const markAsPageableInfo = getMarkAsPageable(context, operation);
        if (markAsPageableInfo) {
            const itemsProperty = diagnostics.pipe(getSdkModelPropertyType(context, markAsPageableInfo.itemsProperty, operation));
            // Set resultSegments to match the behavior of normal paging operations
            baseServiceMethod.response.resultSegments = [itemsProperty];
            if (responseType) {
                context.__pagedResultSet.add(responseType);
            }
            // tcgc will let all paging method return a list of items
            baseServiceMethod.response.type = diagnostics.pipe(getClientTypeWithDiagnostics(context, markAsPageableInfo.itemsProperty.type, operation));
            return diagnostics.wrap({
                ...baseServiceMethod,
                kind: "paging",
                pagingMetadata: {
                    __raw: undefined, // because in this case it is not a real paging operation
                    pageItemsSegments: baseServiceMethod.response.resultSegments,
                },
            });
        }
        else {
            compilerAssert(false, "Unexpected operation should be paged if calling this function");
        }
    }
}
function mapFirstSegmentForResultSegments(resultSegments, response) {
    if (resultSegments === undefined || response === undefined)
        return undefined;
    // TCGC use Http response type as the return type
    // For implicit body response, we need to locate the first segment in the response type
    // Several cases:
    // 1. `op test(): {items, nextLink}`
    // 2. `op test(): {items, nextLink} & {a, b, c}`
    // 3. `op test(): {@bodyRoot body: {items, nextLink}}`
    const responseModel = response.type?.kind === "model"
        ? response.type
        : response.type?.kind === "nullable" && response.type.type.kind === "model"
            ? response.type.type
            : undefined;
    if (resultSegments.length > 0 && responseModel) {
        for (let i = 0; i < resultSegments.length; i++) {
            const segment = resultSegments[i];
            let current = responseModel;
            while (current) {
                for (const property of current.properties ?? []) {
                    if (property.__raw &&
                        findRootSourceProperty(property.__raw) === findRootSourceProperty(segment)) {
                        return [property.__raw, ...resultSegments.slice(i + 1)];
                    }
                }
                current = current.baseModel;
            }
        }
    }
    return resultSegments;
}
export function getPropertySegmentsFromModelOrParameters(source, predicate) {
    const queue = [];
    if (!Array.isArray(source)) {
        if (source.baseModel) {
            const baseResult = getPropertySegmentsFromModelOrParameters(source.baseModel, predicate);
            if (baseResult)
                return baseResult;
        }
    }
    for (const prop of Array.isArray(source) ? source : source.properties.values()) {
        if (predicate(prop)) {
            return [prop];
        }
        if (prop.type.kind === "model") {
            queue.push({ model: prop.type, path: [prop] });
        }
    }
    while (queue.length > 0) {
        const { model, path } = queue.shift();
        for (const prop of model.properties.values()) {
            if (predicate(prop)) {
                return path.concat(prop);
            }
            if (prop.type.kind === "model") {
                queue.push({ model: prop.type, path: path.concat(prop) });
            }
        }
    }
    return undefined;
}
function getSdkLroServiceMethod(context, operation, client) {
    const diagnostics = createDiagnosticCollector();
    const metadata = diagnostics.pipe(getServiceMethodLroMetadata(context, operation, client));
    const baseServiceMethod = diagnostics.pipe(getSdkBasicServiceMethod(context, operation, client));
    baseServiceMethod.response.type = metadata.finalResponse?.result;
    baseServiceMethod.response.resultSegments = metadata.finalResponse?.resultSegments;
    return diagnostics.wrap({
        ...baseServiceMethod,
        kind: "lro",
        __raw_lro_metadata: metadata.__raw,
        lroMetadata: metadata,
        operation: diagnostics.pipe(getSdkServiceOperation(context, metadata.__raw.operation, baseServiceMethod.parameters, client)),
    });
}
function getServiceMethodLroMetadata(context, operation, client) {
    const diagnostics = createDiagnosticCollector();
    const rawMetadata = getTcgcLroMetadata(context, operation, client);
    if (rawMetadata === undefined) {
        return diagnostics.wrap(undefined);
    }
    const finalResponse = getFinalResponse();
    return diagnostics.wrap({
        __raw: rawMetadata,
        finalStateVia: rawMetadata.finalStateVia,
        finalResponse,
        finalStep: getSdkLroServiceFinalStep(context, rawMetadata.finalStep),
        pollingStep: {
            responseBody: rawMetadata.pollingInfo.responseModel
                ? diagnostics.pipe(getClientTypeWithDiagnostics(context, rawMetadata.pollingInfo.responseModel))
                : undefined,
        },
        operation: diagnostics.pipe(getSdkBasicServiceMethod(context, rawMetadata.operation, client))
            .operation,
        logicalResult: diagnostics.pipe(getSdkModelWithDiagnostics(context, rawMetadata.logicalResult)),
        statusMonitorStep: getStatusMonitorStep(context, rawMetadata.statusMonitorStep),
        pollingInfo: getPollingInfo(context, rawMetadata.pollingInfo),
        envelopeResult: diagnostics.pipe(getSdkModelWithDiagnostics(context, rawMetadata.envelopeResult)),
        logicalPath: rawMetadata.logicalPath,
        finalEnvelopeResult: finalResponse ? finalResponse.envelopeResult : "void",
        finalResultPath: rawMetadata.finalResultPath,
    });
    function getSdkLroServiceFinalStep(context, step) {
        if (!step)
            return undefined;
        switch (step.kind) {
            case "finalOperationLink": {
                return {
                    kind: "finalOperationLink",
                    target: getSdkOperationLink(context, step.target),
                };
            }
            case "finalOperationReference": {
                return {
                    kind: "finalOperationReference",
                    target: getSdkOperationReference(context, step.target, client),
                };
            }
            case "pollingSuccessProperty": {
                return {
                    kind: "pollingSuccessProperty",
                    responseModel: getSdkModel(context, step.responseModel),
                    target: diagnostics.pipe(getSdkModelPropertyType(context, step.target)),
                    sourceProperty: step.sourceProperty
                        ? diagnostics.pipe(getSdkModelPropertyType(context, step.sourceProperty))
                        : undefined,
                };
            }
            case "noPollingResult": {
                return {
                    kind: "noPollingResult",
                    responseModel: undefined,
                };
            }
        }
    }
    function getStatusMonitorStep(context, statusMonitorStep) {
        if (!statusMonitorStep)
            return undefined;
        if (statusMonitorStep.kind === "nextOperationLink") {
            return {
                kind: "nextOperationLink",
                responseModel: diagnostics.pipe(getSdkModelWithDiagnostics(context, statusMonitorStep.responseModel)),
                target: getSdkOperationLink(context, statusMonitorStep.target),
            };
        }
        return {
            kind: "nextOperationReference",
            responseModel: diagnostics.pipe(getSdkModelWithDiagnostics(context, statusMonitorStep.responseModel)),
            target: getSdkOperationReference(context, statusMonitorStep.target, client),
        };
    }
    function getSdkOperationLink(context, link) {
        return {
            kind: "link",
            location: link.location,
            property: diagnostics.pipe(getSdkModelPropertyType(context, link.property)),
        };
    }
    function getSdkOperationReference(context, reference, client) {
        const parameters = new Map();
        for (const [key, p] of reference.parameters?.entries() ?? []) {
            parameters.set(key, {
                sourceKind: p.sourceKind,
                source: diagnostics.pipe(getSdkModelPropertyType(context, p.source)),
                target: diagnostics.pipe(getSdkModelPropertyType(context, p.target)),
            });
        }
        return {
            kind: "reference",
            // since these operations may not be included in the client customization, we ignore diagnostics here
            operation: ignoreDiagnostics(getSdkBasicServiceMethod(context, reference.operation, client))
                .operation,
            parameterMap: reference.parameterMap,
            parameters,
            link: reference.link ? getSdkOperationLink(context, reference.link) : undefined,
        };
    }
    function getPollingInfo(context, pollingInfo) {
        const resultProperty = pollingInfo.resultProperty
            ? diagnostics.pipe(getSdkModelPropertyType(context, pollingInfo.resultProperty))
            : undefined;
        const errorProperty = pollingInfo.errorProperty
            ? diagnostics.pipe(getSdkModelPropertyType(context, pollingInfo.errorProperty))
            : undefined;
        return {
            kind: "pollingOperationStep",
            responseModel: diagnostics.pipe(getSdkModelWithDiagnostics(context, pollingInfo.responseModel)),
            terminationStatus: getTerminationStatus(context, pollingInfo.terminationStatus),
            resultProperty,
            errorProperty,
        };
    }
    function getTerminationStatus(context, terminationStatus) {
        switch (terminationStatus.kind) {
            case "status-code":
                return terminationStatus;
            case "model-property":
                return {
                    ...terminationStatus,
                    property: diagnostics.pipe(getSdkModelPropertyType(context, terminationStatus.property)),
                };
        }
    }
    function getFinalResponse() {
        if (rawMetadata?.finalEnvelopeResult === undefined ||
            rawMetadata.finalEnvelopeResult === "void" ||
            rawMetadata.finalResult === undefined ||
            rawMetadata.finalResult === "void") {
            return undefined;
        }
        const envelopeResult = diagnostics.pipe(getClientTypeWithDiagnostics(context, rawMetadata.finalEnvelopeResult));
        const result = diagnostics.pipe(getClientTypeWithDiagnostics(context, rawMetadata.finalResult));
        // find the property inside the envelope result using the final result path
        let sdkProperty = undefined;
        if (envelopeResult.kind === "model") {
            const resultPath = rawMetadata.finalResultPath;
            for (const property of envelopeResult.properties) {
                if (property.__raw === undefined || property.kind !== "property") {
                    continue;
                }
                if (property.__raw?.name === resultPath) {
                    sdkProperty = property;
                    break;
                }
            }
        }
        return {
            envelopeResult,
            result,
            resultSegments: sdkProperty !== undefined ? [sdkProperty] : undefined,
        };
    }
}
function getSdkMethodResponse(context, operation, sdkOperation, client) {
    const responses = sdkOperation.responses;
    const allResponseBodies = [];
    let containsResponseWithoutBody = false;
    responses.forEach((response) => {
        if (response.type) {
            allResponseBodies.push(response.type);
        }
        else {
            containsResponseWithoutBody = true;
        }
    });
    const responseTypes = new Set(allResponseBodies.map((x) => getHashForType(x)));
    let type = undefined;
    if (getResponseAsBool(context, operation)) {
        type = getSdkBuiltInType(context, $(context.program).builtin.boolean);
    }
    else {
        if (responseTypes.size > 1) {
            // return union of all the different types
            type = {
                __raw: operation,
                kind: "union",
                access: "public",
                usage: UsageFlags.Output,
                variantTypes: allResponseBodies,
                name: createGeneratedName(context, operation, "UnionResponse"),
                isGeneratedName: true,
                namespace: client.namespace,
                crossLanguageDefinitionId: `${getCrossLanguageDefinitionId(context, operation)}.UnionResponse`,
                decorators: [],
            };
        }
        else if (responseTypes.size === 1) {
            type = allResponseBodies[0];
        }
    }
    // Set optional property based on whether responses have bodies
    // If type is undefined (no response), optional remains undefined
    let optional = undefined;
    if (type !== undefined) {
        // If we have a response type, set optional based on whether some responses lack bodies
        optional = containsResponseWithoutBody;
    }
    return {
        kind: "method",
        type,
        ...(optional !== undefined && { optional }),
    };
}
export function getSdkBasicServiceMethod(context, operation, client) {
    const diagnostics = createDiagnosticCollector();
    const methodParameters = [];
    // we have to calculate apiVersions first, so that the information is put
    // in __tspTypeToApiVersions before we call parameters since method wraps parameter
    const apiVersions = getAvailableApiVersions(context, operation, getActualClientType(client.__raw));
    let clientParams = context.__clientParametersCache.get(client.__raw);
    if (!clientParams) {
        clientParams = [];
        context.__clientParametersCache.set(client.__raw, clientParams);
    }
    const override = getOverriddenClientMethod(context, operation);
    const params = (override ?? operation).parameters.properties.values();
    for (const param of params) {
        if (isNeverOrVoidType(param.type))
            continue;
        const sdkMethodParam = diagnostics.pipe(getSdkMethodParameter(context, param, operation));
        if (sdkMethodParam.onClient) {
            // add API version and subscription ID parameters to the client parameters
            if (sdkMethodParam.isApiVersionParam) {
                if (!clientParams.find((x) => x.isApiVersionParam)) {
                    clientParams.push(sdkMethodParam);
                }
            }
            else if (isSubscriptionId(context, param)) {
                if (!clientParams.find((x) => isSubscriptionId(context, x))) {
                    clientParams.push(sdkMethodParam);
                }
            }
        }
        else {
            methodParameters.push(sdkMethodParam);
        }
    }
    const serviceOperation = diagnostics.pipe(getSdkServiceOperation(context, operation, methodParameters, client));
    const response = getSdkMethodResponse(context, operation, serviceOperation, client);
    const name = getLibraryName(context, operation);
    return diagnostics.wrap({
        __raw: operation,
        kind: "basic",
        name,
        access: getAccess(context, operation) ?? "public",
        parameters: methodParameters,
        doc: getClientDoc(context, operation),
        summary: getSummary(context.program, operation),
        operation: serviceOperation,
        response,
        apiVersions,
        crossLanguageDefinitionId: getCrossLanguageDefinitionId(context, operation),
        decorators: diagnostics.pipe(getTypeDecorators(context, operation)),
        generateConvenient: shouldGenerateConvenient(context, operation),
        generateProtocol: shouldGenerateProtocol(context, operation),
        isOverride: override !== undefined,
    });
}
function getSdkServiceMethod(context, operation, client) {
    const lro = getTcgcLroMetadata(context, operation, client);
    // `@disablePageable` disables paging even for operations with @list
    const pagingDisabled = getDisablePageable(context, operation);
    const paging = !pagingDisabled &&
        (isList(context.program, operation) || getMarkAsPageable(context, operation));
    if (lro && paging) {
        return getSdkLroPagingServiceMethod(context, operation, client);
    }
    else if (paging) {
        return getSdkPagingServiceMethod(context, operation, client);
    }
    else if (lro) {
        return getSdkLroServiceMethod(context, operation, client);
    }
    return getSdkBasicServiceMethod(context, operation, client);
}
export function getSdkMethodParameter(context, type, operation) {
    const diagnostics = createDiagnosticCollector();
    let property = context.__methodParameterCache?.get(type);
    if (!property) {
        // for parameter that has elevated to client or parent client, we will use the client parameter directly
        if (operation) {
            const correspondingClientParam = getCorrespondingClientParam(context, type, operation);
            if (correspondingClientParam)
                return diagnostics.wrap(correspondingClientParam);
        }
        property = {
            ...diagnostics.pipe(getSdkModelPropertyTypeBase(context, type, operation)),
            kind: "method",
        };
        context.__methodParameterCache.set(type, property);
    }
    return diagnostics.wrap(property);
}
export function createSdkMethods(context, client, sdkClientType) {
    const diagnostics = createDiagnosticCollector();
    const retval = [];
    for (const operation of listOperationsInOperationGroup(context, client)) {
        retval.push(diagnostics.pipe(getSdkServiceMethod(context, operation, sdkClientType)));
    }
    return diagnostics.wrap(retval);
}
