// @azure-tools/typespec-client-generator-core/dist/src/package.js
import { createDiagnosticCollector, getNamespaceFullName, ignoreDiagnostics, } from "@typespec/compiler";
import { prepareClientAndOperationCache } from "./cache.js";
import { createSdkClientType } from "./clients.js";
import { listClients } from "./decorators.js";
import { filterApiVersionsWithDecorators, getActualClientType, getTypeDecorators, } from "./internal-utils.js";
import { getLicenseInfo } from "./license.js";
import { getCrossLanguagePackageId, getNamespaceFromType } from "./public-utils.js";
import { getAllReferencedTypes, handleAllTypes } from "./types.js";
export function createSdkPackage(context) {
    const diagnostics = createDiagnosticCollector();
    populateApiVersionInformation(context);
    diagnostics.pipe(handleAllTypes(context));
    const crossLanguagePackageId = diagnostics.pipe(getCrossLanguagePackageId(context));
    const allReferencedTypes = getAllReferencedTypes(context);
    const versions = context.getPackageVersions();
    // Create apiVersions map for multiple services
    const apiVersionsMap = new Map();
    for (const [namespace, versionList] of versions.entries()) {
        const fullName = getNamespaceFullName(namespace);
        const latestVersion = versionList.at(-1);
        if (latestVersion) {
            // When apiVersion config is "all" for single service, store "all" in the map as well
            const versionValue = context.apiVersion === "all" && versions.size === 1 ? "all" : latestVersion;
            apiVersionsMap.set(fullName, versionValue);
        }
    }
    const sdkPackage = {
        clients: listClients(context).map((c) => diagnostics.pipe(createSdkClientType(context, c))),
        models: allReferencedTypes.filter((x) => x.kind === "model"),
        enums: allReferencedTypes.filter((x) => x.kind === "enum"),
        unions: allReferencedTypes.filter((x) => x.kind === "union" || x.kind === "nullable"),
        crossLanguagePackageId,
        namespaces: [],
        licenseInfo: getLicenseInfo(context),
        metadata: {
            apiVersion: context.apiVersion === "all" && versions.size === 1
                ? "all"
                : versions.size === 1
                    ? [...versions.values()][0].at(-1)
                    : undefined,
            apiVersions: apiVersionsMap,
        },
    };
    organizeNamespaces(context, sdkPackage);
    return diagnostics.wrap(sdkPackage);
}
function organizeNamespaces(context, sdkPackage) {
    const clients = [...sdkPackage.clients];
    while (clients.length > 0) {
        const client = clients.shift();
        getSdkNamespace(context, sdkPackage, client).clients.push(client);
        if (client.children && client.children.length > 0) {
            clients.push(...client.children);
        }
    }
    for (const model of sdkPackage.models) {
        getSdkNamespace(context, sdkPackage, model).models.push(model);
    }
    for (const enumType of sdkPackage.enums) {
        getSdkNamespace(context, sdkPackage, enumType).enums.push(enumType);
    }
    for (const unionType of sdkPackage.unions) {
        getSdkNamespace(context, sdkPackage, unionType).unions.push(unionType);
    }
}
function getSdkNamespace(context, sdkPackage, type) {
    if (!("namespace" in type)) {
        return sdkPackage;
    }
    const namespace = type.namespace;
    const segments = namespace.split(".");
    let current = sdkPackage;
    let fullName = "";
    for (const segment of segments) {
        fullName = fullName === "" ? segment : `${fullName}.${segment}`;
        const ns = current.namespaces.find((ns) => ns.name === segment);
        if (ns === undefined) {
            const rawNamespace = getNamespaceFromType(type.__raw);
            const newNs = {
                __raw: rawNamespace,
                name: segment,
                fullName,
                clients: [],
                models: [],
                enums: [],
                unions: [],
                namespaces: [],
                decorators: rawNamespace ? ignoreDiagnostics(getTypeDecorators(context, rawNamespace)) : [],
            };
            current.namespaces.push(newNs);
            current = newNs;
        }
        else {
            current = ns;
        }
    }
    return current;
}
function populateApiVersionInformation(context) {
    if (context.__rawClientsOperationGroupsCache === undefined) {
        prepareClientAndOperationCache(context);
    }
    // Get the package versions map once (this handles both single and multi-service scenarios)
    const packageVersions = context.getPackageVersions();
    for (const client of context.__rawClientsOperationGroupsCache.values()) {
        const clientType = getActualClientType(client);
        // Multiple service case. Set empty result.
        if (client.services.length > 1) {
            context.setApiVersionsForType(clientType, []);
            context.__clientApiVersionDefaultValueCache.set(client, undefined);
        }
        else {
            const versions = filterApiVersionsWithDecorators(context, clientType, packageVersions.get(client.services[0]) || []);
            context.setApiVersionsForType(clientType, versions);
            context.__clientApiVersionDefaultValueCache.set(client, versions[versions.length - 1]);
        }
    }
}
