// @azure-tools/typespec-client-generator-core/dist/src/tsp-index.js
import { $access, $alternateType, $apiVersion, $client, $clientApiVersions, $clientDefaultValue, $clientDoc, $clientInitialization, $clientLocation, $clientName, $clientNamespace, $clientOption, $convenientAPI, $deserializeEmptyStringAsNull, $disablePageable, $flattenProperty, $legacyHierarchyBuilding, $markAsLro, $markAsPageable, $nextLinkVerb, $operationGroup, $override, $paramAlias, $protocolAPI, $responseAsBool, $scope, $usage, $useSystemTextJsonConverter, } from "./decorators.js";
export { $lib } from "./lib.js";
export { $onValidate } from "./validate.js";
/** @internal */
export const $decorators = {
    "Azure.ClientGenerator.Core": {
        clientName: $clientName,
        convenientAPI: $convenientAPI,
        protocolAPI: $protocolAPI,
        client: $client,
        operationGroup: $operationGroup,
        usage: $usage,
        access: $access,
        override: $override,
        useSystemTextJsonConverter: $useSystemTextJsonConverter,
        clientInitialization: $clientInitialization,
        paramAlias: $paramAlias,
        apiVersion: $apiVersion,
        clientNamespace: $clientNamespace,
        alternateType: $alternateType,
        scope: $scope,
        clientApiVersions: $clientApiVersions,
        deserializeEmptyStringAsNull: $deserializeEmptyStringAsNull,
        responseAsBool: $responseAsBool,
        clientDoc: $clientDoc,
        clientLocation: $clientLocation,
        clientOption: $clientOption,
    },
    "Azure.ClientGenerator.Core.Legacy": {
        hierarchyBuilding: $legacyHierarchyBuilding,
        flattenProperty: $flattenProperty,
        markAsLro: $markAsLro,
        markAsPageable: $markAsPageable,
        disablePageable: $disablePageable,
        nextLinkVerb: $nextLinkVerb,
        clientDefaultValue: $clientDefaultValue,
    },
};
