// oav/dist/lib/liveValidation/liveValidator.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseValidationRequest = exports.legacyParseValidationRequest = exports.LiveValidator = exports.LiveValidatorLoggingTypes = exports.LiveValidatorLoggingLevels = void 0;
exports.formatUrlToExpectedFormat = formatUrlToExpectedFormat;
const http = __importStar(require("http"));
const os = __importStar(require("os"));
const path = __importStar(require("path"));
const path_1 = require("path");
const util = __importStar(require("util"));
const url_1 = require("url");
const glob = __importStar(require("glob"));
const roundTripValidator_1 = require("../armValidator/roundTripValidator");
const models = __importStar(require("../models"));
const requestResponse_1 = require("../models/requestResponse");
const C = __importStar(require("../util/constants"));
const logging_1 = require("../util/logging");
const utils = __importStar(require("../util/utils"));
const inversifyUtils_1 = require("../inversifyUtils");
const loader_1 = require("../swagger/loader");
const errorDefinitions_1 = require("../util/errorDefinitions");
const utils_1 = require("../util/utils");
const liveValidatorLoader_1 = require("./liveValidatorLoader");
const operationSearcher_1 = require("./operationSearcher");
const operationValidator_1 = require("./operationValidator");
var LiveValidatorLoggingLevels;
(function (LiveValidatorLoggingLevels) {
    LiveValidatorLoggingLevels["error"] = "error";
    LiveValidatorLoggingLevels["warn"] = "warn";
    LiveValidatorLoggingLevels["info"] = "info";
    LiveValidatorLoggingLevels["verbose"] = "verbose";
    LiveValidatorLoggingLevels["debug"] = "debug";
    LiveValidatorLoggingLevels["silly"] = "silly";
})(LiveValidatorLoggingLevels || (exports.LiveValidatorLoggingLevels = LiveValidatorLoggingLevels = {}));
var LiveValidatorLoggingTypes;
(function (LiveValidatorLoggingTypes) {
    LiveValidatorLoggingTypes["trace"] = "trace";
    LiveValidatorLoggingTypes["perfTrace"] = "perfTrace";
    LiveValidatorLoggingTypes["error"] = "error";
    LiveValidatorLoggingTypes["incomingRequest"] = "incomingRequest";
    LiveValidatorLoggingTypes["specTrace"] = "specTrace";
})(LiveValidatorLoggingTypes || (exports.LiveValidatorLoggingTypes = LiveValidatorLoggingTypes = {}));
/**
 * @class
 * Live Validator for Azure swagger APIs.
 */
class LiveValidator {
    /**
     * Constructs LiveValidator based on provided options.
     *
     * @param {object} ops The configuration options.
     * @param {callback function} logCallback The callback logger.
     *
     * @returns CacheBuilder Returns the configured CacheBuilder object.
     */
    constructor(options, logCallback) {
        this.swaggerList = [];
        this.loadInBackgroundComplete = false;
        this.logging = (message, level, loggingType, operationName, durationInMilliseconds, validationRequest) => {
            level = level || LiveValidatorLoggingLevels.info;
            loggingType = loggingType || LiveValidatorLoggingTypes.trace;
            operationName = operationName || "";
            durationInMilliseconds = durationInMilliseconds || 0;
            if (this.logFunction !== undefined) {
                if (validationRequest !== undefined && validationRequest !== null) {
                    this.logFunction(message, level, {
                        CorrelationId: validationRequest.correlationId,
                        ActivityId: validationRequest.activityId,
                        ProviderNamespace: validationRequest.providerNamespace,
                        ResourceType: validationRequest.resourceType,
                        ApiVersion: validationRequest.apiVersion,
                        OperationName: operationName,
                        LoggingType: loggingType,
                        DurationInMilliseconds: durationInMilliseconds,
                        SpecName: validationRequest.specName,
                    });
                }
                else {
                    this.logFunction(message, level, {
                        OperationName: operationName,
                        LoggingType: loggingType,
                        DurationInMilliseconds: durationInMilliseconds,
                    });
                }
            }
            else {
                logging_1.log.log(level, message);
            }
        };
        const ops = options || {};
        this.logFunction = logCallback;
        (0, loader_1.setDefaultOpts)(ops, {
            swaggerPaths: [],
            excludedSwaggerPathsPattern: C.DefaultConfig.ExcludedSwaggerPathsPattern,
            directory: path.resolve(os.homedir(), "repo"),
            isPathCaseSensitive: false,
            loadValidatorInBackground: true,
            loadValidatorInInitialize: false,
            isArmCall: false,
        });
        if (!ops.git) {
            ops.git = {
                url: "https://github.com/Azure/azure-rest-api-specs.git",
                shouldClone: false,
            };
        }
        if (!ops.git.url) {
            ops.git.url = "https://github.com/Azure/azure-rest-api-specs.git";
        }
        if (!ops.git.shouldClone) {
            ops.git.shouldClone = false;
        }
        this.options = ops;
        this.logging(`Creating livevalidator with options:${JSON.stringify(this.options)}`);
        this.operationSearcher = new operationSearcher_1.OperationSearcher(this.logging);
    }
    /**
     * Initializes the Live Validator.
     */
    async initialize() {
        var _a, _b, _c, _d;
        const startTime = Date.now();
        // Clone github repository if required
        if (this.options.git.shouldClone && this.options.git.url) {
            const cloneStartTime = Date.now();
            utils.gitClone(this.options.directory, this.options.git.url, this.options.git.branch);
            this.logging(`Clone spec repository ${this.options.git.url}, branch:${this.options.git.branch} in livevalidator.initialize`);
            this.logging(`Clone spec repository ${this.options.git.url}, branch:${this.options.git.branch}`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.perfTrace, "Oav.liveValidator.initialize.gitclone", Date.now() - cloneStartTime);
        }
        // Construct array of swagger paths to be used for building a cache
        this.logging("Get swagger path.");
        const swaggerPaths = await this.getSwaggerPaths();
        const container = (0, inversifyUtils_1.inversifyGetContainer)();
        this.loader = (0, inversifyUtils_1.inversifyGetInstance)(liveValidatorLoader_1.LiveValidatorLoader, {
            container,
            fileRoot: this.options.directory,
            ...this.options,
            loadSuppression: (_a = this.options.loadSuppression) !== null && _a !== void 0 ? _a : Object.keys(errorDefinitions_1.apiValidationErrors),
        });
        this.loader.logging = this.logging;
        // re-set the transform context after set the logging function
        this.loader.setTransformContext();
        const schemaValidator = container.get(inversifyUtils_1.TYPES.schemaValidator);
        this.validateRequestResponsePair = await schemaValidator.compileAsync(requestResponse_1.requestResponseDefinition);
        const allSpecs = [];
        while (swaggerPaths.length > 0) {
            const swaggerPath = swaggerPaths.shift();
            this.swaggerList.push(swaggerPath);
            const spec = await this.getSwaggerInitializer(this.loader, swaggerPath);
            if (spec !== undefined) {
                allSpecs.push(spec);
            }
        }
        this.logging("Apply global transforms for all specs");
        try {
            this.loader.transformLoadedSpecs();
        }
        catch (e) {
            // keeps building validator if it fails to tranform specs coz global transformers catches the exceptions and continue other schema transformings;
            // this error will be reported in validator building or validation runtime.
            const errMsg = `Failed to transform loaded specs, detail error message:${e === null || e === void 0 ? void 0 : e.message}.\nError stack:${e === null || e === void 0 ? void 0 : e.stack}`;
            this.logging(errMsg, LiveValidatorLoggingLevels.error, LiveValidatorLoggingTypes.specTrace, "Oav.liveValidator.initialize.transformLoadedSpecs");
        }
        if (this.options.loadValidatorInInitialize) {
            this.logging("Building validator in initialization time...");
            let spec;
            while (allSpecs.length > 0) {
                try {
                    spec = allSpecs.shift();
                    const loadStart = Date.now();
                    await this.loader.buildAjvValidator(spec);
                    const durationInMs = Date.now() - loadStart;
                    this.logging(`Complete building validator for ${spec._filePath} in initialization time`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.perfTrace, "Oav.liveValidator.initialize.loader.buildAjvValidator", durationInMs);
                    this.logging(`Complete building validator for spec ${spec._filePath}`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.specTrace, "Oav.liveValidator.initialize.loader.buildAjvValidator", durationInMs, {
                        providerNamespace: (_b = spec._providerNamespace) !== null && _b !== void 0 ? _b : "unknown",
                        apiVersion: spec.info.version,
                        specName: spec._filePath,
                    });
                }
                catch (e) {
                    this.logging(`Failed to build validator for spec ${spec === null || spec === void 0 ? void 0 : spec._filePath}.\nErrorMessage:${e === null || e === void 0 ? void 0 : e.message}.\nErrorStack:${e === null || e === void 0 ? void 0 : e.stack}`, LiveValidatorLoggingLevels.error, LiveValidatorLoggingTypes.specTrace, "Oav.liveValidator.initialize.loader.buildAjvValidator", undefined, {
                        providerNamespace: (_c = spec === null || spec === void 0 ? void 0 : spec._providerNamespace) !== null && _c !== void 0 ? _c : "unknown",
                        apiVersion: (_d = spec === null || spec === void 0 ? void 0 : spec.info.version) !== null && _d !== void 0 ? _d : "unknown",
                        specName: spec === null || spec === void 0 ? void 0 : spec._filePath,
                    });
                }
            }
            this.loader = undefined;
        }
        this.logging("Cache initialization complete.");
        const elapsedTime = Date.now() - startTime;
        this.logging(`Cache complete initialization with DurationInMs:${elapsedTime}`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.trace, "Oav.liveValidator.initialize");
        this.logging(`Cache complete initialization`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.perfTrace, "Oav.liveValidator.initialize", elapsedTime);
        if (this.options.loadValidatorInBackground) {
            this.logging("Building validator in background...");
            // eslint-disable-next-line @typescript-eslint/no-floating-promises
            this.loadAllSpecValidatorInBackground(allSpecs);
        }
    }
    isLoadInBackgroundCompleted() {
        return this.loadInBackgroundComplete;
    }
    async loadAllSpecValidatorInBackground(allSpecs) {
        var _a, _b, _c;
        const backgroundStartTime = Date.now();
        utils.shuffleArray(allSpecs);
        while (allSpecs.length > 0) {
            let spec;
            try {
                spec = allSpecs.shift();
                const startTime = Date.now();
                this.logging(`Start building validator for ${spec._filePath} in background`, LiveValidatorLoggingLevels.debug, LiveValidatorLoggingTypes.trace, "Oav.liveValidator.loadAllSpecValidatorInBackground");
                await this.loader.buildAjvValidator(spec, { inBackground: true });
                const elapsedTime = Date.now() - startTime;
                this.logging(`Complete building validator for ${spec._filePath} in background with DurationInMs:${elapsedTime}.`, LiveValidatorLoggingLevels.debug, LiveValidatorLoggingTypes.trace, "Oav.liveValidator.loadAllSpecValidatorInBackground");
                this.logging(`Complete building validator for ${spec._filePath} in background`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.perfTrace, "Oav.liveValidator.loadAllSpecValidatorInBackground", elapsedTime);
                this.logging(`Complete building validator for spec ${spec._filePath}`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.specTrace, "Oav.liveValidator.loadAllSpecValidatorInBackground", elapsedTime, {
                    providerNamespace: (_a = spec._providerNamespace) !== null && _a !== void 0 ? _a : "unknown",
                    apiVersion: spec.info.version,
                    specName: spec._filePath,
                });
            }
            catch (e) {
                this.logging(`Failed to build validator for spec ${spec === null || spec === void 0 ? void 0 : spec._filePath}.\nErrorMessage:${e === null || e === void 0 ? void 0 : e.message}.\nErrorStack:${e === null || e === void 0 ? void 0 : e.stack}`, LiveValidatorLoggingLevels.error, LiveValidatorLoggingTypes.specTrace, "Oav.liveValidator.loadAllSpecValidatorInBackground", undefined, {
                    providerNamespace: (_b = spec === null || spec === void 0 ? void 0 : spec._providerNamespace) !== null && _b !== void 0 ? _b : "unknown",
                    apiVersion: (_c = spec === null || spec === void 0 ? void 0 : spec.info.version) !== null && _c !== void 0 ? _c : "unknown",
                    specName: spec === null || spec === void 0 ? void 0 : spec._filePath,
                });
            }
        }
        this.loader = undefined;
        this.loadInBackgroundComplete = true;
        const elapsedTimeForBuild = Date.now() - backgroundStartTime;
        this.logging(`Completed building validator for all specs in background with DurationInMs:${elapsedTimeForBuild}.`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.trace, "Oav.liveValidator.loadAllSpecValidatorInBackground");
        this.logging(`Completed building validator for all specs in background`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.perfTrace, "Oav.liveValidator.loadAllSpecValidatorInBackground", elapsedTimeForBuild);
    }
    /**
     *  Validates live request.
     */
    async validateLiveRequest(liveRequest, options = {}, operationInfo) {
        const startTime = Date.now();
        const { info, error } = this.getOperationInfo(liveRequest, operationInfo);
        if (error !== undefined) {
            this.logging(`ErrorMessage:${error.message}.ErrorStack:${error.stack}`, LiveValidatorLoggingLevels.error, LiveValidatorLoggingTypes.error, "Oav.liveValidator.validateLiveRequest", undefined, info.validationRequest);
            return {
                isSuccessful: undefined,
                errors: [],
                runtimeException: error,
                operationInfo: info,
            };
        }
        if (!liveRequest.query) {
            liveRequest.query = (0, utils_1.kvPairsToObject)(new url_1.URL(liveRequest.url, "https://management.azure.com").searchParams);
        }
        let errors = [];
        let runtimeException;
        try {
            errors = await (0, operationValidator_1.validateSwaggerLiveRequest)(liveRequest, info, this.loader, options.includeErrors, this.options.isArmCall, this.logging);
        }
        catch (reqValidationError) {
            const msg = `An error occurred while validating the live request for operation ` +
                `"${info.operationId}". The error is:\n ` +
                `${util.inspect(reqValidationError, { depth: null })}`;
            runtimeException = { code: C.ErrorCodes.RequestValidationError.name, message: msg };
            this.logging(msg, LiveValidatorLoggingLevels.error, LiveValidatorLoggingTypes.error, "Oav.liveValidator.validateLiveRequest", undefined, info.validationRequest);
        }
        const elapsedTime = Date.now() - startTime;
        this.logging(`Complete request validation`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.perfTrace, "Oav.liveValidator.validateLiveRequest", elapsedTime, info.validationRequest);
        if (!options.includeOperationMatch) {
            delete info.operationMatch;
            delete info.validationRequest;
        }
        return {
            isSuccessful: runtimeException ? undefined : errors.length === 0,
            operationInfo: info,
            errors,
            runtimeException,
        };
    }
    /**
     * Validates live response.
     */
    async validateLiveResponse(liveResponse, specOperation, options = {}, operationInfo) {
        const startTime = Date.now();
        const { info, error } = this.getOperationInfo(specOperation, operationInfo);
        if (error !== undefined) {
            this.logging(`ErrorMessage:${error.message}.ErrorStack:${error.stack}`, LiveValidatorLoggingLevels.error, LiveValidatorLoggingTypes.error, "Oav.liveValidator.validateLiveResponse", undefined, info.validationRequest);
            return {
                isSuccessful: undefined,
                errors: [],
                runtimeException: error,
                operationInfo: { apiVersion: C.unknownApiVersion, operationId: C.unknownOperationId },
            };
        }
        let errors = [];
        let runtimeException;
        this.transformResponseStatusCode(liveResponse);
        try {
            errors = await (0, operationValidator_1.validateSwaggerLiveResponse)(liveResponse, info, this.loader, options.includeErrors, this.options.isArmCall, this.logging);
        }
        catch (resValidationError) {
            const msg = `An error occurred while validating the live response for operation ` +
                `"${info.operationId}". The error is:\n ` +
                `${util.inspect(resValidationError, { depth: null })}`;
            runtimeException = { code: C.ErrorCodes.RequestValidationError.name, message: msg };
            this.logging(msg, LiveValidatorLoggingLevels.error, LiveValidatorLoggingTypes.error, "Oav.liveValidator.validateLiveResponse", undefined, info.validationRequest);
        }
        const elapsedTime = Date.now() - startTime;
        this.logging(`Complete response validation`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.perfTrace, "Oav.liveValidator.validateLiveResponse", elapsedTime, info.validationRequest);
        if (!options.includeOperationMatch) {
            delete info.operationMatch;
            delete info.validationRequest;
        }
        return {
            isSuccessful: runtimeException ? undefined : errors.length === 0,
            operationInfo: info,
            errors,
            runtimeException,
        };
    }
    /**
     * Validates live request and response.
     */
    async validateLiveRequestResponse(requestResponseObj, options) {
        const validationResult = {
            requestValidationResult: {
                errors: [],
                operationInfo: { apiVersion: C.unknownApiVersion, operationId: C.unknownOperationId },
            },
            responseValidationResult: {
                errors: [],
                operationInfo: { apiVersion: C.unknownApiVersion, operationId: C.unknownOperationId },
            },
        };
        if (!requestResponseObj) {
            const message = 'requestResponseObj cannot be null or undefined and must be of type "object".';
            return {
                ...validationResult,
                runtimeException: {
                    code: C.ErrorCodes.IncorrectInput.name,
                    message,
                },
            };
        }
        const errors = this.validateRequestResponsePair({}, requestResponseObj);
        if (errors.length > 0) {
            const error = errors[0];
            const message = `Found errors "${error.message}" in the provided input in path ${error.jsonPathsInPayload[0]}:\n` +
                `${util.inspect(requestResponseObj, { depth: null })}.`;
            return {
                ...validationResult,
                runtimeException: {
                    code: C.ErrorCodes.IncorrectInput.name,
                    message,
                },
            };
        }
        const request = requestResponseObj.liveRequest;
        const response = requestResponseObj.liveResponse;
        this.transformResponseStatusCode(response);
        const requestValidationResult = await this.validateLiveRequest(request, {
            ...options,
            includeOperationMatch: true,
        });
        const info = requestValidationResult.operationInfo;
        const responseValidationResult = requestValidationResult.isSuccessful === undefined &&
            requestValidationResult.runtimeException === undefined
            ? requestValidationResult
            : await this.validateLiveResponse(response, request, {
                ...options,
                includeOperationMatch: false,
            }, info);
        delete info.validationRequest;
        delete info.operationMatch;
        return {
            requestValidationResult,
            responseValidationResult,
        };
    }
    transformResponseStatusCode(liveResponse) {
        // If status code is passed as a status code string (e.g. "OK") transform it to the status code
        // number (e.g. '200').
        if (!http.STATUS_CODES[liveResponse.statusCode] &&
            utils.statusCodeStringToStatusCode[liveResponse.statusCode.toLowerCase()]) {
            liveResponse.statusCode =
                utils.statusCodeStringToStatusCode[liveResponse.statusCode.toLowerCase()];
        }
    }
    getOperationInfo(request, operationInfo) {
        const info = operationInfo !== null && operationInfo !== void 0 ? operationInfo : {
            apiVersion: C.unknownApiVersion,
            operationId: C.unknownOperationId,
        };
        try {
            if (info.validationRequest === undefined) {
                info.validationRequest = (0, exports.parseValidationRequest)(request.url, request.method, request.headers);
            }
            if (info.operationMatch === undefined) {
                const result = this.operationSearcher.search(info.validationRequest);
                info.apiVersion = result.apiVersion;
                info.operationMatch = result.operationMatch;
            }
            info.operationId = info.operationMatch.operation.operationId;
            return { info };
        }
        catch (error) {
            return { info, error };
        }
    }
    getResolvedOperationInfo(request, operationInfo) {
        const info = operationInfo !== null && operationInfo !== void 0 ? operationInfo : {
            apiVersion: C.unknownApiVersion,
            operationId: C.unknownOperationId,
        };
        try {
            if (info.validationRequest === undefined) {
                info.validationRequest = (0, exports.parseValidationRequest)(request.url, request.method, request.headers);
            }
            if (info.operationMatch === undefined) {
                const result = this.operationSearcher.search(info.validationRequest);
                info.apiVersion = result.apiVersion;
                info.operationMatch = result.operationMatch;
            }
            info.operationId = info.operationMatch.operation.operationId;
            return { info };
        }
        catch (error) {
            return { info, error };
        }
    }
    async getMatchedPaths(jsonsPattern) {
        const startTime = Date.now();
        let matchedPaths = [];
        if (typeof jsonsPattern === "string") {
            matchedPaths = glob
                .sync(jsonsPattern, {
                ignore: this.options.excludedSwaggerPathsPattern,
                nodir: true,
            })
                // Sort alphabetically for compat with glob <= 8 (https://github.com/isaacs/node-glob/blob/v7.2.3/common.js#L20).
                // Tests pass without sorting, but safer to keep sorting for consistent behavior.
                .sort((a, b) => a.localeCompare(b, "en"));
        }
        else {
            for (const pattern of jsonsPattern) {
                const res = glob
                    .sync(pattern, {
                    ignore: this.options.excludedSwaggerPathsPattern,
                    nodir: true,
                })
                    // Sort alphabetically for compat with glob <= 8 (https://github.com/isaacs/node-glob/blob/v7.2.3/common.js#L20).
                    // Tests pass without sorting, but safer to keep sorting for consistent behavior.
                    .sort((a, b) => a.localeCompare(b, "en"));
                for (const path of res) {
                    if (!matchedPaths.includes(path)) {
                        matchedPaths.push(path);
                    }
                }
            }
        }
        this.logging(`Using swaggers found from directory: "${this.options.directory}" and pattern: "${jsonsPattern.toString()}".
      Total paths count: ${matchedPaths.length}`, LiveValidatorLoggingLevels.info);
        this.logging(`Using swaggers found from directory: "${this.options.directory}" and pattern: "${jsonsPattern.toString()}".
      Total paths count: ${matchedPaths.length}`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.perfTrace, "Oav.livevalidator.getMatchedPaths", Date.now() - startTime);
        return matchedPaths;
    }
    async validateRoundTrip(requestResponseObj) {
        var _a;
        const startTime = Date.now();
        if (this.operationSearcher === undefined) {
            const msg = "operationSearcher should be initialized before this call.";
            const runtimeException = { code: C.ErrorCodes.RoundtripValidationError.name, message: msg };
            return {
                isSuccessful: undefined,
                errors: [],
                operationInfo: {
                    operationId: "",
                    apiVersion: "",
                },
                runtimeException: runtimeException,
            };
        }
        const { info, error } = this.getResolvedOperationInfo(requestResponseObj.liveRequest);
        console.log(info.operationId);
        if (error !== undefined) {
            this.logging(`ErrorMessage:${error.message}.ErrorStack:${error.stack}`, LiveValidatorLoggingLevels.error, LiveValidatorLoggingTypes.error, "Oav.liveValidator.validateRoundTrip", undefined, info.validationRequest);
            return {
                isSuccessful: undefined,
                errors: [],
                runtimeException: error,
                operationInfo: info,
            };
        }
        let errors = [];
        let runtimeException;
        try {
            const res = (0, roundTripValidator_1.diffRequestResponse)(requestResponseObj, info, (_a = this.loader) === null || _a === void 0 ? void 0 : _a.getResolvedJsonLoader());
            for (const re of res) {
                if (re !== undefined) {
                    errors.push(re);
                }
            }
        }
        catch (validationErr) {
            const msg = `An error occurred while validating the live request for operation ` +
                `"${info.operationId}". The error is:\n ` +
                `${util.inspect(validationErr, { depth: null })}`;
            runtimeException = { code: C.ErrorCodes.RoundtripValidationError.name, message: msg };
            this.logging(msg, LiveValidatorLoggingLevels.error, LiveValidatorLoggingTypes.error, "Oav.liveValidator.validateRoundTrip", undefined, info.validationRequest);
            return {
                isSuccessful: undefined,
                operationInfo: info,
                errors,
                runtimeException,
            };
        }
        const elapsedTime = Date.now() - startTime;
        this.logging(`Complete roundtrip validation`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.perfTrace, "Oav.liveValidator.validateRoundTrip", elapsedTime, info.validationRequest);
        delete info.validationRequest;
        delete info.operationMatch;
        return {
            isSuccessful: runtimeException ? undefined : errors.length === 0,
            operationInfo: info,
            errors,
            runtimeException,
        };
    }
    async getSwaggerPaths() {
        if (this.options.swaggerPaths.length !== 0) {
            this.logging(`Using user provided swagger paths by options.swaggerPaths. Total paths count: ${this.options.swaggerPaths.length}`);
            return this.options.swaggerPaths;
        }
        else {
            const allJsonsPattern = path.join(this.options.directory, "/specification/**/*.json");
            const swaggerPathPatterns = [];
            if (this.options.swaggerPathsPattern === undefined ||
                this.options.swaggerPathsPattern.length === 0) {
                return this.getMatchedPaths(allJsonsPattern);
            }
            else {
                this.options.swaggerPathsPattern.map((item) => {
                    swaggerPathPatterns.push(path.join(this.options.directory, item));
                });
                return this.getMatchedPaths(swaggerPathPatterns);
            }
        }
    }
    async getSwaggerInitializer(loader, swaggerPath) {
        var _a;
        const startTime = Date.now();
        this.logging(`Building cache from:${swaggerPath}`, LiveValidatorLoggingLevels.debug);
        let spec;
        let resolvedSpec;
        try {
            spec = await loader.load((0, path_1.resolve)(swaggerPath));
            const elapsedTimeLoadSpec = Date.now() - startTime;
            this.logging(`Load spec ${swaggerPath}`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.perfTrace, "Oav.liveValidator.getSwaggerInitializer.loader.load", elapsedTimeLoadSpec);
            const startTimeAddSpecToCache = Date.now();
            if (this.options.enableRoundTripValidator) {
                resolvedSpec = await loader.load((0, path_1.resolve)(swaggerPath), true);
                this.operationSearcher.addSpecToCache(resolvedSpec);
            }
            else {
                this.operationSearcher.addSpecToCache(spec);
            }
            // TODO: add data-plane RP to cache.
            this.logging(`Add spec to cache ${swaggerPath}`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.perfTrace, "Oav.liveValidator.getSwaggerInitializer.operationSearcher.addSpecToCache", Date.now() - startTimeAddSpecToCache);
            const elapsedTime = Date.now() - startTime;
            this.logging(`Complete loading spec ${spec._filePath}`, LiveValidatorLoggingLevels.info, LiveValidatorLoggingTypes.specTrace, "Oav.liveValidator.getSwaggerInitializer", elapsedTime, {
                providerNamespace: (_a = spec._providerNamespace) !== null && _a !== void 0 ? _a : "unknown",
                apiVersion: spec.info.version,
                specName: spec._filePath,
            });
            return spec;
        }
        catch (err) {
            this.logging(`Unable to initialize "${swaggerPath}" file from SpecValidator. We are ` +
                `ignoring this swagger file and continuing to build cache for other valid specs.\nErrorMessage: ${err === null || err === void 0 ? void 0 : err.message};\nErrorStack: ${err === null || err === void 0 ? void 0 : err.stack}`, LiveValidatorLoggingLevels.warn, LiveValidatorLoggingTypes.error);
            const pathProvider = (0, utils_1.getProviderFromSpecPath)(swaggerPath);
            this.logging(`Failed to load spec ${swaggerPath}`, LiveValidatorLoggingLevels.error, LiveValidatorLoggingTypes.specTrace, "Oav.liveValidator.getSwaggerInitializer", undefined, {
                providerNamespace: pathProvider ? pathProvider.provider : "unknown",
                apiVersion: spec ? spec.info.version : "unknown",
                specName: swaggerPath,
            });
            return undefined;
        }
    }
}
exports.LiveValidator = LiveValidator;
/**
 * OAV expects the url that is sent to match exactly with the swagger path. For this we need to keep only the part after
 * where the swagger path starts. Currently those are '/subscriptions' and '/providers'.
 */
function formatUrlToExpectedFormat(requestUrl) {
    return requestUrl.substring(requestUrl.search("/?(subscriptions|providers)/i"));
}
/**
 * Parse the validation request information.
 *
 * @param  requestUrl The url of service api call.
 *
 * @param requestMethod The http verb for the method to be used for lookup.
 *
 * @param correlationId The id to correlate the api calls.
 *
 * @param activityId The id maps to request id, used by RPaaS.
 *
 * @returns parsed ValidationRequest info.
 *
 * @deprecated use parseValidationRequest instead.
 */
const legacyParseValidationRequest = (requestUrl, requestMethod, correlationId, activityId) => {
    if (requestUrl === undefined ||
        requestUrl === null ||
        typeof requestUrl.valueOf() !== "string" ||
        !requestUrl.trim().length) {
        const msg = "An error occurred while trying to parse validation payload." +
            'requestUrl is a required parameter of type "string" and it cannot be an empty string.';
        const e = new models.LiveValidationError(C.ErrorCodes.PotentialOperationSearchError.name, msg);
        throw e;
    }
    if (requestMethod === undefined ||
        requestMethod === null ||
        typeof requestMethod.valueOf() !== "string" ||
        !requestMethod.trim().length) {
        const msg = "An error occurred while trying to parse validation payload." +
            'requestMethod is a required parameter of type "string" and it cannot be an empty string.';
        const e = new models.LiveValidationError(C.ErrorCodes.PotentialOperationSearchError.name, msg);
        throw e;
    }
    return (0, exports.parseValidationRequest)(requestUrl, requestMethod, {
        "x-ms-correlation-request-id": correlationId,
        "x-ms-request-id": activityId,
    });
};
exports.legacyParseValidationRequest = legacyParseValidationRequest;
/**
 * Parse the validation request information.
 *
 * @param requestUrl The url of service api call.
 *
 * @param requestMethod The http verb for the method to be used for lookup.
 *
 * @param headers Optional headers
 *
 * @returns parsed ValidationRequest info.
 */
const parseValidationRequest = (requestUrl, requestMethod, headers) => {
    let queryStr;
    let apiVersion = "";
    let resourceType = "";
    let providerNamespace = "";
    const parsedUrl = new url_1.URL(requestUrl, "https://management.azure.com");
    const pathStr = parsedUrl.pathname || "";
    if (pathStr !== "") {
        // Lower all the keys and values of query parameters before searching for `api-version`
        const queryObject = {};
        parsedUrl.searchParams.forEach((value, key) => {
            queryObject[key.toLowerCase()] = value.toLowerCase();
        });
        apiVersion = (queryObject["api-version"] || C.unknownApiVersion);
        providerNamespace = (0, utils_1.getProviderFromPathTemplate)(pathStr) || C.unknownResourceProvider;
        resourceType = utils.getResourceType(pathStr, providerNamespace);
        // Provider would be provider found from the path or Microsoft.Unknown
        providerNamespace = providerNamespace || C.unknownResourceProvider;
        if (providerNamespace === C.unknownResourceProvider) {
            //apiVersion = C.unknownApiVersion;
        }
        providerNamespace = providerNamespace.toLowerCase();
        apiVersion = apiVersion.toLowerCase();
        queryStr = queryObject;
        requestMethod = requestMethod.toLowerCase();
    }
    return {
        providerNamespace,
        resourceType,
        apiVersion,
        requestMethod: requestMethod,
        host: parsedUrl.host,
        pathStr,
        query: queryStr,
        requestUrl: requestUrl,
        correlationId: (headers === null || headers === void 0 ? void 0 : headers["x-ms-correlation-request-id"]) || "",
        activityId: (headers === null || headers === void 0 ? void 0 : headers["x-ms-request-id"]) || "",
    };
};
exports.parseValidationRequest = parseValidationRequest;
