// oav/dist/lib/liveValidation/liveValidatorLoader.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiveValidatorLoader = void 0;
const openapi_tools_common_1 = require("@azure-tools/openapi-tools-common");
const inversify_1 = require("inversify");
const inversifyUtils_1 = require("../inversifyUtils");
const jsonLoader_1 = require("../swagger/jsonLoader");
const loader_1 = require("../swagger/loader");
const swaggerLoader_1 = require("../swagger/swaggerLoader");
const allOfTransformer_1 = require("../transform/allOfTransformer");
const context_1 = require("../transform/context");
const discriminatorTransformer_1 = require("../transform/discriminatorTransformer");
const noAdditionalPropertiesTransformer_1 = require("../transform/noAdditionalPropertiesTransformer");
const nullableTransformer_1 = require("../transform/nullableTransformer");
const pathRegexTransformer_1 = require("../transform/pathRegexTransformer");
const pureObjectTransformer_1 = require("../transform/pureObjectTransformer");
const referenceFieldsTransformer_1 = require("../transform/referenceFieldsTransformer");
const resolveNestedDefinitionTransformer_1 = require("../transform/resolveNestedDefinitionTransformer");
const schemaV4ToV7Transformer_1 = require("../transform/schemaV4ToV7Transformer");
const transformer_1 = require("../transform/transformer");
const traverseSwagger_1 = require("../transform/traverseSwagger");
const xmsPathsTransformer_1 = require("../transform/xmsPathsTransformer");
const lazyBuilder_1 = require("../util/lazyBuilder");
const utils_1 = require("../util/utils");
let LiveValidatorLoader = class LiveValidatorLoader {
    constructor(opts, jsonLoader, swaggerLoader, schemaValidator) {
        this.opts = opts;
        this.jsonLoader = jsonLoader;
        this.swaggerLoader = swaggerLoader;
        this.schemaValidator = schemaValidator;
        this.getResponseValidator = (0, lazyBuilder_1.getLazyBuilder)("_validate", (response) => {
            var _a;
            const schema = { properties: {} };
            if (response.schema !== undefined && response.schema.type !== "file") {
                schema.properties.body = response.schema;
                (0, openapi_tools_common_1.copyInfo)(response, response.schema);
                this.addRequiredToSchema(schema, "body");
            }
            if (response.headers !== undefined) {
                const headerSchema = { properties: {}, required: [] };
                (0, openapi_tools_common_1.copyInfo)(response.headers, headerSchema);
                schema.properties.headers = headerSchema;
                for (const headerName of Object.keys(response.headers)) {
                    // Per swagger 2.0 spec, header is optional by default
                    const name = headerName.toLowerCase();
                    const sch = response.headers[headerName];
                    headerSchema.properties[name] = sch;
                    addParamTransform(response, { type: sch.type, name, in: "header" });
                }
                if (((_a = headerSchema.required) === null || _a === void 0 ? void 0 : _a.length) === 0) {
                    headerSchema.required = undefined;
                }
            }
            else {
                schema.properties.headers = {};
            }
            return this.schemaValidator.compileAsync(schema);
        });
        this.getRequestValidator = (0, lazyBuilder_1.getLazyBuilder)("_validate", (operation) => {
            const schema = {
                properties: {
                    query: { properties: {} },
                    headers: { properties: {} },
                    path: { properties: {} },
                    formData: { properties: {} },
                },
            };
            this.addParamToSchema(schema, operation._path.parameters, operation);
            this.addParamToSchema(schema, operation.parameters, operation);
            return this.schemaValidator.compileAsync(schema);
        });
        (0, loader_1.setDefaultOpts)(opts, {
            transformToNewSchemaFormat: false,
        });
        this.setTransformContext();
    }
    setTransformContext() {
        this.transformContext = (0, context_1.getTransformContext)(this.jsonLoader, this.schemaValidator, [
            xmsPathsTransformer_1.xmsPathsTransformer,
            resolveNestedDefinitionTransformer_1.resolveNestedDefinitionTransformer,
            this.opts.transformToNewSchemaFormat ? schemaV4ToV7Transformer_1.schemaV4ToV7Transformer : undefined,
            referenceFieldsTransformer_1.referenceFieldsTransformer,
            pathRegexTransformer_1.pathRegexTransformer,
            discriminatorTransformer_1.discriminatorTransformer,
            allOfTransformer_1.allOfTransformer,
            noAdditionalPropertiesTransformer_1.noAdditionalPropertiesTransformer,
            nullableTransformer_1.nullableTransformer,
            pureObjectTransformer_1.pureObjectTransformer,
        ], this.logging);
    }
    async load(specFilePath, keepRefSiblings) {
        const spec = await this.swaggerLoader.load(specFilePath, keepRefSiblings);
        (0, transformer_1.applySpecTransformers)(spec, this.transformContext);
        return spec;
    }
    getResolvedJsonLoader() {
        return this.swaggerLoader.getResolvedJsonLoader();
    }
    transformLoadedSpecs() {
        (0, transformer_1.applyGlobalTransformers)(this.transformContext);
    }
    async buildAjvValidator(spec, options) {
        return (0, traverseSwagger_1.traverseSwaggerAsync)(spec, {
            onOperation: async (operation) => {
                await this.getRequestValidator(operation);
                if (options === null || options === void 0 ? void 0 : options.inBackground) {
                    await (0, utils_1.waitUntilLowLoad)();
                }
            },
            onResponse: async (response) => {
                await this.getResponseValidator(response);
                if (options === null || options === void 0 ? void 0 : options.inBackground) {
                    await (0, utils_1.waitUntilLowLoad)();
                }
            },
        });
    }
    addRequiredToSchema(schema, requiredName) {
        if (schema.required === undefined) {
            schema.required = [];
        }
        if (!schema.required.includes(requiredName)) {
            schema.required.push(requiredName);
        }
    }
    addParamToSchema(schema, params, operation) {
        var _a, _b;
        if (params === undefined) {
            return;
        }
        const properties = schema.properties;
        for (const p of params) {
            const param = this.jsonLoader.resolveRefObj(p);
            switch (param.in) {
                case "body":
                    // ignore validation in case of file type
                    if (((_a = param.schema) === null || _a === void 0 ? void 0 : _a.type) === "file") {
                        break;
                    }
                    properties.body = (_b = param.schema) !== null && _b !== void 0 ? _b : {};
                    if (param.required) {
                        (0, openapi_tools_common_1.copyInfo)(param, schema);
                        this.addRequiredToSchema(schema, "body");
                    }
                    else {
                        operation._bodyTransform = bodyTransformIfNotRequiredAndEmpty;
                    }
                    break;
                case "header":
                    const name = param.name.toLowerCase();
                    if (param.required) {
                        this.addRequiredToSchema(properties.headers, name);
                    }
                    param.required = undefined;
                    properties.headers.properties[name] = param;
                    addParamTransform(operation, param);
                    break;
                case "query":
                    if (shouldSkipQueryParam(param)) {
                        break;
                    }
                    if (param.required) {
                        this.addRequiredToSchema(properties.query, param.name);
                    }
                    // Remove param.required as it have different meaning in swagger and json schema
                    param.required = undefined;
                    properties.query.properties[param.name] = param;
                    addParamTransform(operation, param);
                    break;
                case "path":
                    if (shouldSkipPathParam(param)) {
                        break;
                    }
                    // if (!param.required) {
                    //   throw new Error("Path property mush be required");
                    // }
                    this.addRequiredToSchema(properties.path, param.name);
                    param.required = undefined;
                    properties.path.properties[param.name] = param;
                    addParamTransform(operation, param);
                    break;
                case "formData":
                    // ignore validation in case of file type
                    if (param.type === "file") {
                        break;
                    }
                    if (param.required) {
                        this.addRequiredToSchema(properties.formData, param.name);
                    }
                    // Remove param.required as it have different meaning in swagger and json schema
                    param.required = undefined;
                    properties.formData.properties[param.name] = param;
                    addParamTransform(operation, param);
                    break;
                default:
                    throw new Error(`Not Supported parameter in: ${param}`);
            }
        }
    }
};
exports.LiveValidatorLoader = LiveValidatorLoader;
exports.LiveValidatorLoader = LiveValidatorLoader = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(inversifyUtils_1.TYPES.opts)),
    __param(3, (0, inversify_1.inject)(inversifyUtils_1.TYPES.schemaValidator)),
    __metadata("design:paramtypes", [Object, jsonLoader_1.JsonLoader,
        swaggerLoader_1.SwaggerLoader, Object])
], LiveValidatorLoader);
const skipPathParamProperties = {
    in: "path",
    name: true,
    type: "string",
    description: true,
    required: true,
};
const shouldSkipPathParam = (param) => {
    for (const key of Object.keys(param)) {
        const val = skipPathParamProperties[key];
        if (val !== true && val !== param[key]) {
            return false;
        }
    }
    return true;
};
const skipQueryParamProperties = {
    in: "query",
    name: true,
    type: "string",
    description: true,
    required: true,
};
const shouldSkipQueryParam = (param) => {
    if (param.name !== "api-version") {
        return false;
    }
    for (const key of Object.keys(param)) {
        const val = skipQueryParamProperties[key];
        if (val !== true && val !== param[key]) {
            return false;
        }
    }
    return true;
};
const paramTransformInteger = (data) => {
    const val = Number(data);
    return isNaN(val) ? data : val;
};
const paramTransformBoolean = (data) => {
    const val = data.toLowerCase();
    return val === "true" ? true : val === "false" ? false : data;
};
const parameterTransform = {
    number: paramTransformInteger,
    integer: paramTransformInteger,
    boolean: paramTransformBoolean,
};
const bodyTransformIfNotRequiredAndEmpty = (body) => {
    if (body && Object.keys(body).length === 0 && body.constructor === Object) {
        return undefined;
    }
    else {
        return body;
    }
};
const addParamTransform = (it, param) => {
    const transform = parameterTransform[param.type];
    if (transform === undefined) {
        return;
    }
    if (param.in === "query") {
        const op = it;
        if (op._queryTransform === undefined) {
            op._queryTransform = {};
        }
        op._queryTransform[param.name] = transform;
    }
    else if (param.in === "header") {
        if (it._headerTransform === undefined) {
            it._headerTransform = {};
        }
        it._headerTransform[param.name.toLowerCase()] = transform;
    }
    else if (param.in === "path") {
        const op = it;
        if (op._pathTransform === undefined) {
            op._pathTransform = {};
        }
        op._pathTransform[param.name] = transform;
    }
};
