// oav/dist/lib/liveValidation/operationSearcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OperationSearcher = void 0;
const models_1 = require("../models");
const traverseSwagger_1 = require("../transform/traverseSwagger");
const constants_1 = require("../util/constants");
const errorDefinitions_1 = require("../util/errorDefinitions");
const utils_1 = require("../util/utils");
const liveValidator_1 = require("./liveValidator");
class OperationSearcher {
    constructor(logging) {
        this.logging = logging;
        // Key provider, e.g. Microsoft.ApiManagement
        this.cache = new Map();
    }
    addSpecToCache(spec) {
        (0, traverseSwagger_1.traverseSwagger)(spec, {
            onOperation: (operation, path, method) => {
                var _a, _b;
                const httpMethod = method.toLowerCase();
                const pathObject = path;
                const pathStr = pathObject._pathTemplate;
                let apiVersion = spec.info.version;
                let provider = (0, utils_1.getProviderFromPathTemplate)(pathStr);
                const addOperationToCache = () => {
                    let apiVersions = this.cache.get(provider);
                    if (apiVersions === undefined) {
                        apiVersions = new Map();
                        this.cache.set(provider, apiVersions);
                    }
                    let allMethods = apiVersions.get(apiVersion);
                    if (allMethods === undefined) {
                        allMethods = new Map();
                        apiVersions.set(apiVersion, allMethods);
                    }
                    let operationsForHttpMethod = allMethods.get(httpMethod);
                    if (operationsForHttpMethod === undefined) {
                        operationsForHttpMethod = [];
                        allMethods.set(httpMethod, operationsForHttpMethod);
                    }
                    operationsForHttpMethod.push(operation);
                };
                this.logging(`${apiVersion}, ${operation.operationId}, ${pathStr}, ${httpMethod}`, liveValidator_1.LiveValidatorLoggingLevels.debug);
                if (!apiVersion) {
                    this.logging(`Unable to find apiVersion for path : "${pathObject._pathTemplate}".`, liveValidator_1.LiveValidatorLoggingLevels.error, liveValidator_1.LiveValidatorLoggingTypes.specTrace, "Oav.OperationSearcher.addSpecToCache", undefined, {
                        providerNamespace: (_a = spec._providerNamespace) !== null && _a !== void 0 ? _a : "unknown",
                        apiVersion: spec.info.version,
                        specName: spec._filePath,
                    });
                    apiVersion = constants_1.unknownApiVersion;
                }
                apiVersion = apiVersion.toLowerCase();
                if (!provider) {
                    const title = spec.info.title;
                    // Whitelist lookups: Look up knownTitleToResourceProviders
                    // Putting the provider namespace onto operation for future use
                    if (title && constants_1.knownTitleToResourceProviders[title]) {
                        operation.provider = constants_1.knownTitleToResourceProviders[title];
                    }
                    // Put the operation into 'Microsoft.Unknown' RPs
                    provider = constants_1.unknownResourceProvider;
                    this.logging(`Unable to find provider for path : "${pathObject._pathTemplate}". ` +
                        `Bucketizing into provider: "${provider}"`, liveValidator_1.LiveValidatorLoggingLevels.warn, liveValidator_1.LiveValidatorLoggingTypes.specTrace, "Oav.OperationSearcher.addSpecToCache", undefined, {
                        providerNamespace: (_b = spec._providerNamespace) !== null && _b !== void 0 ? _b : "unknown",
                        apiVersion: spec.info.version,
                        specName: spec._filePath,
                    });
                }
                provider = provider.toLowerCase();
                addOperationToCache();
            },
        });
    }
    /**
     * Gets the swagger operation based on the HTTP url and method
     */
    search(info) {
        const startTime = Date.now();
        const requestInfo = { ...info };
        const searchOperation = () => {
            const operations = this.getPotentialOperations(requestInfo);
            if (operations.reason !== undefined) {
                this.logging(`${operations.reason.message} with requestUrl ${requestInfo.requestUrl}`, liveValidator_1.LiveValidatorLoggingLevels.info, liveValidator_1.LiveValidatorLoggingTypes.trace, "Oav.OperationSearcher.search.getPotentialOperations", undefined, requestInfo);
            }
            return operations;
        };
        let potentialOperations = searchOperation();
        const firstReason = potentialOperations.reason;
        if (potentialOperations.matches.length === 0) {
            this.logging(`Fallback to ${constants_1.unknownResourceProvider} -> ${constants_1.unknownApiVersion}`, liveValidator_1.LiveValidatorLoggingLevels.info, liveValidator_1.LiveValidatorLoggingTypes.trace, "Oav.OperationSearcher.search", undefined, requestInfo);
            //requestInfo.apiVersion = unknownApiVersion;
            potentialOperations = searchOperation();
        }
        if (potentialOperations.matches.length === 0) {
            throw firstReason !== null && firstReason !== void 0 ? firstReason : potentialOperations.reason;
        }
        if (potentialOperations.matches.length > 1) {
            const operationInfos = [];
            potentialOperations.matches.forEach(({ operation }) => {
                const specPath = operation._path._spec._filePath;
                operationInfos.push({
                    id: operation.operationId,
                    path: operation._path._pathTemplate,
                    specPath,
                });
            });
            const msg = `Found multiple matching operations ` +
                `for request url "${requestInfo.requestUrl}" with HTTP Method "${requestInfo.requestMethod}".` +
                `Operation Information: ${JSON.stringify(operationInfos)}`;
            this.logging(msg, liveValidator_1.LiveValidatorLoggingLevels.info, liveValidator_1.LiveValidatorLoggingTypes.trace, "Oav.OperationSearcher.Search", undefined, requestInfo);
            const e = new models_1.LiveValidationError(constants_1.ErrorCodes.MultipleOperationsFound.name, msg);
            throw e;
        }
        this.logging("Complete operation search", liveValidator_1.LiveValidatorLoggingLevels.info, liveValidator_1.LiveValidatorLoggingTypes.perfTrace, "Oav.OperationSearcher.Search", Date.now() - startTime, requestInfo);
        return {
            operationMatch: potentialOperations.matches[0],
            apiVersion: potentialOperations.apiVersion,
        };
    }
    /**
     * Gets list of potential operations objects for given url and method.
     *
     * @param  requestInfo The parsed request info for which to find potential operations.
     *
     * @returns Potential operation result object.
     */
    getPotentialOperations(requestInfo) {
        if (this.cache.size === 0) {
            const msgStr = `Please call "liveValidator.initialize()" before calling this method, ` +
                `so that cache is populated.`;
            throw new Error(msgStr);
        }
        const ret = {
            matches: [],
            resourceProvider: requestInfo.providerNamespace,
            apiVersion: requestInfo.apiVersion,
        };
        if (requestInfo.pathStr === "") {
            ret.reason = new models_1.LiveValidationError(constants_1.ErrorCodes.PathNotFoundInRequestUrl.name, `Could not find path from requestUrl: "${requestInfo.requestUrl}".`);
            return ret;
        }
        // Search using provider
        const allApiVersions = this.cache.get(requestInfo.providerNamespace);
        let meta;
        if (allApiVersions === undefined) {
            // provider does not exist in cache
            meta = (0, errorDefinitions_1.getOavErrorMeta)(constants_1.ErrorCodes.OperationNotFoundInCacheWithProvider.name, {
                providerNamespace: requestInfo.providerNamespace,
            });
            ret.reason = new models_1.LiveValidationError(constants_1.ErrorCodes.OperationNotFoundInCacheWithProvider.name, meta.message);
            return ret;
        }
        // Search using api-version found in the requestUrl
        if (!requestInfo.apiVersion) {
            ret.reason = new models_1.LiveValidationError(constants_1.ErrorCodes.OperationNotFoundInCacheWithApi.name, `Could not find api-version in requestUrl "${requestInfo.requestUrl}".`);
            return ret;
        }
        const allMethods = allApiVersions.get(requestInfo.apiVersion);
        if (allMethods === undefined) {
            meta = (0, errorDefinitions_1.getOavErrorMeta)(constants_1.ErrorCodes.OperationNotFoundInCacheWithApi.name, {
                apiVersion: requestInfo.apiVersion,
                providerNamespace: requestInfo.providerNamespace,
            });
            ret.reason = new models_1.LiveValidationError(constants_1.ErrorCodes.OperationNotFoundInCacheWithApi.name, meta.message);
            return ret;
        }
        const operationsForHttpMethod = allMethods === null || allMethods === void 0 ? void 0 : allMethods.get(requestInfo.requestMethod);
        // Search using requestMethod provided by user
        if (operationsForHttpMethod === undefined) {
            meta = (0, errorDefinitions_1.getOavErrorMeta)(constants_1.ErrorCodes.OperationNotFoundInCacheWithVerb.name, {
                requestMethod: requestInfo.requestMethod,
                apiVersion: requestInfo.apiVersion,
                providerNamespace: requestInfo.providerNamespace,
            });
            ret.reason = new models_1.LiveValidationError(constants_1.ErrorCodes.OperationNotFoundInCacheWithVerb.name, meta.message);
            return ret;
        }
        // Find the best match using regex on path
        ret.matches = getMatchedOperations(requestInfo.host, requestInfo.pathStr, operationsForHttpMethod, requestInfo.query);
        if (ret.matches.length === 0 && ret.reason === undefined) {
            meta = (0, errorDefinitions_1.getOavErrorMeta)(constants_1.ErrorCodes.OperationNotFoundInCache.name, {
                requestMethod: requestInfo.requestMethod,
                apiVersion: requestInfo.apiVersion,
                providerNamespace: requestInfo.providerNamespace,
            });
            ret.reason = new models_1.LiveValidationError(constants_1.ErrorCodes.OperationNotFoundInCache.name, meta.message);
        }
        return ret;
    }
}
exports.OperationSearcher = OperationSearcher;
/**
 * Gets list of matched operations objects for given url.
 *
 * @param {string} requestUrl The url for which to find matched operations.
 *
 * @param {Array<Operation>} operations The list of operations to search.
 *
 * @returns {Array<Operation>} List of matched operations with the request url.
 */
const getMatchedOperations = (host, pathStr, operations, query) => {
    const queryMatchResult = []; // Priority 0
    const result = []; // Priority 1
    const multiParamResult = []; // Priority 2
    for (const operation of operations) {
        const path = operation._path;
        // Validate query first so we could match operation in x-ms-paths
        const queryMatch = path._validateQuery === undefined
            ? undefined
            : path._validateQuery({ isResponse: false }, query).length === 0;
        if (queryMatch === false) {
            continue;
        }
        const toMatch = path._pathRegex._hostTemplate ? host + pathStr : pathStr;
        const pathMatch = path._pathRegex.exec(toMatch);
        if (pathMatch === null) {
            continue;
        }
        (queryMatch !== undefined
            ? queryMatchResult
            : path._pathRegex._hasMultiPathParam
                ? multiParamResult
                : result).push({
            operation,
            pathRegex: path._pathRegex,
            pathMatch,
        });
    }
    return queryMatchResult.length > 0
        ? queryMatchResult
        : result.length > 0
            ? result
            : multiParamResult;
};
