// oav/dist/lib/liveValidation/operationValidator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.issueFromErrorCode = exports.schemaValidateIssueToLiveValidationIssue = exports.transformMapValue = exports.transformLiveHeader = exports.transformBodyValue = exports.validateSwaggerLiveResponse = exports.validateSwaggerLiveRequest = void 0;
const openapi_tools_common_1 = require("@azure-tools/openapi-tools-common");
const ajvSchemaValidator_1 = require("../swaggerValidator/ajvSchemaValidator");
const jsonUtils_1 = require("../util/jsonUtils");
const pathRegexTransformer_1 = require("../transform/pathRegexTransformer");
const errorDefinitions_1 = require("../util/errorDefinitions");
const liveValidator_1 = require("./liveValidator");
const validateSwaggerLiveRequest = async (request, operationContext, loader, includeErrors, isArmCall, logging) => {
    var _a;
    const { operation } = operationContext.operationMatch;
    const { body, query } = request;
    const result = [];
    let validate = operation._validate;
    if (validate === undefined) {
        if (loader === undefined) {
            throw new Error("Loader is undefined but request validator isn't built yet");
        }
        const startTimeToBuild = Date.now();
        validate = await loader.getRequestValidator(operation);
        const elapsedTime = Date.now() - startTimeToBuild;
        if (logging) {
            logging(`On-demand build request validator with DurationInMs:${elapsedTime}`, liveValidator_1.LiveValidatorLoggingLevels.debug, liveValidator_1.LiveValidatorLoggingTypes.trace, "Oav.OperationValidator.validateSwaggerLiveRequest.loader.getRequestValidator", undefined, operationContext.validationRequest);
            logging(`On-demand build request validator`, liveValidator_1.LiveValidatorLoggingLevels.info, liveValidator_1.LiveValidatorLoggingTypes.perfTrace, "Oav.OperationValidator.validateSwaggerLiveRequest.loader.getRequestValidator", elapsedTime, operationContext.validationRequest);
        }
    }
    const pathParam = (0, pathRegexTransformer_1.extractPathParamValue)(operationContext.operationMatch);
    (0, exports.transformMapValue)(pathParam, operation._pathTransform);
    (0, exports.transformMapValue)(query, operation._queryTransform);
    const headers = (0, exports.transformLiveHeader)((_a = request.headers) !== null && _a !== void 0 ? _a : {}, operation);
    validateContentType(operation.consumes, headers, true, result);
    // for rpaas calls, temp solution to log invalid_type errors for additional properties
    // rather than returning the error to rpaas
    const ctx = { isResponse: false, includeErrors: includeErrors };
    const errors = validate(ctx, {
        path: pathParam,
        body: (0, exports.transformBodyValue)(body, operation),
        headers,
        query,
    });
    (0, exports.schemaValidateIssueToLiveValidationIssue)(errors, operation, ctx, result, operationContext, isArmCall, logging, body);
    return result;
};
exports.validateSwaggerLiveRequest = validateSwaggerLiveRequest;
const validateSwaggerLiveResponse = async (response, operationContext, loader, includeErrors, isArmCall, logging) => {
    var _a;
    const { operation } = operationContext.operationMatch;
    const { statusCode, body } = response;
    const rspDef = operation.responses;
    const result = [];
    let rsp = rspDef[statusCode];
    const realCode = parseInt(statusCode, 10);
    if (rsp === undefined && 400 <= realCode && realCode <= 599) {
        rsp = rspDef.default;
    }
    if (rsp === undefined) {
        result.push((0, exports.issueFromErrorCode)("INVALID_RESPONSE_CODE", { statusCode }, rspDef));
        return result;
    }
    let validate = rsp._validate;
    if (validate === undefined) {
        if (loader === undefined) {
            throw new Error("Loader is undefined but request validator isn't built yet");
        }
        const startTimeToBuild = Date.now();
        validate = await loader.getResponseValidator(rsp);
        const elapsedTime = Date.now() - startTimeToBuild;
        if (logging) {
            logging(`On-demand build response validator with DurationInMs:${elapsedTime}`, liveValidator_1.LiveValidatorLoggingLevels.debug, liveValidator_1.LiveValidatorLoggingTypes.trace, "Oav.OperationValidator.validateSwaggerLiveResponse.loader.getResponseValidator", undefined, operationContext.validationRequest);
            logging(`On-demand build request validator`, liveValidator_1.LiveValidatorLoggingLevels.info, liveValidator_1.LiveValidatorLoggingTypes.perfTrace, "Oav.OperationValidator.validateSwaggerLiveResponse.loader.getResponseValidator", elapsedTime, operationContext.validationRequest);
        }
    }
    const headers = (0, exports.transformLiveHeader)((_a = response.headers) !== null && _a !== void 0 ? _a : {}, rsp);
    if (rsp.schema !== undefined) {
        validateContentType(operation.produces, headers, false, result);
        if (isArmCall && realCode >= 200 && realCode < 300) {
            validateLroOperation(operation, statusCode, headers, result);
        }
    }
    const ctx = {
        isResponse: true,
        includeErrors: includeErrors,
        statusCode,
        httpMethod: operation._method,
    };
    const errors = validate(ctx, {
        headers,
        body,
    });
    (0, exports.schemaValidateIssueToLiveValidationIssue)(errors, operation, ctx, result, operationContext, isArmCall, logging, body);
    return result;
};
exports.validateSwaggerLiveResponse = validateSwaggerLiveResponse;
const transformBodyValue = (body, operation) => {
    return operation._bodyTransform === undefined ? body : operation._bodyTransform(body);
};
exports.transformBodyValue = transformBodyValue;
const transformLiveHeader = (headers, it) => {
    const result = {};
    for (const headerName of Object.keys(headers)) {
        result[headerName.toLowerCase()] = headers[headerName];
    }
    (0, exports.transformMapValue)(result, it._headerTransform);
    return result;
};
exports.transformLiveHeader = transformLiveHeader;
const transformMapValue = (data, transforms) => {
    if (transforms === undefined || data === undefined) {
        return;
    }
    for (const key of Object.keys(transforms)) {
        const transform = transforms[key];
        const val = data[key];
        if (typeof val === "string") {
            data[key] = transform(val);
        }
        else if (Array.isArray(val)) {
            data[key] = val.map(transform);
        }
    }
};
exports.transformMapValue = transformMapValue;
const validateContentType = (allowedContentTypes, headers, isRequest, result) => {
    var _a;
    const contentType = ((_a = headers["content-type"]) === null || _a === void 0 ? void 0 : _a.split(";")[0]) || (isRequest ? undefined : "application/octet-stream");
    if (contentType !== undefined && !allowedContentTypes.includes(contentType)) {
        // in some cases, produces value could have colon in type like 'application/json;odata=minimalmetadata'
        for (const allowedContentType of allowedContentTypes) {
            if (allowedContentType.includes(";")) {
                const subAllowedContentType = allowedContentType.split(";")[0];
                if (subAllowedContentType.includes(contentType)) {
                    return;
                }
            }
        }
        result.push((0, exports.issueFromErrorCode)("INVALID_CONTENT_TYPE", {
            contentType,
            supported: allowedContentTypes.join(", "),
        }));
    }
};
/**
 * Finds the resource ID for a given JSON path. Example inputs:
 *   - $.properties.lastname
 *   - $.properties.groups[0].variable
 * @param bodyPayload The full payload object.
 * @param jsonPath The JSON path referring to the problematic property.
 * @returns The resource ID, or undefined if not found.
 */
const findResourceId = (bodyPayload, jsonPath) => {
    // schemaValidateIssueToLiveValidationIssue will provide a valid jsonPath to this function
    const keys = jsonPath
        .replace(/^\$\./, "") // Remove the leading "$."
        .split(/\.|\[(\d+)\]/) // Split by dots or array brackets
        .filter((key) => key !== undefined && key !== "");
    let current = bodyPayload;
    const stack = [];
    for (const key of keys) {
        if (current && typeof current === "object") {
            stack.push(current);
            if (Array.isArray(current)) {
                // Handle array indices
                const index = parseInt(key, 10);
                if (!isNaN(index) && index < current.length) {
                    current = current[index];
                }
                else {
                    return undefined;
                }
            }
            else if (key in current) {
                current = current[key];
            }
            else {
                return undefined;
            }
        }
        else {
            return undefined;
        }
    }
    // notice we only ever check for parent id. This means that we will never accidentally grab the value of an id
    // that has been added erroneously to the payload. (for instance if payload is to an id field that SHOULD NOT be set.)
    for (let i = stack.length - 1; i >= 0; i--) {
        const parent = stack[i];
        if (parent && typeof parent === "object" && "id" in parent) {
            return parent.id;
        }
    }
    return undefined; // No `id` field found
};
const schemaValidateIssueToLiveValidationIssue = (input, operation, ctx, output, _operationContext, _isArmCall, _logging, _bodyPayload) => {
    for (const i of input) {
        const issue = i;
        issue.resourceIds = [];
        issue.documentationUrl = "";
        const source = issue.source;
        if (!source.url) {
            source.url = operation._path._spec._filePath;
        }
        let skipIssue = false;
        issue.pathsInPayload = issue.jsonPathsInPayload.map((path, idx) => {
            if (issue.code === "MISSING_RESOURCE_ID") {
                // ignore this error for sub level resources
                if (path.includes("properties")) {
                    skipIssue = true;
                    return "";
                }
            }
            const isMissingRequiredProperty = issue.code === "OBJECT_MISSING_REQUIRED_PROPERTY";
            const isBodyIssue = path.startsWith(".body");
            if (isBodyIssue && (path.length > 5 || !isMissingRequiredProperty)) {
                path = "$" + path.substr(5);
                issue.jsonPathsInPayload[idx] = path;
                const resolvedJsonPath = (0, jsonUtils_1.jsonPathToPointer)(path);
                if (ctx.isResponse && isBodyIssue) {
                    const resourceId = findResourceId(_bodyPayload, path);
                    if (resourceId) {
                        issue.resourceIds.push(resourceId);
                    }
                }
                return resolvedJsonPath;
            }
            if (isMissingRequiredProperty) {
                if (ctx.isResponse) {
                    if (isBodyIssue) {
                        issue.code = "INVALID_RESPONSE_BODY";
                        // If a long running operation with code 201 or 202 then it could has empty body
                        if (operation["x-ms-long-running-operation"] &&
                            (ctx.statusCode === "201" || ctx.statusCode === "202")) {
                            skipIssue = true;
                        }
                    }
                    else if (path.startsWith(".headers")) {
                        issue.code = "INVALID_RESPONSE_HEADER";
                    }
                }
                else {
                    // In request
                    issue.code = "MISSING_REQUIRED_PARAMETER";
                }
                const meta = (0, errorDefinitions_1.getOavErrorMeta)(issue.code, { missingProperty: issue.params[0] });
                issue.severity = meta.severity;
                issue.message = meta.message;
            }
            const resolvedJsonPath = (0, jsonUtils_1.jsonPathToPointer)(path);
            if (ctx.isResponse && isBodyIssue) {
                const resourceId = findResourceId(_bodyPayload, path);
                if (resourceId) {
                    issue.resourceIds.push(resourceId);
                }
            }
            return resolvedJsonPath;
        });
        if (!skipIssue) {
            output.push(issue);
        }
    }
};
exports.schemaValidateIssueToLiveValidationIssue = schemaValidateIssueToLiveValidationIssue;
const validateLroOperation = (operation, statusCode, headers, result) => {
    if (operation["x-ms-long-running-operation"] === true) {
        if (operation._method === "post") {
            if (statusCode === "202" || statusCode === "201") {
                validateLroHeader(operation, statusCode, headers, result);
            }
            else if (statusCode !== "200" && statusCode !== "204") {
                result.push((0, exports.issueFromErrorCode)("LRO_RESPONSE_CODE", { statusCode }, operation.responses));
            }
        }
        else if (operation._method === "patch") {
            if (statusCode === "202" || statusCode === "201") {
                validateLroHeader(operation, statusCode, headers, result);
            }
            else if (statusCode !== "200") {
                result.push((0, exports.issueFromErrorCode)("LRO_RESPONSE_CODE", { statusCode }, operation.responses));
            }
        }
        else if (operation._method === "delete") {
            if (statusCode === "202") {
                validateLroHeader(operation, statusCode, headers, result);
            }
            else if (statusCode !== "200" && statusCode !== "204") {
                result.push((0, exports.issueFromErrorCode)("LRO_RESPONSE_CODE", { statusCode }, operation.responses));
            }
        }
        else if (operation._method === "put") {
            if (statusCode === "202" || statusCode === "201") {
                validateLroHeader(operation, statusCode, headers, result);
            }
            else if (statusCode !== "200") {
                result.push((0, exports.issueFromErrorCode)("LRO_RESPONSE_CODE", { statusCode }, operation.responses));
            }
        }
    }
};
const validateLroHeader = (operation, statusCode, headers, result) => {
    if (statusCode === "201") {
        // Ignore LRO header check cause RPC says azure-AsyncOperation is optional if using 201/200+ provisioningState
        return;
    }
    if ((headers.location === undefined || headers.location === "") &&
        (headers["azure-AsyncOperation"] === undefined || headers["azure-AsyncOperation"] === "") &&
        (headers["azure-asyncoperation"] === undefined || headers["azure-asyncoperation"] === "")) {
        result.push((0, exports.issueFromErrorCode)("LRO_RESPONSE_HEADER", {
            header: "location or azure-AsyncOperation",
        }, operation.responses));
    }
};
const issueFromErrorCode = (code, param, relatedSchema) => {
    const meta = (0, errorDefinitions_1.getOavErrorMeta)(code, param);
    return {
        code,
        severity: meta.severity,
        message: meta.message,
        jsonPathsInPayload: [],
        pathsInPayload: [],
        schemaPath: "",
        source: (0, ajvSchemaValidator_1.sourceMapInfoToSourceLocation)((0, openapi_tools_common_1.getInfo)(relatedSchema)),
        documentationUrl: "",
    };
};
exports.issueFromErrorCode = issueFromErrorCode;
