// @azure-tools/openapi-tools-common/dist/libs/async-iterator.js
"use strict";
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncDelegator = (this && this.__asyncDelegator) || function (o) {
    var i, p;
    return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
    function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
};
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.asyncFilter = exports.asyncEmpty = exports.asyncFlatMap = exports.asyncFlatten = exports.asyncMap = exports.asyncEntries = exports.asyncToArray = exports.asyncFold = exports.asyncFromPromise = exports.asyncFromSequence = exports.asyncFromSync = exports.asyncIterable = void 0;
exports.asyncIterable = (createIterator) => {
    const property = (f) => (...p) => f(result, ...p);
    const result = {
        [Symbol.asyncIterator]: createIterator,
        fold: property(exports.asyncFold),
        toArray: property(exports.asyncToArray),
        entries: property(exports.asyncEntries),
        map: property(exports.asyncMap),
        flatMap: property(exports.asyncFlatMap),
        filter: property(exports.asyncFilter),
    };
    return result;
};
exports.asyncFromSync = (input) => exports.asyncIterable(function () { return __asyncGenerator(this, arguments, function* () { yield __await(yield* __asyncDelegator(__asyncValues(input))); }); });
exports.asyncFromSequence = (...a) => exports.asyncFromSync(a);
exports.asyncFromPromise = (p) => exports.asyncIterable(function () { return __asyncGenerator(this, arguments, function* () { yield __await(yield* __asyncDelegator(__asyncValues(yield __await(p)))); }); });
exports.asyncFold = async (input, func, init) => {
    var e_1, _a;
    // tslint:disable-next-line:no-let
    let result = init;
    try {
        /* tslint:disable-next-line:no-loop-statement */
        for (var _b = __asyncValues(exports.asyncEntries(input)), _c; _c = await _b.next(), !_c.done;) {
            const [index, value] = _c.value;
            /* tslint:disable-next-line:no-expression-statement */
            result = await func(result, value, index);
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) await _a.call(_b);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return result;
};
exports.asyncToArray = (input) => exports.asyncFold(input, (a, i) => [...a, i], new Array());
exports.asyncEntries = (input) => exports.asyncIterable(function () {
    return __asyncGenerator(this, arguments, function* () {
        var e_2, _a;
        // tslint:disable-next-line:no-if-statement
        if (input === undefined) {
            return yield __await(void 0);
        }
        // tslint:disable-next-line:no-let
        let index = 0;
        try {
            // tslint:disable-next-line:no-loop-statement
            for (var input_1 = __asyncValues(input), input_1_1; input_1_1 = yield __await(input_1.next()), !input_1_1.done;) {
                const value = input_1_1.value;
                yield yield __await([index, value]
                // tslint:disable-next-line:no-expression-statement
                );
                // tslint:disable-next-line:no-expression-statement
                index += 1;
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (input_1_1 && !input_1_1.done && (_a = input_1.return)) yield __await(_a.call(input_1));
            }
            finally { if (e_2) throw e_2.error; }
        }
    });
});
exports.asyncMap = (input, func) => exports.asyncIterable(function () {
    return __asyncGenerator(this, arguments, function* () {
        var e_3, _a;
        try {
            /* tslint:disable-next-line:no-loop-statement */
            for (var _b = __asyncValues(exports.asyncEntries(input)), _c; _c = yield __await(_b.next()), !_c.done;) {
                const [index, value] = _c.value;
                yield yield __await(func(value, index));
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) yield __await(_a.call(_b));
            }
            finally { if (e_3) throw e_3.error; }
        }
    });
});
exports.asyncFlatten = (input) => exports.asyncIterable(function () {
    return __asyncGenerator(this, arguments, function* () {
        var e_4, _a;
        // tslint:disable-next-line:no-if-statement
        if (input === undefined) {
            return yield __await(void 0);
        }
        try {
            // tslint:disable-next-line:no-loop-statement
            for (var input_2 = __asyncValues(input), input_2_1; input_2_1 = yield __await(input_2.next()), !input_2_1.done;) {
                const v = input_2_1.value;
                // tslint:disable-next-line:no-if-statement
                if (v !== undefined) {
                    yield __await(yield* __asyncDelegator(__asyncValues(v)));
                }
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (input_2_1 && !input_2_1.done && (_a = input_2.return)) yield __await(_a.call(input_2));
            }
            finally { if (e_4) throw e_4.error; }
        }
    });
});
exports.asyncFlatMap = (input, func) => exports.asyncFlatten(exports.asyncMap(input, func));
// tslint:disable-next-line:no-empty
exports.asyncEmpty = () => exports.asyncIterable(function () { return __asyncGenerator(this, arguments, function* () { }); });
exports.asyncFilter = (input, func) => exports.asyncFlatMap(input, (v, i) => exports.asyncIterable(function () {
    return __asyncGenerator(this, arguments, function* () {
        // tslint:disable-next-line:no-if-statement
        if (yield __await(func(v, i))) {
            yield yield __await(v);
        }
    });
}));
