// @azure-tools/openapi-tools-common/dist/libs/fs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.asyncRecursiveRmdir = exports.asyncRecursiveReaddir = exports.asyncUnlink = exports.asyncRmdir = exports.asyncMkdir = exports.asyncReaddir = exports.asyncExists = exports.asyncWriteFile = exports.asyncReadFile = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const asyncIt = __importStar(require("./async-iterator"));
const util = __importStar(require("util"));
exports.asyncReadFile = util.promisify(fs.readFile);
exports.asyncWriteFile = util.promisify(fs.writeFile);
exports.asyncExists = util.promisify(fs.exists);
exports.asyncReaddir = util.promisify(fs.readdir);
exports.asyncMkdir = util.promisify(fs.mkdir);
exports.asyncRmdir = util.promisify(fs.rmdir);
exports.asyncUnlink = util.promisify(fs.unlink);
exports.asyncRecursiveReaddir = (dir) => asyncIt.asyncFromPromise(exports.asyncReaddir(dir, { withFileTypes: true })).flatMap(f => {
    const p = path.join(dir, f.name);
    return f.isDirectory() ? exports.asyncRecursiveReaddir(p) : asyncIt.asyncFromSequence(p);
});
exports.asyncRecursiveRmdir = async (dir) => {
    const list = await exports.asyncReaddir(dir, { withFileTypes: true });
    await Promise.all(list.map(async (f) => {
        const p = path.join(dir, f.name);
        if (f.isDirectory()) {
            await exports.asyncRecursiveRmdir(p);
        }
        else {
            await exports.asyncUnlink(p);
        }
    }));
    await exports.asyncRmdir(dir);
};
