// @azure-tools/openapi-tools-common/dist/libs/json-parser.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseJson = exports.tokenize = exports.defaultErrorReport = void 0;
const json = __importStar(require("./json"));
const iterator = __importStar(require("./iterator"));
const sourceMap = __importStar(require("./source-map"));
const addPosition = __importStar(require("./add-position"));
var fa;
(function (fa) {
    fa.applyState = (input, state) => iterator.iterable(function* () {
        for (const c of input) {
            if (state.next === undefined) {
                break;
            }
            const result = state.next(c);
            if (result !== undefined) {
                if (result.result !== undefined) {
                    yield* result.result;
                }
                if (result.state !== undefined) {
                    state = result.state;
                }
            }
        }
        if (state.done !== undefined) {
            const r = state.done();
            if (r !== undefined) {
                yield r;
            }
        }
    });
    fa.nextState = (result, state, c) => {
        if (state.next === undefined) {
            return { result, state };
        }
        const rs = state.next(c);
        if (rs === undefined) {
            return { result, state };
        }
        return {
            result: rs.result === undefined ? result : [...result, ...rs.result],
            state: rs.state === undefined ? state : rs.state
        };
    };
})(fa || (fa = {}));
var set;
(function (set_1) {
    set_1.create = (v) => new Set(v);
    set_1.isElement = (set, c) => set.has(c);
})(set || (set = {}));
const symbol = set.create(["{", "}", "[", "]", ",", ":"]);
const whiteSpace = set.create([" ", "\t", "\r", "\n", ""]);
const jsonValue = set.create([
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
    "g",
    "h",
    "i",
    "j",
    "k",
    "l",
    "m",
    "n",
    "o",
    "p",
    "q",
    "r",
    "s",
    "t",
    "u",
    "v",
    "w",
    "x",
    "y",
    "z",
    "A",
    "B",
    "C",
    "D",
    "E",
    "F",
    "G",
    "H",
    "I",
    "J",
    "K",
    "L",
    "M",
    "N",
    "O",
    "P",
    "Q",
    "R",
    "S",
    "T",
    "U",
    "V",
    "W",
    "X",
    "Y",
    "Z",
    "_",
    "+",
    "-",
    ".",
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9"
]);
const isControl = (c) => {
    const code = c.charCodeAt(0);
    return code <= 0x1f || code === 0x7f;
};
const escapeMap = {
    '"': '"',
    "\\": "\\",
    "/": "/",
    b: "\b",
    t: "\t",
    f: "\f",
    r: "\r",
    n: "\n"
};
const hexMap = {
    "0": 0,
    "1": 1,
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
    A: 0xa,
    a: 0xa,
    B: 0xb,
    b: 0xb,
    C: 0xc,
    c: 0xc,
    D: 0xd,
    d: 0xd,
    E: 0xe,
    e: 0xe,
    F: 0xf,
    f: 0xf
};
exports.defaultErrorReport = (e) => {
    throw e;
};
exports.tokenize = (s, reportError = exports.defaultErrorReport, url) => {
    const report = (position, token, code) => reportError({
        kind: "syntax",
        code,
        position,
        token,
        message: `${code}, token: ${token}, line: ${position.line}, column: ${position.column}`,
        url
    });
    const whiteSpaceState = {
        next: cp => {
            if (cp.c === '"') {
                return { state: stringState(cp.position) };
            }
            if (set.isElement(symbol, cp.c)) {
                return { result: [{ kind: cp.c, position: cp.position }] };
            }
            if (set.isElement(jsonValue, cp.c)) {
                return { state: jsonValueState(cp) };
            }
            if (!set.isElement(whiteSpace, cp.c)) {
                report(cp.position, cp.c, "invalid symbol");
            }
            return;
        }
    };
    const stringState = (position) => {
        let value = "";
        const getResult = () => ({ kind: "value", value, position });
        const done = () => {
            report(position, value, "unexpected end of string");
            return getResult();
        };
        const state = {
            next: cp => {
                if (cp.c === '"') {
                    return {
                        result: [getResult()],
                        state: whiteSpaceState
                    };
                }
                if (isControl(cp.c)) {
                    report(cp.position, cp.c, "invalid symbol");
                }
                if (cp.c === "\\") {
                    return { state: escapeState };
                }
                value += cp.c;
                return;
            },
            done
        };
        const escapeState = {
            next: cp => {
                if (cp.c === "u") {
                    return { state: unicodeState() };
                }
                const e = escapeMap[cp.c];
                if (e === undefined) {
                    report(cp.position, cp.c, "invalid escape symbol");
                    value += cp.c;
                }
                else {
                    value += e;
                }
                return { state };
            },
            done
        };
        // UNICODE escape sequence
        const unicodeState = () => {
            let i = 0;
            let u = 0;
            return {
                next: cp => {
                    const h = hexMap[cp.c];
                    if (h === undefined) {
                        report(cp.position, cp.c, "invalid escape symbol");
                        return { state };
                    }
                    u = (u << 4) | h;
                    ++i;
                    // always for symbols https://json.org/
                    if (i < 4) {
                        return;
                    }
                    value += String.fromCharCode(u);
                    return { state };
                },
                done
            };
        };
        return state;
    };
    const jsonValueState = (prior) => {
        let value = prior.c;
        const getResultValue = () => {
            switch (value) {
                case "true":
                    return true;
                case "false":
                    return false;
                case "null":
                    return null;
            }
            const number = parseFloat(value);
            if (isNaN(number)) {
                report(prior.position, value, "invalid token");
                return value;
            }
            return number;
        };
        const done = () => ({
            kind: "value",
            value: getResultValue(),
            position: prior.position
        });
        return {
            next: cp => {
                if (set.isElement(jsonValue, cp.c)) {
                    value += cp.c;
                    return;
                }
                return fa.nextState([done()], whiteSpaceState, cp);
            },
            done
        };
    };
    return fa.applyState(addPosition.addPosition(s), whiteSpaceState);
};
exports.parseJson = (url, context, reportError = exports.defaultErrorReport) => {
    const report = (position, token, code) => reportError({
        kind: "structure",
        code,
        position,
        token,
        message: `${code}, token: ${token}, line: ${position.line}, column: ${position.column}`,
        url
    });
    const reportToken = (token, message) => report(token.position, token.kind === "value" ? JSON.stringify(token.value) : token.kind, message);
    const endState = {
        next: t => {
            reportToken(t, "unexpected token");
            return { state: {} };
        }
    };
    const objectState = (os) => {
        const separatorState = {
            next: t => {
                switch (t.kind) {
                    case "}":
                        return { state: os.state };
                    case ",":
                        return { state: propertyState };
                }
                reportToken(t, "unexpected token");
                return;
            }
        };
        const propertyValueState = (name) => ({
            next: t => {
                if (t.kind === ":") {
                    return {
                        state: valueState(separatorState, (v, position, primitiveProperties) => {
                            os.value[name] = v;
                            if (json.isPrimitive(v)) {
                                os.primitiveProperties[name] = position;
                            }
                            return {
                                isChild: true,
                                position,
                                parent: os.value,
                                property: name,
                                primitiveProperties
                            };
                        })
                    };
                }
                reportToken(t, "unexpected token");
                return;
            }
        });
        const propertyState = {
            next: t => {
                if (t.kind !== "value") {
                    reportToken(t, "unexpected token");
                    return;
                }
                let name = t.value;
                if (name === null) {
                    reportToken(t, "expecting property name");
                    name = "null";
                }
                else if (typeof name !== "string") {
                    reportToken(t, "expecting property name");
                    name = name.toString();
                }
                return { state: propertyValueState(name) };
            }
        };
        return {
            next: t => {
                if (t.kind === "}") {
                    return { state: os.state };
                }
                return propertyState.next === undefined
                    ? undefined
                    : propertyState.next(t);
            }
        };
    };
    const arrayState = (as) => {
        const separatorState = {
            next: t => {
                switch (t.kind) {
                    case "]":
                        return { state: as.state };
                    case ",":
                        return { state: itemState };
                }
                reportToken(t, "unexpected token");
                return;
            }
        };
        const itemState = valueState(separatorState, (v, position, primitiveProperties) => {
            const property = as.value.push(v) - 1;
            if (json.isPrimitive(v)) {
                as.primitiveProperties[property] = position;
            }
            return {
                isChild: true,
                parent: as.value,
                position,
                property,
                primitiveProperties
            };
        });
        return {
            next: t => {
                if (t.kind === "]") {
                    return { state: as.state };
                }
                return itemState.next !== undefined ? itemState.next(t) : undefined;
            }
        };
    };
    const valueState = (state, set) => ({
        next: t => {
            const updateRef = (value) => {
                const primitiveProperties = {};
                const info = set(value, t.position, primitiveProperties);
                return {
                    state,
                    value: sourceMap.setInfo(value, info),
                    primitiveProperties
                };
            };
            switch (t.kind) {
                case "value":
                    set(t.value, t.position, {});
                    return { state };
                case "{":
                    const objectRef = updateRef({});
                    return { state: objectState(objectRef) };
                case "[":
                    const arrayRef = updateRef([]);
                    return { state: arrayState(arrayRef) };
            }
            reportToken(t, "unexpected token");
            return;
        }
    });
    const tokens = exports.tokenize(context, reportError, url);
    let value;
    fa.applyState(tokens, valueState(endState, (v, position, primitiveProperties) => {
        value = v;
        return {
            isChild: false,
            position,
            url,
            primitiveProperties
        };
    })).toArray();
    if (value === undefined) {
        report({ line: 1, column: 1 }, "", "unexpected end of file");
        return null;
    }
    return value;
};
