// @azure-tools/openapi-tools-common/dist/libs/markdown-transformer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.markDownExToString = exports.parseMarkdown = exports.iterate = exports.createCodeBlock = exports.createText = exports.createNode = void 0;
const cm = __importStar(require("commonmark"));
const it = __importStar(require("./iterator"));
const fm = require('front-matter');
exports.createNode = (type, ...children) => {
    const result = new cm.Node(type);
    for (const c of children) {
        result.appendChild(c);
    }
    return result;
};
exports.createText = (literal) => {
    const result = exports.createNode("text");
    result.literal = literal;
    return result;
};
exports.createCodeBlock = (info, literal) => {
    const result = exports.createNode("code_block");
    result.info = info;
    result.literal = literal;
    return result;
};
exports.iterate = (node) => it.iterable(function* () {
    let c = node.firstChild;
    while (c !== null) {
        yield c;
        c = c.next;
    }
});
exports.parseMarkdown = (fileContent) => {
    const result = fm(fileContent);
    const parser = new cm.Parser();
    return {
        frontMatter: result.frontmatter,
        markDown: parser.parse(result.body)
    };
};
exports.markDownExToString = (mde) => {
    const md = unescape(commonmarkToString(mde.markDown));
    return mde.frontMatter === undefined ? md : `---\n${mde.frontMatter}\n---\n${md}`;
};
const commonmarkToString = (root) => {
    const walker = root.walker();
    let event;
    let output = "";
    while ((event = walker.next())) {
        const curNode = event.node;
        const leaving = render.leaving[curNode.type];
        if (!event.entering && leaving !== undefined) {
            output += leaving(curNode, event.entering);
        }
        const entering = render.entering[curNode.type];
        if (event.entering && entering !== undefined) {
            output += entering(curNode, event.entering);
        }
    }
    output = output.replace(/\n$/, "");
    return output;
};
const indent = (node) => node !== null ? indent(node.parent) + (node.type === "item" ? "  " : "") : "";
const render = {
    entering: {
        text: (node) => node.literal,
        softbreak: () => "\n",
        linebreak: () => "\n",
        emph: () => "*",
        strong: () => "**",
        html_inline: () => "`",
        link: () => "[",
        image: () => { },
        code: (node) => `\`${node.literal}\``,
        document: () => "",
        paragraph: () => "",
        block_quote: () => "> ",
        item: (node) => `${indent(node.parent)}${{ bullet: "*", ordered: `1${node.listDelimiter}` }[node.listType]} `,
        list: () => "",
        heading: (node) => Array(node.level)
            .fill("#")
            .join("") + " ",
        code_block: (node) => `\`\`\` ${node.info}\n${node.literal}\`\`\`\n\n`,
        html_block: (node) => node.literal,
        thematic_break: () => "---\n\n",
        custom_inline: () => { },
        custom_block: () => { },
    },
    leaving: {
        heading: () => "\n\n",
        paragraph: () => "\n\n",
        link: (node) => `](${node.destination})`,
        strong: () => "**",
        emph: () => "*",
    }
};
