// @azure-tools/openapi-tools-common/dist/libs/property-set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.asMutable = exports.copyCreate = exports.create = exports.setMutableProperty = exports.copyProperty = exports.forEach = void 0;
exports.forEach = (source, func) => {
    // tslint:disable-next-line:forin
    for (const key in source) {
        func(source[key], key);
    }
};
exports.copyProperty = (value) => (k) => value[k];
const fromMutableOptional = (v) => v;
/**
 * Use this functions to assign optional properties.
 * The function will delete the property `k` if the given value is `undefined`.
 *
 * @param mutable an object
 * @param k a name of property
 * @param v a property value
 */
exports.setMutableProperty = (mutable, k, v) => {
    if (v !== undefined) {
        mutable[k] = v;
    }
    else {
        // tslint:disable-next-line:no-dynamic-delete
        delete mutable[k];
    }
};
exports.create = (factory) => {
    const result = {};
    exports.forEach(factory, (propertyFactory, k) => { exports.setMutableProperty(result, k, propertyFactory(k)); });
    return fromMutableOptional(result);
};
exports.copyCreate = (source, factory) => {
    const result = {};
    exports.forEach(source, (v, k) => {
        exports.setMutableProperty(result, k, v);
    });
    let changes = false;
    exports.forEach(factory, (propertyFactory, k) => {
        // tslint:disable-next-line:strict-type-predicates
        if (propertyFactory !== undefined) {
            const sourceValue = source[k];
            const newValue = propertyFactory(sourceValue, k);
            // tslint:disable-next-line:strict-comparisons
            if (sourceValue !== newValue) {
                exports.setMutableProperty(result, k, newValue);
                changes = true;
            }
        }
    });
    return changes ? fromMutableOptional(result) : source;
};
exports.asMutable = (source) => source;
