// @azure-tools/openapi-tools-common/dist/libs/string-map.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isMapEmpty = exports.isMapEqual = exports.isMatch = exports.merge = exports.buildMap = exports.stringMap = exports.groupBy = exports.values = exports.keys = exports.objectEntries = exports.mapEntries = exports.objectAllKeys = exports.allKeys = exports.toStringMap = exports.entryValue = exports.entryKey = void 0;
const _ = __importStar(require("./iterator"));
// export const entry: <T>(key: string, value: T) => Entry<T> = tuple2
exports.entryKey = (e) => e[0 /* Key */];
exports.entryValue = (e) => e[1 /* Value */];
exports.toStringMap = (v) => v;
exports.allKeys = (input) => exports.objectAllKeys(input);
exports.objectAllKeys = (input) => _.iterable(function* () {
    // tslint:disable-next-line:no-if-statement
    if (input === undefined || input === null) {
        return;
    }
    // tslint:disable-next-line:no-loop-statement
    for (const key in input) {
        yield key;
    }
});
exports.mapEntries = (input) => exports.objectEntries(input);
exports.objectEntries = (input) => {
    // tslint:disable-next-line:no-if-statement
    if (input === undefined || input === null) {
        return _.empty();
    }
    return exports.objectAllKeys(input).filterMap((key) => {
        const value = input[key];
        return value !== undefined ? [key, value] : undefined;
    });
};
exports.keys = (input) => exports.mapEntries(input).map(exports.entryKey);
exports.values = (input) => exports.mapEntries(input).map(exports.entryValue);
exports.groupBy = (input, reduceFunc) => {
    /* tslint:disable-next-line:readonly-keyword */
    const result = {};
    _.forEach(input, ([key, value]) => {
        const prior = result[key];
        /* tslint:disable-next-line:no-object-mutation no-expression-statement */
        result[key] = prior === undefined ? value : reduceFunc(prior, value);
    });
    return result;
};
exports.stringMap = (input) => 
// tslint:disable-next-line:variable-name
exports.groupBy(input, (_a, b) => b);
exports.buildMap = (source, f) => exports.stringMap(exports.mapEntries(source).map(([k, v]) => [k, f(v, k)]));
exports.merge = (...a) => exports.stringMap(_.map(a, exports.mapEntries).flatMap((v) => v));
// Performs a partial deep comparison between object and source to determine if object contains
// equivalent property values.
// See also https://lodash.com/docs/4.17.10#isMatch
exports.isMatch = (object, source) => exports.mapEntries(source).every(([key, value]) => _.isStrictEqual(object[key], value));
exports.isMapEqual = (a, b) => _.isStrictEqual(a, b) || (exports.isMatch(a, b) && exports.isMatch(b, a));
exports.isMapEmpty = (a) => exports.mapEntries(a).isEmpty();
