// @azure-tools/openapi-tools-common/dist/libs/virtual-fs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.pathDirName = exports.exists = exports.pathJoin = exports.pathResolve = exports.readFile = exports.urlParse = void 0;
const fsp = __importStar(require("./fs"));
const it = __importStar(require("./iterator"));
const node_fetch_1 = __importDefault(require("node-fetch"));
const path = __importStar(require("path"));
const async_retry_1 = __importDefault(require("async-retry"));
const protocolSeparator = "://";
const toUrlString = (url) => url.protocol + protocolSeparator + url.path;
exports.urlParse = (dir) => {
    const split = dir.split(protocolSeparator);
    return split.length === 2
        ? {
            protocol: split[0],
            path: split[1]
        }
        : undefined;
};
exports.readFile = async (pathStr) => {
    const result = exports.urlParse(pathStr);
    return result === undefined
        ? (await fsp.asyncReadFile(pathStr)).toString()
        : await getByUrl(pathStr);
};
const getByUrl = async (url) => {
    return async_retry_1.default(async (bail, retryNr) => {
        try {
            const response = await node_fetch_1.default(url);
            const body = await response.text();
            if (response.status !== 200) {
                const msg = `StatusCode: "${response.status}", ResponseBody: "${body}."`;
                bail(new Error(msg));
            }
            return body;
        }
        catch (fetchError) {
            const message = `Request to ${url} failed with error ${fetchError} on retry number ${retryNr}.`;
            throw new Error(message);
        }
    }, {
        retries: 3,
        factor: 1
    });
};
exports.pathResolve = (dir) => exports.urlParse(dir) !== undefined ? dir : path.resolve(dir);
exports.pathJoin = (dir, value) => {
    const url = exports.urlParse(dir);
    return url !== undefined
        ? toUrlString({
            protocol: url.protocol,
            path: url.path.split("/").concat([value]).join("/")
        })
        : path.join(dir, value);
};
exports.exists = async (dir) => {
    if (exports.urlParse(dir) !== undefined) {
        try {
            return await async_retry_1.default(async () => {
                const { status } = await node_fetch_1.default(dir);
                if (status !== 200) {
                    const msg = `StatusCode "${status}" for "${dir}"`;
                    throw new Error(msg);
                }
                return true;
            }, {
                retries: 3,
                factor: 1
            });
        }
        catch (e) {
            return false;
        }
    }
    else {
        return fsp.asyncExists(dir);
    }
};
exports.pathDirName = (dir) => {
    const url = exports.urlParse(dir);
    if (url === undefined) {
        return path.dirname(dir);
    }
    const split = url.path.split("/");
    return toUrlString({
        protocol: url.protocol,
        path: split.length <= 1 ? url.path : it.join(it.dropRight(split), "/")
    });
};
