// @typespec/openapi3/dist/src/cli/actions/convert/utils/context.js
import { SchemaToExpressionGenerator } from "../generators/generate-types.js";
import { generateNamespaceName } from "./generate-namespace-name.js";
export function createContext(openApi3Doc, logger, namespace) {
    const rootNamespace = namespace
        ? generateNamespaceName(namespace)
        : generateNamespaceName(openApi3Doc.info.title);
    const schemaExpressionGenerator = new SchemaToExpressionGenerator(rootNamespace);
    // Track schemas that are used in multipart forms with their encoding information
    const multipartSchemas = new Map();
    // Track schemas that represent SSE events
    const sseEventSchemas = new Set();
    // Track if SSE features are used
    let sseUsed = false;
    // Create a default no-op logger if none provided
    const defaultLogger = {
        trace: () => { },
        warn: () => { },
        error: () => { },
    };
    const context = {
        openApi3Doc,
        rootNamespace,
        logger: logger ?? defaultLogger,
        getRefName(ref, callingScope) {
            return schemaExpressionGenerator.getRefName(ref, callingScope);
        },
        generateTypeFromRefableSchema(schema, callingScope, isHttpPart = false, encoding) {
            return schemaExpressionGenerator.generateTypeFromRefableSchema(schema, callingScope, isHttpPart, encoding, this);
        },
        getSchemaByRef(ref) {
            return this.getByRef(ref);
        },
        getByRef(ref) {
            const splitRef = ref.split("/");
            const componentKind = splitRef[2]; // #/components/schemas/Pet -> components
            const componentName = splitRef[3]; // #/components/schemas/Pet -> Pet
            switch (componentKind) {
                case "schemas":
                    return openApi3Doc.components?.schemas?.[componentName];
                case "responses":
                    return openApi3Doc.components?.responses?.[componentName];
                case "parameters":
                    return openApi3Doc.components?.parameters?.[componentName];
                case "examples":
                    return openApi3Doc.components?.examples?.[componentName];
                case "requestBodies":
                    return openApi3Doc.components?.requestBodies?.[componentName];
                case "headers":
                    return openApi3Doc.components?.headers?.[componentName];
                case "securitySchemes":
                    return openApi3Doc.components?.securitySchemes?.[componentName];
                case "links":
                    return openApi3Doc.components?.links?.[componentName];
                case "callbacks":
                    return openApi3Doc.components?.callbacks?.[componentName];
                default:
                    this.logger.warn(`Unsupported component kind in $ref: ${ref}`);
                    return undefined;
            }
        },
        registerMultipartSchema(ref, encoding) {
            multipartSchemas.set(ref, encoding ?? null);
        },
        getMultipartSchemaEncoding(ref) {
            const result = multipartSchemas.get(ref);
            return result === null ? undefined : result;
        },
        isSchemaReferenceRegisteredForMultipartForm(ref) {
            return multipartSchemas.has(ref);
        },
        getPartType(propType, name, isHttpPart, encoding, isEnumType, isUnionType) {
            return schemaExpressionGenerator.getPartType(propType, name, isHttpPart, encoding, isEnumType, isUnionType);
        },
        registerSSEEventSchema(ref) {
            sseEventSchemas.add(ref);
        },
        isSSEEventSchema(ref) {
            return sseEventSchemas.has(ref);
        },
        markSSEUsage() {
            sseUsed = true;
        },
        isSSEUsed() {
            return sseUsed;
        },
    };
    return context;
}
