// @azure-tools/typespec-autorest/dist/src/emit.js
import { createTCGCContext } from "@azure-tools/typespec-client-generator-core";
import { compilerAssert, emitFile, getDirectoryPath, getNamespaceFullName, getService, interpolatePath, listServices, NoTarget, reportDeprecated, resolvePath, } from "@typespec/compiler";
import { unsafe_mutateSubgraphWithNamespace, } from "@typespec/compiler/experimental";
import { resolveInfo } from "@typespec/openapi";
import { getVersioningMutators } from "@typespec/versioning";
import { getTracer, reportDiagnostic } from "./lib.js";
import { createDocumentProxy, getOpenAPIForService, sortOpenAPIDocument, } from "./openapi.js";
const defaultOptions = {
    "output-file": "{emitter-output-dir}/{service-name}/{version-status}/{version}/openapi.json",
    "new-line": "lf",
    "include-x-typespec-name": "never",
    "xml-strategy": "xml-service",
};
export async function $onEmit(context) {
    const tracer = getTracer(context.program);
    const options = resolveAutorestOptions(context.program, context.emitterOutputDir, context.options);
    tracer.trace("options", JSON.stringify(options, null, 2));
    await emitAllServiceAtAllVersions(context.program, options);
}
export function resolveAutorestOptions(program, emitterOutputDir, options) {
    const resolvedOptions = {
        ...defaultOptions,
        ...options,
    };
    const armTypesDir = interpolatePath(resolvedOptions["arm-types-dir"] ?? "{project-root}/../../common-types/resource-management", {
        "project-root": program.projectRoot,
        "emitter-output-dir": emitterOutputDir,
    });
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    if (resolvedOptions["examples-directory"]) {
        reportDeprecated(program, `examples-directory option is deprecated use examples-dir instead or remove it if examples are located in {project-root}/examples`, NoTarget);
    }
    return {
        outputFile: resolvedOptions["output-file"],
        outputDir: emitterOutputDir,
        // eslint-disable-next-line @typescript-eslint/no-deprecated
        azureResourceProviderFolder: resolvedOptions["azure-resource-provider-folder"],
        // eslint-disable-next-line @typescript-eslint/no-deprecated
        examplesDirectory: resolvedOptions["examples-dir"] ?? resolvedOptions["examples-directory"],
        version: resolvedOptions["version"],
        newLine: resolvedOptions["new-line"],
        omitUnreachableTypes: resolvedOptions["omit-unreachable-types"],
        versionEnumStrategy: resolvedOptions["version-enum-strategy"],
        includeXTypeSpecName: resolvedOptions["include-x-typespec-name"],
        armTypesDir,
        useReadOnlyStatusSchema: resolvedOptions["use-read-only-status-schema"],
        emitLroOptions: resolvedOptions["emit-lro-options"],
        emitCommonTypesSchema: resolvedOptions["emit-common-types-schema"],
        xmlStrategy: resolvedOptions["xml-strategy"],
        outputSplitting: resolvedOptions["output-splitting"],
    };
}
function getEmitterContext(program, service, options, multiService = false, version) {
    const tcgcSdkContext = createTCGCContext(program, "@azure-tools/typespec-autorest");
    tcgcSdkContext.enableLegacyHierarchyBuilding = true;
    return {
        program,
        outputFile: resolveOutputFile(program, service, multiService, options, version),
        service: service,
        tcgcSdkContext,
        proxy: createDocumentProxy(program, service, options, version),
        version: version,
        multiService: multiService,
    };
}
export async function getAllServicesAtAllVersions(program, options) {
    const services = listServices(program);
    if (services.length === 0) {
        services.push({ type: program.getGlobalNamespaceType() });
    }
    const serviceRecords = [];
    for (const service of services) {
        const versions = getVersioningMutators(program, service.type);
        if (versions === undefined) {
            const context = getEmitterContext(program, service, options, services.length > 1);
            const results = await getOpenAPIForService(context, options);
            for (const result of results) {
                const newResult = { ...result };
                newResult.outputFile = resolveOutputFile(program, service, services.length > 1, options, undefined, result.feature);
                serviceRecords.push({
                    service,
                    versioned: false,
                    ...newResult,
                });
            }
        }
        else if (versions.kind === "transient") {
            const context = getEmitterContext(program, service, options, services.length > 1);
            const results = await getVersionSnapshotDocument(context, versions.mutator, options, services.length > 1);
            for (const result of results) {
                const newResult = { ...result };
                newResult.outputFile = resolveOutputFile(program, service, services.length > 1, options, undefined, result.feature);
                serviceRecords.push({
                    service,
                    versioned: false,
                    ...newResult,
                });
            }
        }
        else {
            const filteredVersions = versions.snapshots.filter((v) => !options.version || options.version === v.version?.value);
            if (filteredVersions.length === 0 && options.version) {
                reportDiagnostic(program, { code: "no-matching-version-found", target: service.type });
            }
            const serviceRecord = {
                service,
                versioned: true,
                versions: [],
            };
            serviceRecords.push(serviceRecord);
            for (const record of filteredVersions) {
                const context = getEmitterContext(program, service, options, services.length > 1, record.version?.value);
                const results = await getVersionSnapshotDocument(context, record.mutator, options, services.length > 1);
                for (const result of results) {
                    const newResult = { ...result };
                    newResult.outputFile = resolveOutputFile(program, service, services.length > 1, options, record.version.value, result.feature);
                    serviceRecord.versions.push({
                        ...newResult,
                        service,
                        version: record.version.value,
                    });
                }
            }
        }
    }
    return serviceRecords;
}
async function getVersionSnapshotDocument(context, mutator, options, multiService = false) {
    const subgraph = unsafe_mutateSubgraphWithNamespace(context.program, [mutator], context.service.type);
    compilerAssert(subgraph.type.kind === "Namespace", "Should not have mutated to another type");
    const service = getService(context.program, subgraph.type);
    const newContext = getEmitterContext(context.program, service, options, multiService, context.version);
    const document = await getOpenAPIForService(newContext, options);
    return document;
}
async function emitAllServiceAtAllVersions(program, options) {
    const services = await getAllServicesAtAllVersions(program, options);
    if (program.compilerOptions.dryRun || program.hasError()) {
        return;
    }
    for (const serviceRecord of services) {
        if (serviceRecord.versioned) {
            for (const documentRecord of serviceRecord.versions) {
                await emitOutput(program, documentRecord, options);
            }
        }
        else {
            await emitOutput(program, serviceRecord, options);
        }
    }
}
async function emitOutput(program, result, options) {
    const currentFeature = result.feature;
    if (currentFeature !== undefined && result.context.proxy !== undefined)
        result.context.proxy.setCurrentFeature(currentFeature);
    const sortedDocument = sortOpenAPIDocument(result.document);
    // Write out the OpenAPI document to the output path
    await emitFile(program, {
        path: result.outputFile,
        content: prettierOutput(JSON.stringify(sortedDocument, null, 2)),
        newLine: options.newLine,
    });
    // Copy examples to the output directory
    if (result.operationExamples.length > 0) {
        const examplesPath = resolvePath(getDirectoryPath(result.outputFile), "examples");
        await program.host.mkdirp(examplesPath);
        for (const { examples } of result.operationExamples) {
            if (examples) {
                for (const { relativePath, file } of Object.values(examples)) {
                    await emitFile(program, {
                        path: resolvePath(examplesPath, relativePath),
                        content: file.text,
                        newLine: options.newLine,
                    });
                }
            }
        }
    }
}
function prettierOutput(output) {
    return output + "\n";
}
export function resolveOutputFile(program, service, multipleServices, options, version, feature) {
    const azureResourceProviderFolder = options.azureResourceProviderFolder;
    const info = resolveInfo(program, service.type);
    version = version ?? info?.version;
    const interpolated = interpolatePath(options.outputFile, {
        "azure-resource-provider-folder": azureResourceProviderFolder,
        "service-name": multipleServices || azureResourceProviderFolder
            ? getNamespaceFullName(service.type)
            : undefined,
        "version-status": version && (version.includes("preview") ? "preview" : "stable"),
        version,
        feature,
    });
    return resolvePath(options.outputDir, interpolated);
}
