// @azure-tools/typespec-autorest/dist/src/openapi.js
import { FinalStateValue, extractLroStates, getArmResourceIdentifierConfig, getAsEmbeddingVector, getLroMetadata, getUnionAsEnum, hasUniqueItems, } from "@azure-tools/typespec-azure-core";
import { getArmCommonTypeOpenAPIRef, getArmIdentifiers, getArmKeyIdentifiers, getCustomResourceOptions, getExternalTypeRef, getFeature, getInlineAzureType, getResourceFeatureSet, isArmCommonType, isArmExternalType, isArmProviderNamespace, isAzureResource, } from "@azure-tools/typespec-azure-resource-manager";
import { getClientDefaultValue, getClientNameOverride, getLegacyHierarchyBuilding, getMarkAsLro, shouldFlattenProperty, } from "@azure-tools/typespec-client-generator-core";
import { NoTarget, compilerAssert, createDiagnosticCollector, explainStringTemplateNotSerializable, getAllTags, getAnyExtensionFromPath, getDirectoryPath, getDiscriminator, getDoc, getEncode, getFormat, getLifecycleVisibilityEnum, getMaxItems, getMaxLength, getMaxValue, getMinItems, getMinLength, getMinValue, getNamespaceFullName, getPagingOperation, getPattern, getProperty, getRelativePathFromDirectory, getRootLength, getSummary, getVisibilityForClass, ignoreDiagnostics, interpolatePath, isArrayModelType, isDeprecated, isErrorModel, isErrorType, isGlobalNamespace, isList, isNeverType, isNullType, isNumeric, isRecordModelType, isSecret, isService, isTemplateDeclaration, isTemplateDeclarationOrInstance, isVoidType, joinPaths, navigateTypesInNamespace, normalizePath, reportDeprecated, resolveEncodedName, resolvePath, serializeValueAsJson, } from "@typespec/compiler";
import { SyntaxKind } from "@typespec/compiler/ast";
import { $ } from "@typespec/compiler/typekit";
import { TwoLevelMap } from "@typespec/compiler/utils";
import { Visibility, createMetadataInfo, getHeaderFieldOptions, getHttpService, getServers, getStatusCodeDescription, getVisibilitySuffix, isHttpFile, isSharedRoute, reportIfNoRoutes, resolveAuthentication, resolveRequestVisibility, } from "@typespec/http";
import { checkDuplicateTypeName, getExtensions, getExternalDocs, getOpenAPITypeName, getParameterKey, isReadonlyProperty, resolveInfo, } from "@typespec/openapi";
import { getVersionsForEnum } from "@typespec/versioning";
import { AutorestOpenAPISchema } from "./autorest-openapi-schema.js";
import { getExamples, getRef } from "./decorators.js";
import { sortWithJsonSchema } from "./json-schema-sorter/sorter.js";
import { createDiagnostic, reportDiagnostic } from "./lib.js";
import { LateBoundReference, } from "./types.js";
import { getClientName, isSupportedAutorestFormat, resolveOperationId, shouldInline, } from "./utils.js";
import { resolveXmlModule } from "./xml.js";
export async function getOpenAPIForService(context, options) {
    const { program, service } = context;
    const proxy = context.proxy ?? createDefaultDocumentProxy(program, service, options, context.version);
    const typeNameOptions = {
        // shorten type names by removing TypeSpec and service namespace
        namespaceFilter(ns) {
            return !isService(program, ns);
        },
    };
    const httpService = ignoreDiagnostics(getHttpService(program, service.type));
    proxy.addAdditionalInfo(resolveInfo(program, service.type));
    const auth = processAuth(service.type);
    if (auth?.securitySchemes)
        proxy.addSecuritySchemes(auth.securitySchemes);
    if (auth?.security)
        proxy.addSecurityRequirements(auth.security);
    const xml = await resolveXmlModule();
    const xmlStrategy = options.xmlStrategy;
    proxy.addHostInfo(resolveHost(program, service.type));
    let currentEndpoint;
    let currentConsumes;
    let currentProduces;
    const metadataInfo = createMetadataInfo(program, {
        canonicalVisibility: Visibility.Read,
        canShareProperty: canSharePropertyUsingReadonlyOrXMSMutability,
    });
    // Keep a map of all Types+Visibility combinations that were encountered
    // that need schema definitions.
    const pendingSchemas = new TwoLevelMap();
    // Reuse a single ref object per Type+Visibility combination.
    const refs = new TwoLevelMap();
    // Keep track of inline types still in the process of having their schema computed
    // This is used to detect cycles in inline types, which is an
    const inProgressInlineTypes = new Set();
    // Map model properties that represent shared parameters to their parameter
    // definition that will go in #/parameters. Inlined parameters do not go in
    // this map.
    const params = new Map();
    // Keep track of types that were processed indirectly and shouldn't be added if `omit-unreachable-types` is not set.
    // This include:
    // - Models that have had properties spread into parameters.
    // - Multipart models
    const indirectlyProcessedTypes = new Set();
    const operationIdsWithExample = new Set();
    const [exampleMap, diagnostics] = await loadExamples(program, options, context.version);
    program.reportDiagnostics(diagnostics);
    const routes = httpService.operations;
    reportIfNoRoutes(program, routes);
    const xmlEnabled = xmlStrategy !== "none";
    // The set of produces/consumes values found in all operations
    let allResponseContentTypes = routes
        .flatMap((route) => route.responses)
        .flatMap((res) => res.responses)
        .flatMap((res) => res.body?.contentTypes ?? [])
        .filter((ct) => 
    // XML and JSON are privileged to be the only content types that can live at the top level and inform global
    // serializer/naming behavior.
    ct === "application/json" || (xmlEnabled && ct === "application/xml"));
    if (allResponseContentTypes.length === 0)
        allResponseContentTypes = ["application/json"];
    const globalProduces = new Set(allResponseContentTypes);
    let allRequestContentTypes = routes
        .flatMap((route) => route.parameters)
        .flatMap((param) => param.body?.contentTypes ?? [])
        .filter((ct) => !!ct && (ct === "application/json" || (xmlEnabled && ct === "application/xml")));
    if (allRequestContentTypes.length === 0)
        allRequestContentTypes = ["application/json"];
    const globalConsumes = new Set(allRequestContentTypes);
    const shouldEmitXml = xmlEnabled && (globalProduces.has("application/xml") || globalConsumes.has("application/xml"));
    const specIsOnlyXml = xmlEnabled &&
        globalProduces.size === 1 &&
        globalProduces.has("application/xml") &&
        globalConsumes.size === 1 &&
        globalConsumes.has("application/xml");
    routes.forEach(emitOperation);
    emitParameters();
    emitSchemas(service.type);
    proxy.setGlobalConsumes([...globalConsumes]);
    proxy.setGlobalProduces([...globalProduces]);
    proxy.writeExamples(exampleMap, operationIdsWithExample);
    return proxy.resolveDocuments(context);
    function resolveHost(program, namespace) {
        const servers = getServers(program, namespace);
        if (servers === undefined) {
            return {};
        }
        // If there is more than one server we then just make a custom host with a parameter asking for the full url.
        if (servers.length > 1) {
            return {
                "x-ms-parameterized-host": {
                    hostTemplate: "{url}",
                    useSchemePrefix: false,
                    parameters: [
                        {
                            name: "url",
                            in: "path",
                            description: "Url",
                            type: "string",
                            format: "uri",
                            "x-ms-skip-url-encoding": true,
                        },
                    ],
                },
            };
        }
        const server = servers[0];
        if (server.parameters.size === 0) {
            const [scheme, host] = server.url.split("://");
            return {
                host,
                schemes: [scheme],
            };
        }
        const parameters = [];
        for (const prop of server.parameters.values()) {
            const param = getOpenAPI2Parameter({
                kind: "path",
                path: [],
                property: prop,
                options: {
                    allowReserved: false,
                    explode: false,
                    style: "simple",
                    name: prop.name,
                },
            }, {
                visibility: Visibility.Read,
                ignoreMetadataAnnotations: false,
            });
            if (prop.type.kind === "Scalar" &&
                $(program).type.isAssignableTo(prop.type, $(program).builtin.url, prop.type)) {
                param["x-ms-skip-url-encoding"] = true;
            }
            parameters.push(param);
        }
        return {
            "x-ms-parameterized-host": {
                hostTemplate: server.url,
                useSchemePrefix: false,
                parameters,
            },
        };
    }
    function resolveXmsPageable(program, operation) {
        if (isList(program, operation.operation)) {
            const pagedInfo = ignoreDiagnostics(getPagingOperation(program, operation.operation));
            return pagedInfo && getXmsPageableForPagingOperation(pagedInfo);
        }
        return undefined;
    }
    function getXmsPageableForPagingOperation(paging) {
        if (paging.output.nextLink) {
            const itemsName = paging.output.pageItems.property.name;
            return {
                nextLinkName: paging.output.nextLink.property.name,
                itemName: itemsName === "value" ? undefined : itemsName,
            };
        }
        return undefined;
    }
    function getFinalStateVia(metadata) {
        switch (metadata.finalStateVia) {
            case FinalStateValue.azureAsyncOperation:
                return "azure-async-operation";
            case FinalStateValue.location:
                return "location";
            case FinalStateValue.operationLocation:
                return "operation-location";
            case FinalStateValue.originalUri:
                return "original-uri";
            default:
                return undefined;
        }
    }
    function getFinalStateSchema(metadata) {
        if (metadata.finalResult !== undefined &&
            metadata.finalResult !== "void" &&
            metadata.finalResult.name.length > 0) {
            const model = metadata.finalResult;
            const schemaOrRef = resolveExternalRef(metadata.finalResult);
            let overrideName = undefined;
            if (model.kind === "Model" && isArrayModelType(model)) {
                const itemType = getModelOrScalarTypeIfNullable(model.indexer?.value);
                if (itemType?.kind === "Model" && itemType.name.length > 0) {
                    overrideName = (n, v) => `${n.replaceAll(/[[\]]/g, "")}${getVisibilitySuffix(v, Visibility.Read)}Array`;
                }
            }
            if (schemaOrRef !== undefined) {
                const ref = proxy.createExternalRef(schemaOrRef.$ref);
                return { "final-state-schema": ref };
            }
            const pending = pendingSchemas.getOrAdd(metadata.finalResult, Visibility.Read, () => ({
                type: model,
                visibility: Visibility.Read,
                getSchemaNameOverride: overrideName,
                ref: refs.getOrAdd(model, Visibility.Read, () => proxy.createLocalRef(model)),
            }));
            return { "final-state-schema": pending.ref };
        }
        return undefined;
    }
    function emitOperation(operation) {
        const { operation: op, verb, parameters } = operation;
        currentEndpoint = proxy.createOrGetEndpoint(operation, context);
        currentConsumes = new Set();
        currentProduces = new Set();
        const currentTags = getAllTags(program, op);
        if (currentTags) {
            currentEndpoint.tags = [...currentTags.values()];
            for (const tag of currentTags) {
                // Add to root tags if not already there
                proxy.addTag(tag, op);
            }
        }
        applyExternalDocs(op, currentEndpoint);
        // Set up basic endpoint fields
        currentEndpoint.summary = getSummary(program, op);
        currentEndpoint.description = getDoc(program, op);
        currentEndpoint.parameters = [];
        currentEndpoint.responses = {};
        const lroMetadata = getLroMetadata(program, op);
        // We ignore GET operations because they cannot be LROs per our guidelines and this
        // ensures we don't add the x-ms-long-running-operation extension to the polling operation,
        // which does have LRO metadata.
        if (lroMetadata !== undefined && operation.verb !== "get") {
            currentEndpoint["x-ms-long-running-operation"] = true;
            if (options.emitLroOptions !== "none") {
                const finalState = getFinalStateVia(lroMetadata);
                if (finalState !== undefined) {
                    const finalSchema = getFinalStateSchema(lroMetadata);
                    let lroOptions = {
                        "final-state-via": finalState,
                    };
                    if (finalSchema !== undefined && options.emitLroOptions === "all") {
                        lroOptions = {
                            "final-state-via": finalState,
                            ...finalSchema,
                        };
                    }
                    currentEndpoint["x-ms-long-running-operation-options"] = lroOptions;
                }
            }
        }
        if (getMarkAsLro(context.tcgcSdkContext, op) === true) {
            currentEndpoint["x-ms-long-running-operation"] = true;
        }
        // Extract paged metadata from Azure.Core.Page
        const pageable = resolveXmsPageable(program, operation);
        if (pageable) {
            currentEndpoint["x-ms-pageable"] = pageable;
        }
        const visibility = resolveRequestVisibility(program, operation.operation, verb);
        emitEndpointParameters(parameters, visibility);
        emitResponses(operation.responses);
        applyEndpointConsumes();
        applyEndpointProduces();
        if (isDeprecated(program, op)) {
            currentEndpoint.deprecated = true;
        }
        const examples = getExamples(program, op);
        if (examples) {
            currentEndpoint["x-ms-examples"] = examples.reduce((acc, example) => ({ ...acc, [example.title]: { $ref: example.pathOrUri } }), {});
        }
        const autoExamples = exampleMap.get(currentEndpoint.operationId);
        if (autoExamples && currentEndpoint.operationId) {
            operationIdsWithExample.add(currentEndpoint.operationId);
            currentEndpoint["x-ms-examples"] = currentEndpoint["x-ms-examples"] || {};
            for (const [title, example] of Object.entries(autoExamples)) {
                currentEndpoint["x-ms-examples"][title] = { $ref: `./examples/${example.relativePath}` };
            }
        }
        // Attach additional extensions after main fields
        attachExtensions(op, currentEndpoint);
    }
    function applyEndpointProduces() {
        if (currentProduces.size > 0 && !checkLocalAndGlobalEqual(globalProduces, currentProduces)) {
            currentEndpoint.produces = [...currentProduces];
        }
    }
    function applyEndpointConsumes() {
        if (currentConsumes.size > 0 && !checkLocalAndGlobalEqual(globalConsumes, currentConsumes)) {
            currentEndpoint.consumes = [...currentConsumes];
        }
    }
    function checkLocalAndGlobalEqual(global, local) {
        if (global.size !== local.size) {
            return false;
        }
        for (const entry of local) {
            if (!global.has(entry)) {
                return false;
            }
        }
        return true;
    }
    function isBytes(type) {
        return $(program).type.isAssignableTo(type, $(program).builtin.bytes, type);
    }
    function isBinaryPayload(body, contentType) {
        const types = new Set(typeof contentType === "string" ? [contentType] : contentType);
        return (body.kind === "Scalar" &&
            body.name === "bytes" &&
            !types.has("application/json") &&
            !types.has("text/plain"));
    }
    function emitResponses(responses) {
        for (const response of responses) {
            for (const statusCode of getOpenAPI2StatusCodes(response.statusCodes, response.type)) {
                emitResponseObject(statusCode, response);
            }
        }
    }
    function getOpenAPI2StatusCodes(statusCodes, diagnosticTarget) {
        if (statusCodes === "*") {
            return ["default"];
        }
        else if (typeof statusCodes === "number") {
            return [String(statusCodes)];
        }
        else {
            return rangeToOpenAPI(statusCodes, diagnosticTarget);
        }
    }
    function rangeToOpenAPI(range, diagnosticTarget) {
        const reportInvalid = () => reportDiagnostic(program, {
            code: "unsupported-status-code-range",
            format: { start: String(range.start), end: String(range.end) },
            target: diagnosticTarget,
        });
        const codes = [];
        let start = range.start;
        let end = range.end;
        if (range.start < 100) {
            reportInvalid();
            start = 100;
            codes.push("default");
        }
        else if (range.end > 599) {
            reportInvalid();
            codes.push("default");
            end = 599;
        }
        const groups = [1, 2, 3, 4, 5];
        for (const group of groups) {
            if (start > end) {
                break;
            }
            const groupStart = group * 100;
            const groupEnd = groupStart + 99;
            if (start >= groupStart && start <= groupEnd) {
                codes.push(`${group}XX`);
                if (start !== groupStart || end < groupEnd) {
                    reportInvalid();
                }
                start = groupStart + 100;
            }
        }
        return codes;
    }
    function getResponseDescriptionForStatusCode(statusCode) {
        if (statusCode === "default") {
            return "An unexpected error response.";
        }
        return getStatusCodeDescription(statusCode) ?? "unknown";
    }
    function emitResponseObject(statusCode, response) {
        const openapiResponse = currentEndpoint.responses[statusCode] ?? {
            description: response.description ?? getResponseDescriptionForStatusCode(statusCode),
        };
        if (isErrorModel(program, response.type) && statusCode !== "default") {
            openapiResponse["x-ms-error-response"] = true;
        }
        const contentTypes = [];
        let body;
        for (const data of response.responses) {
            if (data.headers && Object.keys(data.headers).length > 0) {
                openapiResponse.headers ??= {};
                for (const [key, value] of Object.entries(data.headers)) {
                    openapiResponse.headers[key] = getResponseHeader(value);
                }
            }
            if (data.body) {
                if (body && body.type !== data.body.type) {
                    reportDiagnostic(program, {
                        code: "duplicate-body-types",
                        target: response.type,
                    });
                }
                body = data.body;
                contentTypes.push(...data.body.contentTypes);
            }
        }
        if (body) {
            openapiResponse.schema = getSchemaForResponseBody(body, contentTypes);
        }
        for (const contentType of contentTypes) {
            currentProduces.add(contentType);
        }
        currentEndpoint.responses[statusCode] = openapiResponse;
    }
    function getSchemaForResponseBody(body, contentTypes) {
        const isBinary = contentTypes.every((t) => isBinaryPayload(body.type, t));
        if (body.bodyKind === "file" || isBinary) {
            return { type: "file" };
        }
        if (body.bodyKind === "multipart") {
            // OpenAPI2 doesn't support multipart responses, so we just return a string schema
            return { type: "string" };
        }
        return getSchemaOrRef(body.type, {
            visibility: Visibility.Read,
            ignoreMetadataAnnotations: body.isExplicit && body.containsMetadataAnnotations,
        });
    }
    function getResponseHeader(prop) {
        const header = getOpenAPI2HeaderParameter(prop, {
            visibility: Visibility.Read,
            ignoreMetadataAnnotations: false,
        });
        Object.assign(header, applyIntrinsicDecorators(prop, {
            type: header.type,
            format: header.format,
        }));
        delete header.in;
        delete header.name;
        delete header["x-ms-client-name"];
        delete header.required;
        return header;
    }
    function expandRef(ref) {
        const absoluteRef = interpolatePath(ref, {
            "arm-types-dir": options.armTypesDir,
        });
        if (getRootLength(absoluteRef) === 0) {
            return absoluteRef; // It is already relative.
        }
        return getRelativePathFromDirectory(getDirectoryPath(context.outputFile), absoluteRef, false);
    }
    function resolveExternalRef(type) {
        const refUrl = getRef(program, type);
        if (refUrl) {
            return {
                $ref: expandRef(refUrl),
            };
        }
        const externalTypeRefUrl = getExternalTypeRef(program, type);
        if (externalTypeRefUrl) {
            return {
                $ref: expandRef(externalTypeRefUrl),
            };
        }
        if (isArmCommonType(type) &&
            (type.kind === "Model" ||
                type.kind === "ModelProperty" ||
                type.kind === "Enum" ||
                type.kind === "Union")) {
            const ref = getArmCommonTypeOpenAPIRef(program, type, {
                version: context.version,
                service: context.service,
            });
            if (ref) {
                return {
                    $ref: expandRef(ref),
                };
            }
        }
        return undefined;
    }
    function shouldInlineCoreScalarProperty(type) {
        if (type.kind !== "ModelProperty" ||
            type.type.kind !== "Scalar" ||
            type.type.namespace === undefined)
            return false;
        const nsName = getNamespaceFullName(type.type.namespace);
        return nsName === "Azure.Core" && getInlineAzureType(program, type) === true;
    }
    function getSchemaOrRef(type, schemaContext, namespace) {
        let schemaNameOverride = undefined;
        const ref = resolveExternalRef(type);
        if (ref) {
            if (options.emitCommonTypesSchema === "never" ||
                !metadataInfo.isTransformed(type, schemaContext.visibility)) {
                return ref;
            }
            // Reference schemas will only be generated when they differ from READ
            schemaNameOverride = (n, v) => `${n}${getVisibilitySuffix(v, Visibility.Read)}`;
        }
        if (type.kind === "Scalar" && program.checker.isStdType(type)) {
            return getSchemaForScalar(type);
        }
        if (type.kind === "String" || type.kind === "Number" || type.kind === "Boolean") {
            // For literal types, we just want to emit them directly as well.
            return getSchemaForLiterals(type);
        }
        if (type.kind === "StringTemplate") {
            return getSchemaForStringTemplate(type);
        }
        if (type.kind === "Intrinsic" && type.name === "unknown") {
            return getSchemaForIntrinsicType(type);
        }
        if (type.kind === "ModelProperty") {
            return resolveProperty(type, schemaContext);
        }
        type = metadataInfo.getEffectivePayloadType(type, schemaContext.visibility);
        const name = getOpenAPITypeName(program, type, typeNameOptions);
        if (shouldInline(program, type)) {
            const schema = getSchemaForInlineType(type, name, schemaContext, namespace);
            if (schema === undefined && isErrorType(type)) {
                // Exit early so that syntax errors are exposed.  This error will
                // be caught and handled in emitOpenAPI.
                throw new ErrorTypeFoundError();
            }
            // helps to read output and correlate to TypeSpec
            if (schema && options.includeXTypeSpecName !== "never") {
                schema["x-typespec-name"] = name;
            }
            return schema;
        }
        else {
            // Use shared schema when type is not transformed by visibility from the canonical read visibility.
            if (!metadataInfo.isTransformed(type, schemaContext.visibility)) {
                schemaContext = { ...schemaContext, visibility: Visibility.Read };
            }
            const pending = pendingSchemas.getOrAdd(type, schemaContext.visibility, () => ({
                type,
                visibility: schemaContext.visibility,
                ref: refs.getOrAdd(type, schemaContext.visibility, () => proxy.createLocalRef(type)),
                getSchemaNameOverride: schemaNameOverride,
            }));
            return { $ref: pending.ref };
        }
    }
    function getSchemaForInlineType(type, name, context, namespace) {
        if (inProgressInlineTypes.has(type)) {
            reportDiagnostic(program, {
                code: "inline-cycle",
                format: { type: name },
                target: type,
            });
            return {};
        }
        inProgressInlineTypes.add(type);
        const schema = getSchemaForType(type, context, namespace);
        inProgressInlineTypes.delete(type);
        return schema;
    }
    function getParamPlaceholder(property) {
        let spreadParam = false;
        if (property.sourceProperty) {
            // chase our sources all the way back to the first place this property
            // was defined.
            spreadParam = true;
            property = property.sourceProperty;
            while (property.sourceProperty) {
                property = property.sourceProperty;
            }
        }
        const ref = resolveExternalRef(property);
        if (ref) {
            return ref;
        }
        const parameter = params.get(property);
        if (parameter) {
            return parameter;
        }
        const placeholder = {};
        // only parameters inherited by spreading from non-inlined type are shared in #/parameters
        if (spreadParam && property.model && !shouldInline(program, property.model)) {
            params.set(property, placeholder);
            indirectlyProcessedTypes.add(property.model);
        }
        return placeholder;
    }
    function emitEndpointParameters(methodParams, visibility) {
        const consumes = methodParams.body?.contentTypes ?? [];
        for (const httpProperty of methodParams.properties) {
            const shared = params.get(httpProperty.property);
            if (shared) {
                currentEndpoint.parameters.push(shared);
                continue;
            }
            if (!isHttpParameterProperty(httpProperty)) {
                continue;
            }
            if (httpProperty.kind === "cookie") {
                reportDiagnostic(program, { code: "cookies-unsupported", target: httpProperty.property });
                continue;
            }
            emitParameter(httpProperty.property, () => getOpenAPI2Parameter(httpProperty, { visibility, ignoreMetadataAnnotations: false }));
        }
        if (consumes.length === 0 && methodParams.body) {
            // we didn't find an explicit content type anywhere, so infer from body.
            if (getModelOrScalarTypeIfNullable(methodParams.body.type)) {
                consumes.push("application/json");
            }
        }
        for (const consume of consumes) {
            currentConsumes.add(consume);
        }
        if (methodParams.body && !isVoidType(methodParams.body.type)) {
            emitBodyParameters(methodParams.body, visibility);
        }
    }
    function emitBodyParameters(body, visibility) {
        switch (body.bodyKind) {
            case "single":
                emitSingleBodyParameters(body, visibility);
                break;
            case "multipart":
                emitMultipartBodyParameters(body, visibility);
                break;
            case "file":
                const bodySchema = { type: "string", format: "binary" };
                const { property } = body;
                if (property) {
                    emitParameter(property, () => getOpenAPI2BodyParameter(property, property.name, bodySchema));
                }
                else {
                    currentEndpoint.parameters.push({
                        name: "body",
                        in: "body",
                        schema: {
                            type: "string",
                            format: "binary",
                        },
                        required: true,
                    });
                }
                break;
        }
    }
    function emitSingleBodyParameters(body, visibility) {
        const isBinary = isBinaryPayload(body.type, body.contentTypes);
        const schemaContext = {
            visibility,
            ignoreMetadataAnnotations: body.isExplicit && body.containsMetadataAnnotations,
        };
        const schema = isBinary
            ? { type: "string", format: "binary" }
            : getSchemaOrRef(body.type, schemaContext);
        if (body.property) {
            const prop = body.property;
            emitParameter(prop, () => getOpenAPI2BodyParameter(prop, resolveEncodedName(program, prop, "application/json"), schema));
        }
        else {
            currentEndpoint.parameters.push({
                name: "body",
                in: "body",
                schema,
                required: true,
            });
        }
    }
    function emitMultipartBodyParameters(body, visibility) {
        indirectlyProcessedTypes.add(body.type);
        for (const [index, part] of body.parts.entries()) {
            const partName = part.name ?? `part${index}`;
            let schema = getFormDataSchema(part.body.type, { visibility, ignoreMetadataAnnotations: false }, partName, part.body.type);
            if (schema) {
                if (part.multi) {
                    schema = {
                        type: "array",
                        items: schema.type === "file" ? { type: "string", format: "binary" } : schema,
                    };
                }
                const param = {
                    name: partName,
                    in: "formData",
                    required: !part.optional,
                    ...schema,
                };
                if (part.property) {
                    param.description = getDoc(program, part.property);
                    if (part.property.name !== partName) {
                        param["x-ms-client-name"] = part.property.name;
                    }
                }
                currentEndpoint.parameters.push(param);
            }
        }
    }
    function getModelOrScalarTypeIfNullable(type) {
        if (type.kind === "Model" || type.kind === "Scalar") {
            return type;
        }
        else if (type.kind === "Union") {
            // Remove all `null` types and make sure there's a single model type
            const nonNulls = [...type.variants.values()]
                .map((x) => x.type)
                .filter((variant) => !isNullType(variant));
            if (nonNulls.every((t) => t.kind === "Model" || t.kind === "Scalar")) {
                return nonNulls.length === 1 ? nonNulls[0] : undefined;
            }
        }
        return undefined;
    }
    function emitParameter(prop, resolve) {
        if (isNeverType(prop.type)) {
            return;
        }
        const ph = getParamPlaceholder(prop);
        currentEndpoint.parameters.push(ph);
        // If the parameter already has a $ref, don't bother populating it
        if (!("$ref" in ph)) {
            Object.assign(ph, resolve());
        }
    }
    function getSchemaForPrimitiveItems(type, schemaContext, paramName, target, multipart) {
        const fullSchema = getSchemaForType(type, schemaContext);
        if (fullSchema === undefined) {
            return undefined;
        }
        if (fullSchema.type === "object") {
            reportDiagnostic(program, {
                code: multipart ? "unsupported-multipart-type" : "unsupported-param-type",
                format: { part: paramName },
                target,
            });
            return { type: "string" };
        }
        return fullSchema;
    }
    function getFormDataSchema(type, schemaContext, paramName, target) {
        if (isBytes(type) || isHttpFile(program, type)) {
            return { type: "file" };
        }
        if (type.kind === "Model" && isArrayModelType(type)) {
            const elementType = type.indexer.value;
            if (isBytes(elementType)) {
                return { type: "array", items: { type: "string", format: "binary" } };
            }
            const schema = getSchemaForPrimitiveItems(elementType, schemaContext, paramName, target, true);
            if (schema === undefined) {
                return undefined;
            }
            delete schema.description;
            return {
                type: "array",
                items: schema,
            };
        }
        else {
            const schema = getSchemaForPrimitiveItems(type, schemaContext, paramName, target, true);
            if (schema === undefined) {
                return undefined;
            }
            return schema;
        }
    }
    function getOpenAPI2ParameterBase(param, name) {
        const base = {
            name: name ?? param.name,
            required: !param.optional,
            description: getDoc(program, param),
        };
        const clientName = getClientName(context, param);
        if (name !== clientName) {
            base["x-ms-client-name"] = clientName;
        }
        attachExtensions(param, base);
        return base;
    }
    function getOpenAPI2BodyParameter(param, name, bodySchema) {
        const result = {
            in: "body",
            ...getOpenAPI2ParameterBase(param, name),
            schema: bodySchema,
        };
        if (shouldFlattenProperty(context.tcgcSdkContext, param) === true) {
            result["x-ms-client-flatten"] = true;
        }
        const jsonName = resolveEncodedName(program, param, "application/json");
        if (jsonName !== param.name) {
            // Special case to be able to keep pre-existing cases where you have both the body parameter name and x-ms-client-name
            reportDeprecated(program, "Using encodedName for the body property is meaningless. That property is not serialized as Json. If wanting to rename it use @Azure.ClientGenerator.Core.clientName", param.decorators.find((x) => x.definition?.name === "@encodedName")?.node ?? param);
            result.name = jsonName;
            if (!result["x-ms-client-name"]) {
                result["x-ms-client-name"] = param.name;
            }
        }
        else {
            // For body parameter the only value of the name is in the client so no need to keep the original one
            if (result["x-ms-client-name"]) {
                result.name = result["x-ms-client-name"];
                delete result["x-ms-client-name"];
            }
        }
        return result;
    }
    function getSimpleParameterSchema(param, schemaContext, name) {
        if (param.type.kind === "Model" && isArrayModelType(param.type)) {
            const itemSchema = getSchemaForPrimitiveItems(param.type.indexer.value, schemaContext, name, param);
            const schema = itemSchema && {
                ...itemSchema,
            };
            delete schema.description;
            return { type: "array", items: schema };
        }
        else {
            return getSchemaForPrimitiveItems(param.type, schemaContext, name, param);
        }
    }
    function isEncodingHandledByParameter(encoding) {
        return (encoding === "ArrayEncoding.commaDelimited" ||
            encoding === "ArrayEncoding.pipeDelimited" ||
            encoding === "ArrayEncoding.spaceDelimited");
    }
    function getCollectionFormat(type, explode) {
        if ($(program).array.is(type.type)) {
            if (explode) {
                return "multi";
            }
            const encode = getEncode(context.program, type);
            if (encode) {
                if (encode?.encoding === "ArrayEncoding.commaDelimited") {
                    return "csv";
                }
                if (encode?.encoding === "ArrayEncoding.pipeDelimited") {
                    return "pipes";
                }
                if (encode?.encoding === "ArrayEncoding.spaceDelimited") {
                    return "ssv";
                }
                reportDiagnostic(program, { code: "invalid-multi-collection-format", target: type });
            }
            return "csv";
        }
        return undefined;
    }
    function getOpenAPI2QueryParameter(httpProp, schemaContext) {
        const property = httpProp.property;
        const base = getOpenAPI2ParameterBase(property, httpProp.options.name);
        const collectionFormat = getCollectionFormat(httpProp.property, httpProp.options.explode);
        const schema = getSimpleParameterSchema(property, schemaContext, base.name);
        return {
            in: "query",
            collectionFormat,
            default: property.defaultValue && getDefaultValue(property.defaultValue, property),
            ...base,
            ...schema,
        };
    }
    function getOpenAPI2PathParameter(httpProp, schemaContext) {
        const property = httpProp.property;
        const base = getOpenAPI2ParameterBase(property, httpProp.options.name);
        if (base.required === false) {
            reportDiagnostic(program, {
                code: "unsupported-optional-path-param",
                format: { name: property.name },
                target: property,
            });
        }
        const result = {
            in: "path",
            default: property.defaultValue && getDefaultValue(property.defaultValue, property),
            ...base,
            ...getSimpleParameterSchema(property, schemaContext, base.name),
            required: true,
        };
        if (httpProp.options.allowReserved) {
            result["x-ms-skip-url-encoding"] = true;
        }
        return result;
    }
    function getOpenAPI2HeaderParameter(prop, schemaContext, name) {
        const base = getOpenAPI2ParameterBase(prop, name);
        const headerOptions = getHeaderFieldOptions(program, prop);
        const collectionFormat = getCollectionFormat(prop, headerOptions.explode);
        return {
            in: "header",
            default: prop.defaultValue && getDefaultValue(prop.defaultValue, prop),
            ...base,
            collectionFormat: collectionFormat,
            ...getSimpleParameterSchema(prop, schemaContext, base.name),
        };
    }
    function getOpenAPI2ParameterInternal(httpProperty, schemaContext) {
        switch (httpProperty.kind) {
            case "query":
                return getOpenAPI2QueryParameter(httpProperty, schemaContext);
            case "path":
                return getOpenAPI2PathParameter(httpProperty, schemaContext);
            case "header":
                return getOpenAPI2HeaderParameter(httpProperty.property, schemaContext, httpProperty.options.name);
            case "cookie":
                compilerAssert(false, "Should verify cookies before");
                break;
            default:
                const _assertNever = httpProperty;
                compilerAssert(false, "Unreachable");
        }
    }
    function getOpenAPI2Parameter(httpProp, schemaContext) {
        const value = getOpenAPI2ParameterInternal(httpProp, schemaContext);
        // Apply decorators to a copy of the parameter definition.  We use
        // Object.assign here because applyIntrinsicDecorators returns a new object
        // based on the target object and we need to apply its changes back to the
        // original parameter.
        Object.assign(value, applyIntrinsicDecorators(httpProp.property, {
            type: value.type,
            format: value.format,
        }, "parameter"));
        return value;
    }
    function emitParameters() {
        for (const [property, param] of params) {
            // Add an extension which tells AutoRest that this is a shared operation
            // parameter definition
            if (param["x-ms-parameter-location"] === undefined) {
                param["x-ms-parameter-location"] = "method";
            }
            const key = getParameterKey(program, property, param, proxy.getParameterMap(), typeNameOptions);
            proxy.writeParameter(key, property, param);
            const refedParam = param;
            for (const key of Object.keys(param)) {
                delete refedParam[key];
            }
            refedParam["$ref"] = "#/parameters/" + encodeURIComponent(key);
        }
    }
    function emitSchemas(serviceNamespace) {
        const processedSchemas = new TwoLevelMap();
        processSchemas();
        if (!options.omitUnreachableTypes) {
            processUnreferencedSchemas();
        }
        // Emit the processed schemas. Only now can we compute the names as it
        // depends on whether we have produced multiple schemas for a single
        // TYPESPEC type.
        for (const group of processedSchemas.values()) {
            for (const [visibility, processed] of group) {
                let name = getClientNameOverride(context.tcgcSdkContext, processed.type);
                if (name === undefined) {
                    name = getOpenAPITypeName(program, processed.type, typeNameOptions);
                }
                if (processed.getSchemaNameOverride !== undefined) {
                    name = processed.getSchemaNameOverride(name, visibility);
                }
                else if (group.size > 1) {
                    name += getVisibilitySuffix(visibility, Visibility.Read);
                }
                checkDuplicateTypeName(program, processed.type, name, proxy.getDefinitionMap());
                processed.ref.setLocalValue(program, encodeURIComponent(name), processed.type);
                if (processed.schema) {
                    if (shouldEmitXml)
                        attachXml(processed.type, name, processed);
                    proxy.writeDefinition(name, processed, processed.schema);
                }
            }
        }
        function processSchemas() {
            // Process pending schemas. Note that getSchemaForType may pull in new
            // pending schemas so we iterate until there are no pending schemas
            // remaining.
            while (pendingSchemas.size > 0) {
                for (const [type, group] of pendingSchemas) {
                    for (const [visibility, pending] of group) {
                        processedSchemas.getOrAdd(type, visibility, () => ({
                            ...pending,
                            schema: getSchemaForType(type, {
                                visibility: visibility,
                                ignoreMetadataAnnotations: false,
                            }),
                        }));
                    }
                    pendingSchemas.delete(type);
                }
            }
        }
        function processUnreferencedSchemas() {
            const addSchema = (type) => {
                if (!processedSchemas.has(type) &&
                    !indirectlyProcessedTypes.has(type) &&
                    !shouldInline(program, type) &&
                    !shouldOmitThisUnreachableType(type)) {
                    getSchemaOrRef(type, { visibility: Visibility.Read, ignoreMetadataAnnotations: false });
                }
            };
            const skipSubNamespaces = isGlobalNamespace(program, serviceNamespace);
            navigateTypesInNamespace(serviceNamespace, {
                model: addSchema,
                scalar: addSchema,
                enum: addSchema,
                union: addSchema,
            }, { skipSubNamespaces });
            processSchemas();
        }
        function shouldOmitThisUnreachableType(type) {
            if (options.versionEnumStrategy !== "include" &&
                type.kind === "Enum" &&
                isVersionEnum(program, type)) {
                return true;
            }
            return false;
        }
    }
    function isVersionEnum(program, enumObj) {
        const [_, map] = getVersionsForEnum(program, enumObj);
        if (map !== undefined && map.getVersions()[0].enumMember.enum === enumObj) {
            return true;
        }
        return false;
    }
    function getSchemaForType(type, schemaContext, namespace) {
        const builtinType = getSchemaForLiterals(type);
        if (builtinType !== undefined) {
            return builtinType;
        }
        switch (type.kind) {
            case "Intrinsic":
                return getSchemaForIntrinsicType(type);
            case "Model":
                return getSchemaForModel(type, schemaContext, namespace);
            case "ModelProperty":
                return getSchemaForType(type.type, schemaContext);
            case "Scalar":
                return getSchemaForScalar(type);
            case "Union":
                return getSchemaForUnion(type, schemaContext);
            case "EnumMember":
                return getSchemaForEnumMember(type);
            case "UnionVariant":
                return getSchemaForUnionVariant(type, schemaContext);
            case "Enum":
                return getSchemaForEnum(type);
            case "Tuple":
                return { type: "array", items: {} };
        }
        reportDiagnostic(program, {
            code: "invalid-schema",
            format: { type: type.kind },
            target: type,
        });
        return undefined;
    }
    function getSchemaForIntrinsicType(type) {
        switch (type.name) {
            case "unknown":
                return {};
        }
        reportDiagnostic(program, {
            code: "invalid-schema",
            format: { type: type.name },
            target: type,
        });
        return {};
    }
    /**
     * Version enum is special so we we just render the current version with modelAsString: true
     */
    function getSchemaForVersionEnum(e, currentVersion) {
        const member = [...e.members.values()].find((x) => (x.value ?? x.name) === currentVersion);
        compilerAssert(member, `Version enum ${e.name} does not have a member for ${currentVersion}.`, e);
        return {
            type: "string",
            description: getDoc(program, e),
            enum: [member.value ?? member.name],
            "x-ms-enum": {
                name: e.name,
                modelAsString: true,
                values: [
                    {
                        name: member.name,
                        value: member.value ?? member.name,
                        description: getDoc(program, member),
                    },
                ],
            },
        };
    }
    function getSchemaForEnum(e) {
        const values = [];
        if (e.members.size === 0) {
            reportUnsupportedUnion("empty");
            return {};
        }
        const type = getEnumMemberType(e.members.values().next().value);
        for (const option of e.members.values()) {
            if (type !== getEnumMemberType(option)) {
                reportUnsupportedUnion();
                continue;
            }
            else {
                values.push(option.value ?? option.name);
            }
        }
        // If we are rendering a specific version and trying to render the version enum we should treat it specially to only include the current version.
        if (isVersionEnum(program, e) && context.version) {
            return getSchemaForVersionEnum(e, context.version);
        }
        const schema = { type, description: getDoc(program, e) };
        if (values.length > 0) {
            schema.enum = values;
            addXMSEnum(e, schema);
        }
        if (options.useReadOnlyStatusSchema) {
            const [values, _] = extractLroStates(program, e);
            if (values !== undefined) {
                schema.readOnly = true;
            }
        }
        return schema;
        function getEnumMemberType(member) {
            if (typeof member.value === "number") {
                return "number";
            }
            return "string";
        }
        function reportUnsupportedUnion(messageId = "default") {
            reportDiagnostic(program, { code: "union-unsupported", messageId, target: e });
        }
    }
    function getSchemaForUnionEnum(union, e) {
        const values = [];
        let foundCustom = false;
        for (const [name, member] of e.flattenedMembers.entries()) {
            const description = getDoc(program, member.type);
            const memberClientName = getClientNameOverride(context.tcgcSdkContext, member.type);
            values.push({
                name: memberClientName ?? (typeof name === "string" ? name : `${member.value}`),
                value: member.value,
                description,
            });
            if (description || typeof name === "string") {
                foundCustom = true;
            }
        }
        const clientName = getClientName(context, union);
        const schema = {
            type: e.kind,
            enum: [...e.flattenedMembers.values()].map((x) => x.value),
            "x-ms-enum": {
                name: clientName ?? union.name,
                modelAsString: e.open,
            },
        };
        if (foundCustom) {
            schema["x-ms-enum"].values = values;
        }
        if (e.nullable) {
            schema["x-nullable"] = true;
        }
        if (options.useReadOnlyStatusSchema) {
            const [values, _] = extractLroStates(program, union);
            if (values !== undefined) {
                schema.readOnly = true;
            }
        }
        return applyIntrinsicDecorators(union, schema);
    }
    function getSchemaForUnion(union, schemaContext) {
        const nonNullOptions = [...union.variants.values()]
            .map((x) => x.type)
            .filter((t) => !isNullType(t));
        const nullable = union.variants.size !== nonNullOptions.length;
        if (nonNullOptions.length === 0) {
            reportDiagnostic(program, { code: "union-null", target: union });
            return {};
        }
        if (nonNullOptions.length === 1) {
            const type = nonNullOptions[0];
            // Get the schema for the model type
            const schema = getSchemaOrRef(type, schemaContext);
            if (schema.$ref) {
                return { ...schema, "x-nullable": nullable };
            }
            else {
                schema["x-nullable"] = nullable;
                return schema;
            }
        }
        else {
            const [asEnum, _] = getUnionAsEnum(union);
            if (asEnum) {
                return getSchemaForUnionEnum(union, asEnum);
            }
            reportDiagnostic(program, {
                code: "union-unsupported",
                target: union,
            });
            return {};
        }
    }
    function ifArmIdentifiersDefault(armIdentifiers) {
        return armIdentifiers.every((identifier) => identifier === "id" || identifier === "name");
    }
    function getSchemaForEnumMember(member) {
        const value = member.value ?? member.name;
        const type = typeof value === "number" ? "number" : "string";
        return {
            type,
            enum: [value],
            description: getDoc(program, member),
        };
    }
    function getSchemaForUnionVariant(variant, schemaContext) {
        return getSchemaForType(variant.type, schemaContext);
    }
    function getDefaultValue(defaultType, modelProperty) {
        return serializeValueAsJson(program, defaultType, modelProperty);
    }
    function includeDerivedModel(model) {
        return (!resolveExternalRef(model) &&
            !isTemplateDeclaration(model) &&
            (model.templateMapper?.args === undefined ||
                model.templateMapper?.args.length === 0 ||
                model.derivedModels.length > 0));
    }
    function getDiscriminatorValue(model) {
        let discriminator;
        let current = model;
        while (current.baseModel) {
            discriminator = getDiscriminator(program, current.baseModel);
            if (discriminator) {
                break;
            }
            current = current.baseModel;
        }
        if (discriminator === undefined) {
            return undefined;
        }
        const prop = getProperty(model, discriminator.propertyName);
        if (prop) {
            const values = getStringValues(prop.type);
            if (values.length === 1) {
                return values[0];
            }
        }
        return undefined;
    }
    function getSchemaForModel(model, schemaContext, namespace) {
        const array = getArrayType(model, schemaContext, namespace);
        if (array) {
            return array;
        }
        const rawBaseModel = getLegacyHierarchyBuilding(context.tcgcSdkContext, model);
        const modelSchema = {
            type: "object",
            description: getDoc(program, model),
        };
        applyIntrinsicDecorators(model, modelSchema);
        if (model.baseModel) {
            const discriminatorValue = getDiscriminatorValue(model);
            if (discriminatorValue) {
                const extensions = getExtensions(program, model);
                if (!extensions.has("x-ms-discriminator-value")) {
                    modelSchema["x-ms-discriminator-value"] = discriminatorValue;
                }
            }
        }
        const properties = {};
        if (isRecordModelType(model)) {
            modelSchema.additionalProperties = getSchemaOrRef(model.indexer.value, schemaContext);
        }
        const derivedModels = resolveExternalRef(model)
            ? []
            : model.derivedModels.filter(includeDerivedModel);
        // getSchemaOrRef on all children to push them into components.schemas
        for (const child of derivedModels) {
            getSchemaOrRef(child, schemaContext);
        }
        const discriminator = getDiscriminator(program, model);
        if (discriminator) {
            const { propertyName } = discriminator;
            modelSchema.discriminator = propertyName;
            // Push discriminator into base type, but only if it is not already there
            if (!model.properties.get(propertyName)) {
                properties[propertyName] = {
                    type: "string",
                    description: `Discriminator property for ${model.name}.`,
                };
                modelSchema.required = [propertyName];
            }
        }
        applySummary(model, modelSchema);
        applyExternalDocs(model, modelSchema);
        for (const prop of model.properties.values()) {
            if (rawBaseModel && rawBaseModel.properties.has(prop.name)) {
                const baseProp = rawBaseModel.properties.get(prop.name);
                if (baseProp?.name === prop.name && baseProp.type === prop.type) {
                    // If the property is the same as the base model, skip it
                    continue;
                }
            }
            if (!metadataInfo.isPayloadProperty(prop, schemaContext.visibility, schemaContext.ignoreMetadataAnnotations)) {
                continue;
            }
            if (isNeverType(prop.type)) {
                // If the property has a type of 'never', don't include it in the schema
                continue;
            }
            const propertySchemaName = specIsOnlyXml
                ? resolveEncodedName(program, prop, "application/xml")
                : resolveEncodedName(program, prop, "application/json");
            const clientName = getClientName(context, prop);
            const description = getDoc(program, prop);
            // if this property is a discriminator property, remove it to keep autorest validation happy
            if (model.baseModel) {
                const { propertyName } = getDiscriminator(program, model.baseModel) || {};
                if (propertySchemaName === propertyName) {
                    continue;
                }
            }
            if (!metadataInfo.isOptional(prop, schemaContext.visibility) ||
                prop.name === discriminator?.propertyName) {
                if (!modelSchema.required) {
                    modelSchema.required = [];
                }
                modelSchema.required.push(propertySchemaName);
            }
            // Apply decorators on the property to the type's schema
            properties[propertySchemaName] = resolveProperty(prop, schemaContext);
            const property = properties[propertySchemaName];
            if (propertySchemaName !== clientName) {
                property["x-ms-client-name"] = clientName;
            }
            if (shouldEmitXml && xml.available) {
                const xmlName = resolveEncodedName(program, prop, "application/xml");
                if (xmlName !== propertySchemaName) {
                    property.xml ??= {};
                    property.xml.name = xmlName;
                }
            }
            if (description) {
                property.description = description;
            }
            applySummary(prop, property);
            if (prop.defaultValue && !("$ref" in property)) {
                property.default = getDefaultValue(prop.defaultValue, prop);
            }
            applyExternalDocs(prop, property);
            if (isReadonlyProperty(program, prop)) {
                property.readOnly = true;
            }
            else {
                const lifecycle = getLifecycleVisibilityEnum(program);
                const vis = getVisibilityForClass(program, prop, lifecycle);
                const { read, create, update } = {
                    read: lifecycle.members.get("Read"),
                    create: lifecycle.members.get("Create"),
                    update: lifecycle.members.get("Update"),
                };
                if (vis.size !== lifecycle.members.size) {
                    const mutability = [];
                    if (vis.has(read)) {
                        mutability.push("read");
                    }
                    if (vis.has(update)) {
                        mutability.push("update");
                    }
                    if (vis.has(create)) {
                        mutability.push("create");
                    }
                    if (mutability.length > 0) {
                        property["x-ms-mutability"] = mutability;
                    }
                }
            }
            // Attach any additional OpenAPI extensions
            attachExtensions(prop, property);
        }
        // Special case: if a model type extends a single *templated* base type and
        // has no properties of its own, absorb the definition of the base model
        // into this schema definition.  The assumption here is that any model type
        // defined like this is just meant to rename the underlying instance of a
        // templated type.
        if (model.baseModel &&
            isTemplateDeclarationOrInstance(model.baseModel) &&
            Object.keys(properties).length === 0) {
            // Take the base model schema but carry across the documentation property
            // that we set before
            const baseSchema = getSchemaForType(model.baseModel, schemaContext);
            Object.assign(modelSchema, baseSchema, { description: modelSchema.description });
        }
        else if (model.baseModel) {
            const baseSchema = getSchemaOrRef(rawBaseModel ?? model.baseModel, schemaContext);
            modelSchema.allOf = [baseSchema];
        }
        if (Object.keys(properties).length > 0) {
            modelSchema.properties = properties;
        }
        // Attach any OpenAPI extensions
        attachExtensions(model, modelSchema);
        return modelSchema;
    }
    function canSharePropertyUsingReadonlyOrXMSMutability(prop) {
        const sharedVisibilities = ["Read", "Create", "Update"];
        const lifecycle = getLifecycleVisibilityEnum(program);
        const visibilities = getVisibilityForClass(program, prop, lifecycle);
        // If the property does not have default visibility (all Lifecycle modifiers)
        // then we have to look at the active modifiers to determine if it has any
        // visibility other than read, create, or update, since those are compatible
        // with x-ms-mutability.
        if (visibilities.size !== lifecycle.members.size) {
            for (const visibility of visibilities) {
                if (!sharedVisibilities.includes(visibility.name)) {
                    return false;
                }
            }
        }
        // Otherwise, the property can be shared if it has default visibility or only
        // shared visibilities, but not if it is _invisible_. The property is invisible
        // if it has no active modifiers.
        return visibilities.size !== 0;
    }
    function resolveProperty(prop, context) {
        let propSchema;
        if (prop.type.kind === "Enum" && prop.defaultValue) {
            propSchema = getSchemaForEnum(prop.type);
        }
        else if (prop.type.kind === "Union" && prop.defaultValue) {
            const [asEnum, _] = getUnionAsEnum(prop.type);
            if (asEnum) {
                propSchema = getSchemaForUnionEnum(prop.type, asEnum);
                if (propSchema["x-ms-enum"] && !propSchema["x-ms-enum"].name) {
                    const variants = [...prop.type.variants.values()];
                    const nonNullVariants = variants.filter((v) => !isNullType(v.type));
                    if (nonNullVariants.length === 1 &&
                        variants.length > 1 &&
                        nonNullVariants[0].type.kind === "Enum") {
                        propSchema["x-ms-enum"].name = nonNullVariants[0].type.name;
                    }
                }
            }
            else {
                propSchema = getSchemaOrRef(prop.type, context);
            }
        }
        else {
            propSchema = shouldInlineCoreScalarProperty(prop)
                ? getSchemaForInlineType(prop.type, getOpenAPITypeName(program, prop.type, typeNameOptions), context)
                : getSchemaOrRef(prop.type, context);
            applyArmIdentifiersDecorator(prop.type, propSchema, prop);
        }
        if (shouldEmitXml && xml.available) {
            attachPropertyXml(prop, propSchema);
        }
        return applyIntrinsicDecorators(prop, propSchema);
    }
    function attachExtensions(type, emitObject) {
        // Attach any OpenAPI extensions
        const extensions = getExtensions(program, type);
        if (type.kind === "Model" &&
            (isAzureResource(program, type) ||
                getCustomResourceOptions(program, type)?.isAzureResource === true)) {
            emitObject["x-ms-azure-resource"] = true;
        }
        if (getAsEmbeddingVector(program, type) !== undefined) {
            emitObject["x-ms-embedding-vector"] = true;
        }
        if (type.kind === "Model" && isArmExternalType(program, type) === true) {
            emitObject["x-ms-external"] = true;
        }
        if (type.kind === "Model" && isSecret(program, type) === true) {
            emitObject["x-ms-secret"] = true;
        }
        if (type.kind === "Scalar") {
            const ext = getArmResourceIdentifierConfig(program, type);
            if (ext) {
                emitObject["x-ms-arm-id-details"] = ext;
            }
        }
        if (extensions) {
            for (const key of extensions.keys()) {
                emitObject[key] = extensions.get(key);
            }
        }
    }
    function attachXml(type, schemaName, processed) {
        if (!xml.available)
            return;
        const ns = xml.module.getNs(program, type);
        if ("name" in type && type.name !== undefined && typeof type.name === "string") {
            const xmlName = resolveEncodedName(program, type, "application/xml");
            if (xmlName && xmlName !== schemaName)
                setXmlField("name", xmlName);
        }
        if (ns) {
            if (ns.namespace)
                setXmlField("namespace", ns.namespace);
            if (ns.prefix)
                setXmlField("prefix", ns.prefix);
        }
        function setXmlField(key, value) {
            processed.schema.xml ??= {};
            processed.schema.xml[key] = value;
        }
    }
    function attachPropertyXml(prop, propSchema) {
        if (!xml.available)
            return;
        if (xml.module.isAttribute(program, prop)) {
            setXmlField("attribute", true);
        }
        if (prop.type.kind === "Model" && isArrayModelType(prop.type)) {
            const wrapped = !xml.module.isUnwrapped(program, prop);
            setXmlField("wrapped", wrapped);
        }
        let encode = getEncode(program, prop);
        let resolvedEncodeType = encode?.type ?? prop.type;
        while (resolvedEncodeType.kind === "Scalar" &&
            (encode = getEncode(program, resolvedEncodeType))) {
            resolvedEncodeType = encode.type;
        }
        if (resolvedEncodeType.kind === "Scalar" &&
            $(program).scalar.extendsString(resolvedEncodeType)) {
            if (xml.module.isUnwrapped(program, prop)) {
                setXmlField("x-ms-text", true);
            }
        }
        const xmlNs = xml.module.getNs(program, prop);
        if (xmlNs) {
            setXmlField("namespace", xmlNs.namespace);
            if (xmlNs.prefix)
                setXmlField("prefix", xmlNs.prefix);
        }
        function setXmlField(key, value) {
            propSchema.xml ??= {};
            propSchema.xml[key] = value;
        }
    }
    // Return any string literal values for type
    function getStringValues(type) {
        switch (type.kind) {
            case "String":
                return [type.value];
            case "Union":
                return [...type.variants.values()].flatMap((x) => getStringValues(x.type)).filter((x) => x);
            case "EnumMember":
                return typeof type.value !== "number" ? [type.value ?? type.name] : [];
            case "UnionVariant":
                return getStringValues(type.type);
            default:
                return [];
        }
    }
    function applyIntrinsicDecorators(typespecType, target, usage) {
        const newTarget = { ...target };
        const docStr = getDoc(program, typespecType);
        if (docStr) {
            newTarget.description = docStr;
        }
        const title = getSummary(program, typespecType);
        if (title) {
            target.title = title;
        }
        const formatStr = getFormat(program, typespecType);
        if (formatStr) {
            if (!isSupportedAutorestFormat(formatStr)) {
                reportDiagnostic(program, {
                    code: "unknown-format",
                    format: { schema: "string", format: formatStr },
                    target: typespecType,
                });
            }
            else {
                newTarget.format = formatStr;
            }
        }
        const pattern = getPattern(program, typespecType);
        if (pattern) {
            newTarget.pattern = pattern;
        }
        const minLength = getMinLength(program, typespecType);
        if (minLength !== undefined) {
            newTarget.minLength = minLength;
        }
        const maxLength = getMaxLength(program, typespecType);
        if (maxLength !== undefined) {
            newTarget.maxLength = maxLength;
        }
        const minValue = getMinValue(program, typespecType);
        if (minValue !== undefined) {
            newTarget.minimum = minValue;
        }
        const maxValue = getMaxValue(program, typespecType);
        if (maxValue !== undefined) {
            newTarget.maximum = maxValue;
        }
        const minItems = getMinItems(program, typespecType);
        if (!target.minItems && minItems !== undefined) {
            newTarget.minItems = minItems;
        }
        const maxItems = getMaxItems(program, typespecType);
        if (!target.maxItems && maxItems !== undefined) {
            newTarget.maxItems = maxItems;
        }
        const uniqueItems = (typespecType.kind === "ModelProperty" || typespecType.kind === "Model") &&
            hasUniqueItems(program, typespecType);
        if (uniqueItems && !target.uniqueItems) {
            newTarget.uniqueItems = true;
        }
        if (isSecret(program, typespecType)) {
            newTarget.format = "password";
            newTarget["x-ms-secret"] = true;
        }
        if (typespecType.kind === "ModelProperty" &&
            shouldFlattenProperty(context.tcgcSdkContext, typespecType)) {
            newTarget["x-ms-client-flatten"] = true;
        }
        if (typespecType.kind === "ModelProperty") {
            const clientDefault = getClientDefaultValue(context.tcgcSdkContext, typespecType);
            if (clientDefault) {
                newTarget["x-ms-client-default"] = isNumeric(clientDefault)
                    ? clientDefault.asNumber()
                    : clientDefault;
            }
        }
        attachExtensions(typespecType, newTarget);
        return typespecType.kind === "Scalar" || typespecType.kind === "ModelProperty"
            ? applyEncoding(typespecType, newTarget, usage ?? "body")
            : newTarget;
    }
    function applyEncoding(typespecType, target, usage) {
        const encodeData = getEncode(program, typespecType);
        if (encodeData) {
            const newTarget = { ...target };
            const newType = getSchemaForScalar(encodeData.type);
            // If the target already has a format it takes priority. (e.g. int32)
            const newFormat = mergeFormatAndEncoding(newTarget.format, encodeData.encoding, newType.format);
            if (newFormat) {
                if (!(usage === "parameter" && isEncodingHandledByParameter(newFormat))) {
                    if (!isSupportedAutorestFormat(newFormat)) {
                        reportDiagnostic(program, {
                            code: "unknown-format",
                            format: { schema: "string", format: newFormat },
                            messageId: "encoding",
                            target: typespecType,
                        });
                    }
                    else {
                        newTarget.format = newFormat;
                    }
                    newTarget.type = newType.type;
                    if (newTarget.type !== "array") {
                        delete newTarget.items;
                    }
                    if (newTarget.type !== "object") {
                        delete newTarget.additionalProperties;
                        delete newTarget.properties;
                    }
                }
            }
            return newTarget;
        }
        return target;
    }
    function mergeFormatAndEncoding(format, encoding, encodeAsFormat) {
        switch (format) {
            case undefined:
                return encodeAsFormat ?? encoding ?? format;
            case "date-time":
                switch (encoding) {
                    case "rfc3339":
                        return "date-time";
                    case "unixTimestamp":
                        return "unixtime";
                    case "rfc7231":
                        return "date-time-rfc7231";
                    default:
                        return encoding;
                }
            case "duration":
                switch (encoding) {
                    case "ISO8601":
                        return "duration";
                    default:
                        return encodeAsFormat ?? encoding;
                }
            case "byte":
                switch (encoding) {
                    case "base64":
                        return "byte";
                    default:
                        return encodeAsFormat ?? encoding ?? format;
                }
            default:
                return encodeAsFormat ?? encoding ?? format;
        }
    }
    function applySummary(typespecType, target) {
        const summary = getSummary(program, typespecType);
        if (summary) {
            target.title = summary;
        }
    }
    function applyExternalDocs(typespecType, target) {
        const externalDocs = getExternalDocs(program, typespecType);
        if (externalDocs) {
            target.externalDocs = externalDocs;
        }
    }
    function addXMSEnum(type, schema) {
        if (type.node && type.node.parent && type.node.parent.kind === SyntaxKind.ModelStatement) {
            schema["x-ms-enum"] = {
                name: type.node.parent.id.sv,
                modelAsString: false,
            };
        }
        else if (type.kind === "String") {
            schema["x-ms-enum"] = {
                modelAsString: false,
            };
        }
        else if (type.kind === "Enum") {
            const clientName = getClientName(context, type);
            schema["x-ms-enum"] = {
                name: clientName ?? type.name,
                modelAsString: false,
            };
            const values = [];
            let foundCustom = false;
            for (const member of type.members.values()) {
                const description = getDoc(program, member);
                const memberClientName = getClientName(context, member);
                values.push({
                    name: member.name,
                    value: member.value ?? memberClientName,
                    description,
                });
                if (description || member.value !== undefined) {
                    foundCustom = true;
                }
            }
            if (foundCustom) {
                schema["x-ms-enum"].values = values;
            }
        }
        return schema;
    }
    function getSchemaForStringTemplate(stringTemplate) {
        if (stringTemplate.stringValue === undefined) {
            program.reportDiagnostics(explainStringTemplateNotSerializable(stringTemplate).map((x) => ({
                ...x,
                severity: "warning",
            })));
            return { type: "string" };
        }
        return { type: "string", enum: [stringTemplate.stringValue] };
    }
    function getSchemaForLiterals(typespecType) {
        switch (typespecType.kind) {
            case "Number":
                return { type: "number", enum: [typespecType.value] };
            case "String":
                return addXMSEnum(typespecType, { type: "string", enum: [typespecType.value] });
            case "Boolean":
                return { type: "boolean", enum: [typespecType.value] };
            default:
                return undefined;
        }
    }
    /**
     * If the model is an array model return the OpenAPI2Schema for the array type.
     */
    function getArrayType(typespecType, context, namespace) {
        if (isArrayModelType(typespecType)) {
            const array = {
                type: "array",
                items: getSchemaOrRef(typespecType.indexer.value, {
                    ...context,
                    visibility: context.visibility | Visibility.Item,
                }),
            };
            const armKeyIdentifiers = getArmKeyIdentifiers(program, typespecType);
            if (isArrayTypeArmProviderNamespace(typespecType, namespace) &&
                hasValidArmIdentifiers(armKeyIdentifiers)) {
                array["x-ms-identifiers"] = armKeyIdentifiers;
            }
            return applyIntrinsicDecorators(typespecType, array);
        }
        return undefined;
    }
    function applyArmIdentifiersDecorator(typespecType, schema, property) {
        const armIdentifiers = getArmIdentifiers(program, property);
        if (typespecType.kind !== "Model" || !isArrayModelType(typespecType) || !armIdentifiers) {
            return;
        }
        schema["x-ms-identifiers"] = armIdentifiers;
    }
    function hasValidArmIdentifiers(armIdentifiers) {
        return (armIdentifiers !== undefined &&
            armIdentifiers.length > 0 &&
            !ifArmIdentifiersDefault(armIdentifiers));
    }
    function isArrayTypeArmProviderNamespace(typespecType, namespace) {
        if (typespecType === undefined) {
            return false;
        }
        if (isArmProviderNamespace(program, namespace)) {
            return true;
        }
        if (typespecType.indexer?.value.kind === "Model") {
            return isArmProviderNamespace(program, typespecType.indexer.value.namespace);
        }
        return false;
    }
    function getSchemaForScalar(scalar) {
        let result = {};
        const isStd = program.checker.isStdType(scalar);
        if (isStd) {
            result = getSchemaForStdScalars(scalar);
        }
        else if (scalar.baseScalar) {
            result = getSchemaForScalar(scalar.baseScalar);
        }
        const withDecorators = applyIntrinsicDecorators(scalar, result);
        if (isStd) {
            // Standard types are going to be inlined in the spec and we don't want the description of the scalar to show up
            delete withDecorators.description;
        }
        return withDecorators;
    }
    function getSchemaForStdScalars(scalar) {
        function reportNonspecificScalar(scalarName, chosenScalarName) {
            reportDiagnostic(program, {
                code: "nonspecific-scalar",
                format: { type: scalarName, chosenType: chosenScalarName },
                target: scalar,
            });
        }
        switch (scalar.name) {
            case "bytes":
                return { type: "string", format: "byte" };
            case "numeric":
                reportNonspecificScalar("numeric", "int64");
                return { type: "integer", format: "int64" };
            case "integer":
                reportNonspecificScalar("integer", "int64");
                return { type: "integer", format: "int64" };
            case "int8":
                return { type: "integer", format: "int8" };
            case "int16":
                return { type: "integer", format: "int16" };
            case "int32":
                return { type: "integer", format: "int32" };
            case "int64":
                return { type: "integer", format: "int64" };
            case "safeint":
                return { type: "integer", format: "int64" };
            case "uint8":
                return { type: "integer", format: "uint8" };
            case "uint16":
                return { type: "integer", format: "uint16" };
            case "uint32":
                return { type: "integer", format: "uint32" };
            case "uint64":
                return { type: "integer", format: "uint64" };
            case "float":
                reportNonspecificScalar("float", "float64");
                return { type: "number" };
            case "float64":
                return { type: "number", format: "double" };
            case "float32":
                return { type: "number", format: "float" };
            case "decimal":
                return { type: "number", format: "decimal" };
            case "decimal128":
                return { type: "number", format: "decimal" };
            case "string":
                return { type: "string" };
            case "boolean":
                return { type: "boolean" };
            case "plainDate":
                return { type: "string", format: "date" };
            case "utcDateTime":
            case "offsetDateTime":
                return { type: "string", format: "date-time" };
            case "plainTime":
                return { type: "string", format: "time" };
            case "duration":
                return { type: "string", format: "duration" };
            case "url":
                return { type: "string", format: "uri" };
            default:
                const _assertNever = scalar.name;
                return {};
        }
    }
    function processAuth(serviceNamespace) {
        const authentication = resolveAuthentication(httpService);
        if (authentication) {
            return processServiceAuthentication(authentication, serviceNamespace);
        }
        return undefined;
    }
    function processServiceAuthentication(authentication, serviceNamespace) {
        const oaiSchemes = {};
        for (const scheme of authentication.schemes) {
            const result = getOpenAPI2Scheme(scheme, serviceNamespace);
            if (result !== undefined) {
                oaiSchemes[scheme.id] = result;
            }
        }
        const security = getOpenAPISecurity(oaiSchemes, authentication.defaultAuth);
        return { securitySchemes: oaiSchemes, security };
    }
    function getOpenAPI2Scheme(auth, serviceNamespace) {
        switch (auth.type) {
            case "http":
                if (auth.scheme.toLowerCase() !== "basic") {
                    reportDiagnostic(program, {
                        code: "unsupported-http-auth-scheme",
                        target: serviceNamespace,
                        format: { scheme: auth.scheme },
                    });
                    return undefined;
                }
                return { type: "basic", description: auth.description };
            case "apiKey":
                if (auth.in === "cookie") {
                    return undefined;
                }
                return { type: "apiKey", description: auth.description, in: auth.in, name: auth.name };
            case "oauth2":
                const flow = auth.flows[0];
                if (flow === undefined) {
                    return undefined;
                }
                const oaiFlowName = getOpenAPI2Flow(flow.type);
                return {
                    type: "oauth2",
                    description: auth.description,
                    flow: oaiFlowName,
                    authorizationUrl: flow.authorizationUrl,
                    tokenUrl: flow.tokenUrl,
                    scopes: Object.fromEntries(flow.scopes.map((x) => [x.value, x.description ?? ""])),
                };
            case "openIdConnect":
            default:
                reportDiagnostic(program, {
                    code: "unsupported-auth",
                    format: { authType: auth.type },
                    target: service.type,
                });
                return undefined;
        }
    }
    function getOpenAPISecurity(oaiSchemes, authReference) {
        const security = authReference.options
            .map((authOption) => {
            const securityOption = {};
            for (const httpAuthRef of authOption.all) {
                const scopes = getScopesForAuthReference(httpAuthRef);
                if (httpAuthRef.auth.id in oaiSchemes && scopes) {
                    securityOption[httpAuthRef.auth.id] = scopes;
                }
            }
            return securityOption;
        })
            .filter((x) => Object.keys(x).length > 0);
        return security;
    }
    function getScopesForAuthReference(httpAuthRef) {
        switch (httpAuthRef.kind) {
            case "noAuth":
                // should emit "{}" as a security option https://github.com/OAI/OpenAPI-Specification/issues/14#issuecomment-297457320
                return undefined;
            case "oauth2":
                return httpAuthRef.scopes;
            default:
                return [];
        }
    }
    function getOpenAPI2Flow(flow) {
        switch (flow) {
            case "authorizationCode":
                return "accessCode";
            case "clientCredentials":
                return "application";
            case "implicit":
                return "implicit";
            case "password":
                return "password";
            default:
                const _assertNever = flow;
                compilerAssert(false, "Unreachable");
        }
    }
}
class ErrorTypeFoundError extends Error {
    constructor() {
        super("Error type found in evaluated TypeSpec output");
    }
}
export function sortOpenAPIDocument(doc) {
    // Doing this to make sure the classes with toJSON are resolved.
    const unsorted = JSON.parse(JSON.stringify(doc));
    const sorted = sortWithJsonSchema(unsorted, AutorestOpenAPISchema);
    return sorted;
}
async function checkExamplesDirExists(host, dir) {
    try {
        return (await host.stat(dir)).isDirectory();
    }
    catch (err) {
        return false;
    }
}
async function searchExampleJsonFiles(program, exampleDir) {
    const host = program.host;
    const exampleFiles = [];
    // Recursive file search
    async function recursiveSearch(dir) {
        const fileItems = await host.readDir(dir);
        for (const item of fileItems) {
            const fullPath = joinPaths(dir, item);
            const relativePath = getRelativePathFromDirectory(exampleDir, fullPath, false);
            if ((await host.stat(fullPath)).isDirectory()) {
                await recursiveSearch(fullPath);
            }
            else if ((await host.stat(fullPath)).isFile() &&
                getAnyExtensionFromPath(item) === ".json") {
                exampleFiles.push(normalizePath(relativePath));
            }
        }
    }
    await recursiveSearch(exampleDir);
    return exampleFiles;
}
async function loadExamples(program, options, version) {
    const host = program.host;
    const diagnostics = createDiagnosticCollector();
    const examplesBaseDir = options.examplesDirectory ?? resolvePath(program.projectRoot, "examples");
    const exampleDir = version ? resolvePath(examplesBaseDir, version) : resolvePath(examplesBaseDir);
    if (!(await checkExamplesDirExists(host, exampleDir))) {
        if (options.examplesDirectory) {
            diagnostics.add(createDiagnostic({
                code: "example-loading",
                messageId: "noDirectory",
                format: { directory: exampleDir },
                target: NoTarget,
            }));
        }
        return diagnostics.wrap(new Map());
    }
    const map = new Map();
    const exampleFiles = await searchExampleJsonFiles(program, exampleDir);
    for (const fileName of exampleFiles) {
        try {
            const exampleFile = await host.readFile(resolvePath(exampleDir, fileName));
            const example = JSON.parse(exampleFile.text);
            if (!example.operationId || !example.title) {
                diagnostics.add(createDiagnostic({
                    code: "example-loading",
                    messageId: "noOperationId",
                    format: { filename: fileName },
                    target: { file: exampleFile, pos: 0, end: 0 },
                }));
                continue;
            }
            if (!map.has(example.operationId)) {
                map.set(example.operationId, {});
            }
            const examples = map.get(example.operationId);
            if (example.title in examples) {
                diagnostics.add(createDiagnostic({
                    code: "duplicate-example-file",
                    target: { file: exampleFile, pos: 0, end: 0 },
                    format: {
                        filename: fileName,
                        operationId: example.operationId,
                        title: example.title,
                    },
                }));
            }
            examples[example.title] = {
                relativePath: fileName,
                file: exampleFile,
                data: example,
            };
        }
        catch (err) {
            diagnostics.add(createDiagnostic({
                code: "example-loading",
                messageId: "default",
                format: { filename: fileName, error: err?.toString() ?? "" },
                target: NoTarget,
            }));
        }
    }
    return diagnostics.wrap(map);
}
function isHttpParameterProperty(httpProperty) {
    return ["header", "query", "path", "cookie"].includes(httpProperty.kind);
}
export function createDocumentProxy(program, service, options, version) {
    const features = getResourceFeatureSet(program, service.type);
    if (options.outputSplitting === undefined ||
        options.outputSplitting !== "legacy-feature-files" ||
        features === undefined ||
        features.size < 2) {
        return createDefaultDocumentProxy(program, service, options, version);
    }
    else {
        return createFeatureDocumentProxy(program, service, options, version);
    }
}
export function createDefaultDocumentProxy(program, service, options, version) {
    const root = initializeOpenApi2Document(program, service, version);
    const tags = new Set();
    const definitions = new Map();
    const parameters = new Map();
    let examples = new Map();
    let operationIdsWithExamples = new Set();
    return {
        getDefinitionMap() {
            const result = {};
            for (const [name, schema] of definitions) {
                result[name] = schema;
            }
            return result;
        },
        writeDefinition: (name, processed, schema) => {
            definitions.set(name, schema);
        },
        getParameterMap() {
            const result = {};
            for (const [name, [_, parameter]] of parameters) {
                result[name] = parameter;
            }
            return result;
        },
        writeParameter(key, prop, parameter) {
            parameters.set(key, [prop, parameter]);
            let defaultParameters = root.parameters;
            if (defaultParameters === undefined) {
                defaultParameters = {};
                root.parameters = defaultParameters;
            }
            defaultParameters[key] = { ...parameter };
        },
        createOrGetEndpoint(op, context) {
            const pathItem = initPathItem(program, op, root);
            if (!pathItem[op.verb]) {
                pathItem[op.verb] = { parameters: [] };
            }
            const resolvedOp = pathItem[op.verb];
            resolvedOp.operationId = resolveOperationId(context, op.operation);
            return resolvedOp;
        },
        addTag(tag, op) {
            tags.add(tag);
        },
        setGlobalConsumes(consumes) {
            root.consumes = consumes;
        },
        setGlobalProduces(produces) {
            root.produces = produces;
        },
        addAdditionalInfo(info) {
            if (info !== undefined) {
                Object.assign(root.info, info);
            }
        },
        addHostInfo(hostInfo) {
            Object.assign(root, hostInfo);
        },
        addSecurityRequirements(requirements) {
            if (requirements !== undefined && requirements.length > 0) {
                root.security = requirements;
            }
        },
        addSecuritySchemes(schemes) {
            if (schemes !== undefined && Object.keys(schemes).length > 0) {
                root.securityDefinitions = schemes;
            }
        },
        writeExamples(inExamples, exampleIds) {
            examples = inExamples;
            operationIdsWithExamples = exampleIds;
        },
        resolveDocuments(context) {
            root.definitions = {};
            for (const [name, schema] of definitions) {
                root.definitions[name] = schema;
            }
            finalizeOpenApi2Document(root, tags);
            return Promise.resolve([
                {
                    document: root,
                    operationExamples: [...operationIdsWithExamples]
                        .map((operationId) => {
                        const data = examples.get(operationId);
                        if (data) {
                            return { operationId, examples: Object.values(data) };
                        }
                        else {
                            return undefined;
                        }
                    })
                        .filter((x) => x),
                    outputFile: context.outputFile,
                    context: context,
                },
            ]);
        },
        createLocalRef(type) {
            const feature = getFeature(program, type);
            const result = new LateBoundReference();
            result.file = feature.fileName;
            return result;
        },
        createExternalRef(absoluteRef) {
            const result = new LateBoundReference();
            result.setRemoteValue(absoluteRef);
            return result;
        },
        getCurrentFeature() {
            return undefined;
        },
        setCurrentFeature(feature) {
            return;
        },
        getParameterRef(key) {
            return `#/parameters/${encodeURIComponent(key)}`;
        },
    };
}
function createFeatureDocumentProxy(program, service, options, version) {
    const features = getResourceFeatureSet(program, service.type);
    if (features === undefined || features.size < 2)
        return createDefaultDocumentProxy(program, service, options, version);
    const root = new Map();
    const operationFeatures = new Map();
    let examples = new Map();
    let operationIdsWithExamples = new Set();
    for (const featureName of features.keys()) {
        const featureOptions = features.get(featureName);
        root.set(featureName.toLowerCase(), initializeOpenAPIDocumentItem(program, service, featureOptions, version));
    }
    const defaultFeature = [...root.entries()].filter(([key, _]) => key.toLowerCase() === "common")[0][1];
    const definitions = new Map();
    const resolvedParameters = new Map();
    const commonFeatureKey = [...features.keys()].find((key) => key.toLowerCase() === "common");
    const commonFeature = features.get(commonFeatureKey);
    let currentFeature = "";
    return {
        getDefinitionMap() {
            const result = {};
            for (const [name, [_, schema]] of definitions) {
                result[name] = schema;
            }
            return result;
        },
        writeDefinition: (name, processed, schema) => {
            const feature = getFeatureKey(program, processed.type);
            definitions.set(name, [feature, schema]);
        },
        getParameterMap() {
            const result = {};
            for (const [name, [_, parameter]] of resolvedParameters) {
                result[name] = parameter;
            }
            return result;
        },
        writeParameter(key, prop, parameter) {
            resolvedParameters.set(key, [prop, parameter]);
            let defaultParameters = defaultFeature.document.parameters;
            if (defaultParameters === undefined) {
                defaultParameters = {};
                defaultFeature.document.parameters = defaultParameters;
            }
            defaultParameters[key] = { ...parameter };
        },
        createOrGetEndpoint(op, context) {
            const options = getFeature(program, op.operation);
            const item = root.get(options.featureName.toLowerCase());
            const pathItem = initPathItem(program, op, item.document);
            if (!pathItem[op.verb]) {
                pathItem[op.verb] = { parameters: [] };
            }
            const resolvedOp = pathItem[op.verb];
            const opId = resolveOperationId(context, op.operation);
            addFeatureOperation(opId, options.featureName);
            resolvedOp.operationId = opId;
            return resolvedOp;
        },
        setGlobalConsumes(mimeTypes) {
            for (const featureItem of root.values()) {
                featureItem.document.consumes = [];
                for (const mimeType of mimeTypes) {
                    featureItem.document.consumes.push(mimeType);
                }
            }
        },
        setGlobalProduces(mimeTypes) {
            for (const featureItem of root.values()) {
                featureItem.document.produces = [];
                for (const mimeType of mimeTypes) {
                    featureItem.document.produces.push(mimeType);
                }
            }
        },
        addTag(tag, op) {
            const feature = getFeatureKey(program, op);
            const item = root.get(feature);
            item.tags.add(tag);
        },
        addAdditionalInfo(info) {
            if (info !== undefined) {
                for (const featureItem of root.values())
                    Object.assign(featureItem.document.info, info);
            }
        },
        addHostInfo(hostInfo) {
            for (const featureItem of root.values())
                Object.assign(featureItem.document, hostInfo);
        },
        addSecurityRequirements(requirements) {
            if (requirements !== undefined && requirements.length > 0) {
                for (const featureItem of root.values())
                    featureItem.document.security = requirements;
            }
        },
        addSecuritySchemes(schemes) {
            if (schemes !== undefined && Object.keys(schemes).length > 0) {
                for (const featureItem of root.values())
                    featureItem.document.securityDefinitions = schemes;
            }
        },
        writeExamples(inExamples, exampleIds) {
            examples = inExamples;
            operationIdsWithExamples = exampleIds;
        },
        resolveDocuments(context) {
            const docs = [];
            for (const [featureName, featureItem] of root.entries()) {
                const exampleIds = operationFeatures.get(featureName) || new Set();
                const featureExamples = [...exampleIds]
                    .filter((id) => operationIdsWithExamples.has(id))
                    .map((operationId) => {
                    const data = examples.get(operationId);
                    if (data) {
                        return { operationId, examples: Object.values(data) };
                    }
                    else {
                        return undefined;
                    }
                })
                    .filter((x) => x);
                const definitionsForFeature = Array.from(definitions.entries()).filter(([_, [definitionFeature, __]]) => definitionFeature === featureName);
                featureItem.document.definitions = {};
                for (const [defName, [_, defSchema]] of definitionsForFeature) {
                    featureItem.document.definitions[defName] = defSchema;
                }
                finalizeOpenApi2Document(featureItem.document, featureItem.tags);
                docs.push({
                    document: featureItem.document,
                    operationExamples: featureExamples,
                    outputFile: context.outputFile,
                    feature: featureItem.options.fileName,
                    context: context,
                });
            }
            // Collect operation examples for this feature
            return Promise.resolve(docs);
        },
        createLocalRef(type) {
            let feature = getFeature(program, type);
            currentFeature = feature.featureName;
            if (currentFeature.toLowerCase() === "common" && commonFeature !== undefined) {
                feature = commonFeature;
            }
            const result = new LateBoundReference();
            result.useFeatures = true;
            result.file = feature.fileName;
            if (commonFeature !== undefined) {
                result.commonFile = commonFeature.fileName;
            }
            result.getFileContext = () => this.getCurrentFeature();
            return result;
        },
        createExternalRef(absoluteRef) {
            const result = new LateBoundReference();
            result.setRemoteValue(absoluteRef);
            return result;
        },
        getCurrentFeature() {
            return currentFeature;
        },
        setCurrentFeature(feature) {
            currentFeature = feature;
        },
        getParameterRef(key) {
            return `./${commonFeature?.fileName ?? "common"}.json#/parameters/${encodeURIComponent(key)}`;
        },
    };
    function getFeatureKey(program, type) {
        const feature = getFeature(program, type);
        return feature.featureName.toLowerCase();
    }
    function addFeatureOperation(operationId, featureName) {
        featureName = featureName.toLowerCase();
        if (!operationFeatures.has(featureName)) {
            operationFeatures.set(featureName, new Set([operationId]));
            return;
        }
        const ops = operationFeatures.get(featureName);
        ops.add(operationId);
    }
}
function initializeOpenAPIDocumentItem(program, service, options, version) {
    return {
        document: initializeOpenApi2Document(program, service, version),
        operationExamples: new Map(),
        tags: new Set(),
        options,
    };
}
function initializeOpenApi2Document(program, service, version) {
    return {
        swagger: "2.0",
        info: {
            title: "(title)",
            version: version ?? "0000-00-00",
            "x-typespec-generated": [{ emitter: "@azure-tools/typespec-autorest" }],
        },
        schemes: ["https"],
        externalDocs: getExternalDocs(program, service.type),
        produces: [], // Pre-initialize produces and consumes so that
        consumes: [], // they show up at the top of the document
        tags: [],
        paths: {},
        "x-ms-paths": {},
        definitions: {},
        parameters: {},
    };
}
function finalizeOpenApi2Document(root, tags) {
    const xMsPaths = root["x-ms-paths"];
    if (xMsPaths && Object.keys(xMsPaths).length === 0) {
        delete root["x-ms-paths"];
    }
    const rootSecurity = root.security;
    if (rootSecurity && Object.keys(rootSecurity).length === 0) {
        delete root.security;
    }
    const securityDefs = root.securityDefinitions;
    if (securityDefs && Object.keys(securityDefs).length === 0) {
        delete root.securityDefinitions;
    }
    if (root.consumes !== undefined && root.consumes.length === 0) {
        delete root.consumes;
    }
    if (root.produces !== undefined && root.produces.length === 0) {
        delete root.produces;
    }
    root.tags = Array.from(tags).map((tagName) => ({ name: tagName }));
}
function initPathItem(program, operation, root) {
    let { path, operation: op, verb } = operation;
    let pathsObject = root.paths;
    if (root.paths[path]?.[verb] === undefined && !path.includes("?")) {
        pathsObject = root.paths;
    }
    else if (requiresXMsPaths(program, path, op)) {
        // if the key already exists in x-ms-paths, append the operation id.
        if (path.includes("?")) {
            if (root["x-ms-paths"]?.[path] !== undefined) {
                path += `&_overload=${operation.operation.name}`;
            }
        }
        else {
            path += `?_overload=${operation.operation.name}`;
        }
        pathsObject = root["x-ms-paths"];
    }
    else {
        // This should not happen because http library should have already validated duplicate path or the routes must have been using shared routes and so goes in previous condition.
        compilerAssert(false, `Duplicate route "${path}". This is unexpected.`);
    }
    if (!pathsObject[path]) {
        pathsObject[path] = {};
    }
    return pathsObject[path];
}
function requiresXMsPaths(program, path, operation) {
    const isShared = isSharedRoute(program, operation) ?? false;
    if (path.includes("?")) {
        return true;
    }
    return isShared;
}
