// @azure-tools/typespec-autorest/dist/src/types.js
import { getFeature } from "@azure-tools/typespec-azure-resource-manager";
import { compilerAssert, } from "@typespec/compiler";
/**
 * Represents a node that will hold a JSON reference. The value is computed
 * at the end so that we can defer decisions about the name that is
 * referenced.
 */
export class LateBoundReference {
    isLocal;
    file;
    value;
    useFeatures = false;
    commonFile = "common";
    getFileContext = () => undefined;
    setLocalValue(program, inValue, type) {
        if (type) {
            switch (type.kind) {
                case "Model":
                case "ModelProperty":
                    if (this.useFeatures) {
                        const feature = getFeature(program, type);
                        let featureFile = feature.fileName;
                        if (feature.featureName.toLowerCase() === "common") {
                            featureFile = this.commonFile;
                        }
                        this.file = featureFile;
                    }
                    else {
                        this.file = undefined;
                    }
                    break;
                default:
                    this.file = this.useFeatures ? this.commonFile : undefined;
            }
        }
        this.isLocal = true;
        this.value = inValue;
    }
    setRemoteValue(inValue) {
        this.isLocal = false;
        this.value = inValue;
    }
    toJSON() {
        compilerAssert(this.value, "Reference value never set.");
        const referencingFile = this.getFileContext();
        if (!this.isLocal)
            return this.value;
        if (referencingFile === undefined || this.file === undefined || referencingFile === this.file)
            return `#/definitions/${this.value}`;
        return `./${this.file}.json#/definitions/${this.value}`;
    }
}
