// @azure-tools/typespec-autorest/dist/src/utils.js
import { getClientLocation, getClientNameOverride, } from "@azure-tools/typespec-client-generator-core";
import { getFriendlyName, getLifecycleVisibilityEnum, getVisibilityForClass, isGlobalNamespace, isService, isTemplateInstance, } from "@typespec/compiler";
import { capitalize } from "@typespec/compiler/casing";
import { getOperationId } from "@typespec/openapi";
export function getClientName(context, type) {
    const clientName = getClientNameOverride(context.tcgcSdkContext, type);
    return clientName ?? type.name;
}
/**
 * Determines whether a type will be inlined in OpenAPI rather than defined
 * as a schema and referenced.
 *
 * All anonymous types (anonymous models, arrays, tuples, etc.) are inlined.
 *
 * Template instantiations are inlined unless they have a friendly name.
 *
 * A friendly name can be provided by the user using `@friendlyName`
 * decorator, or chosen by default in simple cases.
 */
export function shouldInline(program, type) {
    if (getFriendlyName(program, type)) {
        return false;
    }
    switch (type.kind) {
        case "Model":
        case "Union":
            return !type.name || isTemplateInstance(type);
        case "Scalar":
            return program.checker.isStdType(type) || isTemplateInstance(type);
        case "Enum":
            return !type.name;
        default:
            return true;
    }
}
/**
 * Resolve the OpenAPI operation ID for the given operation using the following logic:
 * - If @operationId was specified use that value
 * - If @clientLocation was specified:
 *   - If the target is a string, use the string value as the prefix of the operation ID
 *   - If the target is an Interface, use the interface name as the prefix
 *   - If the target is a Namespace and it's not the service namespace or global namespace, use the namespace name as the prefix
 *   - If the target is the service namespace or global namespace, use the operation name as the operation ID
 * - If operation is defined at the root or under the service namespace return `<operation.name>`
 * - Otherwise(operation is under another namespace or interface) return `<namespace/interface.name>_<operation.name>`
 *
 * @param program TypeSpec Program
 * @param operation Operation
 * @returns Operation ID in this format `<name>` or `<group>_<name>`
 */
export function resolveOperationId(context, operation) {
    const { program } = context;
    const explicitOperationId = getOperationId(program, operation);
    if (explicitOperationId) {
        return explicitOperationId;
    }
    const operationName = getClientName(context, operation);
    // Check for `@clientLocation` decorator
    const clientLocation = getClientLocation(context.tcgcSdkContext, operation);
    if (clientLocation) {
        if (typeof clientLocation === "string") {
            return standardizeOperationId(`${clientLocation}_${operationName}`);
        }
        if (clientLocation.kind === "Interface") {
            return standardizeOperationId(`${getClientName(context, clientLocation)}_${operationName}`);
        }
        if (clientLocation.kind === "Namespace") {
            if (isGlobalNamespace(program, clientLocation) || isService(program, clientLocation)) {
                return standardizeOperationId(operationName);
            }
            return standardizeOperationId(`${getClientName(context, clientLocation)}_${operationName}`);
        }
    }
    if (operation.interface) {
        return standardizeOperationId(`${getClientName(context, operation.interface)}_${operationName}`);
    }
    const namespace = operation.namespace;
    if (namespace === undefined ||
        isGlobalNamespace(program, namespace) ||
        isService(program, namespace)) {
        return standardizeOperationId(operationName);
    }
    return standardizeOperationId(`${getClientName(context, namespace)}_${operationName}`);
}
/**
 * Determines if a property is read-only, which is defined as having only the
 * Lifecycle.Read visibility active within the Lifecycle visibility class.
 *
 * If there is more than one Lifecycle visibility modifier active, then the
 * property is not read-only. For example, `@visibility(Lifecycle.Read, Lifecycle.Update)`
 * does not designate a read-only property.
 */
export function isReadonlyProperty(program, property) {
    const lifecycle = getLifecycleVisibilityEnum(program);
    const read = lifecycle.members.get("Read");
    const visibility = getVisibilityForClass(program, property, lifecycle);
    // note: multiple visibilities that include read are not handled using
    // readonly: true, but using separate schemas.
    return visibility.size === 1 && visibility.has(read);
}
function standardizeOperationId(name) {
    return name
        .split("_")
        .map((s) => capitalize(s))
        .join("_");
}
const allowedAutorestFormats = new Set([
    // number format
    "int32",
    "int64",
    "float",
    "double",
    "unixtime",
    "decimal",
    // OAS-defined formats
    "byte",
    "binary",
    "date",
    "date-time",
    "password",
    // Additional formats recognized by autorest
    "char",
    "time",
    "date-time-rfc1123",
    "date-time-rfc7231", // Support for https://github.com/Azure/autorest/issues/4740
    "duration",
    "uuid",
    "base64url",
    "url",
    "odata-query",
    "certificate",
    // ajv supported format
    "uri",
    "uri-reference",
    "uri-template",
    "email",
    "hostname",
    "ipv4",
    "ipv6",
    "regex",
    "json-pointer",
    "relative-json-pointer",
    // for arm id purpose
    "arm-id",
    // Custom exclusions
    "duration-constant",
]);
/**
 * Check if the given format is supported by Autorest.
 * Those formats are validated by https://github.com/Azure/azure-openapi-validator/blob/main/packages/rulesets/src/spectral/functions/schema-format.ts#L17
 */
export function isSupportedAutorestFormat(formatStr) {
    return allowedAutorestFormats.has(formatStr.toLowerCase());
}
