// oav/dist/lib/models/errorWrapper.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorWrapper = void 0;
const errorResponse_1 = require("./errorResponse");
/**
 * @class
 * Initializes a new instance of the ErrorWrapper class.
 * @constructor
 * Wrapper object
 *
 * @member {array} [errors] An array of inner errors.
 *
 */
class ErrorWrapper extends errorResponse_1.ErrorResponse {
    constructor() {
        super();
    }
    /**
     * Defines the metadata of ErrorWrapper
     *
     * @returns {object} metadata of ErrorWrapper
     *
     */
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    mapper() {
        return {
            required: false,
            serializedName: "ErrorWrapper",
            type: {
                name: "Composite",
                className: "ErrorWrapper",
                modelProperties: {
                    code: {
                        required: false,
                        serializedName: "code",
                        type: {
                            name: "String",
                        },
                    },
                    message: {
                        required: false,
                        serializedName: "message",
                        type: {
                            name: "String",
                        },
                    },
                    description: {
                        required: false,
                        serializedName: "description",
                        type: {
                            name: "String",
                        },
                    },
                    params: {
                        required: false,
                        serializedName: "params",
                        type: {
                            name: "Sequence",
                            element: {
                                required: false,
                                serializedName: "StringElementType",
                                type: {
                                    name: "String",
                                },
                            },
                        },
                    },
                    path: {
                        required: false,
                        serializedName: "path",
                        type: {
                            name: "Sequence",
                            element: {
                                required: false,
                                serializedName: "StringElementType",
                                type: {
                                    name: "String",
                                },
                            },
                        },
                    },
                    errors: {
                        required: false,
                        serializedName: "errors",
                        type: {
                            name: "Sequence",
                            element: {
                                required: false,
                                serializedName: "ErrorResponseElementType",
                                type: {
                                    name: "Composite",
                                    className: "ErrorResponse",
                                },
                            },
                        },
                    },
                },
            },
        };
    }
}
exports.ErrorWrapper = ErrorWrapper;
