// @microsoft.azure/openapi-validator-rulesets/dist/native/functions/arm-resource-validation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.xmsPageableListByRGAndSubscriptions = exports.resourcesHaveRequiredProperties = exports.operationsAPIImplementation = exports.bodyTopLevelProperties = exports.armResourcePropertiesBag = exports.trackedResourcesHavePatch = exports.allResourcesHaveDelete = exports.trackedResourceBeyondsThirdLevel = exports.trackedResourcesMustHavePut = void 0;
const arm_helper_1 = require("../utilities/arm-helper");
const rules_helper_1 = require("../utilities/rules-helper");
function* trackedResourcesMustHavePut(openapiSection, options, ctx) {
    const armHelper = new arm_helper_1.ArmHelper(ctx === null || ctx === void 0 ? void 0 : ctx.document, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
    const allTrackedResources = armHelper.getTrackedResources();
    for (const re of allTrackedResources) {
        if (!re.operations.some((op) => op.httpMethod === "put")) {
            yield {
                location: ["definitions", re.modelName],
                message: `The tracked resource ${re.modelName} does not have a corresponding put operation.`,
            };
        }
    }
}
exports.trackedResourcesMustHavePut = trackedResourcesMustHavePut;
function* trackedResourceBeyondsThirdLevel(openapiSection, options, ctx) {
    const armHelper = new arm_helper_1.ArmHelper(ctx === null || ctx === void 0 ? void 0 : ctx.document, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
    const allTrackedResources = armHelper.getTrackedResources();
    const regex = /^.*\/providers\/microsoft\.\w+(?:\/\w+\/(\w+|{\w+})){4,}/gi;
    for (const re of allTrackedResources) {
        if (re.operations.some((op) => regex.test(op.apiPath))) {
            yield {
                location: ["definitions", re.modelName],
                message: `The tracked resource ${re.modelName} is beyond third level of nesting.`,
            };
        }
    }
}
exports.trackedResourceBeyondsThirdLevel = trackedResourceBeyondsThirdLevel;
function* allResourcesHaveDelete(openapiSection, options, ctx) {
    var _a;
    const armHelper = new arm_helper_1.ArmHelper(ctx === null || ctx === void 0 ? void 0 : ctx.document, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
    const allResources = options.isTrackedResource ? armHelper.getTrackedResources() : armHelper.getProxyResources();
    for (const re of allResources) {
        const apiPath = (_a = re.operations.find((op) => op.apiPath)) === null || _a === void 0 ? void 0 : _a.apiPath;
        if (apiPath && !apiPath.toLowerCase().includes("privateendpointconnectionproxies")) {
            if ((options.isTrackedResource || armHelper.findOperation(apiPath, "put")) && !armHelper.findOperation(apiPath, "delete")) {
                yield {
                    location: ["definitions", re.modelName],
                    message: `The resource ${re.modelName} does not have a corresponding delete operation.`,
                };
            }
        }
    }
}
exports.allResourcesHaveDelete = allResourcesHaveDelete;
function* trackedResourcesHavePatch(openapiSection, options, ctx) {
    var _a;
    const armHelper = new arm_helper_1.ArmHelper(ctx === null || ctx === void 0 ? void 0 : ctx.document, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
    const allTrackedResources = armHelper.getTrackedResources();
    for (const re of allTrackedResources) {
        const apiPath = (_a = re.operations.find((op) => op.apiPath)) === null || _a === void 0 ? void 0 : _a.apiPath;
        if (apiPath && !apiPath.toLowerCase().includes("privateendpointconnectionproxies")) {
            if (!armHelper.findOperation(apiPath, "patch")) {
                yield {
                    location: ["definitions", re.modelName],
                    message: `Tracked resource '${re.modelName}' must have patch operation that at least supports the update of tags.`,
                };
            }
        }
    }
}
exports.trackedResourcesHavePatch = trackedResourcesHavePatch;
function* armResourcePropertiesBag(openapiSection, options, ctx) {
    const armHelper = new arm_helper_1.ArmHelper(ctx === null || ctx === void 0 ? void 0 : ctx.document, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
    const propertiesBagTracked = [
        "name",
        "type",
        "id",
        "location",
        "tags",
        "plan",
        "sku",
        "etag",
        "managedby",
        "identity",
        "kind",
        "zones",
        "systemdata",
        "extendedlocation",
    ];
    const trackedResources = armHelper.getTrackedResources();
    yield* processResources(armHelper, trackedResources, propertiesBagTracked);
    const propertiesBagProxy = ["name", "id", "type", "etag", "systemdata"];
    const proxyResources = armHelper.getProxyResources();
    yield* processResources(armHelper, proxyResources, propertiesBagProxy);
}
exports.armResourcePropertiesBag = armResourcePropertiesBag;
function checkPropertiesBag(armHelper, model, resourceName, propertiesBag, propertiesPath) {
    let messages = [];
    const properties = armHelper.getProperty(model, "properties");
    if (properties) {
        propertiesPath.push("properties");
        for (const p of propertiesBag) {
            if (armHelper.getProperty(properties, p)) {
                messages.push(`Top level property names should not be repeated inside the properties bag for ARM resource '${resourceName}'. Properties [${propertiesPath
                    .concat(p)
                    .join(".")}] conflict with ARM top level properties. Please rename these.`);
            }
        }
        const subResult = checkPropertiesBag(armHelper, properties, resourceName, propertiesBag, propertiesPath);
        messages = messages.concat(subResult);
    }
    return messages;
}
function* processResources(armHelper, resources, propertiesBag) {
    for (const resource of resources) {
        const model = armHelper.getResourceByName(resource.modelName);
        const messages = checkPropertiesBag(armHelper, model, resource.modelName, propertiesBag, []);
        for (const message of messages) {
            yield {
                location: ["definitions", resource.modelName],
                message,
            };
        }
    }
}
function* bodyTopLevelProperties(openapiSection, options, ctx) {
    const armHelper = new arm_helper_1.ArmHelper(ctx === null || ctx === void 0 ? void 0 : ctx.document, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
    const allResources = armHelper.getAllResources(true, false);
    for (const re of allResources) {
        const allowedBodyTopLevelProperties = [
            "name",
            "type",
            "id",
            "location",
            "properties",
            "tags",
            "plan",
            "sku",
            "etag",
            "managedby",
            "identity",
            "kind",
            "zones",
            "systemdata",
            "extendedlocation",
        ];
        const properties = armHelper.getResourceProperties(re.modelName);
        for (const propName of Object.keys(properties)) {
            if (!allowedBodyTopLevelProperties.includes(propName.toLowerCase())) {
                yield {
                    location: ["definitions", re.modelName],
                    message: `Top level properties should be one of name, type, id, location, properties, tags, plan, sku, etag, managedBy, identity, zones. Model definition '${re.modelName}' has extra properties ['${propName}'].`,
                };
            }
        }
    }
}
exports.bodyTopLevelProperties = bodyTopLevelProperties;
function* operationsAPIImplementation(openapiSection, options, ctx) {
    const armHelper = new arm_helper_1.ArmHelper(ctx === null || ctx === void 0 ? void 0 : ctx.document, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
    const operationsList = armHelper.getOperationApi();
    if (!operationsList) {
        const resourceProvider = (0, rules_helper_1.getResourceProvider)(ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
        yield {
            location: [],
            message: `Operations API must be implemented for '${resourceProvider}'.`,
        };
    }
}
exports.operationsAPIImplementation = operationsAPIImplementation;
function* resourcesHaveRequiredProperties(openapiSection, options, ctx) {
    var _a;
    const armHelper = new arm_helper_1.ArmHelper(ctx === null || ctx === void 0 ? void 0 : ctx.document, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
    const allResources = armHelper.getAllResources(true, false, true);
    for (const re of allResources) {
        const requiredProperties = ["name", "type", "id"];
        const properties = armHelper.getResourceProperties(re.modelName);
        if (properties["common-types"]) {
            continue;
        }
        for (const propName of requiredProperties) {
            const prop = properties[propName];
            if (!prop || ((_a = armHelper.getAttribute(prop, "readOnly")) === null || _a === void 0 ? void 0 : _a.value) !== true) {
                yield {
                    location: ["definitions", re.modelName],
                    message: `Model definition '${re.modelName}' must have the properties 'name', 'id' and 'type' in its hierarchy and these properties must be marked as readonly.`,
                };
                break;
            }
        }
    }
}
exports.resourcesHaveRequiredProperties = resourcesHaveRequiredProperties;
function* xmsPageableListByRGAndSubscriptions(openapiSection, options, ctx) {
    const armHelper = new arm_helper_1.ArmHelper(ctx === null || ctx === void 0 ? void 0 : ctx.document, ctx === null || ctx === void 0 ? void 0 : ctx.specPath, ctx === null || ctx === void 0 ? void 0 : ctx.inventory);
    const trackedResources = armHelper.getTrackedResources();
    const collectionApiInfos = armHelper.getCollectionApiInfo();
    function isListByRgAndSubscription(apiPaths) {
        return apiPaths.some((p) => armHelper.isPathByResourceGroup(p)) && apiPaths.some((p) => armHelper.isPathBySubscription(p));
    }
    for (const collectionApiInfo of collectionApiInfos) {
        if (isListByRgAndSubscription(collectionApiInfo.collectionGetPath)) {
            const trackedResource = trackedResources.find((r) => r.modelName === collectionApiInfo.childModelName);
            const isXmsPageableResult = collectionApiInfo.collectionGetPath
                .map((p) => armHelper.isPathXmsPageable(p))
                .reduce((result, cur) => (result !== cur ? false : true));
            if (trackedResource && !isXmsPageableResult) {
                yield {
                    location: ["definitions", trackedResource.modelName],
                    message: `For the tracked resource '${trackedResource.modelName}', the x-ms-pageable extension values must be same for list by resource group and subscriptions operations.`,
                };
            }
        }
    }
}
exports.xmsPageableListByRGAndSubscriptions = xmsPageableListByRGAndSubscriptions;
