// @scalar/openapi-types/dist/schemas/extensions/index.js
import { XAdditionalPropertiesNameSchema } from "./x-additional-properties-name.js";
import { XCodeSampleSchema, XCodeSamplesSchema } from "./x-code-samples.js";
import { XEnumDescriptionsSchema } from "./x-enum-descriptions.js";
import { XInternalSchema } from "./x-internal.js";
import { PostResponseSchema, XPostResponseSchema } from "./x-post-response.js";
import { XScalarIconSchema } from "./x-scalar-icon.js";
import { XScalarIgnoreSchema } from "./x-scalar-ignore.js";
import { XScalarRedirectUriSchema } from "./x-scalar-redirect-uri.js";
import { XScalarSdkInstallationSchema } from "./x-scalar-sdk-installation.js";
import { XScalarSecurityQuery } from "./x-scalar-security-query.js";
import { XScalarStabilitySchema, XScalarStabilityValues } from "./x-scalar-stability.js";
import { XTagGroupSchema, XTagGroupsSchema } from "./x-tag-groups.js";
import { XUsePkceValues, XusePkceSchema } from "./x-use-pkce.js";
export {
  PostResponseSchema,
  XAdditionalPropertiesNameSchema,
  XCodeSampleSchema,
  XCodeSamplesSchema,
  XEnumDescriptionsSchema,
  XInternalSchema,
  XPostResponseSchema,
  XScalarIconSchema,
  XScalarIgnoreSchema,
  XScalarRedirectUriSchema,
  XScalarSdkInstallationSchema,
  XScalarSecurityQuery,
  XScalarStabilitySchema,
  XScalarStabilityValues,
  XTagGroupSchema,
  XTagGroupsSchema,
  XUsePkceValues,
  XusePkceSchema
};
