// oav/lib/models/liveResponse.ts
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

/**
 * @class
 * Initializes a new instance of the LiveResponse class.
 * @constructor
 * Properties of the response.
 *
 * @member {string} statusCode The Response status code.
 *
 * @member {object} headers Headers of the response.
 *
 * @member {object} [body] Body of the response.
 *
 * @member {string} [encoding] The encoding of the response body when the body
 * is a buffer.
 *
 */
export class LiveResponse {
  /**
   * Defines the metadata of LiveResponse
   *
   * @returns {object} metadata of LiveResponse
   *
   */
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  public mapper() {
    return {
      required: false,
      serializedName: "LiveResponse",
      type: {
        name: "Composite",
        className: "LiveResponse",
        modelProperties: {
          statusCode: {
            required: true,
            serializedName: "statusCode",
            type: {
              name: "String",
            },
          },
          headers: {
            required: true,
            serializedName: "headers",
            type: {
              name: "Dictionary",
              value: {
                required: false,
                serializedName: "StringElementType",
                type: {
                  name: "String",
                },
              },
            },
          },
          body: {
            required: false,
            serializedName: "body",
            type: {
              name: "Object",
            },
          },
          encoding: {
            required: false,
            serializedName: "encoding",
            type: {
              name: "String",
            },
          },
        },
      },
    };
  }
}
