// oav/lib/models/operationInfo.ts
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

/**
 * @class
 * Initializes a new instance of the OperationInfo class.
 * @constructor
 * Describes the operation against which the live request and response
 * validation happened.
 *
 * @member {string} operationId The id of the operation against which
 * validation happened. This will help find the problematic information in the
 * spec and will be useful in preparing report.
 *
 * @member {string} apiVersion Describes the api-version of the OpenAPI
 * specification. This will help find the OpenAPI spec corresponding to that
 * api-version and will be useful in preparing report.
 *
 */
export class OperationInfo {
  /**
   * Defines the metadata of OperationInfo
   *
   * @returns metadata of OperationInfo
   *
   */
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  public mapper() {
    return {
      required: false,
      serializedName: "OperationInfo",
      type: {
        name: "Composite",
        className: "OperationInfo",
        modelProperties: {
          operationId: {
            required: true,
            serializedName: "operationId",
            type: {
              name: "String",
            },
          },
          apiVersion: {
            required: true,
            serializedName: "apiVersion",
            type: {
              name: "String",
            },
          },
        },
      },
    };
  }
}
