// @azure-tools/typespec-azure-core/dist/src/decorators/operation-link.js
import { useStateMap } from "@typespec/compiler/utils";
import { AzureCoreStateKeys } from "../lib.js";
import { getLroOperationInfo } from "../lro-info.js";
const [
/**
 * Returns the collection of `OperationLinkMetadata` for a given operation, if any, or undefined.
 */
getOperationLinks, setOperationLinks,] = useStateMap(AzureCoreStateKeys.operationLink);
export { getOperationLinks };
export const $operationLink = (context, entity, linkedOperation, linkType, parameters) => {
    if (parameters && parameters.kind !== "Model") {
        return;
    }
    const { program } = context;
    const [operationInfo, diagnostics] = getLroOperationInfo(program, entity, linkedOperation, linkType, parameters);
    if (diagnostics.length > 0) {
        program.reportDiagnostics(diagnostics);
    }
    // An operation may have many operationLinks, so treat them as a collection
    let items = getOperationLinks(program, entity);
    if (items === undefined) {
        items = new Map();
    }
    items.set(linkType, {
        parameters: parameters,
        linkedOperation: linkedOperation,
        linkType: linkType,
        link: operationInfo?.getOperationLink(),
        parameterMap: operationInfo?.getInvocationInfo()?.parameterMap,
        result: operationInfo?.getResultInfo(),
    });
    setOperationLinks(context.program, entity, items);
};
/**
 * Returns the `OperationLinkMetadata` for a given operation and link type, or undefined.
 */
export function getOperationLink(program, entity, linkType) {
    const items = getOperationLinks(program, entity);
    if (items !== undefined) {
        return items.get(linkType);
    }
    return items;
}
