// @azure-tools/typespec-azure-core/dist/src/decorators/polling-location.js
import { createDiagnosticCollector, getEffectiveModelType, ignoreDiagnostics, isNeverType, isUnknownType, isVoidType, } from "@typespec/compiler";
import { $ } from "@typespec/compiler/typekit";
import { useStateMap, useStateSet } from "@typespec/compiler/utils";
import { isMetadata } from "@typespec/http";
import { AzureCoreStateKeys } from "../lib.js";
import { getLroErrorResult } from "./lro-error-result.js";
import { getLroResult } from "./lro-result.js";
import { extractLroStates, findLroStatusProperty, getLongRunningStates, } from "./lro-status.js";
export const [
/**
 * Gets polling information stored with a field that contains a link to an Lro polling endpoint
 * @param program The program to check
 * @param target The ModelProperty to check for polling info
 */
getPollingLocationInfo, setPollingLocationInfo,] = useStateMap(AzureCoreStateKeys.pollingLocationInfo);
export const [isPollingLocation, setPollingLocationParameter] = useStateSet(AzureCoreStateKeys.pollingLocationInfo);
// keys of the pollingOptions type
const optionsKindKey = "kind";
const finalPropertyKey = "finalProperty";
const pollingModelKey = "pollingModel";
const finalResultKey = "finalResult";
export var pollingOptionsKind;
(function (pollingOptionsKind) {
    pollingOptionsKind["StatusMonitor"] = "statusMonitor";
})(pollingOptionsKind || (pollingOptionsKind = {}));
export const $pollingLocation = (context, entity, options) => {
    const { program } = context;
    if (options) {
        if (isNeverType(options))
            return;
        const info = extractPollingLocationInfo(program, entity, options);
        if (info) {
            setPollingLocationInfo(program, entity, info);
        }
    }
    setPollingLocationParameter(program, entity);
};
function extractUnionVariantValue(type) {
    if (type.kind === "UnionVariant" && type.type.kind === "String")
        return type.type.value;
    return undefined;
}
function extractPollingLocationInfo(program, target, options) {
    if (options.kind !== "Model")
        return undefined;
    const kind = options.properties.get(optionsKindKey);
    if (kind === undefined)
        return undefined;
    const kindValue = extractUnionVariantValue(kind.type);
    if (kindValue === undefined)
        return undefined;
    const pollingInfo = { target: target };
    const pollingModel = options.properties.get(pollingModelKey)?.type;
    if (pollingModel && pollingModel.kind === "Model")
        pollingInfo.pollingModel = pollingModel;
    if (pollingModel && isVoidType(pollingModel))
        pollingInfo.pollingModel = $(program).intrinsic.void;
    const finalResult = options.properties.get(finalResultKey)?.type;
    if (finalResult && finalResult.kind === "Model")
        pollingInfo.finalResult = finalResult;
    if (finalResult && isVoidType(finalResult))
        pollingInfo.finalResult = $(program).intrinsic.void;
    switch (kindValue) {
        case pollingOptionsKind.StatusMonitor:
            return extractStatusMonitorLocationInfo(program, options, pollingInfo);
        default:
            return undefined;
    }
}
function extractStatusMonitorLocationInfo(program, options, baseInfo) {
    const kind = options.properties.get(optionsKindKey);
    if (kind === undefined || extractUnionVariantValue(kind.type) !== "statusMonitor")
        return undefined;
    if (baseInfo.pollingModel === undefined || baseInfo.pollingModel.kind === "Intrinsic")
        return undefined;
    const finalProperty = options.properties.get(finalPropertyKey)?.type;
    let finalPropertyValue = undefined;
    if (finalProperty && finalProperty.kind === "ModelProperty")
        finalPropertyValue = finalProperty;
    if (finalProperty &&
        finalProperty.kind === "String" &&
        baseInfo.pollingModel.properties.has(finalProperty.value)) {
        finalPropertyValue = baseInfo.pollingModel.properties.get(finalProperty.value);
    }
    if (finalPropertyValue === undefined)
        finalPropertyValue = ignoreDiagnostics(getLroResult(program, baseInfo.pollingModel, true));
    const statusProperty = findLroStatusProperty(program, baseInfo.pollingModel);
    if (statusProperty === undefined)
        return undefined;
    const statusMonitor = ignoreDiagnostics(extractStatusMonitorInfo(program, baseInfo.pollingModel, statusProperty));
    if (statusMonitor === undefined)
        return undefined;
    statusMonitor.successProperty = finalPropertyValue;
    baseInfo.finalResult =
        finalPropertyValue?.type?.kind === "Model" || finalPropertyValue?.type?.kind === "Intrinsic"
            ? finalPropertyValue.type
            : $(program).intrinsic.void;
    return {
        kind: pollingOptionsKind.StatusMonitor,
        info: statusMonitor,
        ...baseInfo,
    };
}
export function extractStatusMonitorInfo(program, model, statusProperty) {
    const diagnosticsToToss = createDiagnosticCollector();
    const diagnosticsToKeep = createDiagnosticCollector();
    const lroResult = diagnosticsToKeep.pipe(getLroResult(program, model, true));
    const successProperty = lroResult?.kind === "ModelProperty" ? lroResult : undefined;
    const errorProperty = diagnosticsToKeep.pipe(getLroErrorResult(program, model, true));
    const states = getLongRunningStates(program, statusProperty) ??
        diagnosticsToToss.pipe(extractLroStates(program, statusProperty));
    if (!states || !statusProperty)
        return diagnosticsToKeep.wrap(undefined);
    return diagnosticsToKeep.wrap({
        monitorType: getEffectiveModelType(program, model, (p) => !isMetadata(program, p)) ?? model,
        successProperty: successProperty,
        errorProperty: errorProperty,
        statusProperty: statusProperty,
        lroStates: states,
        errorType: errorProperty?.type.kind === "Model" ? errorProperty.type : undefined,
        successType: (successProperty?.type && isUnknownType(successProperty.type)) ||
            successProperty?.type?.kind === "Model"
            ? successProperty.type
            : $(program).intrinsic.void,
        terminationInfo: {
            kind: "model-property",
            property: statusProperty,
            canceledState: states.canceledState,
            failedState: states.failedState,
            succeededState: states.succeededState,
        },
    });
}
