// @scalar/json-magic/dist/bundle/bundle.js
import { path } from "@scalar/helpers/node/path";
import { isObject } from "@scalar/helpers/object/is-object";
import { convertToLocalRef } from "../helpers/convert-to-local-ref.js";
import { getId, getSchemas } from "../helpers/get-schemas.js";
import { getValueByPath } from "../helpers/get-value-by-path.js";
import { escapeJsonPointer } from "../helpers/escape-json-pointer.js";
import { getSegmentsFromPath } from "../helpers/get-segments-from-path.js";
import { isJsonObject } from "../helpers/is-json-object.js";
import { isYaml } from "../helpers/is-yaml.js";
import { getHash, uniqueValueGeneratorFactory } from "./value-generator.js";
const hasRef = (value) => isObject(value) && "$ref" in value && typeof value["$ref"] === "string";
function isRemoteUrl(value) {
  try {
    const url = new URL(value);
    return url.protocol === "http:" || url.protocol === "https:";
  } catch {
    return false;
  }
}
function isFilePath(value) {
  return !isRemoteUrl(value) && !isYaml(value) && !isJsonObject(value);
}
function isLocalRef(value) {
  return value.startsWith("#");
}
function resolveContents(value, plugins) {
  const plugin = plugins.find((p) => p.validate(value));
  if (plugin) {
    return plugin.exec(value);
  }
  return Promise.resolve({
    ok: false
  });
}
function setValueAtPath(obj, path2, value) {
  if (path2 === "") {
    throw new Error("Cannot set value at root ('') pointer");
  }
  const parts = getSegmentsFromPath(path2);
  let current = obj;
  for (let i = 0; i < parts.length; i++) {
    const key = parts[i];
    const isLast = i === parts.length - 1;
    const nextKey = parts[i + 1];
    const shouldBeArray = /^\d+$/.test(nextKey ?? "");
    if (isLast) {
      current[key] = value;
    } else {
      if (!(key in current) || typeof current[key] !== "object") {
        current[key] = shouldBeArray ? [] : {};
      }
      current = current[key];
    }
  }
}
function resolveReferencePath(base, relativePath) {
  if (isRemoteUrl(relativePath)) {
    return relativePath;
  }
  if (isRemoteUrl(base)) {
    const url = new URL(base);
    if (relativePath.startsWith("/")) {
      url.pathname = relativePath;
      return url.toString();
    }
    const mergedPath = path.join(path.dirname(url.pathname), relativePath);
    return new URL(mergedPath, base).toString();
  }
  return path.join(path.dirname(base), relativePath);
}
function prefixInternalRef(input, prefix) {
  if (!isLocalRef(input)) {
    throw "Please provide an internal ref";
  }
  return `#/${prefix.map(escapeJsonPointer).join("/")}${input.substring(1)}`;
}
function prefixInternalRefRecursive(input, prefix) {
  if (Array.isArray(input)) {
    input.forEach((el) => prefixInternalRefRecursive(el, prefix));
    return;
  }
  if (!isObject(input)) {
    return;
  }
  Object.values(input).forEach((el) => prefixInternalRefRecursive(el, prefix));
  if (typeof input === "object" && "$ref" in input && typeof input["$ref"] === "string") {
    const ref = input["$ref"];
    if (!isLocalRef(ref)) {
      return;
    }
    input["$ref"] = prefixInternalRef(ref, prefix);
  }
}
const resolveAndCopyReferences = (targetDocument, sourceDocument, referencePath, externalRefsKey, documentKey, bundleLocalRefs = false, processedNodes = /* @__PURE__ */ new Set()) => {
  const referencedValue = getValueByPath(sourceDocument, getSegmentsFromPath(referencePath)).value;
  if (processedNodes.has(referencedValue)) {
    return;
  }
  processedNodes.add(referencedValue);
  setValueAtPath(targetDocument, referencePath, referencedValue);
  const traverse = (node) => {
    if (!node || typeof node !== "object") {
      return;
    }
    if ("$ref" in node && typeof node["$ref"] === "string") {
      if (node["$ref"].startsWith(`#/${externalRefsKey}/${escapeJsonPointer(documentKey)}`)) {
        resolveAndCopyReferences(
          targetDocument,
          sourceDocument,
          node["$ref"].substring(1),
          externalRefsKey,
          documentKey,
          bundleLocalRefs,
          processedNodes
        );
      } else if (bundleLocalRefs) {
        resolveAndCopyReferences(
          targetDocument,
          sourceDocument,
          node["$ref"].substring(1),
          externalRefsKey,
          documentKey,
          bundleLocalRefs,
          processedNodes
        );
      }
    }
    for (const value of Object.values(node)) {
      traverse(value);
    }
  };
  traverse(referencedValue);
};
const extensions = {
  /**
   * Custom OpenAPI extension key used to store external references.
   * This key will contain all bundled external documents.
   * The x-ext key is used to maintain a clean separation between the main
   * OpenAPI document and its bundled external references.
   */
  externalDocuments: "x-ext",
  /**
   * Custom OpenAPI extension key used to maintain a mapping between
   * hashed keys and their original URLs in x-ext.
   * This mapping is essential for tracking the source of bundled references
   */
  externalDocumentsMappings: "x-ext-urls"
};
async function bundle(input, config) {
  config.externalDocumentsKey = config.externalDocumentsKey ?? extensions.externalDocuments;
  config.externalDocumentsMappingsKey = config.externalDocumentsMappingsKey ?? extensions.externalDocumentsMappings;
  const cache = config.cache ?? /* @__PURE__ */ new Map();
  const loaderPlugins = config.plugins.filter((it) => it.type === "loader");
  const lifecyclePlugin = config.plugins.filter((it) => it.type === "lifecycle");
  const resolveInput = async () => {
    if (typeof input !== "string") {
      return input;
    }
    const result = await resolveContents(input, loaderPlugins);
    if (result.ok && typeof result.data === "object") {
      return result.data;
    }
    throw new Error(
      "Failed to resolve input: Please provide a valid string value or pass a loader to process the input"
    );
  };
  const rawSpecification = await resolveInput();
  const documentRoot = config.root ?? rawSpecification;
  const schemas = getSchemas(documentRoot);
  const isPartialBundling = config.root !== void 0 && config.root !== rawSpecification || config.depth !== void 0;
  const processedNodes = config.visitedNodes ?? /* @__PURE__ */ new Set();
  const defaultOrigin = () => {
    if (config.origin) {
      return config.origin;
    }
    if (typeof input !== "string") {
      return "";
    }
    if (isRemoteUrl(input) || isFilePath(input)) {
      return input;
    }
    return "";
  };
  if (documentRoot[config.externalDocumentsMappingsKey] === void 0) {
    documentRoot[config.externalDocumentsMappingsKey] = {};
  }
  const { generate } = uniqueValueGeneratorFactory(
    config.compress ?? getHash,
    documentRoot[config.externalDocumentsMappingsKey]
  );
  const executeHooks = async (type, ...args) => {
    const hook = config.hooks?.[type];
    if (hook) {
      await hook(...args);
    }
    for (const plugin of lifecyclePlugin) {
      const pluginHook = plugin[type];
      if (pluginHook) {
        await pluginHook(...args);
      }
    }
  };
  const bundler = async (root, origin = defaultOrigin(), isChunkParent = false, depth = 0, currentPath = [], parent = null) => {
    if (config.depth !== void 0 && depth > config.depth) {
      return;
    }
    if (!isObject(root) && !Array.isArray(root)) {
      return;
    }
    if (processedNodes.has(root)) {
      return;
    }
    processedNodes.add(root);
    const context = {
      path: currentPath,
      resolutionCache: cache,
      parentNode: parent,
      rootNode: documentRoot,
      loaders: loaderPlugins
    };
    await executeHooks("onBeforeNodeProcess", root, context);
    const id = getId(root);
    if (hasRef(root)) {
      const ref = root["$ref"];
      const isChunk = "$global" in root && typeof root["$global"] === "boolean" && root["$global"];
      const localRef = convertToLocalRef(ref, id ?? origin, schemas);
      if (localRef !== void 0) {
        if (isPartialBundling) {
          const segments = getSegmentsFromPath(`/${localRef}`);
          const parent2 = segments.length > 0 ? getValueByPath(documentRoot, segments.slice(0, -1)).value : void 0;
          const targetValue = getValueByPath(documentRoot, segments);
          await bundler(targetValue.value, targetValue.context, isChunkParent, depth + 1, segments, parent2);
        }
        await executeHooks("onAfterNodeProcess", root, context);
        return;
      }
      const [prefix, path2 = ""] = ref.split("#", 2);
      const resolvedPath = resolveReferencePath(id ?? origin, prefix);
      const compressedPath = await generate(resolvedPath);
      const seen = cache.has(resolvedPath);
      if (!seen) {
        cache.set(resolvedPath, resolveContents(resolvedPath, loaderPlugins));
      }
      await executeHooks("onResolveStart", root);
      const result = await cache.get(resolvedPath);
      if (result.ok) {
        if (!seen) {
          if (!isChunk) {
            prefixInternalRefRecursive(result.data, [extensions.externalDocuments, compressedPath]);
          }
          await bundler(result.data, isChunk ? origin : resolvedPath, isChunk, depth + 1, [
            config.externalDocumentsKey,
            compressedPath,
            documentRoot[config.externalDocumentsMappingsKey]
          ]);
          setValueAtPath(
            documentRoot,
            `/${config.externalDocumentsMappingsKey}/${escapeJsonPointer(compressedPath)}`,
            resolvedPath
          );
        }
        if (config.treeShake === true) {
          resolveAndCopyReferences(
            documentRoot,
            { [config.externalDocumentsKey]: { [compressedPath]: result.data } },
            prefixInternalRef(`#${path2}`, [config.externalDocumentsKey, compressedPath]).substring(1),
            config.externalDocumentsKey,
            compressedPath
          );
        } else if (!seen) {
          setValueAtPath(documentRoot, `/${config.externalDocumentsKey}/${compressedPath}`, result.data);
        }
        root.$ref = prefixInternalRef(`#${path2}`, [config.externalDocumentsKey, compressedPath]);
        await executeHooks("onResolveSuccess", root);
        await executeHooks("onAfterNodeProcess", root, context);
        return;
      }
      await executeHooks("onResolveError", root);
      await executeHooks("onAfterNodeProcess", root, context);
      return console.warn(
        `Failed to resolve external reference "${resolvedPath}". The reference may be invalid, inaccessible, or missing a loader for this type of reference.`
      );
    }
    await Promise.all(
      Object.entries(root).map(async ([key, value]) => {
        if (key === config.externalDocumentsKey || key === config.externalDocumentsMappingsKey) {
          return;
        }
        await bundler(value, id ?? origin, isChunkParent, depth + 1, [...currentPath, key], root);
      })
    );
    await executeHooks("onAfterNodeProcess", root, context);
  };
  await bundler(rawSpecification);
  if (!config.urlMap && !isPartialBundling) {
    delete documentRoot[config.externalDocumentsMappingsKey];
  }
  return rawSpecification;
}
export {
  bundle,
  extensions,
  isFilePath,
  isLocalRef,
  isRemoteUrl,
  prefixInternalRef,
  prefixInternalRefRecursive,
  resolveAndCopyReferences,
  setValueAtPath
};
