// oav/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Constants = exports.TrafficValidator = exports.SemanticValidator = exports.SwaggerSemanticValidator = exports.NewModelValidator = exports.SwaggerExampleValidator = exports.JsonLoaderRefError = exports.JsonLoader = exports.SuppressionLoader = exports.SwaggerLoader = exports.FileLoader = exports.parseValidationRequest = exports.LiveValidatorLoggingTypes = exports.LiveValidator = exports.getProvider = exports.getResourceType = exports.errorCodeToErrorMetadata = exports.ValidationResultSource = exports.validateTrafficAgainstSpec = exports.validateExamples = exports.validateSpec = void 0;
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
require("reflect-metadata");
const C = __importStar(require("./lib/util/constants"));
// Easy to use methods from validate.ts
var validate_1 = require("./lib/validate");
Object.defineProperty(exports, "validateSpec", { enumerable: true, get: function () { return validate_1.validateSpec; } });
Object.defineProperty(exports, "validateExamples", { enumerable: true, get: function () { return validate_1.validateExamples; } });
Object.defineProperty(exports, "validateTrafficAgainstSpec", { enumerable: true, get: function () { return validate_1.validateTraffic; } });
var validationResultSource_1 = require("./lib/util/validationResultSource");
Object.defineProperty(exports, "ValidationResultSource", { enumerable: true, get: function () { return validationResultSource_1.ValidationResultSource; } });
var validationError_1 = require("./lib/util/validationError");
Object.defineProperty(exports, "errorCodeToErrorMetadata", { enumerable: true, get: function () { return validationError_1.errorCodeToErrorMetadata; } });
var utils_1 = require("./lib/util/utils");
Object.defineProperty(exports, "getResourceType", { enumerable: true, get: function () { return utils_1.getResourceType; } });
Object.defineProperty(exports, "getProvider", { enumerable: true, get: function () { return utils_1.getProvider; } });
// Classes
var liveValidator_1 = require("./lib/liveValidation/liveValidator");
Object.defineProperty(exports, "LiveValidator", { enumerable: true, get: function () { return liveValidator_1.LiveValidator; } });
Object.defineProperty(exports, "LiveValidatorLoggingTypes", { enumerable: true, get: function () { return liveValidator_1.LiveValidatorLoggingTypes; } });
Object.defineProperty(exports, "parseValidationRequest", { enumerable: true, get: function () { return liveValidator_1.legacyParseValidationRequest; } });
var fileLoader_1 = require("./lib/swagger/fileLoader");
Object.defineProperty(exports, "FileLoader", { enumerable: true, get: function () { return fileLoader_1.FileLoader; } });
var swaggerLoader_1 = require("./lib/swagger/swaggerLoader");
Object.defineProperty(exports, "SwaggerLoader", { enumerable: true, get: function () { return swaggerLoader_1.SwaggerLoader; } });
var suppressionLoader_1 = require("./lib/swagger/suppressionLoader");
Object.defineProperty(exports, "SuppressionLoader", { enumerable: true, get: function () { return suppressionLoader_1.SuppressionLoader; } });
var jsonLoader_1 = require("./lib/swagger/jsonLoader");
Object.defineProperty(exports, "JsonLoader", { enumerable: true, get: function () { return jsonLoader_1.JsonLoader; } });
Object.defineProperty(exports, "JsonLoaderRefError", { enumerable: true, get: function () { return jsonLoader_1.JsonLoaderRefError; } });
var modelValidator_1 = require("./lib/swaggerValidator/modelValidator");
Object.defineProperty(exports, "SwaggerExampleValidator", { enumerable: true, get: function () { return modelValidator_1.SwaggerExampleValidator; } });
Object.defineProperty(exports, "NewModelValidator", { enumerable: true, get: function () { return modelValidator_1.NewModelValidator; } });
var semanticValidator_1 = require("./lib/swaggerValidator/semanticValidator");
Object.defineProperty(exports, "SwaggerSemanticValidator", { enumerable: true, get: function () { return semanticValidator_1.SwaggerSemanticValidator; } });
Object.defineProperty(exports, "SemanticValidator", { enumerable: true, get: function () { return semanticValidator_1.SemanticValidator; } });
var trafficValidator_1 = require("./lib/swaggerValidator/trafficValidator");
Object.defineProperty(exports, "TrafficValidator", { enumerable: true, get: function () { return trafficValidator_1.TrafficValidator; } });
// Constants
exports.Constants = C;
